---
title: "TE landscape Draft"
author: "Erica Pehrsson"
date: "`r Sys.Date()`"
output: word_document
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(dev = 'pdf',dev.args=list(pdf = list(useDingbats = FALSE)),fig.width=7, fig.height=8, fig.path='TE_landscape_draft_figures/', cache.path="TE_landscape_draft_cache/")
```

# Setup

## Load required libraries

```{r load libraries}
library(plyr)
packageVersion("plyr")
library(reshape2)
packageVersion("reshape2")
library(ggplot2)
packageVersion("ggplot2")
library(NMF)
packageVersion("NMF")
library(gridExtra)
packageVersion("gridExtra")
library(RColorBrewer)
packageVersion("RColorBrewer")
library(scales)
packageVersion("scales")
library(mgcv)
packageVersion("mgcv")
library(extrafont)
packageVersion("extrafont")
library(rcompanion)
packageVersion("rcompanion")
```

## Set universal plot parameters

```{r set plot parameters}
theme_set(theme_bw(base_size=10) %+replace%  theme(panel.background=element_rect(fill="white"),panel.border=element_rect(color="black",fill=NA),panel.grid=element_blank()))
```

## Load functions

```{r load functions}
source("R_scripts/functions.R")
```

## Define color and label vectors

```{bash state lists, eval=FALSE}
# List of 15 chromHMM states
chromHMM_states.txt

# List of 4 methylation states
methylation_states.txt
```

```{r define vectors}
# TE coordinates
TE_coordinates = c("chromosome","start","stop","subfamily","family","class","strand")
hg19_coordinates = c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")
mm10_coordinates = c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10")

# Chromosomes
standard_chromosomes = c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY","chrM")

# chromHMM
chromHMM_states = c("1_TssA","2_TssAFlnk","3_TxFlnk","4_Tx","5_TxWk","6_EnhG","7_Enh","8_ZNF/Rpts","9_Het","10_TssBiv","11_BivFlnk","12_EnhBiv","13_ReprPC","14_ReprPCWk","15_Quies")
chromHMM_labels = setNames(c("Active TSS","Flanking active TSS","Transcription at gene 5'/3'","Strong transcription","Weak transcription","Genic enhancer","Enhancer","ZNF genes/repeats","Heterochromatin","Poised TSS","Flanking poised TSS/enhancer","Poised enhancer","Repressed Polycomb","Weak repressed Polycomb","Quiescent"),chromHMM_states)
chromHMM_colors = setNames(c(rgb(255,0,0,maxColorValue=255),rgb(255,69,0,maxColorValue=255),rgb(50,205,50,maxColorValue=255),rgb(0,128,0,maxColorValue=255),rgb(0,100,0,maxColorValue=255),rgb(194,225,5,maxColorValue=255),rgb(255,255,0,maxColorValue=255),rgb(102,205,170,maxColorValue=255),rgb(138,145,208,maxColorValue=255),rgb(205,92,92,maxColorValue=255),rgb(233,150,122,maxColorValue=255),rgb(189,183,107,maxColorValue=255),rgb(128,128,128,maxColorValue=255),rgb(192,192,192,maxColorValue=255),"grey90"),chromHMM_states)

chromHMM_states_18 = setNames(c(rgb(255,0,0,maxColorValue=255),rgb(255,69,0,maxColorValue=255),rgb(255,69,0,maxColorValue=255),
                            rgb(255,69,0,maxColorValue=255),rgb(0,128,0,maxColorValue=255),rgb(0,100,0,maxColorValue=255),
                            rgb(194,225,5,maxColorValue=255),rgb(194,225,5,maxColorValue=255),rgb(255,195,77,maxColorValue=255),
                            rgb(255,195,77,maxColorValue=255),rgb(255,255,0,maxColorValue=255),rgb(102,205,170,maxColorValue=255),
                            rgb(138,145,208,maxColorValue=255),rgb(205,92,92,maxColorValue=255),rgb(189,183,107,maxColorValue=255),
                            rgb(128,128,128,maxColorValue=255),rgb(192,192,192,maxColorValue=255),rgb(255,255,255,maxColorValue=255)),
                          c("1_TssA","2_TssFlnk","3_TssFlnkU","4_TssFlnkD","5_Tx","6_TxWk","7_EnhG1","8_EnhG2","9_EnhA1","10_EnhA2","11_EnhWk",
                            "12_ZNF/Rpts","13_Het","14_TssBiv","15_EnhBiv","16_ReprPC","17_ReprPCWk","18_Quies"))

imputed_states = setNames(c(rgb(255,0,0,maxColorValue=255),rgb(255,69,0,maxColorValue=255),rgb(255,69,0,maxColorValue=255),
                            rgb(255,69,0,maxColorValue=255),rgb(0,128,0,maxColorValue=255),rgb(0,128,0,maxColorValue=255),rgb(0,128,0,maxColorValue=255),
                            rgb(0,150,0,maxColorValue=255),rgb(194,225,5,maxColorValue=255),rgb(194,225,5,maxColorValue=255),rgb(194,225,5,maxColorValue=255),
                            rgb(194,225,5,maxColorValue=255),rgb(255,195,77,maxColorValue=255),rgb(255,195,77,maxColorValue=255),rgb(255,195,77,maxColorValue=255),
                            rgb(255,255,0,maxColorValue=255),rgb(255,255,0,maxColorValue=255),rgb(255,255,0,maxColorValue=255),rgb(255,255,102,maxColorValue=255),
                            rgb(102,205,170,maxColorValue=255),rgb(138,145,208,maxColorValue=255),rgb(230,184,183,maxColorValue=255),
                            rgb(112,48,160,maxColorValue=255),rgb(128,128,128,maxColorValue=255),rgb(255,255,255,maxColorValue=255)),
                          c("1_TssA","2_PromU","3_PromD1","4_PromD2","5_Tx5'","6_Tx","7_Tx3'","8_TxWk",
                            "9_TxReg","10_TxEnh5'","11_TxEnh3'","12_TxEnhW","13_EnhA1","14_EnhA2","15_EnhAF","16_EnhW1",
                            "17_EnhW2","18_EnhAc","19_DNase","20_ZNF/Rpts","21_Het","22_PromP","23_PromBiv","24_ReprPC","25_Quies"))

# WGBS
meth_states = c("Hypomethylated","Intermediate","Hypermethylated","Missing")
meth_labels = setNames(c("Hypomethylated","Intermediate meth","Hypermethylated","Missing meth data"),meth_states)
meth_colors = setNames(c("#F8766D","#00BFC4","#7CAE00","#C77CFF"),meth_states)

# All states
states = c(chromHMM_states,meth_states,"DNase","H3K27ac","Expressed_samples")
all_state_colors = c(chromHMM_colors,meth_colors,"aquamarine","tan1","black","black","black") 
names(all_state_colors)[20:24] = c("DNase","H3K27ac","Expressed_samples","Bases","CpGs")
all_state_labels = setNames(c(chromHMM_states,meth_labels,"DHS","H3K27ac","RPKM > 1"),states)
all_pc_states = c(states[1:21],unlist(lapply(states[1:21],function(x) paste(x,"PC",sep="_"))))

# Metric labels
metric_labels = c("chromHMM","WGBS","DNase","H3K27ac","RPKM > 1")
names(metric_labels) = c("chromHMM","WGBS","DNase","H3K27ac","Expressed_samples")

# TE classification
class_colors = setNames(c("#4A72E8","#FF6600","#006600","#cc0000","lightseagreen","#5C5C5C"),c("DNA","LINE","LTR","SINE","SVA","Other"))

# Sample classification
sample_categories = c("Group","Anatomy","Type","Age","Cancer","Germline")
category_labels = setNames(c("Group","Anatomy","Type","Age","Cancer","Germlayer"),sample_categories)

group_colors = setNames(c("#924965","#4178AE","#E41A1C","#69608A","#B65C73","#FF9D0C","#678C69","#55A354","#E67326","#FFD924","#AF5B39","#D56F80","#999999","#C5912B","#C58DAA","#F182BC","#C2655D","#DAB92E","#000000"),c("ESC","ES-deriv","IMR90","iPSC","Mesench","Epithelial","HSC & B-cell","Blood & T-cell","Myosat","Neurosph","Adipose","Heart","Other","Brain","Digestive","Sm. Muscle","Muscle","Thymus","ENCODE2012"))
anatomy_colors = setNames(c("#B2DF8A","#33A02C","#FB9A99","#FFED6F","#80B1D3","#666666","#924965","#4178AE","#FDBF6F","#CAB2D6","#6A3D9A","#BEBADA","#FCCDE5","#BC80BD","#E7298A","#FF7F00","#69608A","#B3DE69","#CCEBC5","#000000","#FB8072","#FDB462","#8DD3C7","#D9D9D9","#E31A1C","#A6CEE3","#D95F02","#E6AB02","#66A61E","#B15928","#E31A1C","#8DD3C7","#E7298A","#A6761D","#666666"),c("ADRENAL","BLOOD","BONE","BRAIN","BREAST","CERVIX","ESC","ESC_DERIVED","FAT","GI_COLON","GI_DUODENUM","GI_ESOPHAGUS","GI_INTESTINE","GI_RECTUM","GI_STOMACH","HEART","IPSC","KIDNEY","LIVER","LUNG","MUSCLE","MUSCLE_LEG","OVARY","PANCREAS","PLACENTA","SKIN","SPLEEN","STROMAL_CONNECTIVE","THYMUS","VASCULAR",NA,NA,NA,NA,NA))
type_colors = setNames(brewer.pal(5,"Greens"),c("PrimaryCulture","PrimaryCell","PrimaryTissue","CellLine","ESCDerived"))
age_colors = setNames(brewer.pal(4,"YlOrRd"),c("Cell_line","Unknown","Non_fetal","Fetal"))
cancer_colors = setNames(c("lightgrey","red"),c("No","Yes"))
germline_colors = setNames(brewer.pal(6,"Dark2"),c("Pluripotent","Mesoderm","Ectoderm","Endoderm","Mixed","Other"))

group_order = c("ESC","iPSC","ES-deriv","Blood & T-cell","HSC & B-cell","Thymus","Brain","Neurosph","Digestive","Epithelial","Adipose","Mesench","Myosat","Muscle","Sm. Muscle","Heart","Other","IMR90","ENCODE2012")

age_labels = setNames(c("Cell line","Fetal","Non-fetal","Unknown"),c("Cell_line","Fetal","Non_fetal","Unknown"))

# Genic features
features = c("cpgIsland","promoters","5UTR","CDS","3UTR","exons","introns","intergenic")
genic_labels = setNames(c("CpG island","Promoter","5'UTR","CDS","3'UTR","Exon","Intron","Intergenic","CDS","Vista enhancers"),c(features,"coding_exon","Vista_enhancers"))

cohorts = c("cpgIsland","promoters","promoters_pc","promoters_nc","5UTR","5UTR_pc","5UTR_nc","coding_exon","coding_exon_pc","3UTR","3UTR_pc","3UTR_nc",
                     "exons","exons_pc","exons_nc","introns","introns_pc","introns_nc","intergenic","blacklist","Vista_enhancers")

# Feature labels
coding_labels = c("All","Protein-coding","Non-coding")
names(coding_labels) = c("All","pc","nc")

# Colnames vectors
measure_states_extra = c("States.chromHMM","Max_states_intra","States.WGBS","Max_expression")
measure_metrics = c("Length","mappability","JC_distance","CpGs","CpGs_per_length")
measure_metrics_subfam = c("Count","Total_length","Length","Mappability","Age","CpGs","Count_CpGs","CpGs_per_TE","CpGs_per_kbp")

enrichment_names = c("Length_ijk","Length_ik","Length_jk","Length_k","Enrichment","Length_percent_jk","Members","Count","Percent")

# Metric labels
variable_labels = setNames(c("Length (bp)","Mappability","Jukes-Cantor evolutionary distance","CpGs","CpGs/bp"),measure_metrics)
measure_labels = setNames(c("Number of TEs","Total length (bp)","Median length (bp)","Median mappability","Median age (JC)","Total CpGs","TEs with CpGs","CpGs/TE","CpGs/kbp"),measure_metrics_subfam)

# Mouse-human chromHMM states
mouse_chromHMM_states = c("TssA","TssAFlnk1","TssAFlnk2","Tx1","Tx2","Enh","EnhLo1","EnhLo2","HetCons","HetFac","TssBiv","EnhPois1","EnhPois2","QuiesG","Quies")
mouse_chromHMM_colors = setNames(c("red","orangered2","orangered4","green4","darkgreen","yellow","gold1","goldenrod2","cornflowerblue","steelblue2","indianred3","burlywood1","lightgoldenrod2","grey","lightgrey"),mouse_chromHMM_states)

overlap_labels = c("% genome","% TEs (either strand)","% TEs (same strand)","% feature in TEs")
names(overlap_labels) = c("Genome_percent_genome","TEs_percent_TE","TEs_strand_percent_TE","Genome_percent_TE")
```

## Load metadata

```{bash sample lists, eval=FALSE}
# Samples with chromHMM annotations (127, 121 without cancer cell lines and IMR90)
mnemonics.txt
mnemonics_noCancer.txt

# Samples with WGBS data (37)
WGBS_samples.txt

# Samples with DNase data (53)
DNase_samples.txt

# Samples with H3K27ac data (98)
H3K27ac_samples.txt

# Samples with RNA data, strand-agnostic (56)
RNA_samples_agnostic.txt
```

```{r metadata}
source("R_scripts/metadata.R")
```

## Load sample statistics

```{r load sample stats}
# Samples with each metric: all, not excluded, no Y, both
sample_counts = rbind(apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x)),])[1]),
                      apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x) & metadata$Exclude == "Include"),])[1]),
                      apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x) & metadata$chrY == "Yes"),])[1]),
                      apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x) & metadata$chrY == "Yes" & metadata$Exclude == "Include"),])[1]))
colnames(sample_counts)[1] = "chromHMM"
rownames(sample_counts) = c("All","Include","chrY","Include.chrY")

mark_labels = c(paste("chromHMM (n=",sample_counts["All","chromHMM"],")",sep=""),
                paste("WGBS (n=",sample_counts["All","WGBS"],")",sep=""),
                paste("DHS (n=",sample_counts["All","DNase"],")",sep=""),
                paste("H3K27ac (n=",sample_counts["All","H3K27ac"],")",sep=""))
names(mark_labels) = c("chromHMM","WGBS","DNase","H3K27ac")

mark_labels_simple = setNames(c("chromHMM","WGBS","DHS","H3K27ac","RPKM > 1"),c("chromHMM","WGBS","DNase","H3K27ac","RNA"))
```

## Create essential matrices

```{bash TEs, eval=FALSE}
# RepeatMasker file for hg19	 
#TE_landscape/raw_data/rmsk.txt.gz -> /bar/genomes/hg19/rmsk/rmsk.txt.gz	
#TE_landscape/features/rmsk.txt	

# RepeatMasker file restricted to all TE classes, chr 1-22, X, Y, M	 

#TE_landscape/features/TEs/rmsk_TE.txt	
awk -v OFS='\t' '{if(($12 == "LTR" || $12 == "DNA" || $12 == "SINE" || $12 == "LINE") && ($6 !~ /_/)) print $6, $7, $8, $11, $12, $13, $10}' rmsk.txt > rmsk_TE.txt

#TE_landscape/features/TEs/rmsk_other.txt	
awk -v OFS='\t' '{if(($12=="Unknown"||$12=="Unknown?"||$12=="DNA?"||$12=="LINE?"||$12=="SINE?"||$12=="LTR?"||$12=="Other"||$12=="RC") && ($6 !~ /_/))print $6, $7, $8, $11, $12, $13, $10}' rmsk.txt > rmsk_other.txt

#TE_landscape/features/TEs/rmsk_TEother.txt	
cat rmsk_TE.txt rmsk_other.txt > rmsk_TEother.txt
```

```{bash TE mappability, eval=FALSE}
# Rsmk 36bp mappability file (from Sundaram et al), standard chromosomes
#TE_landscape/mappability/rmsk.txt.mapability.36mer -> /bar/genomes/hg19/rmsk/rmsk.txt.mapability.36mer

# TE 36bp mappability file	 

#TE_landscape/mappability/rmsk_TE_mappability_36mer.txt	
while read line; do awk -v subfam=$line '{if($4 == subfam) print $0}' rmsk.txt.mapability.36mer >> rmsk_TE_mappability_36mer.txt; done < subfamilies.txt

#TE_landscape/mappability/rmsk_other_mappability_36mer.txt	
while read line; do awk -v subfam=$line '{if($4 == subfam) print $0}' rmsk.txt.mapability.36mer >> rmsk_other_mappability_36mer.txt; done < other_subfamilies.txt
```

```{bash TE age, eval=FALSE}
# JC evolutionary distance for rmsk repeats (takes zipped rmsk file as input)
#TE_landscape/age/rmsk.txt.gz.JCage	
# Python script takes *rmsk.txt.gz as input
python calcAge_generalized.py rmsk.txt.gz.JCage

# JC evolutionary matrix for TEs, standard chromosomes	 

#TE_landscape/age/rmsk_TE_JCage.txt	
awk -v OFS='\t' '{if(($6 == "LTR" || $6 == "DNA" || $6 == "SINE" || $6 == "LINE") && ($1 !~ /_/)) print $1, $2, $3, $4, $5, $6, $7, $8, $9}' rmsk.txt.gz.JCage > rmsk_TE_JCage.txt

#TE_landscape/age/rmsk_other_JCage.txt	
awk -v OFS='\t' '{if(($6 == "LTR?" || $6 == "DNA?" || $6 == "SINE?" || $6 == "LINE?" || $6 == "Unknown?" || $6 == "Unknown" || $6 == "Other" || $6 == "RC") && ($1 !~ /_/)) print $1, $2, $3, $4, $5, $6, $7, $8, $9}' rmsk.txt.gz.JCage > rmsk_other_JCage.txt
```

CpGs per TE loaded from ''TE_CpG_count.txt'', generated in "TE CpG count" below.

```{bash download RefSeq features, eval=FALSE}
# hg19 RefSeq features from UCSC Table Browser
#genic_features/refseq_3UTR.txt
#genic_features/refseq_5UTR.txt
#genic_features/refseq_coding_exon.txt
#genic_features/refseq_exons.txt
#genic_features/refseq_introns.txt
#genic_features/refseq_genes.txt

# Regions 2000bp upstream of RefSeq gene TSS
#genic_features/refseq_up2000.txt

# hg19 CpG islands from UCSC Table Browser
#genic_features/cpgIslandExtUnmasked.txt

# File of chromosome sizes from UCSC	 
#TE_landscape/raw_data/hg19.genome	
mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -e "select chrom, size from hg19.chromInfo"  > hg19.genome
```

```{bash process RefSeq features, eval=FALSE}
# Expand 500bp downstream of RefSeq gene TSS	 
#genic_features/refseq_promoters.txt	
bedtools slop -i refseq_up2000.txt -g hg19.genome -l 0 -r 500 -s > refseq_promoters.txt

# Intergenic regions	 
#genic_features/refseq_intergenic.txt	
bedtools complement -i refseq_genes.txt -g hg19.genome.standard > refseq_intergenic.txt

# File of chromosome sizes from UCSC, filtered to chr 1-22, X, Y	 
#TE_landscape/features/hg19_standard.genome	
head -n25 hg19.genome > hg19_standard.genome
```

```{bash merge RefSeq features, eval=FALSE}
# Merged features
#genic_features/refseq_3UTR_merge.txt
#genic_features/refseq_5UTR_merge.txt
#genic_features/refseq_coding_exon_merge.txt
#genic_features/refseq_exons_merge.txt
#genic_features/refseq_introns_merge.txt
for file in refseq_*.txt ; do sort -k1,1 -k2,2n $file | bedtools merge -i - > $file\_merge; done
rename 's/.txt_merge/_merge.txt/' *.txt_merge

#genic_features/refseq_promoters_merge.txt
bedtools slop -i refseq_up2000.txt -g hg19.genome -l 0 -r 500 -s | sort -k1,1 -k2,2n - | bedtools merge -i - > refseq_promoters_merge.txt

#genic_features/cpgIslandExtUnmasked_merge.txt
sort -k1,1 -k2,2n cpgIslandExtUnmasked.txt | bedtools merge -i - > cpgIslandExtUnmasked_merge.txt

# Merged features, coding/non-coding

#genic_features/refseq_[feature]_pc_merge.txt [6 files]
for file in refseq*.txt; do awk -v OFS='\t' '{if(substr($4,0,2) == "NM") print $1, $2, $3}' $file | sort -k1,1V -k2,2n - | bedtools merge -i - > $(basename "$file" .txt)\_pc_merge.txt; done

#genic_features/refseq_[feature]_nc_merge.txt [5 files]
for file in refseq*.txt; do awk -v OFS='\t' '{if(substr($4,0,2) == "NR") print $1, $2, $3}' $file | sort -k1,1V -k2,2n - | bedtools merge -i - > $(basename "$file" .txt)\_nc_merge.txt; done
```

```{bash TE RefSeq, eval=FALSE}
# Intersection between TEs, RefSeq features	 
#TE_landscape/features/intersect_features/rmsk_TEother_refseq_[feature].txt [7 files]	
for file in ~/genic_features/refseq*merge*txt; do bedtools intersect -wo -a rmsk_TEother.txt -b $file | awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=$11}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' - > rmsk_TEother_$(basename "$file" .txt)\.txt; done

bedtools intersect -wo -a rmsk_TEother.txt -b ~/genic_features/refseq_intergenic.txt | awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=$11}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' - > rmsk_TEother_refseq_intergenic.txt

# Intersection between TEs, Refseq features (coding/non-coding)	 
#TE_landscape/features/intersect_features/rmsk_TEother_refseq_[feature]_pc_merge.txt [6 files]	
#TE_landscape/features/intersect_features/rmsk_TEother_refseq_[feature]_nc_merge.txt [5 files]		
for file in ~/genic_features/RefSeq/refseq_*c_merge.txt; do bedtools intersect -wo -a ../TEs/rmsk_TEother.txt -b $file | awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=$11}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' - > rmsk_TEother_$(basename "$file" .txt)\.txt; done

# Intersection between TEs, CpG islands	 
#TE_landscape/features/intersect_features/rmsk_TEother_cpgIsland.txt	
bedtools intersect -wo -a rmsk_TEother.txt -b ~/genic_features/cpgIslandExtUnmasked_merge.txt | awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=$11}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' - > rmsk_TEother_cpgIsland.txt
```

```{bash TE blacklist, eval=FALSE}
# TEs that overlap the hg19 blacklist	 

#TE_landscape/features/intersect_features/TE_blacklist.bed	
bedtools intersect -wao -a rmsk_TE.txt -b ../../genomes/hg19/blacklist.bed > TE_blacklist.bed

#TE_landscape/features/intersect_features/other_blacklist.bed	
bedtools intersect -wao -a rmsk_other.txt -b ../../genomes/hg19/blacklist.bed > other_blacklist.bed
```

There are 1,790 human VISTA enhancers, 934 of which positively validated.

```{bash TE VISTA, eval=FALSE}
# VISTA enhancers	
#TE_landscape/features/vista_enhancers/vista_enhancers.txt	
https://enhancer.lbl.gov/cgi-bin/imagedb3.pl?form=search&show=1&search.form=no&search.result=yes

# Human VISTA enhancer headers
#TE_landscape/features/vista_enhancers/vista_enhancers_human.txt	

grep '^>Human' features/vista_enhancers/vista_enhancers.txt

# All human VISTA enhancers in bed format
awk -v FS=':|-|\t' -v OFS='\t' '{print $1, $2, $3, $4, $5}' features/vista_enhancers/vista_enhancers_human.txt > features/vista_enhancers/vista_enhancers_human_all.bed

# Human, positively validated VISTA enhancers in bed format
#TE_landscape/features/vista_enhancers/vista_enhancers_human.bed	

# Human, positively validated VISTA enhancers intersected with TEs	 
#TE_landscape/features/intersect_features/vista_enhancers_rmsk_TE.txt	
bedtools intersect -wo -a vista_enhancers_human.bed -b rmsk_TE.txt > vista_enhancers_rmsk_TE.txt

#TE_landscape/features/intersect_features/vista_enhancers_rmsk_other.txt	
bedtools intersect -wo -a vista_enhancers_human.bed -b rmsk_other.txt > vista_enhancers_rmsk_other.txt

cat features/intersect_features/vista_enhancers_rmsk_*.txt | awk -v OFS='\t' '{a[$4,$5,$6,$7,$8,$9,$10]+=1}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' -  > features/intersect_features/vista_enhancers_TE.txt
```

```{r create TEs, eval=FALSE}
source("R_scripts/TE_matrix.R")
```

```{bash get epigenetic data, eval=FALSE}
# Roadmap

# chromHMM annotations (15-state)
#TE_landscape/raw_data/chromHMM/all.mnemonics.bedFiles.tgz
#TE_landscape/raw_data/chromHMM/E#_15_coreMarks_mnemonics.bed [127 files]
http://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/all.mnemonics.bedFiles.tgz

# WGBS
# By-chromosome WGBS fractional methylation matrices of CpG x sample 	
#/bar/mchoudhary/2chromTE/Meth/FractionalMethylation_Removed_E027_E064_Fixed_E012/

# DNase narrow peak bedfiles
#TE_landscape/raw_data/DNase/DNase_narrow_peaks/E#-DNase.macs2.narrowPeak [53 files]
# DNase_peaks.txt is a list of all 53 files
while read line; do wget http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/narrowPeak/$line\.gz; gunzip $line\.gz; done < DNase_peaks.txt

# H3K27ac narrow peak bedfiles
#TE_landscape/raw_data/H3K27ac/H3K27ac_narrow_peaks/E#-H3K27ac.narrowPeak [98 files]
while read line; do wget http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/narrowPeak/$line\-H3K27ac.narrowPeak.gz; done < H3K27ac_samples.txt

# RNA-seq normalization factors
#TE_landscape/raw_data/RNAseq/all.EGID.N.readlength

# 18-state chromHMM bedfiles
wget https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/core_K27ac/jointModel/final/all.mnemonics.bedFiles.tgz

# Imputed 25-state chromHMM bedfiles
while read line; do wget https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/imputed12marks/jointModel/final/$line\_25_imputed12marks_mnemonics.bed.gz; done < ../../sample_lists/mnemonics.txt

# Mouse (ENCODE)

# chromHMM (mm10)
#TE_landscape/bash_scripts/get_mouse_chromHMM.sh
#TE_landscape/raw_data/mouse/chromHMM/ENCFF#.bed.gz [12 files]

# WGBS (mm10)
#TE_landscape/bash_scripts/get_meth.txt
#TE_landscape/raw_data/mouse/WGBS/ENCFF#.bed [9 files]
```

There are 15,478,399 windows. All of the intersections are 200bp. 

The total width of the windows (3095677412) is slightly larger than what is covered by chromHMM (3095675000; no chrM), but my earlier analyses suggest that the missing 2,412 bp is at the end of the chromosomes and does not overlap any TEs (both windows and chromHMM start at base 0 for all chromosomes). Confirming this, only 24 windows do not have chromHMM annotations (one per chromosome). 

```{bash 200bp windows, eval=FALSE}
# Divided up the genome into 200bp windows
bedtools makewindows -g features/hg19_standard.genome -w 200 > features/hg19_standard.windows

# Intersected with chromHMM
while read line; do bedtools intersect -wo -b features/hg19_standard.windows -a raw_data/chromHMM/$line\_15_coreMarks_mnemonics.bed > chromHMM/genome/windows/windows_$line\.bed; done < sample_lists/mnemonics.txt

# Reformmated to bedfiles and sorted
for file in chromHMM/genome/windows/windows_E*.bed; do awk -v OFS='\t' '{print $5, $6, $7, $4}' $file | sort -k1,1 -k2,2n - > chromHMM/genome/windows/$( basename $file .bed ); done

# QC 
## Total width of windows
awk '{sum+=$3-$2}END{print sum}' features/hg19_standard.windows

## Total width of chromHMM
awk '{if($1 != "chrM") sum+=$3-$2}END{print sum}' raw_data/chromHMM/E001_15_coreMarks_mnemonics.bed
```

```{bash split TEs summit, eval=FALSE}
# Intersect TEs with 200bp windows
bedtools intersect -wo -a features/TEs/rmsk_TEother.txt -b features/hg19_standard.windows > features/TEs/rmsk_TEother_windows.txt

# Identify summit TEs (span the center of a 200bp chromHMM annotation)
awk '{if(($9+99.5 >= $2) && ($9+99.5 < $3)) print $0}' features/TEs/rmsk_TEother_windows.txt > features/TEs/rmsk_TEother_summit.txt
cut -f1-7 features/TEs/rmsk_TEother_summit.txt | sort -k1,1 -k2,2n | uniq > features/TEs/rmsk_TEother_summit
mv features/TEs/rmsk_TEother_summit features/TEs/rmsk_TEother_summit.txt

# Identify majority TEs
comm -23 <(sort features/TEs/rmsk_TEother.txt) <(sort features/TEs/rmsk_TEother_summit.txt) > features/TEs/rmsk_TEother_majority.txt
sort -k1,1 -k2,2n features/TEs/rmsk_TEother_majority.txt > features/TEs/rmsk_TEother_majority
mv features/TEs/rmsk_TEother_majority features/TEs/rmsk_TEother_majority.txt
```

```{bash TE chromHMM matrix, eval=FALSE}
# Intersect with individual TEs

#TE_landscape/chromHMM/TEs/intersect/E#_15_coreMarks_mnemonics.bed_TE [127 files]
for file in chromHMM_bedfiles/*.bed ; do bedtools intersect -wo -a rmsk_TE.txt -b $file > $file\_TE; done

#TE_landscape/chromHMM/TEs/intersect/E#_15_coreMarks_mnemonics.bed_other [127 files]
for file in chromHMM_bedfiles/*.bed; do bedtools intersect -wo -a rmsk_other.txt -b $file > $file\_other; done

# Original matrix creation

#TE_landscape/chromHMM/all_chromHMM_TE.txt
for file in chromHMM_bedfiles/*.bed_TE; do suffix=$(basename $file | cut -d '_' -f1); awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7, $11]+=$12;}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sep[8], a[i];}}' $file | awk -v x=$suffix 'BEGIN{OFS="\t";}{print $0, x}' - ; done >> all_chromHMM_TE.txt

#TE_landscape/chromHMM/all_chromHMM_TE_sorted.txt
sort -k1,1V -k2,2n -k3,3n -k6,6 -k10,10 all_chromHMM_TE.txt > all_chromHMM_TE_sorted.txt

#TE_landscape/chromHMM/all_chromHMM_other.txt
for file in chromHMM_bedfiles/*.bed_other; do suffix=$(basename $file | cut -d '_' -f1); awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7, $11]+=$12;}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sep[8], a[i];}}' $file | awk -v x=$suffix 'BEGIN{OFS="\t";}{print $0, x}' - ; done >> all_chromHMM_other.txt

#TE_landscape/chromHMM/all_chromHMM_other_sorted.txt
sort -k1,1V -k2,2n -k3,3n -k6,6 -k10,10 all_chromHMM_other.txt > all_chromHMM_other_sorted.txt

# Combined into one file

# Filtering to summit
while read line; do python ~/bin/TE_landscape/filter_summit.py chromHMM/TEs/intersect/$line\_15_coreMarks_mnemonics.bed_TE 1 8 chromHMM/summit/TEs/rmsk_TE_$line\_chromHMM.bed chromHMM; python ~/bin/TE_landscape/filter_summit.py chromHMM/TEs/intersect/$line\_15_coreMarks_mnemonics.bed_other 1 8 chromHMM/summit/TEs/rmsk_other_$line\_chromHMM.bed chromHMM; cat chromHMM/summit/TEs/rmsk_TE_$line\_chromHMM.bed chromHMM/summit/TEs/rmsk_other_$line\_chromHMM.bed > chromHMM/summit/TEs/rmsk_TEother_$line\_chromHMM.bed; rm chromHMM/summit/TEs/rmsk_TE_$line\_chromHMM.bed; rm chromHMM/summit/TEs/rmsk_other_$line\_chromHMM.bed; done < sample_lists/mnemonics.txt

# Including those with no summit
## From original TE x sample x state file, split by sample
while read line; do python identify_no_summit.py $line rmsk_TEother_$line\_chromHMM.bed rmsk_TEother_$line\_chromHMM_noSummit.txt 7 8; done < mnemonics.txt

# Matrix of TE x sample x state, including summit and majority TEs
while read line; do awk -v OFS='\t' -v sample=$line '{a[$1, $2, $3, $4, $5, $6, $7, $11]+=$12}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sample, a[i], sep[8], "summit"}}' rmsk_TEother_$line\_chromHMM.bed >> rmsk_TEother_chromHMM.txt; awk -v OFS='\t' -v sample=$line '{a[$1, $2, $3, $4, $5, $6, $7, $8]+=$9}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sample, a[i], sep[8], "majority"}}' rmsk_TEother_$line\_chromHMM_noSummit.txt >> rmsk_TEother_chromHMM.txt; done < mnemonics.txt

sort -k1,1V -k2,2n -k3,3n -k4,4 -k8,8 rmsk_TEother_chromHMM.txt > ~/TE_landscape/chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt
```

```{bash TE summit QC, eval=FALSE}
# Number of states per TE overlapping summit
for file in chromHMM/summit/TEs/rmsk_TEother_E*_chromHMM.bed; do cut -f1-7,11 $file | sort | uniq | awk '{a[$1, $2, $3, $4, $5, $6, $7]+=1}END{for(i in a){print a[i]}}' - | awk '{b[$1]+=1}END{for(i in b){print i, b[i]}}' -; done

# TEs not completely covered with summit rules
python ~/bin/TE_landscape/check_missing_bp.py chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt rmsk_TEother_total_chromHMM.txt

# Missing bp per sample
awk '{if($8 == "summit") a[$9]+=($3-$2-$10)}END{for(i in a){print i, a[i]}}' rmsk_TEother_total_chromHMM.txt
awk '{if($8 == "majority") a[$9]+=($3-$2-$10)}END{for(i in a){print i, a[i]}}' rmsk_TEother_total_chromHMM.txt

# Number of TEs missing bp per sample
awk '{if($8 == "summit") a[$9]+=1}END{for(i in a){print i, a[i]}}' rmsk_TEother_total_chromHMM.txt
awk '{if($8 == "majority") a[$9]+=1}END{for(i in a){print i, a[i]}}' rmsk_TEother_total_chromHMM.txt
```

```{bash TE chromHMM potential, eval=FALSE}
# All
python ~/bin/TE_landscape/potential.py chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt features/TEs/rmsk_TEother.txt chromHMM/chromHMM_states.txt sample_lists/mnemonics.txt chromHMM/potential/rmsk_TEother_chromHMM_summit_potential.txt 0 7 9 8

# No cancer
python ~/bin/TE_landscape/potential.py chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt features/TEs/rmsk_TEother.txt chromHMM/chromHMM_states.txt sample_lists/mnemonics_noCancer.txt chromHMM/potential/rmsk_TEother_chromHMM_summit_potential_noCancer.txt 0 7 9 8
```

```{bash TE chromHMM max intra, eval=FALSE}
# Maximum states per TE in any sample
python ~/bin/TE_landscape/state_sharing_intra_max.py chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt chromHMM/rmsk_TEother_chromHMM_summit_max.txt 7
```

```{bash process WGBS, eval=FALSE}
# Bedfile of CpG methylation level across all chromsomes, both strands 	 
#TE_landscape/WGBS/all_CpG_Meth.bedGraph	
#TE_landscape/WGBS/all_CpG_Meth.bed	
for file in /bar/mchoudhary/2chromTE/Meth/FractionalMethylation_Removed_E027_E064_Fixed_E012/chr*.fm; do awk -v chr=$(basename "$file" .fm) -v OFS='\t' '{print chr, $1, $1+1, $0}' $file >> all_CpG_Meth.bedGraph; done &
cut -f1-3,5- all_CpG_Meth.bedGraph | sort -k1,1V -k2,2n -k3,3n - > all_CpG_Meth.bed
```

```{bash TE WGBS matrix, eval=FALSE}
# Intersect with individual TEs
#TE_landscape/WGBS/TE_CpG_Meth_new.bed
split -l 1000000 ~/TE_landscape/all_CpG_Meth.bed
for file in xa*; do echo $file; bedtools intersect -wo -a ~/TE_landscape/rmsk_TEother.txt -b $file >>  TE_CpG_Meth_new.bed ; done

# Average methylation level of each TE in each sample	 
#TE_landscape/WGBS/TE_CpG_Meth_new_average.txt	
python ~/bin/TE_landscape/average_methylation.py TE_CpG_Meth_new.bed ~/TE_landscape/rmsk_TEother.txt ~/TE_landscape/WGBS_samples.txt TE_CpG_Meth_new_average.txt
```

```{bash TE CpG count, eval=FALSE}
# Number of CpGs per TE	 
#TE_landscape/WGBS/TE_CpG_count.txt	
awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' TE_CpG_Meth_new.bed > TE_CpG_count.txt
```

```{bash TE DHS matrix, eval=FALSE}
# Intersect with individual TEs
TE_landscape/DNase/intersect_sample_list_DNase_peak.sh
#TE_landscape/DNase/intersect/TEs/rmsk_TEother_E#-DNase.macs2.narrowPeak [53 files]

# Filtering to summit
for file in DNase/intersect/TEs/rmsk_TEother_E*-DNase.macs2.narrowPeak; do python ~/bin/TE_landscape/filter_summit.py $file 1 8 DNase/summit/TEs/$( basename $file -DNase.macs2.narrowPeak)_DNase_summit.txt; done

# Matrix of TE x sample x state
while read line; do awk -v OFS='\t' -v sample=$line '{a[$1, $2, $3, $4, $5, $6, $7]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sample, a[i]}}' DNase/summit/TEs/rmsk_TEother_$line\_DNase_summit.txt >> DNase/rmsk_TEother_DNase_summit.txt; done < sample_lists/DNase_samples.txt
```

```{bash TE H3K27ac matrix, eval=FALSE}
# Intersect with individual TEs
#TE_landscape/H3K27ac/intersect/TEs/rmsk_TEother_E#-H3K27ac.narrowPeak [98 files]
while read line; do bedtools intersect -wo -a ../rmsk_TEother.txt -b H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak > H3K27ac_TEs/rmsk_TEother_$line\-H3K27ac.narrowPeak; done < H3K27ac_samples.txt

# Filtering to summit
for file in H3K27ac/intersect/TEs/rmsk_TEother_E*-H3K27ac.narrowPeak; do python ~/bin/TE_landscape/filter_summit.py $file 1 8 H3K27ac/summit/TEs/$( basename $file -H3K27ac.narrowPeak)_H3K27ac_summit.txt; done

# Matrix of TE x sample x state
while read line; do awk -v OFS='\t' -v sample=$line '{a[$1, $2, $3, $4, $5, $6, $7]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sample, a[i]}}' H3K27ac/summit/TEs/rmsk_TEother_$line\_H3K27ac_summit.txt >> H3K27ac/rmsk_TEother_H3K27ac_summit.txt; done < sample_lists/H3K27ac_samples.txt
```

```{bash TE RNA matrix, eval=FALSE}
# Average RNA-seq expression

# Intersect TEs with RNA-seq bedfiles (sorted)
intersect_RNA_agnostic.sh

# Average expression per TE, raw, combined samples
#TE_landscape/RNAseq/rmsk_TEother_average.txt
cut -f1-7 rmsk_TEother.txt.sorted_E003_average.txt > rmsk_TEother_average.txt
while read line; do cut -f10 rmsk_TEother.txt.sorted_$line\_average.txt | paste rmsk_TEother_average.txt - > rmsk_TEother_average; mv rmsk_TEother_average rmsk_TEother_average.txt ; done < RNA_samples_agnostic.txt
```

```{bash individual RefSeq exons, eval=FALSE}
# Unique Refseq exon coordinates, filtered to standard chromosomes
awk -v OFS='\t' '{if($1 !~ /_/) print $1, $2, $3, $6}' ~/genic_features/RefSeq/refseq_exons.txt | sort | uniq > ~/genic_features/RefSeq/refseq_exons_unique.txt

# Sorted
sort -k1,1 -k2,2n ~/genic_features/RefSeq/refseq_exons_unique.txt > RNAseq/refseq_exons_unique.txt.sorted

# Number of exons overlapping each other
sort -k1,1V -k2,2n -k3,3n ~/genic_features/RefSeq/refseq_exons_unique.txt | bedtools merge -i - | wc -l
```

```{bash exon RNA matrix, eval=FALSE}
# Intersect exons with RNA-seq bedfiles
intersect_RNA_agnostic.sh

# Average expression per exon, raw, combined samples
#TE_landscape/RNAseq/refseq_exons_average.txt
cut -f1-4 refseq_exons_unique.txt.sorted_E003_average.txt > refseq_exons_average.txt
while read line; do cut -f7 refseq_exons_unique.txt.sorted_$line\_average.txt | paste refseq_exons_average.txt - > refseq_exons_average; mv refseq_exons_average refseq_exons_average.txt ; done < RNA_samples_agnostic.txt
```

Split into subfamily x state - TEs ever in state

```{bash by subfamily state, eval=FALSE}
# TEs by subfamily and state
awk '{if($10 != "8_ZNF/Rpts") print > "chromHMM/subfamily/by_state/"$4"_"$10".txt"; else print > "chromHMM/subfamily/by_state/"$4"_8_ZNF.Rpts.txt"}' chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt

# WGBS
awk '{print > "WGBS/subfamily/by_state/"$4"_"$10".txt"}' WGBS/TE_WGBS_state_sorted.txt

# TEs by subfamily and state
awk '{print $0 > "DNase/subfamily/"$4"_DNase.txt"}' DNase/rmsk_TEother_DNase_summit.txt

# TEs by subfamily and state
awk '{print $0 > "H3K27ac/subfamily/"$4"_H3K27ac.txt"}' H3K27ac/rmsk_TEother_H3K27ac_summit.txt
```

```{r create TE epigenetics, eval=FALSE}
source("R_scripts/chromHMM_matrix.R")
source("R_scripts/WGBS_TE_avg_methylation.R")
source("R_scripts/DNase_peaks.R")
source("R_scripts/H3K27ac_peaks.R")
source("R_scripts/RNAseq.R")
```

```{bash individual RefSeq promoters, eval=FALSE}
# Unique Refseq promoter coordinates	 
#genic_features/RefSeq/refseq_promoters_unique.txt	
awk -v OFS='\t' '{print $1, $2, $3, $6}' ~/genic_features/RefSeq/refseq_promoters.txt | sort | uniq > refseq_promoters_unique.txt

# Filtered to standard chromosomes	 
#genic_features/RefSeq/refseq_promoters_unique_std.txt	
awk -v OFS='\t' '{if($1 !~ /_/) print $0}' refseq_promoters_unique.txt > refseq_promoters_unique_std.txt

# Number of promoters overlapping each other
sort -k1,1V -k2,2n -k3,3n ~/genic_features/RefSeq/refseq_promoters_unique_std.txt | bedtools merge -i - | wc -l
```

```{bash promoter chromHMM matrix, eval=FALSE}
# Intersect with Refseq promoters
#TE_landscape/chromHMM/Refseq_promoters/chromHMM_refseq_promoters_unique.txt
for file in ../../raw_data/chromHMM/E*.bed; do suffix=$(basename $file | cut -d '_' -f1); bedtools intersect -wo -a refseq_promoters_unique.txt -b $file | awk -v OFS='\t' -v sample=$suffix '{print $0, sample}' - >> chromHMM_refseq_promoters_unique.txt; done

# Filtering to summit
python ~/bin/TE_landscape/filter_summit.py chromHMM/Refseq_promoters/chromHMM_refseq_promoters_unique.txt 1 5 chromHMM/summit/promoters/chromHMM_refseq_promoters_unique_summit.txt chromHMM

awk -v OFS='\t' '{a[$1, $2, $3, $4, $8, $10]+=$9}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[6], a[i], sep[5]}}' chromHMM/summit/promoters/chromHMM_refseq_promoters_unique_summit.txt | sort -k1,1V -k2,2n -k3,3n -k5,5 - > chromHMM/refseq_promoters_unique_chromHMM_summit_sorted.txt
```

```{bash promoter chromHMM potential, eval=FALSE}
# Refseq promoters
python ~/bin/TE_landscape/potential.py chromHMM/refseq_promoters_unique_chromHMM_summit_sorted.txt ~/genic_features/RefSeq/refseq_promoters_unique_std.txt chromHMM/chromHMM_states.txt sample_lists/mnemonics.txt chromHMM/potential/refseq_promoters_chromHMM_summit_potential.txt 0 4 6 5
```

```{bash promoter WGBS matrix, eval=FALSE}
# Intersect with Refseq promoters
#TE_landscape/WGBS/Refseq_promoters/refseq_promoter_unique_CpG_Meth.bed
bedtools intersect -wo -a ~/genic_features/RefSeq/refseq_promoters_unique.txt -b ../all_CpG_Meth.bed > refseq_promoter_unique_CpG_Meth.bed

# Average methylation of unique Refseq promoters	 
#TE_landscape/WGBS/Refseq_promoters/refseq_promoter_unique_CpG_Meth_average.txt	
python ~/bin/TE_landscape/average_methylation_promoter.py refseq_promoter_unique_CpG_Meth.bed ~/genic_features/RefSeq/refseq_promoters_unique_std.txt ../../sample_lists/WGBS_samples.txt refseq_promoter_unique_CpG_Meth_average.txt
```

```{bash promoter DHS matrix, eval=FALSE}
# Intersect with Refseq promoters
#TE_landscape/DNase/Refseq_promoters/refseq_promoter_unique_E*-DNase.macs2.narrowPeak [53 files]
while read line; do bedtools intersect -wo -a ~/genic_features/RefSeq/refseq_promoters_unique.txt -b ../../raw_data/DNase/DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak > refseq_promoter_unique_$line\-DNase.macs2.narrowPeak; done < ../../sample_lists/DNase_samples.txt

# Filter to summit
for file in DNase/Refseq_promoters/refseq_promoter_unique_E*-DNase.macs2.narrowPeak; do python ~/bin/TE_landscape/filter_summit.py $file 1 5 DNase/summit/promoters/$( basename $file -DNase.macs2.narrowPeak)_DNase_summit.txt; done

while read line; do awk -v OFS='\t' -v sample=$line '{a[$1, $2, $3, $4]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sample, a[i]}}' DNase/summit/promoters/refseq_promoter_unique_$line\_DNase_summit.txt >> DNase/refseq_promoter_unique_DNase_summit.txt; done < sample_lists/DNase_samples.txt
```

```{bash promoter H3K27ac matrix, eval=FALSE}
# Intersect with Refseq promoters
#TE_landscape/H3K27ac/Refseq_promoters/refseq_promoter_unique_E*-H3K27ac.narrowPeak [98 files]
while read line; do bedtools intersect -wo -a ~/genic_features/RefSeq/refseq_promoters_unique.txt -b ../../raw_data/H3K27ac/H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak > refseq_promoter_unique_$line\-H3K27ac.narrowPeak; done < ../../sample_lists/H3K27ac_samples.txt

# Filter summit
for file in H3K27ac/Refseq_promoters/refseq_promoter_unique_E*-H3K27ac.narrowPeak; do python ~/bin/TE_landscape/filter_summit.py $file 1 5 H3K27ac/summit/promoters/$( basename $file -H3K27ac.narrowPeak)_H3K27ac_summit.txt; done

while read line; do awk -v OFS='\t' -v sample=$line '{a[$1, $2, $3, $4]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sample, a[i]}}' H3K27ac/summit/promoters/refseq_promoter_unique_$line\_H3K27ac_summit.txt >> H3K27ac/refseq_promoter_unique_H3K27ac_summit.txt; done < sample_lists/H3K27ac_samples.txt
```

```{r create promoter epigenetics, eval=FALSE}
source("R_scripts/promoter_matrices.R")
```

All shuffled files have 4430788 TEs, same number of class, same number of subfamily, but not same number on each chromosome (3/4/18). 

```{bash shuffled TEs, eval=FALSE}
# hg19 gaps, sorted	 
#TE_landscape/features/shuffled_TEs/gap_sorted.txt	
tail -n+2 /bar/genomes/hg19/gap/gap.txt | awk -v OFS='\t' '{print $2, $3, $4}' - | sort -k1,1V -k2,2n -k3,3n - > gap_sorted.txt

# Shuffled TE positions (no gaps)	 
#TE_landscape/features/shuffled_TEs/rmsk_TE_shuffle_#.txt [10 files]	
for i in {1..10}; do bedtools shuffle -i ~/TE_landscape/features/TEs/rmsk_TEother.txt -g ~/TE_landscape/features/hg19_standard.genome -excl gap_sorted.txt > rmsk_TE_shuffle_$i\.txt; done

# Sorted shuffled TEs (2/20/19)
for i in {1..10}; do sort -k1,1 -k2,2n features/shuffled_TEs/rmsk_TE_shuffle_$i\.txt > features/shuffled_TEs/rmsk_TE_shuffle_$i\_sorted.txt; done
for i in {1..10}; do mv features/shuffled_TEs/rmsk_TE_shuffle_$i\_sorted.txt features/shuffled_TEs/rmsk_TE_shuffle_$i\.txt; done
```

```{bash shuffled feature overlap, eval=FALSE}
# Intersection between shuffled TEs, Refseq features
#TE_landscape/features/shuffled_TEs/intersect_features/rmsk_TE_shuffle_#_cpgIsland.txt [10 files]
#TE_landscape/features/shuffled_TEs/intersect_features/rmsk_TE_shuffle_#_refseq_intergenic.txt [10 files]
#TE_landscape/features/shuffled_TEs/intersect_features/rmsk_TE_shuffle_#_refseq_[feature]_merge.txt [70 files]
#TE_landscape/features/shuffled_TEs/intersect_features/rmsk_TE_shuffle_#_refseq_[feature]_nc_merge.txt [60 files]
#TE_landscape/features/shuffled_TEs/intersect_features/rmsk_TE_shuffle_#_refseq_[feature]_pc_merge.txt [70 files]

for j in {1..10}; do for file in ~/genic_features/RefSeq/*_merge.txt; do bedtools intersect -wo -a features/shuffled_TEs/rmsk_TE_shuffle_$j\.txt -b $file | awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=$11}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' - > features/shuffled_TEs/intersect_features/rmsk_TE_shuffle_$j\_$(basename "$file" .txt)\.txt; done; bedtools intersect -wo -a features/shuffled_TEs/rmsk_TE_shuffle_$j\.txt -b ~/genic_features/RefSeq/refseq_intergenic.txt | awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=$11}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' - > features/shuffled_TEs/intersect_features/rmsk_TE_shuffle_$j\_refseq_intergenic.txt; bedtools intersect -wo -a features/shuffled_TEs/rmsk_TE_shuffle_$j\.txt -b ~/genic_features/cpgIslandExtUnmasked_merge.txt | awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=$11}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' - > features/shuffled_TEs/intersect_features/rmsk_TE_shuffle_$j\_cpgIsland.txt; done
```

```{bash shuffled chromHMM matrix, eval=FALSE}
# Summit
## Identify summit TEs by intersecting with 200bp windows
for i in {1..10}; do bedtools intersect -wo -a features/shuffled_TEs/rmsk_TE_shuffle_$i\.txt -b features/hg19_standard.windows > features/shuffled_TEs/rmsk_TE_shuffle_$i\_windows.txt; done

## Unique shuffled TEs that overlap any chromHMM bin summit (i.e., center of 200bp windows)
for i in {1..10}; do awk '{if(($9+99.5 >= $2) && ($9+99.5 < $3)) print $0}' features/shuffled_TEs/rmsk_TE_shuffle_$i\_windows.txt > features/shuffled_TEs/rmsk_TE_shuffle_$i\_summit.txt; done
for file in features/shuffled_TEs/rmsk_TE_shuffle_*_summit.txt; do cut -f1-7 $file | sort -k1,1 -k2,2n | uniq > features/shuffled_TEs/$(basename $file .txt); mv features/shuffled_TEs/$(basename $file .txt) $file; done

## Intersect with 200bp chromHMM windows 
for i in {1..2}; do while read line; do bedtools intersect -wo -sorted -a features/shuffled_TEs/rmsk_TE_shuffle_$i\_summit.txt -b chromHMM/genome/windows/windows_$line\.bed > /scratch/ecp/shuffled/rmsk_TE_shuffle_$i\_$line\_summit.txt; done < sample_lists/mnemonics.txt; done

## Filter to bin summit overlap and sum state within TE
for file in rmsk_TE_shuffle_1_E*_summit.txt; do awk -v OFS='\t' '{if(($9+99.5 >= $2) && ($9+99.5 < $3)) a[$1, $2, $3, $4, $5, $6, $7, $11]+=$12}END{for(i in a) {split(i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sep[8], a[i]}}' $file > $(basename $file .txt); done &
for i in {1..7}; do for file in rmsk_TE_shuffle_$i\_E*_summit; do mv $file $file\.txt; done; done

# Majority
## TEs that are not summit overlapping (majority), sorted
for i in {1..10}; do comm -23 <(sort features/shuffled_TEs/rmsk_TE_shuffle_$i\.txt) <(cut -f1-7 features/shuffled_TEs/rmsk_TE_shuffle_$i\_summit.txt | sort) > features/shuffled_TEs/rmsk_TE_shuffle_$i\_majority.txt; done
for file in features/shuffled_TEs/rmsk_TE_shuffle_*_majority.txt; do sort -k1,1 -k2,2n $file > features/shuffled_TEs/$(basename $file .txt); mv features/shuffled_TEs/$(basename $file .txt) $file; done

## Intersection with chromHMM states
## Each can only overlap at most 2 states
for i in {1..10}; do while read line; do bedtools intersect -wo -a features/shuffled_TEs/rmsk_TE_shuffle_$i\_majority.txt -b raw_data/chromHMM/$line\_15_coreMarks_mnemonics.bed > /scratch/ecp/shuffled/rmsk_TE_shuffle_$i\_$line\_majority.txt; done < sample_lists/mnemonics.txt; done &

## Pick the majority state
for file in rmsk_TE_shuffle_1_E*_majority.txt; do python ~/bin/TE_landscape/pick_majority.py $file $(basename $file .txt) 10 11; done
for i in {2..10}; do for file in rmsk_TE_shuffle_$i\_E*_majority.txt; do python ~/bin/TE_landscape/pick_majority.py $file $(basename $file .txt) 10 11; done; done &
for file in rmsk_TE_shuffle_*_E*_majority; do mv $file $file\.txt; done

# Combined summit and majority TEs
for i in {1..7}; do while read line; do awk -v OFS='\t' -v sample=$line '{print $0, sample, "summit"}' rmsk_TE_shuffle_$i\_$line\_summit.txt >> rmsk_TE_shuffle_$i\_chromHMM.txt; awk -v OFS='\t' -v sample=$line '{print $0, sample, "majority"}' rmsk_TE_shuffle_$i\_$line\_majority.txt >> rmsk_TE_shuffle_$i\_chromHMM.txt; done < ~/TE_landscape/sample_lists/mnemonics.txt; done &

# Sorted
# Where chromosomes.txt is chr 1-22 and sex chromosomes
for i in {1..10}; do awk '{print>$1}' rmsk_TE_shuffle_$i\_chromHMM.txt; while read line; do sort -k1,1V -k2,2n -k3,3n -k4,4 -k10,10 $line >> rmsk_TE_shuffle_$i\_chromHMM_sorted.txt; rm $line; done < chromosomes.txt; done
```

chromHMM overlap bp is not the same for shuffled TEs as for real summit TEs!

```{bash shuffled chromHMM potential, eval=FALSE}
for i in {1..10}; do python ~/bin/TE_landscape/potential.py rmsk_TE_shuffle_$i\_chromHMM_sorted.txt ~/TE_landscape/features/shuffled_TEs/rmsk_TE_shuffle_$i\.txt chromHMM_states.txt mnemonics.txt rmsk_TE_shuffle_$i\_chromHMM_potential.txt 0 9 7 8; done &
```

```{bash shuffled WGBS matrix, eval=FALSE}
# Intersect with shuffled TEs
split -l 1000000 ~/TE_landscape/WGBS/all_CpG_Meth.bed
for i in {1..10}; do for file in x*; do bedtools intersect -wo -a ../rmsk_TE_shuffle_$i\.txt -b $file >> rmsk_TE_shuffle_$i\_Meth.bed; done; done

# Average methylation per shuffled TE	 
#TE_landscape/WGBS/shuffled/rmsk_TE_shuffle_#_Meth_average.txt [10 files]	
for i in {1..10}; do python ~/bin/TE_landscape/average_methylation.py rmsk_TE_shuffle_$i\_Meth.bed ../rmsk_TE_shuffle_$i\.txt ~/TE_landscape/sample_lists/WGBS_samples.txt rmsk_TE_shuffle_$i\_Meth_average.txt; done
```

```{bash shuffled CpG count, eval=FALSE}
#TE_landscape/WGBS/shuffled/TE_CpG_count_#.txt [10 files]
for j in {1..10}; do awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=1}END{for(k in a){split(k,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[k];}}' rmsk_TE_shuffle_$j\_Meth.bed > TE_CpG_count_$j.txt; done
```

```{bash shuffled DHS, eval=FALSE}
for i in {1..10}; do while read line; do bedtools intersect -wo -a features/shuffled_TEs/rmsk_TE_shuffle_$i\.txt -b raw_data/DNase/DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak > DNase/shuffled/rmsk_TE_$line\_$i\-DNase.narrowPeak; done < sample_lists/DNase_samples.txt; done

for file in DNase/shuffled/rmsk_TE_E*-DNase.narrowPeak; do python ~/bin/TE_landscape/filter_summit.py $file 1 8 DNase/shuffled/$( basename $file -DNase.narrowPeak)_DNase_summit.txt DNase; done

for j in {1..10}; do while read line; do awk -v OFS='\t' -v sample=$line '{a[$1, $2, $3, $4, $5, $6, $7]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sample, a[i]}}' DNase/shuffled/rmsk_TE_$line\_$j\_DNase_summit.txt >> DNase/shuffled/rmsk_TE_$j\_DNase_summit.txt; done < sample_lists/DNase_samples.txt; done &
```

```{bash shuffled H3K27ac, eval=FALSE}
for i in {1..10}; do while read line; do bedtools intersect -wo -a features/shuffled_TEs/rmsk_TE_shuffle_$i\.txt -b raw_data/H3K27ac/H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak > H3K27ac/shuffled/rmsk_TE_$line\_$i\-H3K27ac.narrowPeak; done < sample_lists/H3K27ac_samples.txt; done

for file in H3K27ac/shuffled/rmsk_TE_E*-H3K27ac.narrowPeak; do python ~/bin/TE_landscape/filter_summit.py $file 1 8 H3K27ac/shuffled/$( basename $file -H3K27ac.narrowPeak)_H3K27ac_summit.txt H3K27ac; done

for j in {1..10}; do while read line; do awk -v OFS='\t' -v sample=$line '{a[$1, $2, $3, $4, $5, $6, $7]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sample, a[i]}}' H3K27ac/shuffled/rmsk_TE_$line\_$j\_H3K27ac_summit.txt >> H3K27ac/shuffled/rmsk_TE_$j\_H3K27ac_summit.txt; done < sample_lists/H3K27ac_samples.txt; done
```

```{r create shuffled matrices, eval=FALSE}
source("R_scripts/shuffled_matrix.R")
source("R_scripts/shuffled_matrix_chromHMM.R")
source("R_scripts/shuffled_matrix_WGBS.R")
source("R_scripts/shuffled_matrix_DNase.R")
source("R_scripts/shuffled_matrix_H3K27ac.R")
```

## Load essential matrices

```{r load essential matrices}
load("R_datasets/rmsk_TE.RData")
load("R_datasets/chromHMM_TE_state.RData")
load("R_datasets/TE_meth_average.RData")
load("R_datasets/TE_DNase_peaks.RData")
load("R_datasets/TE_H3K27ac_peaks.RData")
load("R_datasets/rna.RData")
```

## Load feature sizes

```{bash merged TEs, eval=FALSE}
# Merged all TE RepeatMasker file	 
#TE_landscape/features/TEs/merge/rmsk_TEother_merge.txt	
cat rmsk_TE.txt rmsk_other.txt | sort -k1,1 -k2,2n - | bedtools merge -i - > rmsk_TEother_merge.txt
```

```{bash merged TE width, eval=FALSE}
awk '{sum+=$3-$2}END{print sum}' features/TEs/merge/rmsk_TEother_merge.txt
awk '{if($1 != "chrY") sum+=$3-$2}END{print sum}' features/TEs/merge/rmsk_TEother_merge.txt
```

```{bash count CpGs, eval=FALSE}
wc -l WGBS/all_CpG_Meth.bed

#TE_landscape/WGBS/CpG_TE_Meth.bed
bedtools intersect -a all_CpG_Meth.bed -b rmsk_TEother_merge.txt > CpG_TE_Meth.bed &

wc -l WGBS/CpG_TE_Meth.bed
```

```{bash get feature length, eval=FALSE}
bash feature_overlap_length.sh
```

```{r load feature sizes}
# Genome
hg19_genome = read.table("raw_data/hg19.genome",sep='\t',header=TRUE)
GENOME_WIDTH = sum(as.numeric(hg19_genome[which(hg19_genome$chrom %in% standard_chromosomes),]$size))
GENOME_WIDTH_noY = sum(as.numeric(hg19_genome[which(hg19_genome$chrom %in% standard_chromosomes & hg19_genome$chrom != "chrY"),]$size))

# Total widths
CHROMHMM_TOTAL_WIDTH = sample_counts["chrY","chromHMM"]*GENOME_WIDTH + (sample_counts["All","chromHMM"]-sample_counts["chrY","chromHMM"])*GENOME_WIDTH_noY
DNASE_TOTAL_WIDTH = sample_counts["chrY","DNase"]*GENOME_WIDTH + (sample_counts["All","DNase"]-sample_counts["chrY","DNase"])*GENOME_WIDTH_noY
H3K27AC_TOTAL_WIDTH = sample_counts["chrY","H3K27ac"]*GENOME_WIDTH + (sample_counts["All","H3K27ac"]-sample_counts["chrY","H3K27ac"])*GENOME_WIDTH_noY

# TEs 
NUM_TE = dim(rmsk_TE)[1]
NUM_TE_noY = dim(rmsk_TE[which(rmsk_TE$chromosome != "chrY"),])[1]
NUM_TE_WGBS = dim(TE_meth_average)[1]
  
MERGED_TE_WIDTH = 1389947349
MERGED_TE_WIDTH_noY = 1375898142

# Total widths
CHROMHMM_TE_WIDTH = sample_counts["chrY","chromHMM"]*MERGED_TE_WIDTH + (sample_counts["All","chromHMM"]-sample_counts["chrY","chromHMM"])*MERGED_TE_WIDTH_noY
DNASE_TE_WIDTH = sample_counts["chrY","DNase"]*MERGED_TE_WIDTH + (sample_counts["All","DNase"]-sample_counts["chrY","DNase"])*MERGED_TE_WIDTH_noY
H3K27AC_TE_WIDTH = sample_counts["chrY","H3K27ac"]*MERGED_TE_WIDTH + (sample_counts["All","H3K27ac"]-sample_counts["chrY","H3K27ac"])*MERGED_TE_WIDTH_noY

# CpGs
ALL_CPGS = 56434896/2
TE_CPGS = 28373958/2

# Feature lengths
feature_overlap = as.data.frame(matrix(readLines("features/intersect_features/feature_overlap.txt"),ncol=5,byrow=TRUE))
colnames(feature_overlap) = c("Cohort","Genome","Genome_noY","TEs","TEs_stranded")
feature_overlap$Cohort = gsub("refseq_","",feature_overlap$Cohort)
feature_overlap = split_coding(feature_overlap)
feature_overlap[,2:5] = apply(feature_overlap[,2:5],2,function(x) as.numeric(x))
feature_overlap = feature_overlap[,c(1,6:7,2:5)]
```

## Calculate TE class and subfamily stats

```{bash classes, eval=FALSE}
# Four large TE classes (DNA, LINE, LTR, SINE)
TE_class.txt

# 8 other TE classes from the rmsk file
other_class.txt
```

```{bash merged class, eval=FALSE}
# Merged file for each class	
#TE_landscape/features/TEs/class/rmsk_[class].txt [13 files]	
while read line; do awk -v OFS='\t' -v class=$line '{if($5 == class)print $0}' rmsk_TE.txt | bedtools merge -i - > TE_classes/rmsk_$line\.txt; done < TE_class.txt &
while read line; do awk -v OFS='\t' -v class=$line '{if($5 == class)print $0}' rmsk_other.txt | bedtools merge -i - > TE_classes/rmsk_$line\.txt; done < other_class.txt &

# Original Unconfident class (without RC)
awk -v OFS='\t' '{if(($5=="LTR?")||($5=="DNA?")||($5=="LINE?")||($5=="SINE?")||($5=="Unknown?")||($5=="Unknown")) print $0}' ../rmsk_other.txt | bedtools merge -i - > rmsk_Unconfident.txt

# Adding RC to Unconfident class	 
#TE_landscape/features/TEs/class/rmsk_Unconfident_RC.txt	
awk -v OFS='\t' '{if(($5=="LTR?")||($5=="DNA?")||($5=="LINE?")||($5=="SINE?")||($5=="Unknown?")||($5=="Unknown")||($5=="RC")) print $0}' ../rmsk_TEother.txt | bedtools merge -i - > rmsk_Unconfident_RC.txt
    
# Merged TE bases by class	 
# Where TEother_class.txt is DNA, LINE, LTR, SINE, Other, RC, and Unconfident
#TE_landscape/features/TEs/class/TEother_class_merge.txt	
while read line; do awk -v OFS='\t' -v class=$line '{print $0, class}' rmsk_$line\.txt >> TEother_class_merge.txt ; done < TEother_class.txt

# Adding merged Unconfident-RC bases	 
awk -v OFS='\t' '{print $0, "Unconfident_RC"}' rmsk_Unconfident_RC.txt >> TEother_class_merge.txt
```

```{bash class CpG count, eval=FALSE}
# Number of CpGs per class	 
#TE_landscape/WGBS/class/TE_CpG_class.txt	
awk -v OFS='\t' '{print $5, $8, $9, $10}' TE_CpG_Meth_new.bed | sort | uniq | awk -v OFS='\t' '{a[$1]+=1}END{for(i in a){print i, a[i];}}' - > TE_CpG_class.txt 
awk -v OFS='\t' '{if($5 == "LTR?" || $5 == "DNA?" || $5 == "SINE?" || $5 == "LINE?" || $5 == "Unknown?" || $5 == "Unknown") print $8, $9, $10}' TE_CpG_Meth_new.bed | sort | uniq | wc -l >> TE_CpG_class.txt
awk -v OFS='\t' '{if($5 == "LTR?" || $5 == "DNA?" || $5 == "SINE?" || $5 == "LINE?" || $5 == "Unknown?" || $5 == "Unknown" || $5 == "RC") print $8, $9, $10}' TE_CpG_Meth_new.bed | sort | uniq | wc -l >> class/TE_CpG_class.txt #Had to add name manually
```

```{bash class length, eval=FALSE}
# Class lengths with/without chrY	 
#TE_landscape/features/TEs/class/class_lengths.txt	
awk -v OFS='\t' '{a[$4]+=$3-$2}END{for(i in a){print i, a[i]}}' TEother_class_merge.txt > class_lengths.txt

#TE_landscape/features/TEs/class/class_lengths_noY.txt		 
awk -v OFS='\t' '{if($1 != "chrY") a[$4]+=$3-$2}END{for(i in a){print i, a[i]}}' TEother_class_merge.txt > class_lengths_noY.txt
```

```{bash subfamilies, eval=FALSE}
# 903 subfamilies in the DNA/LINE/SINE/LTR classes OR all 968 subfamilies
subfamilies.txt

# 65 subfamilies belonging to Other and SVA classes
other_subfamilies.txt
```

```{bash merged subfamilies, eval=FALSE}
# Merged file for each subfamily	 
#TE_landscape/features/TEs/subfamily/TEother_subfamily_merge.txt	
while read line; do awk -v OFS='\t' -v subfam=$line '{if($4 == subfam)print $0}' ../rmsk_TE.txt | bedtools merge -i - > rmsk_$line\.txt; done < subfamilies.txt
while read line; do awk -v OFS='\t' -v subfam=$line '{if($4 == subfam)print $0}' ../rmsk_other.txt | bedtools merge -i - > rmsk_$line\.txt; done < other_subfamilies.txt

while read line; do awk -v OFS='\t' -v subfam=$line '{print $0, subfam}' rmsk_$line\.txt >> TEother_subfamily_merge.txt; rm rmsk_$line\.txt; done < subfamilies.txt
while read line; do awk -v OFS='\t' -v subfam=$line '{print $0, subfam}' rmsk_$line\.txt >> TEother_subfamily_merge.txt; done < other_subfamilies.txt
while read line; do rm rmsk_$line\.txt; done < other_subfamilies.txt
```

```{bash subfamily length, eval=FALSE}
# Subfamily lengths with/without chrY	 
#TE_landscape/features/TEs/subfamily/subfamily_lengths.txt	
awk -v OFS='\t' '{a[$4]+=$3-$2}END{for(i in a){print i, a[i]}}' TEother_subfamily_merge.txt > subfamily_lengths.txt

#TE_landscape/features/TEs/subfamily/subfamily_lengths_noY.txt		 
awk -v OFS='\t' '{if($1 != "chrY") a[$4]+=$3-$2}END{for(i in a){print i, a[i]}}' TEother_subfamily_merge.txt > subfamily_lengths_noY.txt
```

```{bash subfamily CpG count, eval=FALSE}
# Number of CpGs per subfamily	 
#TE_landscape/WGBS/subfamily/TE_CpG_subfamily.txt	
awk -v OFS='\t' '{print $4, $8, $9, $10}' TE_CpG_Meth_new.bed | sort | uniq | awk -v OFS='\t' '{a[$1]+=1}END{for(i in a){print i, a[i];}}' - > TE_CpG_subfamily.txt
```

```{r calculate TE stats, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/TE_class_stats.R")
source("R_scripts/TE_subfamily_stats.R")
```

# Introduction

## Basic sample stats

```{r basic sample stats}
sample_counts
```

## Basic TE statistics

```{r basic TE stats}
# Number of TEs
NUM_TE

# % of genome
MERGED_TE_WIDTH/GENOME_WIDTH

# % CpGs overlapping TEs
TE_CPGS/ALL_CPGS

# Number of families
length(unique(rmsk_TE$family))

# Number of subfamilies
length(unique(rmsk_TE$subfamily))

# Length of TEs
ddply(rmsk_TE,.(),summarise,Min=min(Length),Median=median(Length),Max=max(Length))

# % TEs overlapping CpGs
NUM_TE_WGBS
NUM_TE_WGBS/NUM_TE
NUM_TE-NUM_TE_WGBS
ddply(TE_meth_average,.(),summarise,Min=min(CpGs),Median=median(CpGs),Max=max(CpGs))
```

# Results

## Contribution

```{bash genome chromHMM, eval=FALSE}
# Whole genome

# Number of bases in each state in each sample	 
#TE_landscape/chromHMM/genome/sample_summaries/E#_15_coreMarks_mnemonics.bed_state [127 files]	
for file in chromHMM_bedfiles/E*_15_coreMarks_mnemonics.bed; do awk 'BEGIN{SUM=0}{SUM+=$3-$2}END{print SUM}' $file > $file\_state; while read line; do grep $line $file | awk 'BEGIN{SUM=0}{SUM+=$3-$2}END{print SUM}' - >> $file\_state; done < chromHMM_states.txt ; done

# Table of number of bases in each state in each sample	(combines bp counts of state per sample into one file)
#TE_landscape/chromHMM/genome/mnemonics_state.txt	
bash combine_states.sh chromHMM_states.txt mnemonics.txt chromHMM_bedfiles/*state
```

```{bash TE chromHMM, eval=FALSE}
# Merged TEs

# Intersect with merged TEs
#TE_landscape/chromHMM/TEs/intersect/E#_15_coreMarks_mnemonics.bed_TEother_merge [127 files]
for file in chromHMM_bedfiles/E*_15_coreMarks_mnemonics.bed; do output=$(basename $file); bedtools intersect -wo -a rmsk_TEother_merge.txt -b $file > chromHMM_other/$output\_TEother_merge; done

# Sum number of bases in state
#TE_landscape/chromHMM/TEs/sample_summaries/TEother_merge/E#_15_coreMarks_mnemonics.bed_TEother_merge_state [127 files]
for file in chromHMM_other/*; do awk '{SUM+=$8}END{print SUM}' $file > $file\_state; while read line; do grep $line $file | awk '{SUM+=$8}END{print SUM}' - >> $file\_state; done < chromHMM_states.txt ; done

# Table of number of bases in each state in merged TEs in each sample	
#TE_landscape/chromHMM/mnemonics_TEother_merge_states.txt	
bash combine_states.sh chromHMM_states.txt mnemonics.txt chromHMM_other/*state
```

```{bash feature chromHMM, eval=FALSE}
# Intersect with merged features
#TE_landscape/chromHMM/chromHMM_refseq_features.txt
for file in ~/genic_features/RefSeq/*_merge.txt; do feature=$(basename "$file" _merge.txt); while read line; do bedtools intersect -wo -a $file -b raw_data/chromHMM/$line\_15_coreMarks_mnemonics.bed | awk -v OFS='\t' -v sample=$line -v feature=$feature '{print sample, feature, $0}' - >> chromHMM/Refseq_features/chromHMM_features.txt; done < sample_lists/mnemonics.txt; done

while read line; do bedtools intersect -wo -a ~/genic_features/RefSeq/refseq_intergenic.txt -b raw_data/chromHMM/$line\_15_coreMarks_mnemonics.bed | awk -v OFS='\t' -v sample=$line '{print sample, "intergenic", $0}' - >> chromHMM/Refseq_features/chromHMM_features.txt; done < sample_lists/mnemonics.txt

# Sum overlap
#TE_landscape/chromHMM/Refseq_features/chromHMM_CDS_states.txt
#TE_landscape/chromHMM/Refseq_features/chromHMM_3UTR_states.txt
#TE_landscape/chromHMM/Refseq_features/chromHMM_5UTR_states.txt
#TE_landscape/chromHMM/Refseq_features/chromHMM_exon_states.txt
#TE_landscape/chromHMM/Refseq_features/chromHMM_intron_states.txt
#TE_landscape/chromHMM/Refseq_features/chromHMM_promoter_states.txt
#TE_landscape/chromHMM/Refseq_features/chromHMM_refseq_intergenic_states.txt
#TE_landscape/chromHMM/Refseq_features/chromHMM_repeats_states.txt
awk -v OFS='\t' '{a[$1, $2, $9]+=$10}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], a[i]}}' chromHMM/Refseq_features/chromHMM_features.txt > chromHMM/chromHMM_refseq_features.txt
```

chromHMM misses 2583 bp from all chromosomes and 2417 bp from all but chrY in comparison to chromosome length. All feature bases are annotated, indicating all unannoted bases are intergenic. 

```{bash chromHMM QC, eval=FALSE}
# Difference between hg19 chromosome lengths and chromHMM annotation
awk -v OFS='\t' '{chr[$1]+=$3-$2}END{for(i in chr){print i, chr[i]}}' raw_data/chromHMM/E001_15_coreMarks_mnemonics.bed

# Difference between feature lengths and chromHMM annotation
for file in chromHMM/Refseq_features/intersect/chromHMM_*.bed; do awk -v OFS='\t' -v feature=$(basename $file .bed) '{a[$9]+=$8}END{for(i in a){print feature, a[i]}}' $file | sort | uniq; done < features/features.txt
```

```{bash genome WGBS, eval=FALSE}
# Number of CpGs in each state, all CpGs, each sample
#TE_landscape/WGBS/all_CpG_Meth_states.txt		 
awk '{for (i=4;i<=NF;i++){if($i == -1) miss[i]+=1; else if ($i < 0.3) hypo[i]+=1; else if ($i > 0.7) hyper[i]+=1; else if (($i <= 0.7) && ($i >= 0.3)) inter[i]+=1;}}; END{for (i in hyper) print i" "hypo[i]" "inter[i]" "hyper[i]" "miss[i];}' all_CpG_Meth.bed > all_CpG_Meth_states.txt
```

```{bash TE WGBS, eval=FALSE}
# Number of CpGs in each state, TE CpGs, each sample
#TE_landscape/WGBS/CpG_TE_Meth_states.txt		 
awk '{for (i=4;i<=NF;i++){if($i == -1) miss[i]+=1; else if ($i < 0.3) hypo[i]+=1; else if ($i > 0.7) hyper[i]+=1; else if (($i <= 0.7) && ($i >= 0.3)) inter[i]+=1;}}; END{for (i in hyper) print i" "hypo[i]" "inter[i]" "hyper[i]" "miss[i];}' CpG_TE_Meth.bed > CpG_TE_Meth_states.txt
```

```{bash feature WGBS, eval=FALSE}
# Merged features
for file in ~/genic_features/RefSeq/*_merge.txt; do echo $file; feature=$(basename "$file" _merge.txt); bedtools intersect -a WGBS/all_CpG_Meth.bed -b $file | awk -v OFS='\t' -v feature=$feature '{print feature, $0}' >> WGBS/feature_CpG_Meth.bed; done

bedtools intersect -a WGBS/all_CpG_Meth.bed -b ~/genic_features/RefSeq/refseq_intergenic.txt | awk -v OFS='\t' '{print "intergenic", $0}' >> WGBS/feature_CpG_Meth.bed

# Number of CpGs in each state, CpGs overlapping Refseq features, each sample
awk -v OFS='\t' '{for (i=5;i<=NF;i++){if($i == -1) miss[$1, i]+=1; else if ($i < 0.3) hypo[$1, i]+=1; else if ($i > 0.7) hyper[$1, i]+=1; else if (($i <= 0.7) && ($i >= 0.3)) inter[$1, i]+=1}}END{for(i in hypo){split(i,sep,SUBSEP); print "Hypomethylated", sep[1], sep[2], hypo[i]}; for(i in hyper){split(i,sep,SUBSEP); print "Hypermethylated", sep[1], sep[2], hyper[i]}; for(i in inter){split(i,sep,SUBSEP); print "Intermediate", sep[1], sep[2], inter[i]}; for(i in miss){split(i,sep,SUBSEP); print "Missing", sep[1], sep[2], miss[i]}}' WGBS/feature_CpG_Meth.bed > WGBS/feature_CpG_Meth_states.txt
```

```{bash genome DHS, eval=FALSE}
#TE_landscape/DNase/DNase_stats.txt	

# Number and width of DNase peaks overall
while read line; do wget http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/narrowPeak/$line\.gz; gunzip $line\.gz; echo $line >> DNase_stats.txt; wc -l $line>> DNase_stats.txt; awk '{sum+=$3-$2}END{print sum}' $line >> DNase_stats.txt; done < DNase_peaks.txt

# Peaks overlapping TEs per sample (no summit rule)
while read line; do awk '{print $8, $9, $10}' rmsk_TEother_$line\-DNase.macs2.narrowPeak | sort | uniq | wc -l >> test; done < ../DNase_samples.txt
```

```{bash TE DHS, eval=FALSE}
# Intersect with merged TEs
#TE_landscape/DNase/rmsk_TEother_merge_DNase.txt
while read line; do bedtools intersect -wo -a DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak -b ../rmsk_TEother_merge.txt | awk -v OFS='\t' -v sample=$line '{print $0, sample}' - >> rmsk_TEother_merge_DNase.txt; done < DNase_samples.txt

#TE_landscape/DNase/rmsk_TEother_merge_DNase_contribution.txt
awk -v OFS='\t' '{a[$15]+=$14}END{for(i in a){print i, a[i]}}' rmsk_TEother_merge_DNase.txt | sort > rmsk_TEother_merge_DNase_contribution.txt

# Peaks overlapping TEs per sample, including summit rules
while read line; do awk -v OFS='\t' '{print $8, $9, $10}' DNase/summit/TEs/rmsk_TEother_$line\_DNase_summit.txt | sort | uniq | wc -l >> DNase/rmsk_TEother_DNase_summit_stats.txt; done < sample_lists/DNase_samples.txt
```

```{bash feature DHS, eval=FALSE}
for file in ~/genic_features/RefSeq/*_merge.txt; do feature=$(basename "$file" _merge.txt); echo $feature; while read line; do echo $line; bedtools intersect -wo -a raw_data/DNase/DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak -b $file | awk -v OFS='\t' -v sample=$line -v feature=$feature '{print sample, feature, $0}' - >> DNase/refseq_features_DNase_intersect.txt; done < sample_lists/DNase_samples.txt; done

while read line; do echo $line; bedtools intersect -wo -a raw_data/DNase/DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak -b ~/genic_features/RefSeq/refseq_intergenic.txt | awk -v OFS='\t' -v sample=$line '{print sample, "intergenic", $0}' - >> DNase/refseq_features_DNase_intersect.txt; done < sample_lists/DNase_samples.txt

# Merged features
awk -v OFS='\t' '{a[$1, $2]+=$16}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], a[i]}}' DNase/refseq_features_DNase_intersect.txt > DNase/refseq_features_DNase.txt
```

```{bash genome H3K27ac, eval=FALSE}
#TE_landscape/H3K27ac/H3K27ac_stats.txt	

# Number and width of H3K27ac peaks overall
while read line; do echo $line >> H3K27ac_stats.txt; wc -l H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak >> H3K27ac_stats.txt; awk '{sum+=$3-$2}END{print sum}' H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak >> H3K27ac_stats.txt; done < H3K27ac_samples.txt

# Peaks overlapping TEs per sample (no summit rule)
while read line; do awk '{print $8, $9, $10}' H3K27ac_TEs/rmsk_TEother_$line\-H3K27ac.narrowPeak | sort | uniq | wc -l >> test; done < ../H3K27ac_samples.txt #Combine in Excel
```

```{bash TE H3K27ac, eval=FALSE}
# Intersect with merged TEs
#TE_landscape/H3K27ac/rmsk_TEother_merge_H3K27ac.txt
while read line; do bedtools intersect -wo -a H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak -b ../rmsk_TEother_merge.txt | awk -v OFS='\t' -v sample=$line '{print $0, sample}' - >> rmsk_TEother_merge_H3K27ac.txt; done < H3K27ac_samples.txt

#TE_landscape/H3K27ac/rmsk_TEother_merge_H3K27ac_contribution.txt
awk -v OFS='\t' '{a[$15]+=$14}END{for(i in a){print i, a[i]}}' rmsk_TEother_merge_H3K27ac.txt | sort > rmsk_TEother_merge_H3K27ac_contribution.txt

# Peaks overlapping TEs per sample, including summit rules
while read line; do awk -v OFS='\t' '{print $8, $9, $10}' H3K27ac/summit/TEs/rmsk_TEother_$line\_H3K27ac_summit.txt | sort | uniq | wc -l >> H3K27ac/rmsk_TEother_H3K27ac_summit_stats.txt; done < sample_lists/H3K27ac_samples.txt
```

```{bash feature H3K27ac, eval=FALSE}
for file in ~/genic_features/RefSeq/*_merge.txt; do feature=$(basename "$file" _merge.txt); echo $feature; while read line; do echo $line; bedtools intersect -wo -a raw_data/H3K27ac/H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak -b $file | awk -v OFS='\t' -v sample=$line -v feature=$feature '{print sample, feature, $0}' - >> H3K27ac/refseq_features_H3K27ac_intersect.txt; done < sample_lists/H3K27ac_samples.txt; done

while read line; do echo $line; bedtools intersect -wo -a raw_data/H3K27ac/H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak -b ~/genic_features/RefSeq/refseq_intergenic.txt | awk -v OFS='\t' -v sample=$line '{print sample, "intergenic", $0}' - >> H3K27ac/refseq_features_H3K27ac_intersect.txt; done < sample_lists/H3K27ac_samples.txt

# Merged features
awk -v OFS='\t' '{a[$1, $2]+=$16}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], a[i]}}' H3K27ac/refseq_features_H3K27ac_intersect.txt > H3K27ac/refseq_features_H3K27ac.txt
```

```{bash class chromHMM, eval=FALSE}
# TE merged classes
#TE_landscape/chromHMM/TEs/intersect/class/rmsk_[class].txt_chromHMM.bed [13 files]
for class in TE_classes/rmsk_*.txt; do for file in chromHMM_bedfiles/E*.bed; do suffix=$(basename $file | cut -d '_' -f1); bedtools intersect -wo -a $class -b $file | awk -v OFS='\t' -v tag=$suffix '{print $0, tag}' - >> $class\_chromHMM.bed; done; done

#Unconfident class
for file in ../chromHMM_bedfiles/E*.bed; do suffix=$(basename $file | cut -d '_' -f1); bedtools intersect -wo -a rmsk_Unconfident.txt -b $file | awk -v OFS='\t' -v tag=$suffix '{print $0, tag}' - >> rmsk_Unconfident.txt_chromHMM.bed; done

#TE_landscape/chromHMM/TEs/intersect/class/rmsk_Unconfident_RC.txt_chromHMM.bed
for file in ../../raw_data/chromHMM/E*.bed; do suffix=$(basename $file | cut -d '_' -f1); bedtools intersect -wo -a ../../features/TEs/class/rmsk_Unconfident_RC.txt -b $file | awk -v OFS='\t' -v tag=$suffix '{print $0, tag}' - >> rmsk_Unconfident_RC.txt_chromHMM.bed; done

# Merged TE classes
# Number of bases in each state in merged TE classes in each sample
# Where classes.txt is DNA, LINE, LTR, SINE, Other (SVA), and Unconfident_RC
while read line; do awk -v OFS='\t' -v class=$line '{a[$7,$9]+=$8}END{for(i in a) {split (i, sep, SUBSEP); print class, sep[1], sep[2], a[i];}}' chromHMM/TEs/intersect/class/rmsk_$line\.txt_chromHMM.bed >> chromHMM/class/class_state_sample.txt; done < chromHMM/class/classes.txt
```

```{bash class WGBS, eval=FALSE}
# Number of CpGs in each state, TE CpGs, by sample x class

#TE_landscape/WGBS/class_CpG_Meth_states.txt		 
while read line; do awk -v OFS='\t' -v class=$line '{if($5 == class) print $0}' TE_CpG_Meth_new.bed | cut -f8- - | sort | uniq | awk -v OFS='\t' -v class=$line '{for (i=4;i<=NF;i++){if($i == -1) miss[i]+=1; else if ($i < 0.3) hypo[i]+=1; else if ($i > 0.7) hyper[i]+=1; else if (($i <= 0.7) && ($i >= 0.3)) inter[i]+=1;}}; END{for (i in hyper) print i, hypo[i], inter[i], hyper[i], miss[i], class;}' -; done < ../features/TEs/class/TEother_class.txt >> TE_class_hypo.txt

awk -v OFS='\t' '{if(($5 == "LINE?") || ($5 == "SINE?") || ($5 == "DNA?") || ($5 == "LTR?") || ($5 == "Unknown?") || ($5 == "Unknown")) print $0}' TE_CpG_Meth_new.bed | cut -f8- - | sort | uniq | awk -v OFS='\t' '{for (i=4;i<=NF;i++){if($i == -1) miss[i]+=1; else if ($i < 0.3) hypo[i]+=1; else if ($i > 0.7) hyper[i]+=1; else if (($i <= 0.7) && ($i >= 0.3)) inter[i]+=1;}}; END{for (i in hyper) print i, hypo[i], inter[i], hyper[i], miss[i], "Unconfident";}' - >> TE_class_hypo.txt (class_CpG_Meth_states.txt)

awk -v OFS='\t' '{if(($5 == "LINE?") || ($5 == "SINE?") || ($5 == "DNA?") || ($5 == "LTR?") || ($5 == "Unknown?") || ($5 == "Unknown") || ($5 == "RC")) print $0}' TE_CpG_Meth_new.bed | cut -f8- - | sort | uniq | awk -v OFS='\t' '{for (i=4;i<=NF;i++){if($i == -1) miss[i]+=1; else if ($i < 0.3) hypo[i]+=1; else if ($i > 0.7) hyper[i]+=1; else if (($i <= 0.7) && ($i >= 0.3)) inter[i]+=1;}}; END{for (i=4;i<=NF;i++) {print i, hypo[i], inter[i], hyper[i], miss[i], "Unconfident_RC";}}' - >> class_CpG_Meth_states.txt
```

```{bash class DHS, eval=FALSE}
# Merged TE classes
#TE_landscape/DNase/rmsk_TEother_class_Dnase.txt
while read line; do bedtools intersect -wo -a ../TE_classes/TEother_class_merge.txt -b DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak | awk -v OFS='\t' -v sample=$line '{print $0, sample}' - >> rmsk_TEother_class_Dnase.txt ; done < DNase_samples.txt

while read line; do grep "Unconfident_RC" ../features/TEs/class/TEother_class_merge.txt | bedtools intersect -wo -a - -b ../raw_data/DNase/DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak | awk -v OFS='\t' -v sample=$line '{print $0, sample}' - >> rmsk_TEother_class_Dnase.txt ; done < ../sample_lists/DNase_samples.txt

# Merged TE classes
#TE_landscape/DNase/class_DNase_sample.txt
awk -v OFS='\t' '{a[$4, $16]+=$15}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], a[i];}}' rmsk_TEother_class_Dnase.txt > class_DNase_sample.txt
```

```{bash class H3K27ac, eval=FALSE}
# Merged TE classes
#TE_landscape/H3K27ac/rmsk_TEother_class_H3K27ac.txt
while read line; do bedtools intersect -wo -a ../features/TEs/class/TEother_class_merge.txt -b ../raw_data/H3K27ac/H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak | awk -v OFS='\t' -v sample=$line '{print $0, sample}' - >> rmsk_TEother_class_H3K27ac.txt ; done < ../sample_lists/H3K27ac_samples.txt

# Merged TE classes
#TE_landscape/H3K27ac/class_H3K27ac_sample.txt
awk -v OFS='\t' '{a[$4, $16]+=$15}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], a[i];}}' rmsk_TEother_class_H3K27ac.txt > class_H3K27ac_sample.txt
```

### Figure 1

```{r Figure 1, echo=FALSE}
source("R_scripts/proportion.R")
source("R_scripts/proportion_class.R") 

a = ggplot(combined_proportion[which(combined_proportion$Coding == "All"),],aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + xlab("Feature") + ylab("Proportion in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + scale_x_discrete(labels=c("TE","Genome",genic_labels)) + facet_wrap(~Mark,nrow=1,labeller=labeller(Mark = mark_labels)) 

b = ggplot(combine_class_proportion,aes(x=class,y=Proportion,fill=State)) + geom_bar(stat="identity") + xlab("TE class") + ylab("Proportion in state") + scale_fill_manual(values=all_state_colors,labels=all_state_labels) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.position="bottom",legend.box.margin = margin(t=-12),legend.key.size = unit(3,'mm')) + facet_wrap(~Cohort,nrow=1,labeller=labeller(Cohort = mark_labels)) + guides(fill = guide_legend(ncol = 8,title.position = "top"))

c = ggplot(contribution,aes(x=State,y=Proportion,fill=State)) + geom_bar(stat="identity") + coord_flip() + scale_y_reverse(limits=c(1,0),expand=c(0.01,0.01)) + xlab("State") + ylab("Proportion of state overlapping TEs") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(labels=all_state_labels,limits=c(rev(states[16:19]),"CpGs",rev(states[c(1:15,20:21)]),"Bases")) + geom_text(aes(label=round(Proportion,2)),hjust=1.01,size=3) + theme(axis.title.y = element_text(margin = margin(r = -10))) + geom_vline(xintercept=5.5) + geom_vline(xintercept=4.5,linetype="dashed",color="black") + geom_vline(xintercept=6.5,linetype="dashed",color="grey") + geom_vline(xintercept=7.5,linetype="dashed",color="grey") + geom_vline(xintercept=22.5,linetype="dashed",color="black") 

d = ggplot(contribution_class,aes(x=State,y=Proportion,fill=forcats::fct_rev(class))) + geom_bar(stat="identity") + coord_flip() + ylab("Proportion in each TE class") + theme(axis.title.y=element_blank(),axis.text.y=element_blank(),axis.ticks.y = element_blank()) + scale_x_discrete(limits=c(rev(states[16:19]),"CpGs",rev(states[c(1:15,20:21)]),"Bases")) + scale_fill_manual(values=class_colors,guide=FALSE) + geom_vline(xintercept=5.5) + geom_vline(xintercept=4.5,linetype="dashed",color="black") + geom_vline(xintercept=6.5,linetype="dashed",color="grey") + geom_vline(xintercept=7.5,linetype="dashed",color="grey") + geom_vline(xintercept=22.5,linetype="dashed",color="black") + scale_y_continuous(limits=c(0,1),expand=c(0.01,0.01)) 

legend_class = get_legend(ggplot(contribution_class,aes(x=State,y=Proportion,fill=class)) + geom_bar(stat="identity") + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.text.y=element_blank(),legend.position="bottom",legend.direction = "horizontal",legend.key.size = unit(3,'mm')) + scale_fill_manual(values=class_colors,name="Class") + guides(fill = guide_legend(nrow = 2)))

grid.arrange(a,b,c,d,legend_class, nrow = 4,layout_matrix = rbind(c(1),c(2),c(3,4),c(NA,5)),heights=c(0.25,0.3,0.4,0.05),widths=c(0.55,0.45))
```

### Proportion/contribution analysis

```{r analysis Fig 1}
# Proportion of each feature in each state, across all samples
combined_proportion[which(combined_proportion$Coding == "All"),]

# Specific states, TEs, promoters, and exons
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "TE" & combined_proportion$State %in% chromHMM_states[c(1:3,6:7)]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "TE" & combined_proportion$State %in% chromHMM_states[4:5]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "promoters" & combined_proportion$State %in% chromHMM_states[c(1:3,6:7)]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "exons" & combined_proportion$State %in% chromHMM_states[4:5]),]$Proportion)

# Does any feature have less % in state than TEs?
ddply(combined_proportion[which(combined_proportion$Coding == "All"),],.(State),function(x) x[which(x$Proportion > x[which(x$Feature == "TE"),]$Proportion),c("State","Cohort","Feature","Coding","Proportion")])

# By-class proportion
dcast(combine_class_proportion,class~State,value.var="Proportion")
test = ddply(combine_class_proportion,.(State),function(x) x[which.max(x$Proportion),])
test[order(test$class),]

# Proportion of all genomic bases in TEs
CHROMHMM_TE_WIDTH/CHROMHMM_TOTAL_WIDTH

# Proportion of all/TE CpGs in TE classes
rmsk_TE_class[,c("class_update","CpGs","Percent_TE_CpGs","Percent_all_CpGs")]

# Contribution
contribution

# Contribution, composite states
contribution_composite

# Contribution by class
dcast(contribution_class,class~State,value.var="Proportion")
```

### VISTA enhancers

```{bash count vista enhancers, eval=FALSE}
# Number of positively validated VISTA enhancers from human
wc -l features/vista_enhancers/vista_enhancers_human.bed
# 934

# Proportion overlapping TEs
cat features/intersect_features/vista_enhancers_rmsk_*.txt | awk '{print $1, $2, $3}' - | sort | uniq | wc -l
# 355

# 355/934
```

### Supplementary Figure 1

```{r Figure S1, echo=FALSE}
# Plot bases in genome/TEs in each state by sample, increasing order
all_state_proportion_matrix = dcast(all_state_proportion[which(all_state_proportion$Cohort %in% c("Genome","TE")),],State+Sample+Group~Cohort,value.var="Bases")
all_state_proportion_corr = ddply(all_state_proportion_matrix,.(State),summarise,Cor=as.numeric(unlist(cor.test(Genome,TE,method="spearman"))["estimate.rho"]),Pvalue=as.numeric(unlist(cor.test(Genome,TE,method="spearman"))["p.value"]))

ggplot(all_state_proportion_matrix,aes(x=log10(Genome),y=log10(TE))) + geom_point(aes(color=Group)) + scale_color_manual(values=group_colors) + facet_wrap(~State,scales="free",ncol=6,labeller=labeller(State=all_state_labels)) + xlab("Bases/CpGs in genome (log10)") + ylab("Bases/CpGs in TEs (log10)") + geom_smooth(method="lm",se=FALSE,color="black") + geom_text(data=all_state_proportion_corr,aes(x = -Inf, y = Inf,label=round(Cor,2)),hjust=-0.1,vjust=1.5,size=3) + scale_y_continuous(breaks=pretty_breaks(n=3)) + scale_x_continuous(breaks=pretty_breaks(n=3)) + theme(legend.position="bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm'))
```

### Compare TE bases to non-TE bases by sample

```{r proportion investigation}
ddply(all_state_proportion[which(all_state_proportion$Cohort %in% c("Genome","TE")),],.(State,Cohort),summarise,Median=median(Bases),Mean=mean(Bases),SD=sd(Bases),Min=min(Bases),Max=max(Bases),CV=sd(Bases)/mean(Bases))

ddply(all_state_proportion[which(all_state_proportion$Cohort == "Genome"),],.(State,Cohort),summarise,Median=median(Proportion))

dcast(all_state_proportion[which(all_state_proportion$Mark == "WGBS" & all_state_proportion$Cohort == "Genome"),c("Sample","State","Proportion")],Sample~State,value.var="Proportion")

test = ddply(all_state_proportion[which(all_state_proportion$Feature == "Genome"),],.(State),transform,Mean=mean(Proportion),SD=sd(Proportion))
tail(test[order(abs(test$Proportion-test$Mean)/test$SD),c("Sample","State","Proportion","Mean","SD")],n=15)

test = ddply(all_state_proportion_matrix,.(State,Group),summarise,Genome_median=median(Genome))
test[order(test$State,test$Genome_median),]

# Not MHC corrected (0.05/21 ~ 0.0024)
ddply(all_state_proportion_matrix,.(State),summarise,Genome_corr = unlist(cor.test(Genome-TE,TE,method="spearman"))["estimate.rho"],Genome_pvalue = unlist(cor.test(Genome-TE,TE,method="spearman")["p.value"]))
ddply(all_state_proportion_matrix,.(State),summarise,Genome_corr = unlist(cor.test(Genome,TE/Genome,method="spearman"))["estimate.rho"],Genome_pvalue = unlist(cor.test(Genome,TE/Genome,method="spearman")["p.value"]))
```

### Supplementary Figure 2

```{r Figure S2, echo=FALSE}
source("R_scripts/feature_overlap.R")

# Feature overlap
a = ggplot(feature_overlap_long,aes(x=Feature,y=Percent,fill=Coding)) + geom_bar(position="dodge",stat="identity") + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),legend.key.size = unit(2,'mm')) + scale_x_discrete(labels=genic_labels) + scale_fill_discrete(labels=coding_labels,name="Feature") + facet_wrap(~Measure,nrow=1,labeller=labeller(Measure=overlap_labels)) + xlab("Feature") + ylab("Proportion")

## Average proportion by sample
b = ggplot(ddply(all_state_proportion[which(all_state_proportion$Coding == "All"),],.(Mark,State,Feature),summarise,Proportion=mean(Proportion)),aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + xlab("Feature") + ylab("Proportion in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + scale_x_discrete(labels=c("TE","Genome",genic_labels)) + facet_wrap(~Mark,nrow=1,labeller=labeller(Mark = mark_labels)) 

# Total proportion split by protein-coding/non-coding
c = ggplot(combined_proportion[which(combined_proportion$Feature %in% c("promoters","5UTR","3UTR","exons","introns")),],aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + xlab("Feature") + ylab("Proportion in state") + scale_fill_manual(values=all_state_colors,labels=all_state_labels) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.position="bottom",legend.box.margin = margin(t=-12),legend.key.size = unit(3,'mm')) + scale_x_discrete(labels=c("TE","Genome",genic_labels)) + facet_grid(Coding~Mark,labeller=labeller(Coding=coding_labels,Mark=mark_labels)) + guides(fill = guide_legend(ncol = 8,title.position = "top"))

grid.arrange(a,b,c,nrow=3,heights=c(0.25,0.22,0.53))
```

### RefSeq feature analysis

```{r analysis Fig S2}
# Feature %
feature_overlap_long

# Proportion of each feature in each state, across all samples
combined_proportion
```

### Supplementary Figure 3

```{r Figure S3, echo=FALSE}
a = ggplot(rmsk_TE_class,aes(x=factor(1),y=Total_length,fill=class_update,label=paste(round(Total_length/1000000,1)," (",round(Total_length/sum(Total_length)*100,0),"%) ",sep=""))) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(name="Class",values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank(),plot.title = element_text(size=8)) + geom_text(position = position_fill(vjust = 0.5),size=2) + ggtitle("Total length (Mb)") + scale_y_continuous()

b = ggplot(rmsk_TE_class,aes(x=factor(1),y=Count,fill=class_update,label=paste(format(Count, big.mark=",", scientific=FALSE)," (",round(Count/sum(Count)*100,0),"%) ",sep=""))) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(name="Class",values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank(),plot.title = element_text(size=8)) + geom_text(position = position_fill(vjust = 0.5),size=2) + ggtitle("TEs")

c = ggplot(rmsk_TE_class,aes(x=factor(1),y=Subfamilies,fill=class_update,label=paste(format(Subfamilies, big.mark=",", scientific=FALSE)," (",round(Subfamilies/sum(Subfamilies)*100,0),"%) ",sep=""))) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(name="Class",values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank(),plot.title = element_text(size=8)) + geom_text(position = position_fill(vjust = 0.5),size=2) + ggtitle("Subfamilies")

d = ggplot(rmsk_TE,aes(x=Length/1000,y=..scaled..,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("Length (kbp)") + ylab("Density") + facet_wrap(~class_update,nrow=1) + scale_y_continuous(breaks=c(0,0.5,1))

e = ggplot(rmsk_TE,aes(x=CpGs,y=..scaled..,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("CpGs") + ylab("Density") + facet_wrap(~class_update,nrow=1) + scale_x_continuous(breaks=pretty_breaks(n=4)) + scale_y_continuous(breaks=c(0,0.5,1))

f = ggplot(rmsk_TE,aes(x=CpGs_per_length*1000,y=..scaled..,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE)+ xlab("CpGs/kbp") + ylab("Density") + facet_wrap(~class_update,nrow=1) + scale_x_continuous(breaks=pretty_breaks(n=4)) + scale_y_continuous(breaks=c(0,0.5,1))

g = ggplot(rmsk_TE,aes(x=mappability,y=..scaled..,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE)+ xlab("Mappability") + ylab("Density") + facet_wrap(~class_update,nrow=1) + scale_y_continuous(breaks=c(0,0.5,1)) + scale_x_continuous(breaks=c(0,0.5,1))

h = ggplot(rmsk_TE,aes(x=JC_distance,y=..scaled..,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("Jukes-Cantor evolutionary distance") + ylab("Density") + facet_wrap(~class_update,nrow=1) + scale_x_continuous(breaks=c(0,0.5,1)) + scale_y_continuous(breaks=c(0,0.5,1),limits=c(0,1))

grid.arrange(a,b,c,legend_class,d,e,f,g,h,nrow=7,layout_matrix=rbind(c(1,2,3),c(4),c(5),c(6),c(7),c(8),c(9)),heights=c(0.2,0.05,0.15,0.15,0.15,0.15,0.15))
```

### TE metric stats

```{r metrics analysis, cache=TRUE}
# Median TE metrics by overall and by class, with p-values
p.adjust(apply(rmsk_TE[,measure_metrics],2,function(x) unlist(kruskal.test(x~rmsk_TE$class_update))["p.value"]),method="bonf")

apply(rmsk_TE[,measure_metrics],2,median)
aggregate(data=rmsk_TE[,c("class_update",measure_metrics)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)
sort(apply(aggregate(data=rmsk_TE[,c("class_update",measure_metrics)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)[,2:6],2,function(x) IQR(x)/median(x)))

# Max TE length by class
aggregate(data=rmsk_TE,Length~class_update,max)
```

## Potential

```{bash TE chromHMM counts, eval=FALSE}
# Number of TEs in state in sample
awk -v OFS='\t' '{a[$8,$10]+=1}END{for(i in a){split (i, sep, SUBSEP); print sep[1], sep[2], a[i];}}' chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt > chromHMM/state_sample_counts_summit.txt
```

```{bash class chromHMM counts, eval=FALSE}
# TEs by class
while read line ; do awk -v OFS='\t' -v class=$line '{if($5 == class) print $0}' chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt > chromHMM/chromHMM_summit_$line\.txt; done < features/TEs/class/TE_class.txt
awk -v OFS='\t' '{if($5 == "Other") print $0}' chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt > chromHMM/chromHMM_summit_SVA.txt
awk -v OFS='\t' '{if(($5 != "Other") && ($5 != "DNA") && ($5 != "LINE") && ($5 != "LTR") && ($5 != "SINE")) print $0}' chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt > chromHMM/chromHMM_summit_Other.txt

# By class
for file in chromHMM/chromHMM_summit_*.txt; do awk -v OFS='\t' -v class=$(basename $file .txt) '{a[$8, $10]+=1}END{for(i in a) {split (i, sep, SUBSEP); print class, sep[1], sep[2], a[i];}}' $file >> chromHMM/class_state_sample_summit.txt; done
```

```{r Figure 2 scripts, echo=FALSE, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/potential.R")
source("R_scripts/potential_class.R")
```

### Figure 2

```{r Figure 2, echo=FALSE}
a = ggplot(combine_boxplot,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y = element_text(margin = margin(r = -10))) + labs(y="Proportion of TEs\nin state") + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + geom_point(data=combine_stats,aes(x=State,y=Proportion_ever),color="red",size=2) 

b = ggplot(combine_class_ever,aes(x=State,y=Proportion,fill=forcats::fct_rev(Class))) + geom_bar(stat="identity") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + labs(y="Proportion of TEs\never in state by class") + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + scale_y_continuous(expand=c(0.01,0.01))

c = ggplot(combine_potential[which(combine_potential$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples\nin state") + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + geom_boxplot(width=0.2,outlier.size = 0.1) + scale_y_continuous(expand=c(0.01,0.01))

d = ggplot(combine_potential[which(combine_potential$Sample.Proportion == 1),],aes(x=State,y=log10(Count+1),fill=State)) + geom_bar(stat="identity") + geom_text(aes(label=format(Count, big.mark=",", scientific=FALSE)),hjust=1.01,size=2.5) + scale_y_reverse(limits=c(8,0)) + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y = element_text(margin = margin(r = -10))) + ylab("Number of TEs in state in\n100% of samples (log10)")

e = ggplot(ddply(combine_potential[which(combine_potential$Sample.Proportion >= 0.9),],.(State),summarise,Count=sum(Count)),aes(x=State,y=log10(Count+1),fill=State)) + geom_bar(stat="identity") + geom_text(aes(label=format(Count, big.mark=",", scientific=FALSE)),hjust=0,size=2.5) + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Number of TEs in state in\n90% of samples (log10)") + scale_y_continuous(limits=c(0,8))

f = ggplot(ddply(combine_potential_class[which(combine_potential_class$Sample.Proportion >= 0.9),],.(State,Class),summarise,TEs.Proportion=sum(Count)/sum(Total)),aes(x=State,y=TEs.Proportion,fill=forcats::fct_rev(Class))) + geom_bar(stat="identity") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + labs(y="Proportion of TEs in 90%\nof samples by class") + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + scale_y_continuous(expand=c(0.01,0.01))

grid.arrange(a,b,c,d,e,f,legend_class, nrow = 3, layout_matrix=rbind(c(1,2,3),c(4,5,6),c(NA,NA,7)),widths=c(0.42,0.29,0.29),heights=c(0.475,0.475,0.05))
```

### Potential analysis

```{r analysis Fig 2a, cache=TRUE}
ddply(combine_boxplot,.(State),summarise,Median=median(Proportion),Mean=mean(Proportion))

# Number of TEs ever in composite states
## Roadmap Active (states 1-8)
sum(apply(chromHMM_TE_state[8:15],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Roadmap Inactive (states 9-15)
sum(apply(chromHMM_TE_state[16:22],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Active regulatory states (states 1-3, 6-7)
sum(apply(chromHMM_TE_state[c(8:10,13:14)],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Transcribed states (states 4-5)
sum(apply(chromHMM_TE_state[11:12],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Poised regulatory states (states 10-12)
sum(apply(chromHMM_TE_state[17:19],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Repressed states (states 9, 13-14)
sum(apply(chromHMM_TE_state[c(16,20,21)],1,function(x) sum(x > 0)) > 0)/NUM_TE

# Number of TEs ever in each state
ddply(combine_potential,.(State),summarise,Count=sum(Count[which(Samples != 0)]))
combine_stats
ddply(rmsk_TE_class,.(class_update,Count),summarise,TE_prop=Count/NUM_TE)
dcast(combine_class_ever,Class~State,value.var="Proportion")
combine_potential_class[which(combine_potential_class$Sample.Proportion == 1),]

# Median samples in state (ever)
ddply(combine_potential[which(combine_potential$Samples != 0),],.(State),summarise,first=quantile(rep(Sample.Proportion,times=Count),0.25),Median=median(rep(Sample.Proportion,times=Count)),third=quantile(rep(Sample.Proportion,times=Count),0.75))
dcast(ddply(combine_potential_class[which(combine_potential_class$Samples != 0),],.(State,Class),summarise,Median=median(rep(Sample.Proportion,times=Count))),Class~State,value.var="Median")
```

### TEs in each state in x% of samples

```{r potential percent}
ddply(combine_potential,.(State),summarise,In_100=sum(Count[which(Sample.Proportion == 1)]),
      In_90=sum(Count[which(Sample.Proportion >= 0.9)]),
      In_75=sum(Count[which(Sample.Proportion >= 0.75)]),
      In_50=sum(Count[which(Sample.Proportion >= 0.5)]),
      In_25=sum(Count[which(Sample.Proportion >= 0.25)]))
```

### TEs always in single state

```{r analysis Fig 2c}
always_TEs = c(setNames(lapply(chromHMM_states,function(x) chromHMM_TE_state[which(chromHMM_TE_state[[x]] == sample_counts["All","chromHMM"]),TE_coordinates]),chromHMM_states),
                  setNames(lapply(meth_states,function(x) TE_meth_average[which(TE_meth_average[[x]] == sample_counts["All","WGBS"]),TE_coordinates]),meth_states))
always_TEs$DNase = TE_DNase_peaks[which(TE_DNase_peaks$Samples == sample_counts["All","DNase"]),TE_coordinates]
always_TEs$H3K27ac = TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$Samples == sample_counts["All","H3K27ac"]),TE_coordinates]
always_TEs$RNA = RNA_TE[which(RNA_TE$Expressed_samples == sample_counts["All","RNA"]),TE_coordinates]
always_TEs = ldply(always_TEs)
colnames(always_TEs)[1] = "State"

always_TEs_noY = setNames(lapply(chromHMM_states,function(x) chromHMM_TE_state[which(chromHMM_TE_state$chromosome == "chrY" & chromHMM_TE_state[[x]] == sample_counts["chrY","chromHMM"]),TE_coordinates]),chromHMM_states)
always_TEs_noY$DNase = TE_DNase_peaks[which(TE_DNase_peaks$chromosome == "chrY" & TE_DNase_peaks$Samples == sample_counts["chrY","DNase"]),TE_coordinates]
always_TEs_noY$H3K27ac = TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$chromosome == "chrY" & TE_H3K27ac_peaks$Samples == sample_counts["chrY","H3K27ac"]),TE_coordinates]
always_TEs_noY$RNA = RNA_TE[which(RNA_TE$chromosome == "chrY" & RNA_TE$Expressed_samples == sample_counts["chrY","RNA"]),TE_coordinates]
always_TEs_noY = ldply(always_TEs_noY)
colnames(always_TEs_noY)[1] = "State"

## Summarise
ddply(always_TEs,.(State),summarise,Count=length(chromosome),Proportion=length(chromosome)/NUM_TE,Proportion_CpG=length(chromosome)/NUM_TE_WGBS)
ddply(always_TEs_noY,.(State),summarise,Count=length(chromosome),Proportion=length(chromosome)/NUM_TE,Proportion_CpG=length(chromosome)/NUM_TE_WGBS)

### Unique always in chromHMM state
dim(unique(always_TEs[which(always_TEs$State %in% chromHMM_states),2:8]))

## Genomic locations
always_TEs_location = merge(always_TEs,rmsk_TE[,c(TE_coordinates,"class_update","Length",cohorts)],by=TE_coordinates,all.x=TRUE)
ddply(always_TEs_location,.(State),function(x) apply(x[,cohorts],2,function(y) sum(!is.na(y))))

dim(always_TEs_location[which(always_TEs_location$State == "RNA" & is.na(always_TEs_location$exons) & is.na(always_TEs_location$introns) & is.na(always_TEs_location$promoters)),])

## Phylogeny
ddply(always_TEs_location,.(State,class_update),summarise,Count=length(chromosome))

always_TE_class = ddply(always_TEs_location[,c(TE_coordinates,"class_update","State")],.(State,class_update),summarise,Count=length(chromosome))
always_TE_class = merge(rmsk_TE_class[,c("class_update","Count")],always_TE_class,by="class_update",all.y=TRUE)
always_TE_class[is.na(always_TE_class)] = 0
colnames(always_TE_class)[c(2,4)] = c("Count.Class","Count")
always_TE_class = ddply(always_TE_class,.(State),transform,Count.State=sum(Count))
always_TE_class$Enrichment = log2((always_TE_class$Count/always_TE_class$Count.State)/(always_TE_class$Count.Class/NUM_TE))
always_TE_class = always_TE_class[order(always_TE_class$Enrichment),]
dim(always_TE_class[which(always_TE_class$Enrichment > 0),])
always_TE_class[which(always_TE_class$Enrichment > 5),]

always_TE_family = ddply(always_TEs_location[,c(TE_coordinates,"class_update","State")],.(State,class_update,family),summarise,Count=length(chromosome))
always_TE_family = merge(ddply(rmsk_TE_subfamily,.(family,class_update),summarise,Count=sum(Count)),always_TE_family,by=c("family","class_update"),all.y=TRUE)
always_TE_family[is.na(always_TE_family)] = 0
colnames(always_TE_family)[c(3,5)] = c("Count.Family","Count")
always_TE_family = ddply(always_TE_family,.(State),transform,Count.State=sum(Count))
always_TE_family$Enrichment = log2((always_TE_family$Count/always_TE_family$Count.State)/(always_TE_family$Count.Family/NUM_TE))
always_TE_family = always_TE_family[order(always_TE_family$Enrichment),]
dim(always_TE_family[which(always_TE_family$Enrichment > 0),])
always_TE_family[which(always_TE_family$Enrichment > 0.5 & always_TE_family$Count > 10),]

always_TE_subfamily = ddply(always_TEs_location[,c(TE_coordinates,"class_update","State")],.(State,class_update,family,subfamily),summarise,Count=length(chromosome))
always_TE_subfamily = merge(rmsk_TE_subfamily[,c("subfamily","family","class_update","Count")],always_TE_subfamily,by=c("subfamily","family","class_update"),all.y=TRUE)
always_TE_subfamily[is.na(always_TE_subfamily)] = 0
colnames(always_TE_subfamily)[c(4,6)] = c("Count.Subfamily","Count")
always_TE_subfamily = ddply(always_TE_subfamily,.(State),transform,Count.State=sum(Count))
always_TE_subfamily$Enrichment = log2((always_TE_subfamily$Count/always_TE_subfamily$Count.State)/(always_TE_subfamily$Count.Subfamily/NUM_TE))
dim(always_TE_subfamily[which(always_TE_subfamily$Enrichment > 0),])
always_TE_subfamily = always_TE_subfamily[order(always_TE_subfamily$State,always_TE_subfamily$Enrichment),]
always_TE_subfamily[which(always_TE_subfamily$Enrichment > 0.5 & always_TE_subfamily$Count > 10),]

## Print out TEs
write.table(always_TEs[,c(TE_coordinates,"State")],file="always_TEs.bed",sep='\t',row.names=FALSE,quote=FALSE,col.names = FALSE)

# 90% TEs
always_TEs_90 = c(setNames(lapply(chromHMM_states,function(x) chromHMM_TE_state[which(chromHMM_TE_state[[x]] >= sample_counts["All","chromHMM"]*.9),TE_coordinates]),chromHMM_states),
                  setNames(lapply(meth_states,function(x) TE_meth_average[which(TE_meth_average[[x]] >= sample_counts["All","WGBS"]*.9),TE_coordinates]),meth_states))
always_TEs_90$DNase = TE_DNase_peaks[which(TE_DNase_peaks$Samples >= sample_counts["All","DNase"]*.9),TE_coordinates]
always_TEs_90$H3K27ac = TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$Samples >= sample_counts["All","H3K27ac"]*.9),TE_coordinates]
always_TEs_90$RNA = RNA_TE[which(RNA_TE$Expressed_samples >= sample_counts["All","RNA"]*.9),TE_coordinates]
always_TEs_90 = ldply(always_TEs_90)
colnames(always_TEs_90)[1] = "State"

## Print out TEs
write.table(always_TEs_90[,c(TE_coordinates,"State")],file="always_TEs_90.bed",sep='\t',row.names=FALSE,quote=FALSE,col.names = FALSE)

# Proportion in each state
dcast(ddply(combine_potential_class[which(combine_potential_class$Sample.Proportion >= 0.9),],.(State,Class),summarise,TEs.Proportion=sum(Count)/sum(Total)),State~Class,value.var="TEs.Proportion")
```

```{bash investigate always TEs, eval=FALSE}
# TEs always in a state

# Find the nearest gene
sort -k1,1 -k2,2n always_TEs.bed > always_TEs_sorted.bed
sort -k1,1 -k2,2n ~/genic_features/RefSeq/refseq_genes.txt | bedtools closest -a always_TEs_sorted.bed -b - -D b -t all > always_TEs_genes.txt

# Find the distance to the nearest gene
awk -v FS='\t' '{if(($8 == "DNase") && ($21 > 5000)) print $0}' always_TEs_genes.txt | wc -l
```

### Background proportion of genome in active regulatory state

Proportion of the genome that is in an active regulatory state in any Roadmap sample. 

From the 200bp window-chromHMM intersections, I pulled out those in the 5 active regulatory states. Then, I found the number of unique windows that are ever in one of those states. There are 6,644,958, 43% of all windows.  

```{bash window potential, eval=FALSE}
# 200bp windows in an active regulatory state
while read line; do awk -v OFS='\t' -v sample=$line '{if(($4 == "1_TssA") || ($4 == "2_TssAFlnk") || ($4 == "3_TxFlnk") || ($4 == "6_EnhG") || ($4 == "7_Enh")) print $5, $6, $7, $4, sample}' chromHMM/genome/windows/windows_$line\.bed >> chromHMM/genome/windows/windows_active_reg.bed; done < sample_lists/mnemonics.txt

# Number of unique windows that are ever in an active regulatory state
cut -f1-3 chromHMM/genome/windows/windows_active_reg.bed | sort | uniq | wc -l
```

Because these windows are smaller than TEs, I will confirm the results using the shuffled TEs. Found the number of shuffled TEs (times 10 iterations) that are ever in an active regulatory state. The mean proportion is 48.66% +/- 0.03%, a very tight distribution that is slightly higher than the value for actual TEs. 

```{bash shuffled potential, eval=FALSE}
# Number of shuffled TEs per iteration that are ever in an active regulatory state
for i in {1..10}; do tail -n +2 chromHMM/shuffled_TEs/rmsk_TE_shuffle_$i\_chromHMM_potential.txt | awk '{if(($8 > 0) || ($9 > 0) || ($10 > 0) || ($13 > 0) || ($14 > 0)) print $0}' | wc -l; done
```

### Supplementary Figure 4 (block/summit and 18-state)

'''Block/summit'''

How many TEs are responsible for their chromHMM region? Found TEs that overlap the center of chromHMM blocks.

```{bash block control, eval=FALSE}
# Filtered TE x chromHMM state intersections to only those where the TE overlaps the center of the block
while read line; do awk -v OFS='\t' -v sample=$line '{mid=$9+(($10-$9)/2)-0.5;if((mid >= $2) && (mid < $3)) print $0, sample}' chromHMM/TEs/intersect/$line\_15_coreMarks_mnemonics.bed_TE >> chromHMM/blocks/rmsk_TE_block_summit.txt ; done < sample_lists/mnemonics.txt
while read line; do awk -v OFS='\t' -v sample=$line '{mid=$9+(($10-$9)/2)-0.5;if((mid >= $2) && (mid < $3)) print $0, sample}' chromHMM/TEs/intersect/$line\_15_coreMarks_mnemonics.bed_other >> chromHMM/blocks/rmsk_TE_block_summit.txt ; done < sample_lists/mnemonics.txt

# Unique
awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7, $11, $13]+=$12}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sep[8], sep[9], a[i]}}' chromHMM/rmsk_TE_block_summit.txt | sort -k1,1V -k2,2n -k3,3n -k4,4 -k9,9 - > chromHMM/rmsk_TE_block_summit_sorted.txt

# Number of TEs in each state per sample
awk -v OFS='\t' '{a[$8, $9]+=1}END{for(i in a) {split(i, sep, SUBSEP); print sep[1], sep[2], a[i]}}' chromHMM/block/rmsk_TE_block_summit_sorted.txt > chromHMM/block/rmsk_TE_block_summit_counts.txt

# Number of samples each TE is in each state
awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7, $8]+=1}END{for(i in a) {split(i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sep[8], a[i]}}' chromHMM/rmsk_TE_block_summit_sorted.txt > chromHMM/rmsk_TE_block_potential.txt
```

As a QC, found the number of TEs per state that are identified using block summit overlap, but not our usual analysis, and compared it to the total number identified using block summit overlap. The proportion is small, usually <1%. There are no subfamily x state combinations that are present in block overlap but not normally.

```{bash block QC, eval=FALSE}
# TEs overlapping a block summit that are not counted in the original analysis
awk -v OFS='\t' '{if($8 != "8_ZNF/Rpts") print $1, $2, $3, $4, $5, $6, $7, $9, $10, $8 > $4"_"$8".txt"; else print $1, $2, $3, $4, $5, $6, $7, $9, $10, $8 > $4"_8_ZNF.Rpts.txt"}' rmsk_TE_block_summit_sorted.txt

while read state; do while read subfam; do echo $subfam $state; comm -23 <(sort chromHMM/block/$subfam\_$state\.txt) <(cut -f1-10 chromHMM/subfamily/by_state/$subfam\_$state\.txt | sort) >> chromHMM/block/block_exclusive.txt; done < features/TEs/subfamily/subfamilies.txt; done < chromHMM/chromHMM_states.txt
while read subfam; do echo $subfam; comm -23 <(sort chromHMM/block/$subfam\_8_ZNF.Rpts.txt) <(cut -f1-10 chromHMM/subfamily/by_state/$subfam\_8_ZNF.Rpts.txt | sort) >> chromHMM/block/block_exclusive.txt; done < features/TEs/subfamily/subfamilies.txt

# Subfamily x state combinations present using block overlap but not normally
while read state; do while read subfam; do if [ ! -f chromHMM/subfamily/by_state/$subfam\_$state\.txt ]; then if [ -f chromHMM/block/$subfam\_$state\.txt ]; then echo $subfam $state; fi; fi; done < features/TEs/subfamily/subfamilies.txt; done < chromHMM/chromHMM_states.txt >> chromHMM/block/not_all.txt
while read subfam; do if [ ! -f chromHMM/subfamily/by_state/$subfam\_8_ZNF.Rpts.txt ]; then if [ -f chromHMM/block/$subfam\_8_ZNF.Rpts.txt ]; then echo $subfam; fi; fi; done < features/TEs/subfamily/subfamilies.txt

# Number of TEs per state identified using block overlap but not normally vs. total
awk -v OFS='\t' '{a[$10]+=1}END{for(i in a){print i, a[i]}}' chromHMM/block/block_exclusive.txt
awk -v OFS='\t' '{a[$8]+=1}END{for(i in a){print i, a[i]}}' chromHMM/block/rmsk_TE_block_summit_sorted.txt
```

As an intermediate between the chromHMM overlap rules in the paper and block center overlap, I repeated the analysis with just TEs that overlap 200bp bin summits. 

```{bash summit control, eval=FALSE}
# TEs per sample, restricted to those overlapping summits
awk -v OFS='\t' '{if($11 == "summit") a[$8,$10]+=1}END{for(i in a){split (i, sep, SUBSEP); print sep[1], sep[2], a[i];}}' chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt > chromHMM/state_sample_counts_summit_only.txt
```

The number of TEs in each state per sample is much smaller, with a median of 17% for 1_TssA (up to 43% for 11_BivFlnk).

For 1_TssA and 7_Enh, >50% of the TEs identified as ever in the state using the rules in the paper are also identified when I just consider block centers, although this number is as low as 29% for 14_ReprPCWk and 36% for 3_TxFlnk. Overall, 26% of TEs can be in an active regulatory state in any sample, about half of the value in the paper. 

```{r block analysis}
source("R_scripts/chromHMM_block_potential.R")

# Median samples in state
test2 = ddply(test,.(State,Category),summarise,Median=median(na.omit(Ratio)))
test2[order(test2$Median),]

# Ever
potential_ever[order(potential_ever$Ratio),]

# TEs in any active regulatory state in any sample
## Overlapping block centers
dim(unique(block_potential[which(block_potential$State %in% chromHMM_states[c(1:3,6:7)]),TE_coordinates]))[1]/NUM_TE

## Overlapping bin summits
sum(apply(summit_potential[c(8:10,13:14)],1,function(x) sum(x > 0)) > 0)/NUM_TE
```

'''18-state model'''

The Roadmap project has 98 samples annotated with the [https://egg2.wustl.edu/roadmap/web_portal/chr_state_learning.html#exp_18state](18-state model) (which includes H3K27ac). 15,478,375 windows have annotations for most samples (the same as for the 15-state model), 15181508 for the samples without chrY. The annotations do not go to the end of the chromosomes.

```{bash 18 state, eval=FALSE}
# Intersect 200bp windows with chromHMM, 18-state model
while read line; do bedtools intersect -wo -a raw_data/chromHMM_18state/$line\_18_core_K27ac_mnemonics.bed -b features/hg19_standard.windows | awk -v OFS='\t' '{print $5, $6, $7, $4}' - | sort -k1,1 -k2,2n - > chromHMM/genome/windows/windows_18state_$line\.bed; done < sample_lists/H3K27ac_samples.txt

# Summit TEs
while read line; do bedtools intersect -wo -sorted -a features/TEs/rmsk_TEother_summit.txt -b chromHMM/genome/windows/windows_18state_$line\.bed > /scratch/ecp/TE_landscape/state18/rmsk_TEother_summit_$line\.txt; done < sample_lists/H3K27ac_samples.txt
for file in rmsk_TEother_summit_E*.txt; do awk -v OFS='\t' '{if(($9+99.5 >= $2) && ($9+99.5 < $3)) a[$1, $2, $3, $4, $5, $6, $7, $11]+=$12}END{for(i in a) {split(i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sep[8], a[i]}}' $file > $(basename $file .txt); done
for file in rmsk_TEother_summit_E[0-9][0-9][0-9] ; do mv $file $file\.txt; done

# Majority TEs
while read line; do bedtools intersect -wo -a features/TEs/rmsk_TEother_majority.txt -b raw_data/chromHMM_18state/$line\_18_core_K27ac_mnemonics.bed > /scratch/ecp/TE_landscape/state18/rmsk_TEother_majority_$line\.txt; done < sample_lists/H3K27ac_samples.txt
for file in rmsk_TEother_majority_*.txt; do python ~/bin/TE_landscape/pick_majority.py $file $(basename $file .txt) 10 11; done
for file in rmsk_TEother_majority_E[0-9][0-9][0-9]; do mv $file $file\.txt; done

# Combine and sort
while read line; do awk -v OFS='\t' -v sample=$line '{print $0, sample, "summit"}' rmsk_TEother_summit_$line\.txt >> rmsk_TEother_18state.txt; awk -v OFS='\t' -v sample=$line '{print $0, sample, "majority"}' rmsk_TEother_majority_$line\.txt >> rmsk_TEother_18state.txt; done < ~/TE_landscape/sample_lists/H3K27ac_samples.txt &
awk '{print>$1}' rmsk_TEother_18state.txt
while read line; do sort -k1,1V -k2,2n -k3,3n -k4,4 -k10,10 $line >> rmsk_TEother_18state_sorted.txt; rm $line; done < chromosomes.txt
```

```{bash 18 state potential, eval=FALSE}
# List of states in 18-state model
sample_lists/chromHMM_18_states.txt

# Potential
python ~/bin/TE_landscape/potential.py chromHMM/rmsk_TEother_18state_sorted.txt features/TEs/rmsk_TEother.txt sample_lists/chromHMM_18_states.txt sample_lists/H3K27ac_samples.txt chromHMM/potential/rmsk_TEother_18state_potential.txt 0 9 7 8

# TEs in state per sample
awk -v OFS='\t' '{a[$8,$10]+=1}END{for(i in a){split (i, sep, SUBSEP); print sep[1], sep[2], a[i];}}' chromHMM/rmsk_TEother_18state_sorted.txt > chromHMM/state_sample_counts_18state.txt
```

```{r 18 state potential load}
# Load matrix
chromHMM_18state = read.table("chromHMM/potential/rmsk_TEother_18state_potential.txt",sep='\t',header=TRUE,quote="")
colnames(chromHMM_18state)[8:25] = names(chromHMM_states_18)

# Potential
## Distribution
chromHMM_18state_dist = sample_distribution(chromHMM_18state,c(8:25),sample_counts["All","H3K27ac"])
chromHMM_18state_potential = melt(chromHMM_18state_dist,id.vars="Samples")
colnames(chromHMM_18state_potential) = c("Samples","State","Count")
chromHMM_18state_potential = ddply(chromHMM_18state_potential,.(State),transform,Sample.Proportion = Samples/(length(Samples)-1))

## Stats
chromHMM_18state_stats = potential_stats(chromHMM_18state_dist,18,sample_counts["All","H3K27ac"])
chromHMM_18state_stats$State = factor(rownames(chromHMM_18state_stats),levels=names(chromHMM_states_18))

## Number of TEs in each chromHMM state by sample
state_sample_18state = read.table("chromHMM/state_sample_counts_18state.txt",sep='\t',quote="")
colnames(state_sample_18state) = c("State","Sample","Count")
state_sample_18state$Count = as.numeric(state_sample_18state$Count)
state_sample_18state$Proportion = ifelse(metadata[match(state_sample_18state$Sample,metadata$Sample),]$chrY == "Yes",state_sample_18state$Count/NUM_TE,state_sample_18state$Count/NUM_TE_noY)
state_sample_18state$State = factor(state_sample_18state$State,levels=names(chromHMM_states_18))
```

The number of TEs in the quiescent state is back down, to a median of 62% per sample. Additionally, the number of TEs ever in an active regulatory state is back up to 48%, despite losing ~30 samples! Specifically, 39% can be in the 11_EnhWk state (weak enhancer), 25% in 10_EnhA2 (active enhancer), and 17% in 9_EnhA1 (active enhancer), although only 2% are ever in the 1_TssA state. 73% can be transcribed - very similar to what we see with the 15-state model.

```{r 18 state potential analysis, cache=TRUE}
ddply(state_sample_18state,.(State),summarise,Median=median(Proportion))

## Active regulatory states (states 1-4, 7-11)
sum(apply(chromHMM_18state[c(8:11,14:18)],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Transcribed states (states 5-6)
sum(apply(chromHMM_18state[12:13],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Poised regulatory states (states 14-15)
sum(apply(chromHMM_18state[21:22],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Repressed states (states 13, 16-17)
sum(apply(chromHMM_18state[c(20,23:24)],1,function(x) sum(x > 0)) > 0)/NUM_TE

ddply(chromHMM_18state_potential,.(State),summarise,Count=sum(Count[which(Samples != 0)]))
chromHMM_18state_stats
```

Block/summit analysis and potential analysis (Fig 2A/C) using 18-state chromHMM model.  

```{r Figure S4, echo=FALSE}
# Plot (x-axis in increasing order of median block size)
a = ggplot(test,aes(x=State,y=Ratio,fill=Category)) + geom_boxplot() + ylim(0,1) + 
  ylab("TEs in state per sample\n(proportion of Fig. 2a TEs)") + scale_x_discrete(limits=chromHMM_states[rev(c(2,10:12,3,6:8,1,13,9,4:5,14:15))]) + 
  geom_hline(yintercept = 0.82,linetype="dashed") + scale_fill_discrete(guide=FALSE) + coord_flip()

b = ggplot(potential_ever,aes(x=State,y=Ratio,color=Category)) + geom_point(size=4) + ylim(0,1) +
  theme(axis.text.y = element_blank(),axis.title.y = element_blank(),axis.ticks.y = element_blank()) + 
  ylab("TEs ever in state\n(proportion of Fig. 2a TEs)") + scale_x_discrete(limits=chromHMM_states[rev(c(2,10:12,3,6:8,1,13,9,4:5,14:15))]) + 
  geom_hline(yintercept = 0.82,linetype="dashed") + scale_color_discrete(labels=setNames(c("Center of chromHMM block","Center of 200bp bin"),c("Block","Summit")),guide=FALSE) + coord_flip()

category_legend = get_legend(ggplot(potential_ever,aes(x=State,y=Ratio,color=Category)) + geom_point(size=4) + ylim(0,1) +
  theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.position = "bottom") + 
  ylab("TEs ever in state\n(proportion of Fig. 2a TEs)") + scale_x_discrete(limits=chromHMM_states[c(2,10:12,3,6:8,1,13,9,4:5,14:15)]) + 
  geom_hline(yintercept = 0.82,linetype="dashed") + scale_color_discrete(labels=setNames(c("Center of chromHMM block","Center of 200bp bin"),c("Block","Summit")),name="Overlap") + coord_flip())

c = ggplot(state_sample_18state,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=chromHMM_states_18,guide=FALSE) + scale_x_discrete(limits=rev(names(chromHMM_states_18))) + theme(axis.title.y = element_text(margin = margin(r = -5))) + labs(y="Proportion of TEs in state") + ylim(0,1) + coord_flip() + geom_point(data=chromHMM_18state_stats,aes(x=State,y=Proportion_ever),color="red",size=2) + scale_y_continuous(limits=c(0,1),expand=c(0.01,0.01))

d = ggplot(chromHMM_18state_potential[which(chromHMM_18state_potential$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=chromHMM_states_18,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits=rev(names(chromHMM_states_18))) + geom_boxplot(width=0.2,outlier.size = 0.1) + scale_y_continuous(expand=c(0.01,0.01))

grid.arrange(a,b,category_legend,c,d,nrow=3,layout_matrix=rbind(c(1,2),c(3),c(4,5)),heights=c(0.5,0.05,0.45),widths=c(0.55,0.45))
```

### Supplementary Figure 5

```{r Figure S5, echo=FALSE}
a = ggplot(combine_boxplot_class,aes(x=State,y=Proportion,fill=Class)) + geom_boxplot() + scale_fill_manual(values=class_colors,guide=FALSE) + coord_flip() + ylab("Proportion of TEs in state") + scale_x_discrete(labels=rev(all_state_labels),limits=rev(levels(combine_boxplot_class$State))) + geom_point(data=combine_stats_class,aes(x=State,y=Proportion_ever,color=Class),position=position_dodge(width=0.5)) + scale_color_manual(values=class_colors) + theme(legend.position="bottom",axis.title.y = element_text(margin = margin(r = -5)),legend.key.size = unit(1,'mm'),legend.box.margin = margin(t=-10)) + guides(color = guide_legend(nrow = 2)) 

state_order = ddply(combine_stats_class,"State",summarise,CV=sd(Samples_avg_all)/mean(Samples_avg_all), Range=max(Samples_avg_all)-min(Samples_avg_all))

rmsk_TE_class_feature = melt(rmsk_TE_class[,c("class_update",cohorts[1:19])],id.vars="class_update")
colnames(rmsk_TE_class_feature) = c("Class","Cohort","Proportion")
rmsk_TE_class_feature = split_coding(rmsk_TE_class_feature)

b = ggplot(rmsk_TE_class_feature,aes(x=Class,y=Proportion,fill=Feature)) + geom_bar(stat="identity") + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.position="bottom",legend.box.margin = margin(t=-10),legend.key.size = unit(3,'mm')) + xlab("TE class") + ylab("Proportion of TEs overlapping feature") + scale_fill_manual(values=c(brewer.pal(8,"Spectral")),labels=genic_labels[c(8,1:7)],name="Feature") + facet_wrap(~Coding,ncol=1,labeller=labeller(Coding=coding_labels),scales="free_y") + guides(fill=guide_legend(title.position = "top"))

c = ggplot(combine_potential_class[which(combine_potential_class$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors,guide=FALSE) + xlab("State") + ylab("Proportion of samples in state") + scale_x_discrete(labels=rev(all_state_labels),limits = rev(states)) + facet_wrap(~Class,nrow=1) + geom_boxplot(width=0.2,outlier.size = 0.1)  + theme(axis.title.y = element_text(margin = margin(r = -10))) + scale_y_continuous(breaks=pretty_breaks(n=3),expand=c(0.01,0.01)) 

grid.arrange(a,b,c,nrow=2,layout_matrix=rbind(c(1,2),c(3)),heights=c(0.6,0.4),widths=c(0.58,0.42))
```

### Class potential analysis

```{r class potential analysis}
combine_stats_class
state_order
```

### Feature overlap and CpGs per class

```{r analysis Fig S4}
# Any TEs not in a feature? Number of features each TE overlaps
table(apply(rmsk_TE[,cohorts[2:19]],1,function(x) sum(!is.na(x))))
table(apply(rmsk_TE[,c(features[2:3],"coding_exon",features[5:8])],1,function(x) sum(!is.na(x))))

# Number/proportion of TEs in each feature, overall and by class
feature_proportion = apply(rmsk_TE[,cohorts[1:19]],2,function(x) length(na.omit(x)))
feature_proportion/NUM_TE
rmsk_TE_class[,c("class_update",cohorts[1:20])]

# TEs with CpGs per class
rmsk_TE_class[,c("class_update","Count","TEs_wCpG","TEs_wCpG_per")]

# Median CpGs per TE by class
median(rmsk_TE$CpGs)
ddply(rmsk_TE,.(class_update),summarise,Median_CpGs=median(CpGs),Median_CpGs_wCpGs=median(CpGs[which(CpGs > 0)]))
```

### Supplementary Figure 6

```{bash promoter chromHMM counts, eval=FALSE}
# Number of promoters in state in sample
awk -v OFS='\t' '{if($1 !~ /_/) a[$5, $7]+=1}END{for(i in a){split (i, sep, SUBSEP); print sep[1], sep[2], a[i];}}' chromHMM/refseq_promoters_unique_chromHMM_summit_sorted.txt > chromHMM/promoters_state_sample_counts_summit.txt
```

```{r Figure S6 scripts, echo=FALSE, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/potential_promoter.R")
source("R_scripts/potential_exon.R")
```

```{r Figure S6, echo=FALSE}
# Promoters
a = ggplot(combine_boxplot_prom,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors[1:21],guide=FALSE) + theme(axis.title.y = element_text(margin = margin(r = -10))) + labs(y="Proportion of promoters in state") + coord_flip() + scale_x_discrete(limits=rev(states[1:21]),labels=all_state_labels[1:21]) + geom_point(data=combine_stats_prom,aes(x=State,y=Proportion_ever),color="red",size=2) + scale_y_continuous(limits=c(0,1),expand=c(0.05,0.05))

b = ggplot(combine_potential_prom[which(combine_potential_prom$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors[1:21],guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states[1:21]),labels=all_state_labels[1:21]) + geom_boxplot(width=0.2,outlier.size = 0.1) + scale_y_continuous(expand=c(0.01,0.01))

# Exons
c = ggplot(RNA_exon_sample,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values="black",guide=FALSE) + labs(y="Proportion of exons in state") + coord_flip() + scale_x_discrete(labels="       RPKM > 1") + geom_point(data=RNA_potential_exon_stats,aes(x=State,y=Proportion_ever),color="red",size=3) + scale_y_continuous(limits=c(0,1),expand=c(0.05,0.05))

d = ggplot(RNA_potential_exon_long[which(RNA_potential_exon_long$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values="black",guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples RPKM > 1") + scale_x_discrete(labels=all_state_labels[22]) + geom_boxplot(width=0.2,outlier.size = 0.1) + scale_y_continuous(expand=c(0.01,0.01))

# No cancer/IMR90
e = ggplot(combine_boxplot_noCancer_IMR90,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y = element_text(margin = margin(r = -10))) + labs(y="Proportion of TEs in state") + coord_flip() + scale_x_discrete(limits = rev(levels(combine_boxplot_noCancer_IMR90$State)),labels=rev(all_state_labels)) + geom_point(data=combine_stats_noCancer_IMR90,aes(x=State,y=Proportion_ever),color="red",size=2) + scale_y_continuous(limits=c(0,1),expand=c(0.05,0.05))

f = ggplot(combine_potential_noCancer[which(combine_potential_noCancer$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + geom_boxplot(width=0.2,outlier.size = 0.1) + scale_y_continuous(expand=c(0.01,0.01))

grid.arrange(a,b,c,d,e,f,nrow = 3,layout_matrix=rbind(c(1,2),c(3,4),c(5,6)),heights=c(0.45,0.1,0.45),widths=c(0.6,0.4))
```

### Promoter statistics

```{bash promoter CpG count, eval=FALSE}
# Refseq promoters
#TE_landscape/WGBS/Refseq_promoters/refseq_promoter_unique_CpG_count.txt
awk -v OFS='\t' '{a[$1, $2, $3, $4]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], a[i];}}' refseq_promoter_unique_CpG_Meth.bed > refseq_promoter_unique_CpG_count.txt
```

```{r promoter analysis}
# Number of unique promoters
NUM_PROMOTER

combine_stats_prom
ddply(combine_boxplot_prom,.(State),summarise,Mean=mean(Proportion),Median=median(Proportion))

# Promoter CpG stats
promoter_CpG_count = read.table("WGBS/Refseq_promoters/refseq_promoter_unique_CpG_count.txt",sep='\t')
promoter_CpG_count$V5 = promoter_CpG_count$V5/2
dim(promoter_CpG_count)[1] # Number of promoters with CpGs
dim(promoter_CpG_count)[1]/NUM_PROMOTER # % of promoters with CpGs
ddply(promoter_CpG_count,.(),summarise,Min=min(V5),Median=median(V5),Mean=mean(V5),Max=max(V5))
mean(c(promoter_CpG_count$V5,rep(0,(NUM_PROMOTER-dim(promoter_CpG_count)[1]))))
median(c(promoter_CpG_count$V5,rep(0,(NUM_PROMOTER-dim(promoter_CpG_count)[1]))))

# % in state in 90% of samples
ddply(combine_potential_prom,.(State),summarise,Total=sum(Count[which(Sample.Proportion >= 0.9)])/NUM_PROMOTER)
```

### Exon statistics (expression potential control)

```{bash exon CpG count, eval=FALSE}
# Refseq exons
split -l 1000000 ~/TE_landscape/WGBS/all_CpG_Meth.bed
for file in x*; do echo $file; bedtools intersect -wo -a ~/genic_features/RefSeq/refseq_exons_unique.txt -b $file >> ~/TE_landscape/WGBS/refseq_exons_unique_CpG_Meth.bed ; done

# Exons (1/3/18)
awk -v OFS='\t' '{a[$1, $2, $3, $4]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], a[i];}}' WGBS/refseq_exons_unique_CpG_Meth.bed > WGBS/refseq_exons_unique_CpG_count.txt
```

```{r exon control analysis, cache=TRUE}
# Number of unique exons
NUM_EXON

RNA_potential_exon_stats
mean(RNA_exon_sample$Proportion)
median(RNA_exon_sample$Proportion)

# Exon CpG stats
exon_CpG_count = read.table("WGBS/refseq_exons_unique_CpG_count.txt",sep='\t')
exon_CpG_count$V5 = exon_CpG_count$V5/2
dim(exon_CpG_count)[1] # Number of exons with CpGs
dim(exon_CpG_count)[1]/NUM_EXON # % of exons with CpGs
ddply(exon_CpG_count,.(),summarise,Min=min(V5),Median=median(V5),Mean=mean(V5),Max=max(V5))
mean(c(exon_CpG_count$V5,rep(0,(NUM_EXON-dim(exon_CpG_count)[1]))))
median(c(exon_CpG_count$V5,rep(0,(NUM_EXON-dim(exon_CpG_count)[1]))))

# Number of samples expressed
table(RNA_TE$Expressed_samples)
table(RNA_TE$Expressed_samples)/dim(RNA_TE)[1]

table(RNA_refseq_exon$Expressed_samples)
table(RNA_refseq_exon$Expressed_samples)/NUM_EXON

# % in state in 90% of samples
ddply(RNA_potential_exon_long,.(State),summarise,Total=sum(Count[which(Sample.Proportion >= 0.9)])/NUM_EXON)

# Maximum expression
median(RNA_TE$Max_expression)
median(RNA_TE[which(RNA_TE$Expressed_samples > 0),]$Max_expression)
max(RNA_TE$Max_expression)
tail(RNA_TE[order(RNA_TE$Max_expression),],n=20)

median(RNA_refseq_exon$Max_expression)
median(RNA_refseq_exon[which(RNA_refseq_exon$Expressed_samples > 0),]$Max_expression)
max(RNA_refseq_exon$Max_expression)
tail(RNA_refseq_exon[order(RNA_refseq_exon$Max_expression),],n=20)

cor.test(RNA_TE$Max_expression,RNA_TE$Expressed_samples,method="spearman")

# Max expression by class
merge(ddply(RNA_TE,.(class_update),summarise,Proportion_expressed=sum(Expressed_samples > 0)/length(Expressed_samples),Max_median=median(Max_expression)),
      ddply(RNA_TE[which(RNA_TE$Expressed_samples > 0),],.(class_update),summarise,Max_median_expressed=median(Max_expression)),by="class_update")

# Average TE vs. exon expression
TE_express = unlist(RNA_TE[,8:63])
median(TE_express)
median(TE_express[which(TE_express > 1)])
rm(TE_express)
exon_express = unlist(RNA_refseq_exon[,5:60])
median(exon_express)
median(exon_express[which(exon_express > 1)])
rm(exon_express)
```

### Cancer/IMR90 outlier analysis

```{r analysis Fig S5a}
combine_stats_noCancer_IMR90

# Average proportion of TEs in each state per sample, no IMR90
ddply(combine_boxplot,.(State),summarise,Mean=mean(Proportion),SD=sd(Proportion),Outlier=mean(Proportion)+2*sd(Proportion),Median=median(Proportion),IQR=IQR(Proportion))
ddply(combine_boxplot[which(combine_boxplot$Sample != "E017"),],.(State),summarise,Mean=mean(Proportion),SD=sd(Proportion),Median=median(Proportion),IQR=IQR(Proportion))

# Proportion of TEs in each state, IMR90
combine_boxplot[which(combine_boxplot$Sample == "E017"),]
```

### Response Figure 1 (replicates)

Identifying pairs of samples that are so similar they could be considered replicates. Relied on Table S1 from the Roadmap publication. 

Files of all TE x sample in the state by state and subfamily, for each active regulatory state (see Figure 5). For each TE, each state, each category and grouping (e.g., Group: Brain), I found the number of samples the TE is in the state, the number of samples the TE is in the state total, and the number of samples in the category.

```{bash active reg matrices, eval=FALSE}
# Lists of individual TEs in each active regulatory state by sample
cat chromHMM/subfamily/by_state/*1_TssA* > chromHMM_1_TssA.txt
cat chromHMM/subfamily/by_state/*2_TssAFlnk* > chromHMM_2_TssAFlnk.txt
cat chromHMM/subfamily/by_state/*3_TxFlnk* > chromHMM_3_TxFlnk.txt
cat chromHMM/subfamily/by_state/*6_EnhG* > chromHMM_6_EnhG.txt
cat chromHMM/subfamily/by_state/*7_Enh* > chromHMM_7_Enh.txt
```

For each replicate sample pair, I found the fraction of TEs that are in both samples over the fraction in either sample (or both). The proportions are very low. However, this may be due to the fact that I exclude any TEs that are in the state outside the pair, which biases away from those in many samples. We also know that the structure of the samples is such that there are many clusters of similar samples, not strict replicates. Therefore, I reran the analysis with several thresholds:

+ Exclusive to the sample pair (mean reproducible 6 +/- 1%)
+ <5 samples total (8 +/- 1%)
+ Exclusive to that group (or groups for the 3 pairs where the samples are from multiple groups) (10 +/- 2%)
+ No threshold (32 +/- 8%)

However, with the exception of no threshold, these only marginally improve the results.  

```{r load replicate matrices, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/replicate_matrices.R")
```

```{r run replicates, cache=TRUE, cache.lazy=FALSE}
# Select replicates
replicates = as.data.frame(matrix(c("E019","E020",
                                    "E008","E014",
                                    "E003","E016",
                                    "E055","E056",
                                    "E059","E061",
                                    "E101","E102"),ncol=2,byrow = TRUE))
colnames(replicates) = c("Sample 1","Sample 2")
replicates$Set = rep("Replicate",6)

test = as.data.frame(cbind(rep(as.vector(replicates$`Sample 1`),each=4),
                           rep(c("E032","E049","E066","E071"),times=6),
                           rep("Tissue",24)))
colnames(test) = c("Sample 1","Sample 2","Set")

replicates = rbind(replicates,test)
replicates$Pair = paste(replicates$`Sample 1`,replicates$`Sample 2`,sep="/")
rm(test)

# Apply over replicates
## Number of samples per TE x state
chromHMM_replicates = apply(replicates,1,function(z) ddply(chromHMM_active_matrix[which(chromHMM_active_matrix$Sample %in% as.vector(z[1:2])),],.(State,chromosome,start,stop,subfamily,family,class,strand),summarise,Samples=length(Sample)))
names(chromHMM_replicates) = replicates$Pair
chromHMM_replicates = ldply(chromHMM_replicates)
colnames(chromHMM_replicates)[1] = "Pair"
chromHMM_replicates = merge(chromHMM_replicates,replicates[,c("Pair","Set")],by="Pair")
```

```{r run replicates window, cache=TRUE, cache.lazy=FALSE, eval=FALSE}
## Number of samples per window x state
window_replicates = apply(replicates,1,function(z) ddply(windows_active[which(windows_active$Sample %in% as.vector(z[1:2])),],.(State,chr,start,stop),summarise,Samples=length(Sample)))
names(window_replicates) = replicates$Pair
window_replicates = ldply(window_replicates)
colnames(window_replicates)[1] = "Pair"
window_replicates = merge(window_replicates,replicates[,c("Pair","Set")],by="Pair")
```

```{r filter replictes, eval=FALSE}
# Combine all active regulatory states
chromHMM_replicates_active = ddply(chromHMM_replicates,.(Pair,Set,chromosome,start,stop,subfamily,family,class,strand),summarise,Samples=max(Samples))

# Fraction of TEs in either sample found in both
chromHMM_replicates_all = ddply(chromHMM_replicates_active,.(Pair,Set,Samples),summarise,Count=length(Samples))

## Combine all
chromHMM_replicates_count = dcast(chromHMM_replicates_all,Pair+Set~Samples,value.var="Count")
chromHMM_replicates_count[is.na(chromHMM_replicates_count)] = 0
chromHMM_replicates_count$Fraction = chromHMM_replicates_count$`2`/(chromHMM_replicates_count$`1`+chromHMM_replicates_count$`2`)
```

```{r replicate analysis, eval=FALSE}
dcast(ddply(chromHMM_replicates_count[which(chromHMM_replicates_count$Set == "Replicate"),],.(Threshold,State),summarise,Denominator=median(`1`+`2`)),State~Threshold,value.var="Denominator")

replicates_median = dcast(ddply(chromHMM_replicates_count,.(Threshold,State,Set),summarise,Median=median(Fraction)),Threshold+State~Set,value.var="Median")
replicates_median$Ratio = replicates_median$Replicate/replicates_median$Tissue
replicates_median
```

```{r Figure R1, echo=FALSE, eval=FALSE}
a = ggplot(chromHMM_replicates_count,aes(x=Pair,y=Fraction,fill=State,shape=Threshold)) + geom_point(color="black",stroke=1) +   scale_fill_manual(values=all_state_colors,guide=FALSE) + xlab("Sample pair") + ylab("Fraction of TEs in both samples / either sample") +   theme(axis.text.x=element_text(angle=90,vjust=0.5,size=7),legend.position = "bottom",strip.text.y = element_text(size=5)) + scale_shape_manual(values=c(21,22,24,25),guide=FALSE) + facet_grid(State~Set,scales="free_x",space="free_x",labeller=labeller(Set=setNames(c("Replicates","Different tissue"),c("Replicate","Tissue")))) + ylim(0,0.7)

b = ggplot(replicates_shuffled_count,aes(x=Pair,y=Fraction,fill=State,shape=Threshold)) + geom_point(color="black",stroke=1) + 
scale_fill_manual(values=all_state_colors,guide=FALSE) + xlab("Sample pair") + ylab("Fraction of TEs in both samples / either sample") + 
theme(axis.text.x=element_text(angle=90,vjust=0.5,size=7),strip.text.y = element_text(size=5)) + scale_shape_manual(values=c(21,22,25),guide=FALSE) + facet_grid(State~Set,scales="free_x",space="free_x",labeller=labeller(Set=setNames(c("Replicates","Different tissue"),c("Replicate","Tissue")))) + ylim(0,0.7)

legend_replicate = get_legend(ggplot(chromHMM_replicates_count,aes(x=Pair,y=Fraction,fill=State,shape=Threshold)) + geom_point(color="black",stroke=1) +   scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x=element_text(angle=90,vjust=0.5,size=7),legend.position = "bottom",strip.text.y = element_text(size=6)) + scale_shape_manual(values=c(21,22,24,25),labels=c("Exclusive to pair","<5 samples","Exclusive to group","None")) + facet_grid(State~Set,scales="free_x",space="free_x",labeller=labeller(Set=setNames(c("Replicates","Different tissue"),c("Replicate","Tissue")))))

grid.arrange(a,b,legend_replicate,heights=c(0.47,0.47,0.06))
```

### Supplementary Figure 7 (orthogonal)

Creates sets for analysis of TEs with restricted active regulatory activity versus constitutively active, non-TE bases. First, I identified TEs that are in each of the active regulatory states in 1-5 samples total ("highly restricted"). For my positive control, I could use promoters, but they are much longer than the average TE. Instead, I will use shuffled TEs that are in each state in >50 samples. With all 10 iterations of shuffled TEs, this is about the same number of TEs as in the first set. Finally, I got those that are not in the active regulatory states in any sample. As a bonus, although the reviewer asked for non-TE, non-restricted regions, I also found the TEs in each active regulatory state in >50 samples.

```{r create orthogonal sets, cache=TRUE, cache.lazy=FALSE}
# Restricted TEs
restricted = lapply(states[c(1:3,6:7)],function(x) merge(chromHMM_TE_state[which(chromHMM_TE_state[[x]] > 0 & chromHMM_TE_state[[x]] <= 5),TE_coordinates],rmsk_TE,by=TE_coordinates))
names(restricted) = states[c(1:3,6:7)]

# Positive control, shuffled TEs
load("R_datasets/shuffled_TEs.RData")
load("R_datasets/shuffled_chromHMM.RData")
shuffled_pos = lapply(seq(1,10,1),
                      function(y) {a = lapply(states[c(1:3,6:7)],function(x) merge(shuffled_chromHMM_potential[[y]][which(shuffled_chromHMM_potential[[y]][[paste("X",x,sep="")]] > 50),TE_coordinates],
                                                                                   rmsk_TE_shuffled[[y]],by=TE_coordinates));
                        names(a) = states[c(1:3,6:7)]; return(a)})
rm(list=c("shuffled_chromHMM_potential","rmsk_TE_shuffled"))

# Positive control, TEs
TE_pos = lapply(states[c(1:3,6:7)],function(x) merge(chromHMM_TE_state[which(chromHMM_TE_state[[x]] > 50),c(TE_coordinates)],rmsk_TE,by=TE_coordinates))
names(TE_pos) = states[c(1:3,6:7)]

# Negative control
TE_neg = merge(chromHMM_TE_state[which(apply(chromHMM_TE_state[,states[c(1:3,6:7)]],1,function(x) sum(x) == 0)),c(TE_coordinates)],rmsk_TE,by=TE_coordinates)
```

```{r write orthogonal sets, eval=FALSE}
lapply(states[c(1:3,6:7)],function(x) write.table(restricted[[x]][,TE_coordinates[c(1:4,6,7,5)]],file=paste("orthogonal/restricted_",x,".txt",sep=""),row.names=FALSE,col.names=FALSE,quote=FALSE,sep='\t'))

lapply(seq(1,10,1),function(y) lapply(states[c(1:3,6:7)],function(x) write.table(shuffled_pos[[y]][[x]][,TE_coordinates[c(1:4,6,7,5)]],file=paste("orthogonal/shuffle_",y,"_constitutive_",x,".txt",sep=""),row.names=FALSE,col.names=FALSE,quote=FALSE,sep='\t')))

lapply(states[c(1:3,6:7)],function(x) write.table(TE_pos[[x]][,TE_coordinates[c(1:4,6,7,5)]],file=paste("orthogonal/constitutive_",x,".txt",sep=""),row.names=FALSE,col.names=FALSE,quote=FALSE,sep='\t'))

write.table(TE_neg[,TE_coordinates[c(1:4,6,7,5)]],file="orthogonal/not_active.txt",row.names=FALSE,col.names=FALSE,quote=FALSE,sep='\t')
```

Combined the shuffled files together. 

How many of the shuffled constitutive TEs are redundant? I found the total number of shuffled TEs per state, then compared it to the unique number. There are almost no redundancies, just one in the 2_TssAFlnk state, although many are clearly close together. This is likely because it is very unlikely for the exact shuffled TE to land at the same location in different iterations. Also looked at the overlapping bp within the sets. This is a large fraction of the total bp in each state, particularly for the 1_TssA state. 

```{bash combined shuffle, eval=FALSE}
while read line; do cat orthogonal/shuffle_*_constitutive_$line\.txt > orthogonal/shuffle_constitutive_$line\.txt; done < sample_lists/chromHMM_states.txt

# Shuffled constitutive QC
## Redundant TEs
while read line; do cat orthogonal/shuffle_*_constitutive_$line\.txt | wc -l; done < sample_lists/chromHMM_states.txt
while read line; do cat orthogonal/shuffle_*_constitutive_$line\.txt | sort | uniq | wc -l; done < sample_lists/chromHMM_states.txt

## Overlapping bp
for state in 1_TssA 2_TssAFlnk 3_TxFlnk 6_EnhG 7_Enh; do cat orthogonal/shuffle_*_constitutive_$state\.txt | awk '{sum+=$3-$2}END{print sum}'; done
for state in 1_TssA 2_TssAFlnk 3_TxFlnk 6_EnhG 7_Enh; do cat orthogonal/shuffle_*_constitutive_$state\.txt | cut -f1-3| sort -k1,1 -k2,2n | bedtools merge -i - | awk '{sum+=$3-$2}END{print sum}' -; done
```

```{r orthogonal metadata}
orthogonal_sets = data.frame(cbind(State=c(rep(states[c(1:3,6:7)],3),NA),
                                   Set=c(rep("constitutive",5),rep("restricted",5),rep("shuffle_constitutive",5),"not_active")))
orthogonal_sets$File = paste(orthogonal_sets$Set,orthogonal_sets$State,sep="_")
orthogonal_sets$File[16] = "not_active"

orthogonal_labels = setNames(c("Restricted activity","Shuffled constitutive","Constitutively active","Inactive"),c("restricted","shuffle_constitutive","constitutive","not_active"))
```

Characteristics of each of the orthogonal evidence sets. Both the constitutively active and restricted true TEs are shorter than the shuffled constitutively active TEs, although TEs with restricted activity are slightly longer, particularly for the promoter states. Inactive TEs are intermediate in length. Constitutively active true TEs also have a higher proportion of SINE elements at the expense of LINE/LTR elements compared to restricted ones, likely due to exclusion of those TEs from canonical promoters. Finally, true TEs with constitutive promoter active are far more likely to overlap promoters, while those with restricted 7_Enh activity are more likely to be intergenic vs. intronic than TEs constitutively in the state.

```{r orthogonal characteristics}
orthogonal_profile = rbind(ldply(shuffled_pos,function(y) profile_set(y)),profile_set(restricted),profile_set(TE_pos))
test = profile_set(list(TE_neg))
test$`.id` = NA
orthogonal_profile = rbind(orthogonal_profile,test)
rm(test)

colnames(orthogonal_profile)[1:3] = c("State","Number","Length")
orthogonal_profile$SVA = orthogonal_profile$Other
orthogonal_profile$Other = orthogonal_profile$`DNA?`+orthogonal_profile$`LINE?` + orthogonal_profile$`LTR?` + orthogonal_profile$RC + orthogonal_profile$`SINE?` + orthogonal_profile$Unknown + orthogonal_profile$`Unknown?`
orthogonal_profile$Set = factor(c(rep("shuffle_constitutive",50),rep("restricted",5),rep("constitutive",5),"not_active"),levels=c("constitutive","shuffle_constitutive","restricted","not_active"))

#ggplot(orthogonal_profile,aes(x=Set,y=Length,fill=State)) + geom_boxplot() + scale_fill_manual(values=all_state_colors) + ylim(0,350) + ylab("Median length (bp)") + scale_x_discrete(labels=orthogonal_labels)

#ggplot(melt(orthogonal_profile[,c("Set","State",cohorts[c(1:2,5,8,10,13,16,19)])],id.vars=c("Set","State")),aes(x=Set,y=value,fill=State)) + geom_boxplot() + scale_fill_manual(values=all_state_colors,guide=FALSE) + ylim(0,1) + scale_x_discrete(labels=orthogonal_labels) + facet_grid(State~variable) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + ylab("Proportion overlapping feature")

#ggplot(melt(orthogonal_profile[,c("Set","State","DNA","LINE","LTR","SINE","SVA","Other")],id.vars=c("Set","State")),aes(x=Set,y=value,fill=State)) + geom_boxplot() + scale_fill_manual(values=all_state_colors,guide=FALSE) + ylim(0,1) +  scale_x_discrete(labels=orthogonal_labels) + facet_grid(State~variable) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + ylab("Proportion in class")
```

'''VISTA'''

Intersected the VISTA enhancers with each set of TEs. 9/10 shuffled 3_TxFlnk constitutive sets do not overlap any VISTA enhancers, as do the real TE 1_TssA and 3_TxFlnk constitutive sets.  

```{bash orthogonal vista, eval=FALSE}
for file in orthogonal/*.txt; do bedtools intersect -wo -a $file -b features/vista_enhancers/vista_enhancers_human_all.bed > orthogonal/vista/$(basename $file .txt)_vista.txt; done

# Combined shuffled
for file in orthogonal/shuffle_constitutive_*.txt; do bedtools intersect -wo -a $file -b features/vista_enhancers/vista_enhancers_human_all.bed > orthogonal/vista/$(basename $file .txt)_vista.txt; done
```

Across all of the sets, there are 7,630 TEs that overlap VISTA enhancers, and in only 42 of the cases does the TE overlaps more than one VISTA enhancer (2 or 3). However, many TEs can overlap the same element. 

I found, for each set, the total number of VISTA enhancers overlapping TEs and the fraction that positively validated. The rate of validation for enhancers overlapping TEs with restricted activity is not lower than it is for the shuffled positive control, except for the 7_Enh state. The validation rate is lower for TEs that are never active. 

```{r orthogonal vista analysis}
vista_val = lapply(as.vector(orthogonal_sets$File),function(x) read.table(paste("orthogonal/vista/",x,"_vista.txt",sep=""),sep="\t",
                                          col.names=c(TE_coordinates[c(1:4,6,7,5)],"chr_vista","start_vista","stop_vista","element","Validation","Overlap")))
names(vista_val) = as.vector(orthogonal_sets$File)

# Repeat inactive results for each state
vista_val = vista_val[c(1:15,rep(16,5))]

vista_val = vista_val[sapply(vista_val, function(x) dim(x)[1]) > 0]
vista_val = ldply(vista_val,.id = "File")
vista_val = merge(vista_val,orthogonal_sets,by="File",all.x=TRUE)
vista_val[which(vista_val$File == "not_active"),]$State = rep(states[c(1:3,6:7)],each=dim(vista_val[which(vista_val$File == "not_active"),])[1]/5)
vista_val$Set = factor(vista_val$Set,levels=names(orthogonal_labels))
  
# Unique VISTA enhancers overlapping TEs in that set
vista_val_fraction = ddply(vista_val,.(Set,State),summarise,Total=length(unique(element)),
      Pos=length(unique(element[which(Validation == "positive")])),
      Neg=length(unique(element[which(Validation == "negative")])),
      Fraction=Pos/Total)

# Fisher's exact test
lapply(states[c(1:3,6:7)],function(x) fisher.test(vista_val_fraction[which(vista_val_fraction$State == x & vista_val_fraction$Set %in% c("restricted","shuffle_constitutive")),c("Pos","Neg")]))
lapply(states[c(2,6:7)],function(x) fisher.test(vista_val_fraction[which(vista_val_fraction$State == x & vista_val_fraction$Set %in% c("constitutive","shuffle_constitutive")),c("Pos","Neg")]))
lapply(states[c(1:3,6:7)],function(x) fisher.test(vista_val_fraction[which(vista_val_fraction$State %in% c(x,NA) & vista_val_fraction$Set %in% c("restricted","not_active")),c("Pos","Neg")]))
```

'''HOMER'''

Ran HOMER (known motifs only) on each of the orthogonal sets. 

```{bash orthogonal homer, eval=FALSE}
for file in orthogonal/*.txt; do echo $file; findMotifsGenome.pl $file hg19 orthogonal/homer/$(basename $file .txt) -size given -nomotif 2>> orthogonal/homer/output.txt; done

## Combined shuffled
for file in orthogonal/shuffle_constitutive_*.txt; do echo $file; findMotifsGenome.pl $file hg19 orthogonal/homer/$(basename $file .txt) -size given -nomotif 2>> orthogonal/homer/output.txt; done
```

HOMER gives more detail about the motifs [http://homer.ucsd.edu/homer/motif/motifDatabase.html| here].

```{r load homer results}
# Load members with motif (no background)
homer_files = list.files(path="orthogonal/homer/individual/",pattern="knownResults.txt",recursive=TRUE,full.names=TRUE)

homer_known = lapply(homer_files,function(x) read.table(x,sep='\t',header=TRUE,comment.char=""))
names(homer_known) = homer_files
homer_known = lapply(homer_known,setNames,nm=c("Motif Name","Consensus","P-value","Log P-value","q-value (Benjamini)",
                                               "Target Seqs with Motif","% of Target Sequences with Motif",
                                               "Background Seqs with Motif","% of Background Sequences with Motif"))
# Repeat inactive results for each state
homer_known = homer_known[c(1:5,7:16,rep(6,5))]

homer_known = ldply(homer_known,.id = "File")
homer_known$File = gsub("/knownResults.txt","",gsub("orthogonal/homer/individual//","",homer_known$File))
homer_known$State = rep(rep(states[c(1:3,6:7)],each=364),4)
homer_known$Set = factor(rep(c("constitutive","restricted","shuffle_constitutive","not_active"),each=1820),levels=names(orthogonal_labels))
homer_known$`% of Target Sequences with Motif` = as.numeric(gsub("%","",homer_known$`% of Target Sequences with Motif`))
homer_known$`% of Background Sequences with Motif` = as.numeric(gsub("%","",homer_known$`% of Background Sequences with Motif`))

# Add number of TEs per set
homer_known = merge(homer_known,ddply(orthogonal_profile,.(State,Set),summarise,Number=sum(Number)),by=c("State","Set"),all.x=TRUE)
homer_known[which(homer_known$Set == "not_active"),]$Number = orthogonal_profile[which(orthogonal_profile$Set == "not_active"),]$Number
```

```{r homer statistics}
a = ddply(homer_known[which(homer_known$State %in% states),],.(`Motif Name`,State),function(x) {y <- x[which(x$Set %in% c("restricted","shuffle_constitutive")),c("Target Seqs with Motif","Number")]; y$`Target without` = y$Number - y$`Target Seqs with Motif`; return(unlist(chisq.test(y[,c("Target Seqs with Motif","Target without")]))["p.value"])})
a$Comparison = rep("restricted_shuffle",dim(a)[1])

b = ddply(homer_known[which(homer_known$State %in% states),],.(`Motif Name`,State),function(x) {y <- x[which(x$Set %in% c("constitutive","shuffle_constitutive")),c("Target Seqs with Motif","Number")]; y$`Target without` = y$Number - y$`Target Seqs with Motif`; return(unlist(chisq.test(y[,c("Target Seqs with Motif","Target without")]))["p.value"])})
b$Comparison = rep("constitutive_shuffle",dim(a)[1])

c = ddply(homer_known[which(homer_known$State %in% states),],.(`Motif Name`,State),function(x) {y <- x[which(x$Set %in% c("restricted","not_active")),c("Target Seqs with Motif","Number")]; y$`Target without` = y$Number - y$`Target Seqs with Motif`; return(unlist(chisq.test(y[,c("Target Seqs with Motif","Target without")]))["p.value"])})
c$Comparison = rep("restricted_inactive",dim(b)[1])

homer_stats = rbind(a,b,c)
homer_stats$Padj = p.adjust(homer_stats$p.value,method = "fdr")
```

```{r orthogonal homer analysis}
# Number of significant comparisons
table(droplevels(homer_stats[which(homer_stats$Padj < 0.05),c("Comparison","State")]))

# Number of significant comparisons at each p-value threshold
test = lapply(seq(0,350,1),function(x) melt(as.matrix(table(homer_stats[which(as.numeric(homer_stats$p.value) < 10^-x),c("Comparison","State")]))))
names(test) = lapply(seq(0,350,1),function(x) 10^-x)
test = ldply(test,.id="Cutoff")
test$Cutoff = as.numeric(test$Cutoff)
ggplot(test,aes(x=Cutoff,y=value,fill=Comparison)) + geom_bar(position="stack",stat="identity") + facet_wrap(~State)

# Promoter motifs
homer_stats[which(homer_stats$`Motif Name` %in% grep("Promoter",unique(homer_stats$`Motif Name`),value=TRUE) & homer_stats$Padj < 0.05),]
```

'''Population'''

Conserved elements generated by phastCons, based on the primate alignment only, from UCSC. Removed the bin field to turn it into a bed file. 

```{bash phastCons, eval=FALSE}
wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/phastConsElements46wayPrimates.txt.gz
awk -v OFS='\t' '{print $2, $3, $4, $5, $6}' orthogonal/pop/phastConsElements46wayPrimates.txt > orthogonal/pop/phastCons.txt
```

Downloaded the hg19 GERP elements and combined into one file. There are 1,044,996 elements. 

```{bash GERP, eval=FALSE}
wget http://mendel.stanford.edu/SidowLab/downloads/gerp/hg19.GERP_elements.tar.gz
tar -xvzf hg19.GERP_elements.tar.gz
for file in GERP/*; do awk -v OFS='\t' -v chr=$file '{print chr, $1, $2, $3, $4, $5}' $file >> GERP_hg19_elems.txt; done
sed 's/GERP\/hg19_//' GERP_hg19_elems.txt | sed 's/_elems\.txt//' | sort -k1,1 -k2,2n > GERP_hg19_elems_sorted.txt
```

Intersection with all sets, found the length of overlap per element. 

```{bash phastCons intersection, eval=FALSE}
for file in orthogonal/*.txt; do bedtools intersect -wao -a $file -b orthogonal/pop/phastCons.txt > orthogonal/pop/$(basename $file .txt)_phastCons.txt; done
for file in orthogonal/pop/*phastCons.txt; do awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $7, $6]+=$13}END{for(i in a) {split(i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i]}}' $file > orthogonal/pop/$(basename $file .txt)_total.txt; done
```

```{bash GERP intersection, eval=FALSE}
for file in orthogonal/*.txt; do sort -k1,1 -k2,2n $file | bedtools intersect -wao -a - -b orthogonal/pop/GERP_hg19_elems_sorted.txt -sorted > orthogonal/pop/$(basename $file .txt )_gerp.txt; done
for file in orthogonal/pop/*gerp.txt; do awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $7, $6]+=$14}END{for(i in a) {split(i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i]}}' $file > orthogonal/pop/$(basename $file .txt)_total.txt; done
```

```{r phastCons GERP, cache=TRUE, cache.lazy=FALSE}
phastcons = lapply(as.vector(orthogonal_sets$File),function(x) read.table(paste("orthogonal/pop/",x,"_phastCons_total.txt",sep=""),sep='\t',col.names=c(TE_coordinates[c(1:4,6,5,7)],"Overlap")))
names(phastcons) = as.vector(orthogonal_sets$File)

# Repeat inactive results for each state
phastcons = phastcons[c(1:15,rep(16,5))]

phastcons = ldply(phastcons,.id = "File")
phastcons$Database = rep("PhastCons",dim(phastcons)[1])

gerp = lapply(as.vector(orthogonal_sets$File),function(x) read.table(paste("orthogonal/pop/",x,"_gerp_total.txt",sep=""),sep='\t',col.names=c(TE_coordinates[c(1:4,6,5,7)],"Overlap")))
names(gerp) = as.vector(orthogonal_sets$File)

# Repeat inactive results for each state
gerp = gerp[c(1:15,rep(16,5))]

gerp = ldply(gerp,.id = "File")
gerp$Database = rep("GERP",dim(gerp)[1])

cons = rbind(phastcons,gerp)
rm(list=c("phastcons","gerp"))

cons = merge(cons,orthogonal_sets,by="File",all.x=TRUE)
cons[which(cons$File == "not_active"),]$State = rep(rep(states[c(1:3,6:7)],each=dim(cons[which(cons$File == "not_active"),])[1]/10),2)
cons$Proportion = cons$Overlap/(cons$stop-cons$start)
cons$Set = factor(cons$Set,levels=names(orthogonal_labels))

ddply(cons,.(Set,State,Database),summarise,Mean=mean(Proportion),Overlapping=sum(Overlap > 10)/length(Overlap))

# Tests
## Proportion of element covered by conserved block (Wilcox)
lapply(states[c(1:3,6:7)],function(x) wilcox.test(data=cons[which(cons$State == x & cons$Set %in% c("restricted","shuffle_constitutive") & cons$Database == "PhastCons"),],Proportion~Set))
lapply(states[c(1:3,6:7)],function(x) wilcox.test(data=cons[which(cons$State == x & cons$Set %in% c("constitutive","shuffle_constitutive") & cons$Database == "PhastCons"),],Proportion~Set))
lapply(states[c(1:3,6:7)],function(x) wilcox.test(data=cons[which(cons$State %in% c(x,NA) & cons$Set %in% c("restricted","not_active") & cons$Database == "PhastCons"),],Proportion~Set))

lapply(states[c(1:3,6:7)],function(x) wilcox.test(data=cons[which(cons$State == x & cons$Set %in% c("restricted","shuffle_constitutive") & cons$Database == "GERP"),],Proportion~Set))
lapply(states[c(1:3,6:7)],function(x) wilcox.test(data=cons[which(cons$State == x & cons$Set %in% c("constitutive","shuffle_constitutive") & cons$Database == "GERP"),],Proportion~Set))
lapply(states[c(1:3,6:7)],function(x) wilcox.test(data=cons[which(cons$State %in% c(x,NA) & cons$Set %in% c("restricted","not_active") & cons$Database == "GERP"),],Proportion~Set))

## Proportion of elements with >10bp of conserved block (Chi-squared)
cons_10bp = ddply(cons,.(Set,State,Database),summarise,Overlapping=sum(Overlap > 10),Total=length(Overlap),Not=Total-Overlapping)

lapply(states[c(1:3,6:7)],function(x) chisq.test(cons_10bp[which(cons_10bp$State == x & cons_10bp$Set %in% c("restricted","shuffle_constitutive") & cons_10bp$Database == "PhastCons"),c("Overlapping","Not")]))
lapply(states[c(1:3,6:7)],function(x) chisq.test(cons_10bp[which(cons_10bp$State == x & cons_10bp$Set %in% c("constitutive","shuffle_constitutive") & cons_10bp$Database == "PhastCons"),c("Overlapping","Not")]))
lapply(states[c(1:3,6:7)],function(x) chisq.test(cons_10bp[which(cons_10bp$State %in% c(x,NA) & cons_10bp$Set %in% c("restricted","not_active") & cons_10bp$Database == "PhastCons"),c("Overlapping","Not")]))

lapply(states[c(1:3,6:7)],function(x) chisq.test(cons_10bp[which(cons_10bp$State == x & cons_10bp$Set %in% c("restricted","shuffle_constitutive") & cons_10bp$Database == "GERP"),c("Overlapping","Not")]))
lapply(states[c(1:3,6:7)],function(x) chisq.test(cons_10bp[which(cons_10bp$State == x & cons_10bp$Set %in% c("constitutive","shuffle_constitutive") & cons_10bp$Database == "GERP"),c("Overlapping","Not")]))
lapply(states[c(1:3,6:7)],function(x) chisq.test(cons_10bp[which(cons_10bp$State %in% c(x,NA) & cons_10bp$Set %in% c("restricted","not_active") & cons_10bp$Database == "GERP"),c("Overlapping","Not")]))
```

```{r Figure S7, echo=FALSE}
# VISTA
a = ggplot(vista_val_fraction,aes(x=State,y=Fraction,shape=Set)) + geom_point() + ylim(0,1) + ylab("Proportion of VISTA enhancers\npositively validated") + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + scale_shape_discrete(guide=FALSE)

# Population
b = ggplot(cons_10bp,aes(x=State,y=Overlapping/Total,shape=Set)) + geom_point() + ylim(0,1) + facet_wrap(~Database) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + ylab("Proportion of TEs overlapping\nphastCons/GERP elements") + scale_shape_discrete(guide=FALSE)

# HOMER (promoter motifs)
c = ggplot(homer_known[which(homer_known$`Motif Name` %in% grep("Promoter",unique(homer_known$`Motif Name`),value=TRUE)),],aes(x=`Motif Name`,y=`% of Target Sequences with Motif`/100,shape=Set)) + geom_point() + xlab("Transcription factor") + ylab("Proportion of TEs with motif") + scale_shape_discrete(labels=orthogonal_labels,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + facet_wrap(~State) + scale_x_discrete(labels=sapply(grep("Promoter",unique(homer_known$`Motif Name`),value=TRUE),function(x) unlist(strsplit(as.character(x),split='[(]'))[[1]][1])) 

set_legend = get_legend(ggplot(vista_val_fraction,aes(x=State,y=Fraction,shape=Set)) + geom_point() + theme(legend.position = "bottom",legend.box.margin = margin(t=-10),legend.key.size = unit(3,'mm')) + scale_shape_discrete(labels=orthogonal_labels))

grid.arrange(a,b,c,set_legend,layout_matrix = rbind(c(1,2),c(3),c(4)),widths=c(0.35,0.65),heights=c(0.35,0.6,0.05))
```

### Figure 3

```{bash chromHMM intra, eval=FALSE}
# Updated 5/17/18 to 6/5/2018 with summit overlap

# State switching intra, chromHMM
#TE_landscape/chromHMM/state_switching/rmsk_TEother_chromHMM_intra.txt
python ~/bin/TE_landscape/state_sharing_intra2.py chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt chromHMM/chromHMM_states.txt chromHMM/state_switching/rmsk_TEother_chromHMM_intra.txt 7 9
```

```{bash chromHMM inter, eval=FALSE}
# State switching inter, chromHMM
#TE_landscape/chromHMM/state_switching/rmsk_TEother_chromHMM_inter.txt
python ~/bin/TE_landscape/state_sharing_inter.py chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt chromHMM/chromHMM_states.txt chromHMM/state_switching/rmsk_TEother_chromHMM_inter.txt 7 9
```

```{bash chromHMM inter class, eval=FALSE}
# State switching inter, chromHMM, by class
#TE_landscape/chromHMM/state_switching/class/rmsk_TEother_chromHMM_inter_[class].txt [6 files]
while read line; do python ~/bin/TE_landscape/state_sharing_inter.py chromHMM/chromHMM_summit_$line\.txt chromHMM/chromHMM_states.txt chromHMM/state_switching/class/rmsk_TEother_chromHMM_inter_$line\.txt 7 9; done < features/TEs/class/TE_class.txt

python ~/bin/TE_landscape/state_sharing_inter.py chromHMM/chromHMM_summit_SVA.txt chromHMM/chromHMM_states.txt chromHMM/state_switching/class/rmsk_TEother_chromHMM_inter_SVA.txt 7 9

python ~/bin/TE_landscape/state_sharing_inter.py chromHMM/chromHMM_summit_Other.txt chromHMM/chromHMM_states.txt chromHMM/state_switching/class/rmsk_TEother_chromHMM_inter_Other.txt 7 9
```

```{bash WGBS inter, eval=FALSE}
# Methylation level of each TE x sample	 
#TE_landscape/WGBS/TE_WGBS_state.txt (from WGBS_TE_avg_methylation.R)
# Removed header

# Methylation level of each TE x sample, sorted	 
#TE_landscape/WGBS/TE_WGBS_state_sorted.txt	
sort -k1,1V -k2,2n -k3,3n -k4,4 -k8,8 WGBS/TE_WGBS_state.txt > WGBS/TE_WGBS_state_sorted.txt

# Adding methylation states	 
awk -v OFS='\t' '{if($9 == "NA") print $0, "Missing"; else if ($9 < 0.3) print $0, "Hypomethylated"; else if ($9 > 0.7) print $0, "Hypermethylated"; else if (($9 <= 0.7) && ($9 >= 0.3)) print $0, "Intermediate";}' TE_WGBS_state_sorted.txt > TE_WGBS_state_sorted

# WGBS inter switching
#TE_landscape/WGBS/TE_WGBS_state_inter.txt		 
python ~/bin/TE_landscape/state_sharing_inter.py WGBS/TE_WGBS_state_sorted.txt WGBS/methylation_states.txt WGBS/TE_WGBS_state_inter.txt 7 9
```

```{bash WGBS inter class, eval=FALSE}
# Methylation level/state of each TE x sample, sorted, by class	 
#TE_landscape/WGBS/class/[class]_WGBS_state_sorted.txt [6 files]	
while read line ; do awk -v OFS='\t' -v class=$line '{if($5 == class) print $0}' TE_WGBS_state_sorted.txt > $line\_WGBS_state_sorted.txt; done < ../features/TEs/class/TE_class.txt

awk -v OFS='\t' '{if($5 == "Other") print $0}' TE_WGBS_state_sorted.txt > SVA_WGBS_state_sorted.txt

awk -v OFS='\t' '{if(($5 != "Other") && ($5 != "LINE") && ($5 != "SINE") && ($5 != "DNA") && ($5 != "LTR")) print $0}' TE_WGBS_state_sorted.txt > Other_WGBS_state_sorted.txt

# WGBS inter switching, by class
#TE_landscape/WGBS/class/[class]_WGBS_state_sorted.txt_inter.txt [6 files]		
for file in WGBS/class/*WGBS_state_sorted.txt; do python ~/bin/TE_landscape/state_sharing_inter.py $file WGBS/methylation_states.txt $file\_inter.txt 7 9; done &
```

```{r Figure 3, echo=FALSE, fig.width=7, fig.height=6}
source("R_scripts/state_switching.R")

# chromHMM states
a = ggplot(chromHMM_TE_state,aes(x=States,y = ..density..,fill=class_update)) + geom_histogram(binwidth=1,color="grey") + xlab("Number of chromHMM states") + ylab("Proportion of TEs") + scale_fill_manual(values=class_colors,guide=FALSE) + facet_wrap(~class_update)   

# WGBS states
b = ggplot(TE_meth_average,aes(x=States,y = ..density..,fill=class_update)) + geom_histogram(binwidth=1,color="grey") + xlab("Number of methylation states") + ylab("Proportion of TEs") + scale_fill_manual(values=class_colors,guide=FALSE) + facet_wrap(~class_update) 

# State switching inter, chromHMM
c = ggplot(melt(as.matrix(state_switching_inter)),aes(x=Var2,y=Var1,fill=value/sample_counts["All","chromHMM"])) + geom_tile() + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(limits=rev(chromHMM_states)) + scale_fill_gradient(low="white",high="darkmagenta",limits=c(0,1),guide=FALSE) + coord_equal() + xlab("State 2") + ylab("State 1")

# State switching inter, WGBS
d = ggplot(melt(as.matrix(ss_inter_meth)),aes(x=Var2,y=Var1,fill=value/sample_counts["All","WGBS"])) + geom_tile() + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + scale_y_discrete(limits=rev(meth_states)) + scale_fill_gradient(low="white",high="darkmagenta",limits=c(0,1),guide=FALSE) + coord_equal() + xlab("State 2") + ylab("State 1")

scale_switching = get_legend(ggplot(melt(as.matrix(ss_inter_meth)),aes(x=Var2,y=Var1,fill=value/sample_counts["All","WGBS"])) + geom_tile() + theme(legend.position = "bottom",legend.direction = "horizontal") + scale_fill_gradient(low="white",high="darkmagenta",limits=c(0,1),name="Average proportion\nof samples in state"))

grid.arrange(a,b,c,d,scale_switching,nrow = 3, layout_matrix=rbind(c(1,2),c(3,4),c(5,5)),heights=c(0.3,0.6,0.1))
```

### Number of states per TE, by state

TEs ever in each state versus the total number of states they are in across all samples. 

```{r state dynamics}
chromHMM_ever_total = ldply(chromHMM_states,function(x) median(chromHMM_TE_state[which(chromHMM_TE_state[[x]] > 0),]$States))
chromHMM_ever_total$State = chromHMM_states
chromHMM_ever_total

chromHMM_dynamics = melt(chromHMM_TE_state[,c(chromHMM_states,"States")],id.var="States")
chromHMM_dynamics = chromHMM_dynamics[which(chromHMM_dynamics$value > 0),]

meth_ever_total = ldply(meth_states,function(x) median(TE_meth_average[which(TE_meth_average[[x]] > 0),]$States))
meth_ever_total$State = meth_states
meth_ever_total

meth_dynamics = melt(TE_meth_average[,c(meth_states,"States")],id.var="States")
meth_dynamics = meth_dynamics[which(meth_dynamics$value > 0),]
```

### Supplementary Figure 8

```{r Figure S8, echo=FALSE}
# State dynamics
a = ggplot(chromHMM_dynamics,aes(x=States,y=..scaled..,fill=forcats::fct_rev(variable))) + geom_density(adjust=10,alpha=0.5) + scale_fill_manual(values=chromHMM_colors,name="State",limits=chromHMM_states) + xlim(0,15) + ylab("Density") + theme(legend.position="bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(3,'mm')) + guides(fill=guide_legend(nrow=4,title.position = "top")) + xlab("Number of chromHMM states")

b = ggplot(meth_dynamics,aes(x=States,y=..scaled..,fill=forcats::fct_rev(variable))) + geom_density(adjust=10,alpha=0.8) + scale_fill_manual(values=meth_colors,name="State",limits=meth_states,labels=meth_labels) + ylab("Density") + theme(legend.position="bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(3,'mm')) + guides(fill=guide_legend(nrow=4,title.position = "top")) + xlab("Number of methylation states")

# chromHMM states intra
c = ggplot(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit"),],aes(x=Max_states_intra,y = ..density..,fill=class_update)) + geom_histogram(binwidth=1,color="grey") + ylab("Proportion of TEs") + xlab("Number of chromHMM states") + facet_wrap(~class_update) + scale_fill_manual(values=class_colors,guide=FALSE) + theme(legend.position="bottom")

# State switching intra, chromHMM
d = ggplot(melt(as.matrix(state_switching_intra)),aes(x=Var2,y=Var1,fill=value)) + geom_tile() + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.title.y = element_text(margin = margin(t = -10))) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(labels=rev(chromHMM_states),limits=rev(levels(melt(as.matrix(state_switching_intra))$Var1))) + scale_fill_gradient(low="white",high="darkblue",limits=c(0,1),name="Proportion of\nTE instances") + coord_equal() + xlab("State 2") + ylab("State 1")

# State switching inter, WGBS, by class
e = ggplot(melt(ss_inter_meth_class,id.vars=c("Class","State")),aes(x=variable,y=State,fill=value/sample_counts["All","WGBS"])) + geom_tile() + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + scale_y_discrete(limits=rev(meth_states)) + scale_fill_gradient(low="white",high="darkmagenta",limits=c(0,1),guide=FALSE) + facet_wrap(~Class,nrow=1) + coord_equal() + xlab("State 2") + ylab("State 1")

grid.arrange(a,b,c,d,e,scale_switching,nrow=4,layout_matrix=rbind(c(1,2),c(3,4),c(5),c(6)),heights=c(0.35,0.3,0.3,0.05),widths=c(0.55,0.45))
```

### State switching analysis

```{r state switching analysis}
state_switching_intra
state_switching_inter

dim(chromHMM_TE_state[which(chromHMM_TE_state$`10_TssBiv` > 0 | chromHMM_TE_state$`11_BivFlnk` > 0),])
```

### Number of states per TE, overall and by class, chromHMM and WGBS

```{r states overall}
table(chromHMM_TE_state$States)
ddply(chromHMM_TE_state,.(),summarise,Mean=mean(States),SD=sd(States),Median=median(States),IQR=IQR(States))
ddply(chromHMM_TE_state,.(class_update),summarise,Mean=mean(States),SD=sd(States),Median=median(States),IQR=IQR(States))

states_outlier = chromHMM_TE_state[which(chromHMM_TE_state$States == 15),]
ddply(states_outlier,.(),summarise,Length=median(stop-start),States_intra=median(Tissues)/127,Max_states_intra=median(Max_states_intra))
states_outlier = merge(states_outlier,rmsk_TE[,c(TE_coordinates,cohorts)],by=TE_coordinates,all.x=TRUE)
states_outlier
write.table(states_outlier[,c(TE_coordinates,"class_update",cohorts)],file="States_15_TEs.bed",sep='\t',row.names=FALSE,quote=FALSE)

table(TE_meth_average$States)
ddply(TE_meth_average,.(),summarise,Mean=mean(States),SD=sd(States),Median=median(States),IQR=IQR(States))
ddply(TE_meth_average,.(class_update),summarise,Mean=mean(States),SD=sd(States),Median=median(States),IQR=IQR(States))

# Correlation of TE length with number of states ever
cor.test((chromHMM_TE_state$stop-chromHMM_TE_state$start),chromHMM_TE_state$States,method="spearman")
cor.test((TE_meth_average$stop-TE_meth_average$start),TE_meth_average$States,method="spearman")

# Correlation of number of states ever, max in one sample
by(chromHMM_TE_state,chromHMM_TE_state$Category,function(x) cor.test(x$Max_states_intra,x$States,method="spearman"))

# By chromosome
ddply(chromHMM_TE_state,.(chromosome,Category),summarise,Max_states_intra=median(Max_states_intra),States=median(States))
```

### Number of states per TE x sample

```{bash chromHMM states per TE, eval=FALSE}
# Number of states per TE x sample
python ~/bin/TE_landscape/state_sharing_intra_lite.py chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt chromHMM/rmsk_TEother_chromHMM_summit_state_counts.txt 7 9
```

```{bash WGBS states per TE, eval=FALSE}
# TEs overlapping a CpG in each methylation state by sample
#TE_landscape/WGBS/TE_CpG_Meth_state.txt
python ~/bin/TE_landscape/count_CpG_state.py TE_CpG_Meth_new.bed TE_CpG_count.txt ../sample_lists/WGBS_samples.txt TE_CpG_Meth_state.txt

# Number of TE x sample with CpGs in more than one state
python ~/bin/TE_landscape/shared_meth_CpG.py WGBS/TE_CpG_Meth_state.txt WGBS/TE_CpG_state_counts.txt
```

```{r states per TE-sample}
# chromHMM
test = read.table("chromHMM/rmsk_TEother_chromHMM_summit_state_counts.txt",sep='\t')
colnames(test) = c("States","Total")
## Remove erroneous Quies entry
test[1,2] = test[1,2]-1
test$Proportion = test$Total/sum(test$Total)
test
## Proportion of those with >1 state in only 2 states
test$Total[2]/sum(test$Total[2:15])

table(chromHMM_TE_state[,c("Max_states_intra","Category")])
max_intra_outlier = chromHMM_TE_state[which(chromHMM_TE_state$Max_states_intra > 7),]
ddply(max_intra_outlier,.(),summarise,Length=median(stop-start),Tissues=median(Tissues)/127)
max_intra_outlier = merge(max_intra_outlier,rmsk_TE[,c(TE_coordinates,cohorts)],by=TE_coordinates,all.x=TRUE)
max_intra_outlier
write.table(max_intra_outlier[,c(TE_coordinates,"class_update",cohorts)],file="States_intra_8_TEs.bed",sep='\t',row.names=FALSE,quote=FALSE)

# Number of TEs per category
table(chromHMM_TE_state$Category)/NUM_TE

# Number of TEs overlapping bin summits by class
table(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit"),]$class_update)

## Max states per TE
ddply(chromHMM_TE_state,.(Category),summarise,Mean=mean(Max_states_intra),SD=sd(Max_states_intra),Median=median(Max_states_intra),IQR(Max_states_intra))
ddply(chromHMM_TE_state,.(class_update,Category),summarise,Count=length(Max_states_intra),Mean=mean(Max_states_intra),SD=sd(Max_states_intra),Median=median(Max_states_intra),IQR(Max_states_intra))

## TEs with >1 max states in a sample
dim(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit" & chromHMM_TE_state$Max_states_intra > 1),])[1]
dim(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit" & chromHMM_TE_state$Max_states_intra > 1),])[1]/dim(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit"),])[1]

## Correlation of TE length with max in one sample
by(data=chromHMM_TE_state,chromHMM_TE_state$Category,function(x) cor.test(x$stop-x$start,x$Max_states_intra,method="spearman"))

# WGBS
test = read.table("WGBS/TE_CpG_state_counts.txt",sep='\t')
test$Proportion = test$V2/sum(test$V2)
test
sum(test$Proportion[2:4])

## Proportion of those with >2 states
sum(test$Proportion[3:4])/sum(test$Proportion)
```

### DNase/H3K27ac peak analysis

```{r peak analysis, cache=TRUE}
# Number of DNase peaks per TE
d = table(unlist(TE_DNase_peaks[,8:60]))
d

# % of TE x sample with 1 peak
d[2]/sum(d[2:length(d)])

# Number of H3K27ac peaks per TE
e = table(unlist(TE_H3K27ac_peaks[,8:105]))
e

# % of TE x sample with 1 peak
e[2]/sum(e[2:length(e)])
```

### Supplementary Figure 9 (shuffled state switching)

```{bash shuffled dynamics}
# Inter state switching for shuffled TEs
## chromHMM
for i in {1..10}; do python ~/bin/TE_landscape/state_sharing_inter.py rmsk_TE_shuffle_$i\_chromHMM_sorted.txt chromHMM_states.txt rmsk_TE_shuffle_$i\_chromHMM_inter.txt 9 7; done &

## WGBS
for i in {1..10}; do python ~/bin/TE_landscape/state_sharing_inter_meth.py WGBS/shuffled/rmsk_TE_shuffle_$i\_Meth_average.txt WGBS/shuffled/TE_CpG_count_$i\.txt WGBS/shuffled/TE_shuffle_$i\_inter.txt; done
```

```{r shuffled dynamics script, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/shuffled_state_switching.R")
```

```{r Figure S9, echo=FALSE}
iteration_labels = unlist(setNames(lapply(seq(1,10,1),function(x) paste("Iteration ",x,sep="")),seq(1,10,1)))

a = ggplot(shuffled_chromHMM_inter,aes(x=State2,y=State1,fill=Mean_samples/sample_counts["All","chromHMM"])) + geom_tile() +
  theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.title = element_blank()) + scale_y_discrete(limits=rev(chromHMM_states)) + scale_x_discrete(labels=chromHMM_states) + scale_fill_gradient(low="white",high="darkmagenta",limits=c(0,1),guide=FALSE) + coord_equal() + facet_wrap(~Iteration,nrow=2,labeller=labeller(Iteration=iteration_labels)) + xlab("State 2") + ylab("State 1")

b = ggplot(shuffled_meth_inter,aes(x=State2,y=State1,fill=Mean_samples/sample_counts["All","WGBS"])) + geom_tile() +
  theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.position="bottom") + scale_y_discrete(limits=rev(meth_states)) + 
  scale_fill_gradient(name="Average proportion\nof samples in state",low="white",high="darkmagenta",limits=c(0,1)) + coord_equal() + facet_wrap(~Iteration,nrow=2,labeller=labeller(Iteration=iteration_labels)) + xlab("State 2") + ylab("State 1")

grid.arrange(a,b)
```

```{r shuffled dynamics analysis}
## Median total states per TE
table(shuffled_chromHMM_states$States,shuffled_chromHMM_states$Iteration)
ddply(shuffled_chromHMM_states,.(Iteration),summarise,Median=median(States),IQR=IQR(States))
ddply(shuffled_chromHMM_states,.(Iteration,class),summarise,Median=median(States),IQR=IQR(States))

table(shuffled_WGBS_states$States,shuffled_WGBS_states$Iteration)
ddply(shuffled_WGBS_states,.(Iteration),summarise,Median=median(States),IQR=IQR(States))
ddply(shuffled_WGBS_states,.(Iteration,class),summarise,Median=median(States),IQR=IQR(States))

# Matrices
shuffled_chromHMM_inter
ddply(shuffled_chromHMM_inter,.(State1,State2),summarise,Proportion=mean(Mean_samples/sample_counts["All","chromHMM"]),SD=sd(Mean_samples/sample_counts["All","chromHMM"]))

shuffled_meth_inter

## TEs ever in poised promoter states
ldply(shuffled_chromHMM_potential,function(x) dim(x[which(x$X10_TssBiv > 0 | x$X11_BivFlnk > 0),])[1]/NUM_TE)

# Dynamics
shuffled_chromHMM_dynamics
shuffled_meth_dynamics

rm(list=c("shuffled_chromHMM_potential","shuffled_WGBS_average"))
```

### Supplementary Figure 10

Across all TEs x sample with DHS information, 2% overlap a peak summit; for H3K27ac, it is 0.7%. If they were independent, the expectation would be 0.02% of TEs overlapping both peaks. If I restrict the comparison to TEs x sample with both DHS and H3K27ac information (by default also chromHMM), the numbers become 2% and 0.8% (essentially the same), but the number of TEs x sample with both is 0.2%, 10X the expectation. So even without restricting to 1_TssA, there is already an enrichment between these two states, which we knew from the Chi-squared analyses.

For TEs also annotated with 1_TssA, the numbers overlapping DHS and H3K27ac peaks are 20% and 11%, which makes the number of TEs overlapping both (4%) only ~twice the expectation. However, this is still 20X the amount seen for all TEs.

```{bash compare marks, eval=FALSE}
# Combine multiple epigenetic marks
# Updated 5/7/18 and 5/22/18 with pandas script

# Combined matrix with each TE x sample and chromHMM, WGBS, DNase, H3K27ac, and RNA states, by sample
# WGBS and RNA matrices are output by WGBS_TE_avg_methylation.R and RNAseq.R
ln -s /bar/epehrsson/TE_landscape/chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt chromHMM/.
ln -s /bar/epehrsson/TE_landscape/WGBS/TE_WGBS_state_sorted.txt WGBS/.
ln -s /bar/epehrsson/TE_landscape/DNase/rmsk_TEother_DNase_summit.txt DNase/.
ln -s /bar/epehrsson/TE_landscape/H3K27ac/rmsk_TEother_H3K27ac_summit.txt H3K27ac/.
ln -s /bar/epehrsson/TE_landscape/RNAseq/rmsk_TE_rpkm.txt RNA/. 

awk '{print>"chromHMM/"$8}' rmsk_TEother_chromHMM_summit_sorted.txt &
awk '{print>"WGBS/"$8}' TE_WGBS_state_sorted.txt &
awk '{print>"H3K27ac/"$8}' rmsk_TEother_H3K27ac_summit.txt &
awk '{print>"DNase/"$8}' rmsk_TEother_DNase_summit.txt &
awk '{print>"RNA/"$8}' rmsk_TE_rpkm.txt & 

# Input: TE states by sample, split into a folder for each metric
# Output: Single combined file per sample
python combine_marks_pandas.py ~/TE_landscape/sample_lists/mnemonics.txt &

# Count number of TEs x sample in each state combination (all, unique)
# Python script outputs two files (hardcoded)
#TE_landscape/compare_marks/combine_marks_counts.txt
#TE_landscape/compare_marks/combine_marks_counts_unique.txt
python aggregate_marks_pandas.py mnemonics.txt
```

```{r Figure S10, echo=FALSE}
source("R_scripts/compare_marks.R")

a = ggplot(by_chromHMM,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~chromHMM,labeller=labeller(Mark=mark_labels_simple,chromHMM=label_wrap_gen())) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),strip.text.x = element_text(size = 7), plot.title=element_text(size=8)) + ylab("Proportion of TE x sample") + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) + ggtitle("chromHMM state")

b = ggplot(by_WGBS,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~WGBS,labeller=labeller(WGBS=setNames(c("Hypo","Inter","Hyper","Missing"),meth_states),Mark=mark_labels_simple)) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.title=element_text(size=8),strip.text.x = element_text(size = 7)) + ylab("Proportion of TE x sample") + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) + ggtitle("WGBS state")

c = ggplot(by_DNase,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~DNase,labeller=labeller(Mark=mark_labels_simple)) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),axis.title.y=element_blank(), plot.title=element_text(size=8)) + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) + ggtitle("DHS peak overlap")

d = ggplot(by_H3K27ac,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~H3K27ac,labeller=labeller(Mark=mark_labels_simple)) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),axis.title.y=element_blank(), plot.title=element_text(size=8)) + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) + ggtitle("H3K27ac peak overlap")

e = ggplot(by_RNA,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~RNA,labeller=labeller(Mark=mark_labels_simple)) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),axis.title.y=element_blank(), plot.title=element_text(size=8)) + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) + ggtitle("RPKM > 1")

legend_states = get_legend(ggplot(by_DNase,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  theme(legend.position = "bottom",legend.key.size = unit(3,'mm')) + scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),labels=c(all_state_labels,"True","False")) + guides(fill=guide_legend(ncol=7,title.position = "top")))

grid.arrange(a,b,c,d,e,legend_states, nrow=3, layout_matrix=rbind(c(1),c(2,3,4,5),c(6)),widths=c(0.36,0.21,0.21,0.22),heights=c(0.45,0.45,0.1))
```

### Compare mark analysis (get p-values and residuals)

```{r compare mark analysis}
# Number of samples with each combination of data
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$DNase)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$H3K27ac)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$DNase) & !is.na(metadata$H3K27ac)),])[1]
dim(metadata[which(!is.na(metadata$DNase) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$H3K27ac) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$DNase) & !is.na(metadata$H3K27ac) & !is.na(metadata$RNA)),])[1]

# Tables for each mark
ddply(WGBS_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(DNase_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(H3K27ac_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(RNA_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(compare_marks_all,.(chromHMM),summarise,TE_sample=sum(TE_sample),Proportion=sum(TE_sample)/sum(compare_marks_unique$TE_sample))

# Proportion of each state in other states
by_chromHMM
by_WGBS
by_DNase
by_H3K27ac
by_RNA

# Chi-squared for each mark combination (chromHMM not normalized)
table_list = list(xtabs(TE_sample~chromHMM+WGBS,compare_marks_all),
                  xtabs(TE_sample~chromHMM+DNase,compare_marks_all),
                  xtabs(TE_sample~chromHMM+H3K27ac,compare_marks_all),
                  xtabs(TE_sample~chromHMM+RNA,compare_marks_all),
                  xtabs(TE_sample~WGBS+DNase,compare_marks_unique),
                  xtabs(TE_sample~WGBS+H3K27ac,compare_marks_unique),
                  xtabs(TE_sample~WGBS+RNA,compare_marks_unique),
                  xtabs(TE_sample~DNase+H3K27ac,compare_marks_unique),
                  xtabs(TE_sample~DNase+RNA,compare_marks_unique),
                  xtabs(TE_sample~H3K27ac+RNA,compare_marks_unique))
chisq_list = lapply(table_list,function(x) chisq.test(x))
p.adjust(unlist(lapply(chisq_list,function(x) x$p.value)),method="bonf")
lapply(chisq_list,function(x) x$residuals)

# Proportion of TEs x sample overlapping a DHS peak, an H3K27ac peak, or both
test1 = ddply(compare_marks_unique[which(!is.na(compare_marks_unique$DNase) & !is.na(compare_marks_unique$H3K27ac)),],.(DNase,H3K27ac),summarise,TE_sample=sum(TE_sample))
ddply(test1,.(),summarise,DHS=sum(TE_sample[which(DNase == "True")])/sum(TE_sample),H3K=sum(TE_sample[which(H3K27ac == "True")])/sum(TE_sample),Both=sum(TE_sample[which(DNase == "True" & H3K27ac == "True")])/sum(TE_sample))

# Proportion of TEs x sample in each chromHMM state also overlapping DNase peak and H3K27ac peak
test1 = ddply(compare_marks_all[which(!is.na(compare_marks_all$DNase) & !is.na(compare_marks_all$H3K27ac)),],.(chromHMM,DNase,H3K27ac),summarise,TE_sample=sum(TE_sample))
test2 = ddply(test1,.(chromHMM),summarise,DHS=sum(TE_sample[which(DNase == "True")])/sum(TE_sample),H3K=sum(TE_sample[which(H3K27ac == "True")])/sum(TE_sample),Both=sum(TE_sample[which(DNase == "True" & H3K27ac == "True")])/sum(TE_sample))
test2[order(test2$Both),]

# Proportion of TEs x sample in each WGBS state also overlapping DNase peak and H3K27ac peak
test1 = ddply(compare_marks_unique[which(!is.na(compare_marks_unique$DNase) & !is.na(compare_marks_unique$H3K27ac) & !is.na(compare_marks_unique$WGBS)),],.(WGBS,DNase,H3K27ac),summarise,TE_sample=sum(TE_sample))
test2 = ddply(test1,.(WGBS),summarise,DHS=sum(TE_sample[which(DNase == "True")])/sum(TE_sample),H3K=sum(TE_sample[which(H3K27ac == "True")])/sum(TE_sample),Both=sum(TE_sample[which(DNase == "True" & H3K27ac == "True")])/sum(TE_sample))
test2[order(test2$Both),]

# QC
## Total TE x sample x chromHMM states = 570405028 - matches total TEs in rmsk_TEother_chromHMM_summit_sorted.txt and sum(compare_marks_all$TE_sample)
## Total TE x sample = 562520596 - matches total unique TEs (121*NUM_TE + 6*NUM_TE_noY) and sum(compare_marks_unique$TE_sample)
## Total TE x sample correct for each metric
apply(sample_counts,2,function(x) (x[3]*NUM_TE) + ((x[1]-x[3])*NUM_TE_noY))
apply(compare_marks_unique,2,function(x) sum(compare_marks_unique[which(!is.na(x)),]$TE_sample))
```

## By sample

### By-sample analysis

```{r analysis Fig 4a, cache=TRUE, cache.lazy=FALSE}
# Stats
by_sample_stats = ddply(by_sample_all,~State,summarise, Range = max(Proportion)-min(Proportion), Median = median(Proportion), Mean = mean(Proportion), CV = sd(Proportion)/mean(Proportion))
by_sample_stats[order(by_sample_stats$Range),]

ddply(by_sample_all,~State,function(x) x[which.min(x$Proportion),])
ddply(by_sample_all,~State,function(x) x[which.max(x$Proportion),])

# Kruskal-Wallis test by sample classification
# MHC adjusted for every state (not category)
by_sample_KW = ddply(by_sample_all,~State,summarise,Group = p.adjust(unlist(kruskal.test(Proportion,Group))["p.value"],method="bonf"),
      Anatomy = p.adjust(unlist(kruskal.test(Proportion,Anatomy))["p.value"],method="bonf"),
      Type = p.adjust(unlist(kruskal.test(Proportion,Type))["p.value"],method="bonf"),
      Age = p.adjust(unlist(kruskal.test(Proportion,Age))["p.value"],method="bonf"),
      Germline = p.adjust(unlist(kruskal.test(Proportion,Germline))["p.value"],method="bonf"))
by_sample_KW = merge(by_sample_KW,ddply(droplevels(by_sample_all[which(!(by_sample_all$State %in% meth_states)),]),~State,summarise,Cancer = p.adjust(unlist(kruskal.test(Proportion,Cancer))["p.value"],method="bonf")),by="State",all.x=TRUE)
by_sample_KW = melt(by_sample_KW,id.var="State")
colnames(by_sample_KW)[2:3] = c("Category","Pvalue")
by_sample_KW[which(by_sample_KW$Pvalue < 0.05),]

# Permutation tests by sample classification
# MHC adjusted for category and state, not direction
permute_up = ddply(by_sample_all[,c("State","Sample","Proportion","Contribution")],.(State),function(y) permute_by_sample(y,"Proportion","+",unique(y$Contribution),"Proportion",-1))
permute_up = permute_up[which(permute_up$Samples.x > 0),]
permute_up$Padjust = p.adjust(permute_up$Pvalue,method="fdr")
permute_up[which(permute_up$Padjust < 0.1 & permute_up$Category == "Group" & permute_up$State %in% states[c(1:7,16:17,20:21)]),]
write.table(permute_up,file="sample_permute_up.txt",sep='\t',row.names=FALSE,quote=FALSE)

permute_down = ddply(by_sample_all[,c("State","Sample","Proportion","Contribution")],.(State),function(y) permute_by_sample(y,"Proportion","-",unique(y$Contribution),"Proportion",-1))
permute_down = permute_down[which(permute_down$Samples.y - permute_down$Samples.x > 0),]
permute_down$Padjust = p.adjust(permute_down$Pvalue,method="fdr")
permute_down[which(permute_down$Padjust < 0.1 & permute_down$Category == "Group" & permute_down$State %in% states[c(1:7,16:17,20:21)]),]
write.table(permute_down,file="sample_permute_down.txt",sep='\t',row.names=FALSE,quote=FALSE)

# Mean Group proportion by state 
test = aggregate(data=by_sample_all,Proportion~State+Group,mean)
test[order(test$State,test$Proportion),]
```

### Figure 4

```{bash RNA coverage, eval=FALSE}
intersect_RNA_features.sh
intersect_RNA_chrY.sh
```

```{r Figure 4, echo=FALSE}
#source("R_scripts/proportion.R")
#source("R_scripts/proportion_class.R")
source("R_scripts/RNA_coverage.R")

# add_stars(c(1:9,11:15,20:21,16,18),round(ddply(by_sample_all,.(State),summarise,Max=max(Proportion))$Max[c(1:9,11:15,20:21,16,18)]+0.05,2),size=6)
by_sample_KW$Stars = ifelse(by_sample_KW$Pvalue < 0.05,"*","")

a = ggplot(by_sample_all,aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.8),size=1) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x = element_blank(),axis.ticks.x = element_blank(),strip.background = element_rect(fill=NA,color=NA)) + ylab("Proportion of state in TEs") + scale_color_manual(values=group_colors,guide=FALSE) + geom_errorbar(aes(ymax=Contribution,ymin=Contribution),position=position_dodge(),color="black") + geom_hline(yintercept = MERGED_TE_WIDTH/GENOME_WIDTH,linetype="dashed") + geom_hline(yintercept = TE_CPGS/ALL_CPGS,linetype="dotdash") + scale_y_continuous(limits=c(0,0.85)) + scale_x_discrete(limits=states[c(1:15,20:21,16:19)],labels=all_state_labels) + geom_vline(xintercept=15.5,linetype="dashed",color="grey") + geom_vline(xintercept=16.5,linetype="dashed",color="grey") + geom_vline(xintercept=17.5,linetype="dashed",color="grey") + geom_text(data=by_sample_KW[which(by_sample_KW$Category == "Group"),],aes(x=State,y=0.8,label=Stars),color="red",size=7) + facet_grid(rows=vars(""))

b = ggplot(by_sample_class,aes(x=State,y=Proportion_state,color=Group)) + geom_point(position = position_jitterdodge(dodge.width = 0.8),size=1) + theme(panel.grid=element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.title.x = element_text(margin = margin(t = -10))) + ylab("Proportion of state in class") + scale_color_manual(values=group_colors,guide=FALSE) + scale_x_discrete(limits=states[c(1:15,20:21,16:19)],labels=all_state_labels) + geom_errorbar(aes(ymax=Contribution,ymin=Contribution),position=position_dodge(),color="black") + facet_grid(rows=vars(class),scales="free") + geom_hline(data=by_sample_class[which(by_sample_class$State == "1_TssA" & by_sample_class$Sample == "E001"),],aes(yintercept=Bases_class/GENOME_WIDTH),linetype="dashed") + geom_hline(data=by_sample_class[which(by_sample_class$State == "Hypomethylated" & by_sample_class$Sample == "E003"),],aes(yintercept=Bases_class/ALL_CPGS),linetype="dotdash") + geom_vline(xintercept=15.5,linetype="dashed",color="grey") + geom_vline(xintercept=16.5,linetype="dashed",color="grey") + geom_vline(xintercept=17.5,linetype="dashed",color="grey") + scale_y_continuous(limits=c(0,0.6))

# Average expression of TEs vs. genome
c = ggplot(RNA_proportion,aes(x=1,y=Ratio,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.5),size=1) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x = element_blank(),axis.ticks.x = element_blank(),strip.background = element_rect(fill=NA,color=NA)) + scale_color_manual(values=group_colors,guide=FALSE) + ylab("TE:Genome\nexpression") + scale_y_continuous(limits=c(0,1)) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black") + facet_grid(~"")

d = ggplot(class_RNA,aes(x=1,y=Ratio,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.5),size=1) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x = element_blank(),axis.ticks.x = element_blank()) + scale_color_manual(values=group_colors,guide=FALSE) + ylab("Class:Genome\nexpression") + ylim(0,1) + facet_wrap(~class_update,nrow=1) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black")

legend = get_legend(ggplot(by_sample_all,aes(x=State,y=Proportion,color=Group)) + geom_point(position = position_jitter(width = 0.3),size=2) + scale_color_manual(values=group_colors) + theme(legend.margin=margin(0,0,0,0),legend.key.size = unit(5,'mm')))

grid.arrange(a,b,c,d,legend,nrow=3,ncol=3,layout_matrix=rbind(c(1,1,5),c(2,2,5),c(3,4,5)),heights=c(0.2,0.55,0.15),widths=c(0.2,0.65,0.15))
```

### By sample and class analysis

```{r analysis Fig S11a, cache=TRUE, cache.lazy=FALSE}
by_sample_stats_class = ddply(na.omit(by_sample_class),.(State,class),summarise, Range = max(Proportion_state)-min(Proportion_state), Median = median(Proportion_state), Mean = mean(Proportion_state),CV = sd(Proportion_state)/mean(Proportion_state))
by_sample_stats_class[order(by_sample_stats_class$Range),]
by_sample_stats_class[order(by_sample_stats_class$CV),]

# Kruskal-Wallis
# MHC adjusted for state and class
by_sample_KW_class = ddply(by_sample_class,.(State,class),summarise,Group = p.adjust(unlist(kruskal.test(Proportion_state,Group))["p.value"],method="bonf"))
by_sample_KW_class[which(by_sample_KW_class$Group < 0.05),]

# Number of enriched samples by Group
# MHC adjusted for state and class
permute_up = ddply(by_sample_class[,c("class","State","Sample","Proportion_state","Contribution")],.(State,class),function(y) permute_by_sample(y,"Proportion_state","+",unique(y$Contribution),"Proportion_state",-1))
permute_up = permute_up[which(permute_up$Samples.x > 0),]
permute_up$Padjust = p.adjust(permute_up$Pvalue,method="fdr")
permute_up[which(permute_up$Padjust < 0.1 & permute_up$Category == "Group" & permute_up$State %in% states[c(1:7,16:17,20:21)]),]

permute_down = ddply(by_sample_class[,c("class","State","Sample","Proportion_state","Contribution")],.(State,class),function(y) permute_by_sample(y,"Proportion_state","-",unique(y$Contribution),"Proportion_state",-1))
permute_down = permute_down[which(permute_down$Samples.y - permute_down$Samples.x > 0),]
permute_down$Padjust = p.adjust(permute_down$Pvalue,method="fdr")
permute_down[which(permute_down$Padjust < 0.1 & permute_down$Category == "Group" & permute_down$State %in% states[c(1:7,16:17,20:21)]),]

test = ddply(by_sample_class,.(State,class,Group),summarise,Proportion_state=mean(Proportion_state),Contribution=unique(Contribution))
test[order(test$State,test$class,test$Proportion),]

# Enrichments over expectation
table(by_sample_class[which(by_sample_class$Enrichment > 0),c("State","class")])

by_sample_class[which(by_sample_class$class %in% c("DNA","LINE","SINE","LTR") & by_sample_class$State %in% states[c(1:3,6:7,16,20:21)] & by_sample_class$Enrichment > 0),c("Sample","class","State","Enrichment",sample_categories)]
```

### DNase/H3K27ac peak analysis (vs. bp)

```{r peak analysis pt2}
# % of DNase peaks overlapping TEs, overall
sum(DNase_stats$Summit_in_TE)/sum(DNase_stats$Peaks)

## By sample
ddply(DNase_stats,.(),summarise,Median=median(Summit_in_TE/Peaks),Mean=mean(Summit_in_TE/Peaks),SD=sd(Summit_in_TE/Peaks),Min=min(Summit_in_TE/Peaks),Max=max(Summit_in_TE/Peaks))
DNase_stats[which.min(DNase_stats$Summit_in_TE/DNase_stats$Peaks),]
DNase_stats[which.max(DNase_stats$Summit_in_TE/DNase_stats$Peaks),]

# % of H3K27ac peaks overlapping TEs, overall
sum(H3K27ac_stats$Summit_in_TE)/sum(H3K27ac_stats$Peaks)

## By sample
ddply(H3K27ac_stats,.(),summarise,Median=median(Summit_in_TE/Peaks),Mean=mean(Summit_in_TE/Peaks),SD=sd(Summit_in_TE/Peaks),Min=min(Summit_in_TE/Peaks),Max=max(Summit_in_TE/Peaks))
H3K27ac_stats[which.min(H3K27ac_stats$Summit_in_TE/H3K27ac_stats$Peaks),]
H3K27ac_stats[which.max(H3K27ac_stats$Summit_in_TE/H3K27ac_stats$Peaks),]
```

### Expression analysis

```{r ratio expression}
range(RNA_proportion$Ratio)
median(RNA_proportion$Ratio)
ddply(RNA_proportion,.(Group),summarise,Median=median(Ratio))

ddply(class_RNA,.(class_update),summarise,Median=median(Ratio))
dcast(ddply(class_RNA,.(Group,class_update),summarise,Median=median(Ratio)),class_update~Group,value.var="Median")
```

### Supplementary Figure 11

```{r Figure S11, echo=FALSE}
# Other sample classifications
by_sample_all_long = melt(by_sample_all[,c("State","Sample","Proportion","Contribution",sample_categories[2:6])],id.vars=c("State","Sample","Proportion","Contribution"))
colnames(by_sample_all_long)[5:6] = c("Category","Group")

a = ggplot(by_sample_all_long,aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6),size=1) + theme(panel.grid=element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.title.x = element_text(margin = margin(t = -10))) + ylab("Proportion of state in TEs") + scale_color_manual(values=c(anatomy_colors,type_colors,age_colors,cancer_colors,germline_colors),guide=FALSE) + geom_errorbar(aes(ymax=Contribution,ymin=Contribution),position=position_dodge(),color="black") + geom_hline(yintercept = MERGED_TE_WIDTH/GENOME_WIDTH,linetype="dashed") + geom_hline(yintercept = TE_CPGS/ALL_CPGS,linetype="dotdash") + scale_y_continuous(limits=c(0,0.85)) + scale_x_discrete(limits=states[c(1:15,20:21,16:19)],labels=all_state_labels) + geom_vline(xintercept=15.5,linetype="dashed",color="grey") + geom_vline(xintercept=16.5,linetype="dashed",color="grey") + geom_vline(xintercept=17.5,linetype="dashed",color="grey") + facet_wrap(~Category,nrow=5,labeller=labeller(Category=category_labels)) + geom_text(data=by_sample_KW[which(by_sample_KW$Category != "Group"),],aes(x=State,y=0.8,label=Stars),color="red",size=7)

legend_anatomy = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Anatomy"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(anatomy_colors),name="Anatomy")) 

legend_type = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Type"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(type_colors),name="Type") + guides(color = guide_legend(nrow = 2)))

legend_age = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Age"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(age_colors),name="Age",labels=age_labels) + guides(color = guide_legend(nrow = 2)))

legend_cancer = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Cancer"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(cancer_colors),name="Cancer"))

legend_germline = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Germline"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(germline_colors),name="Germlayer"))

grid.arrange(a,legend_anatomy,legend_type,legend_age,legend_cancer,legend_germline,nrow=4,layout_matrix=rbind(c(1,1),c(2,2),c(3,4),c(6,5)),heights=c(0.8,0.1,0.05,0.05),widths=c(0.6,0.4))
```

## Subfamily enrichment

### Set thresholds 

```{r set enrichment thresholds}
THRESHOLD_IJK_MEMBER = 10
THRESHOLD_IK_MEMBER = 30
THRESHOLD_LOR = 1.5
THRESHOLD_PC = 0.01

THRESHOLD_IJK_BASE = 600
THRESHOLD_IK_BASE = 5000
THRESHOLD_IJK_CPG = 6
THRESHOLD_IK_CPG = 50
```

### Figure 5

```{bash subfamily chromHMM, eval=FALSE}
# Intersect with merged TE subfamilies
#TE_landscape/chromHMM/subfamily/subfamily_state_sample.bed
for file in ../chromHMM_bedfiles/E*.bed; do suffix=$(basename $file | cut -d '_' -f1); bedtools intersect -wo -a TEother_subfamily_merge.txt -b $file | awk -v OFS='\t' -v tag=$suffix '{print $0, tag}' - >> subfamily_state_sample.bed; done

#TE_landscape/chromHMM/subfamily/subfamily_state_sample.txt
awk -v OFS='\t' '{a[$4, $8, $10]+=$9;}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], a[i];}}' subfamily_state_sample.bed > subfamily_state_sample.txt
```

```{bash subfamily chromHMM members, eval=FALSE}
# By subfamily
awk -v OFS='\t' '{a[$4, $8, $10]+=1}END{for(i in a){split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], a[i];}}' chromHMM/rmsk_TEother_chromHMM_summit_sorted.txt > chromHMM/subfamily/subfamily_state_sample_summit.txt
```

```{bash subfamily DHS, eval=FALSE}
# Merged TE subfamilies (unique peaks overlapping subfamily)
while read line; do while read line2; do awk -v sample=$line -v subfam=$line2 -v OFS='\t' '{if($4 == subfam) a[$8, $9, $10]+=1}END{print sample, subfam, length(a)}' DNase/summit/TEs/rmsk_TEother_$line\_DNase_summit.txt >> DNase/subfamily_DNase_sample_summit.txt; done < features/TEs/subfamily/subfamilies.txt; done < sample_lists/DNase_samples.txt
```

```{bash subfamily H3K27ac, eval=FALSE}
# Merged TE subfamilies (unique peaks overlapping subfamily)
while read line; do while read line2; do awk -v sample=$line -v subfam=$line2 -v OFS='\t' '{if($4 == subfam) a[$8, $9, $10]+=1}END{print sample, subfam, length(a)}' H3K27ac/summit/TEs/rmsk_TEother_$line\_H3K27ac_summit.txt >> H3K27ac/subfamily_H3K27ac_sample_summit.txt; done < features/TEs/subfamily/subfamilies.txt; done < sample_lists/H3K27ac_samples.txt
```

```{bash subfamily WGBS, eval=FALSE}
# Number of CpGs in each state, TE CpGs, by sample x subfamily
#TE_landscape/WGBS/subfamily_CpG_Meth_states.txt	
while read line; do awk -v OFS='\t' -v subfam=$line '{if($4 == subfam) print $0}' TE_CpG_Meth_new.bed | cut -f8- - | sort | uniq | awk -v OFS='\t' -v subfam=$line '{for (i=4;i<=40;i++){if($i == -1) miss[i]+=1; else if ($i < 0.3) hypo[i]+=1; else if ($i > 0.7) hyper[i]+=1; else if (($i <= 0.7) && ($i >= 0.3)) inter[i]+=1;}}; END{for (i=4;i<=40;i++){print i, hypo[i], inter[i], hyper[i], miss[i], subfam;}}' -; done < ../features/TEs/subfamily/subfamilies.txt >> subfamily_CpG_Meth_states.txt
```

```{bash subfamily shuffle, eval=FALSE}
# Merged each subfamily and sorted
for i in {1..10}; do awk '{print>$4}' features/shuffled_TEs/rmsk_TE_shuffle_$i\.txt; while read line; do bedtools merge -i $line | awk -v OFS='\t' -v subfam=$line '{print $0, subfam}' - >> features/shuffled_TEs/subfamily/subfamily_merge_$i\.txt; rm $line; done < features/TEs/subfamily/subfamilies.txt; done
for i in {1..10}; do sort -k1,1 -k2,2n features/shuffled_TEs/subfamily/subfamily_merge_$i\.txt > features/shuffled_TEs/subfamily/subfamily_merge_$i\_sorted.txt; mv features/shuffled_TEs/subfamily/subfamily_merge_$i\_sorted.txt features/shuffled_TEs/subfamily/subfamily_merge_$i\.txt; done

## chromHMM
# Intersect with epigenetic data
for i in {1..10}; do while read line; do bedtools intersect -wo -a features/shuffled_TEs/subfamily/subfamily_merge_$i\.txt -b raw_data/chromHMM/$line\_15_coreMarks_mnemonics.bed | awk -v OFS='\t' -v sample=$line '{print $0, sample}' - >> chromHMM/shuffled_TEs/subfamily/subfamily_state_sample_$i\.bed; done < sample_lists/mnemonics.txt; done

# Number of bp per subfamily x sample
for j in {1..10}; do awk -v OFS='\t' '{a[$4, $8, $10]+=$9}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], a[i]}}' chromHMM/shuffled_TEs/subfamily/subfamily_state_sample_$j\.bed > chromHMM/shuffled_TEs/subfamily/subfamily_state_sample_$j\.txt; done

# Number of subfamily members in state per sample
for j in {1..10}; do awk -v OFS='\t' '{a[$4, $8, $10]+=1}END{for(i in a){split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], a[i]}}' chromHMM/shuffled_TEs/rmsk_TE_shuffle_$j\_chromHMM_sorted.txt > chromHMM/shuffled_TEs/subfamily/subfamily_state_sample_summit_$j\.txt; done

## WGBS
# Intersect with epigenetic data
split -l 1000000 ~/TE_landscape/WGBS/all_CpG_Meth.bed
for i in {1..10}; do for file in x*; do echo $file; bedtools intersect -wo -a ~/TE_landscape/features/shuffled_TEs/subfamily/subfamily_merge_$i\.txt -b $file >>  subfam_CpG_Meth_$i\.bed ; done; done

# Number of CpGs in state per subfamily x sample
for j in {1..10}; do awk '{print>"indv/"$4}' subfam_CpG_Meth_$j\.bed; while read line; do awk -v OFS='\t' -v subfam=$line '{for (i=8;i<=44;i++){if($i == -1) miss[i]+=1; else if ($i < 0.3) hypo[i]+=1; else if ($i > 0.7) hyper[i]+=1; else if (($i <= 0.7) && ($i >= 0.3)) inter[i]+=1;}}END{for (i=8;i<=44;i++){print i, hypo[i], inter[i], hyper[i], miss[i], subfam}}' indv/$line; rm indv/$line; done < ~/TE_landscape/sample_lists/subfamilies.txt >> subfamily_CpG_Meth_states_$j\.txt; done &

# Number of CpGs per subfamily (Length_ik)
for j in {1..10}; do awk -v OFS='\t' '{a[$4]+=1}END{for(i in a){print i, a[i]}}' /scratch/ecp/TE_landscape/WGBS/subfam_CpG_Meth_$j\.bed > ~/TE_landscape/WGBS/shuffled/subfamily/subfamily_CpG_count_$j\.txt; done

## DHS
# Intersect with epigenetic data
for i in {1..10}; do while read line; do bedtools intersect -wo -a features/shuffled_TEs/subfamily/subfamily_merge_$i\.txt -b raw_data/DNase/DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak | awk -v OFS='\t' -v sample=$line '{print $0, sample}' - >> DNase/shuffled/subfamily/rmsk_TEother_subfamily_DNase_$i\.txt ; done < sample_lists/DNase_samples.txt; done

# Filtered to only peaks where the summit overlaps the subfamily
for i in {1..10}; do python ~/bin/TE_landscape/filter_summit.py DNase/shuffled/subfamily/rmsk_TEother_subfamily_DNase_$i\.txt 1 5 DNase/shuffled/subfamily/rmsk_TEother_subfamily_DNase_$i\_summit.txt DNase; done

# Number of peaks per subfamily x sample
for j in {1..10}; do awk -v OFS='\t' '{a[$4, $16]+=1}END{for(i in a) {split(i, sep, SUBSEP); print sep[1], sep[2], a[i]}}' DNase/shuffled/subfamily/rmsk_TEother_subfamily_DNase_$j\_summit.txt > DNase/shuffled/subfamily/subfamily_DNase_sample_summit_$j\.txt; done

## H3K27ac
# Intersect with epigenetic data
for i in {1..10}; do while read line; do bedtools intersect -wo -a features/shuffled_TEs/subfamily/subfamily_merge_$i\.txt -b raw_data/H3K27ac/H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak | awk -v OFS='\t' -v sample=$line '{print $0, sample}' - >> H3K27ac/shuffled/subfamily/rmsk_TEother_subfamily_H3K27ac_$i\.txt ; done < sample_lists/H3K27ac_samples.txt; done

# Filtered to only peaks where the summit overlaps the subfamily
for i in {1..10}; do python ~/bin/TE_landscape/filter_summit.py H3K27ac/shuffled/subfamily/rmsk_TEother_subfamily_H3K27ac_$i\.txt 1 5 H3K27ac/shuffled/subfamily/rmsk_TEother_subfamily_H3K27ac_$i\_summit.txt H3K27ac; done

# Number of peaks per subfamily x sample
for j in {1..10}; do awk -v OFS='\t' '{a[$4, $16]+=1}END{for(i in a) {split(i, sep, SUBSEP); print sep[1], sep[2], a[i]}}' H3K27ac/shuffled/subfamily/rmsk_TEother_subfamily_H3K27ac_$j\_summit.txt > H3K27ac/shuffled/subfamily/subfamily_H3K27ac_sample_summit_$j\.txt; done
```

''enrichment_list.txt'' generated from results of analysis in this section (subfamilies of interest).

```{bash HOMER, eval=FALSE}
bash run_homer.sh enrichment/enrichment_list.txt
```

```{r Figure 5 scripts, echo=FALSE, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/WGBS_subfamily_enrichment_TE.R")
source("R_scripts/subfamily_enrichment.R") #Requires proportion.R
source("R_scripts/subfamily_potential.R")
source("R_scripts/subfamily_PCA.R")
source("R_scripts/subfamily_potential_shuffle.R")
source("R_scripts/shuffled_subfamily_enrichment.R")
```

```{r Figure 5, echo=FALSE}
# Potential
a = ggplot(subfam_states[which(subfam_states$variable == "States"),],aes(x=value)) + geom_histogram(binwidth=1,fill="black") + ylab("Subfamilies") + xlab("Number of states") + facet_wrap(~Metric,scales="free_x",nrow=1,labeller=labeller(Metric=mark_labels_simple)) + scale_y_continuous(limits=c(0,968)) + expand_limits(x=0)

# Clustering
b = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=age_labels) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep="")) + theme(aspect.ratio = 1,legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm'))

family_shapes=setNames(c(17,rep(16,20),18,rep(16,23)),sort(as.vector(unique(rmsk_TE_subfamily$family))))
subfamily_chromHMM_pca$eigenvectors$family_shape = ifelse(subfamily_chromHMM_pca$eigenvectors$family == "L1","L1",ifelse(subfamily_chromHMM_pca$eigenvectors$family == "Alu","Alu","Other"))

c = ggplot(subfamily_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family_shape)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,name="Family") + labs(x=paste("PC1 (",round(subfamily_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(subfamily_chromHMM_pca$eigenvalues[2],1),"%)",sep="")) + theme(aspect.ratio = 1,legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) 

# Enrichment
d = ggplot(subfamily_state_sample_counts,aes(x=State,y=Sample.Proportion,color=class_update)) + geom_point(position=position_jitterdodge(jitter.width=0.5)) + coord_flip() + ylab("Proportion of samples enriched LOR > 1.5") + scale_color_manual(values=class_colors,guide=FALSE) + geom_vline(xintercept=seq(1.5,20.5,1), colour='grey') + scale_x_discrete(limits = rev(states[c(1:15,20:21,16:19)]),labels=setNames(paste(subfamily_state_sample_counts_combine$State,"\n(n=",subfamily_state_sample_counts_combine$V1,")",sep=""),subfamily_state_sample_counts_combine$State)) + theme(axis.title.y = element_text(margin = margin(r = -10)))

# HOMER
homer = read.table("enrichment/homer/HOMER_LTR22A_7_Enh_enriched/knownResults.txt",sep='\t',header=TRUE,comment.char = "@")
homer_top5 = head(homer[order(homer$q.value..Benjamini.),],n=5)
homer_top5 = melt(homer_top5[,c(1,5,7,9)],id.vars=c("Motif.Name","q.value..Benjamini."))
homer_top5$PC = as.numeric(gsub("%","",homer_top5$value))

e = ggplot(homer_top5,aes(x=reorder(Motif.Name,`q.value..Benjamini.`,median),y=PC,fill=variable)) + geom_bar(stat="identity",position="dodge") + ylab("% TEs with motif") + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.position=c(0.1,0.9),legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm'),legend.title = element_blank()) + scale_x_discrete(labels=c("Nkx2.1","Nkx2.2","Tcf3","Fli1","HNF4a")) + xlab("Transcription factor") + scale_fill_discrete(labels=c("Active TEs","Inactive TEs"))

legend_group = get_legend(ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group)) + geom_point() + scale_color_manual(values=group_colors) + theme(aspect.ratio = 1,legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + guides(color=guide_legend(ncol=8)))

grid.arrange(a,b,c,d,e,legend_group,legend_class,nrow=6,ncol=2,layout_matrix=rbind(c(1),c(6),c(2,4),c(3,4),c(5,4),c(5,7)),heights=c(0.2,0.05,0.25,0.25,0.2,0.05),widths=c(0.4,0.6))
```

### Number of enrichments by state and number of enriched subfamilies by state, overall and by class

```{r enrichment table}
enrichment_table = merge(as.data.frame(table(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),]$State)),subfamily_state_sample_counts_combine,by.x="Var1",by.y="State")
colnames(enrichment_table)[1:3] = c("State","Enrichments","Subfamilies")
enrichment_table = enrichment_table[match(states[1:21],enrichment_table$State),]

sum(enrichment_table$Enrichments)
enrichment_table$Enrichments_pc = enrichment_table$Enrichments/sum(enrichment_table$Enrichments)
enrichment_table

# Number of subfamilies enriched at least once, overall and active states
length(unique(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),]$subfamily))
length(unique(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR & subfamily_state_sample_filter$State %in% c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")),]$subfamily))

# Enriched classes
test = dcast(aggregate(data=subfamily_state_sample_counts,V1~State+class_update,function(x) sum(x > 0)),class_update~State,value.var = "V1")
table(rmsk_TE_subfamily$class_update)/968
ddply(enrichment_table,.(State),function(x) as.numeric(x[4:9])/sum(as.numeric(x[4:9])))
apply(test[,2:22],2,function(x) log2((x/sum(x))/(table(rmsk_TE_subfamily$class_update)/968)))

# Enriched families
test = dcast(aggregate(data=subfamily_state_sample_counts,V1~State+family,function(x) sum(x > 0)),family~State,value.var = "V1")
table(rmsk_TE_subfamily$family)/968
test_log = as.data.frame(apply(test[,2:22],2,function(x) log2((x/sum(x))/(table(rmsk_TE_subfamily$family)/968))))
test_log$Subfamilies = as.numeric(table(rmsk_TE_subfamily$family)[rownames(test_log)])
lapply(colnames(test_log)[1:21],function(x) test_log[which(test_log[,x] > 0 & test_log$Subfamilies > 0),c(x,"Subfamilies")])
```

### Supplementary Figure 12

```{r Figure S12, echo=FALSE}
# Potential
shuffle_state_potential = ddply(subfamily_state_potential_shuffle,.(Iteration,Metric,subfamily),summarise,States=sum(Samples > 0))

a = ggplot(shuffle_state_potential,aes(x=States,fill=Iteration)) + geom_histogram(binwidth=1,position="dodge") + ylab("Subfamilies") + xlab("Number of states") + facet_wrap(~Metric,scales="free_x",nrow=1,labeller=labeller(Metric=mark_labels_simple)) + scale_y_continuous(limits=c(0,968)) + expand_limits(x=0) + scale_fill_grey()

test = subfamily_state_potential
test$Iteration = rep("Real",dim(subfamily_state_potential)[1])
test = rbind(test,subfamily_state_potential_shuffle)
test$Iteration = factor(test$Iteration,levels=c(seq(1,10,1),"Real"))
  
b = ggplot(test,aes(x=Sample.Proportion,y=..scaled..,fill=Iteration)) + geom_density(alpha=0.5) + facet_wrap(~State,labeller=labeller(State=all_state_labels)) + ylab("Density") + scale_fill_manual(values=c(rep("grey",10),"red")) + xlab("Proportion of samples in state") + theme(legend.position="bottom") + guides(fill=guide_legend(nrow=1))

# Plot number of enrichments per state for true TEs vs. shuffled TEs
c = ggplot(shuffled_enrichment_counts_combine,aes(x=State,y=Enrichments,fill=State)) + geom_boxplot() + geom_point(data=enrichment_table,aes(x=State,y=Enrichments),color="red") + ylab("Number of enrichments") +  theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(labels=all_state_labels)

d = ggplot(shuffled_enrichment_counts_combine,aes(x=State,y=Subfamilies,fill=State)) + geom_boxplot() +   geom_point(data=enrichment_table,aes(x=State,y=Subfamilies),color="red") + ylab("Number of enriched subfamilies") +  theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(labels=all_state_labels)

grid.arrange(a,b,c,d,layout_matrix=rbind(c(1),c(2),c(3,4)),nrow=3,heights=c(0.15,0.55,0.3))
```

### Subfamily potential

```{r subfam potential}
# Number of subfamilies ever in each state and median samples in state
ddply(subfamily_state_potential,.(State),summarise,Subfamilies=sum(Samples > 0),Samples=median(Samples))

# Number of states per subfamily (across all samples, max in one sample)
by(subfam_states,subfam_states[,c("variable","Metric")],function(x) table(x$value))

ddply(subfam_states,.(variable,Metric),summarise,Median=median(value))

# Outliers
subfam_states[which(subfam_states$value < 11 & subfam_states$variable == "States" & subfam_states$Metric == "chromHMM"),]
subfam_states[which(subfam_states$value < 10 & subfam_states$variable == "Intra"),]

# Correlation with length
by(subfam_states,subfam_states[,c("variable","Metric")],function(x) cor.test(x$value,x$Total_length))
by(subfam_states,subfam_states[,c("variable","Metric")],function(x) cor.test(x$value,x$CpGs))

# Distribution of samples in state
ddply(subfamily_state_potential[which(subfamily_state_potential$State %in% chromHMM_states & subfamily_state_potential$Samples >= sample_counts["All","chromHMM"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
ddply(subfamily_state_potential[which(subfamily_state_potential$State %in% meth_states & subfamily_state_potential$Samples >= sample_counts["All","WGBS"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
ddply(subfamily_state_potential[which(subfamily_state_potential$State == "DNase" & subfamily_state_potential$Samples >= sample_counts["All","DNase"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
ddply(subfamily_state_potential[which(subfamily_state_potential$State == "H3K27ac" & subfamily_state_potential$Samples >= sample_counts["All","H3K27ac"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
```

### Shuffled subfamily potential

For DNase and H3K27ac, the number of subfamilies in each state in at least one sample is 964.5 +/- 1.5 and 963 +/- 1.6, versus 967 and 964, respectively. So, they are about the same as expectation. Three iterations have one subfamily without any CpGs.

```{r subfam potential shuffled}
# Total number of states per subfamily, across samples
subfamily_potential_shuffle_ever = ddply(shuffle_state_potential,.(Iteration,Metric,States),summarise,Subfamilies=length(subfamily))
ddply(subfamily_potential_shuffle_ever,.(Metric,States),summarise,Mean=mean(Subfamilies),SD=sd(Subfamilies))

# Distribution of samples in state
ddply(subfamily_state_potential_shuffle,.(Iteration,State),summarise,Subfamilies=sum(Sample.Proportion > 0.75))
```

```{r shuffled subfamily enrichment}
ddply(shuffled_enrichment_counts_combine,.(State),summarise,Enrichments=median(Enrichments),Subfamilies=median(Subfamilies))
```

### Supplementary Figure 13

```{r Figure S13, echo=FALSE}
## Samples
a = ggplot(sample_WGBS_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_WGBS_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_WGBS_pca$eigenvalues[2],1),"%)",sep=""))+ scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=age_labels,guide=FALSE) + theme(aspect.ratio = 1)

b = ggplot(sample_WGBS_pca$eigenvectors,aes(x=PC1,y=PC3,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_WGBS_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC3 (",round(sample_WGBS_pca$eigenvalues[3],1),"%)",sep=""))+ scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=age_labels,guide=FALSE) + theme(aspect.ratio = 1)

c = ggplot(sample_DNase_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_DNase_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_DNase_pca$eigenvalues[2],1),"%)",sep=""))+ scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=age_labels,guide=FALSE) + theme(aspect.ratio = 1)

d = ggplot(sample_H3K27ac_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_H3K27ac_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_H3K27ac_pca$eigenvalues[2],1),"%)",sep=""))+ scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=age_labels,guide=FALSE) + theme(aspect.ratio = 1)

## chromHMM other groupings
e = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Anatomy)) + geom_point() + scale_color_manual(values=anatomy_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

f = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Type)) + geom_point() + scale_color_manual(values=type_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

g = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Germline)) + geom_point() + scale_color_manual(values=germline_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

h = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Cancer)) + geom_point() + scale_color_manual(values=cancer_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

## Subfamilies
i = ggplot(subfamily_WGBS_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE)+ labs(x=paste("PC1 (",round(subfamily_WGBS_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(subfamily_WGBS_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

j = ggplot(subfamily_DNase_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE)+ labs(x=paste("PC1 (",round(subfamily_DNase_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(subfamily_DNase_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

k = ggplot(subfamily_H3K27ac_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE)+ labs(x=paste("PC1 (",round(subfamily_H3K27ac_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(subfamily_H3K27ac_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

# Legends
legend_class_dot = get_legend(ggplot(subfamily_H3K27ac_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,name="Class") + scale_shape_manual(values=family_shapes,guide=FALSE) + theme(aspect.ratio = 1,legend.position="bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')))

legend_shape_age = get_legend(ggplot(sample_H3K27ac_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=age_labels) + theme(aspect.ratio = 1,legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')))

legend_family = get_legend(ggplot(subfamily_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family_shape)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,name="Family") + theme(aspect.ratio = 1,legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')))

grid.arrange(a,b,c,d,legend_group,legend_shape_age,e,f,g,h,legend_anatomy,legend_type,legend_germline,legend_cancer,i,j,k,legend_class_dot,legend_family,nrow=9,layout_matrix=rbind(c(1,2,3,4),c(5),c(6),c(7,8,9,10),c(11),c(12),c(13,14),c(15,16,17,NA),c(18,19)),heights=c(0.23,0.05,0.04,0.23,0.1,0.04,0.04,0.23,0.04))
```

### PCA

```{r subfam clustering}
# Matrix dimensions
dim(sample_chromHMM_3D)
dim(subfamily_chromHMM_3D)
dim(sample_WGBS_3D)
dim(subfamily_WGBS_3D)
dim(sample_DNase_enrichment)
dim(subfamily_DNase_enrichment)
dim(sample_H3K27ac_enrichment)
dim(subfamily_H3K27ac_enrichment)
```

### Enrichment thresholds

```{r excluded subfamilies}
# Number of subfamilies excluded by enrichment thresholds
# chromHMM, all chromosomes
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count <= THRESHOLD_IK_MEMBER),])[1]
# chromHMM, no Y
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count > THRESHOLD_IK_MEMBER & rmsk_TE_subfamily$Count_noY <= THRESHOLD_IK_MEMBER),])[1]

# WGBS, all chromosomes (excluded only from WGBS)
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count_CpGs <= THRESHOLD_IK_MEMBER & rmsk_TE_subfamily$Count > THRESHOLD_IK_MEMBER),])[1]

rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count_noY <= THRESHOLD_IK_MEMBER | rmsk_TE_subfamily$Count_CpGs <= THRESHOLD_IK_MEMBER),c("subfamily","family","class_update","Count","Count_noY","Count_CpGs","Total_length","Total_length_noY","CpGs")]

# Subfamilies with no CpGs
rmsk_TE_subfamily[which(rmsk_TE_subfamily$CpGs == 0),c("subfamily","Count","Count_noY","Total_length","Total_length_noY","CpGs")]

# Number of enrichments excluded by enrichment thresholds
test = table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR & subfamily_state_sample_combined$Count > THRESHOLD_IK_MEMBER & subfamily_state_sample_combined$Members <= THRESHOLD_IJK_MEMBER),]$State)
test
sum(test)
```

### Average members in a state for subfamilies in a state

```{r bp members percent in state}
# All instances
ddply(subfamily_state_sample_combined,~State,summarize,Percent_median=median(na.omit(Percent)), Members_median=median(na.omit(Members)),Percent_min=min(na.omit(Percent)), Members_min=min(na.omit(Members)), Percent_max=max(na.omit(Percent)), Members_max=max(na.omit(Members)))   

# Enriched
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],~State,summarise,Percent_median=median(Percent), Members_median=median(Members),Percent_min=min(Percent), Members_min=min(Members), Percent_max=max(Percent), Members_max=max(Members))

# >1%
ddply(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],~State,summarise, Percent_median=median(Percent),Members_median=median(Members),Percent_min=min(Percent), Members_min=min(Members), Percent_max=max(Percent), Members_max=max(Members))

# Max by state (outliers)
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],.(State),function(x) x[which.max(x$Percent),])
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],.(State),function(x) tail(x[order(x$Percent),c("subfamily","family","class_update","State","Sample",sample_categories,"Members","Percent")],n=10))
```

### WGBS subfamily state stats

```{r WGBS subfamily stats}
TE_meth_subfamily_percent = subfamily_CpG_meth[,c("subfamily","family","class_update","State","Sample","Count","Percent")]

# Average, max, and range of proportion of subfamily in methylation state plus all members in a state in a sample
TE_meth_subfamily_stats = ddply(TE_meth_subfamily_percent,.(subfamily,State),summarise,Median=median(Percent),Range=max(Percent)-min(Percent),Max=max(Percent),All_members=sum(Percent == 1))

table(TE_meth_subfamily_stats[which(TE_meth_subfamily_stats$All_members > 0),]$State)
sort(table(droplevels(subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == "Hypermethylated" & subfamily_state_sample_combined$Percent == 1),]$subfamily)))
merge(TE_meth_subfamily_stats[which(TE_meth_subfamily_stats$All_members > 0),],rmsk_TE_subfamily[,c("subfamily","Count","Count_CpGs","Total_length","Total_length_noY","CpGs")],by="subfamily")

# Average proportion in state, with/without IMR90
ddply(TE_meth_subfamily_stats,.(State),summarise,Median=mean(Median),Range=mean(Range),Max=mean(Max))
```

### Supplementary Figure 14

GO BP results downloaded in tsv format from GREAT. 

```{r Figure S14, echo=FALSE}
source("R_scripts/TE_correlation_subfamily.R") # Requires subfamily_enrichment.R
source("R_scripts/subfamily_correlate.R") # Requires TE_correlation_subfamily.R

a = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% measure_metrics_subfam[c(1,3,5,8)] & correlate_subfamily$class_update == "LTR"),],aes(x=State,y=Metric,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.text=element_text(size=7),axis.title.x = element_text(margin = margin(r = -10)),axis.title.y=element_text(size=9)) + scale_x_discrete(limits=states[c(1:14,16:17,19:21)],labels=all_state_labels) + ylab("Subfamily\ncharacteristic") + xlab("Samples enriched in state") +  scale_y_discrete(limits=rev(measure_metrics_subfam[c(1,3,5,8)]),labels=measure_labels) 
  
b = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% c(features[1:3],"coding_exon",features[5:8]) & correlate_subfamily$class_update == "LTR"),],aes(x=State,y=Metric,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(limits=c(-1,1),high="darkred",low="darkblue",name="Spearman's\nrho") + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.text=element_text(size=7),axis.title.x = element_text(margin = margin(r = -10)),axis.title.y=element_text(size=9)) + scale_x_discrete(limits=states[c(1:14,16:17,19:21)],labels=all_state_labels) + ylab("% members\noverlapping feature") + xlab("Samples enriched in state") +  scale_y_discrete(limits=rev(c(features[1:3],"coding_exon",features[5:8])),labels=genic_labels) 

# GREAT GO BP
great_names = gsub(".txt","",list.files("enrichment/great/"))
great = lapply(great_names,function(x) read.table(paste("enrichment/great/",x,".txt",sep=""),sep='\t'))
names(great) = great_names
great = ldply(great)
colnames(great) = c("set","Term Name","Binom Rank","Binom Raw P-Value","Binom FDR Q-Val","Binom Fold Enrichment","Binom Observed Region Hits","Binom Region Set Coverage",
                    "Hyper Rank","Hyper FDR Q-Val","Hyper Fold Enrichment","Hyper Observed Gene Hits","Hyper Total Genes","Hyper Gene Set Coverage")

c = ggplot(great[which(great$set == "AmnSINE1_7_Enh"),],aes(x=reorder(`Term Name`,`Binom Raw P-Value`,median),y=log10(`Binom Raw P-Value`))) + geom_bar(stat="identity") + coord_flip() + theme(axis.title.y = element_blank(),plot.title = element_text(hjust = 1,size=7),axis.text=element_text(size=6),plot.margin = margin(0,0,0,0)) + ylab("-Log10(p-value)") + ggtitle("AmnSINE1, 7_Enh")

d = ggplot(great[which(great$set == "AmnSINE2_DNase"),],aes(x=reorder(`Term Name`,`Binom Raw P-Value`,median),y=log10(`Binom Raw P-Value`))) + geom_bar(stat="identity") + coord_flip() + theme(axis.title.y = element_blank(),plot.title = element_text(hjust = 1,size=7),axis.text=element_text(size=6),plot.margin = margin(0,0,0,0)) + ylab("-Log10(p-value)") + ggtitle("AmnSINE2, DHS")

e = ggplot(great[which(great$set == "MER121_7_Enh"),],aes(x=reorder(`Term Name`,`Binom Raw P-Value`,median),y=log10(`Binom Raw P-Value`))) + geom_bar(stat="identity") + coord_flip() + theme(axis.title.y = element_blank(),plot.title = element_text(hjust = 1,size=7),axis.text=element_text(size=6),plot.margin = margin(0,0,0,0)) + ylab("-Log10(p-value)") + ggtitle("MER121, 7_Enh")

f = ggplot(great[which(great$set == "MER131_DNase"),],aes(x=reorder(`Term Name`,`Binom Raw P-Value`,median),y=log10(`Binom Raw P-Value`))) + geom_bar(stat="identity") + coord_flip() + theme(axis.title.y = element_blank(),plot.title = element_text(hjust = 1,size=7),axis.text=element_text(size=6),plot.margin = margin(0,0,0,0)) + ylab("-Log10(p-value)") + ggtitle("MER131, DHS")

grid.arrange(a,b,c,d,e,f,nrow=5,layout_matrix=rbind(c(1,2),c(3),c(4),c(5),c(6)),heights=c(0.25,0.2,0.2,0.2,0.15),widths=c(0.45,0.55))
```

### Subfamily enrichment correlation

Not MHC corrected

```{r subfamily enrichment correlation}
# Metric vs. metric
## Metrics
correlate_subfam_feature[which(abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% measure_metrics_subfam & correlate_subfam_feature$Metric %in% measure_metrics_subfam & correlate_subfam_feature$class_update == "All" & correlate_subfam_feature$p.value < 0.05),]

## Features
correlate_subfam_feature[which(abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfam_feature$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfam_feature$class_update == "All" & correlate_subfam_feature$State != correlate_subfam_feature$Metric & correlate_subfam_feature$p.value < 0.05),]

## Chromosomes
correlate_subfam_feature[which(abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% standard_chromosomes & correlate_subfam_feature$Metric %in% standard_chromosomes & correlate_subfam_feature$class_update == "All" & correlate_subfam_feature$State != correlate_subfam_feature$Metric & correlate_subfam_feature$p.value < 0.05),]

# State vs. state
correlate_subfam_state[which(abs(correlate_subfam_state$estimate.rho) > 0.3 & correlate_subfam_state$State %in% states & correlate_subfam_state$State2 %in% states & correlate_subfam_state$class_update == "All" & correlate_subfam_state$State != correlate_subfam_state$State2 & correlate_subfam_state$p.value < 0.05),]

## Range of correlation for active regulatory states, subfamily enrichment
median(correlate_subfam_state[which(correlate_subfam_state$class_update == "All" & correlate_subfam_state$State %in% states[c(1:2,7,16:17,20:21)] & correlate_subfam_state$State2 %in% states[c(1:2,7,16:17,20:21)] & correlate_subfam_state$State != correlate_subfam_state$State2),]$estimate.rho)
median(correlate_subfam_state[which(correlate_subfam_state$class_update == "LTR" & correlate_subfam_state$State %in% states[c(1:2,7,16:17,20:21)] & correlate_subfam_state$State2 %in% states[c(1:2,7,16:17,20:21)] & correlate_subfam_state$State != correlate_subfam_state$State2),]$estimate.rho)

# State vs. metric
## Metrics
correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% measure_metrics_subfam & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.05),]

correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% measure_metrics_subfam & correlate_subfamily$class_update != "All"  & correlate_subfamily$p.value < 0.05),]

## Features
correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.05),]

correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfamily$class_update != "All" & correlate_subfamily$p.value < 0.05),]

## Chromosomes
correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% standard_chromosomes & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.05),]

correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% standard_chromosomes & correlate_subfamily$class_update != "All" & correlate_subfamily$p.value < 0.05),]
```

### Summary of subfamily metrics

```{r subfamily metrics}
# Subfamily metrics, overall and by class
apply(rmsk_TE_subfamily[,measure_metrics_subfam],2,median)
aggregate(data=rmsk_TE_subfamily[,c("class_update",measure_metrics_subfam)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)

# Subfamily feature overlap by class
apply(rmsk_TE_subfamily[,cohorts],2,median)
aggregate(data=rmsk_TE_subfamily[,c("class_update",cohorts)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)

# Subfamily chromosome location by class
apply(rmsk_TE_subfamily[,standard_chromosomes[1:24]],2,median)
aggregate(data=rmsk_TE_subfamily[,c("class_update",standard_chromosomes[1:24])],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)

# Age vs. number of members
correlate_subfam_feature[which(correlate_subfam_feature$Metric == "Count" & correlate_subfam_feature$State == "Age"),]

# Subfamilies with very low mappability
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),])[1]
table(droplevels(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),]$family))
rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),c("subfamily","family","class_update","Count","Mappability")]
```

### SINE subfamilies

```{r SINE subfamilies}
# SINE subfamily active regulatory state enrichments
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)] & subfamily_state_sample_counts$class_update == "SINE"),]
test
ddply(test,.(subfamily),summarise,sum(V1))

test = rmsk_TE_subfamily[which(rmsk_TE_subfamily$class_update == "SINE"),]

# Age of SINE subfamilies
test[order(test$Age),c("subfamily","family","class_update","Count","Age")]

# CpG density of SINE subfamiles
test[order(test$CpGs_per_kbp),c("subfamily","family","class_update","Count","Age","CpGs_per_kbp")]
median(test[which(test$family == "Alu"),]$CpGs_per_kbp)

# Overlap with exons
test[order(test$exons),c("subfamily","family","class_update","Count","Age","exons","exons_pc","exons_nc")]
test[order(test$exons_pc),c("subfamily","family","class_update","Count","Age","exons","exons_pc","exons_nc")]
test[order(test$exons_nc),c("subfamily","family","class_update","Count","Age","exons","exons_pc","exons_nc")]

# Overlap with VISTA enhancers
test[order(test$Vista_enhancers),c("subfamily","family","class_update","Count","Age","Vista_enhancers")]

# Chromosomes
test[order(test$chr2),c("subfamily","family","class_update","Count","Age",standard_chromosomes[1:24])]
```

### VISTA enhancers

```{r vista enhancers analysis}
dim(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),])
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),c(TE_coordinates,"Vista_enhancers")]$subfamily)))
sort(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),c(TE_coordinates,"Vista_enhancers")]$subfamily)))

correlate_subfamily[which(correlate_subfamily$Metric == "Vista_enhancers" & correlate_subfamily$State %in% states & abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$p.value < 0.05),] # Not MHC corrected
```

### LTR subfamilies

```{r LTR subfamilies}
# LTR subfamily active regulatory state enrichments
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)] & subfamily_state_sample_counts$class_update == "LTR"),]
test[order(test$V1),]
ddply(test,.(subfamily),summarise,sum(V1))
```

### LINE subfamilies

```{r LINE subfamilies}
test = ddply(subfamily_state_sample_counts[which(subfamily_state_sample_counts$class_update == "LINE" & subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)]),],.(subfamily),summarise,Count=sum(V1))
test
sum(test$Count)

# Age of LINE subfamilies
test = rmsk_TE_subfamily[which(rmsk_TE_subfamily$class_update == "LINE"),]
test[order(test$Age),c("subfamily","family","class_update","Count","Age","CpGs_per_kbp")]
test[order(test$CpGs_per_kbp),c("subfamily","family","class_update","Count","Age","CpGs_per_kbp")]
```

### Genic subfamilies

```{r genic subfamilies}
# 4_Tx
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State == "4_Tx"),]
test[order(test$V1),]

# 6_EnhG
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State == "6_EnhG"),]
test[order(test$V1),]

# Intron overlap
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$State %in% c("3_TxFlnk","4_Tx","5_TxWk","6_EnhG") & subfamily_state_sample_counts$V1 > 0),]
test = merge(test,rmsk_TE_subfamily[,c("subfamily","Count","introns","introns_pc","introns_nc")],by="subfamily")
head(test[order(test$introns),],n=40)
```

### MER57E3

```{r mer57e3}
subfamily_state_sample_counts[which(subfamily_state_sample_counts$subfamily == "MER57E3"),]
MER57E3 = rmsk_TE[which(rmsk_TE$subfamily == "MER57E3"),]
dim(MER57E3)[1]
table(MER57E3$chromosome)
apply(MER57E3[,cohorts],2,function(x) dim(MER57E3[which(!is.na(x)),])[1])
```

### Young subfamilies

```{r young subfamilies}
quantile(rmsk_TE_subfamily$Age,0.05)
young_subfamilies = as.vector(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Age < quantile(rmsk_TE_subfamily$Age,0.05)),]$subfamily)
rmsk_TE_subfamily[which(rmsk_TE_subfamily$subfamily %in% young_subfamilies),c("subfamily","family","class_update","Age")]

young_subfamilies_enrichment = dcast(subfamily_state_sample_counts[which(subfamily_state_sample_counts$subfamily %in% young_subfamilies & subfamily_state_sample_counts$Sample.Proportion > 0),],subfamily~State,value.var="Sample.Proportion")
rownames(young_subfamilies_enrichment) = young_subfamilies_enrichment$subfamily
young_subfamilies_enrichment = young_subfamilies_enrichment[,2:16]

sort(apply(young_subfamilies_enrichment,2,function(x) sum(!is.na(x))))
sort(apply(young_subfamilies_enrichment,1,function(x) sum(!is.na(x))))

young_subfamilies_enrichment
```

### Subfamily factors

```{r subfamily correlation}
subfamily_measure_long = melt(rmsk_TE_subfamily_measure[,c("class_update","family","subfamily",cohorts,standard_chromosomes[1:24],measure_metrics_subfam,states[1:21])],id.vars=c("class_update","family","subfamily",cohorts,standard_chromosomes[1:24],measure_metrics_subfam))
subfamily_measure_long = melt(subfamily_measure_long,id.vars=c("class_update","family","subfamily","variable","value"))
colnames(subfamily_measure_long)[4:7] = c("State","Samples","Metric","Value")
```

### Subfamily heatmaps

```{r heatmaps, echo=FALSE, eval=FALSE}
a = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","1_TssA",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
b = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","2_TssAFlnk",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
c = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","3_TxFlnk",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
d = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","6_EnhG",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
e = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","7_Enh",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
f = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","4_Tx",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
g = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","5_TxWk",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
h = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Hypomethylated",1,sample_counts["All","WGBS"],THRESHOLD_LOR)
i = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Intermediate",1,sample_counts["All","WGBS"],THRESHOLD_LOR)
j = plot_binary_heatmap(subfamily_state_sample_filter,"DNase","DNase",1,sample_counts["All","DNase"],THRESHOLD_LOR)
k = plot_binary_heatmap(subfamily_state_sample_filter,"H3K27ac","H3K27ac",1,sample_counts["All","H3K27ac"],THRESHOLD_LOR)
```

### Subfamily permutation for candidates
### RERUN when updated

Looks for subfamily x state x sample group combinations enriched in more samples than expectation. Filtered on enrichment threshold (LOR>1.5) and 10 members in the state; excludes subfamilies with <30 members (accounts for state/chrY). Runs all metadata categories at once.

MHC on only subfamilies x states x categories/groupings ever enriched in state, active states only, categories with 1+ enrichments. 

```{r subfam permutation, cache=TRUE, cache.lazy=FALSE, eval=FALSE}
# Subfamilies ever enriched in a state
subfamily_state_ever = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)]),c("subfamily","State")]
permutation_input = merge(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Count > THRESHOLD_IK_MEMBER),],subfamily_state_ever,by=c("subfamily","State"))

# Interesting candidates
permute_subfam = ddply(permutation_input[,c("subfamily","State","Sample","Enrichment","Members")],.(subfamily,State),function(x) permute_by_sample(matrix=x,metric="Enrichment",direction="+",threshold=THRESHOLD_LOR,filtering="Members",threshold2=THRESHOLD_IJK_MEMBER))
permute_subfam = permute_subfam[which(permute_subfam$Samples.x > 0),]
permute_subfam$Padjust = p.adjust(permute_subfam$Pvalue,method="fdr")
dim(permute_subfam[which(permute_subfam$Padjust < 0.05),])

permute_subfam_wide = dcast(aggregate(data = permute_subfam[which(permute_subfam$Padjust < 0.05),],Grouping~State+subfamily+Category, paste, collapse = ','),State+subfamily~Category,value.var="Grouping")
permute_subfam_wide = merge(permute_subfam_wide,subfamily_state_sample_counts,by=c("subfamily","State"))[,c(1,10,9,2:8,11)]
write.table(permute_subfam_wide,file="enrichment/state_subfamily_clusters_enriched.txt",sep='\t',row.names = FALSE,quote = FALSE)
```

```{r subfam permutation pc, cache=TRUE, cache.lazy=FALSE, eval=FALSE}
# Subfamilies ever enriched in a state
subfamily_state_ever_pc = subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0 & subfamily_state_sample_counts_pc$State %in% states[c(1:7,16:17,20:21)]),c("subfamily","State")]
permutation_input_pc = merge(subfamily_state_sample_combined,subfamily_state_ever_pc,by=c("subfamily","State"))

# Interesting candidates
permute_subfam_pc = ddply(permutation_input_pc[,c("subfamily","State","Sample","Length_percent_jk","Members")],.(subfamily,State),function(x) permute_by_sample(x,"Length_percent_jk","+",THRESHOLD_PC,"Members",0))
permute_subfam_pc = permute_subfam_pc[which(permute_subfam_pc$Samples.x > 0),]
permute_subfam_pc$Padjust = p.adjust(permute_subfam_pc$Pvalue,method="fdr")
dim(permute_subfam_pc[which(permute_subfam_pc$Padjust < 0.05),])

permute_subfam_pc_wide = dcast(aggregate(data = permute_subfam_pc[which(permute_subfam_pc$Padjust < 0.05),],Grouping~State+subfamily+Category, paste, collapse = ','),State+subfamily~Category,value.var="Grouping")
permute_subfam_pc_wide = merge(permute_subfam_pc_wide,subfamily_state_sample_counts_pc,by=c("subfamily","State"))[,c(1,10,9,2:8,11)]
write.table(permute_subfam_pc_wide,file="enrichment/state_subfamily_clusters_pc.txt",sep='\t',row.names = FALSE,quote = FALSE)
```

### Print bedfiles for candidates (enriched and never)

''enrichment_candidates.txt'' is a list of subfamilies of interst, generated from above analysis.

```{r print bedfiles, eval=FALSE}
source("R_scripts/TE_correlation.R")

candidates = read.table(file="enrichment/enrichment_candidates.txt",sep='\t') #Results of candidate Excel analysis
colnames(candidates) = c("State","subfamily")

# Subfamilies enriched in many samples
candidates_many = subfamily_state_sample_counts[which(subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)] & subfamily_state_sample_counts$Sample.Proportion > 0.25),]
candidates_many[order(candidates_many$State,candidates_many$Sample.Proportion),]

# Combine enriched and many lists
candidates_all = unique(rbind(candidates,candidates_many[,c("State","subfamily")]))

# Enriched bedfiles
candidates_individual = apply(candidates_all,1,function(x) get_subfamily_enriched(x[2],x[1]))
names(candidates_individual) = candidates_all$State
candidates_individual = ldply(candidates_individual)
colnames(candidates_individual)[1] = "State"

# Never in state bedfiles
apply(candidates_all,1,function(x) write.table(rmsk_TE_measure[which(rmsk_TE_measure$subfamily == x[2] & rmsk_TE_measure[[x[1]]] == 0),c(colnames(rmsk_TE_measure)[1:4],x[1],"strand")],sep='\t',row.names=FALSE,col.names=FALSE,quote=FALSE,file=paste("enrichment/bedfiles/",x[2],"_never_",x[1],".bed",sep="")))

write.table(candidates_individual,file="enrichment/enrichment_candidates_indv.txt",sep='\t',row.names = FALSE,quote = FALSE)
```

### Explore candidates

```{r explore candidates, eval=FALSE}
# Overlap of enriched TEs with genic features
candidates_features = merge(candidates_individual,rmsk_TE[,c("chromosome","start","stop","subfamily","strand","family","class_update",features[c(1:3,5:8)],"coding_exon","Vista_enhancers")],all.x=TRUE,by=c("chromosome","start","stop","subfamily","strand"))

candidates_features_summary = ddply(candidates_features,.(subfamily,class_update,State),function(x) apply(x[,c(features[c(1:3,5:8)],"coding_exon","Vista_enhancers")],2,function(x) sum(!is.na(x))/length(x)))

# Percent overlap of each candidate subfamily with RefSeq features (only members in state when enriched)
write.table(candidates_features_summary,file="enrichment/candidates_features_summary.txt",sep='\t',row.names=FALSE,quote=FALSE)
```

### Supplementary Figure 15

Samples are ordered by group. Red lines split ESC/iPSC/ES-derived; Blood & T-cell/HSC & B-cell/Thymus; Brain/Neurosph; Digestive; Epithelial; Adipose/Mesench/Myosat/Muscle/Sm. Muscle/Heart/Other; IMR90/ENCODE2012

```{r Figure S15, echo=FALSE}
# Subfamilies with at least one instance of >1% of state in an active state
pc_subfamilies = as.vector(unique(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_combined$State %in% c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")),]$subfamily))

# Plot
a = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$subfamily %in% pc_subfamilies & subfamily_state_sample_combined$State %in% chromHMM_states[1:7]),],aes(x=reorder(Sample,match(Group,group_order)),y=reorder(subfamily,rmsk_TE_subfamily[match(subfamily,rmsk_TE_subfamily$subfamily),]$Total_length),fill=Length_percent_jk)) + geom_tile(linetype="blank") + theme(axis.text.x = element_blank(),axis.ticks.x=element_blank(),axis.title.y = element_text(margin = margin(r = -10))) + scale_fill_gradient(low="white",high="darkblue",limits=c(0,0.12),guide=FALSE) +  facet_wrap(~State,nrow=2) + xlab("Sample") + ylab("Subfamily") + geom_vline(xintercept=22.5,color="black") + geom_vline(xintercept=47.5,color="black") + geom_vline(xintercept=59.5,color="black") + geom_vline(xintercept=71.5,color="black") + geom_vline(xintercept=79.5,color="black") +  geom_vline(xintercept=111.5,color="black")

b = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$subfamily %in% pc_subfamilies & subfamily_state_sample_combined$State == "DNase"),],aes(x=reorder(Sample,match(Group,group_order)),y=reorder(subfamily,rmsk_TE_subfamily[match(subfamily,rmsk_TE_subfamily$subfamily),]$Total_length),fill=Length_percent_jk)) + geom_tile(linetype="blank") + theme(axis.text.x = element_blank(),axis.ticks.x=element_blank(),axis.title.y = element_text(margin = margin(r = -10))) + scale_fill_gradient(low="white",high="darkblue",limits=c(0,0.12),guide=FALSE) +  facet_wrap(~State,ncol=3,labeller=labeller(State=all_state_labels)) + xlab("Sample") + ylab("Subfamily") + geom_vline(xintercept=8.5,color="black") + geom_vline(xintercept=16.5,color="black") + geom_vline(xintercept=18.5,color="black") + geom_vline(xintercept=23.5,color="black") + geom_vline(xintercept=28.5,color="black") +  geom_vline(xintercept=38.5,color="black")

c = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$subfamily %in% pc_subfamilies & subfamily_state_sample_combined$State == "H3K27ac"),],aes(x=reorder(Sample,match(Group,group_order)),y=reorder(subfamily,rmsk_TE_subfamily[match(subfamily,rmsk_TE_subfamily$subfamily),]$Total_length),fill=Length_percent_jk)) + geom_tile(linetype="blank") + theme(axis.text.x = element_blank(),axis.ticks.x=element_blank(),axis.text.y = element_blank(),axis.ticks.y=element_blank(),axis.title.y=element_blank()) + scale_fill_gradient(low="white",high="darkblue",limits=c(0,0.12),guide=FALSE) +  facet_wrap(~State,ncol=3) + xlab("Sample") + ylab("Subfamily") + geom_vline(xintercept=16.5,color="black") + geom_vline(xintercept=35.5,color="black") + geom_vline(xintercept=42.5,color="black") + geom_vline(xintercept=52.5,color="black") + geom_vline(xintercept=57.5,color="black") +  geom_vline(xintercept=81.5,color="black")

d = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$subfamily %in% pc_subfamilies & subfamily_state_sample_combined$State %in% meth_states[1:2]),],aes(x=reorder(Sample,match(Group,group_order)),y=reorder(subfamily,rmsk_TE_subfamily[match(subfamily,rmsk_TE_subfamily$subfamily),]$Total_length),fill=Length_percent_jk)) + geom_tile(linetype="blank") + theme(axis.text.x = element_blank(),axis.ticks.x=element_blank(),axis.text.y = element_blank(),axis.ticks.y=element_blank(),axis.title.y=element_blank()) + scale_fill_gradient(low="white",high="darkblue",limits=c(0,0.12),guide=FALSE) +  facet_wrap(~State,ncol=3) + xlab("Sample") + ylab("Subfamily") + geom_vline(xintercept=13.5,color="black") + geom_vline(xintercept=15.5,color="black") + geom_vline(xintercept=19.5,color="black") + geom_vline(xintercept=25.5,color="black") + geom_vline(xintercept=35.5,color="black") + geom_vline(xintercept=36.5,color="black")

scale_length_jk = get_legend(ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$subfamily %in% pc_subfamilies & subfamily_state_sample_combined$State %in% chromHMM_states[1:7]),],aes(x=reorder(Sample,match(Group,group_order)),y=reorder(subfamily,rmsk_TE_subfamily[match(subfamily,rmsk_TE_subfamily$subfamily),]$Total_length),fill=Length_percent_jk)) + geom_tile(linetype="blank") + scale_fill_gradient(name="Proportion of state\nin subfamily",low="white",high="darkblue",limits=c(0,0.12)) + theme(legend.position = "bottom"))

grid.arrange(a,b,c,d,scale_length_jk,layout_matrix=rbind(c(1,1,1),c(2,3,4),c(5)),heights=c(0.6,0.35,0.05),widths=c(0.33,0.37,0.3))
```

### Subfamily stats, >1%

```{r pc table}
# Max % in subfamily by state
ddply(subfamily_state_sample_combined,~State,function(x) x[which.max(x$Length_percent_jk),])

# Max % of state in subfamily without enrichment
ddply(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment <= THRESHOLD_LOR),],~State,function(x) x[which.max(x$Length_percent_jk),])

# Number of instances by state and number of >1% subfamilies by state, overall and by class
pc_table = merge(as.data.frame(table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),]$State)),subfamily_state_sample_counts_pc_combine,by.x="Var1",by.y="State")
colnames(pc_table)[1:3] = c("State","Instances","Subfamilies")
sum(pc_table$Instances)

# Instances >1% bp of state, also enriched
pc_table = merge(pc_table,as.data.frame(table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR),]$State)),by.x="State",by.y="Var1",all=TRUE)
colnames(pc_table)[10] = "Enriched_Instances"
pc_table$Percent = pc_table$Enriched_Instances/pc_table$Instances
pc_table

# Number of subfamilies >1% across all states
length(unique(subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0),]$subfamily))
length(unique(subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0 & subfamily_state_sample_counts_pc$State %in% states[c(1:7,16:17,20:21)]),]$subfamily))

# Subfamilies with >1% of bases in genome
GENOME_WIDTH*THRESHOLD_PC
rmsk_TE_subfamily[which(rmsk_TE_subfamily$Total_length > GENOME_WIDTH*THRESHOLD_PC),c("subfamily","Count","Total_length","Total_length_noY","CpGs")]

# Subfamilies with >1% of CpGs in genome
ALL_CPGS*THRESHOLD_PC
rmsk_TE_subfamily[which(rmsk_TE_subfamily$CpGs > ALL_CPGS*THRESHOLD_PC),c("subfamily","Count","Total_length","Total_length_noY","CpGs")]

# Largest subfamilies vs. >1% of state
test = merge(dcast(subfamily_state_sample_counts_pc,subfamily+family+class_update~State,value.var = "V1"),rmsk_TE_subfamily[,c("subfamily","Count","Total_length","CpGs","Count_CpGs")],by="subfamily",all=TRUE)

test = test[order(test$Total_length,decreasing=TRUE),]
rownames(test) <- NULL
test$Order = rownames(test)
lapply(c(chromHMM_states,"DNase","H3K27ac"),function(x) if(pc_table[which(pc_table$State == x),]$Enriched_Instances > 0) {join(test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs","Count_CpGs",x)],aggregate(data=subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x & subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],Enrichment~subfamily,function(y) sum(y > THRESHOLD_LOR)),by="subfamily",type="left")} else{test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)]})

test = test[order(test$CpGs,decreasing=TRUE),]
rownames(test) <- NULL
test$Order = rownames(test)
lapply(meth_states,function(x) if(pc_table[which(pc_table$State == x),]$Enriched_Instances > 0) {join(test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs","Count_CpGs",x)],aggregate(data=subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x & subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],Enrichment~subfamily,function(y) sum(y > THRESHOLD_LOR)),by="subfamily",type="left")} else{test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)]})
```

## Factors influencing potential

### Figure 6

```{r Figure 6 scripts, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 2 scripts","Figure 5 scripts")}
source("R_scripts/TE_correlation.R")
source("R_scripts/state_metric_correlation.R")
```

```{r Figure 6, echo=FALSE,fig.height=6}
# SINE/Alu age vs. TE metrics and number of samples in methylation states
rmsk_TE_measure$cpgIsland_binary = as.numeric(ifelse(rmsk_TE_measure$cpgIsland == 0,0,1))

rmsk_TE_measure_SINE = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$class_update == "SINE"),c(TE_coordinates,"class_update",measure_metrics,meth_states,"cpgIsland_binary")])
rmsk_TE_measure_SINE$Length = rmsk_TE_measure_SINE$Length/1000

a = ggplot(rmsk_TE_measure_SINE,aes(x=JC_distance,y=..scaled..)) + geom_density(fill=class_colors["SINE"]) + xlab("Jukes-Cantor Evolutionary Distance") + ylab("Density")

b = ggplot(melt(rmsk_TE_measure_SINE,id.vars=c(TE_coordinates,"class_update","Length","mappability","JC_distance","CpGs","CpGs_per_length","cpgIsland_binary")),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=meth_colors, guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + ylab("Proportion of samples in state") + ylim(-0.05,1)

c = ggplot(melt(rmsk_TE_measure_SINE,id.vars=c(TE_coordinates,"class_update","JC_distance","CpGs",meth_states,"cpgIsland_binary"))) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + ylab("TE characteristic") + geom_smooth(data=rmsk_TE_measure_SINE,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + scale_y_continuous(breaks=seq(0,1.25,0.25))

rmsk_TE_measure_Alu = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$family == "Alu"),c(TE_coordinates,"class_update",measure_metrics,meth_states,"cpgIsland_binary")])
rmsk_TE_measure_Alu$Length = rmsk_TE_measure_Alu$Length/1000

d = ggplot(rmsk_TE_measure_Alu,aes(x=JC_distance,y=..scaled..)) + geom_density(fill=class_colors["SINE"]) + xlab("Jukes-Cantor Evolutionary Distance") + ylab("Density")

e = ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","Length","mappability","JC_distance","CpGs","CpGs_per_length","cpgIsland_binary")),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=meth_colors, guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + ylab("Proportion of samples in state") + ylim(-0.05,1)

f = ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","JC_distance","CpGs",meth_states,"cpgIsland_binary"))) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + ylab("TE characteristic") + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + scale_y_continuous(breaks=seq(0,1.25,0.25),limits=c(0,1.25))

legend1 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","Length","mappability","JC_distance","CpGs","CpGs_per_length","cpgIsland_binary")),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=meth_colors,name="State",labels=all_state_labels) + theme(legend.direction = "horizontal",legend.key.size = unit(3,'mm')))
  
legend2 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","JC_distance","CpGs",meth_states,"cpgIsland_binary"))) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),labels=c("Length (kbp)","Mappability","CpGs/bp"),name="Characteristic") + ylab("Property") + theme(legend.direction = "horizontal",legend.key.size = unit(3,'mm')) + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")))

grid.arrange(a,b,c,d,e,f,legend1,legend2,nrow=5,layout_matrix=rbind(c(1,4),c(2,5),c(7,7),c(3,6),c(8,8)),heights=c(0.2,0.35,0.05,0.35,0.05))
```

### Figure 6 analysis

```{r analysis Fig 6}
# Significant correlations between number of samples in state and metric (not MHC corrected)
correlate_TE_state[which(abs(correlate_TE_state$estimate.rho) > 0.3 & correlate_TE_state$State %in% states),]

# % of SINE that is Alu
dim(rmsk_TE[which(rmsk_TE$family == "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SINE"),])[1]

# Average age for older SINE, Alu
median(rmsk_TE[which(rmsk_TE$class == "SINE"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$class == "SINE"),]$JC_distance)
median(rmsk_TE[which(rmsk_TE$class == "SINE" & rmsk_TE$family != "Alu"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$class == "SINE" & rmsk_TE$family != "Alu"),]$JC_distance)
median(rmsk_TE[which(rmsk_TE$family == "Alu"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$family == "Alu"),]$JC_distance)

# Proportion of older SINE, Alu overlapping CpG islands
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SINE" & rmsk_TE$family != "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SINE" & rmsk_TE$family != "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SINE" & rmsk_TE$family != "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$family == "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$family == "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$family == "Alu"),])[1]

# Other models
test = lapply(c(meth_states,"Length","mappability","CpGs_per_length"),function(x) gam(rmsk_TE_measure_SINE[[x]]~rmsk_TE_measure_SINE$JC_distance))
lapply(test,function(x) summary(x))
test = lapply(c(meth_states,"Length","mappability","CpGs_per_length"),function(x) gam(rmsk_TE_measure_Alu[[x]]~rmsk_TE_measure_Alu$JC_distance))
lapply(test,function(x) summary(x))

# Overlap with CpG islands by age (logistic regression) (all, SINE, Alu)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure,family="binomial")
summary(test)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure_SINE,family="binomial")
summary(test)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure_Alu,family="binomial")
summary(test)

# Age vs. overlap with CpG islands
ddply(rmsk_TE_measure,.(cpgIsland_binary),summarise,Age=median(JC_distance))
ddply(rmsk_TE_measure[which(rmsk_TE_measure$class_update == "SINE"),],.(cpgIsland_binary),summarise,Age=median(JC_distance))
ddply(rmsk_TE_measure[which(rmsk_TE_measure$family == "Alu"),],.(cpgIsland_binary),summarise,Age=median(JC_distance))
```

### Supplementary Figure 16

```{bash chromosome epigenetics, eval=FALSE}
# By chromosome
while read line; do awk -v OFS='\t' -v sample=$line '{a[$1,$4]+=$3-$2}END{for(i in a){split(i,sep,SUBSEP); print sample, sep[1], sep[2], a[i];}}' raw_data/chromHMM/$line\_15_coreMarks_mnemonics.bed; done < sample_lists/mnemonics.txt > chromHMM/chromosome_states.txt

## By chromosome
awk '{print $0 > $1}' ~/TE_landscape/WGBS/all_CpG_Meth.bed
for file in chr*; do awk -v chr=$file -v OFS='\t' '{for (i=4;i<=NF;i++){if($i == -1) miss[i]+=1; else if ($i < 0.3) hypo[i]+=1; else if ($i > 0.7) hyper[i]+=1; else if (($i <= 0.7) && ($i >= 0.3)) inter[i]+=1;}}; END{for (i in hyper) print chr, i, hypo[i], inter[i], hyper[i], miss[i];}' $file >> chr_CpG_meth_states.txt; done &

# Whole genome, number of peaks by chromosome
while read line; do awk -v sample=$line '{a[$1]+=1}END{for(i in a){print sample, i, a[i]}}' raw_data/DNase/DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak >> DNase/chr_DNase.txt ; done < sample_lists/DNase_samples.txt

# Whole genome, number of peaks by chromosome
while read line; do awk -v sample=$line '{a[$1]+=1}END{for(i in a){print sample, i, a[i]}}' raw_data/H3K27ac/H3K27ac_narrow_peaks/$line\-H3K27ac.narrowPeak >> H3K27ac/chr_H3K27ac.txt ; done < sample_lists/H3K27ac_samples.txt
```

```{r Figure S16, echo=FALSE}
source("R_scripts/chromosome_potential.R")

# Epigenetic state by chromosome
a = ggplot(chrom_state_total,aes(x=chromosome,y=Bases,fill=State)) + geom_bar(stat="identity",position="fill") + scale_fill_manual(values = chromHMM_colors) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.position="bottom",legend.box.margin = margin(t=-10),legend.margin=margin(0,0,0,0),legend.key.size = unit(3,'mm')) + ylab("Proportion in state") + xlab("Chromosome") + guides(fill = guide_legend(nrow=4,title.position = "top"))

b = ggplot(chr_state_WGBS_total,aes(x=chromosome,y=CpGs,fill=State)) + geom_bar(position="fill",stat="identity") + scale_fill_manual(values=meth_colors,labels=meth_labels) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.position="bottom",legend.box.margin = margin(t=-10),legend.margin=margin(0,0,0,0),legend.key.size = unit(3,'mm')) + ylab("Proportion in state") + xlab("Chromosome") + guides(fill = guide_legend(nrow=4,title.position = "top"))

c = ggplot(chr_state_DNase_total,aes(x=size/1000000,y=Peaks/1000000,label=chromosome)) + geom_text(size=3) + geom_smooth(method='lm',se=FALSE,color="black",linetype="dashed") + xlab("Chromosome size (Mbp)") + ylab("Peaks (millions)")  + scale_y_continuous(limits=c(0,1.5))

d = ggplot(chr_state_H3K27ac_total,aes(x=size/1000000,y=Peaks/1000000,label=chromosome)) + geom_text(size=3) + geom_smooth(method='lm',se=FALSE,color="black",linetype="dashed") + xlab("Chromosome size (Mbp)") + ylab("Peaks (millions)") + scale_y_continuous(limits=c(0,1.5))

# Chromosome vs. states
e = ggplot(chromosome_potential,aes(x=State,y=Zscore,label=chromosome)) + geom_point() + geom_text(angle=45,hjust=0,vjust=0,size=3) + scale_x_discrete(limits=rev(states),labels=all_state_labels) + scale_y_continuous(limits=c(-5,5)) + theme(panel.grid.major.y=element_line(color="grey"),axis.title.y = element_text(margin = margin(r = -10))) + ylab("Z-score, % of samples in state") + geom_hline(yintercept = 1,color="grey") + geom_hline(yintercept = 2,color="grey") + geom_hline(yintercept = -1,color="grey") + geom_hline(yintercept = -2,color="grey") + geom_hline(yintercept = -3,color="grey") + geom_hline(yintercept = 3,color="grey") + geom_hline(yintercept = -4,color="grey") + geom_hline(yintercept = 4,color="grey") + coord_flip() + geom_hline(yintercept = -5,color="grey") + geom_hline(yintercept = 5,color="grey")

grid.arrange(a,b,c,d,e,layout_matrix=rbind(c(1,2),c(3,4),c(5)),heights=c(0.35,0.25,0.4))
```

### Supplementary Figure 17 (intergenic potential)

```{r Figure S17 scripts, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 6 scripts")}
source("R_scripts/feature_overlap_state.R") #Requires potential.R
source("R_scripts/intergenic_potential.R")
```

For each state and TE class, the proportion of TEs that are genic (including promoters) at each frequency of expression. The dashed lines give the proportion of all members of the class that are intergenic. I split the analysis between all genes and just protein-coding genes, because TEs have a huge contribution to lncRNA.  

For 4_Tx, the proportion of TEs that are intergenic is higher than baseline for those that are never in the state, but it drops off precipitously as the number of samples in the state increases, declining much more slowly thereafter. This is true for all classes, although LTR always has a higher proportion of intergenic TEs for this state. Conclusion: TEs are much more likely to be transcribed when they overlap a gene, but this is true across all classes, not just SINEs. 

Feature-state patterns roughly the same across class. Stronger for protein-coding than non-coding (not shown).

```{r Figure S17, echo=FALSE}
# Feature overlap vs. samples in state, by class
a = ggplot(na.omit(feature_state_mean_class[which(feature_state_mean_class$Coding == "All"),]),aes(x=State,y=forcats::fct_rev(Feature),fill=Enrichment*100)) + geom_tile() + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background = element_rect(fill="lightgrey"),axis.title.x = element_text(margin = margin(t = -10))) + scale_fill_gradient2(name="% increase\nin sample\nproportion",low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-100,5000),na.value="darkblue") + scale_y_discrete(labels=genic_labels) + scale_x_discrete(labels=all_state_labels) + facet_wrap(~Class) + ylab("Feature")

b = ggplot(state_genic[which(!(state_genic$class_update %in% c("SVA","Other")) & state_genic$State %in% states[c(1,4:5,7,20:21)]),],aes(x=Bin,y=Ratio,color=class_update)) + geom_line() + ylim(0,1) + scale_color_manual(values=class_colors,name="Class") + xlab("Proportion of samples in state") + ylab("Proportion of TEs overlapping genes") +  geom_hline(data=rmsk_TE_genic[which(!(rmsk_TE_genic$class_update %in% c("SVA","Other"))),],aes(yintercept=Proportion,color=class_update),linetype="dotted") + facet_grid(State~Genes,labeller = labeller(Genes=setNames(c("All genes","Protein-coding genes"),c("All","PC")),State=all_state_labels))

grid.arrange(a,b,nrow=2)
```

### CpG density vs class repression mechanism

Found the average number of samples a TE is in 9_Het or hypermethylated by class. For 9_Het, it is 0 for SINE and 0.79% for LTR (mean 1.8% and 5.3%). For hypermethylation, it is 97.3% for SINE and 91.9% for LTR (mean 89.5% and 80.2%).

To answer the question of whether CpG density or class has a stronger impact on repression, I created generalized linear models with a quasi-Poisson distribution for the response, which seems to match the distributions. They include only class or class and CpG density as variables. I used raw counts for samples in state and performed anova using the chi-squared test for significance. For a log link function, the base mean is multiplied by exp(coefficient). 

Adding CpG density significantly improves the 9_Het model, and the effect size for class remains about the same. The effect size for CpG density (controlled for class) is small, with each additional CpG per kbp decreasing the number of samples in the 9_Het state by 0.007%. Since most TEs fall within the 0-100 CpGs/kbp range, this results in a range of ~1% of samples. Adding an interaction term also improves the model, although to a smaller extent. This result suggests that increased CpG density is negatively correlated with the number of samples in the 9_Het state, but that the correlation is about half as strong for SINE as for LTR elements. (With GLM: For 9_Het, adding the two additional terms decreases the impact of LTR vs. SINE, but only by a very small amount (35% instead of 34%). In contrast, increasing the CpGs/kbp from 0 to 100 would decrease the number of samples to ~72% of the original number.)

With hypermethylation, including CpG density also significantly improves the model, and the SINE class term remains significant, though with a smaller effect size (6.8 vs. 9.3%). In this case, each additional CpG/kbp increases the number of samples in the state by 0.2%, such that the overall effect of CpG density can be larger than that of class. Adding the interaction term also improves the model, and the increase due to CpG density is higher in LTR than in SINE elements. (With GLM: For hypermethylated samples, the addition of the CpG density term also slightly decreases the SINE coefficient's impact (112% vs. 109%). However, increasing the CpGs/kbp from 0 to 100 increases the number of samples by ~27%, and this is barely impacted by the addition of the interaction term.)

```{r repression modelling}
source("R_scripts/CpG_density_class_represssion.R")

# CpG density between SINE and LTR elements
ddply(rmsk_TE,.(class_update),summarise,CpG_dens=median(CpGs_per_length)*1000,dens=IQR(CpGs_per_length)*1000)

# Average number of samples a TE is 9_Het or hypermethylated
ddply(rmsk_TE_measure,.(class_update),summarise,Het=median(`9_Het`),Hyper=median(na.omit(Hypermethylated)))
ddply(rmsk_TE_measure,.(class_update),summarise,Het=mean(`9_Het`),Hyper=mean(na.omit(Hypermethylated)))

anova(model_het1,model_het2,test="F")
anova(model_het2,model_het3,test="F")

anova(model_hyper1,model_hyper2,test="F")
anova(model_hyper2,model_hyper3,test="F")
```

### Supplementary Figure 18

```{r Figure S18, echo=FALSE,fig.height=6}
#source("R_scripts/state_metric_correlation.R")

a = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background=element_rect(fill="lightgrey"),aspect.ratio = 1) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs")) + xlab("TE characteristic") + ylab("TE characteristic")

b = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background=element_rect(fill="lightgrey"),aspect.ratio = 1) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs")) + xlab("TE characteristic") + ylab("TE characteristic") + facet_wrap(~class_update)

# Metric vs. state
c = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update == "All" & correlate_TE_state$State %in% meth_states),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background=element_rect(fill="lightgrey"),aspect.ratio = 1) + scale_y_discrete(limits=rev(meth_states)) + scale_x_discrete(limits=measure_metrics[c(1,4:5,2:3)],labels=c("Length","CpGs","CpGs/bp","Mappability","Age")) + xlab("TE characteristic") + ylab("Samples in state")

d = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update != "All" & correlate_TE_state$State %in% meth_states),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background=element_rect(fill="lightgrey"),aspect.ratio = 1) + scale_y_discrete(limits=rev(meth_states)) + scale_x_discrete(limits=measure_metrics[c(1,4:5,2:3)],labels=c("Length","CpGs","CpGs/bp","Mappability","Age")) + xlab("TE characteristic") + ylab("Samples in state") + facet_wrap(~class_update) 

scale_legend = get_legend(ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(limits=c(-1,1),high="darkred",low="darkblue",mid="white",name="Spearman's rho") + theme(panel.background=element_rect(fill="lightgrey"),aspect.ratio = 1,legend.position="bottom"))

grid.arrange(a,b,c,d,scale_legend,nrow=3,heights=c(0.45,0.45,0.1),widths=c(0.45,0.55))
```

### TE feature stats

```{r TE feature stats, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 6 scripts")}
# States
apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y > 0),states],na.rm=TRUE)) - apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y == 0),states],na.rm=TRUE))

dcast(feature_state_mean[which(feature_state_mean$Coding == "All"),],Feature~State,value.var = "Enrichment")
feature_state_mean_class[which(feature_state_mean_class$Coding == "All"),]

# Proportion of TEs ever in state overlapping each feature
apply(rmsk_TE_measure[,states],2,function(x) apply(rmsk_TE_measure[,cohorts],2,function(y) dim(rmsk_TE_measure[which(x > 0 & y > 0),])[1]/dim(rmsk_TE_measure[which(x > 0),])[1]))
```

### TE feature vs. metric

```{r feature v metric, eval=FALSE}
source("R_scripts/feature_overlap_metric.R") #Requires TE_correlation.R

feature_metric_mean[order(feature_metric_mean$Enrichment),]
feature_metric_mean_class[order(feature_metric_mean_class$Enrichment),]

apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y > 0),measure_metrics],na.rm=TRUE)) - apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y == 0),measure_metrics],na.rm=TRUE))
```

### Chromosome stats

```{r chromosome stats}
# Proportion of TEs relative to genome length
hg19_TEs[order(hg19_TEs$Prop_TEs/hg19_TEs$Prop),]

# Z-scores
chromosome_potential[order(chromosome_potential$Zscore),]

# Proportion of TEs in each state on each chromosome
apply(rmsk_TE_measure[,states],2,function(x) table(rmsk_TE_measure[which(x > 0),]$chromosome)/dim(rmsk_TE_measure[which(x > 0),])[1])[standard_chromosomes[1:24],]
```

### Potential without chrY test

```{r potential noY, eval=FALSE}
# Potential without chrY
test = chromHMM_TE_state[which(chromHMM_TE_state$chromosome != "chrY"),]
test.a = sample_distribution(test,c(8:22),sample_counts["All","chromHMM"])
test.b = cumulative_distribution(test,c(8:22),sample_counts["All","chromHMM"])
test.b$State = factor(rep(chromHMM_states,each=sample_counts["All","chromHMM"]),levels=chromHMM_states)
test.c = potential_stats(test.a,15,sample_counts["All","chromHMM"])
test.c$State = factor(chromHMM_states,levels=chromHMM_states)
chromHMM_TE_state_dist_stats-test.c

# Average samples in state for chrY TEs
sort(colMeans(chromHMM_TE_state[which(chromHMM_TE_state$chromosome == "chrY"),8:22]))
```

## Mouse

```{bash sample pairs, eval=FALSE}
# Twelve Roadmap samples with corresponding mouse WGBS samples (7) and mouse chromHMM samples (12)
human_mouse_samples.txt
```

```{bash mm10 TEs, eval=FALSE}
# RepeatMasker file for mm10	 
#TE_landscape/features/mouse/rmsk_mm10.txt	
cp /bar/genomes/mm10/rmsk/rmsk.txt.gz rmsk_mm10.txt.gz

# mm10 mouse TEs, chromosomes 1-19, X, Y, M only	 
#TE_landscape/features/mouse/TEs/mm10_rmsk_TE.txt	
awk -v OFS='\t' '{if(($12 == "LTR" || $12 == "DNA" || $12 == "SINE" || $12 == "LINE" || 12 == "Other" || $12 == "RC" || $12 == "Unknown" || $12 == "DNA?" || $12 == "LINE?" || $12 == "LTR?" || $12 == "SINE?" || $12 == "RC?") && ($6 !~ /_/)) print $6, $7, $8, $11, $12, $13, $10}' rmsk_mm10.txt > mm10_rmsk_TE.txt
```

```{bash human-mouse liftover, eval=FALSE}
#To load liftOver
ml kentUCSC

# All TE RepeatMasker bedfile with unique number for each	 
#TE_landscape/features/TEs/rmsk_TEother_numbered.txt	
cat rmsk_TE.txt rmsk_other.txt | awk -v OFS="\t" '{print $1,$2,$3,NR,".",$7}' - > rmsk_TEother_numbered.txt

# Liftover between human TEs and mouse mm10	 
#TE_landscape/Mouse/liftover/rmsk_TE_hg19tomm10.bed	
liftOver rmsk_TEother_numbered.txt ../../genomes/hg19/chainFiles/hg19ToMm10.over.chain.gz rmsk_TE_hg19tomm10.bed out_mm10.txt -minMatch=0.1

# Human TEs that do not lift over to mouse	
#TE_landscape/Mouse/liftover/out_mm10.txt	

# Numbers of human TEs that lifted over to mouse	 
#TE_landscape/Mouse/liftover/numbers.txt	
awk '{print $4}' rmsk_TE_hg19tomm10.bed > numbers.txt

# Pull out human TEs lifted over to mouse by number	 
#TE_landscape/Mouse/liftover/rmsk_TE_in_mm10.txt	
python ../bin/TE_landscape/subset_file.py rmsk_TEother.txt numbers.txt rmsk_TE_in_mm10.txt

# Combine human and lifted over mouse coordinates	 
#TE_landscape/Mouse/rmsk_TE_mm10.bed	
cut -f1-3,6 rmsk_TE_hg19tomm10.bed | paste - rmsk_TE_in_mm10.txt > rmsk_TE_mm10.bed

# Lifted-over human coordinates intersected with mouse TE coordinates	 
#TE_landscape/Mouse/liftover/hg19_mm10_TE_intersect.bed	
bedtools intersect -wo -a rmsk_TE_mm10.bed -b mm10_rmsk_TE.txt > hg19_mm10_TE_intersect.bed

# Human-mouse orthologs TEs with same subfamily name (hg19-mm10)	 
#TE_landscape/Mouse/liftover/hg19_mm10_TE_intersect_same.bed	
awk -v OFS="\t" '{if($8==$15)print $0}' hg19_mm10_TE_intersect.bed > hg19_mm10_TE_intersect_same.bed

# Test to confirm I am not missing orthologs by filtering repeats (1/19/18)
awk -v OFS='\t' '{print $6, $7, $8, $11, $12, $13, $10}' features/mouse/rmsk_mm10.txt | bedtools intersect -wo -a Mouse/rmsk_TE_mm10.bed -b - > Mouse/hg19_mm10_repeat_intersect.bed
awk -v OFS="\t" '{if($8==$15)print $0}' Mouse/hg19_mm10_repeat_intersect.bed > Mouse/hg19_mm10_repeat_intersect_same.bed
```

```{bash mm10 WGBS, eval=FALSE}
# WGBS (mm10)
#TE_landscape/Mouse/WGBS/intersect/TEs/mm10_rmsk_TE_ENCFF#.bed [9 files]
#TE_landscape/Mouse/WGBS/intersect/TEs/mm10_rmsk_TE_WGBS.bed
awk -v OFS='\t' '{print $1, $2, $3, $10, $11}' $file | bedtools intersect -wo -a mm10_rmsk_TE.txt -b - > mm10_rmsk_TE_$file;

# mm10 TE average methylation 
#TE_landscape/Mouse/WGBS/mm10_rmsk_TE_WGBS_avg.txt	
python ~/bin/TE_landscape/average_methylation_stranded.py mm10_rmsk_TE_WGBS.bed mm10_rmsk_TE.txt samples.txt  mm10_rmsk_TE_WGBS_avg.txt
```

```{bash mm10 CpG count, eval=FALSE}
# Mouse TEs (12/7/17)
awk -v OFS='\t' '{a[$1, $2, $3, $4, $5, $6, $7]+=1}END{for(i in a){split(i,sep,SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], a[i];}}' Mouse/WGBS/intersect/TEs/mm10_rmsk_TE_WGBS.bed > Mouse/WGBS/intersect/TEs/mm10_rmsk_TE_CpG_count.txt
```

```{bash hg19 ortholog chromHMM, eval=FALSE}
# chromHMM state for all orthologous hg19 TEs
# Uses the chromHMM state of TEs, split by sample, used for combine marks analysis
## Confirmed that it only pulled out TEs with matching subfamily and that the number of unique TEs is 269096 for all
cut -f5-11 Mouse/liftover/hg19_mm10_TE_intersect_same.bed | uniq | sort -k1,1 -k2,2n > Mouse/hg19_orthologs.txt

for file in E066 E071 E081 E083 E084 E086 E088 E092 E094 E096 E105 E106; do tail -n +2 /scratch/ecp/pandas/$file | sort -k1,1 -k2,2n - | bedtools intersect -sorted -wo -a Mouse/hg19_orthologs.txt -b - -f 1 -r > Mouse/hg19_ortholog_chromHMM_$file\.txt; done

for file in Mouse/chromHMM/hg19_ortholog_chromHMM_*.txt; do cut -f8-18 $file >> Mouse/chromHMM/hg19_chromHMM_TE.txt; done
```

```{bash mm10 chromHMM, eval=FALSE}
## Intersected with TEs with human orthologs
while read a b c; do gunzip raw_data/mouse/chromHMM/$c\.bed.gz; cut -f12-18 Mouse/liftover/hg19_mm10_TE_intersect_same.bed | bedtools intersect -wo -a - -b raw_data/mouse/chromHMM/$c\.bed > Mouse/chromHMM/$c\_TE.txt; gzip raw_data/mouse/chromHMM/$c\.bed; done < Mouse/human_mouse_samples.txt

## Any overlap for mm10 TEs with human ortholog
while read a b c; do uniq Mouse/chromHMM/$c\_TE.txt | awk -v OFS='\t' -v sample=$c '{a[$1, $2, $3, $4, $5, $6, $7, $11]+=$17}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sep[8], sample, a[i]}}' - | sort -k1,1V -k2,2V -k3,3n -k9,9 - >> Mouse/chromHMM/mm10_chromHMM_TE_sorted.txt; done < Mouse/human_mouse_samples.txt
```

```{bash mm10 subfamily, eval=FALSE}
## Intersected with all mm10 TEs
while read a b c; do gunzip raw_data/mouse/chromHMM/$c\.bed.gz; bedtools intersect -wo -a features/mouse/mm10_rmsk_TE.txt -b raw_data/mouse/chromHMM/$c\.bed > Mouse/chromHMM/$c\_TE_all.txt; gzip raw_data/mouse/chromHMM/$c\.bed; done < Mouse/human_mouse_samples.txt &

# chromHMM, any overlap, for all mm10 TEs
while read a b c; do awk -v OFS='\t' -v sample=$c '{a[$1, $2, $3, $4, $5, $6, $7, $11]+=$17}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], sep[4], sep[5], sep[6], sep[7], sep[8], sample, a[i]}}' Mouse/chromHMM/$c\_TE_all.txt | sort -k1,1V -k2,2V -k3,3n -k9,9 - >> Mouse/chromHMM/mm10_chromHMM_TE_sorted_all.txt; done < Mouse/human_mouse_samples.txt

## mm10 TEs in state x subfamily x sample
awk -v OFS='\t' '{a[$4,$8,$9]+=1}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], sep[3], a[i]}}' Mouse/chromHMM/mm10_chromHMM_TE_sorted_all.txt > Mouse/chromHMM/mm10_chromHMM_subfamily.txt
```

```{bash hg19 subfamily active, eval=FALSE}
# Proportion of subfamily in active regulatory state
# Uses chromHMM state of TEs, split by sample, used for combine marks analysis
for file in E066 E071 E081 E083 E084 E086 E088 E092 E094 E096 E105 E106; do awk -v OFS='\t' '{if(($10 == "1_TssA") || ($10 == "2_TssAFlnk") || ($10 == "3_TxFlnk") || ($10 == "6_EnhG") || ($10 == "7_Enh")) print $1, $2, $3, $4, $5, $6, $7, $8}' /scratch/ecp/pandas/$file | uniq | awk '{a[$4,$8]+=1}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], a[i]}}' - >> Mouse/chromHMM/hg19_chromHMM_subfamily_active.txt; done
```

```{bash mm10 subfamily active, eval=FALSE}
## Proportion in any active regulatory state
awk -v OFS='\t' '{if(($8 == "TssA") || ($8 == "TssAFlnk1") || ($8 == "TssAFlnk2") || ($8 == "Enh") || ($8 == "EnhLo1") || ($8 == "EnhLo2")) print $1, $2, $3, $4, $5, $6, $7, $9}' Mouse/chromHMM/mm10_chromHMM_TE_sorted_all.txt | uniq | awk -v OFS='\t' '{a[$4,$8]+=1}END{for(i in a) {split (i, sep, SUBSEP); print sep[1], sep[2], a[i]}}' - > Mouse/chromHMM/mm10_chromHMM_subfamily_active.txt
```

```{r mouse scripts, cache=TRUE, cache.lazy=FALSE}
#source("R_scripts/proportion.R")
#source("R_scripts/subfamily_enrichment.R")
#source("R_scripts/WGBS_subfamily_enrichment_TE.R")
source("R_scripts/human_mouse_orthologs.R")
source("R_scripts/human_mouse_ortholog_WGBS.R") 
source("R_scripts/human_mouse_ortholog_chromHMM.R")
source("R_scripts/human_mouse_subfamily.R")

# TE ortholog analysis
rmsk_TE_class$Mouse_orthologs = table(unique(human_mouse_orthologs_mm10[,c(hg19_coordinates,"class_update")])$class_update)[as.vector(rmsk_TE_class$class_update)]
rmsk_TE_class = merge(rmsk_TE_class,ddply(rmsk_TE_subfamily[which(rmsk_TE_subfamily$subfamily %in% mm10_rmsk_TE$subfamily),],~class_update,summarize,
                                          Mouse_ortholog_subfamily = length(subfamily)),by="class_update",all.x=TRUE)
rmsk_TE_class[which(is.na(rmsk_TE_class$Mouse_ortholog_subfamily)),]$Mouse_ortholog_subfamily = 0

# Table of tissue-specific TEs
hg19_orthologs_specific_active = as.data.frame(t(sapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) tissue_matrix(x,hg19_mm10_chromHMM_active))))
hg19_orthologs_specific_active$Tissue = rownames(hg19_orthologs_specific_active)
hg19_orthologs_specific_active$Other = hg19_orthologs_specific_active$All - hg19_orthologs_specific_active$Specific - hg19_orthologs_specific_active$On - hg19_orthologs_specific_active$Off
hg19_orthologs_specific_active = melt(hg19_orthologs_specific_active[,c("Specific","On","Off","Other","Tissue")],id.vars="Tissue")
colnames(hg19_orthologs_specific_active)[2:3] = c("Subset","Count")
hg19_orthologs_specific_active$State = rep("Active",dim(hg19_orthologs_specific_active)[1])

hg19_orthologs_specific_promoter = as.data.frame(t(sapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) tissue_matrix(x,hg19_mm10_chromHMM_promoter))))
hg19_orthologs_specific_promoter$Tissue = rownames(hg19_orthologs_specific_promoter)
hg19_orthologs_specific_promoter$Other = hg19_orthologs_specific_promoter$All - hg19_orthologs_specific_promoter$Specific - hg19_orthologs_specific_promoter$On - hg19_orthologs_specific_promoter$Off
hg19_orthologs_specific_promoter = melt(hg19_orthologs_specific_promoter[,c("Specific","On","Off","Other","Tissue")],id.vars="Tissue")
colnames(hg19_orthologs_specific_promoter)[2:3] = c("Subset","Count")
hg19_orthologs_specific_promoter$State = rep("Promoter",dim(hg19_orthologs_specific_promoter)[1])

hg19_orthologs_specific = rbind(hg19_orthologs_specific_active,hg19_orthologs_specific_promoter)
hg19_orthologs_specific$Subset = factor(hg19_orthologs_specific$Subset,levels=c("Off","Specific","Other","On"))
hg19_orthologs_specific = ddply(hg19_orthologs_specific,.(Tissue,State),transform,Proportion=Count/sum(Count),Total=sum(Count))
```

### Figure 7

A) Axes are individual TE methylation level - all 13 samples pairs. Excludes TEs missing methylation in either sample of a pair.
B) Y-axis is % of human TEs in each state in that state in mouse; colors as before - all 13 sample pairs
C) Axes are subfamily % hypomethylated (of those with CpGs)

```{r Figure 7, echo=FALSE, fig.height=5}
a = ggplot(human_mouse_orthologs_WGBS_paired,aes(x=Human_methylation,y=Mouse_methylation)) + stat_bin2d(bins=50) + scale_fill_gradient(low="white",high="darkblue",name="Number\nof TEs") + xlab("Methylation level (human)") + ylab("Methylation level\n(mouse)") + coord_equal() + theme(legend.position = "bottom",legend.direction="horizontal",legend.box.margin = margin(t=-10))

state_freq = ddply(human_mouse_orthologs_WGBS_paired,.(Human_state_WGBS),summarise,Entries=length(Human_state_WGBS)/dim(human_mouse_orthologs_WGBS_paired)[1])

b = ggplot(human_mouse_orthologs_WGBS_paired,aes(x=Human_state_WGBS,fill=Mouse_state_WGBS)) + geom_bar(position="fill") + scale_fill_manual(values=meth_colors,name="State (mouse)",labels=meth_labels) + xlab("State (human)") + ylab("Proportion in state\nin mouse") + scale_x_discrete(labels=as.vector(apply(state_freq,1,function(x) paste(x[1],"\n(",round(as.numeric(x[2])*100,0),"%)",sep="")))) + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),legend.direction="horizontal",legend.position="bottom",legend.box.margin = margin(t=-10),legend.key.size = unit(3,'mm')) + guides(fill=guide_legend(nrow=2,title.position="top"))

state_freq_chromHMM = ddply(human_mouse_orthologs_chromHMM_paired,.(Human_state_chromHMM),summarise,Entries=length(Human_state_chromHMM)/dim(human_mouse_orthologs_chromHMM_paired)[1])

c = ggplot(human_mouse_orthologs_chromHMM_paired,aes(x=Human_state_chromHMM,fill=Mouse_state_chromHMM)) + geom_bar(position="fill") + xlab("State (human)") + ylab("Proportion in state\nin mouse") + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.direction="horizontal",legend.position="bottom",legend.box.margin = margin(t=-10),legend.key.size = unit(3,'mm')) + scale_x_discrete(labels=as.vector(apply(state_freq_chromHMM,1,function(x) paste(x[1],"\n(",round(as.numeric(x[2])*100,1),"%)",sep="")))) + guides(fill = guide_legend(nrow=3,title.position = "top")) + scale_fill_manual(values=mouse_chromHMM_colors,name="State (mouse)")

d = ggplot(hg19_orthologs_specific,aes(x=factor(1),y=Count,fill=Subset,label=paste(round(100*Proportion,0),"%",sep=""))) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank(),strip.text.x = element_text(size = 8),legend.key.size = unit(3,'mm'),legend.position="bottom") + facet_grid(State~Tissue,labeller=labeller(Tissue = setNames(c("Brain","Instestine","Stomach","Heart","Lung"),unique(hg19_orthologs_specific$Tissue)))) + scale_fill_discrete(labels=c("<2 samples","Tissue-specific","2-8 samples","> 8 samples"),name="Category") + guides(fill = guide_legend(nrow = 2)) + geom_text(position = position_fill(vjust = 0.5),size=2)

grid.arrange(a,b,c,d,nrow=2,layout_matrix=rbind(c(1,4,4),c(2,3,3)),heights=c(0.45,0.55))
```

### Supplementary Figure 19

```{r Figure S19, echo=FALSE,fig.height=6}
ortholog_stats = melt(rmsk_TE_class[,c("class_update","Count","Mouse_orthologs")],id.var="class_update")
colnames(ortholog_stats)[2:3] = c("Category","TEs")
ortholog_stats = ddply(ortholog_stats,.(Category),transform,Percent=100*TEs/sum(TEs))

a = ggplot(ortholog_stats,aes(x=Category,y=TEs,fill=class_update,label=paste(format(TEs, big.mark=",", scientific=FALSE)," (",round(Percent,0),"%) ",sep=""))) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank(),plot.title = element_text(size=8)) + ggtitle("TEs") + geom_text(position = position_fill(vjust = 0.5),size=2)

ortholog_stats_subfam = melt(rmsk_TE_class[,c("class_update","Subfamilies","Mouse_ortholog_subfamily")],id.var="class_update")
colnames(ortholog_stats_subfam)[2:3] = c("Category","Count")
ortholog_stats_subfam = ddply(ortholog_stats_subfam,.(Category),transform,Percent=100*Count/sum(Count))
  
b = ggplot(ortholog_stats_subfam,aes(x=Category,y=Count,fill=class_update,label=paste(format(Count, big.mark=",", scientific=FALSE)," (",round(Percent,0),"%) ",sep=""))) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank(),plot.title = element_text(size=8)) + ggtitle("Subfamilies") + geom_text(position = position_fill(vjust = 0.5),size=2)

c = ggplot(hg19_mm10_TE_WGBS_subfamily_hypo_paired,aes(x=Human_hypo,y=Mouse_hypo,color=Anatomy,shape=Age)) + geom_point() + scale_color_manual(values=anatomy_colors,guide=FALSE) + xlab("Subfamily proportion (human)") + ylab("Subfamily proportion (mouse)") + geom_abline(slope=1,linetype="dotted") + geom_smooth(aes(linetype=Age),method='lm',fullrange=TRUE,alpha=0.1) + xlim(0,0.4) + ylim(0,0.4) + guides(color=FALSE,shape=FALSE,linetype=FALSE) + coord_equal() + scale_linetype_manual(values=c(1,2),guide=FALSE)

d = ggplot(hg19_mm10_chromHMM_subfamily[which(hg19_mm10_chromHMM_subfamily$Human_state_chromHMM == "1_TssA" & hg19_mm10_chromHMM_subfamily$Mouse_state_chromHMM == "TssA"),],aes(x=Percent.x,y=Percent.y,color=Anatomy,shape=Age)) + geom_point() + scale_color_manual(values=anatomy_colors,guide=FALSE) + xlab("Subfamily proportion (human)") + ylab("Subfamily proportion (mouse)") + geom_abline(slope=1,linetype="dotted") + geom_smooth(aes(linetype=Age),method='lm',fullrange=TRUE,alpha=0.1) + xlim(0,0.05) + ylim(0,0.05) + guides(color=FALSE,shape=FALSE,linetype=FALSE) + coord_equal() + scale_linetype_manual(values=c(1,2),guide=FALSE)

e = ggplot(hg19_mm10_chromHMM_subfamily_active,aes(x=Percent.x,y=Percent.y,color=Anatomy,shape=Age)) + geom_point() + scale_color_manual(values=anatomy_colors,guide=FALSE) + xlab("Subfamily proportion (human)") + ylab("Subfamily proportion (mouse)") + geom_abline(slope=1,linetype="dotted") + geom_smooth(aes(linetype=Age),method='lm',fullrange=TRUE,alpha=0.1) + xlim(0,0.4) + ylim(0,0.4) + guides(color=FALSE,shape=FALSE,linetype=FALSE) + coord_equal() + scale_linetype_manual(values=c(1,2),guide=FALSE)

legend = get_legend(ggplot(hg19_mm10_chromHMM_subfamily_active,aes(x=Percent.x,y=Percent.y,color=Anatomy,shape=Age)) + geom_point() + scale_color_manual(values=anatomy_colors) + scale_shape_discrete(labels=age_labels) + geom_smooth(aes(linetype=Age),method='lm',fullrange=TRUE,alpha=0.1) + scale_linetype_manual(labels=age_labels,values=c(1,2)) + xlab("Human") + ylab("Mouse") + theme(legend.position = "bottom",legend.direction = "horizontal",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')))

grid.arrange(a,b,legend_class,c,d,e,legend,nrow=4,layout_matrix=rbind(c(1,NA,2),c(3),c(4,5,6),c(7)),heights=c(0.45,0.05,0.45,0.05))
```

### Human-mouse ortholog analysis

```{bash count liftover, eval=FALSE}
# Number of hg19 TEs that liftover to mm10
wc -l Mouse/liftover/rmsk_TE_hg19tomm10.bed
```

```{r mouse ortholog analysis}
# Number of mouse TEs
MOUSE_TE = dim(mm10_rmsk_TE)[1]
MOUSE_TE

# Number of human TEs with orthologs
HG19_TE = dim(unique(human_mouse_orthologs_mm10[,hg19_coordinates]))[1]

# Number of mouse TEs with orthologs
MM10_TE = dim(unique(human_mouse_orthologs_mm10[,mm10_coordinates]))[1]

# Number of hg19 TEs that liftover
1265775
1265775/NUM_TE

# Number of unique orthologous TEs in hg19->mm10, hg19, and mm10
dim(unique(human_mouse_orthologs_mm10[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")]))[1]
HG19_TE
HG19_TE/NUM_TE
MM10_TE

# Number, % of human TEs with 2+ human mm10 TEs
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_hg19))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_mm10))[2:11])/length(unique(human_mouse_orthologs_mm10$human_TE_mm10))

# Number, % of human TEs with 2+ mouse TEs
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$human_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$human_TE_mm10))[2:10])/length(unique(human_mouse_orthologs_mm10$human_TE_mm10))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$human_TE_hg19))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$human_TE_hg19))[2:10])/length(unique(human_mouse_orthologs_mm10$human_TE_hg19))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$mouse_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$mouse_TE_mm10))[2:15])/length(unique(human_mouse_orthologs_mm10$mouse_TE_mm10))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$mouse_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$mouse_TE_mm10))[2:25])/length(unique(human_mouse_orthologs_mm10$mouse_TE_mm10))

# Average length of those TEs
median(unique(human_mouse_orthologs_mm10[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")])$human_stop_mm10-
         unique(human_mouse_orthologs_mm10[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")])$human_start_mm10)
median(unique(human_mouse_orthologs_mm10[,hg19_coordinates])$human_stop_hg19-unique(human_mouse_orthologs_mm10[,hg19_coordinates])$human_start_hg19)
median(unique(human_mouse_orthologs_mm10[,mm10_coordinates])$mouse_stop_mm10-unique(human_mouse_orthologs_mm10[,mm10_coordinates])$mouse_start_mm10)

# Class distribution
test = rmsk_TE_class[,c("class_update","Count","Mouse_orthologs","Subfamilies","Mouse_ortholog_subfamily")]
test[,2:5] = apply(test[,2:5],2,function(x) x/sum(x))
test 

# Proportion of hg19 TEs and subfamilies with ortholog in mouse, by class
rmsk_TE_class$Mouse_orthologs/rmsk_TE_class$Count
rmsk_TE_class$Mouse_ortholog_subfamily/rmsk_TE_class$Subfamilies

# Average age of TEs with mouse orthologs
median(rmsk_TE$JC_distance)
median(merge(rmsk_TE,unique(human_mouse_orthologs_mm10[,hg19_coordinates]),by.x=TE_coordinates,by.y=hg19_coordinates[c(1:4,6,5,7)])$JC_distance)
```

### Human-mouse ortholog WGBS analysis

Finding the effect size for Chi-squared tests. For all methylation states, all samples, the effect size is small (0.1227), according to [http://rcompanion.org/handbook/H_10.html| rcompanion's website]. For each sample pair, Cramer's V ranges from 0.099 to 0.143 (all small effect sizes). The effect size is larger for hypomethylated vs. not (0.1954 for all samples, range 0.1632 to 0.2278 for each pair, which are mostly medium effect sizes). For the chromHMM states, the effect sizes for all samples is 0.2308, range 0.216 to 0.2623, all medium effect sizes.

```{r mouse WGBS analysis}
# Number, % of hg19 orthologs with CpGs
x = merge(TE_meth_average[,c(TE_coordinates,"CpGs")],unique(human_mouse_orthologs_mm10[,hg19_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=hg19_coordinates)
dim(x)
dim(x)[1]/HG19_TE

# Average number of CpGs per hg19 ortholog with CpGs
median(x$CpGs)
rm(x)

# Number, % of mm10 orthologs with CpGs
y = merge(mm10_rmsk_TE_WGBS,unique(human_mouse_orthologs_mm10[,mm10_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=mm10_coordinates)
dim(y)
dim(y)[1]/MM10_TE

# Average number of CpGs per mm10 ortholog with CpGs
median(y$CpGs)
rm(y)

# Number of unique orthologous TEs with CpGs in mm10 and hg19 (both have CpGs)
dim(human_mouse_orthologs_WGBS)[1]
dim(unique(human_mouse_orthologs_WGBS[,mm10_coordinates]))[1]
dim(unique(human_mouse_orthologs_WGBS[,hg19_coordinates]))[1]

# TEs with meth data in at least one mouse/human sample
test = human_mouse_orthologs_WGBS[which(apply(human_mouse_orthologs_WGBS,1,function(x) sum(is.na(x[17:23])) < 7) & apply(human_mouse_orthologs_WGBS,1,function(x) sum(is.na(x[24:30])) < 7)),]

## Number of human (hg19) TEs
dim(unique(human_mouse_orthologs_WGBS[,hg19_coordinates]))[1]-dim(unique(test[,hg19_coordinates]))[1]

## Number of mouse TEs
dim(unique(human_mouse_orthologs_WGBS[,mm10_coordinates]))[1]-dim(unique(test[,mm10_coordinates]))[1]

# Number of TEs with methylation data per sample
ddply(human_mouse_orthologs_WGBS_paired,.(Human_sample,Mouse_sample_WGBS),summarise,Both=sum(!is.na(Human_methylation) & !is.na(Mouse_methylation)))

# Table of human/mouse states, overall
table(human_mouse_orthologs_WGBS_paired[,c("Mouse_state_WGBS","Human_state_WGBS")])

# Chi-sq test for human-mouse pairs, all methylation states, overall and by sample
ddply(human_mouse_orthologs_WGBS_paired,.(),summarise,Pvalue=unlist(chisq.test(x=Human_state_WGBS,y=Mouse_state_WGBS))["p.value"])
cramerV(x=as.vector(human_mouse_orthologs_WGBS_paired$Human_state_WGBS),y=as.vector(human_mouse_orthologs_WGBS_paired$Mouse_state_WGBS))

ddply(human_mouse_orthologs_WGBS_paired,.(Human_sample),summarise,Pvalue=unlist(chisq.test(x=Human_state_WGBS,y=Mouse_state_WGBS))["p.value"])
by(human_mouse_orthologs_WGBS_paired,human_mouse_orthologs_WGBS_paired$Human_sample,function(x) cramerV(x=as.vector(x$Human_state_WGBS),y=as.vector(x$Mouse_state_WGBS)))

# Hypomethylated vs. all other states
test = human_mouse_orthologs_WGBS_paired[,c("Human_sample","Human_state_WGBS","Mouse_state_WGBS")]
test$Human_state_WGBS = mapvalues(test$Human_state_WGBS,meth_states[2:3],rep("Missing",2))
test$Mouse_state_WGBS = mapvalues(test$Mouse_state_WGBS,meth_states[2:3],rep("Missing",2))

ddply(test,.(),summarise,Pvalue=unlist(chisq.test(x=Human_state_WGBS,y=Mouse_state_WGBS))["p.value"])
cramerV(x=as.vector(test$Human_state_WGBS),y=as.vector(test$Mouse_state_WGBS))

ddply(test,.(Human_sample),summarise,Pvalue=unlist(chisq.test(x=Human_state_WGBS,y=Mouse_state_WGBS))["p.value"])
by(test,test$Human_sample,function(x) cramerV(x=as.vector(x$Human_state_WGBS),y=as.vector(x$Mouse_state_WGBS)))

# TE orthologs hypomethylated in both human and mouse
human_mouse_orthologs_WGBS_hypo = human_mouse_orthologs_WGBS_paired[which(human_mouse_orthologs_WGBS_paired$Mouse_state_WGBS == "Hypomethylated" & human_mouse_orthologs_WGBS_paired$Human_state_WGBS == "Hypomethylated"),]

# Number of pairs where TE is hypomethylated in both
dim(human_mouse_orthologs_WGBS_hypo)
dim(unique(human_mouse_orthologs_WGBS_hypo[,hg19_coordinates]))
dim(unique(human_mouse_orthologs_WGBS_hypo[,mm10_coordinates]))
ddply(human_mouse_orthologs_WGBS_hypo,.(Human_sample),function(x) dim(x)[1])
```

### Age analysis (not shown)

```{r age analysis, eval=FALSE}
test = ddply(human_mouse_orthologs_WGBS_paired,.(human_chr_hg19,human_start_hg19,human_subfamily,human_class,mouse_chr_mm10,mouse_start_mm10,JC_distance),summarise,Concordant=sum(Human_state_WGBS == Mouse_state_WGBS),Con_hypo=sum(Human_state_WGBS == Mouse_state_WGBS & Human_state_WGBS == "Hypomethylated"))
cor.test(test$JC_distance,test$Con_hypo,method="spearman")
cor.test(test$JC_distance,test$Concordant,method="spearman")
by(test,test$human_class,function(x) cor.test(x$JC_distance,x$Con_hypo,method="spearman"))
by(test,test$human_class,function(x) cor.test(x$JC_distance,x$Concordant,method="spearman"))
```

### Human-mouse ortholog chromHMM analysis

```{r mouse chromHMM analysis}
# Confirming all orthologous TEs have chromHMM state
count_na(human_mouse_orthologs_chromHMM_paired)
dim(unique(hg19_orthologs_chromHMM[,hg19_coordinates]))[1]
dim(hg19_orthologs_chromHMM)
dim(unique(human_mouse_orthologs_chromHMM_paired[,c(hg19_coordinates,"Human_sample","Human_state_chromHMM")]))[1]
dim(unique(mm10_orthologs_chromHMM[,mm10_coordinates]))[1]
dim(mm10_orthologs_chromHMM)
dim(unique(human_mouse_orthologs_chromHMM_paired[,c(mm10_coordinates,"Mouse_sample_chromHMM","Mouse_state_chromHMM")]))[1]

# Number of orthologous TEs in each chromHMM state
human_mouse_chromHMM_table = table(human_mouse_orthologs_chromHMM_paired$Human_state_chromHMM,human_mouse_orthologs_chromHMM_paired$Mouse_state_chromHMM)
human_mouse_chromHMM_table

# Chi-squared, overall and by sample
ddply(human_mouse_orthologs_chromHMM_paired,.(),summarise,Pvalue=unlist(chisq.test(x=Human_state_chromHMM,y=Mouse_state_chromHMM))["p.value"])
cramerV(x=as.vector(human_mouse_orthologs_chromHMM_paired$Human_state_chromHMM),y=as.vector(human_mouse_orthologs_chromHMM_paired$Mouse_state_chromHMM))

ddply(human_mouse_orthologs_chromHMM_paired,.(Human_sample),summarise,Pvalue=unlist(chisq.test(x=Human_state_chromHMM,y=Mouse_state_chromHMM))["p.value"])
by(human_mouse_orthologs_chromHMM_paired,human_mouse_orthologs_chromHMM_paired$Human_sample,function(x) cramerV(x=as.vector(x$Human_state_chromHMM),y=as.vector(x$Mouse_state_chromHMM)))

# Number of pairs and human/mouse TEs, both in promoter state
x = human_mouse_orthologs_chromHMM_paired[which(human_mouse_orthologs_chromHMM_paired$Human_state_chromHMM == "1_TssA" & human_mouse_orthologs_chromHMM_paired$Mouse_state_chromHMM == "TssA"),]
dim(x)[1]
dim(unique(x[,hg19_coordinates]))[1]
dim(unique(x[,mm10_coordinates]))[1]
ddply(x,.(Human_sample),function(x) dim(x)[1])

# hg19 TE class per state
table(unique(x[,hg19_coordinates])$human_class)

# Number of pairs and human/mouse TEs, both in active regulatory state
y = unique(human_mouse_orthologs_chromHMM_paired[which(human_mouse_orthologs_chromHMM_paired$Human_state_chromHMM %in% chromHMM_states[c(1:3,6:7)] & human_mouse_orthologs_chromHMM_paired$Mouse_state_chromHMM %in% mouse_chromHMM_states[c(1:3,6:8)]),c(hg19_coordinates,mm10_coordinates,"Human_sample","Mouse_sample_chromHMM")])
dim(y)[1]
dim(unique(y[,hg19_coordinates]))[1]
dim(unique(y[,mm10_coordinates]))[1]
ddply(y,.(Human_sample),function(z) dim(z)[1])

# hg19 TE class per state
table(unique(y[,hg19_coordinates])$human_class)

# Tissue-specific TEs
## Promoter state
as.data.frame(t(sapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) tissue_matrix(x,hg19_mm10_chromHMM_promoter))))

## Active regulatory state
as.data.frame(t(sapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) tissue_matrix(x,hg19_mm10_chromHMM_active))))

# TEs constitutively on/off
## Promoter state
dim(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples > 8 & hg19_mm10_chromHMM_promoter$Mouse_samples > 8),])
dim(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples > 8 & hg19_mm10_chromHMM_promoter$Mouse_samples < 2),])
dim(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples < 2 & hg19_mm10_chromHMM_promoter$Mouse_samples > 8),])

## Active regulatory states
dim(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples > 8 & hg19_mm10_chromHMM_active$Mouse_samples > 8),])
dim(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples > 8 & hg19_mm10_chromHMM_active$Mouse_samples < 2),])
dim(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples < 2 & hg19_mm10_chromHMM_active$Mouse_samples > 8),])
```

### Subfamily analysis

```{r ortholog subfamily analysis}
# Number of mouse subfamilies
length(unique(mm10_rmsk_TE$subfamily))

# Number of subfamilies shared between human and mouse
dim(hg19_mm10_subfamily_count)[1]

# Average age of subfamilies also in mouse 
median(rmsk_TE_subfamily$Age)
median(rmsk_TE_subfamily[which(rmsk_TE_subfamily$subfamily %in% mm10_rmsk_TE$subfamily),]$Age)

# Number of subfamilies with at least 30 members in both
dim(hg19_mm10_subfamily_count[which(hg19_mm10_subfamily_count$Count.x > THRESHOLD_IK_MEMBER & hg19_mm10_subfamily_count$Count.y > THRESHOLD_IK_MEMBER),])[1]
dim(hg19_mm10_subfamily_count[which(hg19_mm10_subfamily_count$Count_CpGs.x > THRESHOLD_IK_MEMBER & hg19_mm10_subfamily_count$Count_CpGs.y > THRESHOLD_IK_MEMBER),])[1]

# Average ratio of subfamily members for shared subfamilies
median(hg19_mm10_subfamily_count$Count.y/hg19_mm10_subfamily_count$Count.x)
median(hg19_mm10_subfamily_count$Count_CpGs.y/hg19_mm10_subfamily_count$Count_CpGs.x)
head(hg19_mm10_subfamily_count[order(hg19_mm10_subfamily_count$Count.y/hg19_mm10_subfamily_count$Count.x),])
tail(hg19_mm10_subfamily_count[order(hg19_mm10_subfamily_count$Count.y/hg19_mm10_subfamily_count$Count.x),])

# Linear correlations
by(hg19_mm10_TE_WGBS_subfamily_hypo_paired,hg19_mm10_TE_WGBS_subfamily_hypo_paired$Human_sample,function(x) lm(Mouse_hypo~Human_hypo,x))
by(hg19_mm10_chromHMM_subfamily[which(hg19_mm10_chromHMM_subfamily$Human_state_chromHMM == "1_TssA" & hg19_mm10_chromHMM_subfamily$Mouse_state_chromHMM == "TssA"),],droplevels(hg19_mm10_chromHMM_subfamily[which(hg19_mm10_chromHMM_subfamily$Human_state_chromHMM == "1_TssA" & hg19_mm10_chromHMM_subfamily$Mouse_state_chromHMM == "TssA"),]$Human_sample),function(x) lm(Percent.y~Percent.x,x))
by(hg19_mm10_chromHMM_subfamily_active,hg19_mm10_chromHMM_subfamily_active$Human_sample,function(x) lm(Percent.y~Percent.x,x))
```

### Correlation analysis

Reviewer 1 brought to our attention that there is controversy over the mouseENCODE result that species cluster more strongly than tissues. They suggested the paper, "A reanalysis of mouse ENCODE comparative gene expression data" (F1000Research 2015). Essentially, the original study (Lin et al, "Comparison of the transcriptional landscape between human and mouse tissues", PNAS 2014 - with Mike Snyder, Joe Ecker, Tom Gingeras, Mike Beer) split human and mouse samples onto separate lanes, sometimes on separate machines. When the authors correct for this batch effect using ComBat (sva package), they see that tissues cluster more strongly than species (as they expected). More relevant for my paper, the original authors (Lin et al) analyzed the previously published Roadmap data but did not focus on the results because they acknowledged that they were confounded by batch effects, particularly by lab. 

```{r correlation boxplots, eval=FALSE}
c = ggplot(human_mouse_spearman,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + ylab("Spearman correlation") + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse")) + scale_y_continuous(limits=c(0,1)) + scale_fill_discrete(guide=FALSE) + geom_segment(x=1,xend=1.25,y=0.8,yend=0.8,color="red") + annotate("text",x=1.125,y=0.81,label="*",color="red",size=5) + xlab("Comparison")

d = ggplot(human_mouse_spearman_subfamily,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + ylab("Spearman correlation") + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse")) + scale_y_continuous(limits=c(0,1)) + scale_fill_discrete(guide=FALSE) + geom_segment(x=1,xend=1.25,y=0.87,yend=0.87,color="red") + annotate("text",x=1.125,y=0.88,label="*",color="red",size=5) + xlab("Comparison")

e = ggplot(human_mouse_spearman_subfamily_TssA,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + ylab("Spearman correlation") + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse")) + scale_y_continuous(limits=c(0,1)) + scale_fill_discrete(guide=FALSE) + geom_segment(x=1.75,xend=1.99,y=0.85,yend=0.85,color="red") + geom_segment(x=2.01,xend=2.25,y=0.85,yend=0.85,color="red") + geom_segment(x=1.75,xend=2.25,y=0.9,yend=0.9,color="red") + annotate("text",x=1.875,y=0.86,label="*",color="red",size=5) + annotate("text",x=2,y=0.91,label="*",color="red",size=5) + annotate("text",x=2.125,y=0.86,label="*",color="red",size=5) + xlab("Comparison")

f = ggplot(human_mouse_spearman_subfamily_active,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + ylab("Spearman correlation") + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse")) + scale_y_continuous(limits=c(0,1)) + scale_fill_discrete(guide=FALSE) + geom_segment(x=1,xend=1.25,y=0.9,yend=0.9,color="red") + annotate("text",x=1.125,y=0.91,label="*",color="red",size=5) + geom_segment(x=1.75,xend=1.99,y=0.85,yend=0.85,color="red") + geom_segment(x=2.01,xend=2.25,y=0.85,yend=0.85,color="red") + geom_segment(x=1.75,xend=2.25,y=0.9,yend=0.9,color="red") + annotate("text",x=1.875,y=0.86,label="*",color="red",size=5) + annotate("text",x=2,y=0.91,label="*",color="red",size=5) + annotate("text",x=2.125,y=0.86,label="*",color="red",size=5) + xlab("Comparison")

legend = get_legend(ggplot(human_mouse_spearman_subfamily_active,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + theme(legend.direction = "horizontal") + scale_fill_discrete(name="Tissue/Age",labels=c("Same/Same","Same/Different","Different/Same","Different/Different")))
```

```{r correlation analysis, eval=FALSE}
# Individual TE Spearman correlations
ddply(human_mouse_spearman,.(Test),summarise,Median=median(Correlation),IQR=IQR(Correlation))
pairwise.wilcox.test(human_mouse_spearman$Correlation,human_mouse_spearman$Test)
by(human_mouse_spearman,human_mouse_spearman$Test,function(x) pairwise.wilcox.test(x$Correlation,x$Tissue.Age))

# Subfamily Spearman correlations 
## Hypomethylated
ddply(human_mouse_spearman_subfamily,.(Test),summarise,Median=median(Correlation),IQR=IQR(Correlation))
pairwise.wilcox.test(human_mouse_spearman_subfamily$Correlation,human_mouse_spearman_subfamily$Test)
by(human_mouse_spearman_subfamily,human_mouse_spearman_subfamily$Test,function(x) pairwise.wilcox.test(x$Correlation,x$Tissue.Age))

## 1_TssA
ddply(human_mouse_spearman_subfamily_TssA,.(Test),summarise,Median=median(Correlation),IQR=IQR(Correlation))
pairwise.wilcox.test(human_mouse_spearman_subfamily_TssA$Correlation,human_mouse_spearman_subfamily_TssA$Test)
by(human_mouse_spearman_subfamily_TssA,human_mouse_spearman_subfamily_TssA$Test,function(x) pairwise.wilcox.test(x$Correlation,x$Tissue.Age))

## Active regulatory
ddply(human_mouse_spearman_subfamily_active,.(Test),summarise,Median=median(Correlation),IQR=IQR(Correlation))
pairwise.wilcox.test(human_mouse_spearman_subfamily_active$Correlation,human_mouse_spearman_subfamily_active$Test)
by(human_mouse_spearman_subfamily_active,human_mouse_spearman_subfamily_active$Test,function(x) pairwise.wilcox.test(x$Correlation,x$Tissue.Age))
```

### Human-mouse tissue-specific orthologs in same chromHMM state

```{r print candidates}
write.table(ldply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples < 5 & hg19_mm10_chromHMM_promoter[[paste(x,".x",sep="")]] == 2 & hg19_mm10_chromHMM_promoter$Mouse_samples < 5 & hg19_mm10_chromHMM_promoter[[paste(x,".y",sep="")]] == 2),]),file="Mouse/tissue_specific_promoter.txt",sep='\t',quote=FALSE,row.names=FALSE)

lapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) write.table(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples < 5 & hg19_mm10_chromHMM_active[[paste(x,".x",sep="")]] == 2 & hg19_mm10_chromHMM_active$Mouse_samples < 5 & hg19_mm10_chromHMM_active[[paste(x,".y",sep="")]] == 2),],file=paste("Mouse/tissue_specific_active_",x,".txt",sep=""),sep='\t',quote=FALSE,row.names=FALSE))
```

### Human-mouse constitutive orthologs

```{r print constitutive}
## Promoter state
write.table(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples > 8 & hg19_mm10_chromHMM_promoter$Mouse_samples > 8),],file="Mouse/constitutive_both_promoter.txt",sep='\t',quote=FALSE,row.names=FALSE)
write.table(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples > 8 & hg19_mm10_chromHMM_promoter$Mouse_samples < 2),],file="Mouse/constitutive_human_promoter.txt",sep='\t',quote=FALSE,row.names=FALSE)
write.table(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples < 2 & hg19_mm10_chromHMM_promoter$Mouse_samples > 8),],file="Mouse/constitutive_mouse_promoter.txt",sep='\t',quote=FALSE,row.names=FALSE)

## Active regulatory states
write.table(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples > 8 & hg19_mm10_chromHMM_active$Mouse_samples > 8),],file="Mouse/constitutive_both_active.txt",sep='\t',quote=FALSE,row.names=FALSE)
write.table(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples > 8 & hg19_mm10_chromHMM_active$Mouse_samples < 2),],file="Mouse/constitutive_human_active.txt",sep='\t',quote=FALSE,row.names=FALSE)
write.table(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples < 2 & hg19_mm10_chromHMM_active$Mouse_samples > 8),],file="Mouse/constitutive_mouse_active.txt",sep='\t',quote=FALSE,row.names=FALSE)
```

```{bash investigate ortholog, eval=FALSE}
# Tissue-specific orthologs in human and mouse that are intergenic
bedtools closest -a <(cut -f8-31 Mouse/tissue_specific_promoter.txt | tail -n +2 -  | sort -k1,1 -k2,2n - ) -b <(sort -k1,1 -k2,2n ~/genic_features/RefSeq/refseq_genes.txt) -D b -t all | awk -v OFS='\t' '{if($37 != 0) print $1, $2, $3, $4, $37}' | sort | uniq

## >50kb from the nearest gene
bedtools closest -a <(cat Mouse/tissue_specific_active_*.txt | cut -f8-31 | awk '{if($1 != "human_chr_hg19") print $0}' -  | sort -k1,1 -k2,2n - ) -b <(sort -k1,1 -k2,2n ~/genic_features/RefSeq/refseq_genes.txt) -D b -t all | awk -v OFS='\t' '{if(($37 > 50000) || ($37 < -50000)) print $1, $2, $3, $4, $37}' | sort | uniq | wc -l

## >100kb from the nearest gene
bedtools closest -a <(cat Mouse/tissue_specific_active_*.txt | cut -f8-31 | awk '{if($1 != "human_chr_hg19") print $0}' -  | sort -k1,1 -k2,2n - ) -b <(sort -k1,1 -k2,2n ~/genic_features/RefSeq/refseq_genes.txt) -D b -t all | awk -v OFS='\t' '{if(($37 > 100000) || ($37 < -100000)) print $1, $2, $3, $4, $37}' | sort | uniq | wc -l

# Constitutively active TEs in one or both species, intergenic
bedtools closest -a <(cut -f8-31 Mouse/constitutive_both_promoter.txt | tail -n +2 -  | sort -k1,1 -k2,2n - ) -b <(sort -k1,1 -k2,2n ~/genic_features/RefSeq/refseq_genes.txt) -D b -t all | awk -v OFS='\t' '{if($37 != 0) print $1, $2, $3, $4, $37}' | sort | uniq
```

```{bash make browser, eval=FALSE}
# Process tracks for display on the Browser (in public_html/mouseENCODE)

# mm10 chromHMM categorical tracks
## Fourth column must be positive integer
while read a b c; do gunzip raw_data/mouse/chromHMM/$c\.bed.gz; awk -v OFS='\t' '{if($4 == "TssA"){print $1, $2, $3, 1} else if ($4 == "TssAFlnk1"){print $1, $2, $3, 2} else if ($4 == "TssAFlnk2"){print $1, $2, $3, 3} else if ($4 == "Tx1"){print $1, $2, $3, 4} else if ($4 == "Tx2"){print $1, $2, $3, 5} else if ($4 == "Enh"){print $1, $2, $3, 6} else if ($4 == "EnhLo1"){print $1, $2, $3, 7} else if ($4 == "EnhLo2"){print $1, $2, $3, 8} else if ($4 == "HetCons"){print $1, $2, $3, 9} else if ($4 == "HetFac"){print $1, $2, $3, 10} else if ($4 == "TssBiv"){print $1, $2, $3, 11} else if ($4 == "EnhPois1"){print $1, $2, $3, 12} else if ($4 == "EnhPois2"){print $1, $2, $3, 13} else if ($4 == "QuiesG"){print $1, $2, $3, 14} else if ($4 == "Quies"){print $1, $2, $3, 15}}' raw_data/mouse/chromHMM/$c\.bed | sort -k1,1 -k2,2n - > /tavern/epehrsson/mouseENCODE/signal_tracks/$c\.bed; bgzip /tavern/epehrsson/mouseENCODE/signal_tracks/$c\.bed; tabix -p bed /tavern/epehrsson/mouseENCODE/signal_tracks/$c\.bed.gz; done < Mouse/human_mouse_samples.txt &

# mm10 WGBS tracks 
for file in ~/TE_landscape/raw_data/mouse/WGBS/*.bed; do awk -v OFS='\t' '{print $1, $2, $3, $10}' $file | sort -k1,1 -k2,2n - > $(basename $file .bed)_readDepth.bed; bgzip ~/public_html/mouseENCODE/$(basename $file .bed)_readDepth.bed; done
for file in ~/TE_landscape/raw_data/mouse/WGBS/*.bed; do awk -v OFS='\t' '{print $1, $2, $3, $11}' $file | sort -k1,1 -k2,2n - > $(basename $file .bed)_meth.bed; bgzip ~/public_html/mouseENCODE/$(basename $file .bed)_meth.bed; done
for file in *.bed.gz; do tabix -p bed $file; done
```

http://epigenomegateway.wustl.edu/browser/?genome=hg19&session=CrUnGegbbU&statusId=1624390483 (Remy the chef)

# Methods 

## Basic statistics

```{r QC statistics}
# Proportion of TEs on chrY
NUM_TE-NUM_TE_noY
(NUM_TE-NUM_TE_noY)/NUM_TE

# Total bp in TEs
MERGED_TE_WIDTH

# Total TE bases (redundant)
sum(rmsk_TE$Length)-MERGED_TE_WIDTH
sum(rmsk_TE$Length)/MERGED_TE_WIDTH

# CpGs in TEs, genome
TE_CPGS
ALL_CPGS
```

## Blacklist TEs

```{r blacklist}
# Number of TEs overlapping blacklist
dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1]/NUM_TE

# Number of subfamilies overlapping blacklist
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$subfamily)))

# Which chromosomes?
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$chromosome)))

dim(rmsk_TE_subfamily_measure[which(rmsk_TE_subfamily_measure$blacklist > 0.01),c("subfamily","family","class_update","Count","blacklist")])
tail(rmsk_TE_subfamily_measure[order(rmsk_TE_subfamily_measure$blacklist),c("subfamily","family","class_update","Count","blacklist")])
```

# Supplement

```{bash peak widths, eval=FALSE}
# ChromHMM, block size by state and sample
#TE_landscape/chromHMM/chromHMM_blocks.txt
while read line; do awk -v OFS='\t' -v sample=$line '{print $3-$2, $4, sample}' raw_data/chromHMM/$line\_15_coreMarks_mnemonics.bed >> chromHMM/chromHMM_blocks.txt; done < sample_lists/mnemonics.txt

#TE_landscape/DNase/peak_widths.txt
while read line; do awk -v OFS='\t' -v sample=$line '{print $3-$2, sample}' ../raw_data/DNase/DNase_narrow_peaks/$line\-DNase.macs2.narrowPeak >> peak_widths.txt; done < ../sample_lists/DNase_samples.txt

#TE_landscape/raw_data/H3K27ac/H3K27ac_narrow_peaks/peak_widths.txt
while read line; do awk -v OFS='\t' -v sample=$line '{print $3-$2, sample}' $line\-H3K27ac.narrowPeak >> peak_widths.txt; done < ../../../sample_lists/H3K27ac_samples.txt
```

```{bash histone profile, eval=FALSE}
# Updated with summit TEs 5/29/18 to 6/19/18 (should be run on htcf)

# Each TE extended from center to 10kb region, excluding those that would extend past the ends of the chromosomes.	 

#TE_landscape/features/TEs/rmsk_TE_10kb.txt	
while read chr size; do awk -v OFS="\t" -v chr=$chr -v size=$size '{if($1 == chr && $2 > 4999 && ($3+5000 < size)) print $1,int($2+(($3-$2)/2))-5000,int($2+(($3-$2)/2))+5000,$1,$2,$3,$4,$5,$6,$7}' rmsk_TE.txt >> rmsk_TE_10kb.txt; done < hg19_standard.genome

#TE_landscape/features/TEs/rmsk_other_10kb.txt	
while read chr size; do awk -v OFS="\t" -v chr=$chr -v size=$size '{if($1 == chr && $2 > 4999 && ($3+5000 < size)) print $1,int($2+(($3-$2)/2))-5000,int($2+(($3-$2)/2))+5000,$1,$2,$3,$4,$5,$6,$7}' rmsk_other.txt >> rmsk_other_10kb.txt; done < hg19_standard.genome

# 10kb-extended TE bed file, rearranged so original coordinates are first	 

#TE_landscape/features/TEs/rmsk_TE_10kb.bed	
awk -v OFS="\t" '{print $4,$5,$6,$7,$8,$9,$10,$1,$2,$3}' /bar/epehrsson/TE_landscape/rmsk_TE_10kb.txt > rmsk_TE_10kb.bed

#TE_landscape/features/TEs/rmsk_other_10kb.bed	
awk -v OFS="\t" '{print $4,$5,$6,$7,$8,$9,$10,$1,$2,$3}' rmsk_other_10kb.txt > /scratch/ecp/rmsk_other_10kb.bed

# Combined into one file
cat features/TEs/rmsk_*_10kb.bed > features/TEs/rmsk_TEother_10kb.bed

# TEs in state in sample
awk -v OFS='\t' '{print $0 > "ever/rmsk_TEother_"$8".txt"}' rmsk_TEother_chromHMM_summit_sorted.txt
while read line; do echo $line; awk -v OFS='\t' '{if($10 != "8_ZNF/Rpts") print $0 > "ever/rmsk_TEother_"$8"_"$10".txt"}' ever/rmsk_TEother_$line\.txt; done < mnemonics.txt
while read line; do echo $line; awk -v OFS='\t' '{if($10 == "8_ZNF/Rpts") print $0 > "ever/rmsk_TEother_"$8"_8_ZNF.Rpts.txt"}' ever/rmsk_TEother_$line\.txt; done < mnemonics.txt
 
# Finds level of histone modifications and DNase over 10kb region centered on TEs in bins	
intersect_histone.sh
# Finds level of DNA methylation over 10kb region centered on TEs, in bins
intersect_meth.sh

# Epigenetic profiles by state and mark
#TE_landscape/compare_marks/profile_histone/rmsk_TEother_[state]_[mark]_average.txt [108 files]
# marks.txt is a list of 9 
while read state; do while read mark; do echo $state $mark; cat averages/rmsk_TEother_E*_$state\_$mark\_average.txt > rmsk_TEother_$state\_$mark\_average.tmp; python calculate_bin_average_all.py rmsk_TEother_$state\_$mark\_average.tmp rmsk_TEother_$state\_$mark\_average.txt; rm rmsk_TEother_$state\_$mark\_average.tmp; done < marks.txt; done < chromHMM_states.txt

# Combined histone profiles
while read state; do while read mark; do awk -v OFS='\t' -v mark=$mark -v state=$state '{print $0, mark, state}' rmsk_TEother_$state\_$mark\_average.txt >> TEother_average.txt; done < marks.txt; done < chromHMM_states.txt
```

### Supplementary Figure 21

```{r Figure S21 scripts, echo=FALSE, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/state_lengths.R")

# Number vs. length of blocks/peaks by sample
all_lengths_blocks = ddply(all_lengths,.(State,Sample),summarise,Mean=mean(Length),SD=sd(Length),Median=median(Length),Blocks=length(Length))
all_lengths_blocks = merge(all_lengths_blocks,metadata[,c("Sample",sample_categories)],by="Sample",all.x=TRUE)

all_lengths_blocks_corr = ddply(all_lengths_blocks,.(State),summarise,Cor=as.numeric(unlist(cor.test(Blocks,Median,method="spearman"))["estimate.rho"]),Pvalue=as.numeric(unlist(cor.test(Blocks,Median,method="spearman"))["p.value"]))
```

```{r Figure S23, echo=FALSE}
source("R_scripts/histone_modification.R")

a = ggplot(histones,aes(x=Position/1000,y=Level,color=Mark,group=Mark)) + geom_line(size=1.5,alpha=0.7) + theme(panel.grid=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + ylab("Fold enrichment over input") + scale_color_manual(labels=c(levels(histones$Mark)[1:8],"DNA methylation"),values=c(gg_color_hue(8),"black")) + xlab("Distance from TE center (kbp)")+ylim(0,12) + facet_wrap(~ State,nrow=3) + scale_y_continuous(sec.axis = sec_axis(~./12.5, name = "DNA methylation level"))

b = ggplot(all_lengths_blocks,aes(x=log10(Blocks),y=log10(Median))) + geom_point(aes(color=Group)) + xlab("Number of blocks/peaks (log10)") + ylab("Median length of blocks/peaks (log10(bp))") + scale_color_manual(values=group_colors) + facet_wrap(~State,scales="free",labeller=labeller(State=all_state_labels)) + geom_smooth(method="lm",se=FALSE,color="black") + geom_text(data=all_lengths_blocks_corr,aes(x = -Inf, y = Inf,label=round(Cor,2)),hjust=-0.1,vjust=1.5,size=3) + scale_y_continuous(breaks=pretty_breaks(n=3)) + scale_x_continuous(breaks=pretty_breaks(n=3)) + theme(legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm'))

grid.arrange(a,b,heights=c(0.45,0.55))
```

### State/peak lengths analysis

```{r state peak lengths analysis, cache=TRUE}
# Length of states
ddply(all_lengths,.(State),summarise,Mean=mean(Length),SD=sd(Length),Median=median(Length))

# Differences in length of states by sample
ddply(all_lengths,.(State),summarise,KW_length=unlist(kruskal.test(Length~Sample))["p.value"])

# Differences in number of blocks by sample
test = ddply(all_lengths_blocks,.(State),summarise,Range_number=max(Blocks)-min(Blocks),CV_number=sd(Blocks)/mean(Blocks),Range_length=max(Median)-min(Median),CV_length=sd(Median)/mean(Median))
test[order(test$CV_number),]
test[order(test$CV_length),]

# Comparing number of blocks to length of states
all_lengths_blocks_corr[order(as.numeric(all_lengths_blocks_corr$Cor)),]

rm(all_lengths)
```

### Mappability of genome/TE bases

```{bash average mappability, eval=FALSE}
# 36bp mappability, genome-wide	 
#TE_landscape/mappability/wgEncodeCrgMapabilityAlign36mer.bedGraph	
wget hgdownload.cse.ucsc.edu/gbdb/hg19/bbi/wgEncodeCrgMapabilityAlign36mer.bw

# From raw file
~/bin/bigWigToBedGraph wgEncodeCrgMapabilityAlign36mer.bw.2 wgEncodeCrgMapabilityAlign36mer.bedGraph

# Average mappability, whole genome
awk '{sum+=($4*($3-$2));total+=$3-$2;}END{print sum/total}' wgEncodeCrgMapabilityAlign36mer.bedGraph

# Average mappability, TEs 	 
split -l 5000000 ~/TE_landscape/mappability/wgEncodeCrgMapabilityAlign36mer.bedGraph
for file in x*; do bedtools intersect -wo -a rmsk_TEother_merge.txt -b $file >> rmsk_TEother_merge_map.txt; done
awk '{sum+=($7*$8);total+=$8;}END{print sum/total}' rmsk_TEother_merge_map.txt
```

```{r mappability}
# Mappability whole genome, TEs (8/27/17)
GENOME_MAP=0.789192
GENOME_MAP

TE_MAP=0.624207
TE_MAP

# Mappability vs. length by class
cor.test(rmsk_TE$Length,rmsk_TE$mappability,method="spearman")
by(rmsk_TE,rmsk_TE$class_update,function(x) cor.test(x$Length,x$mappability,method="spearman"))

# Correlations between mappability, CpGs, CpGs per length, number of samples Missing (WGBS)
correlate_TE_state[which(correlate_TE_state$State == "Missing" & correlate_TE_state$Metric %in% c("mappability","CpGs","CpGs_per_length")),]

# Correlation between mappability, number of samples Quiescent
correlate_TE_state[which(correlate_TE_state$Metric == "mappability" & correlate_TE_state$State == "15_Quies"),]
```