---
title: "TE landscape Draft"
author: "Erica Pehrsson"
date: "8/15/2017"
output: word_document
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load metadata and functions

```{r load metadata}
load("R_datasets/metadata.RData")
source("R_scripts/functions.R")
```

# Set universal plot parameters

```{r}
library(plyr)
library(reshape2)
library(vcd)
library(ggplot2)
library(NMF)
library(gridExtra)
library(RColorBrewer)
theme_set(theme_bw() %+replace%  theme(panel.background=element_rect(fill="white"),panel.border=element_rect(color="black",fill=NA),panel.grid=element_blank()))
```

# Define color and label vectors

```{r}
# chromHMM
chromHMM_states = c("1_TssA","2_TssAFlnk","3_TxFlnk","4_Tx","5_TxWk","6_EnhG","7_Enh","8_ZNF/Rpts","9_Het","10_TssBiv","11_BivFlnk","12_EnhBiv","13_ReprPC","14_ReprPCWk","15_Quies")
chromHMM_labels = setNames(c("Active TSS","Flanking active TSS","Transcription at gene 5'/3'","Strong transcription","Weak transcription","Genic enhancer","Enhancer","ZNF genes/repeats","Heterochromatin","Poised TSS","Flanking poised TSS/enhancer","Poised enhancer","Repressed Polycomb","Weak repressed Polycomb","Quiescent"),chromHMM_states)
chromHMM_colors = setNames(c("#66A61E","#1B9E77","#005A32","#E7298A","#F781BF","#41B6C4","#045A8D","#FFD92F","#99000D","#D95F02","#E6AB02","#A6761D","#6A3D9A","#CAB2D6","#666666"),chromHMM_states)
chromHMM_composite = setNames(c("Active regulatory","Active regulatory","Active regulatory","Transcribed","Transcribed","Active regulatory","Active regulatory","Zinc-finger/repeats","Repressed","Poised regulatory","Poised regulatory","Poised regulatory","Repressed","Repressed","Quiescent"),chromHMM_states)

# WGBS
meth_states = c("Hypomethylated","Intermediate","Hypermethylated","Missing")
meth_labels = setNames(c("Hypomethylated (<30%)","Intermediately methylated (30-70%)","Hypermethylated (>70%)","Missing data"),meth_states)
meth_colors = setNames(c("#F8766D","#00BFC4","#7CAE00","#C77CFF"),meth_states)

# All states
all_state_colors = c(chromHMM_colors,meth_colors,"aquamarine","tan1","black")
names(all_state_colors)[20:22] = c("DNase","H3K27ac","Expression")
all_state_labels = setNames(c(chromHMM_states,meth_states,"DNase","H3K27ac","RPKM > 1"),c(c(chromHMM_states,meth_states,"DNase","H3K27ac","Expression")))

# TE classification
class_colors = setNames(c("gold1","azure3","peru","lemonchiffon3","seagreen","lavenderblush4","brown1"),c("DNA","LINE","LTR","SINE","SVA","Other","RC"))

# Sample classification
group_colors = setNames(c("#924965","#4178AE","#E41A1C","#69608A","#B65C73","#FF9D0C","#678C69","#55A354","#E67326","#FFD924","#AF5B39","#D56F80","#999999","#C5912B","#C58DAA","#F182BC","#C2655D","#DAB92E","#000000"),c("ESC","ES-deriv","IMR90","iPSC","Mesench","Epithelial","HSC & B-cell","Blood & T-cell","Myosat","Neurosph","Adipose","Heart","Other","Brain","Digestive","Sm. Muscle","Muscle","Thymus","ENCODE2012"))
anatomy_colors = setNames(c("#B2DF8A","#33A02C","#FB9A99","#FFED6F","#80B1D3","#666666","#924965","#4178AE","#FDBF6F","#CAB2D6","#6A3D9A","#BEBADA","#FCCDE5","#BC80BD","#E7298A","#FF7F00","#69608A","#B3DE69","#CCEBC5","#000000","#FB8072","#FDB462","#8DD3C7","#D9D9D9","#E31A1C","#A6CEE3","#D95F02","#E6AB02","#66A61E","#B15928","#E31A1C","#8DD3C7","#E7298A","#A6761D","#666666"),c("ADRENAL","BLOOD","BONE","BRAIN","BREAST","CERVIX","ESC","ESC_DERIVED","FAT","GI_COLON","GI_DUODENUM","GI_ESOPHAGUS","GI_INTESTINE","GI_RECTUM","GI_STOMACH","HEART","IPSC","KIDNEY","LIVER","LUNG","MUSCLE","MUSCLE_LEG","OVARY","PANCREAS","PLACENTA","SKIN","SPLEEN","STROMAL_CONNECTIVE","THYMUS","VASCULAR",NA,NA,NA,NA,NA))

# Genic features
genic_labels = setNames(c("Promoter","5'UTR","CDS","3'UTR","Exon","Intron","Intergenic","CpG island"),c("promoters","5UTR","coding_exon","3UTR","exons","introns","intergenic","cpgIsland"))
```

# Define facet labels

```{r}
measure_labels = c("% genome","% TEs (either strand)","% TEs (same strand)","% feature in TEs")
names(measure_labels) = c("Genome_percent_genome","TEs_percent_TE","TEs_strand_percent_TE","Genome_percent_TE")

mark_labels = c("chromHMM (n=127)","WGBS (n=37)","DNase (n=53)","H3K27ac (n=98)")
names(mark_labels) = c("chromHMM","WGBS","DNase","H3K27ac")

group_labels = c("chromHMM","WGBS","DNase","H3K27ac","RPKM > 1")
names(group_labels) = c("chromHMM","WGBS","DNase","H3K27ac","Expression")

coding_labels = c("All","Protein-coding","Non-coding")
names(coding_labels) = c("All","pc","nc")

metric_labels = c("Proportion of TEs in state in any sample","Average % of samples in state","Average % of samples in state if ever in state")
names(metric_labels) = c("Proportion_ever","Samples_avg_all","Samples_avg_ever")

variable_labels = c("Mappability","Jukes-Cantor evolutionary distance","Length (bp)","CpGs","CpGs/bp")
names(variable_labels) = c("mappability","JC_distance","Length","CpGs","CpGs_per_length")
```

## Figure 1

Figure 1. Contribution of TEs to active and regulatory epigenetic states. (A) Total proportion of TEs, the entire genome, and RefSeq genic features annotated with each chromHMM state, overlapping DNase or H3K27ac peaks, or in each methylation state (WGBS, CpGs overlapping feature) across all samples annotated with that measure. Normalized by total bases annotated with chromHMM for that feature. Regions of overlap with TEs were excluded from all RefSeq features, and features were merged to remove redundant bases. Promoters represent 2000bp upstream / 500bp downstream of transcription start sites. (B) Total proportion of each state within TEs across all samples, ordered by proportion in TEs. Horizontal line represents the proportion of all genomic bases in TEs (“Total”). 

```{r Figure 1, echo=FALSE}
source("R_scripts/proportion.R")
source("R_scripts/contribution.R")

a = ggplot(combined_proportion[which(combined_proportion$Coding == "All"),],aes(x=Cohort,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Mean proportion of feature in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x=element_blank()) + scale_x_discrete(labels=c("TE","Genome",genic_labels[1:7])) + facet_wrap(~Group,nrow=2,scales="free",labeller=labeller(Group = mark_labels))

b = ggplot(contribution_TE[which(contribution_TE$Cohort == "TE"),],aes(x=State,y=Proportion,fill=State)) + geom_bar(stat="identity") + coord_flip() + ylab("Proportion of state in TEs") + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.ticks.x=element_blank()) + geom_hline(yintercept = 0.44918676) + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=contribution_TE$State[1:21]) + geom_text(aes(label=round(100*Proportion,1)),position="stack",hjust=1)

legend = get_legend(ggplot(combined_proportion[which(combined_proportion$Coding == "All"),],aes(x=Cohort,y=Proportion,fill=State)) + geom_bar(stat="identity") + scale_fill_manual(values=all_state_colors) + theme(axis.text.x = element_text(size=15,angle=90,hjust=1),axis.title.x=element_blank(),legend.title = element_blank(),legend.position="bottom",legend.direction = "horizontal") + facet_wrap(~Group,nrow=4))

grid.arrange(a,b,legend, nrow = 2,layout_matrix = rbind(c(1,2),c(3,3)),heights=c(0.8,0.2))
```

# Average proportion of TE CpGs hypomethylated per sample

```{r}
mean(apply(TE_CpG_meth,1,function(x) x[1]/sum(x)))
```

## Supplementary Figure 1

Supplementary Figure 1. Confirmation of chromHMM annotations. (A) Distribution of hg19 TE lengths by class, plotted on log10 scale. (B) Distribution lengths of chromHMM state blocks, DNase peaks, and H3K27ac peaks from all samples, plotted on log10 scale. (C) Distribution of methylation levels for CpGs overlapping TEs, by sample. Excludes CpGs with missing data. (D) Fold-enrichment ratios over input for ChIP-seq of seven histone modifications (five core histone modifications, H3K27ac, and H3K9ac) and DNase hypersensitivity as well as DNA methylation level over 10kb regions centered on TEs in each chromHMM state (≥1bp) in each sample, calculated in bins of 50bp. Values are averages of all instances of TEs in the state in each of the 127 samples. 

```{r Figure S1}
load("R_datasets/rmsk_TE.RData")

a = ggplot(rmsk_TE,aes(x=reorder(class_update,Length,median),y=log10(Length),fill=class_update)) + geom_boxplot() + ylab("log10(Length) of TEs") + theme(text=element_text(size=15),axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_fill_manual(values=class_colors,guide=FALSE)

source("R_scripts/state_lengths.R")

b = ggplot(all_lengths,aes(x=State,y=log10(Length),fill=State)) + geom_boxplot() + theme(text=element_text(size=15),axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_fill_manual(values=all_state_colors[c(1:15,20:21)],guide=FALSE) + ylab("log10(Length) of blocks") + scale_y_continuous(limits=seq(1,8,1))

load("R_datasets/cpg_density.RData")

c = ggplot(cpg,aes(x=Methylation,fill=Sample)) + geom_density(alpha=0.5,adjust=10) + xlab("Methylation level per CpG") + ylab("Density")

source("R_scripts/histone_modification.R")

d = ggplot(histones,aes(x=Bin,y=Level,color=Mark,group=Mark))+geom_point(size=0.5)+geom_line(size=1.5)+theme(axis.text.x = element_blank(),axis.ticks.x=element_blank(),text=element_text(size=15),panel.grid=element_blank(),legend.position = "bottom",legend.title = element_blank())+ylab("Fold enrichment over input") + scale_color_manual(labels=c(levels(histones$Mark)[1:8],"DNA methylation"),values=c(gg_color_hue(8),"black")) + xlab("Genomic position (+/-5kb around TE)")+ylim(0,12) + facet_wrap(~ State,nrow=3)

grid.arrange(a,b,c,d, nrow=3,layout_matrix=rbind(c(1,3),c(2,3),c(4,4)),heights=c(0.2,0.3,0.5))
```

# State lengths analysis

```{r}
# Stats for state/peak lengths
ddply(all_lengths,~State,Mean=mean(Length),SD=sd(Length))

# H3K27ac peak lengths by sample
mean(aggregate(data=H3K27ac_peaks,V1~V2,sum)$V1)
sd(aggregate(data=H3K27ac_peaks,V1~V2,sum)$V1)
```

## Supplementary Figure 2

Supplementary Figure 2. Overlap of TEs with RefSeq genic features. (A) Length of overlap between TEs and RefSeq features as proportion of genome, feature, or TEs. (B) Average proportion of TEs, the entire genome, and RefSeq genic features in each state across all samples annotated with that measure, split by protein-coding and non-coding features ("NM"/"NR"). Regions of overlap with TEs were excluded from all RefSeq features, and features were merged to remove redundant bases.

```{r Figure S2}
source("R_scripts/feature_overlap.R")
# source("R_scripts/proportion.R")

a = ggplot(feature_overlap_long,aes(x=Feature,y=Percent,fill=Cohort)) + geom_bar(position="dodge",stat="identity") + theme(text=element_text(size=15),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1),legend.title = element_blank())  + scale_x_discrete(labels=genic_labels[1:7]) + scale_fill_discrete(labels=c("All","Protein-Coding","Non-coding")) + facet_wrap(~Measure,nrow=1,labeller=labeller(Measure=measure_labels))

b = ggplot(combined_proportion[which(combined_proportion$Cohort != "genome_noTE"),],aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Mean proportion of feature in state") + scale_fill_manual(values=all_state_colors) + theme(axis.text.x = element_text(size=15,angle=90,hjust=1),axis.title.x=element_blank()) + scale_x_discrete(labels=c("TE","Genome",genic_labels[1:7])) + facet_grid(Coding~Group,labeller=labeller(Coding=coding_labels))

grid.arrange(a,b,nrow=2,heights=c(0.4,0.6))
```

## Figure 2

Figure 2. Potential for individual TEs to be in each epigenetic state. (A) Boxplots indicate the proportion of all TEs (n=4,430,788) in the state in each sample. Red dots are the fraction of all TEs in the state in any sample. For WGBS states, TEs are restricted to those with any CpG (n=3,191,684). (B) Average proportion of samples a TE is in a state. Total is >100% for chromHMM due to TEs in multiple states in a single sample. (C) For TEs in a chromHMM state in at least one sample, cumulative proportion of individual TEs in that state in at least that fraction of samples. Horizontal lines indicate the mean fraction of samples a TE is in each state if it is in that state in at least one sample. 

```{r Figure 2}
source("R_scripts/potential.R")

combine_boxplot = rbind(state_sample_count[,c(1:2,4)],TE_meth_average_state_long,TE_DNase_peaks_sample[,2:4],TE_H3K27ac_peaks_sample[,2:4],RNA_RPKM_sample[,2:4])
combine_stats = rbind(chromHMM_TE_state_dist_stats,TE_meth_average_category_stats,TE_DNase_potential_stats,TE_H3K27ac_potential_stats,RNA_potential_stats)
combine_stats$Group = factor(c(rep("chromHMM",15),rep("WGBS",4),"DNase","H3K27ac","Expression"),levels=c("chromHMM","WGBS","DNase","H3K27ac","Expression"))

a = ggplot(combine_boxplot,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="% of TEs in state") + coord_flip() + scale_x_discrete(limits = rev(levels(combine_boxplot$State)),labels=rev(all_state_labels)) + geom_point(data=combine_stats,aes(x=State,y=Proportion_ever),color="red",size=3)

b = ggplot(combine_stats,aes(x=1,y=Samples_avg_all,fill=State)) + geom_bar(stat="identity",show.legend=FALSE) + scale_fill_manual(values=all_state_colors) + theme(axis.title.x=element_blank(),axis.ticks.x=element_blank(),axis.text.x=element_blank()) + ylab("Average % of samples in state") + facet_wrap(~Group,nrow=1,labeller=labeller(Group = group_labels))

c = ggplot(chromHMM_TE_state_cum,aes(x=Sample_proportion,y=Proportion,color=State,group=State)) + geom_point(size=0.1) + geom_line(size=1.5) + labs(x="Proportion of all samples (n=127)",y="Cumulative proportion of TEs in state") + scale_color_manual(values=chromHMM_colors,guide=FALSE) + geom_vline(xintercept = chromHMM_TE_state_dist_stats$Samples_avg_ever/100,color=chromHMM_colors)

legend = get_legend(ggplot(combine_boxplot,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors) + theme(legend.direction = "horizontal",legend.title = element_blank(),legend.position="bottom"))

grid.arrange(a,b,c,legend, nrow = 3, layout_matrix=rbind(c(1,2),c(1,3),c(4,4)),heights=c(0.4,0.4,0.2))
```

## Number of TEs ever in composite states: Roadmap Active (states 1-8) and Inactive (states 9-15) states; functional category: active regulatory states (states 1-3, 6-7), transcribed states (states 4-5), poised regulatory states (states 10-12), repressed states (states 9, 13-14)

```{r}
sum(apply(chromHMM_TE_state[8:15],1,function(x) sum(x > 0)) > 0)
sum(apply(chromHMM_TE_state[16:22],1,function(x) sum(x > 0)) > 0)
sum(apply(chromHMM_TE_state[c(8:10,13:14)],1,function(x) sum(x > 0)) > 0)
sum(apply(chromHMM_TE_state[11:12],1,function(x) sum(x > 0)) > 0)
sum(apply(chromHMM_TE_state[17:19],1,function(x) sum(x > 0)) > 0)
sum(apply(chromHMM_TE_state[c(16,20,21)],1,function(x) sum(x > 0)) > 0)
```

# RNA analysis

```{r}
mean(apply(RNA_TE_agnostic[,8:60],1,function(x) sum(x > 0)))
mean(RNA_TE_agnostic[which(RNA_TE_agnostic$Expressed_samples > 0),]$Expressed_samples)
median(RNA_TE_agnostic$Max_expression)
median(RNA_TE_agnostic[which(RNA_TE_agnostic$Expressed_samples > 0),]$Max_expression)
```

## Supplementary Figure 3

Supplmentary Figure 3. Comparison of chromHMM, WGBS, DNase, and H3K27ac states. (A) Proportion of all TEs x sample in each methylation state (in samples with WGBS data) annotated with each chromHMM state. Sums are >100% due to TEs annotated with multiple chromHMM states. 

```{r Figure S3}
source("R_scripts/compare_marks.R")

test = xtabs(TE_sample_norm ~ chromHMM+WGBS, data=compare_marks_all)
names(dimnames(test)) = c("chromHMM","WGBS")
a = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample_norm ~ WGBS+chromHMM, data=compare_marks_all)
names(dimnames(test)) = c("WGBS","chromHMM")
b = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample_norm ~ DNase+chromHMM, data=compare_marks_all)
names(dimnames(test)) = c("DNase","chromHMM")
c = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample_norm ~ H3K27ac+chromHMM, data=compare_marks_all)
names(dimnames(test)) = c("H3K27ac","chromHMM")
d = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample_norm ~ RNA+chromHMM, data=compare_marks_all)
names(dimnames(test)) = c("RNA","chromHMM")
e = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample_norm ~ chromHMM+DNase, data=compare_marks_all)
names(dimnames(test)) = c("chromHMM","DNase")
f = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ WGBS+DNase, data=compare_marks_unique)
names(dimnames(test)) = c("WGBS","DNase")
g = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ DNase+WGBS, data=compare_marks_unique)
names(dimnames(test)) = c("DNase","WGBS")
h = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ H3K27ac+WGBS, data=compare_marks_unique)
names(dimnames(test)) = c("H3K27ac","WGBS")
i = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ RNA+WGBS, data=compare_marks_unique)
names(dimnames(test)) = c("RNA","WGBS")
j = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample_norm ~ chromHMM+H3K27ac, data=compare_marks_all)
names(dimnames(test)) = c("chromHMM","H3K27ac")
k = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ WGBS+H3K27ac, data=compare_marks_unique)
names(dimnames(test)) = c("WGBS","H3K27ac")
l = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ DNase+H3K27ac, data=compare_marks_unique)
names(dimnames(test)) = c("DNase","H3K27ac")
m = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ H3K27ac+DNase, data=compare_marks_unique)
names(dimnames(test)) = c("H3K27ac","DNase")
n = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ RNA+DNase, data=compare_marks_unique)
names(dimnames(test)) = c("RNA","DNase")
o = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample_norm ~ chromHMM+RNA, data=compare_marks_all)
names(dimnames(test)) = c("chromHMM","RNA")
p = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ WGBS+RNA, data=compare_marks_unique)
names(dimnames(test)) = c("WGBS","RNA")
q = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ DNase+RNA, data=compare_marks_unique)
names(dimnames(test)) = c("DNase","RNA")
r = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ H3K27ac+RNA, data=compare_marks_unique)
names(dimnames(test)) = c("H3K27ac","RNA")
s = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

test = xtabs(TE_sample ~ RNA+H3K27ac, data=compare_marks_unique)
names(dimnames(test)) = c("RNA","H3K27ac")
t = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

mplot(a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t)

test = xtabs(TE_sample_norm ~ chromHMM+WGBS+DNase+H3K27ac+RNA, data=compare_marks_all)
u = mosaic(test,shade=TRUE,return_grob=TRUE,legend=FALSE)

mplot(u)
```

# Compare mark analysis (get p-values and residuals)

```{r}
# Tables for each mark
aggregate(data=compare_marks_unique,TE_sample~DNase,function(x) sum(x)/sum(compare_marks_unique[which(!is.na(compare_marks_unique$DNase)),]$TE_sample))
aggregate(data=compare_marks_unique,TE_sample~H3K27ac,function(x) sum(x)/sum(compare_marks_unique[which(!is.na(compare_marks_unique$H3K27ac)),]$TE_sample))
aggregate(data=compare_marks_unique,TE_sample~RNA,function(x) sum(x)/sum(compare_marks_unique[which(!is.na(compare_marks_unique$RNA)),]$TE_sample))
aggregate(data=compare_marks_unique,TE_sample~WGBS,function(x) sum(x)/sum(compare_marks_unique[which(!is.na(compare_marks_unique$WGBS)),]$TE_sample))
aggregate(data=compare_marks_all,TE_sample_norm~chromHMM,function(x) sum(x)/sum(compare_marks_all$TE_sample_norm))

# Chi-squared for each mark combination
chisq_list = list(chisq.test(xtabs(TE_sample~H3K27ac+DNase,data=aggregate(data=compare_marks_unique,TE_sample~H3K27ac+DNase,sum))),  
                  chisq.test(xtabs(TE_sample~H3K27ac+WGBS,data=aggregate(data=compare_marks_unique,TE_sample~H3K27ac+WGBS,sum))), 
                  chisq.test(xtabs(TE_sample~DNase+WGBS,data=aggregate(data=compare_marks_unique,TE_sample~DNase+WGBS,sum))), 
                  chisq.test(xtabs(TE_sample~DNase+RNA,data=aggregate(data=compare_marks_unique,TE_sample~DNase+RNA,sum))), 
                  chisq.test(xtabs(TE_sample~RNA+WGBS,data=aggregate(data=compare_marks_unique,TE_sample~RNA+WGBS,sum))), 
                  chisq.test(xtabs(TE_sample~RNA+H3K27ac,data=aggregate(data=compare_marks_unique,TE_sample~RNA+H3K27ac,sum))), 
                  chisq.test(xtabs(TE_sample_norm~chromHMM+DNase,data=aggregate(data=compare_marks_all,TE_sample_norm~chromHMM+DNase,sum))), 
                  chisq.test(xtabs(TE_sample_norm~chromHMM+H3K27ac,data=aggregate(data=compare_marks_all,TE_sample_norm~chromHMM+H3K27ac,sum))),
                  chisq.test(xtabs(TE_sample_norm~chromHMM+RNA,data=aggregate(data=compare_marks_all,TE_sample_norm~chromHMM+RNA,sum))),
                  chisq.test(xtabs(TE_sample_norm~chromHMM+WGBS,data=aggregate(data=compare_marks_all,TE_sample_norm~chromHMM+WGBS,sum))))
lapply(chisq_list,function(x) x$p.value)
lapply(chisq_list,function(x) x$residuals)

# Combining all marks
compare_marks_all[which(compare_marks_all$WGBS == "Hypomethylated" & compare_marks_all$DNase == "yes" & compare_marks_all$H3K27ac == "yes"),]

# Discuss proportion of each state in other states/DNase/H3K27ac, esp. active regulatory
```

## Supplementary Figure 4

Supplementary Figure 4. Potential controls. (A-C) Potential metrics with shuffled TEs (n=4,430,788), 10 iterations. Actual TE values are indicated with red dot. (D-E) Potential metrics with unique Refseq promoters. (F) Expression potential metrics with unique Refseq exons. 

```{r Figure S4}
source("R_scripts/potential_shuffled.R")
source("R_scripts/potential_promoter.R")
source("R_scripts/potential_exon.R")

# Shuffled TEs
combine_boxplot$Iteration = rep("True",dim(combine_boxplot)[1])
combine_shuffled_boxplot = rbind(state_sample_count_shuffled[,c(1:2,4:5)],shuffled_WGBS_average_count,shuffled_DNase_potential_count[,2:5],shuffled_H3K27ac_potential_count[,2:5],combine_boxplot[which(combine_boxplot$State != "Expression"),])
combine_shuffled_boxplot$Iteration = factor(combine_shuffled_boxplot$Iteration,levels=c(seq(1,10,1),"True"))

combine_shuffled_stats = rbind(shuffled_chromHMM_potential_stats,shuffled_WGBS_average_stats,shuffled_DNase_potential_stats,shuffled_H3K27ac_potential_stats)
combine_shuffled_stats$Group = factor(c(rep("chromHMM",150),rep("WGBS",40),rep("DNase",10),rep("H3K27ac",10)),levels=c("chromHMM","WGBS","DNase","H3K27ac"))
combine_shuffled_stats = melt(combine_shuffled_stats,id.vars=c("State","Group","Iteration"))
colnames(combine_shuffled_stats)[4:5] = c("Metric","Value")

combine_stats_long = melt(combine_stats,id.vars=c("State","Group"))
colnames(combine_stats_long)[3:4] = c("Metric","Value")

a = ggplot(combine_shuffled_boxplot,aes(x=State,y=Proportion*100,fill=Iteration,color=Iteration)) + geom_boxplot(position="dodge",outlier.size=0.1) + scale_x_discrete(limits=all_state_labels[1:21]) + labs(y="% of TEs in state") + theme(axis.title.x=element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_fill_manual(values=c(rep("white",10),"white"),guide=FALSE) + scale_color_manual(values=c(rep("grey",10),"red"),guide=FALSE)

b = ggplot(combine_shuffled_stats,aes(x=State,y=Value)) + geom_boxplot() + theme(axis.title.y = element_blank(),axis.title.x = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_x_discrete(limits=c(chromHMM_states,meth_states,"DNase","H3K27ac")) + geom_point(data=combine_stats_long[which(combine_stats_long$State != "Expression"),],aes(x=State,y=Value),color="red",size=1) + facet_wrap(~Metric,nrow=3,labeller=labeller(Metric=metric_labels))

# Promoters
combine_cum_prom = rbind(chromHMM_promoter_state_cum,promoter_meth_average_category_cum,promoter_DNase_potential_cum,promoter_H3K27ac_potential_cum)
combine_stats_prom = rbind(chromHMM_promoter_state_dist_stats,promoter_meth_average_category_stats,promoter_DNase_potential_stats,promoter_H3K27ac_potential_stats)
combine_stats_prom$Group = factor(c(rep("chromHMM",15),rep("WGBS",4),"DNase","H3K27ac"),levels=c("chromHMM","WGBS","DNase","H3K27ac"))
combine_boxplot_prom = rbind(promoter_state_sample_count[,c(1:2,4)],promoter_meth_average_state_long,promoter_DNase_peaks_sample[,2:4],promoter_H3K27ac_peaks_sample[,2:4])

c = ggplot(combine_boxplot_prom,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors[1:21],guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="% of promoters in state") + coord_flip() + scale_x_discrete(limits = rev(levels(combine_boxplot_prom$State))) + geom_point(data=combine_stats_prom,aes(x=State,y=Proportion_ever),color="red",size=2)

d = ggplot(combine_cum_prom,aes(x=Sample_proportion,y=Proportion,color=State,group=State)) + geom_point(size=0.1) + geom_line(size=1.5) + labs(x="Proportion of all samples",y="Cumulative proportion of promoters in state") + scale_color_manual(values=all_state_colors[1:21],guide=FALSE) + geom_vline(xintercept = combine_stats_prom$Samples_avg_ever/100,color=all_state_colors[as.vector(combine_stats_prom$State)])

# Exons
e = ggplot(RNA_exon_sample,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values="black",guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="% of exons in state") + coord_flip() + scale_x_discrete(labels="RPKM > 1") + geom_point(data=RNA_potential_exon_stats,aes(x=State,y=Proportion_ever),color="red",size=3) + scale_y_continuous(limits=c(0,100))

f = ggplot(RNA_potential_exon_cum,aes(x=Sample_proportion,y=Proportion,color=State,group=State)) + geom_point(size=0.1) + geom_line(size=1.5) + labs(x="Proportion of all samples (n=52)",y="Cumulative proportion of exons RPKM > 1") + scale_color_manual(values="black",guide=FALSE) + geom_vline(xintercept = RNA_potential_exon_stats$Samples_avg_ever/100,color="black")

legend = get_legend(ggplot(combine_boxplot,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors) + theme(legend.direction = "horizontal",legend.title = element_blank(),legend.position="bottom"))

grid.arrange(a,b,c,d,e,f,legend, nrow = 5, ncol=2, layout_matrix=cbind(c(1,3,3,4,4),c(2,5,6,6,7)),heights=c(0.5,0.1,0.15,0.1,0.15))
```

# Exon expression control analysis

```{r }
mean(apply(RNA_exon_unique,1,function(x) sum(as.numeric(x[6:56]) > 0)))

#hist(RNA_exon_unique$Expressed_samples)

median(RNA_exon_unique$Max_expression)
median(RNA_exon_unique[which(RNA_exon_unique$Expressed_samples > 0),]$Max_expression)
```

## Supplementary Figure 5

Supplementary Figure 5. Additional potential figures. (A) For TEs in a methylation state, overlapping DNase or H3K27ac peaks, or expressed RPKM >1 in at least one sample, cumulative proportion of individual TEs in that state in at least that proportion of samples. Horizontal lines indicate the mean proportion of samples a TE is in each state if it is in that state in at least one sample. (B) Proportion of TEs with CpGs (n=3,191,684) in each methylation state by sample. Ordered by decreasing proportion of TEs intermediately methylated or hypomethylated. (C) Boxplots indicate the proportion of all TEs in the state per sample, excluding five cancer cell lines and IMR90 (chromHMM n=121, WGBS n=36, DNase n=48, H3K27ac n=92, RNA n=48).  Red dots are the fraction of all TEs in that state in any sample, excluding those six samples. (D) Average proportion of samples a TE is in a state, excluding cancer cell lines and IMR90. Total is >100% for chromHMM due to TEs in multiple states in a single sample.

```{r Figure S5}
# source("R_scripts/potential.R")

combine_cum = rbind(TE_meth_average_category_cum,TE_DNase_potential_cum,TE_H3K27ac_potential_cum,RNA_potential_cum)
combine_boxplot_noCancer_IMR90 = droplevels(combine_boxplot[which(metadata[match(combine_boxplot$Sample,metadata$Sample),]$Exclude == "Include"),])
combine_stats_noCancer_IMR90 = rbind(chromHMM_TE_state_dist_noCancer_stats,TE_meth_average_noIMR90_category_stats,TE_DNase_potential_noCancer_stats,TE_H3K27ac_potential_noCancer_stats,RNA_potential_noCancer_stats)
combine_stats_noCancer_IMR90$Group = factor(c(rep("chromHMM",15),rep("WGBS",4),"DNase","H3K27ac","Expression"),levels=c("chromHMM","WGBS","DNase","H3K27ac","Expression"))

a = ggplot(TE_meth_average_state_long,aes(x=Sample,y=Proportion,fill=State)) + geom_bar(stat="identity") + theme(axis.text.x=element_text(angle=90),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.title.x=element_blank()) + scale_fill_manual(values=meth_colors,guide=FALSE) + ylab("Proportion of TEs in methylation state")

b = ggplot(combine_cum,aes(x=Sample_proportion,y=Proportion,color=State,group=State)) + geom_point(size=0.1) + geom_line(size=1.5) + labs(x="Proportion of samples",y="Cumulative proportion of TEs in state") + scale_color_manual(values=all_state_colors[16:22],guide=FALSE) + geom_vline(xintercept = combine_stats$Samples_avg_ever[16:22]/100,color=all_state_colors[as.vector(combine_stats$State)[16:22]]) 

c = ggplot(combine_boxplot_noCancer_IMR90,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="% of TEs in state") + coord_flip() + scale_x_discrete(limits = rev(levels(combine_boxplot_noCancer_IMR90$State)),labels=rev(all_state_labels)) + geom_point(data=combine_stats_noCancer_IMR90,aes(x=State,y=Proportion_ever),color="red",size=3)

d = ggplot(combine_stats_noCancer_IMR90,aes(x=1,y=Samples_avg_all,fill=State)) + geom_bar(stat="identity",show.legend=FALSE) + scale_fill_manual(values=all_state_colors) + theme(axis.title.x=element_blank(),axis.ticks.x=element_blank(),axis.text.x=element_blank()) + ylab("Average % of samples in state") + facet_wrap(~Group,nrow=1,labeller=labeller(Group = group_labels))

legend = get_legend(ggplot(combine_boxplot,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors) + theme(legend.direction = "horizontal",legend.title = element_blank(),legend.position="bottom"))

grid.arrange(a,b,c,d,legend, nrow = 3,layout_matrix=rbind(c(1,2),c(3,4),c(5,5)),heights=c(0.4,0.4,0.2))
```

# Average TE methylation by sample

```{r}
# Average TE methylation level by sample
TE_meth_average_mean = apply(TE_meth_average[,c(8:44)],2,function(x) mean(na.omit(x)))

# Summaries with/without IMR90
mean(TE_meth_average_mean)
sd(TE_meth_average_mean)
mean(TE_meth_average_mean[c(1:10,12:37)])
sd(TE_meth_average_mean[c(1:10,12:37)])
```

# Average proportion of TEs hypomethylated per sample, no IMR90

```{r}
mean(TE_meth_average_state$Hypomethylated[2:37])
sd(TE_meth_average_state$Hypomethylated[2:37])
```

## Supplementary Figure 6

Supplementary Figure 6. State switching for chromHMM states. (A) Total number of chromHMM states per TE across all samples (blue) and maximum number of chromHMM states per TE in one sample (red). (B) Proportion of instances where two chromHMM states are co-annotated on the same TE in the same sample (n=4,430,788 x 127). Rows normalized by total number of instances of TEs in that state in one sample (diagonal). Color is the proportion of total instances in the first state (rows) co-annotated with the second state (columns). (C) Proportion of TEs (n=4,430,788) in two chromHMM states in different samples. Rows normalized by number of TEs in that state in at least one sample. Color is the proportion of TEs in one state (rows) that are in the second state (columns) in a different sample.

```{r Figure S6}
#source("R_scripts/potential.R")
source("R_scripts/state_switching.R")

# chromHMM states
a = ggplot(chromHMM_TE_state,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + geom_histogram(data = chromHMM_TE_state,aes(x=Max_states_intra),binwidth=1,fill="darkred",color="grey",alpha=0.5) + ylab("Number of TEs") + xlab("Number of chromHMM states")

# chromHMM states by class 
b = ggplot(chromHMM_TE_state,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + scale_x_continuous(limits=c(0,15)) + geom_histogram(data = chromHMM_TE_state,aes(x=Max_states_intra),binwidth=1,fill="darkred",color="grey",alpha=0.5) + ylab("Number of TEs") + xlab("Number of chromHMM states") + facet_wrap(~class_update,scales="free")

# WGBS states
TE_meth_average$States = apply(TE_meth_average[,46:49],1,function(x) sum(x > 0))
TE_meth_average$States_noIMR90 = apply(TE_meth_average[,50:53],1,function(x) sum(x > 0))

c = ggplot(TE_meth_average,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + ylab("Number of TEs") + xlab("Number of methylation states")

# WGBS states by class
d = ggplot(TE_meth_average,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + scale_x_continuous(limits=c(0,4)) + ylab("Number of TEs") + xlab("Number of methylation states") + facet_wrap(~class_update,scales="free")

# Matrices
e = ggplot(melt(as.matrix(state_switching_intra)),aes(x=Var2,y=Var1,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(labels=rev(chromHMM_states),limits=rev(levels(melt(as.matrix(state_switching_intra))$Var1))) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1),guide=FALSE)

f = ggplot(melt(as.matrix(state_switching_inter)),aes(x=Var2,y=Var1,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(labels=rev(chromHMM_states),limits=rev(levels(melt(as.matrix(state_switching_inter))$Var1))) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1),guide=FALSE)

g = ggplot(melt(ss_inter_class,id.vars=c("Class","State")),aes(x=variable,y=State,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(limits=rev(chromHMM_states)) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1),guide=FALSE) + facet_wrap(~Class)

h = ggplot(melt(as.matrix(ss_inter_meth)),aes(x=Var2,y=Var1,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_y_discrete(limits=rev(levels(melt(as.matrix(ss_inter_meth))$Var1))) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1))

i = ggplot(melt(ss_inter_meth_class,id.vars=c("Class","State")),aes(x=variable,y=State,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_y_discrete(limits=rev(meth_states)) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1),guide=FALSE) + facet_wrap(~Class)

grid.arrange(a,b,c,d,e,f,g,h,i, nrow=4, layout_matrix=rbind(c(1,2,2),c(3,4,4),c(5,6,8),c(7,7,9)),heights=c(0.2,0.2,0.3,0.3))
```

# Shuffled number of states (not shown)

```{r}
#source("R_scripts/potential_shuffled.R")

# Shuffled chromHMM
shuffled_chromHMM_states = ldply(shuffled_chromHMM_potential,function(x) x[,23:24])
shuffled_chromHMM_states$Iteration = factor(rep(seq(1,10,1),each=4430788))

ggplot(shuffled_chromHMM_states,aes(x=States,fill=Iteration)) + geom_histogram(binwidth=1,alpha=0.5,position="dodge") + ylab("Number of TEs") + xlab("Number of chromHMM states")
ggplot(shuffled_chromHMM_states,aes(x=Max_states_intra,fill=Iteration)) + geom_histogram(binwidth=1,alpha=0.5,position="dodge") + ylab("Number of TEs") + xlab("Number of chromHMM states")

# Shuffled WGBS
shuffled_WGBS_average = lapply(shuffled_WGBS_average,function(x) transform(x,States = apply(x[,46:49],1,function(y) sum(y > 0))))
shuffled_WGBS_average = lapply(shuffled_WGBS_average,function(x) transform(x,States_noIMR90 = apply(x[,50:53],1,function(y) sum(y > 0))))
shuffled_WGBS_states = ldply(shuffled_WGBS_average,function(x) x[,54:55])
shuffled_WGBS_states$Iteration = factor(unlist(lapply(seq(1,10,1),function(y) rep(y,unlist(lapply(shuffled_WGBS_average,function(x) dim(x)[1]))[y]))))

ggplot(shuffled_WGBS_states,aes(x=States,fill=Iteration)) + geom_histogram(binwidth=1,alpha=0.5,position="dodge") + ylab("Number of shuffled TEs") + xlab("Number of methylation states")
```

## Correlation of TE length with number of states ever, max in one sample

```{r}
cor.test((chromHMM_TE_state$stop-chromHMM_TE_state$start),chromHMM_TE_state$States)
cor.test(chromHMM_TE_state$stop-chromHMM_TE_state$start,chromHMM_TE_state$Max_states_intra)
```

# Number of chromHMM and WGBS states per TE, by class

```{r}
aggregate(data=chromHMM_TE_state,States~class_update,mean)
aggregate(data=chromHMM_TE_state,States~class_update,sd)
aggregate(data=TE_meth_average,States~class_update,mean)
aggregate(data=TE_meth_average,States~class_update,sd)
```

# TEs always in the 5_TxWk state

```{r}
chromHMM_TE_state[which(chromHMM_TE_state$X5_TxWk == chromHMM_TE_state$Tissues),]
```

# Correlation of number of samples in state with other states

```{r}
TE_state_correlation = ldply(colnames(rmsk_TE_measure)[35:63],function(x) correlate_spearman(rmsk_TE_measure,x,c(35:63)))
TE_state_correlation$State2 = rep(colnames(rmsk_TE_measure)[35:63],each=539)
TE_state_correlation = TE_state_correlation[which(TE_state_correlation$State != TE_state_correlation$State2),]
TE_state_correlation$p.value = as.numeric(as.character(TE_state_correlation$p.value))
TE_state_correlation$estimate.rho = as.numeric(as.character(TE_state_correlation$estimate.rho))

ggplot(TE_state_correlation,aes(x=State1,y=State2,fill=estimate.rho)) + geom_tile() + facet_wrap(~class_update) + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1))
```

## Figure 3

Figure 3. Correlation of TE activity with phylogeny and TE features. (A) Total proportion of each TE class in each state across all samples. (B) Average proportion of samples a TE is in a state, by class. Ordered by decreasing range. Coefficient of variation in parentheses. (C) Difference in proportion of samples a TE is in a state by overlap with RefSeq genic features, normalized by overall average. (D) TE subfamily age (Jukes-Cantor evolutionary distance) vs. average number of CpGs per kbp. (E) TE subfamily age (Jukes-Cantor evolutionary distance) vs. range of proportion of TEs hypomethylated per sample (n=37). 

```{r Figure 3}
source("R_scripts/proportion_class.R")
source("R_scripts/potential_class.R")
source("R_scripts/feature_overlap_state.R")
source("R_scripts/state_metric_correlation.R")

# Proportion in state by class
combine_class_proportion = rbind(contribution_class[,c(1:2,4)],class_CpG_meth_total,TE_DNase_class_total,TE_H3K27ac_class_total)
combine_class_proportion$Cohort = factor(c(rep("chromHMM",90),rep("WGBS",24),rep("DNase",6),rep("H3K27ac",6)),levels=c("chromHMM","WGBS","DNase","H3K27ac"))

a = ggplot(combine_class_proportion,aes(x=class,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Total proportion of class in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x = element_blank()) + facet_wrap(~Cohort,nrow=2,scales="free")

# Class potential
combine_stats_class = rbind(chromHMM_TE_state_class_stats,TE_meth_average_class_stats,potential_TE_DNase_class_stats,potential_TE_H3K27ac_class_stats,RNA_potential_class_stats)
combine_stats_class$Group = factor(c(rep("chromHMM",90),rep("WGBS",24),rep("DNase",6),rep("H3K27ac",6),rep("Expression",6)),levels=c("chromHMM","WGBS","DNase","H3K27ac","Expression"))
state_order = ddply(combine_stats_class,"State",summarise,CV=sd(Samples_avg_all)/mean(Samples_avg_all), Range=max(Samples_avg_all)-min(Samples_avg_all))

b = ggplot(combine_stats_class,aes(x=State,y=Samples_avg_all)) + geom_point(aes(color=class_update),size=4) + scale_color_manual(values=class_colors,name="Class") + scale_x_discrete(limits=state_order[order(state_order$Range),]$State,labels=all_state_labels) + scale_y_continuous(limits=c(0,100)) + theme(axis.title.y = element_blank()) + ylab("Average % samples in state") + geom_text(data=state_order,aes(x=State,y=100,label=paste("(",round(CV,2),")",sep=""))) + coord_flip() 

# Feature table
c = ggplot(feature_state_mean[which(feature_state_mean$Coding == "All"),],aes(x=State,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1)) + scale_fill_gradient2(low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-5,30)) + scale_y_discrete(limits=rev(levels(feature_state_mean$Feature)),labels=genic_labels) + scale_x_discrete(labels=all_state_labels)

# SINE/Alu age vs. TE metrics and number of samples in methylation states
rmsk_TE_measure$cpgIsland_binary = as.vector(rmsk_TE_measure$cpgIsland)
rmsk_TE_measure$cpgIsland_binary = gsub("no",0,rmsk_TE_measure$cpgIsland_binary)
rmsk_TE_measure$cpgIsland_binary = gsub("yes",1,rmsk_TE_measure$cpgIsland_binary)
rmsk_TE_measure$cpgIsland_binary = as.numeric(rmsk_TE_measure$cpgIsland_binary)

rmsk_TE_measure_SINE = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$class_update == "SINE"),c(1:10,12:14,52:55,64)])
rmsk_TE_measure_SINE$Length = rmsk_TE_measure_SINE$Length/1000

d = ggplot(rmsk_TE_measure_SINE,aes(x=JC_distance)) + geom_density(fill=class_colors["SINE"]) + theme(axis.title.x = element_blank(),axis.title.y = element_blank())

e = ggplot(melt(rmsk_TE_measure_SINE,id.vars=colnames(rmsk_TE_measure_SINE)[c(1:11,13,18)]),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=setNames(c("#000000",meth_colors),c("CpGs",names(meth_colors))),guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank()) 

f = ggplot(melt(rmsk_TE_measure_SINE,id.vars=colnames(rmsk_TE_measure_SINE)[c(1:7,9,11:12,14:18)])) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + geom_smooth(data=rmsk_TE_measure_SINE,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + theme(axis.title.y = element_blank())

rmsk_TE_measure_Alu = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$family == "Alu"),c(1:10,12:14,52:55,64)])
rmsk_TE_measure_Alu$Length = rmsk_TE_measure_Alu$Length/1000

g = ggplot(rmsk_TE_measure_Alu,aes(x=JC_distance)) + geom_density(fill="lightblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank())

h = ggplot(melt(rmsk_TE_measure_Alu,id.vars=colnames(rmsk_TE_measure_Alu)[c(1:11,13,18)]),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=setNames(c("#000000",meth_colors),c("CpGs",names(meth_colors))),guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank()) 

i = ggplot(melt(rmsk_TE_measure_Alu,id.vars=colnames(rmsk_TE_measure_Alu)[c(1:7,9,11:12,14:18)])) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + theme(axis.title.y = element_blank())

legend1 = get_legend(ggplot(combine_class_proportion,aes(x=class,y=Proportion,fill=State)) + geom_bar(stat="identity") + scale_fill_manual(values=all_state_colors) + theme(legend.direction = "horizontal",legend.title = element_blank()))

legend2 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=colnames(rmsk_TE_measure_Alu)[c(1:11,13,18)]),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=setNames(c("#000000",meth_colors),c("CpGs",names(meth_colors))),labels=c("CpGs/TE",meth_states[c(1,3,2,4)])) + theme(legend.title = element_blank(),legend.direction = "horizontal"))
  
legend3 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=colnames(rmsk_TE_measure_Alu)[c(1:7,9,11:12,14:18)])) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),labels=c("Length (kbp)","Mappability","CpGs/bp")) + ylab("Property") + theme(legend.title = element_blank(),legend.direction = "horizontal") + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")))

grid.arrange(a,c,b,d,e,f,g,h,i,legend1,legend2,legend3,nrow=6,layout_matrix=rbind(c(1,1,2,2),c(10,10,4,7),c(3,3,5,8),c(3,3,6,9),c(3,3,11,11),c(3,3,12,12)),heights=c(0.5,0.1,0.2,0.2,0.05,0.05))
```

# Number of samples in state by class, p-values

```{r}
apply(rmsk_TE_measure[,35:63],2,function(x) unlist(kruskal.test(x~rmsk_TE_measure$class_update))["p.value"])
```

# Number of samples in state by genic feature, p-values

```{r}
TE_pvalue_all_state = apply(rmsk_TE_measure[,15:34],2,function(x) apply(rmsk_TE_measure[,35:63],2,function(y) unlist(wilcox.test(y~x))["p.value"]))
TE_pvalue_class_state = ddply(rmsk_TE_measure,~class_update,function(z) apply(z[,15:34],2,function(x) apply(z[,35:63],2,function(y) unlist(wilcox.test(y~x))["p.value"])))
```

# Overlap with CpG islands by age (logistic regression)

```{r}
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure,family="binomial")
```

# Comparison of CpG % hypomethylated by class

```{r}
kruskal.test(class_CpG_meth$Hypomethylated~class_CpG_meth$class)
pairwise.wilcox.test(class_CpG_meth$Hypomethylated,class_CpG_meth$class,p.adj="bonf")
#kruskal.test(class_CpG_meth[which(class_CpG_meth$Sample != "E017"),]$Hypomethylated~class_CpG_meth[which(class_CpG_meth$Sample != "E017"),]$class)
#pairwise.wilcox.test(class_CpG_meth[which(class_CpG_meth$Sample != "E017"),]$Hypomethylated,class_CpG_meth[which(class_CpG_meth$Sample != "E017"),]$class,p.adj="bonf")
```

## Supplementary Figure 7

Supplementary Figure 7. Potential for an individual TE to be in an epigenetic state, by class. (A) Proportion of TEs in each state in any sample, by class (DNA n = 456,948, LINE n = 1,480,369, LTR n = 708,210, SINE n = 1,769,839, SVA n = 3,608, Other n = 11,814; TEs with CpGs only; DNA n = 273,870, LINE n = 948,864, LTR n = 531,389, SINE n = 1,427,517, SVA n = 3,519, Other n = 6,525). (B) Average proportion of samples in which a TE is in each state, by class. (C) For TEs in a state in at least one sample, average proportion of samples the TE is in that state, by class.

```{r Figure S7}
#source("R_scripts/potential_class.R")

combine_boxplot_class = rbind(class_state_sample[,c(1:3,5)],TE_meth_average_state_class,TE_DNase_peaks_class,TE_H3K27ac_peaks_class,RNA_RPKM_class)

a = ggplot(combine_boxplot_class,aes(x=State,y=Proportion*100,fill=Class)) + geom_boxplot() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y = element_blank()) + coord_flip() + labs(y="% of TEs in state by sample") + scale_x_discrete(labels=rev(all_state_labels),limits=rev(levels(combine_boxplot_class$State)))

b = ggplot(melt(combine_stats_class),aes(x=State,y=value,fill=class_update)) + geom_bar(stat="identity",position="dodge") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1),axis.title.y=element_blank()) + scale_x_discrete(labels=all_state_labels) + scale_y_continuous(limits=c(0,100)) + facet_wrap(~variable,nrow=3,labeller=labeller(variable=metric_labels))

#c = ggplot(combine_stats_class,aes(x=class_update,y=Samples_avg_all,fill=State)) + geom_bar(stat="identity",show.legend=FALSE) + scale_fill_manual(values=all_state_colors) + theme(axis.title.x=element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + ylab("Average % of samples in state") + facet_wrap(~Group,nrow=1,scales="free",labeller=labeller(Group = group_labels))

legend1 = get_legend(ggplot(combine_stats_class,aes(x=State,y=Samples_avg_ever,fill=class_update)) + geom_bar(stat="identity",position="dodge") + scale_fill_manual(values=class_colors,name="Class") + theme(legend.position="bottom",legend.direction="horizontal"))

#legend2 = get_legend(ggplot(combine_stats_class,aes(x=class_update,y=Samples_avg_all,fill=State)) + geom_bar(stat="identity") + scale_fill_manual(values=all_state_colors) + theme(legend.title=element_blank(),legend.position="bottom",legend.direction = "horizontal"))

grid.arrange(a,b,legend1,nrow=2,layout_matrix=rbind(c(1,2),c(3,3)),heights=c(0.9,0.1))
```

# Max expression by class

```{r}
merge(aggregate(data=RNA_TE_agnostic,Expressed_samples~class_update,function(x) sum(x > 0)/length(x)),aggregate(data=RNA_TE_agnostic,Max_expression~class_update,median),by=c("class_update"))
aggregate(data=RNA_TE_agnostic[which(RNA_TE_agnostic$Expressed_samples > 0),],Max_expression~class_update,median)
```

## Supplementary Figure 8

Supplementary Figure 8. Epigenetic state by feature overlap. (A) Proportion of TEs in each class overlapping RefSeq features. TEs may overlap multiple features. (B) Difference in proportion of samples a TE is in a state by overlap with RefSeq genic features, normalized by overall average, including features split by protein-coding/non-coding. (C) By class. 

```{r Figure S8}
#source("R_scripts/feature_overlap_state.R")

rmsk_TE_class_feature = melt(rmsk_TE_class[,c(1,12:30)],id.vars="class_update")
colnames(rmsk_TE_class_feature) = c("Class","Cohort","Proportion")
rmsk_TE_class_feature = split_coding(rmsk_TE_class_feature,2)

a = ggplot(rmsk_TE_class_feature,aes(x=Class,y=Proportion,fill=Feature)) + geom_bar(stat="identity") + theme(axis.title.x = element_blank(),legend.title=element_blank()) +  ylab("Proportion of TEs overlapping feature") + scale_fill_manual(values=c(brewer.pal(8,"Spectral")),labels=genic_labels[c(8,1:7)]) + facet_wrap(~Coding,nrow=1,labeller=labeller(Coding=coding_labels))

b = ggplot(feature_state_mean,aes(x=State,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-5,40)) + scale_y_discrete(limits=rev(levels(feature_state_mean$Feature)),labels=genic_labels) + scale_x_discrete(labels=all_state_labels) + facet_wrap(~Coding,labeller=labeller(Coding=coding_labels))

c = ggplot(feature_state_mean_class[which(feature_state_mean_class$Coding == "All"),],aes(x=State,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-5,120)) + scale_y_discrete(limits=rev(levels(feature_state_mean_class$Feature)),labels=genic_labels) + scale_x_discrete(labels=all_state_labels) + facet_wrap(~Class)

grid.arrange(a,b,c,nrow=3,heights=c(0.25,0.3,0.45))
```

# TE stats

```{r}
# Any TEs not in a feature?
rmsk_TE$Feature_count = apply(rmsk_TE[,13:19],1,function(x) 7-sum(is.na(x)))

# Proportion of TEs in feature
apply(rmsk_TE[,13:20],2,function(x) length(na.omit(x))/4430788)

# Class
# Enrichment of TE classes overlapping features (LOR)
t(apply(rmsk_TEother_feature_class,1,function(x) log2(as.numeric(x)/as.numeric(rmsk_TEother_feature_class[8,]))))

# Subfamily 
# Subfamilies with much higher proportion in feature
apply(rmsk_TEother_feature_subfamily[,6:12],2,function(x) rmsk_TEother_feature_subfamily[which(x > mean(x)+2*sd(x) & rmsk_TEother_feature_subfamily$Count*x > 2),])

# Subfamilies enriched in 3'UTR and hypomethylated
sort(table(droplevels(rmsk_TE_age[which(!(is.na(rmsk_TE_age$'3UTR')) & rmsk_TE_age$Hypomethylated > 10),]$subfamily)))
```

## Supplementary Figure 9

Supplementary Figure 9. TE metrics by class. (A) Proportion of all TEs (left), TE subfamilies (middle), and length (bp; right) in each TE class. Blue slice represents the length of the non-TE genome. (B) Density plot of mappability per TE, by class. (D) Density plot of TE age (Jukes-Cantor evolutionary distance from consensus) per TE, by class. (G) CpGs per TE and proportion of TEs with CpGs by class. 

```{r Figure S9}
source("R_scripts/TE_class_stats.R")
#source("R_scripts/TE_correlation.R")

par(mfrow=c(1,3)) 

pie(rmsk_TE_class$Count,labels=paste(rmsk_TE_class$Count,"(",round(rmsk_TE_class$Count/sum(rmsk_TE_class$Count)*100,0),"%) ",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="TEs",cex=1,cex.main=2,line=-20,radius=0.5)

pie(rmsk_TE_class$Subfamilies,labels=paste(rmsk_TE_class$Subfamilies,"(",round(rmsk_TE_class$Subfamilies/sum(rmsk_TE_class$Subfamilies)*100,0),"%) ",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="Subfamilies",cex=1,cex.main=2,line=-20,radius=0.5)
 
pie(c(3095693983-1389947349,rmsk_TE_class$Total_length),labels=paste(round(c(3095693983-1389947349,rmsk_TE_class$Total_length)/3095693983*100,0),"%",sep=""),col=c("lightblue",class_colors[as.vector(rmsk_TE_class$class_update)]),clockwise=TRUE,cex=1,cex.main=2,main="Length (bp)",line=-20,radius=0.5)

rmsk_TE_measure_metric = melt(rmsk_TE_measure[,c(8:10,12:14)],id.vars="class_update")
rmsk_TE_measure_metric$variable = factor(rmsk_TE_measure_metric$variable,levels(rmsk_TE_measure_metric$variable)[c(3,2,1,4,5)])

b = ggplot(rmsk_TE_measure,aes(x=JC_distance,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("Jukes-Cantor evolutionary distance") + ylab("Density") + scale_x_continuous(limits=c(0,1)) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

c = ggplot(rmsk_TE_measure,aes(x=mappability,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("Mappability (36bp)") + ylab("Density") + scale_x_continuous(limits=c(0,1)) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

d = ggplot(rmsk_TE_measure,aes(x=Length,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("Length (bp)") + ylab("Density") + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

e = ggplot(rmsk_TE_measure,aes(x=CpGs,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("CpGs") + ylab("Density") + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

f = ggplot(rmsk_TE_measure,aes(x=CpGs_per_length,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("CpGs/bp") + ylab("Density") + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

grid.arrange(b,c,d,e,f,nrow=5)
```

# CpGs per class

```{r}
rmsk_TE_class[,c(1:2,12,32,35:38)]
```

# TE stats

```{r}
# Median TE lengths
median(rmsk_TE$Length)

# Mappability whole genome, TEs (see 8/27/17)

# TE metrics by class, with p-values
apply(rmsk_TE_measure[,c(8,10,12:14)],2,function(x) unlist(kruskal.test(x~rmsk_TE_measure$class_update))["p.value"])
aggregate(data=rmsk_TE_measure[,c(8:10,12:14)],.~class_update,mean)
sort(apply(aggregate(data=rmsk_TE_measure[,c(8:10,12:14)],.~class_update,mean)[,2:6],2,function(x) sd(x)/mean(x)))

# Correlation between mappability, % of samples Quiescent
cor.test(rmsk_TE_map36$mappability,rmsk_TE_map36$X15_Quies)

# Mappability of TEs always Quiescent vs not
test = potential_TE_state[which(apply(potential_TE_state[,8:21],1,function(x) sum(x)) == 0),]
test = merge(test[,1:7],rmsk_TE_map36[,1:8])
t.test(test$mappability,rmsk_TE_map36$mappability)

# Average TE, Alu, SVA age
mean(rmsk_TEother_age$JC_distance)
mean(rmsk_TEother_age[which(rmsk_TEother_age$family == "Alu"),]$JC_distance)
mean(rmsk_TEother_age[which(rmsk_TEother_age$family == "Other"),]$JC_distance)

# Correlation between number of CpGs, % of samples Missing
(correlate_TE_state) - all TEs, all classes except SVA; SVA - Length, mappability, CpGs_per_length
```

# Supplementary Figure 10

Supplementary Figure 10. Correlations between TE metrics and number of samples in state, by class. 

```{r Figure S10}
#source("R_scripts/state_metric_correlation.R")
source("R_scripts/feature_overlap_metric.R")

a = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update == "All" & correlate_TE_state$State %in% c("Hypomethylated","Intermediate","Hypermethylated","Missing")),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(limits=c("Missing","Hypermethylated","Intermediate","Hypomethylated")) + scale_x_discrete(limits=c("Length","CpGs","CpGs_per_length","JC_distance","mappability"),labels=c("Length","CpGs","CpGs/bp","Age","Mappability"))

b = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update != "All" & correlate_TE_state$State %in% c("Hypomethylated","Intermediate","Hypermethylated","Missing")),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(limits=c("Missing","Hypermethylated","Intermediate","Hypomethylated")) + scale_x_discrete(limits=c("Length","CpGs","CpGs_per_length","JC_distance","mappability"),labels=c("Length","CpGs","CpGs/bp","Age","Mappability")) + facet_wrap(~class_update) 

c = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs"))

d = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs")) + facet_wrap(~class_update)

e = ggplot(feature_metric_mean[which(feature_metric_mean$Coding == "All"),],aes(x=Metric,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(low="darkblue",high="darkred",mid="white",midpoint=0) + scale_y_discrete(limits=rev(levels(feature_metric_mean$Feature)),labels=genic_labels) + scale_x_discrete(labels=c("Length (bp)","Mappability","Age","CpGs","CpGs/bp"))

f = ggplot(feature_metric_mean_class[which(feature_metric_mean_class$Coding == "All"),],aes(x=Metric,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(low="darkblue",high="darkred",mid="white",midpoint=0) + scale_y_discrete(limits=rev(levels(feature_metric_mean_class$Feature)),labels=genic_labels) + scale_x_discrete(labels=c("Length (bp)","Mappability","Age","CpGs","CpGs/bp")) + facet_wrap(~Class)

grid.arrange(a,b,c,d,e,f,nrow=3)
```

# TE metrics by genic features, p-values

```{r}
TE_pvalue_all_feature = apply(rmsk_TE_measure[,15:34],2,function(x) apply(rmsk_TE_measure[,c(8,10,12:14)],2,function(y) unlist(wilcox.test(y~x))["p.value"]))
TE_pvalue_class_feature = ddply(rmsk_TE_measure,~class_update,function(z) apply(z[,15:34],2,function(x) apply(z[,c(8,10,12:14)],2,function(y) unlist(wilcox.test(y~x))["p.value"])))
```

## Figure 4

Figure 4. Variation in TE activity by sample classification. (A) Proportion of each state in TEs by sample, colored by Roadmap groups. Black line represents median across all samples for that state. For WGBS, proportion is CpGs instead of bases. (B) Average CpG methylation level by sample for all CpGs, CpGs in TEs, and CpGs not in TEs. (C) Principal coordinate analysis on TEs in the 6_EnhG or 7_Enh enhancer states in each sample. 

```{r Figure 4}
source("R_scripts/proportion_by_sample.R")

by_sample_all = rbind(mnemonics_states_TE_proportion,WGBS_proportion,DNase_proportion,H3K27ac_proportion)
by_sample_all = merge(by_sample_all,metadata[,c(1,4:9)],by="Sample")
by_sample_all$Proportion = as.numeric(by_sample_all$Proportion)

a = ggplot(by_sample_all,aes(x=State,y=Proportion,color=Group)) + geom_point(position = position_jitter(width = 0.3)) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylim(0,1) + ylab("Proportion of state in TEs") + scale_color_manual(values=group_colors,guide=FALSE) + scale_x_discrete(limits=levels(by_sample_all$State)) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black",width = 0.8)
 
source("~/TE_landscape/R_scripts/WGBS_sample_CpG_average.R")

b = ggplot(CpG_meth_average,aes(x=Sample,y=Methylation,group=Cohort,color=Cohort)) + geom_point(size=3) + geom_line() + theme(axis.text.x=element_text(angle=90),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.title.x=element_blank(),legend.position = c(0.8, 0.2),legend.title=element_blank(),panel.grid.major.x = element_line(color="black")) + ylab("Average methylation") + scale_color_manual(values=c("black","blue","red"),labels=c("All CpGs","Not in TEs","In TEs")) 

source("R_scripts/enhancer_PCA.R")

c = plot_pca(enhancer_matrix_pca,c(1,2),metadata,"Group","Roadmap Group",group_colors,FALSE)

legend = get_legend(ggplot(by_sample_all,aes(x=State,y=Proportion,color=Group)) + geom_point(position = position_jitter(width = 0.3)) + theme(text=element_text(size=15),panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylim(0,1) + ylab("Proportion of state in TEs") + scale_color_manual(values=group_colors) + scale_x_discrete(limits=levels(by_sample_all$State)) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black",width = 0.4) + guides(color = guide_legend(override.aes = list(size=6.5))))

grid.arrange(a,b,c,legend,nrow=2,layout_matrix=rbind(c(1,1,4),c(2,3,4)),widths=c(0.4,0.4,0.2))

```

## By-sample analysis

```{r}
# Stats
by_sample_stats = ddply(by_sample_all,~State,summarise, Range = max(Proportion)-min(Proportion), Median = median(Proportion), Mean = mean(Proportion), CV = sd(Proportion)/mean(Proportion))

# Kruskal-Wallis test by sample classification
by_sample_KW = ddply(by_sample_all,~State,summarise,Group = p.adjust(unlist(kruskal.test(Proportion,Group))["p.value"],method="bonf"),
      Anatomy = p.adjust(unlist(kruskal.test(Proportion,Anatomy))["p.value"],method="bonf"),
      Type = p.adjust(unlist(kruskal.test(Proportion,Type))["p.value"],method="bonf"),
      Age = p.adjust(unlist(kruskal.test(Proportion,Age))["p.value"],method="bonf"),
      Germline = p.adjust(unlist(kruskal.test(Proportion,Germline))["p.value"],method="bonf"))
by_sample_KW = merge(by_sample_KW,ddply(droplevels(by_sample_all[which(!(by_sample_all$State %in% meth_states)),]),~State,summarise,Cancer = p.adjust(unlist(kruskal.test(Proportion,Cancer))["p.value"],method="bonf")),by="State",all.x=TRUE)

# Wilcox tests by sample classification
by_sample_Wilcox = list(
  merge(ddply(by_sample_all,~State,summarise,Group_pvalue=wilcox_to_all(Proportion,droplevels(Group)),Group=levels(droplevels(Group))),
  aggregate(data=by_sample_all,Proportion~State+Group,mean),by=c("State","Group"),all.x=TRUE),
  merge(ddply(by_sample_all,~State,summarise,Anatomy_pvalue=wilcox_to_all(Proportion,droplevels(Anatomy)),Anatomy=levels(droplevels(Anatomy))),
  aggregate(data=by_sample_all,Proportion~State+Anatomy,mean),by=c("State","Anatomy"),all.x=TRUE),
  merge(ddply(by_sample_all,~State,summarise,Type_pvalue=wilcox_to_all(Proportion,droplevels(Type)),Type=levels(droplevels(Type))),
  aggregate(data=by_sample_all,Proportion~State+Type,mean),by=c("State","Type"),all.x=TRUE),
  merge(ddply(by_sample_all,~State,summarise,Age_pvalue=wilcox_to_all(Proportion,droplevels(Age)),Age=levels(droplevels(Age))),
  aggregate(data=by_sample_all,Proportion~State+Age,mean),by=c("State","Age"),all.x=TRUE),
  merge(ddply(by_sample_all,~State,summarise,Germline_pvalue=wilcox_to_all(Proportion,droplevels(Germline)),Germline=levels(droplevels(Germline))),
  aggregate(data=by_sample_all,Proportion~State+Germline,mean),by=c("State","Germline"),all.x=TRUE))
names(by_sample_Wilcox) = c("Group","Anatomy","Type","Age","Germline")

by_sample_Wilcox$Cancer = 
  merge(ddply(by_sample_all[which(!(by_sample_all$State %in% meth_states)),],~State,summarise,Cancer_pvalue=p.adjust(unlist(wilcox.test(Proportion~Cancer))["p.value"],method="bonf")),
        aggregate(data=by_sample_all,Proportion~State+Cancer,mean),by="State",all.y=TRUE)

# For table
merge(by_sample_stats,merge(by_sample_KW[,1:2],by_sample_Wilcox$Group[which(by_sample_Wilcox$Group$Group_pvalue < 0.05),],by="State",all=TRUE),by="State",all=TRUE)
```

# Enhancer PCA analysis

```{r}
enhancer_matrix_dist = cor(enhancer_matrix[,8:134],use="pairwise.complete.obs")
enhancer_matrix_hclust = hclust(as.dist(enhancer_matrix_dist),method="complete")
plot(enhancer_matrix_hclust,hang=-1)

summary(enhancer_matrix_pca) #To get the proportion and cumulative proportion of variance explained by each PC
```

# Supplementary Figure 11

Supplementary Figure 11. Methylation level of RefSeq features across samples. (B) Average methylation level of CpGs overlapping RefSeq features by sample. 

# Sample groups in each category that are significantly different than all other samples in the proportion of that state in TEs (p<0.05, Wilcox tests with Bonferroni correction).

```{r Figure S11}
#source("R_scripts/proportion_class.R")
#source("R_scripts/WGBS_sample_CpG_average.R")

# By-sample proportion by TE class
by_sample_class = rbind(class_chromHMM[,c(1:3,8)],class_CpG_meth[,c(1:2,4,8)],TE_DNase_class[,c(1:2,4,8)],TE_H3K27ac_class[,c(1:2,4,8)])
by_sample_class = merge(by_sample_class,metadata[,c(1,4:9)],by="Sample")
by_sample_class$Proportion = as.numeric(by_sample_class$Proportion)

a = ggplot(by_sample_class,aes(x=State,y=Proportion,color=Group)) + geom_point(position = position_jitter(width = 0.3)) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=90,hjust=1)) + ylim(0,1) + ylab("Proportion of state in TEs") + scale_color_manual(values=group_colors,guide=FALSE) + scale_x_discrete(limits=levels(by_sample_class$State)) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black") + facet_wrap(~class)

# Feature methylation by sample
b =  ggplot(CpG_feature_meth_average,aes(x=Sample,y=Methylation,group=Cohort,color=Feature,shape=Coding)) + geom_point(size=3) + geom_line() + theme(axis.text.x=element_text(angle=90),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.title.x=element_blank(),legend.title=element_blank(),panel.grid.major.x = element_line(color="black")) + ylab("Average methylation") + scale_color_manual(values=c("black",brewer.pal(8,"Set1"))) + scale_shape_discrete(labels=c("All","Protein-coding","Non-coding")) 
grid.arrange(a,b,nrow=2)
```

# Tests by sample classification and class

```{r}
by_sample_stats_class = ddply(by_sample_class,.(State,class),summarise, Range = max(Proportion)-min(Proportion), Median = median(Proportion), Mean = mean(Proportion),CV = sd(Proportion)/mean(Proportion))

by_sample_KW_class = ddply(by_sample_class,.(State,class),summarise,Group = p.adjust(unlist(kruskal.test(Proportion,Group))["p.value"],method="bonf"))

by_sample_Wilcox_class = merge(ddply(by_sample_class,.(State,class),summarise,Group_pvalue=wilcox_to_all(Proportion,droplevels(Group)),Group=levels(droplevels(Group))),aggregate(data=by_sample_class,Proportion~State+Group+class,mean),by=c("State","Group","class"),all.x=TRUE)

# For table
merge(by_sample_stats_class,merge(by_sample_KW_class,by_sample_Wilcox_class[which(by_sample_Wilcox_class$Group_pvalue < 0.05),],by=c("State","class"),all=TRUE),by=c("State","class"),all=TRUE)
```

# Enrichment of TE class x state x sample

```{r}
class_enrichment = rbind(class_chromHMM[,c(1:3,10)],class_CpG_meth[,c(1:2,4,10)],TE_DNase_class[,c(1:2,4,10)],TE_H3K27ac_class[,c(1:2,4,10)])
```

## CpG average methylation per sample

```{r}
# Stats (TEs and all)
ddply(CpG_meth_average,~Cohort,summarise,Mean=mean(Methylation),SD=sd(Methylation),Outlier=mean(Methylation)-2*sd(Methylation))
ddply(CpG_feature_meth_average,"Cohort",summarise,Mean=mean(Methylation),SD=sd(Methylation))

# Identifying differences by sample Group
ddply(CpG_meth_average,~Cohort,summarise,Group=p.adjust(unlist(kruskal.test(Methylation,metadata[match(Sample,metadata$Sample),]$Group))["p.value"]))
ddply(CpG_meth_average,~Cohort,summarise,Pvalue=wilcox_to_all(Methylation,droplevels(metadata[match(Sample,metadata$Sample),]$Group)),Group=levels(droplevels(metadata[match(Sample,metadata$Sample),]$Group)))
```

## Figure 5

Figure 5. Enrichment of TE subfamilies in epigenetic states in individual samples. (A) Number of samples each subfamily is enriched in each state (log odds ratio > 1.5). The total number of subfamilies enriched LOR > 1.5 in at least one sample are listed under each state. (B) Subfamilies with at least two LOR scores >1.5 in the 7_Enh state (blue squares indicate enrichment in that sample). Binary clustering, complete linkage. Columns are colored by sample metadata, and rows are colored by subfamily class. 

```{r Figure 5}
source("R_scripts/subfamily_enrichment.R")

a = ggplot(subfamily_state_sample_counts,aes(x=State,y=V1,color=class_update)) + geom_point(position=position_jitterdodge(jitter.width=0.5)) + coord_flip() + theme(axis.title.y=element_blank()) + ylab("Number of samples enriched LOR > 1.5") + scale_color_manual(values=class_colors,name="Class") + scale_y_continuous(breaks=c(0,seq(10,120,10),127),expand = c(0.01,0.01)) + geom_rect(xmin=0.5,xmax=1.5,ymin=99,ymax=127,fill="grey",color=NA) + geom_rect(xmin=1.5,xmax=2.5,ymin=54,ymax=127,fill="grey",color=NA) + geom_rect(xmin=2.5,xmax=6.5,ymin=38,ymax=127,fill="grey",color=NA) + geom_vline(xintercept=seq(1.5,20.5,1), colour='grey') + scale_x_discrete(limits = rev(levels(subfamily_state_sample_counts$State)),labels=as.vector(rev(apply(subfamily_state_sample_counts_combine,1,function(x) paste(x[1],"\n(n=",x[2],")",sep="")))))

# Legends for metadata categories
legend_group = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Group)) + geom_tile() + scale_fill_manual(values=group_colors) + theme(legend.direction = "horizontal"))
legend_anatomy = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Anatomy)) + geom_tile() + scale_fill_manual(values=anatomy_colors) + theme(legend.direction = "horizontal"))
legend_type = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Type)) + geom_tile() + scale_fill_manual(values=brewer.pal(5,"Greens")) + theme(legend.direction = "horizontal"))
legend_age = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Age)) + geom_tile() + scale_fill_manual(values=brewer.pal(4,"YlOrRd"),labels=c("Cell line","Fetal","Non-fetal","Unknown")) + theme(legend.direction = "horizontal"))
legend_cancer = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Cancer)) + geom_tile() + scale_fill_manual(values=c("white","red")) + theme(legend.direction = "horizontal"))
legend_germline = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Germline)) + geom_tile() + scale_fill_manual(values=brewer.pal(6,"Dark2")) + theme(legend.direction = "horizontal"))

b = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","7_Enh",2,127,THRESHOLD_LOR)

# Arrange column legends
grid.arrange(legend_group,legend_anatomy,legend_age,legend_cancer,legend_germline,legend_type,ncol=2,widths=c(0.6,0.4),layout_matrix=cbind(c(1,2,6),c(3,4,5)))
```

# Number of enrichments by state and number of enriched subfamilies by state, overall and by class

```{r}
enrichment_table = merge(as.data.frame(table(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),]$State)),subfamily_state_sample_counts_combine,by.x="Var1",by.y="State")
names(enrichment_table)[1:3] = c("State","Enrichments","Subfamilies")
```

## Supplementary Figure 12

Supplementary Figure 12. Subfamilies with at least two logs odds ratio scores >1.5 in the (A) 1_TssA, (B) 2_TssAFlnk, (C) 3_TxFlnk, (D) 4_Tx, (E) 5_TxWk, or (F) 6_EnhG chromHMM states, (G) hypomethylation, (H) DNase peaks, or (I) H3K27ac peaks. Blue squares indicate enrichment in that sample. Binary clustering, complete linkage. Columns are colored by sample metadata, and rows are colored by subfamily class. 

```{r Figure S12}
#source("R_scripts/subfamily_enrichment.R")

a = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","1_TssA",2,127,THRESHOLD_LOR)
b = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","2_TssAFlnk",2,127,THRESHOLD_LOR)
c = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","3_TxFlnk",2,127,THRESHOLD_LOR)
d = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","6_EnhG",2,127,THRESHOLD_LOR)
e = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","4_Tx",2,127,THRESHOLD_LOR)
f = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","5_TxWk",2,127,THRESHOLD_LOR)
g = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Hypomethylated",2,37,THRESHOLD_LOR)
h = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Intermediate",2,37,THRESHOLD_LOR)
i = plot_binary_heatmap(subfamily_state_sample_filter,"DNase","DNase",2,53,THRESHOLD_LOR)
j = plot_binary_heatmap(subfamily_state_sample_filter,"H3K27ac","H3K27ac",2,98,THRESHOLD_LOR)

legend_class = get_legend(ggplot(rmsk_TE_subfamily,aes(x=subfamily,y=class_update,fill=class_update)) + geom_tile() + scale_fill_manual(values=class_colors))

# Arrange column legends
grid.arrange(legend_group,legend_anatomy,legend_age,legend_cancer,legend_germline,legend_type,ncol=2,widths=c(0.6,0.4),layout_matrix=cbind(c(1,2,6),c(3,4,5)))
```

## Supplementary Figure 13

Supplementary Figure 13. Involvement of subfamily members enriched in epigenetic states. (A) Proportion and (B) number of subfamily members in a state for subfamilies enriched in each state. If a subfamily is enriched in a state in multiple samples, each instance is included. No subfamily is enriched in the 15_Quies chromHMM or hypermethylated CpG states. 

```{r Figure S13}

a = ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=Percent,fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Proportion of subfamily members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=c(chromHMM_states[1:14],meth_states[c(1:2,4)],"DNase","H3K27ac"))

b = ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=log10(Members),fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Log10 number of subfamily members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=c(chromHMM_states[1:14],meth_states[c(1:2,4)],"DNase","H3K27ac"))

legend = get_legend(ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=Percent,fill=State)) + geom_boxplot() + scale_fill_manual(values=all_state_colors))

grid.arrange(a,b,legend,nrow=2,layout_matrix=rbind(c(1,3),c(2,3)),widths=c(0.8,0.2))
```

# Average proportion of members in a state for subfamilies enriched in a state

```{r}
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],~State,summarise, Percent=mean(Percent), Members=mean(Members))
```

## Supplementary Figure 14

Supplementary Figure 14. Subfamilies representing >1% of a state. (A) Subfamilies representing more than 1% of a chromHMM state, hypomethylated or intermediately methylated CpGs, or DNase/H3K27ac peak width in a sample. Active regulatory and transcribed chromHMM states are presented. Color represents proportion of the state in that sample in the subfamily. Subfamilies are ordered by decreasing total length.

```{r Figure S14}

ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_filter$State %in% c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")),],aes(x=Sample,y=reorder(subfamily,rmsk_TE_subfamily[match(subfamily,rmsk_TE_subfamily$subfamily),]$Total_length),fill=Length_percent_jk)) + geom_tile(linetype="blank") + theme(axis.ticks.x=element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank()) + scale_fill_gradient(name="% state",low="lightblue",high="darkblue",limits=c(0.009,0.12)) +  facet_wrap(~State)

```

# Subfamily stats, >1%

```{r}
# Number of enrichments by state and number of enriched subfamilies by state, overall and by class
pc_table = merge(as.data.frame(table(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Length_percent_jk > THRESHOLD_PC),]$State)),subfamily_state_sample_counts_pc_combine,by.x="Var1",by.y="State")
names(pc_table)[1:3] = c("State","Instances","Subfamilies")

# Subfamilies with >1% bp of state, not enriched
dim(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_filter$Enrichment <= THRESHOLD_LOR),])

```

# WGBS subfamily hypo/hyper stats

```{r}
source("R_scripts/WGBS_subfamily_hyper_TE.R")

# Subfamilies hypermethylated in all samples
sort(apply(TE_meth_subfamily_hyper[,c(4:13,15:40)],1,function(x) sum(na.omit(x) == 1)))

# Average proportion hypermethylated, with/without IMR90
mean(unlist(TE_meth_subfamily_hyper[,c(4:13,15:40)]),na.rm=TRUE)
mean(unlist(TE_meth_subfamily_hyper[,c(4:40)]),na.rm=TRUE)

source("R_scripts/WGBS_subfamily_enrichment_TE.R")

mean(TE_meth_subfamily_hypo$Range)
mean(TE_meth_subfamily_hypo$Range_noIMR90)
mean(TE_meth_subfamily_hypo$Mean)
mean(TE_meth_subfamily_hypo$Mean_noIMR90)
```

# Subfamily enrichment clustering

```{r}
# Clustering 
enrichment_lor = lapply(c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac"),function(x) enrichment_proportion(subfamily_state_sample_filter[which(subfamily_state_sample_filter$State == x),],"Enrichment",THRESHOLD_LOR))
names(enrichment_lor) = c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")
ldply(enrichment_lor,function(x) x[which(x$Proportion > 0.5 & x$Enriched > 1),])

enrichment_pc = lapply(c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac"),function(x) enrichment_proportion(subfamily_state_sample_filter[which(subfamily_state_sample_filter$State == x),],"Length_percent_jk",THRESHOLD_PC))
names(enrichment_pc) = c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")
ldply(enrichment_pc,function(x) x[which(x$Proportion > 0.5 & x$Enriched > 1),])
```

## Figure 6

MER121 is enriched for the 7_Enh enhancer state in mesodermal samples. (A) GREAT (B) HOMER (C) enrichment

```{r Figure 6}

```

# Subfamily analysis

```{r}
source("R_scripts/TE_correlation_subfamily.R")
# Update all numbers!!

# Input variables (including feature ratio) and number of samples by class
apply(rmsk_TE_subfamily_measure[,c(4:5,7,9,12:162)],2,function(x) unlist(kruskal.test(x~rmsk_TE_subfamily_measure$class_update))["p.value"])
sort(apply(aggregate(data=rmsk_TE_subfamily_measure[,c(3:5,7,9,12:162)],.~class_update,mean)[,2:156],2,function(x) sd(x)/mean(x)))
aggregate(data=rmsk_TE_subfamily_measure[,c(3:5,7,9,12:162)],.~class_update,mean)

# Correlation between input variables (including feature ratio)
correlate_subfam_feature = ldply(colnames(rmsk_TE_subfamily_measure)[c(4:5,7,9,12:35)],function(x) correlate_spearman(rmsk_TE_subfamily_measure,x,c(4:5,7,9,12:35)))
correlate_subfam_feature$Metric = rep(colnames(rmsk_TE_subfamily_measure)[c(4:5,7,9,12:35)],each=196)
correlate_subfam_feature$p.value = as.numeric(as.character(correlate_subfam_feature$p.value))
correlate_subfam_feature$estimate.rho = as.numeric(as.character(correlate_subfam_feature$estimate.rho))
correlate_subfam_feature$class_update = factor(correlate_subfam_feature$class_update,levels=c("DNA","LINE","LTR","SINE","SVA","Other","All"))

# Correlate samples in state, max, mean, range with subfamily features
correlate_subfamily = ldply(colnames(rmsk_TE_subfamily_measure)[c(4:5,7,9,12:35)],function(x) correlate_spearman(rmsk_TE_subfamily_measure,x,c(36:162)))
correlate_subfamily$Metric = rep(colnames(rmsk_TE_subfamily_measure)[c(4:5,7,9,12:35)],each=889)
correlate_subfamily$Metric = factor(correlate_subfamily$Metric,levels=unique(correlate_subfamily$Metric)[c(1:2,26,25,27:28,3:4,21:23,9:13,6:8,14:16,18:20,17,5,24)])
correlate_subfamily$p.value = as.numeric(as.character(correlate_subfamily$p.value))
correlate_subfamily$estimate.rho = as.numeric(as.character(correlate_subfamily$estimate.rho))
correlate_subfamily$class_update = factor(correlate_subfamily$class_update,levels=c("DNA","LINE","LTR","SINE","SVA","Other","All"))

# Divisions of correlations
corr_enrich = unique(correlate_subfamily$State)[c(1:15,76:83,116,121,127)]
corr_mmr = unique(correlate_subfamily$State)[c(31:75,92:115,118:120,123:125,127)]
corr_percent = unique(correlate_subfamily$State)[c(16:30,84:91,117,122)]

# Compare states to each other
correlate_subfam_state = ldply(colnames(rmsk_TE_subfamily_measure)[36:162],function(x) correlate_spearman(rmsk_TE_subfamily_measure,x,c(36:162)))
correlate_subfam_state$State2 = rep(colnames(rmsk_TE_subfamily_measure)[36:162],each=889)
correlate_subfam_state$p.value = as.numeric(as.character(correlate_subfam_state$p.value))
correlate_subfam_state$estimate.rho = as.numeric(as.character(correlate_subfam_state$estimate.rho))
correlate_subfam_state$class_update = factor(correlate_subfam_state$class_update,levels=c("DNA","LINE","LTR","SINE","SVA","Other","All"))

# Plots
ggplot(correlate_subfam_feature,aes(x=State,y=Metric,fill=estimate.rho)) + geom_tile() + facet_wrap(~class_update) + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1))

ggplot(correlate_subfamily,aes(x=State,y=Metric,fill=estimate.rho)) + geom_tile() + facet_wrap(~class_update) + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1))
ggplot(correlate_subfamily[which(correlate_subfamily$State %in% corr_enrich & correlate_subfamily$p.value < 0.001),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + facet_wrap(~class_update) + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1))

ggplot(correlate_subfam_state,aes(x=State,y=State2,fill=estimate.rho)) + geom_tile() + facet_wrap(~class_update) + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1))

# Correlation between age, likelihood enriched in state
wilcox.test(rmsk_TE_age_subfamily[which(subfamily_state_sample_counts[match(rmsk_TE_age_subfamily$subfamily,subfamily_state_sample_counts[which(subfamily_state_sample_counts$State== "6_EnhG"),]$Subfamily),]$V1 == 0),]$JC_distance_mean,rmsk_TE_age_subfamily[which(subfamily_state_sample_counts[match(rmsk_TE_age_subfamily$subfamily,subfamily_state_sample_counts[which(subfamily_state_sample_counts$State== "6_EnhG"),]$Subfamily),]$V1 >0),]$JC_distance_mean)
wilcox.test(rmsk_TEother_age_subfamily[which(rmsk_TEother_age_subfamily$DNase == 0),]$JC_distance_mean,rmsk_TEother_age_subfamily[which(rmsk_TEother_age_subfamily$DNase > 0),]$JC_distance_mean)
wilcox.test(rmsk_TEother_age_subfamily[which(rmsk_TEother_age_subfamily$H3K27ac == 0),]$JC_distance_mean,rmsk_TEother_age_subfamily[which(rmsk_TEother_age_subfamily$H3K27ac > 0),]$JC_distance_mean)

# RNA subfamily analysis - work on
dim(RNA_TE_agnostic_subfamily[which(RNA_TE_agnostic_subfamily$Max_expression > 1),])
```

## Other subfamily analysis

```{r}
# Mean methylation of LTR subfamilies
mean(rmsk_TE_age_subfamily[which(rmsk_TE_age_subfamily$class == "LTR"),]$Mean)
sd(rmsk_TE_age_subfamily[which(rmsk_TE_age_subfamily$class == "LTR"),]$Mean)

# Correlation between age, subfamily average methylation for Alu subfamilies
cor.test(rmsk_TE_age_subfamily[which(rmsk_TE_age_subfamily$family == "Alu"),]$JC_distance_mean,rmsk_TE_age_subfamily[which(rmsk_TE_age_subfamily$family == "Alu"),]$Mean)

# Young subfamilies
rmsk_TEother_age_subfamily[which(rmsk_TEother_age_subfamily$JC_distance_mean < mean(rmsk_TEother_age_subfamily$JC_distance_mean)-2*sd(rmsk_TEother_age_subfamily$JC_distance_mean)),]

# Average age of subfamilies also in mouse 
mean(rmsk_TEother_age_subfamily[which(rmsk_TEother_age_subfamily$subfamily %in% mm10_rmsk_TE$subfamily),]$JC_distance_mean)

# Plotting stats by subfamily 
# (C) Density plot of mean subfamily mappability, by class. (E) Density plot of mean subfamily age, by class.
source("R_scripts/TE_subfamily_stats.R")

ggplot(rmsk_TE_subfamily,aes(x=Mappability,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(panel.grid=element_blank()) + xlab("Mean mappability by subfamily") + ylab("Density") + xlim(0,1) + facet_wrap(~class_update,nrow=2,scales="free_y")

ggplot(rmsk_TE_subfamily,aes(x=Age,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(panel.grid=element_blank()) +  xlab("Mean Jukes-Cantor evolutionary distance by subfamily") + ylab("Density") + facet_wrap(~class_update,nrow=2,scales="free_y")

```

# Linear modelling

```{r}
library(relaimpo)
fit_hypo = lm(data=rmsk_TE_subfamily_measure[,c(4:5,7,9,12:35,113)],Hypomethylated ~ .)
test = calc.relimp(fit_hypo)
```

## Figure 7

Figure 7. Epigenetic state of orthologous TEs in human and mouse. (A) Density plot of the methylation level of TE orthologs in fetal forebrain (ENCFF592SYK) /ganglion eminence-derived primary cultured neurospheres (E054) in human and mouse. (B) Pearson correlation methylation level of orthologous TEs within species, across species in equivalent samples, or across species in randomly paired samples. (C) Pearson correlation for subfamilies present in both human and mouse, proportion of members hypomethylated. (D) Scatter plot of the proportion of hypomethylated members for subfamilies present in both human in mouse in equivalent samples. (E) An example TE hypomethylated in both human and mouse and in a promoter chromHMM state in both. 

```{r Figure 7}
source("R_scripts/human_mouse_ortholog_WGBS.R")

a = ggplot(human_mouse_orthologs_mm10,aes(x=E054,y=ENCFF592SYK/100)) + geom_point(alpha=0.1) + ylab("Mouse TE methylation level") + xlab("Human TE methylation level") + geom_density2d()

b = ggplot(hg19_mm10_TE_WGBS_subfamily_hypo_paired,aes(x=Human,y=Mouse,color=Pair)) + geom_point(alpha=0.5) + scale_color_discrete(guide=FALSE) + xlab("Human subfamily % hypomethylated") + ylab("Mouse subfamily % hypomethylated")

grid.arrange(a,b,nrow=1)
```

## Supplementary Figure 15

Supplementary Figure 15. Mouse TE statistics. (A) Proportion of human TEs and TE subfamilies with orthologs in mouse, by class. (B) chromHMM states in the 7-state model, defining histone modifications, and the 15-state model state with a similar definition.

```{r Figure S14}
source("R_scripts/TE_ortholog_class_stats.R")
#source("R_scripts/human_mouse_ortholog_WGBS.R")

par(mfrow=c(1,2),"mar"=c(5, 4, 4, 3)) 

pie(rmsk_TE_class$Mouse_orthologs,labels=paste(rmsk_TE_class$Mouse_orthologs,"(",round(rmsk_TE_class$Mouse_orthologs/sum(rmsk_TE_class$Mouse_orthologs)*100,0),"%)",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="Orthologous TEs",cex=1,cex.main=2,line=-3,radius=0.5)

pie(rmsk_TE_class$Mouse_ortholog_subfamily,labels=paste(rmsk_TE_class$Mouse_ortholog_subfamily,"(",round(rmsk_TE_class$Mouse_ortholog_subfamily/sum(rmsk_TE_class$Mouse_ortholog_subfamily)*100,0),"%)",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="Orthologous subfamilies",cex=1,cex.main=2,line=-3,radius=0.5)

b = ggplot(human_mouse_samples,aes(x=Test,y=Spearman)) + geom_boxplot() + ylab("Spearman correlation across TEs") + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse, same tissue","Human-mouse, random tissue")) + scale_y_continuous(limits=c(0,1))

c = ggplot(human_mouse_samples,aes(x=Test,y=Subfamily_Spearman)) + geom_boxplot() + ylab("Spearman correlation across subfamilies") + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse, same tissue","Human-mouse, random tissue")) + scale_y_continuous(limits=c(0,1))

grid.arrange(b,c,nrow=1)
```

# Mouse WGBS analysis

```{r}
# Average age of TEs with mouse orthologs
mean(merge(rmsk_TEother_age,unique(human_mouse_orthologs_mm10[,8:14]),by.x=c("chromosome","start","stop","subfamily","family","class","strand"),by.y=c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_family","human_class","human_strand_hg19"))$JC_distance)

# Number of orthologs with CpGs
merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,1:7]),by.x=colnames(TE_CpG_count)[1:7],by.y=colnames(human_mouse_orthologs_mm10)[1:7])

# Number of CpGs per orthologs with CpGs
mean(merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,1:7]),by.x=colnames(TE_CpG_count)[1:7],by.y=colnames(human_mouse_orthologs_mm10)[1:7])$CpGs)

# Chi-sq test for human-mouse pairs
apply(human_mouse_samples[1:13,],1,function(x) unlist(chisq.test(matrix(c(nrow(human_mouse_orthologs_mm10_category[which(human_mouse_orthologs_mm10_category[,x[1]] == "Hypomethylated" & human_mouse_orthologs_mm10_category[,x[2]] == "Hypomethylated"),]),nrow(human_mouse_orthologs_mm10_category[which(human_mouse_orthologs_mm10_category[,x[1]] == "Hypomethylated" & human_mouse_orthologs_mm10_category[,x[2]] %in% c("Hypermethylated","Intermediate")),]),nrow(human_mouse_orthologs_mm10_category[which(human_mouse_orthologs_mm10_category[,x[1]] %in% c("Hypermethylated","Intermediate") & human_mouse_orthologs_mm10_category[,x[2]] == "Hypomethylated"),]),nrow(human_mouse_orthologs_mm10_category[which(human_mouse_orthologs_mm10_category[,x[1]] %in% c("Hypermethylated","Intermediate") & human_mouse_orthologs_mm10_category[,x[2]] %in% c("Hypermethylated","Intermediate")),])),nrow=2)))["p.value"])

# chromHMM analysis
write.table(unique(human_mouse_orthologs_mm10_hypo[which(human_mouse_orthologs_mm10_hypo$Human_state == "1_TssA" & human_mouse_orthologs_mm10_hypo$Mouse_State == "1"),9:15]),row.names=FALSE,col.names=FALSE,quote=FALSE,sep='\t',file="Mouse/hg19_orthologs_hypo_1TssA.bed")
```
