---
title: "TE landscape Draft"
author: "Erica Pehrsson"
date: "`r Sys.Date()`"
output: word_document
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(dev = 'pdf',dev.args=list(pdf = list(useDingbats = FALSE)),fig.width=7, fig.height=8, fig.path='TE_landscape_draft_figures/', cache.path="TE_landscape_draft_cache/")
```

# Setup

## Load required libraries

```{r load libraries}
library(plyr)
packageVersion("plyr")
library(reshape2)
packageVersion("reshape2")
library(ggplot2)
packageVersion("ggplot2")
library(NMF)
packageVersion("NMF")
library(gridExtra)
packageVersion("gridExtra")
library(RColorBrewer)
packageVersion("RColorBrewer")
library(scales)
packageVersion("scales")
library(mgcv)
packageVersion("mgcv")
library(extrafont)
packageVersion("extrafont")
```

## Set universal plot parameters

```{r set plot parameters}
theme_set(theme_bw(base_size=10) %+replace%  theme(panel.background=element_rect(fill="white"),panel.border=element_rect(color="black",fill=NA),panel.grid=element_blank()))
```

## Define color and label vectors

```{r define vectors}
# TE coordinates
TE_coordinates = c("chromosome","start","stop","subfamily","family","class","strand")
hg19_coordinates = c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")
mm10_coordinates = c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10")

# Chromosomes
standard_chromosomes = c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY","chrM")

# chromHMM
chromHMM_states = c("1_TssA","2_TssAFlnk","3_TxFlnk","4_Tx","5_TxWk","6_EnhG","7_Enh","8_ZNF/Rpts","9_Het","10_TssBiv","11_BivFlnk","12_EnhBiv","13_ReprPC","14_ReprPCWk","15_Quies")
chromHMM_labels = setNames(c("Active TSS","Flanking active TSS","Transcription at gene 5'/3'","Strong transcription","Weak transcription","Genic enhancer","Enhancer","ZNF genes/repeats","Heterochromatin","Poised TSS","Flanking poised TSS/enhancer","Poised enhancer","Repressed Polycomb","Weak repressed Polycomb","Quiescent"),chromHMM_states)
chromHMM_colors = setNames(c(rgb(255,0,0,maxColorValue=255),rgb(255,69,0,maxColorValue=255),rgb(50,205,50,maxColorValue=255),rgb(0,128,0,maxColorValue=255),rgb(0,100,0,maxColorValue=255),rgb(194,225,5,maxColorValue=255),rgb(255,255,0,maxColorValue=255),rgb(102,205,170,maxColorValue=255),rgb(138,145,208,maxColorValue=255),rgb(205,92,92,maxColorValue=255),rgb(233,150,122,maxColorValue=255),rgb(189,183,107,maxColorValue=255),rgb(128,128,128,maxColorValue=255),rgb(192,192,192,maxColorValue=255),"grey90"),chromHMM_states)

# WGBS
meth_states = c("Hypomethylated","Intermediate","Hypermethylated","Missing")
meth_labels = setNames(c("Hypomethylated","Intermediate meth","Hypermethylated","Missing meth data"),meth_states)
meth_colors = setNames(c("#F8766D","#00BFC4","#7CAE00","#C77CFF"),meth_states)

# All states
states = c(chromHMM_states,meth_states,"DNase","H3K27ac","Expressed_samples")
all_state_colors = c(chromHMM_colors,meth_colors,"aquamarine","tan1","black","gray28","gray28") 
names(all_state_colors)[20:24] = c("DNase","H3K27ac","Expressed_samples","Bases","CpGs")
all_state_labels = setNames(c(chromHMM_states,meth_labels,"DHS","H3K27ac","RPKM > 1"),states)
all_pc_states = c(states[1:21],unlist(lapply(states[1:21],function(x) paste(x,"PC",sep="_"))))

# Metric labels
metric_labels = c("chromHMM","WGBS","DNase","H3K27ac","RPKM > 1")
names(metric_labels) = c("chromHMM","WGBS","DNase","H3K27ac","Expressed_samples")

# TE classification
class_colors = setNames(c("#4A72E8","#FF6600","#006600","#cc0000","lightseagreen","#5C5C5C"),c("DNA","LINE","LTR","SINE","SVA","Other"))

# Sample classification
sample_categories = c("Group","Anatomy","Type","Age","Cancer","Germline")

group_colors = setNames(c("#924965","#4178AE","#E41A1C","#69608A","#B65C73","#FF9D0C","#678C69","#55A354","#E67326","#FFD924","#AF5B39","#D56F80","#999999","#C5912B","#C58DAA","#F182BC","#C2655D","#DAB92E","#000000"),c("ESC","ES-deriv","IMR90","iPSC","Mesench","Epithelial","HSC & B-cell","Blood & T-cell","Myosat","Neurosph","Adipose","Heart","Other","Brain","Digestive","Sm. Muscle","Muscle","Thymus","ENCODE2012"))
anatomy_colors = setNames(c("#B2DF8A","#33A02C","#FB9A99","#FFED6F","#80B1D3","#666666","#924965","#4178AE","#FDBF6F","#CAB2D6","#6A3D9A","#BEBADA","#FCCDE5","#BC80BD","#E7298A","#FF7F00","#69608A","#B3DE69","#CCEBC5","#000000","#FB8072","#FDB462","#8DD3C7","#D9D9D9","#E31A1C","#A6CEE3","#D95F02","#E6AB02","#66A61E","#B15928","#E31A1C","#8DD3C7","#E7298A","#A6761D","#666666"),c("ADRENAL","BLOOD","BONE","BRAIN","BREAST","CERVIX","ESC","ESC_DERIVED","FAT","GI_COLON","GI_DUODENUM","GI_ESOPHAGUS","GI_INTESTINE","GI_RECTUM","GI_STOMACH","HEART","IPSC","KIDNEY","LIVER","LUNG","MUSCLE","MUSCLE_LEG","OVARY","PANCREAS","PLACENTA","SKIN","SPLEEN","STROMAL_CONNECTIVE","THYMUS","VASCULAR",NA,NA,NA,NA,NA))
type_colors = setNames(brewer.pal(5,"Greens"),c("PrimaryCulture","PrimaryCell","PrimaryTissue","CellLine","ESCDerived"))
age_colors = setNames(brewer.pal(4,"YlOrRd"),c("Cell_line","Unknown","Non_fetal","Fetal"))
cancer_colors = setNames(c("lightgrey","red"),c("No","Yes"))
germline_colors = setNames(brewer.pal(6,"Dark2"),c("Pluripotent","Mesoderm","Ectoderm","Endoderm","Mixed","Other"))

# Genic features
features = c("cpgIsland","promoters","5UTR","CDS","3UTR","exons","introns","intergenic")
genic_labels = setNames(c("CpG island","Promoter","5'UTR","CDS","3'UTR","Exon","Intron","Intergenic","CDS","Vista enhancers"),c(features,"coding_exon","Vista_enhancers"))

cohorts = c("cpgIsland","promoters","promoters_pc","promoters_nc","5UTR","5UTR_pc","5UTR_nc","coding_exon","coding_exon_pc","3UTR","3UTR_pc","3UTR_nc",
                     "exons","exons_pc","exons_nc","introns","introns_pc","introns_nc","intergenic","blacklist","Vista_enhancers")

# Feature labels
coding_labels = c("All","Protein-coding","Non-coding")
names(coding_labels) = c("All","pc","nc")

# C-Gate subfamilies
cgate_subfams = c('AluJb','AluJo','AluS','AluSc','AluSg','AluSg1','AluSp','AluSq','AluSx','AluSz','AluY','FLAM_A','FRAM','HERV3-int','HERV9-int','HERVE-int','HERVH-int','L1HS','L1MB4','L1PA6','L2a','LTR10A','LTR16A','LTR22B','LTR31A','LTR5','LTR5_Hs','LTR9','MER11A','MER21A','MER34','MER39B','MER4B','MER54B','MER74C','Mir m','MIR3','MIRb','MLT1C','MLT2B2','MLT2B3','MSTB2','Plat_L3','RTVL-H','THE1A','THE1C','THE1D','AluJr','AluSq2','L1MB7','L1MC4','L1ME3A','L1PA11','L1PA4','LTR12','LTR12C','LTR12D','LTR13A','LTR2','LTR2B','MIR','MIRc','MLT1B')

# Colnames vectors
measure_states_extra = c("States.chromHMM","Max_states_intra","States.WGBS","Max_expression")
measure_metrics = c("Length","mappability","JC_distance","CpGs","CpGs_per_length")
measure_metrics_subfam = c("Count","Total_length","Length","Mappability","Age","CpGs","Count_CpGs","CpGs_per_TE","CpGs_per_kbp")

enrichment_names = c("Length_ijk","Length_ik","Length_jk","Length_k","Enrichment","Length_percent_jk","Members","Count","Percent")

# Metric labels
variable_labels = setNames(c("Length (bp)","Mappability","Jukes-Cantor evolutionary distance","CpGs","CpGs/bp"),measure_metrics)
measure_labels = setNames(c("Number of TEs","Total length (bp)","Median length (bp)","Median mappability","Median age (JC)","Total CpGs","TEs with CpGs","CpGs/TE","CpGs/kbp"),measure_metrics_subfam)

# Mouse-human chromHMM states
mouse_chromHMM_states = c("TssA","TssAFlnk1","TssAFlnk2","Tx1","Tx2","Enh","EnhLo1","EnhLo2","HetCons","HetFac","TssBiv","EnhPois1","EnhPois2","QuiesG","Quies")
mouse_chromHMM_colors = setNames(c("red","orangered2","orangered4","green4","darkgreen","yellow","gold1","goldenrod2","cornflowerblue","steelblue2","indianred3","burlywood1","lightgoldenrod2","grey","lightgrey"),mouse_chromHMM_states)

overlap_labels = c("% genome","% TEs (either strand)","% TEs (same strand)","% feature in TEs")
names(overlap_labels) = c("Genome_percent_genome","TEs_percent_TE","TEs_strand_percent_TE","Genome_percent_TE")
```

## Load functions

```{r load functions}
source("R_scripts/functions.R")
```

## Create essential matrices

```{r create essential matrices, eval=FALSE}
source("R_scripts/metadata.R")
source("R_scripts/TE_matrix.R")
source("R_scripts/chromHMM_matrix.R")
source("R_scripts/WGBS_TE_avg_methylation.R")
source("R_scripts/DNase_peaks.R")
source("R_scripts/H3K27ac_peaks.R")
source("R_scripts/RNAseq.R")
source("R_scripts/promoter_matrices.R")
```

## Load metadata

```{r load metadata}
load("R_datasets/metadata.RData")
```

## Load sample statistics

```{r load sample stats}
# Samples with each metric: all, not excluded, no Y, both
sample_counts = rbind(apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x)),])[1]),
                      apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x) & metadata$Exclude == "Include"),])[1]),
                      apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x) & metadata$chrY == "Yes"),])[1]),
                      apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x) & metadata$chrY == "Yes" & metadata$Exclude == "Include"),])[1]))
colnames(sample_counts)[1] = "chromHMM"
rownames(sample_counts) = c("All","Include","chrY","Include.chrY")

mark_labels = c(paste("chromHMM (n=",sample_counts["All","chromHMM"],")",sep=""),
                paste("WGBS (n=",sample_counts["All","WGBS"],")",sep=""),
                paste("DHS (n=",sample_counts["All","DNase"],")",sep=""),
                paste("H3K27ac (n=",sample_counts["All","H3K27ac"],")",sep=""))
names(mark_labels) = c("chromHMM","WGBS","DNase","H3K27ac")
```

## Load essential matrices

```{r load essential matrices}
load("R_datasets/rmsk_TE.RData")
load("R_datasets/chromHMM_TE_state.RData")
load("R_datasets/TE_meth_average.RData")
load("R_datasets/TE_DNase_peaks.RData")
load("R_datasets/TE_H3K27ac_peaks.RData")
load("R_datasets/rna.RData")
```

## Load feature sizes

```{r load feature sizes}
# Genome
hg19_genome = read.table("raw_data/hg19.genome",sep='\t',header=TRUE)
GENOME_WIDTH = sum(as.numeric(hg19_genome[which(hg19_genome$chrom %in% standard_chromosomes),]$size))
GENOME_WIDTH_noY = sum(as.numeric(hg19_genome[which(hg19_genome$chrom %in% standard_chromosomes & hg19_genome$chrom != "chrY"),]$size))

# Total widths
CHROMHMM_TOTAL_WIDTH = sample_counts["chrY","chromHMM"]*GENOME_WIDTH + (sample_counts["All","chromHMM"]-sample_counts["chrY","chromHMM"])*GENOME_WIDTH_noY
DNASE_TOTAL_WIDTH = sample_counts["chrY","DNase"]*GENOME_WIDTH + (sample_counts["All","DNase"]-sample_counts["chrY","DNase"])*GENOME_WIDTH_noY
H3K27AC_TOTAL_WIDTH = sample_counts["chrY","H3K27ac"]*GENOME_WIDTH + (sample_counts["All","H3K27ac"]-sample_counts["chrY","H3K27ac"])*GENOME_WIDTH_noY

# TEs 
NUM_TE = dim(rmsk_TE)[1]
NUM_TE_noY = dim(rmsk_TE[which(rmsk_TE$chromosome != "chrY"),])[1]
NUM_TE_WGBS = dim(TE_meth_average)[1]
  
# awk '{sum+=$3-$2}END{print sum}' features/TEs/merge/rmsk_TEother_merge.txt
MERGED_TE_WIDTH = 1389947349
# awk '{if($1 != "chrY") sum+=$3-$2}END{print sum}' features/TEs/merge/rmsk_TEother_merge.txt
MERGED_TE_WIDTH_noY = 1375898142

# Total widths
CHROMHMM_TE_WIDTH = sample_counts["chrY","chromHMM"]*MERGED_TE_WIDTH + (sample_counts["All","chromHMM"]-sample_counts["chrY","chromHMM"])*MERGED_TE_WIDTH_noY
DNASE_TE_WIDTH = sample_counts["chrY","DNase"]*MERGED_TE_WIDTH + (sample_counts["All","DNase"]-sample_counts["chrY","DNase"])*MERGED_TE_WIDTH_noY
H3K27AC_TE_WIDTH = sample_counts["chrY","H3K27ac"]*MERGED_TE_WIDTH + (sample_counts["All","H3K27ac"]-sample_counts["chrY","H3K27ac"])*MERGED_TE_WIDTH_noY

# CpGs
ALL_CPGS = 56434896/2
TE_CPGS = 28373958/2

# Feature lengths
feature_overlap = as.data.frame(matrix(readLines("features/intersect_features/feature_overlap.txt"),ncol=5,byrow=TRUE))
colnames(feature_overlap) = c("Cohort","Genome","Genome_noY","TEs","TEs_stranded")
feature_overlap$Cohort = gsub("refseq_","",feature_overlap$Cohort)
feature_overlap = split_coding(feature_overlap)
feature_overlap[,2:5] = apply(feature_overlap[,2:5],2,function(x) as.numeric(x))
feature_overlap = feature_overlap[,c(1,6:7,2:5)]
```

## Calculate TE class and subfamily stats

```{r calculate TE stats}
source("R_scripts/TE_class_stats.R")
source("R_scripts/TE_subfamily_stats.R")
```

# Introduction

## Basic sample stats

```{r basic sample stats}
sample_counts
```

## Basic TE statistics

```{r basic TE stats}
# Number of TEs
NUM_TE

# % of genome
MERGED_TE_WIDTH/GENOME_WIDTH

# % CpGs overlapping TEs
TE_CPGS/ALL_CPGS

# Number of families
length(unique(rmsk_TE$family))

# Number of subfamilies
length(unique(rmsk_TE$subfamily))

# Length of TEs
ddply(rmsk_TE,.(),summarise,Min=min(Length),Median=median(Length),Max=max(Length))

# % TEs overlapping CpGs
NUM_TE_WGBS
NUM_TE_WGBS/NUM_TE
NUM_TE-NUM_TE_WGBS
ddply(TE_meth_average,.(),summarise,Min=min(CpGs),Median=median(CpGs),Max=max(CpGs))
```

# Results

## Contribution

### Figure 1

```{r Figure 1, echo=FALSE}
source("R_scripts/proportion.R")
source("R_scripts/proportion_class.R") 

a = ggplot(combined_proportion[which(combined_proportion$Coding == "All"),],aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion of feature in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.title.x=element_blank()) + scale_x_discrete(labels=c("TE","Genome",genic_labels)) + facet_wrap(~Mark,nrow=1,labeller=labeller(Mark = mark_labels))

b = ggplot(combine_class_proportion,aes(x=class,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion of class in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.title.x = element_blank()) + facet_wrap(~Cohort,nrow=1)

c = ggplot(contribution,aes(x=State,y=Proportion,fill=State)) + geom_bar(stat="identity") + coord_flip() + scale_y_reverse(limits=c(1,0)) + ylab("Proportion of state in TEs") + theme(axis.title.y=element_blank()) + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(labels=all_state_labels,limits=c(rev(states[16:19]),"CpGs",rev(states[c(1:15,20:21)]),"Bases")) + geom_text(aes(label=round(Proportion,2)),hjust=1.01,size=3) + geom_vline(xintercept=5.5) + geom_vline(xintercept=4.5,linetype="dashed",color="black") + geom_vline(xintercept=6.5,linetype="dashed",color="grey") + geom_vline(xintercept=7.5,linetype="dashed",color="grey") + geom_vline(xintercept=22.5,linetype="dashed",color="black") 

d = ggplot(contribution_class,aes(x=State,y=Proportion,fill=class)) + geom_bar(stat="identity") + coord_flip() + ylab("Proportion in each TE class") + theme(axis.title.y=element_blank(),axis.text.y=element_blank(),axis.ticks.y = element_blank()) + scale_x_discrete(limits=c(rev(states[16:19]),"CpGs",rev(states[c(1:15,20:21)]),"Bases")) + scale_fill_manual(values=class_colors,guide=FALSE) + geom_vline(xintercept=5.5) + geom_vline(xintercept=4.5,linetype="dashed",color="black") + geom_vline(xintercept=6.5,linetype="dashed",color="grey") + geom_vline(xintercept=7.5,linetype="dashed",color="grey") + geom_vline(xintercept=22.5,linetype="dashed",color="black") 

legend_class = get_legend(ggplot(contribution_class,aes(x=State,y=Proportion,fill=class)) + geom_bar(stat="identity") + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.text.y=element_blank(),legend.position="bottom",legend.direction = "horizontal") + scale_fill_manual(values=class_colors,name="Class") + guides(fill = guide_legend(nrow = 1)))

grid.arrange(a,b,c,d,legend_class, nrow = 4,layout_matrix = rbind(c(1,1),c(2,2),c(3,4),c(6,6)),heights=c(0.3,0.3,0.35,0.05),widths=c(0.55,0.45))
```

### Proportion/contribution analysis

```{r analysis Fig 1}
# Proportion of each feature in each state, across all samples
combined_proportion[which(combined_proportion$Coding == "All"),]

# Specific states, TEs, promoters, and exons
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "TE" & combined_proportion$State %in% chromHMM_states[c(1:3,6:7)]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "TE" & combined_proportion$State %in% chromHMM_states[4:5]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "promoters" & combined_proportion$State %in% chromHMM_states[c(1:3,6:7)]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "exons" & combined_proportion$State %in% chromHMM_states[4:5]),]$Proportion)

# Does any feature have less % in state than TEs?
ddply(combined_proportion[which(combined_proportion$Coding == "All"),],.(State),function(x) x[which(x$Proportion > x[which(x$Feature == "TE"),]$Proportion),c("State","Cohort","Feature","Coding","Proportion")])

# By-class proportion
dcast(combine_class_proportion,class~State,value.var="Proportion")
test = ddply(combine_class_proportion,.(State),function(x) x[which.max(x$Proportion),])
test[order(test$class),]

# Proportion of all genomic bases in TEs
CHROMHMM_TE_WIDTH/CHROMHMM_TOTAL_WIDTH

# Proportion of all/TE CpGs in TE classes
rmsk_TE_class[,c("class_update","CpGs","Percent_TE_CpGs","Percent_all_CpGs")]

# Contribution
contribution

# Contribution, composite states
contribution_composite

# Contribution by class
dcast(contribution_class,class~State,value.var="Proportion")
```

### VISTA enhancers

```{r vista enhancers}
# wc -l features/vista_enhancers/vista_enhancers_human.bed
# cat features/intersect_features/vista_enhancers_rmsk_*.txt | awk '{print $1, $2, $3}' - | sort | uniq | wc -l
355

355/934
```

### Supplementary Figure 1

```{r Figure S1, echo=FALSE}
source("R_scripts/histone_modification.R")

# Plot bases in genome/TEs in each state by sample, increasing order
all_state_proportion_matrix = dcast(all_state_proportion[which(all_state_proportion$Cohort %in% c("Genome","TE")),],State+Sample+Group~Cohort,value.var="Bases")
all_state_proportion_corr = ddply(all_state_proportion_matrix,.(State),summarise,Cor=as.numeric(unlist(cor.test(Genome,TE,method="spearman"))["estimate.rho"]),Pvalue=as.numeric(unlist(cor.test(Genome,TE,method="spearman"))["p.value"]))

a = ggplot(all_state_proportion_matrix,aes(x=log10(Genome),y=log10(TE))) + geom_point(aes(color=Group)) + scale_color_manual(values=group_colors,guide=FALSE) + facet_wrap(~State,scales="free",ncol=6) + xlab("Bases/CpGs in genome (log10)") + ylab("Bases/CpGs in TEs (log10)") + geom_smooth(method="lm",se=FALSE,color="black") + geom_text(data=all_state_proportion_corr,aes(x = -Inf, y = Inf,label=round(Cor,2)),hjust=-0.1,vjust=1.5,size=3) + scale_y_continuous(breaks=pretty_breaks(n=3)) + scale_x_continuous(breaks=pretty_breaks(n=3))

b = ggplot(histones,aes(x=Bin,y=Level,color=Mark,group=Mark)) + geom_line(size=1.5,alpha=0.7) + theme(axis.text.x = element_blank(),axis.ticks.x=element_blank(),panel.grid=element_blank(),legend.position = "bottom",legend.title = element_blank())+ylab("Fold enrichment over input") + scale_color_manual(labels=c(levels(histones$Mark)[1:8],"DNA methylation"),values=c(gg_color_hue(8),"black")) + xlab("Genomic position (+/-5kb around TE)")+ylim(0,12) + facet_wrap(~ State,nrow=3) + scale_y_continuous(sec.axis = sec_axis(~./12.5, name = "DNA methylation level"))

grid.arrange(a,b,heights=c(0.5,0.5))
```

### Compare TE bases to non-TE bases by sample

```{r proportion investigation}
ddply(all_state_proportion[which(all_state_proportion$Cohort %in% c("Genome","TE")),],.(State,Cohort),summarise,Median=median(Bases),Mean=mean(Bases),SD=sd(Bases),Min=min(Bases),Max=max(Bases),CV=sd(Bases)/mean(Bases))

ddply(all_state_proportion[which(all_state_proportion$Cohort == "Genome"),],.(State,Cohort),summarise,Median=median(Proportion))

dcast(all_state_proportion[which(all_state_proportion$Mark == "WGBS" & all_state_proportion$Cohort == "Genome"),c("Sample","State","Proportion")],Sample~State,value.var="Proportion")

test = ddply(all_state_proportion[which(all_state_proportion$Feature == "Genome"),],.(State),transform,Mean=mean(Proportion),SD=sd(Proportion))
tail(test[order(abs(test$Proportion-test$Mean)/test$SD),c("Sample","State","Proportion","Mean","SD")],n=15)

test = ddply(all_state_proportion_matrix,.(State,Group),summarise,Genome_median=median(Genome))
test[order(test$State,test$Genome_median),]

# Not MHC corrected (0.05/21 ~ 0.0024)
ddply(all_state_proportion_matrix,.(State),summarise,Genome_corr = unlist(cor.test(Genome-TE,TE,method="spearman"))["estimate.rho"],Genome_pvalue = unlist(cor.test(Genome-TE,TE,method="spearman")["p.value"]))
ddply(all_state_proportion_matrix,.(State),summarise,Genome_corr = unlist(cor.test(Genome,TE/Genome,method="spearman"))["estimate.rho"],Genome_pvalue = unlist(cor.test(Genome,TE/Genome,method="spearman")["p.value"]))
```

### Supplementary Figure 2

```{r Figure S2, echo=FALSE}
source("R_scripts/feature_overlap.R")

## Average proportion by sample
a = ggplot(ddply(all_state_proportion[which(all_state_proportion$Coding == "All"),],.(Mark,State,Feature),summarise,Proportion=mean(Proportion)),aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.title.x=element_blank()) + scale_x_discrete(labels=c("TE","Genome",genic_labels)) + facet_wrap(~Mark,nrow=1,labeller=labeller(Mark = mark_labels))

# Total proportion split by protein-coding/non-coding
b = ggplot(combined_proportion,aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion of feature in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.title.x=element_blank()) + scale_x_discrete(labels=c("TE","Genome",genic_labels)) + facet_grid(Coding~Mark,labeller=labeller(Coding=coding_labels))

# Feature overlap
c = ggplot(feature_overlap_long,aes(x=Feature,y=Percent,fill=Coding)) + geom_bar(position="dodge",stat="identity") + theme(axis.title.x=element_blank(),axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),legend.title = element_blank())  + scale_x_discrete(labels=genic_labels) + scale_fill_discrete(guide=FALSE) + facet_wrap(~Measure,nrow=1,labeller=labeller(Measure=overlap_labels)) + ylab("Proportion")

grid.arrange(a,b,c,nrow=3,heights=c(0.25,0.5,0.25))
```

### RefSeq feature analysis

```{r analysis Fig S2}
# Feature %
feature_overlap_long

# Proportion of each feature in each state, across all samples
combined_proportion
```

### Supplementary Figure 3

```{r Figure S3, echo=FALSE}
a = ggplot(rmsk_TE_class,aes(x=factor(1),y=Total_length,fill=class_update,label=paste(round(Total_length/1000000,1)," (",round(Total_length/sum(Total_length)*100,0),"%) ",sep=""))) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(name="Class",values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank()) + geom_text(position = position_fill(vjust = 0.5),size=3)

b = ggplot(rmsk_TE_class,aes(x=factor(1),y=Count,fill=class_update,label=paste(format(Count, big.mark=",", scientific=FALSE)," (",round(Count/sum(Count)*100,0),"%) ",sep=""))) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(name="Class",values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank()) + geom_text(position = position_fill(vjust = 0.5),size=3) 

c = ggplot(rmsk_TE_class,aes(x=factor(1),y=Subfamilies,fill=class_update,label=paste(format(Subfamilies, big.mark=",", scientific=FALSE)," (",round(Subfamilies/sum(Subfamilies)*100,0),"%) ",sep=""))) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(name="Class",values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank()) + geom_text(position = position_fill(vjust = 0.5),size=3)

d = ggplot(rmsk_TE,aes(x=Length/1000,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank(),axis.title.y = element_blank()) + xlab("Length (kbp)") + facet_wrap(~class_update,nrow=1,scales="free_y") 

e = ggplot(rmsk_TE,aes(x=CpGs,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank(),axis.title.y = element_blank()) + xlab("CpGs") + facet_wrap(~class_update,nrow=1,scales="free_y") + scale_x_continuous(breaks=pretty_breaks(n=4))

f = ggplot(rmsk_TE,aes(x=CpGs_per_length,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank(),axis.title.y = element_blank()) + xlab("CpGs/bp") + facet_wrap(~class_update,nrow=1,scales="free_y") + scale_x_continuous(breaks=pretty_breaks(n=4))

g = ggplot(rmsk_TE,aes(x=mappability,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank(),axis.title.y = element_blank()) + xlab("Mappability") + xlim(0,1) + facet_wrap(~class_update,nrow=1,scales="free_y") 

h = ggplot(rmsk_TE,aes(x=JC_distance,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank(),axis.title.y = element_blank()) +  xlab("Jukes-Cantor evolutionary distance") + facet_wrap(~class_update,nrow=1,scales="free_y") + scale_x_continuous(breaks=pretty_breaks(n=4))

grid.arrange(a,b,c,d,e,f,g,h,nrow=6,layout_matrix=rbind(c(1,2,3),c(4,4,4),c(5,5,5),c(6,6,6),c(7,7,7),c(8,8,8)),heights=c(0.25,0.15,0.15,0.15,0.15,0.15))
```

### TE metric stats

```{r metrics analysis}
# Median TE metrics by overall and by class, with p-values
p.adjust(apply(rmsk_TE[,measure_metrics],2,function(x) unlist(kruskal.test(x~rmsk_TE$class_update))["p.value"]),method="bonf")

apply(rmsk_TE[,measure_metrics],2,median)
aggregate(data=rmsk_TE[,c("class_update",measure_metrics)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)
sort(apply(aggregate(data=rmsk_TE[,c("class_update",measure_metrics)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)[,2:6],2,function(x) IQR(x)/median(x)))

# Max TE length by class
aggregate(data=rmsk_TE,Length~class_update,max)
```

## Potential

### Figure 2

```{r Figure 2, echo=FALSE, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/potential.R")
source("R_scripts/potential_class.R")

a = ggplot(combine_boxplot,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="Proportion of TEs in state") + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + geom_point(data=combine_stats,aes(x=State,y=Proportion_ever),color="red",size=3)

b = ggplot(combine_class_ever,aes(x=State,y=Proportion,fill=Class)) + geom_bar(stat="identity") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + labs(y="TEs ever in state by class") + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels))

c = ggplot(combine_potential[which(combine_potential$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + geom_boxplot(width=0.2,outlier.size = 0.1)

d = ggplot(combine_potential[which(combine_potential$Sample.Proportion == 1),],aes(x=State,y=log10(Count+1),fill=State)) + geom_bar(stat="identity") + geom_text(aes(label=format(Count, big.mark=",", scientific=FALSE)),hjust=1.01,size=3) + scale_y_reverse(limits=c(8,0)) + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + ylab("TEs in state in\n100% of samples (log10)")

e = ggplot(ddply(combine_potential[which(combine_potential$Sample.Proportion >= 0.9),],.(State),summarise,Count=sum(Count)),aes(x=State,y=log10(Count+1),fill=State)) + geom_bar(stat="identity") + geom_text(aes(label=format(Count, big.mark=",", scientific=FALSE)),hjust=0,size=3) + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("TEs in state in\n90% of samples (log10)") + scale_y_continuous(limits=c(0,8))

f = ggplot(ddply(combine_potential_class[which(combine_potential_class$Sample.Proportion >= 0.9),],.(State,Class),summarise,TEs.Proportion=sum(Count)/sum(Total)),aes(x=State,y=TEs.Proportion,fill=Class)) + geom_bar(stat="identity") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + labs(y="Proportion in\neach class (90%)") + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels))

grid.arrange(a,b,c,d,e,f, nrow = 2, layout_matrix=rbind(c(1,2,3),c(4,5,6)),widths=c(0.4,0.3,0.3))
```

### Potential analysis

```{r analysis Fig 2a, cache=TRUE}
ddply(combine_boxplot,.(State),summarise,Median=median(Proportion),Mean=mean(Proportion))

# Number of TEs ever in composite states
## Roadmap Active (states 1-8)
sum(apply(chromHMM_TE_state[8:15],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Roadmap Inactive (states 9-15)
sum(apply(chromHMM_TE_state[16:22],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Active regulatory states (states 1-3, 6-7)
sum(apply(chromHMM_TE_state[c(8:10,13:14)],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Transcribed states (states 4-5)
sum(apply(chromHMM_TE_state[11:12],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Poised regulatory states (states 10-12)
sum(apply(chromHMM_TE_state[17:19],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Repressed states (states 9, 13-14)
sum(apply(chromHMM_TE_state[c(16,20,21)],1,function(x) sum(x > 0)) > 0)/NUM_TE

# Number of TEs ever in each state
ddply(combine_potential,.(State),summarise,Count=sum(Count[which(Samples != 0)]))
combine_stats
ddply(rmsk_TE_class,.(class_update,Count),summarise,TE_prop=Count/NUM_TE)
dcast(combine_class_ever,Class~State,value.var="Proportion")
combine_potential_class[which(combine_potential_class$Sample.Proportion == 1),]

# Median samples in state (ever)
ddply(combine_potential[which(combine_potential$Samples != 0),],.(State),summarise,first=quantile(rep(Sample.Proportion,times=Count),0.25),Median=median(rep(Sample.Proportion,times=Count)),third=quantile(rep(Sample.Proportion,times=Count),0.75))
dcast(ddply(combine_potential_class[which(combine_potential_class$Samples != 0),],.(State,Class),summarise,Median=median(rep(Sample.Proportion,times=Count))),Class~State,value.var="Median")
```

### TEs in each state in x% of samples

```{r potential percent}
ddply(combine_potential,.(State),summarise,In_100=sum(Count[which(Sample.Proportion == 1)]),
      In_90=sum(Count[which(Sample.Proportion >= 0.9)]),
      In_75=sum(Count[which(Sample.Proportion >= 0.75)]),
      In_50=sum(Count[which(Sample.Proportion >= 0.5)]),
      In_25=sum(Count[which(Sample.Proportion >= 0.25)]))
```

### TEs always in single state

```{r analysis Fig 2c}
always_TEs = c(setNames(lapply(chromHMM_states,function(x) chromHMM_TE_state[which(chromHMM_TE_state[[x]] == sample_counts["All","chromHMM"]),TE_coordinates]),chromHMM_states),
                  setNames(lapply(meth_states,function(x) TE_meth_average[which(TE_meth_average[[x]] == sample_counts["All","WGBS"]),TE_coordinates]),meth_states))
always_TEs$DNase = TE_DNase_peaks[which(TE_DNase_peaks$Samples == sample_counts["All","DNase"]),TE_coordinates]
always_TEs$H3K27ac = TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$Samples == sample_counts["All","H3K27ac"]),TE_coordinates]
always_TEs$RNA = RNA_TE[which(RNA_TE$Expressed_samples == sample_counts["All","RNA"]),TE_coordinates]
always_TEs = ldply(always_TEs)
colnames(always_TEs)[1] = "State"

always_TEs_noY = setNames(lapply(chromHMM_states,function(x) chromHMM_TE_state[which(chromHMM_TE_state$chromosome == "chrY" & chromHMM_TE_state[[x]] == sample_counts["chrY","chromHMM"]),TE_coordinates]),chromHMM_states)
always_TEs_noY$DNase = TE_DNase_peaks[which(TE_DNase_peaks$chromosome == "chrY" & TE_DNase_peaks$Samples == sample_counts["chrY","DNase"]),TE_coordinates]
always_TEs_noY$H3K27ac = TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$chromosome == "chrY" & TE_H3K27ac_peaks$Samples == sample_counts["chrY","H3K27ac"]),TE_coordinates]
always_TEs_noY$RNA = RNA_TE[which(RNA_TE$chromosome == "chrY" & RNA_TE$Expressed_samples == sample_counts["chrY","RNA"]),TE_coordinates]
always_TEs_noY = ldply(always_TEs_noY)
colnames(always_TEs_noY)[1] = "State"

## Summarise
ddply(always_TEs,.(State),summarise,Count=length(chromosome),Proportion=length(chromosome)/NUM_TE,Proportion_CpG=length(chromosome)/NUM_TE_WGBS)
ddply(always_TEs_noY,.(State),summarise,Count=length(chromosome),Proportion=length(chromosome)/NUM_TE,Proportion_CpG=length(chromosome)/NUM_TE_WGBS)

### Unique always in chromHMM state
dim(unique(always_TEs[which(always_TEs$State %in% chromHMM_states),2:8]))

## Genomic locations
always_TEs_location = merge(always_TEs,rmsk_TE[,c(TE_coordinates,"class_update","Length",cohorts)],by=TE_coordinates,all.x=TRUE)
ddply(always_TEs_location,.(State),function(x) apply(x[,cohorts],2,function(y) sum(!is.na(y))))

dim(always_TEs_location[which(always_TEs_location$State == "RNA" & is.na(always_TEs_location$exons) & is.na(always_TEs_location$introns) & is.na(always_TEs_location$promoters)),])

## Phylogeny
ddply(always_TEs_location,.(State,class_update),summarise,Count=length(chromosome))

always_TE_class = ddply(always_TEs_location[,c(TE_coordinates,"class_update","State")],.(State,class_update),summarise,Count=length(chromosome))
always_TE_class = merge(rmsk_TE_class[,c("class_update","Count")],always_TE_class,by="class_update",all.y=TRUE)
always_TE_class[is.na(always_TE_class)] = 0
colnames(always_TE_class)[c(2,4)] = c("Count.Class","Count")
always_TE_class = ddply(always_TE_class,.(State),transform,Count.State=sum(Count))
always_TE_class$Enrichment = log2((always_TE_class$Count/always_TE_class$Count.State)/(always_TE_class$Count.Class/NUM_TE))
always_TE_class = always_TE_class[order(always_TE_class$Enrichment),]
dim(always_TE_class[which(always_TE_class$Enrichment > 0),])
always_TE_class[which(always_TE_class$Enrichment > 5),]

always_TE_family = ddply(always_TEs_location[,c(TE_coordinates,"class_update","State")],.(State,class_update,family),summarise,Count=length(chromosome))
always_TE_family = merge(ddply(rmsk_TE_subfamily,.(family,class_update),summarise,Count=sum(Count)),always_TE_family,by=c("family","class_update"),all.y=TRUE)
always_TE_family[is.na(always_TE_family)] = 0
colnames(always_TE_family)[c(3,5)] = c("Count.Family","Count")
always_TE_family = ddply(always_TE_family,.(State),transform,Count.State=sum(Count))
always_TE_family$Enrichment = log2((always_TE_family$Count/always_TE_family$Count.State)/(always_TE_family$Count.Family/NUM_TE))
always_TE_family = always_TE_family[order(always_TE_family$Enrichment),]
dim(always_TE_family[which(always_TE_family$Enrichment > 0),])
always_TE_family[which(always_TE_family$Enrichment > 0.5 & always_TE_family$Count > 10),]

always_TE_subfamily = ddply(always_TEs_location[,c(TE_coordinates,"class_update","State")],.(State,class_update,family,subfamily),summarise,Count=length(chromosome))
always_TE_subfamily = merge(rmsk_TE_subfamily[,c("subfamily","family","class_update","Count")],always_TE_subfamily,by=c("subfamily","family","class_update"),all.y=TRUE)
always_TE_subfamily[is.na(always_TE_subfamily)] = 0
colnames(always_TE_subfamily)[c(4,6)] = c("Count.Subfamily","Count")
always_TE_subfamily = ddply(always_TE_subfamily,.(State),transform,Count.State=sum(Count))
always_TE_subfamily$Enrichment = log2((always_TE_subfamily$Count/always_TE_subfamily$Count.State)/(always_TE_subfamily$Count.Subfamily/NUM_TE))
dim(always_TE_subfamily[which(always_TE_subfamily$Enrichment > 0),])
always_TE_subfamily = always_TE_subfamily[order(always_TE_subfamily$State,always_TE_subfamily$Enrichment),]
always_TE_subfamily[which(always_TE_subfamily$Enrichment > 0.5 & always_TE_subfamily$Count > 10),]

## Print out TEs
write.table(always_TEs[,c(TE_coordinates,"State")],file="always_TEs.bed",sep='\t',row.names=FALSE,quote=FALSE,col.names = FALSE)

# 90% TEs
always_TEs_90 = c(setNames(lapply(chromHMM_states,function(x) chromHMM_TE_state[which(chromHMM_TE_state[[x]] >= sample_counts["All","chromHMM"]*.9),TE_coordinates]),chromHMM_states),
                  setNames(lapply(meth_states,function(x) TE_meth_average[which(TE_meth_average[[x]] >= sample_counts["All","WGBS"]*.9),TE_coordinates]),meth_states))
always_TEs_90$DNase = TE_DNase_peaks[which(TE_DNase_peaks$Samples >= sample_counts["All","DNase"]*.9),TE_coordinates]
always_TEs_90$H3K27ac = TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$Samples >= sample_counts["All","H3K27ac"]*.9),TE_coordinates]
always_TEs_90$RNA = RNA_TE[which(RNA_TE$Expressed_samples >= sample_counts["All","RNA"]*.9),TE_coordinates]
always_TEs_90 = ldply(always_TEs_90)
colnames(always_TEs_90)[1] = "State"

## Print out TEs
write.table(always_TEs_90[,c(TE_coordinates,"State")],file="always_TEs_90.bed",sep='\t',row.names=FALSE,quote=FALSE,col.names = FALSE)

# Proportion in each state
dcast(ddply(combine_potential_class[which(combine_potential_class$Sample.Proportion >= 0.9),],.(State,Class),summarise,TEs.Proportion=sum(Count)/sum(Total)),State~Class,value.var="TEs.Proportion")
```

### Supplementary Figure 4

```{r Figure S4, echo=FALSE}
#source("R_scripts/potential_class.R")

a = ggplot(combine_boxplot_class,aes(x=State,y=Proportion,fill=Class)) + geom_boxplot() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y = element_blank()) + coord_flip() + labs(y="Proportion of TEs in state") + scale_x_discrete(labels=rev(all_state_labels),limits=rev(levels(combine_boxplot_class$State))) + geom_point(data=combine_stats_class,aes(x=State,y=Proportion_ever,color=Class),position=position_dodge(width=0.5)) + scale_color_manual(values=class_colors,guide=FALSE)

state_order = ddply(combine_stats_class,"State",summarise,CV=sd(Samples_avg_all)/mean(Samples_avg_all), Range=max(Samples_avg_all)-min(Samples_avg_all))

b = ggplot(combine_potential_class[which(combine_potential_class$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states)) + facet_wrap(~Class,nrow=2) + geom_boxplot(width=0.2,outlier.size = 0.1)  + scale_y_continuous(breaks=pretty_breaks(n=3))

rmsk_TE_class_feature = melt(rmsk_TE_class[,c("class_update",cohorts[1:19])],id.vars="class_update")
colnames(rmsk_TE_class_feature) = c("Class","Cohort","Proportion")
rmsk_TE_class_feature = split_coding(rmsk_TE_class_feature)

d = ggplot(rmsk_TE_class_feature,aes(x=Class,y=Proportion,fill=Feature)) + geom_bar(stat="identity") + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.title=element_blank(),legend.position="right") +  ylab("Proportion of TEs\noverlapping feature") + scale_fill_manual(values=c(brewer.pal(8,"Spectral")),labels=genic_labels[c(8,1:7)]) + facet_wrap(~Coding,nrow=1,labeller=labeller(Coding=coding_labels)) + guides(fill = guide_legend(nrow = 4))

grid.arrange(a,b,d,nrow=2,layout_matrix=rbind(c(1,2),c(4)),heights=c(0.75,0.25),widths=c(0.45,0.55))
```

### Class potential analysis

```{r class potential analysis}
combine_stats_class
state_order
```

### Feature overlap and CpGs per class

```{r analysis Fig S9a}
# Any TEs not in a feature? Number of features each TE overlaps
table(apply(rmsk_TE[,cohorts[2:19]],1,function(x) sum(!is.na(x))))
table(apply(rmsk_TE[,c(features[2:3],"coding_exon",features[5:8])],1,function(x) sum(!is.na(x))))

# Number/proportion of TEs in each feature, overall and by class
feature_proportion = apply(rmsk_TE[,cohorts[1:19]],2,function(x) length(na.omit(x)))
feature_proportion/NUM_TE
rmsk_TE_class[,c("class_update",cohorts[1:20])]

# TEs with CpGs per class
rmsk_TE_class[,c("class_update","Count","TEs_wCpG","TEs_wCpG_per")]

# Median CpGs per TE by class
median(rmsk_TE$CpGs)
ddply(rmsk_TE,.(class_update),summarise,Median_CpGs=median(CpGs),Median_CpGs_wCpGs=median(CpGs[which(CpGs > 0)]))
```

### Supplementary Figure 5

```{r Figure S5, echo=FALSE, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/potential_promoter.R")
source("R_scripts/potential_exon.R")

# Promoters
a = ggplot(combine_boxplot_prom,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors[1:21],guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="Proportion of promoters in state") + coord_flip() + scale_x_discrete(limits=rev(states[1:21]),labels=all_state_labels[1:21]) + geom_point(data=combine_stats_prom,aes(x=State,y=Proportion_ever),color="red",size=2)

b = ggplot(combine_potential_prom[which(combine_potential_prom$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors[1:21],guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states[1:21]),labels=all_state_labels[1:21]) + geom_boxplot(width=0.2,outlier.size = 0.1)

# Exons
c = ggplot(RNA_exon_sample,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values="black",guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="Proportion of exons in state") + coord_flip() + scale_x_discrete(labels="RPKM > 1") + geom_point(data=RNA_potential_exon_stats,aes(x=State,y=Proportion_ever),color="red",size=3) + scale_y_continuous(limits=c(0,1))

d = ggplot(RNA_potential_exon_long[which(RNA_potential_exon_long$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values="black",guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples RPKM > 1") + scale_x_discrete(labels=all_state_labels[22]) + geom_boxplot(width=0.2,outlier.size = 0.1)

# No cancer/IMR90
e = ggplot(combine_boxplot_noCancer_IMR90,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="Proportion of TEs in state") + coord_flip() + scale_x_discrete(limits = rev(levels(combine_boxplot_noCancer_IMR90$State)),labels=rev(all_state_labels)) + geom_point(data=combine_stats_noCancer_IMR90,aes(x=State,y=Proportion_ever),color="red",size=2)

f = ggplot(combine_potential_noCancer[which(combine_potential_noCancer$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + geom_boxplot(width=0.2,outlier.size = 0.1)

grid.arrange(a,b,c,d,e,f,nrow = 3,layout_matrix=rbind(c(1,2),c(3,4),c(5,6)),heights=c(0.45,0.1,0.45),widths=c(0.6,0.4))
```

### Promoter statistics

```{r promoter analysis}
# Number of unique promoters
NUM_PROMOTER

combine_stats_prom
ddply(combine_boxplot_prom,.(State),summarise,Mean=mean(Proportion),Median=median(Proportion))

# Promoter CpG stats
promoter_CpG_count = read.table("WGBS/Refseq_promoters/refseq_promoter_unique_CpG_count.txt",sep='\t')
promoter_CpG_count$V5 = promoter_CpG_count$V5/2
dim(promoter_CpG_count)[1] # Number of promoters with CpGs
dim(promoter_CpG_count)[1]/NUM_PROMOTER # % of promoters with CpGs
ddply(promoter_CpG_count,.(),summarise,Min=min(V5),Median=median(V5),Mean=mean(V5),Max=max(V5))
mean(c(promoter_CpG_count$V5,rep(0,(NUM_PROMOTER-dim(promoter_CpG_count)[1]))))
median(c(promoter_CpG_count$V5,rep(0,(NUM_PROMOTER-dim(promoter_CpG_count)[1]))))

# % in state in 90% of samples
ddply(combine_potential_prom,.(State),summarise,Total=sum(Count[which(Sample.Proportion >= 0.9)])/NUM_PROMOTER)
```

### Exon statistics (expression potential control)

```{r exon control analysis, cache=TRUE}
# Number of unique exons
NUM_EXON

RNA_potential_exon_stats
mean(RNA_exon_sample$Proportion)
median(RNA_exon_sample$Proportion)

# Exon CpG stats
exon_CpG_count = read.table("WGBS/refseq_exons_unique_CpG_count.txt",sep='\t')
exon_CpG_count$V5 = exon_CpG_count$V5/2
dim(exon_CpG_count)[1] # Number of exons with CpGs
dim(exon_CpG_count)[1]/NUM_EXON # % of exons with CpGs
ddply(exon_CpG_count,.(),summarise,Min=min(V5),Median=median(V5),Mean=mean(V5),Max=max(V5))
mean(c(exon_CpG_count$V5,rep(0,(NUM_EXON-dim(exon_CpG_count)[1]))))
median(c(exon_CpG_count$V5,rep(0,(NUM_EXON-dim(exon_CpG_count)[1]))))

# Number of samples expressed
table(RNA_TE$Expressed_samples)
table(RNA_TE$Expressed_samples)/dim(RNA_TE)[1]

table(RNA_refseq_exon$Expressed_samples)
table(RNA_refseq_exon$Expressed_samples)/NUM_EXON

# % in state in 90% of samples
ddply(RNA_potential_exon_long,.(State),summarise,Total=sum(Count[which(Sample.Proportion >= 0.9)])/NUM_EXON)

# Maximum expression
median(RNA_TE$Max_expression)
median(RNA_TE[which(RNA_TE$Expressed_samples > 0),]$Max_expression)
max(RNA_TE$Max_expression)
tail(RNA_TE[order(RNA_TE$Max_expression),],n=20)

median(RNA_refseq_exon$Max_expression)
median(RNA_refseq_exon[which(RNA_refseq_exon$Expressed_samples > 0),]$Max_expression)
max(RNA_refseq_exon$Max_expression)
tail(RNA_refseq_exon[order(RNA_refseq_exon$Max_expression),],n=20)

cor.test(RNA_TE$Max_expression,RNA_TE$Expressed_samples,method="spearman")

# Max expression by class
merge(ddply(RNA_TE,.(class_update),summarise,Proportion_expressed=sum(Expressed_samples > 0)/length(Expressed_samples),Max_median=median(Max_expression)),
      ddply(RNA_TE[which(RNA_TE$Expressed_samples > 0),],.(class_update),summarise,Max_median_expressed=median(Max_expression)),by="class_update")

# Average TE vs. exon expression
TE_express = unlist(RNA_TE[,8:63])
median(TE_express)
median(TE_express[which(TE_express > 1)])
rm(TE_express)
exon_express = unlist(RNA_refseq_exon[,5:60])
median(exon_express)
median(exon_express[which(exon_express > 1)])
rm(exon_express)
```

### Cancer/IMR90 outlier analysis

```{r analysis Fig S5a}
combine_stats_noCancer_IMR90

# Average proportion of TEs in each state per sample, no IMR90
ddply(combine_boxplot,.(State),summarise,Mean=mean(Proportion),SD=sd(Proportion),Outlier=mean(Proportion)+2*sd(Proportion),Median=median(Proportion),IQR=IQR(Proportion))
ddply(combine_boxplot[which(combine_boxplot$Sample != "E017"),],.(State),summarise,Mean=mean(Proportion),SD=sd(Proportion),Median=median(Proportion),IQR=IQR(Proportion))

# Proportion of TEs in each state, IMR90
combine_boxplot[which(combine_boxplot$Sample == "E017"),]
```

### Figure 3

```{r Figure 3, echo=FALSE, fig.width=7, fig.height=6}
source("R_scripts/state_switching.R")

# chromHMM states
a = ggplot(chromHMM_TE_state,aes(x=States,y = ..density..,fill=class_update)) + geom_histogram(binwidth=1,color="grey") + xlab("Number of chromHMM states") + ylab("Proportion of TEs") + scale_fill_manual(values=class_colors,guide=FALSE) + facet_wrap(~class_update)   

# WGBS states
b = ggplot(TE_meth_average,aes(x=States,y = ..density..,fill=class_update)) + geom_histogram(binwidth=1,color="grey") + xlab("Number of methylation states") + ylab("Proportion of TEs") + scale_fill_manual(values=class_colors,guide=FALSE) + facet_wrap(~class_update) 

# State switching inter, chromHMM
c = ggplot(melt(as.matrix(state_switching_inter)),aes(x=Var2,y=Var1,fill=value/sample_counts["All","chromHMM"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(limits=rev(chromHMM_states)) + scale_fill_gradient(low="white",high="darkmagenta",limits=c(0,1),guide=FALSE) + coord_equal()

# State switching inter, WGBS
d = ggplot(melt(as.matrix(ss_inter_meth)),aes(x=Var2,y=Var1,fill=value/sample_counts["All","WGBS"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + scale_y_discrete(limits=rev(meth_states)) + scale_fill_gradient2(low="white",high="darkmagenta",limits=c(0,1),guide=FALSE) + coord_equal()

legend = get_legend(ggplot(melt(as.matrix(ss_inter_meth)),aes(x=Var2,y=Var1,fill=value/sample_counts["All","WGBS"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank(),legend.position = "bottom",legend.direction = "horizontal") + scale_fill_gradient2(low="white",high="darkmagenta",limits=c(0,1)))

grid.arrange(a,b,c,d,legend,nrow = 3, layout_matrix=rbind(c(1,2),c(3,4),c(5,5)),heights=c(0.3,0.6,0.1))
```

### Supplementary Figure 6

```{r Figure S6, echo=FALSE}
# chromHMM states intra
a = ggplot(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit"),],aes(x=Max_states_intra,y = ..density..,fill=class_update)) + geom_histogram(binwidth=1,color="grey") + ylab("Proportion of TEs") + xlab("Number of chromHMM states") + facet_wrap(~class_update) + scale_fill_manual(values=class_colors,guide=FALSE)

# State switching intra, chromHMM
b = ggplot(melt(as.matrix(state_switching_intra)),aes(x=Var2,y=Var1,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(labels=rev(chromHMM_states),limits=rev(levels(melt(as.matrix(state_switching_intra))$Var1))) + scale_fill_gradient(low="white",high="darkmagenta",limits=c(0,1)) + coord_equal()

# State switching inter, chromHMM, by class 
c = ggplot(melt(ss_inter_class,id.vars=c("Class","State")),aes(x=variable,y=State,fill=value/sample_counts["All","chromHMM"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5,size=7),legend.title = element_blank(),axis.text.y=element_text(size=7)) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(limits=rev(chromHMM_states)) + scale_fill_gradient(low="white",high="darkmagenta",guide=FALSE) + facet_wrap(~Class,nrow=1) + coord_equal()

# State switching inter, WGBS, by class
d = ggplot(melt(ss_inter_meth_class,id.vars=c("Class","State")),aes(x=variable,y=State,fill=value/sample_counts["All","WGBS"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.title = element_blank()) + scale_y_discrete(limits=rev(meth_states)) + scale_fill_gradient2(low="white",high="darkmagenta",limits=c(0,1),guide=FALSE) + facet_wrap(~Class,nrow=1) + coord_equal()

grid.arrange(a,b,c,d, nrow=3, layout_matrix=rbind(c(1,2),c(3,3),c(4,4)),heights=c(0.4,0.3,0.3))
```

### State switching analysis

```{r state switching analysis}
state_switching_intra
state_switching_inter

dim(chromHMM_TE_state[which(chromHMM_TE_state$`10_TssBiv` > 0 | chromHMM_TE_state$`11_BivFlnk` > 0),])
```

### Number of states per TE, overall and by class, chromHMM and WGBS

```{r analysis Fig S6a}
table(chromHMM_TE_state$States)
ddply(chromHMM_TE_state,.(),summarise,Mean=mean(States),SD=sd(States),Median=median(States),IQR=IQR(States))
ddply(chromHMM_TE_state,.(class_update),summarise,Mean=mean(States),SD=sd(States),Median=median(States),IQR=IQR(States))

states_outlier = chromHMM_TE_state[which(chromHMM_TE_state$States == 15),]
ddply(states_outlier,.(),summarise,Length=median(stop-start),States_intra=median(Tissues)/127,Max_states_intra=median(Max_states_intra))
states_outlier = merge(states_outlier,rmsk_TE[,c(TE_coordinates,cohorts)],by=TE_coordinates,all.x=TRUE)
states_outlier
write.table(states_outlier[,c(TE_coordinates,"class_update",cohorts)],file="States_15_TEs.bed",sep='\t',row.names=FALSE,quote=FALSE)

table(TE_meth_average$States)
ddply(TE_meth_average,.(),summarise,Mean=mean(States),SD=sd(States),Median=median(States),IQR=IQR(States))
ddply(TE_meth_average,.(class_update),summarise,Mean=mean(States),SD=sd(States),Median=median(States),IQR=IQR(States))

# Correlation of TE length with number of states ever
cor.test((chromHMM_TE_state$stop-chromHMM_TE_state$start),chromHMM_TE_state$States,method="spearman")
cor.test((TE_meth_average$stop-TE_meth_average$start),TE_meth_average$States,method="spearman")

# Correlation of number of states ever, max in one sample
by(chromHMM_TE_state,chromHMM_TE_state$Category,function(x) cor.test(x$Max_states_intra,x$States,method="spearman"))

# By chromosome
ddply(chromHMM_TE_state,.(chromosome,Category),summarise,Max_states_intra=median(Max_states_intra),States=median(States))
```

### Number of states per TE x sample

```{r states per TE-sample}
# chromHMM
test = read.table("chromHMM/rmsk_TEother_chromHMM_summit_state_counts.txt",sep='\t')
colnames(test) = c("States","Total")
## Remove erroneous Quies entry
test[1,2] = test[1,2]-1
test$Proportion = test$Total/sum(test$Total)
test
## Proportion of those with >1 state in only 2 states
test$Total[2]/sum(test$Total[2:15])

table(chromHMM_TE_state[,c("Max_states_intra","Category")])
max_intra_outlier = chromHMM_TE_state[which(chromHMM_TE_state$Max_states_intra > 7),]
ddply(max_intra_outlier,.(),summarise,Length=median(stop-start),Tissues=median(Tissues)/127)
max_intra_outlier = merge(max_intra_outlier,rmsk_TE[,c(TE_coordinates,cohorts)],by=TE_coordinates,all.x=TRUE)
max_intra_outlier
write.table(max_intra_outlier[,c(TE_coordinates,"class_update",cohorts)],file="States_intra_8_TEs.bed",sep='\t',row.names=FALSE,quote=FALSE)

# Number of TEs per category
table(chromHMM_TE_state$Category)/NUM_TE

# Number of TEs overlapping bin summits by class
table(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit"),]$class_update)

## Max states per TE
ddply(chromHMM_TE_state,.(Category),summarise,Mean=mean(Max_states_intra),SD=sd(Max_states_intra),Median=median(Max_states_intra),IQR(Max_states_intra))
ddply(chromHMM_TE_state,.(class_update,Category),summarise,Count=length(Max_states_intra),Mean=mean(Max_states_intra),SD=sd(Max_states_intra),Median=median(Max_states_intra),IQR(Max_states_intra))

## TEs with >1 max states in a sample
dim(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit" & chromHMM_TE_state$Max_states_intra > 1),])[1]
dim(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit" & chromHMM_TE_state$Max_states_intra > 1),])[1]/dim(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit"),])[1]

## Correlation of TE length with max in one sample
by(data=chromHMM_TE_state,chromHMM_TE_state$Category,function(x) cor.test(x$stop-x$start,x$Max_states_intra,method="spearman"))

# WGBS
test = read.table("WGBS/TE_CpG_state_counts.txt",sep='\t')
test$Proportion = test$V2/sum(test$V2)
test
sum(test$Proportion[2:4])

## Proportion of those with >2 states
sum(test$Proportion[3:4])/sum(test$Proportion)
```

### Number of states per TE, by state

TEs ever in each state versus the total number of states they are in across all samples. 

```{r state dynamics}
lapply(chromHMM_states,function(x) median(chromHMM_TE_state[which(chromHMM_TE_state[[x]] > 0),]$States))
lapply(meth_states,function(x) median(TE_meth_average[which(TE_meth_average[[x]] > 0),]$States))

ggplot(test,aes(x=States,fill=variable)) + geom_density(alpha=0.5,adjust=10) + scale_fill_manual(values=chromHMM_colors)
```

### DNase/H3K27ac peak analysis

```{r peak analysis, cache=TRUE}
# Number of DNase peaks per TE
d = table(unlist(TE_DNase_peaks[,8:60]))
d

# % of TE x sample with 1 peak
d[2]/sum(d[2:length(d)])

# Number of H3K27ac peaks per TE
e = table(unlist(TE_H3K27ac_peaks[,8:105]))
e

# % of TE x sample with 1 peak
e[2]/sum(e[2:length(e)])
```

### Supplementary Figure 7

```{r Figure S7, echo=FALSE}
source("R_scripts/compare_marks.R")

a = ggplot(by_chromHMM,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~chromHMM) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),strip.text.x = element_text(size = 5)) + ylab("Proportion of TE x sample") + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) 

b = ggplot(by_WGBS,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~WGBS,labeller=labeller(WGBS=setNames(c("Hypo","Inter","Hyper","Missing"),meth_states))) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank()) + ylab("Proportion of TE x sample") + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) 

c = ggplot(by_DNase,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~DNase) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),axis.title.y = element_blank()) + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) 

d = ggplot(by_H3K27ac,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~H3K27ac) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),axis.title.y = element_blank()) + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) 

e = ggplot(by_RNA,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=2.5) + facet_grid(Mark~RNA) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),axis.title.y = element_blank()) + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE) 

grid.arrange(a,b,c,d,e, nrow=2, layout_matrix=rbind(c(1,1,1,1),c(2,3,4,5)),widths=c(0.4,0.2,0.2,0.2))
```

### Compare mark analysis (get p-values and residuals)

```{r compare mark analysis}
# Number of samples with each combination of data
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$DNase)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$H3K27ac)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$DNase) & !is.na(metadata$H3K27ac)),])[1]
dim(metadata[which(!is.na(metadata$DNase) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$H3K27ac) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$DNase) & !is.na(metadata$H3K27ac) & !is.na(metadata$RNA)),])[1]

# Tables for each mark
ddply(WGBS_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(DNase_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(H3K27ac_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(RNA_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(compare_marks_all,.(chromHMM),summarise,TE_sample=sum(TE_sample),Proportion=sum(TE_sample)/sum(compare_marks_unique$TE_sample))

# Proportion of each state in other states
by_chromHMM
by_WGBS
by_DNase
by_H3K27ac
by_RNA

# Chi-squared for each mark combination (chromHMM not normalized)
table_list = list(xtabs(TE_sample~chromHMM+WGBS,compare_marks_all),
                  xtabs(TE_sample~chromHMM+DNase,compare_marks_all),
                  xtabs(TE_sample~chromHMM+H3K27ac,compare_marks_all),
                  xtabs(TE_sample~chromHMM+RNA,compare_marks_all),
                  xtabs(TE_sample~WGBS+DNase,compare_marks_unique),
                  xtabs(TE_sample~WGBS+H3K27ac,compare_marks_unique),
                  xtabs(TE_sample~WGBS+RNA,compare_marks_unique),
                  xtabs(TE_sample~DNase+H3K27ac,compare_marks_unique),
                  xtabs(TE_sample~DNase+RNA,compare_marks_unique),
                  xtabs(TE_sample~H3K27ac+RNA,compare_marks_unique))
chisq_list = lapply(table_list,function(x) chisq.test(x))
p.adjust(unlist(lapply(chisq_list,function(x) x$p.value)),method="bonf")
lapply(chisq_list,function(x) x$residuals)

# Proportion of TEs x sample in each chromHMM state also overlapping DNase peak and H3K27ac peak
test1 = ddply(compare_marks_all[which(!is.na(compare_marks_all$DNase) & !is.na(compare_marks_all$H3K27ac)),],.(chromHMM,DNase,H3K27ac),summarise,TE_sample=sum(TE_sample))
test2 = merge(ddply(test1,.(chromHMM),summarise,TE_sample=sum(TE_sample)),ddply(test1[which(test1$DNase == "True" & test1$H3K27ac == "True"),],.(chromHMM),summarise,TE_sample=sum(TE_sample)),by="chromHMM")
test2$TE_sample_proportion = test2$TE_sample.y/test2$TE_sample.x
test2[order(test2$TE_sample_proportion),]

# Proportion of TEs x sample in each WGBS state also overlapping DNase peak and H3K27ac peak
test1 = ddply(compare_marks_unique[which(!is.na(compare_marks_unique$DNase) & !is.na(compare_marks_unique$H3K27ac) & !is.na(compare_marks_unique$WGBS)),],.(WGBS,DNase,H3K27ac),summarise,TE_sample=sum(TE_sample))
test2 = merge(ddply(test1,.(WGBS),summarise,TE_sample=sum(TE_sample)),ddply(test1[which(test1$DNase == "True" & test1$H3K27ac == "True"),],.(WGBS),summarise,TE_sample=sum(TE_sample)),by="WGBS")
test2$TE_sample_proportion = test2$TE_sample.y/test2$TE_sample.x
test2[order(test2$TE_sample_proportion),]

# QC
## Total TE x sample x chromHMM states = 570405028 - matches total TEs in rmsk_TEother_chromHMM_summit_sorted.txt and sum(compare_marks_all$TE_sample)
## Total TE x sample = 562520596 - matches total unique TEs (121*NUM_TE + 6*NUM_TE_noY) and sum(compare_marks_unique$TE_sample)
## Total TE x sample correct for each metric
apply(sample_counts,2,function(x) (x[3]*NUM_TE) + ((x[1]-x[3])*NUM_TE_noY))
apply(compare_marks_unique,2,function(x) sum(compare_marks_unique[which(!is.na(x)),]$TE_sample))
```

## By sample

### By-sample analysis

```{r analysis Fig 4a, cache=TRUE, cache.lazy=FALSE}
# Stats
by_sample_stats = ddply(by_sample_all,~State,summarise, Range = max(Proportion)-min(Proportion), Median = median(Proportion), Mean = mean(Proportion), CV = sd(Proportion)/mean(Proportion))
by_sample_stats[order(by_sample_stats$Range),]

ddply(by_sample_all,~State,function(x) x[which.min(x$Proportion),])
ddply(by_sample_all,~State,function(x) x[which.max(x$Proportion),])

# Kruskal-Wallis test by sample classification
# MHC adjusted for every state (not category)
by_sample_KW = ddply(by_sample_all,~State,summarise,Group = p.adjust(unlist(kruskal.test(Proportion,Group))["p.value"],method="bonf"),
      Anatomy = p.adjust(unlist(kruskal.test(Proportion,Anatomy))["p.value"],method="bonf"),
      Type = p.adjust(unlist(kruskal.test(Proportion,Type))["p.value"],method="bonf"),
      Age = p.adjust(unlist(kruskal.test(Proportion,Age))["p.value"],method="bonf"),
      Germline = p.adjust(unlist(kruskal.test(Proportion,Germline))["p.value"],method="bonf"))
by_sample_KW = merge(by_sample_KW,ddply(droplevels(by_sample_all[which(!(by_sample_all$State %in% meth_states)),]),~State,summarise,Cancer = p.adjust(unlist(kruskal.test(Proportion,Cancer))["p.value"],method="bonf")),by="State",all.x=TRUE)
by_sample_KW = melt(by_sample_KW,id.var="State")
colnames(by_sample_KW)[2:3] = c("Category","Pvalue")
by_sample_KW[which(by_sample_KW$Pvalue < 0.05),]

# Permutation tests by sample classification
# MHC adjusted for category and state, not direction
permute_up = ddply(by_sample_all[,c("State","Sample","Proportion","Contribution")],.(State),function(y) permute_by_sample(y,"Proportion","+",unique(y$Contribution),"Proportion",-1))
permute_up = permute_up[which(permute_up$Samples.x > 0),]
permute_up$Padjust = p.adjust(permute_up$Pvalue,method="fdr")
permute_up[which(permute_up$Padjust < 0.1 & permute_up$Category == "Group" & permute_up$State %in% states[c(1:7,16:17,20:21)]),]
write.table(permute_up,file="sample_permute_up.txt",sep='\t',row.names=FALSE,quote=FALSE)

permute_down = ddply(by_sample_all[,c("State","Sample","Proportion","Contribution")],.(State),function(y) permute_by_sample(y,"Proportion","-",unique(y$Contribution),"Proportion",-1))
permute_down = permute_down[which(permute_down$Samples.y - permute_down$Samples.x > 0),]
permute_down$Padjust = p.adjust(permute_down$Pvalue,method="fdr")
permute_down[which(permute_down$Padjust < 0.1 & permute_down$Category == "Group" & permute_down$State %in% states[c(1:7,16:17,20:21)]),]
write.table(permute_down,file="sample_permute_down.txt",sep='\t',row.names=FALSE,quote=FALSE)

# Mean Group proportion by state 
test = aggregate(data=by_sample_all,Proportion~State+Group,mean)
test[order(test$State,test$Proportion),]
```

### Figure 4

```{r Figure 4, echo=FALSE}
#source("R_scripts/proportion.R")
#source("R_scripts/proportion_class.R")
source("R_scripts/RNA_coverage.R")

# add_stars(c(1:9,11:15,20:21,16,18),round(ddply(by_sample_all,.(State),summarise,Max=max(Proportion))$Max[c(1:9,11:15,20:21,16,18)]+0.05,2),size=6)
by_sample_KW$Stars = ifelse(by_sample_KW$Pvalue < 0.05,"*","")

a = ggplot(by_sample_all,aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.8),size=1) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x = element_blank(),axis.ticks.x = element_blank()) + ylab("Proportion of state in TEs") + scale_color_manual(values=group_colors,guide=FALSE) + geom_errorbar(aes(ymax=Contribution,ymin=Contribution),position=position_dodge(),color="black") + geom_hline(yintercept = MERGED_TE_WIDTH/GENOME_WIDTH,linetype="dashed") + geom_hline(yintercept = TE_CPGS/ALL_CPGS,linetype="dotdash") + scale_y_continuous(limits=c(0,0.85)) + scale_x_discrete(limits=states[c(1:15,20:21,16:19)],labels=all_state_labels) + geom_vline(xintercept=15.5,linetype="dashed",color="grey") +
geom_vline(xintercept=16.5,linetype="dashed",color="grey") + geom_vline(xintercept=17.5,linetype="dashed",color="grey")  + geom_text(data=by_sample_KW[which(by_sample_KW$Category == "Group"),],aes(x=State,y=0.8,label=Stars),color="red",size=7) 

b = ggplot(by_sample_class,aes(x=State,y=Proportion_state,color=Group)) + geom_point(position = position_jitterdodge(dodge.width = 0.8),size=1) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + ylab("Proportion of state in class") + scale_color_manual(values=group_colors,guide=FALSE) + scale_x_discrete(limits=states[c(1:15,20:21,16:19)]) + geom_errorbar(aes(ymax=Contribution,ymin=Contribution),position=position_dodge(),color="black") + facet_grid(rows=vars(class),scales="free") + geom_hline(data=by_sample_class[which(by_sample_class$State == "1_TssA" & by_sample_class$Sample == "E001"),],aes(yintercept=Bases_class/GENOME_WIDTH),linetype="dashed") + geom_hline(data=by_sample_class[which(by_sample_class$State == "Hypomethylated" & by_sample_class$Sample == "E003"),],aes(yintercept=Bases_class/ALL_CPGS),linetype="dotdash") + geom_vline(xintercept=15.5,linetype="dashed",color="grey") + geom_vline(xintercept=16.5,linetype="dashed",color="grey") + geom_vline(xintercept=17.5,linetype="dashed",color="grey") + scale_y_continuous(limits=c(0,0.6))

# Average expression of TEs vs. genome
c = ggplot(RNA_proportion,aes(x=1,y=Ratio,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.5),size=1) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x = element_blank(),axis.ticks.x = element_blank()) + scale_color_manual(values=group_colors,guide=FALSE) + ylab("TE:Genome\nexpression") + scale_y_continuous(limits=c(0,1)) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black") 

d = ggplot(class_RNA,aes(x=1,y=Ratio,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.5),size=1) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x = element_blank(),axis.ticks.x = element_blank()) + scale_color_manual(values=group_colors,guide=FALSE) + ylab("Class:Genome\nexpression") + ylim(0,1) + facet_wrap(~class_update,nrow=1) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black")

legend = get_legend(ggplot(by_sample_all,aes(x=State,y=Proportion,color=Group)) + geom_point(position = position_jitter(width = 0.3),size=2) + scale_color_manual(values=group_colors) + theme(legend.margin=margin(0,0,0,0)))

grid.arrange(a,b,c,d,legend,nrow=3,ncol=3,layout_matrix=rbind(c(1,1,5),c(2,2,5),c(3,4,5)),heights=c(0.3,0.55,0.15),widths=c(0.25,0.62,0.13))
```

### By sample and class analysis

```{r analysis Fig S11a, cache=TRUE, cache.lazy=FALSE}
by_sample_stats_class = ddply(na.omit(by_sample_class),.(State,class),summarise, Range = max(Proportion_state)-min(Proportion_state), Median = median(Proportion_state), Mean = mean(Proportion_state),CV = sd(Proportion_state)/mean(Proportion_state))
by_sample_stats_class[order(by_sample_stats_class$Range),]
by_sample_stats_class[order(by_sample_stats_class$CV),]

# Kruskal-Wallis
# MHC adjusted for state and class
by_sample_KW_class = ddply(by_sample_class,.(State,class),summarise,Group = p.adjust(unlist(kruskal.test(Proportion_state,Group))["p.value"],method="bonf"))
by_sample_KW_class[which(by_sample_KW_class$Group < 0.05),]

# Number of enriched samples by Group
# MHC adjusted for state and class
permute_up = ddply(by_sample_class[,c("class","State","Sample","Proportion_state","Contribution")],.(State,class),function(y) permute_by_sample(y,"Proportion_state","+",unique(y$Contribution),"Proportion_state",-1))
permute_up = permute_up[which(permute_up$Samples.x > 0),]
permute_up$Padjust = p.adjust(permute_up$Pvalue,method="fdr")
permute_up[which(permute_up$Padjust < 0.1 & permute_up$Category == "Group" & permute_up$State %in% states[c(1:7,16:17,20:21)]),]

permute_down = ddply(by_sample_class[,c("class","State","Sample","Proportion_state","Contribution")],.(State,class),function(y) permute_by_sample(y,"Proportion_state","-",unique(y$Contribution),"Proportion_state",-1))
permute_down = permute_down[which(permute_down$Samples.y - permute_down$Samples.x > 0),]
permute_down$Padjust = p.adjust(permute_down$Pvalue,method="fdr")
permute_down[which(permute_down$Padjust < 0.1 & permute_down$Category == "Group" & permute_down$State %in% states[c(1:7,16:17,20:21)]),]

test = ddply(by_sample_class,.(State,class,Group),summarise,Proportion_state=mean(Proportion_state),Contribution=unique(Contribution))
test[order(test$State,test$class,test$Proportion),]

# Enrichments over expectation
table(by_sample_class[which(by_sample_class$Enrichment > 0),c("State","class")])

by_sample_class[which(by_sample_class$class %in% c("DNA","LINE","SINE","LTR") & by_sample_class$State %in% states[c(1:3,6:7,16,20:21)] & by_sample_class$Enrichment > 0),c("Sample","class","State","Enrichment",sample_categories)]
```

### DNase/H3K27ac peak analysis (vs. bp)

```{r peak analysis pt2}
# % of DNase peaks overlapping TEs, overall
sum(DNase_stats$Summit_in_TE)/sum(DNase_stats$Peaks)

## By sample
ddply(DNase_stats,.(),summarise,Median=median(Summit_in_TE/Peaks),Mean=mean(Summit_in_TE/Peaks),SD=sd(Summit_in_TE/Peaks),Min=min(Summit_in_TE/Peaks),Max=max(Summit_in_TE/Peaks))
DNase_stats[which.min(DNase_stats$Summit_in_TE/DNase_stats$Peaks),]
DNase_stats[which.max(DNase_stats$Summit_in_TE/DNase_stats$Peaks),]

# % of H3K27ac peaks overlapping TEs, overall
sum(H3K27ac_stats$Summit_in_TE)/sum(H3K27ac_stats$Peaks)

## By sample
ddply(H3K27ac_stats,.(),summarise,Median=median(Summit_in_TE/Peaks),Mean=mean(Summit_in_TE/Peaks),SD=sd(Summit_in_TE/Peaks),Min=min(Summit_in_TE/Peaks),Max=max(Summit_in_TE/Peaks))
H3K27ac_stats[which.min(H3K27ac_stats$Summit_in_TE/H3K27ac_stats$Peaks),]
H3K27ac_stats[which.max(H3K27ac_stats$Summit_in_TE/H3K27ac_stats$Peaks),]
```

### Expression analysis

```{r ratio expression}
range(RNA_proportion$Ratio)
median(RNA_proportion$Ratio)
ddply(RNA_proportion,.(Group),summarise,Median=median(Ratio))

ddply(class_RNA,.(class_update),summarise,Median=median(Ratio))
dcast(ddply(class_RNA,.(Group,class_update),summarise,Median=median(Ratio)),class_update~Group,value.var="Median")
```

### Supplementary Figure 8

```{r Figure S8, echo=FALSE}
# Other sample classifications
by_sample_all_long = melt(by_sample_all[,c("State","Sample","Proportion","Contribution",sample_categories[2:6])],id.vars=c("State","Sample","Proportion","Contribution"))
colnames(by_sample_all_long)[5:6] = c("Category","Group")

a = ggplot(by_sample_all_long,aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6),size=1) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + ylab("Proportion of state in TEs") + scale_color_manual(values=c(anatomy_colors,type_colors,age_colors,cancer_colors,germline_colors),guide=FALSE) + geom_errorbar(aes(ymax=Contribution,ymin=Contribution),position=position_dodge(),color="black") + geom_hline(yintercept = MERGED_TE_WIDTH/GENOME_WIDTH,linetype="dashed") + geom_hline(yintercept = TE_CPGS/ALL_CPGS,linetype="dotdash") + scale_y_continuous(limits=c(0,0.85)) + scale_x_discrete(limits=states[c(1:15,20:21,16:19)],labels=all_state_labels) + geom_vline(xintercept=15.5,linetype="dashed",color="grey") + geom_vline(xintercept=16.5,linetype="dashed",color="grey") + geom_vline(xintercept=17.5,linetype="dashed",color="grey") + facet_wrap(~Category,nrow=5) + geom_text(data=by_sample_KW[which(by_sample_KW$Category != "Group"),],aes(x=State,y=0.8,label=Stars),color="red",size=7)

legend1 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Anatomy"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(anatomy_colors),name="Anatomy")) 

legend2 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Type"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(type_colors),name="Type") + guides(color = guide_legend(nrow = 2)))

legend3 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Age"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(age_colors),name="Age") + guides(color = guide_legend(nrow = 2)))

legend4 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Cancer"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(cancer_colors),name="Cancer"))

legend5 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Germline"),],aes(x=State,y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')) + scale_color_manual(values=c(germline_colors),name="Germlayer"))

grid.arrange(a,legend1,legend2,legend3,legend4,legend5,nrow=4,layout_matrix=rbind(c(1,1),c(2,2),c(3,4),c(6,5)),heights=c(0.8,0.1,0.05,0.05),widths=c(0.6,0.4))
```

## Subfamily enrichment

### Set thresholds 

```{r set enrichment thresholds}
THRESHOLD_IJK_MEMBER = 10
THRESHOLD_IK_MEMBER = 30
THRESHOLD_LOR = 1.5
THRESHOLD_PC = 0.01

THRESHOLD_IJK_BASE = 600
THRESHOLD_IK_BASE = 5000
THRESHOLD_IJK_CPG = 6
THRESHOLD_IK_CPG = 50
```

### Figure 5

```{r Figure 5, echo=FALSE, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/WGBS_subfamily_enrichment_TE.R")
source("R_scripts/subfamily_enrichment.R") #Requires proportion.R
source("R_scripts/subfamily_potential.R")
source("R_scripts/subfamily_PCA.R")

# Potential
a = ggplot(subfam_states[which(subfam_states$variable == "States"),],aes(x=value)) + geom_histogram(binwidth=1,fill="cornflowerblue") + ylab("Subfamilies") + xlab("States") + facet_wrap(~Metric,scales="free_x",nrow=1) + scale_y_continuous(limits=c(0,968)) + expand_limits(x=0)

# Clustering
family_shapes=setNames(c(17,rep(16,20),18,rep(16,23)),sort(as.vector(unique(rmsk_TE_subfamily$family))))

b = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group,shape=Age)) + geom_point(size=3) + scale_color_manual(values=group_colors,guide=FALSE) + scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=c("Cell line","Fetal","Non-fetal","Unknown"),guide=FALSE) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep="")) + theme(aspect.ratio = 1)

c = ggplot(subfamily_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family)) + geom_point(size=3) + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE) + labs(x=paste("PC1 (",round(subfamily_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(subfamily_chromHMM_pca$eigenvalues[2],1),"%)",sep="")) + theme(aspect.ratio = 1)

# Enrichment
d = ggplot(subfamily_state_sample_counts,aes(x=State,y=Sample.Proportion,color=class_update)) + geom_point(position=position_jitterdodge(jitter.width=0.5)) + coord_flip() + theme(axis.title.y=element_blank()) + ylab("Proportion of samples enriched LOR > 1.5") + scale_color_manual(values=class_colors,guide=FALSE) + geom_vline(xintercept=seq(1.5,20.5,1), colour='grey') + scale_x_discrete(limits = rev(states[c(1:15,20:21,16:19)]),labels=setNames(paste(subfamily_state_sample_counts_combine$State,"\n(n=",subfamily_state_sample_counts_combine$V1,")",sep=""),subfamily_state_sample_counts_combine$State))

# HOMER
homer = read.table("enrichment/homer/HOMER_LTR22A_7_Enh_enriched/knownResults.txt",sep='\t',header=TRUE,comment.char = "@")
homer_top5 = head(homer[order(homer$q.value..Benjamini.),],n=5)
homer_top5 = melt(homer_top5[,c(1,5,7,9)],id.vars=c("Motif.Name","q.value..Benjamini."))
homer_top5$PC = as.numeric(gsub("%","",homer_top5$value))

e = ggplot(homer_top5,aes(x=reorder(Motif.Name,`q.value..Benjamini.`,median),y=PC,fill=variable)) + geom_bar(stat="identity",position="dodge") + ylab("% TEs with motif") + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),axis.title.x = element_blank()) + scale_fill_discrete(guide=FALSE) + scale_x_discrete(labels=c("Nkx2.1","Nkx2.2","Tcf3","Fli1","HNF4a"))

grid.arrange(a,b,c,d,e,nrow=4,ncol=2,layout_matrix=rbind(c(1,1),c(2,4),c(3,4),c(5,4)),heights=c(0.2,0.3,0.3,0.2),widths=c(0.35,0.65))
```

### Supplementary Figure 9

```{r Figure S9, echo=FALSE}
ggplot(subfamily_state_potential,aes(x=Sample.Proportion,fill=State)) + geom_histogram() + facet_wrap(~State) + ylab("Subfamilies") + scale_fill_manual(values=all_state_colors,guide=FALSE) + xlab("Proportion of samples in state")
```

### Subfamily potential

```{r subfam potential}
# Number of subfamilies ever in each state and median samples in state
ddply(subfamily_state_potential,.(State),summarise,Subfamilies=sum(Samples > 0),Samples=median(Samples))

# Number of states per subfamily (across all samples, max in one sample)
by(subfam_states,subfam_states[,c("variable","Metric")],function(x) table(x$value))

ddply(subfam_states,.(variable,Metric),summarise,Median=median(value))

# Outliers
subfam_states[which(subfam_states$value < 11 & subfam_states$variable == "States" & subfam_states$Metric == "chromHMM"),]
subfam_states[which(subfam_states$value < 10 & subfam_states$variable == "Intra"),]

# Correlation with length
by(subfam_states,subfam_states[,c("variable","Metric")],function(x) cor.test(x$value,x$Total_length))
by(subfam_states,subfam_states[,c("variable","Metric")],function(x) cor.test(x$value,x$CpGs))

# Distribution of samples in state
ddply(subfamily_state_potential[which(subfamily_state_potential$State %in% chromHMM_states & subfamily_state_potential$Samples >= sample_counts["All","chromHMM"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
ddply(subfamily_state_potential[which(subfamily_state_potential$State %in% meth_states & subfamily_state_potential$Samples >= sample_counts["All","WGBS"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
ddply(subfamily_state_potential[which(subfamily_state_potential$State == "DNase" & subfamily_state_potential$Samples >= sample_counts["All","DNase"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
ddply(subfamily_state_potential[which(subfamily_state_potential$State == "H3K27ac" & subfamily_state_potential$Samples >= sample_counts["All","H3K27ac"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
```

### Supplementary Figure 10

```{r Figure S10, echo=FALSE}
## Samples
a = ggplot(sample_WGBS_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_WGBS_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_WGBS_pca$eigenvalues[2],1),"%)",sep=""))+ scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=c("Cell line","Fetal","Non-fetal","Unknown"),guide=FALSE) + theme(aspect.ratio = 1)

b = ggplot(sample_WGBS_pca$eigenvectors,aes(x=PC1,y=PC3,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_WGBS_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC3 (",round(sample_WGBS_pca$eigenvalues[3],1),"%)",sep=""))+ scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=c("Cell line","Fetal","Non-fetal","Unknown"),guide=FALSE) + theme(aspect.ratio = 1)

c = ggplot(sample_DNase_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_DNase_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_DNase_pca$eigenvalues[2],1),"%)",sep=""))+ scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=c("Cell line","Fetal","Non-fetal","Unknown"),guide=FALSE) + theme(aspect.ratio = 1)

d = ggplot(sample_H3K27ac_pca$eigenvectors,aes(x=PC1,y=PC2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_H3K27ac_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_H3K27ac_pca$eigenvalues[2],1),"%)",sep=""))+ scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=c("Cell line","Fetal","Non-fetal","Unknown"),guide=FALSE) + theme(aspect.ratio = 1)

## chromHMM other groupings
e = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Anatomy)) + geom_point() + scale_color_manual(values=anatomy_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

f = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Type)) + geom_point() + scale_color_manual(values=type_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

g = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Germline)) + geom_point() + scale_color_manual(values=germline_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

h = ggplot(sample_chromHMM_pca$eigenvectors,aes(x=PC1,y=PC2,color=Cancer)) + geom_point() + scale_color_manual(values=cancer_colors,guide=FALSE) + labs(x=paste("PC1 (",round(sample_chromHMM_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(sample_chromHMM_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

## Subfamilies
i = ggplot(subfamily_WGBS_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE)+ labs(x=paste("PC1 (",round(subfamily_WGBS_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(subfamily_WGBS_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

j = ggplot(subfamily_DNase_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE)+ labs(x=paste("PC1 (",round(subfamily_DNase_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(subfamily_DNase_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

k = ggplot(subfamily_H3K27ac_pca$eigenvectors,aes(x=PC1,y=PC2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE)+ labs(x=paste("PC1 (",round(subfamily_H3K27ac_pca$eigenvalues[1],1),"%)",sep=""),y=paste("PC2 (",round(subfamily_H3K27ac_pca$eigenvalues[2],1),"%)",sep=""))+ theme(aspect.ratio = 1)

grid.arrange(a,b,c,d,e,f,g,h,i,j,k,ncol=3) 
```

### PCA

```{r subfam clustering}
# Matrix dimensions
dim(sample_chromHMM_3D)
dim(subfamily_chromHMM_3D)
dim(sample_WGBS_3D)
dim(subfamily_WGBS_3D)
dim(sample_DNase_enrichment)
dim(subfamily_DNase_enrichment)
dim(sample_H3K27ac_enrichment)
dim(subfamily_H3K27ac_enrichment)
```

### Enrichment thresholds

```{r excluded subfamilies}
# Number of subfamilies excluded by enrichment thresholds
# chromHMM, all chromosomes
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count <= THRESHOLD_IK_MEMBER),])[1]
# chromHMM, no Y
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count > THRESHOLD_IK_MEMBER & rmsk_TE_subfamily$Count_noY <= THRESHOLD_IK_MEMBER),])[1]

# WGBS, all chromosomes (excluded only from WGBS)
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count_CpGs <= THRESHOLD_IK_MEMBER & rmsk_TE_subfamily$Count > THRESHOLD_IK_MEMBER),])[1]

rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count_noY <= THRESHOLD_IK_MEMBER | rmsk_TE_subfamily$Count_CpGs <= THRESHOLD_IK_MEMBER),c("subfamily","family","class_update","Count","Count_noY","Count_CpGs","Total_length","Total_length_noY","CpGs")]

# Subfamilies with no CpGs
rmsk_TE_subfamily[which(rmsk_TE_subfamily$CpGs == 0),c("subfamily","Count","Count_noY","Total_length","Total_length_noY","CpGs")]

# Number of enrichments excluded by enrichment thresholds
test = table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR & subfamily_state_sample_combined$Count > THRESHOLD_IK_MEMBER & subfamily_state_sample_combined$Members <= THRESHOLD_IJK_MEMBER),]$State)
test
sum(test)
```

### Number of enrichments by state and number of enriched subfamilies by state, overall and by class

```{r analysis Fig 5}
enrichment_table = merge(as.data.frame(table(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),]$State)),subfamily_state_sample_counts_combine,by.x="Var1",by.y="State")
colnames(enrichment_table)[1:3] = c("State","Enrichments","Subfamilies")
enrichment_table = enrichment_table[match(states[1:21],enrichment_table$State),]

sum(enrichment_table$Enrichments)
enrichment_table$Enrichments_pc = enrichment_table$Enrichments/sum(enrichment_table$Enrichments)
enrichment_table

# Number of subfamilies enriched at least once, overall and active states
length(unique(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),]$subfamily))
length(unique(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR & subfamily_state_sample_filter$State %in% c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")),]$subfamily))

# Enriched classes
test = dcast(aggregate(data=subfamily_state_sample_counts,V1~State+class_update,function(x) sum(x > 0)),class_update~State,value.var = "V1")
table(rmsk_TE_subfamily$class_update)/968
ddply(enrichment_table,.(State),function(x) as.numeric(x[4:9])/sum(as.numeric(x[4:9])))
apply(test[,2:22],2,function(x) log2((x/sum(x))/(table(rmsk_TE_subfamily$class_update)/968)))

# Enriched families
test = dcast(aggregate(data=subfamily_state_sample_counts,V1~State+family,function(x) sum(x > 0)),family~State,value.var = "V1")
table(rmsk_TE_subfamily$family)/968
test_log = as.data.frame(apply(test[,2:22],2,function(x) log2((x/sum(x))/(table(rmsk_TE_subfamily$family)/968))))
test_log$Subfamilies = as.numeric(table(rmsk_TE_subfamily$family)[rownames(test_log)])
lapply(colnames(test_log)[1:21],function(x) test_log[which(test_log[,x] > 0 & test_log$Subfamilies > 0),c(x,"Subfamilies")])
```

### Supplementary Figure 11

```{r Figure S11, echo=FALSE}
a = ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=Percent,fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Proportion of members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=states[1:21]) + scale_y_continuous(limits=c(0,1))

b = ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=log10(Members),fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Log10 number of members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=states[1:21]) + scale_y_continuous(limits=c(0,6))

c = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],aes(x=State,y=Percent,fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Proportion of members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=states[1:21]) + scale_y_continuous(limits=c(0,1))

d = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],aes(x=State,y=log10(Members),fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Log10 number of members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=states[1:21]) + scale_y_continuous(limits=c(0,6))

grid.arrange(a,c,b,d,nrow=2,layout_matrix=cbind(c(1,2),c(3,4)))
```

### Average members in a state for subfamilies in a state

```{r bp members percent in state}
# All instances
ddply(subfamily_state_sample_combined,~State,summarize,Percent_median=median(na.omit(Percent)), Members_median=median(na.omit(Members)),Percent_min=min(na.omit(Percent)), Members_min=min(na.omit(Members)), Percent_max=max(na.omit(Percent)), Members_max=max(na.omit(Members)))   

# Enriched
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],~State,summarise,Percent_median=median(Percent), Members_median=median(Members),Percent_min=min(Percent), Members_min=min(Members), Percent_max=max(Percent), Members_max=max(Members))

# >1%
ddply(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],~State,summarise, Percent_median=median(Percent),Members_median=median(Members),Percent_min=min(Percent), Members_min=min(Members), Percent_max=max(Percent), Members_max=max(Members))

# Max by state (outliers)
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],.(State),function(x) x[which.max(x$Percent),])
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],.(State),function(x) tail(x[order(x$Percent),c("subfamily","family","class_update","State","Sample",sample_categories,"Members","Percent")],n=10))
```

### WGBS subfamily state stats

```{r analysis Fig S13c}
TE_meth_subfamily_percent = subfamily_CpG_meth[,c("subfamily","family","class_update","State","Sample","Count","Percent")]

# Average, max, and range of proportion of subfamily in methylation state plus all members in a state in a sample
TE_meth_subfamily_stats = ddply(TE_meth_subfamily_percent,.(subfamily,State),summarise,Median=median(Percent),Range=max(Percent)-min(Percent),Max=max(Percent),All_members=sum(Percent == 1))

table(TE_meth_subfamily_stats[which(TE_meth_subfamily_stats$All_members > 0),]$State)
sort(table(droplevels(subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == "Hypermethylated" & subfamily_state_sample_combined$Percent == 1),]$subfamily)))
merge(TE_meth_subfamily_stats[which(TE_meth_subfamily_stats$All_members > 0),],rmsk_TE_subfamily[,c("subfamily","Count","Count_CpGs","Total_length","Total_length_noY","CpGs")],by="subfamily")

# Average proportion in state, with/without IMR90
ddply(TE_meth_subfamily_stats,.(State),summarise,Median=mean(Median),Range=mean(Range),Max=mean(Max))
```

### Supplementary Figure 12

GO BP results downloaded in tsv format from GREAT. 

```{r Figure S12, echo=FALSE}
source("R_scripts/TE_correlation_subfamily.R") # Requires subfamily_enrichment.R
source("R_scripts/subfamily_correlate.R") # Requires TE_correlation_subfamily.R

# State vs. state
a = ggplot(correlate_subfam_state[which(correlate_subfam_state$State %in% states & correlate_subfam_state$State2 %in% states & correlate_subfam_state$class_update == "All"),],aes(x=State,y=State2,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.ticks.x=element_blank(),axis.text=element_text(size=7)) + scale_x_discrete(limits=states[c(1:14,16:17,19:21)]) + scale_y_discrete(limits=rev(states[c(1:14,16:17,19:21)]))

b = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% c(measure_metrics_subfam,features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfamily$class_update == "LTR"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=30,hjust=1,vjust=1),axis.text=element_text(size=7)) + scale_y_discrete(limits=rev(states[c(1:14,16:17,19:21)])) + scale_x_discrete(limits=c(measure_metrics_subfam,features[1:3],"coding_exon",features[5:8],"Vista_enhancers"),labels=c(measure_labels,genic_labels)) + geom_vline(color="black",xintercept=9.5)

# GREAT GO BP
great_names = gsub(".txt","",list.files("enrichment/great/"))
great = lapply(great_names,function(x) read.table(paste("enrichment/great/",x,".txt",sep=""),sep='\t'))
names(great) = great_names
great = ldply(great)
colnames(great) = c("set","Term Name","Binom Rank","Binom Raw P-Value","Binom FDR Q-Val","Binom Fold Enrichment","Binom Observed Region Hits","Binom Region Set Coverage",
                    "Hyper Rank","Hyper FDR Q-Val","Hyper Fold Enrichment","Hyper Observed Gene Hits","Hyper Total Genes","Hyper Gene Set Coverage")

c = ggplot(great[which(great$set == "AmnSINE1_7_Enh"),],aes(x=reorder(`Term Name`,`Binom Raw P-Value`,median),y=log10(`Binom Raw P-Value`))) + geom_bar(stat="identity") + coord_flip() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),plot.title = element_text(hjust = 1),axis.text=element_text(size=7)) 

d = ggplot(great[which(great$set == "AmnSINE2_DNase"),],aes(x=reorder(`Term Name`,`Binom Raw P-Value`,median),y=log10(`Binom Raw P-Value`))) + geom_bar(stat="identity") + coord_flip() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),plot.title = element_text(hjust = 1),axis.text=element_text(size=7))

e = ggplot(great[which(great$set == "MER121_7_Enh"),],aes(x=reorder(`Term Name`,`Binom Raw P-Value`,median),y=log10(`Binom Raw P-Value`))) + geom_bar(stat="identity") + coord_flip() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),plot.title = element_text(hjust = 1),axis.text=element_text(size=7)) 

f = ggplot(great[which(great$set == "MER131_DNase"),],aes(x=reorder(`Term Name`,`Binom Raw P-Value`,median),y=log10(`Binom Raw P-Value`))) + geom_bar(stat="identity") + coord_flip() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),plot.title = element_text(hjust = 1),axis.text=element_text(size=7))

grid.arrange(a,b,c,d,e,f,layout_matrix=rbind(c(1,2),c(3,3),c(4,4),c(5,5),c(6,6)),heights=c(0.25,0.2,0.2,0.2,0.15))
```

### Subfamily enrichment correlation

Not MHC corrected

```{r analysis Fig S15a}
# Metric vs. metric
## Metrics
correlate_subfam_feature[which(abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% measure_metrics_subfam & correlate_subfam_feature$Metric %in% measure_metrics_subfam & correlate_subfam_feature$class_update == "All" & correlate_subfam_feature$p.value < 0.05),]

## Features
correlate_subfam_feature[which(abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfam_feature$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfam_feature$class_update == "All" & correlate_subfam_feature$State != correlate_subfam_feature$Metric & correlate_subfam_feature$p.value < 0.05),]

## Chromosomes
correlate_subfam_feature[which(abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% standard_chromosomes & correlate_subfam_feature$Metric %in% standard_chromosomes & correlate_subfam_feature$class_update == "All" & correlate_subfam_feature$State != correlate_subfam_feature$Metric & correlate_subfam_feature$p.value < 0.05),]

# State vs. state
correlate_subfam_state[which(abs(correlate_subfam_state$estimate.rho) > 0.3 & correlate_subfam_state$State %in% states & correlate_subfam_state$State2 %in% states & correlate_subfam_state$class_update == "All" & correlate_subfam_state$State != correlate_subfam_state$State2 & correlate_subfam_state$p.value < 0.05),]

## Range of correlation for active regulatory states, subfamily enrichment
median(correlate_subfam_state[which(correlate_subfam_state$class_update == "All" & correlate_subfam_state$State %in% states[c(1:2,7,16:17,20:21)] & correlate_subfam_state$State2 %in% states[c(1:2,7,16:17,20:21)] & correlate_subfam_state$State != correlate_subfam_state$State2),]$estimate.rho)
median(correlate_subfam_state[which(correlate_subfam_state$class_update == "LTR" & correlate_subfam_state$State %in% states[c(1:2,7,16:17,20:21)] & correlate_subfam_state$State2 %in% states[c(1:2,7,16:17,20:21)] & correlate_subfam_state$State != correlate_subfam_state$State2),]$estimate.rho)

# State vs. metric
## Metrics
correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% measure_metrics_subfam & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.05),]

correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% measure_metrics_subfam & correlate_subfamily$class_update != "All"  & correlate_subfamily$p.value < 0.05),]

## Features
correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.05),]

correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfamily$class_update != "All" & correlate_subfamily$p.value < 0.05),]

## Chromosomes
correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% standard_chromosomes & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.05),]

correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states[c(1:7,16:17,20:21)] & correlate_subfamily$Metric %in% standard_chromosomes & correlate_subfamily$class_update != "All" & correlate_subfamily$p.value < 0.05),]
```

### Summary of subfamily metrics

```{r subfamily metrics}
# Subfamily metrics, overall and by class
apply(rmsk_TE_subfamily[,measure_metrics_subfam],2,median)
aggregate(data=rmsk_TE_subfamily[,c("class_update",measure_metrics_subfam)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)

# Subfamily feature overlap by class
apply(rmsk_TE_subfamily[,cohorts],2,median)
aggregate(data=rmsk_TE_subfamily[,c("class_update",cohorts)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)

# Subfamily chromosome location by class
apply(rmsk_TE_subfamily[,standard_chromosomes[1:24]],2,median)
aggregate(data=rmsk_TE_subfamily[,c("class_update",standard_chromosomes[1:24])],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)

# CGate subfamilies
length(cgate_subfams)

# Age vs. number of members
correlate_subfam_feature[which(correlate_subfam_feature$Metric == "Count" & correlate_subfam_feature$State == "Age"),]

# Subfamilies with very low mappability
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),])[1]
table(droplevels(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),]$family))
rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),c("subfamily","family","class_update","Count","Mappability")]
```

### SINE subfamilies

```{r SINE subfamilies}
# SINE subfamily active regulatory state enrichments
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)] & subfamily_state_sample_counts$class_update == "SINE"),]
test
ddply(test,.(subfamily),summarise,sum(V1))

test = rmsk_TE_subfamily[which(rmsk_TE_subfamily$class_update == "SINE"),]

# Age of SINE subfamilies
test[order(test$Age),c("subfamily","family","class_update","Count","Age")]

# CpG density of SINE subfamiles
test[order(test$CpGs_per_kbp),c("subfamily","family","class_update","Count","Age","CpGs_per_kbp")]
median(test[which(test$family == "Alu"),]$CpGs_per_kbp)

# Overlap with exons
test[order(test$exons),c("subfamily","family","class_update","Count","Age","exons","exons_pc","exons_nc")]
test[order(test$exons_pc),c("subfamily","family","class_update","Count","Age","exons","exons_pc","exons_nc")]
test[order(test$exons_nc),c("subfamily","family","class_update","Count","Age","exons","exons_pc","exons_nc")]

# Overlap with VISTA enhancers
test[order(test$Vista_enhancers),c("subfamily","family","class_update","Count","Age","Vista_enhancers")]

# Chromosomes
test[order(test$chr2),c("subfamily","family","class_update","Count","Age",standard_chromosomes[1:24])]
```

### VISTA enhancers

```{r vista enhancers analysis}
dim(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),])
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),c(TE_coordinates,"Vista_enhancers")]$subfamily)))
sort(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),c(TE_coordinates,"Vista_enhancers")]$subfamily)))

correlate_subfamily[which(correlate_subfamily$Metric == "Vista_enhancers" & correlate_subfamily$State %in% states & abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$p.value < 0.05),] # Not MHC corrected
```

### LTR subfamilies

```{r LTR subfamilies}
# LTR subfamily active regulatory state enrichments
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)] & subfamily_state_sample_counts$class_update == "LTR"),]
test[order(test$V1),]
ddply(test,.(subfamily),summarise,sum(V1))
```

### LINE subfamilies

```{r LINE subfamilies}
test = ddply(subfamily_state_sample_counts[which(subfamily_state_sample_counts$class_update == "LINE" & subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)]),],.(subfamily),summarise,Count=sum(V1))
test
sum(test$Count)

# Age of LINE subfamilies
test = rmsk_TE_subfamily[which(rmsk_TE_subfamily$class_update == "LINE"),]
test[order(test$Age),c("subfamily","family","class_update","Count","Age","CpGs_per_kbp")]
test[order(test$CpGs_per_kbp),c("subfamily","family","class_update","Count","Age","CpGs_per_kbp")]
```

### Genic subfamilies

```{r genic subfamilies}
# 4_Tx
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State == "4_Tx"),]
test[order(test$V1),]

# 6_EnhG
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State == "6_EnhG"),]
test[order(test$V1),]

# Intron overlap
test = subfamily_state_sample_counts[which(subfamily_state_sample_counts$State %in% c("3_TxFlnk","4_Tx","5_TxWk","6_EnhG") & subfamily_state_sample_counts$V1 > 0),]
test = merge(test,rmsk_TE_subfamily[,c("subfamily","Count","introns","introns_pc","introns_nc")],by="subfamily")
head(test[order(test$introns),],n=40)
```

### MER57E3

```{r mer57e3}
subfamily_state_sample_counts[which(subfamily_state_sample_counts$subfamily == "MER57E3"),]
MER57E3 = rmsk_TE[which(rmsk_TE$subfamily == "MER57E3"),]
dim(MER57E3)[1]
table(MER57E3$chromosome)
apply(MER57E3[,cohorts],2,function(x) dim(MER57E3[which(!is.na(x)),])[1])
```

### Young subfamilies

```{r young subfamilies}
quantile(rmsk_TE_subfamily$Age,0.05)
young_subfamilies = as.vector(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Age < quantile(rmsk_TE_subfamily$Age,0.05)),]$subfamily)
rmsk_TE_subfamily[which(rmsk_TE_subfamily$subfamily %in% young_subfamilies),c("subfamily","family","class_update","Age")]

young_subfamilies_enrichment = dcast(subfamily_state_sample_counts[which(subfamily_state_sample_counts$subfamily %in% young_subfamilies & subfamily_state_sample_counts$Sample.Proportion > 0),],subfamily~State,value.var="Sample.Proportion")
rownames(young_subfamilies_enrichment) = young_subfamilies_enrichment$subfamily
young_subfamilies_enrichment = young_subfamilies_enrichment[,2:16]

sort(apply(young_subfamilies_enrichment,2,function(x) sum(!is.na(x))))
sort(apply(young_subfamilies_enrichment,1,function(x) sum(!is.na(x))))

young_subfamilies_enrichment
```

### Subfamily factors

```{r subfamily correlation}
subfamily_measure_long = melt(rmsk_TE_subfamily_measure[,c("class_update","family","subfamily",cohorts,standard_chromosomes[1:24],measure_metrics_subfam,states[1:21])],id.vars=c("class_update","family","subfamily",cohorts,standard_chromosomes[1:24],measure_metrics_subfam))
subfamily_measure_long = melt(subfamily_measure_long,id.vars=c("class_update","family","subfamily","variable","value"))
colnames(subfamily_measure_long)[4:7] = c("State","Samples","Metric","Value")
```

### Subfamily heatmaps

```{r heatmaps, echo=FALSE, eval=FALSE}
# Legends for metadata categories
legend_group = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Group)) + geom_tile() + scale_fill_manual(values=group_colors) + theme(legend.direction = "horizontal"))
legend_anatomy = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Anatomy)) + geom_tile() + scale_fill_manual(values=anatomy_colors) + theme(legend.direction = "horizontal"))
legend_type = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Type)) + geom_tile() + scale_fill_manual(values=brewer.pal(5,"Greens")) + theme(legend.direction = "horizontal"))
legend_age = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Age)) + geom_tile() + scale_fill_manual(values=brewer.pal(4,"YlOrRd"),labels=c("Cell line","Fetal","Non-fetal","Unknown")) + theme(legend.direction = "horizontal"))
legend_cancer = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Cancer)) + geom_tile() + scale_fill_manual(values=c("white","red")) + theme(legend.direction = "horizontal"))
legend_germline = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Germline)) + geom_tile() + scale_fill_manual(values=brewer.pal(6,"Dark2")) + theme(legend.direction = "horizontal"))

a = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","1_TssA",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
b = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","2_TssAFlnk",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
c = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","3_TxFlnk",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
d = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","6_EnhG",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
e = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","7_Enh",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
f = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","4_Tx",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
g = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","5_TxWk",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
h = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Hypomethylated",1,sample_counts["All","WGBS"],THRESHOLD_LOR)
i = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Intermediate",1,sample_counts["All","WGBS"],THRESHOLD_LOR)
j = plot_binary_heatmap(subfamily_state_sample_filter,"DNase","DNase",1,sample_counts["All","DNase"],THRESHOLD_LOR)
k = plot_binary_heatmap(subfamily_state_sample_filter,"H3K27ac","H3K27ac",1,sample_counts["All","H3K27ac"],THRESHOLD_LOR)

legend_class = get_legend(ggplot(rmsk_TE_subfamily,aes(x=subfamily,y=class_update,fill=class_update)) + geom_tile() + scale_fill_manual(values=class_colors))

# Arrange column legends
grid.arrange(legend_group,legend_anatomy,legend_age,legend_cancer,legend_germline,legend_type,ncol=2,widths=c(0.6,0.4),layout_matrix=cbind(c(1,2,6),c(3,4,5)))
```

### Subfamily permutation for candidates
### RERUN when updated

Looks for subfamily x state x sample group combinations enriched in more samples than expectation. Filtered on enrichment threshold (LOR>1.5) and 10 members in the state; excludes subfamilies with <30 members (accounts for state/chrY). Runs all metadata categories at once.

MHC on only subfamilies x states x categories/groupings ever enriched in state, active states only, categories with 1+ enrichments. 

```{r subfam permutation, cache=TRUE, cache.lazy=FALSE, eval=FALSE}
# Subfamilies ever enriched in a state
subfamily_state_ever = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)]),c("subfamily","State")]
permutation_input = merge(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Count > THRESHOLD_IK_MEMBER),],subfamily_state_ever,by=c("subfamily","State"))

# Interesting candidates
permute_subfam = ddply(permutation_input[,c("subfamily","State","Sample","Enrichment","Members")],.(subfamily,State),function(x) permute_by_sample(matrix=x,metric="Enrichment",direction="+",threshold=THRESHOLD_LOR,filtering="Members",threshold2=THRESHOLD_IJK_MEMBER))
permute_subfam = permute_subfam[which(permute_subfam$Samples.x > 0),]
permute_subfam$Padjust = p.adjust(permute_subfam$Pvalue,method="fdr")
dim(permute_subfam[which(permute_subfam$Padjust < 0.05),])

permute_subfam_wide = dcast(aggregate(data = permute_subfam[which(permute_subfam$Padjust < 0.05),],Grouping~State+subfamily+Category, paste, collapse = ','),State+subfamily~Category,value.var="Grouping")
permute_subfam_wide = merge(permute_subfam_wide,subfamily_state_sample_counts,by=c("subfamily","State"))[,c(1,10,9,2:8,11)]
write.table(permute_subfam_wide,file="enrichment/state_subfamily_clusters_enriched.txt",sep='\t',row.names = FALSE,quote = FALSE)
```

```{r subfam permutation pc, cache=TRUE, cache.lazy=FALSE, eval=FALSE}
# Subfamilies ever enriched in a state
subfamily_state_ever_pc = subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0 & subfamily_state_sample_counts_pc$State %in% states[c(1:7,16:17,20:21)]),c("subfamily","State")]
permutation_input_pc = merge(subfamily_state_sample_combined,subfamily_state_ever_pc,by=c("subfamily","State"))

# Interesting candidates
permute_subfam_pc = ddply(permutation_input_pc[,c("subfamily","State","Sample","Length_percent_jk","Members")],.(subfamily,State),function(x) permute_by_sample(x,"Length_percent_jk","+",THRESHOLD_PC,"Members",0))
permute_subfam_pc = permute_subfam_pc[which(permute_subfam_pc$Samples.x > 0),]
permute_subfam_pc$Padjust = p.adjust(permute_subfam_pc$Pvalue,method="fdr")
dim(permute_subfam_pc[which(permute_subfam_pc$Padjust < 0.05),])

permute_subfam_pc_wide = dcast(aggregate(data = permute_subfam_pc[which(permute_subfam_pc$Padjust < 0.05),],Grouping~State+subfamily+Category, paste, collapse = ','),State+subfamily~Category,value.var="Grouping")
permute_subfam_pc_wide = merge(permute_subfam_pc_wide,subfamily_state_sample_counts_pc,by=c("subfamily","State"))[,c(1,10,9,2:8,11)]
write.table(permute_subfam_pc_wide,file="enrichment/state_subfamily_clusters_pc.txt",sep='\t',row.names = FALSE,quote = FALSE)
```

### Print bedfiles for candidates (enriched and never)

```{r print bedfiles, eval=FALSE}
source('R_scripts/TE_correlation.R')

candidates = read.table(file="enrichment/enrichment_candidates.txt",sep='\t') #Results of candidate Excel analysis
colnames(candidates) = c("State","subfamily")

# Subfamilies enriched in many samples
candidates_many = subfamily_state_sample_counts[which(subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)] & subfamily_state_sample_counts$Sample.Proportion > 0.25),]
candidates_many[order(candidates_many$State,candidates_many$Sample.Proportion),]

# Combine enriched and many lists
candidates_all = unique(rbind(candidates,candidates_many[,c("State","subfamily")]))

# Enriched bedfiles
candidates_individual = apply(candidates_all,1,function(x) get_subfamily_enriched(x[2],x[1]))
names(candidates_individual) = candidates_all$State
candidates_individual = ldply(candidates_individual)
colnames(candidates_individual)[1] = "State"

# Never in state bedfiles
apply(candidates_all,1,function(x) write.table(rmsk_TE_measure[which(rmsk_TE_measure$subfamily == x[2] & rmsk_TE_measure[[x[1]]] == 0),c(colnames(rmsk_TE_measure)[1:4],x[1],"strand")],sep='\t',row.names=FALSE,col.names=FALSE,quote=FALSE,file=paste("enrichment/bedfiles/",x[2],"_never_",x[1],".bed",sep="")))

write.table(candidates_individual,file="enrichment/enrichment_candidates_indv.txt",sep='\t',row.names = FALSE,quote = FALSE)
```

### Explore candidates

```{r explore candidates, eval=FALSE}
# Overlap of enriched TEs with genic features
candidates_features = merge(candidates_individual,rmsk_TE[,c("chromosome","start","stop","subfamily","strand","family","class_update",features[c(1:3,5:8)],"coding_exon","Vista_enhancers")],all.x=TRUE,by=c("chromosome","start","stop","subfamily","strand"))

candidates_features_summary = ddply(candidates_features,.(subfamily,class_update,State),function(x) apply(x[,c(features[c(1:3,5:8)],"coding_exon","Vista_enhancers")],2,function(x) sum(!is.na(x))/length(x)))

# Percent overlap of each candidate subfamily with RefSeq features (only members in state when enriched)
write.table(candidates_features_summary,file="enrichment/candidates_features_summary.txt",sep='\t',row.names=FALSE,quote=FALSE)
```

### Supplementary Figure 13

```{r Figure S13, echo=FALSE}
ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_combined$State %in% c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")),],aes(x=Sample,y=reorder(subfamily,rmsk_TE_subfamily[match(subfamily,rmsk_TE_subfamily$subfamily),]$Total_length),fill=Length_percent_jk)) + geom_tile(linetype="blank") + theme(axis.ticks.x=element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank()) + scale_fill_gradient(name="% state",low="lightblue",high="darkblue",limits=c(0.009,0.12)) +  facet_wrap(~State)
```

### Subfamily stats, >1%

```{r analysis Fig S13a}
# Max % in subfamily by state
ddply(subfamily_state_sample_combined,~State,function(x) x[which.max(x$Length_percent_jk),])

# Max % of state in subfamily without enrichment
ddply(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment <= THRESHOLD_LOR),],~State,function(x) x[which.max(x$Length_percent_jk),])

# Number of instances by state and number of >1% subfamilies by state, overall and by class
pc_table = merge(as.data.frame(table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),]$State)),subfamily_state_sample_counts_pc_combine,by.x="Var1",by.y="State")
colnames(pc_table)[1:3] = c("State","Instances","Subfamilies")
sum(pc_table$Instances)

# Instances >1% bp of state, also enriched
pc_table = merge(pc_table,as.data.frame(table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR),]$State)),by.x="State",by.y="Var1",all=TRUE)
colnames(pc_table)[10] = "Enriched_Instances"
pc_table$Percent = pc_table$Enriched_Instances/pc_table$Instances
pc_table

# Number of subfamilies >1% across all states
length(unique(subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0),]$subfamily))
length(unique(subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0 & subfamily_state_sample_counts_pc$State %in% states[c(1:7,16:17,20:21)]),]$subfamily))

# Subfamilies with >1% of bases in genome
GENOME_WIDTH*THRESHOLD_PC
rmsk_TE_subfamily[which(rmsk_TE_subfamily$Total_length > GENOME_WIDTH*THRESHOLD_PC),c("subfamily","Count","Total_length","Total_length_noY","CpGs")]

# Subfamilies with >1% of CpGs in genome
ALL_CPGS*THRESHOLD_PC
rmsk_TE_subfamily[which(rmsk_TE_subfamily$CpGs > ALL_CPGS*THRESHOLD_PC),c("subfamily","Count","Total_length","Total_length_noY","CpGs")]

# Largest subfamilies vs. >1% of state
test = merge(dcast(subfamily_state_sample_counts_pc,subfamily+family+class_update~State,value.var = "V1"),rmsk_TE_subfamily[,c("subfamily","Count","Total_length","CpGs","Count_CpGs")],by="subfamily",all=TRUE)

test = test[order(test$Total_length,decreasing=TRUE),]
rownames(test) <- NULL
test$Order = rownames(test)
lapply(c(chromHMM_states,"DNase","H3K27ac"),function(x) if(pc_table[which(pc_table$State == x),]$Enriched_Instances > 0) {join(test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs","Count_CpGs",x)],aggregate(data=subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x & subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],Enrichment~subfamily,function(y) sum(y > THRESHOLD_LOR)),by="subfamily",type="left")} else{test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)]})

test = test[order(test$CpGs,decreasing=TRUE),]
rownames(test) <- NULL
test$Order = rownames(test)
lapply(meth_states,function(x) if(pc_table[which(pc_table$State == x),]$Enriched_Instances > 0) {join(test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs","Count_CpGs",x)],aggregate(data=subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x & subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],Enrichment~subfamily,function(y) sum(y > THRESHOLD_LOR)),by="subfamily",type="left")} else{test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)]})
```

## Factors influencing potential

### Figure 6

```{r Figure 6, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 1","Figure 2","Figure 5")}
source("R_scripts/TE_correlation.R")
source("R_scripts/state_metric_correlation.R")

# SINE/Alu age vs. TE metrics and number of samples in methylation states
rmsk_TE_measure$cpgIsland_binary = as.numeric(ifelse(rmsk_TE_measure$cpgIsland == 0,0,1))

rmsk_TE_measure_SINE = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$class_update == "SINE"),c(TE_coordinates,"class_update",measure_metrics,meth_states,"cpgIsland_binary")])
rmsk_TE_measure_SINE$Length = rmsk_TE_measure_SINE$Length/1000

a = ggplot(rmsk_TE_measure_SINE,aes(x=JC_distance,y=..scaled..)) + geom_density(fill=class_colors["SINE"]) + theme(axis.title.x = element_blank(),axis.title.y = element_blank())

b = ggplot(melt(rmsk_TE_measure_SINE,id.vars=c(TE_coordinates,"class_update","Length","mappability","JC_distance","CpGs","CpGs_per_length","cpgIsland_binary")),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=meth_colors, guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank()) 

c = ggplot(melt(rmsk_TE_measure_SINE,id.vars=c(TE_coordinates,"class_update","JC_distance","CpGs",meth_states,"cpgIsland_binary"))) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + geom_smooth(data=rmsk_TE_measure_SINE,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + theme(axis.title.y = element_blank())

rmsk_TE_measure_Alu = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$family == "Alu"),c(TE_coordinates,"class_update",measure_metrics,meth_states,"cpgIsland_binary")])
rmsk_TE_measure_Alu$Length = rmsk_TE_measure_Alu$Length/1000

d = ggplot(rmsk_TE_measure_Alu,aes(x=JC_distance,y=..scaled..)) + geom_density(fill="lightblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank())

e = ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","Length","mappability","JC_distance","CpGs","CpGs_per_length","cpgIsland_binary")),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=meth_colors, guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank()) 

f = ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","JC_distance","CpGs",meth_states,"cpgIsland_binary"))) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + theme(axis.title.y = element_blank()) 

legend1 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","Length","mappability","JC_distance","CpGs","CpGs_per_length","cpgIsland_binary")),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=meth_colors) + theme(legend.title = element_blank(),legend.direction = "horizontal"))
  
legend2 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","JC_distance","CpGs",meth_states,"cpgIsland_binary"))) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),labels=c("Length (kbp)","Mappability","CpGs/bp")) + ylab("Property") + theme(legend.title = element_blank(),legend.direction = "horizontal") + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")))

grid.arrange(a,b,c,d,e,f,legend1,legend2,nrow=5,layout_matrix=rbind(c(1,4),c(2,5),c(3,6),c(7,7),c(8,8)),heights=c(0.2,0.3,0.4,0.05,0.05))
```

### Figure 6 analysis

```{r analysis Fig 6}
# Significant correlations between number of samples in state and metric (not MHC corrected)
correlate_TE_state[which(abs(correlate_TE_state$estimate.rho) > 0.3 & correlate_TE_state$State %in% states),]

# % of SINE that is Alu
dim(rmsk_TE[which(rmsk_TE$family == "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SINE"),])[1]

# Average age for older SINE, Alu
median(rmsk_TE[which(rmsk_TE$class == "SINE"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$class == "SINE"),]$JC_distance)
median(rmsk_TE[which(rmsk_TE$class == "SINE" & rmsk_TE$family != "Alu"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$class == "SINE" & rmsk_TE$family != "Alu"),]$JC_distance)
median(rmsk_TE[which(rmsk_TE$family == "Alu"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$family == "Alu"),]$JC_distance)

# Proportion of older SINE, Alu overlapping CpG islands
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SINE" & rmsk_TE$family != "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SINE" & rmsk_TE$family != "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SINE" & rmsk_TE$family != "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$family == "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$family == "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$family == "Alu"),])[1]

# Other models
test = lapply(c(meth_states,"Length","mappability","CpGs_per_length"),function(x) gam(rmsk_TE_measure_SINE[[x]]~rmsk_TE_measure_SINE$JC_distance))
lapply(test,function(x) summary(x))
test = lapply(c(meth_states,"Length","mappability","CpGs_per_length"),function(x) gam(rmsk_TE_measure_Alu[[x]]~rmsk_TE_measure_Alu$JC_distance))
lapply(test,function(x) summary(x))

# Overlap with CpG islands by age (logistic regression) (all, SINE, Alu)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure,family="binomial")
summary(test)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure_SINE,family="binomial")
summary(test)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure_Alu,family="binomial")
summary(test)

# Age vs. overlap with CpG islands
ddply(rmsk_TE_measure,.(cpgIsland_binary),summarise,Age=median(JC_distance))
ddply(rmsk_TE_measure[which(rmsk_TE_measure$class_update == "SINE"),],.(cpgIsland_binary),summarise,Age=median(JC_distance))
ddply(rmsk_TE_measure[which(rmsk_TE_measure$family == "Alu"),],.(cpgIsland_binary),summarise,Age=median(JC_distance))
```

### Supplementary Figure 14

Feature-state patterns roughly the same across class (not shown). Stronger for protein-coding than non-coding (not shown).

Legends: a) Spearman's rho, f) Average % incrase

```{r Figure S14, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 6")}
source("R_scripts/chromosome_potential.R")
source("R_scripts/feature_overlap_state.R") #Requires potential.R

# Epigenetic state by chromosome
a = ggplot(chrom_state_total,aes(x=chromosome,y=Bases,fill=State)) + geom_bar(stat="identity",position="fill") + scale_fill_manual(values = chromHMM_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + ylab("Proportion")

b = ggplot(chr_state_WGBS_total,aes(x=chromosome,y=CpGs,fill=State)) + geom_bar(position="fill",stat="identity") + scale_fill_manual(values=meth_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)) + ylab("Proportion")

c = ggplot(chr_state_DNase_total,aes(x=size/1000000,y=Peaks/1000000,color=chromosome,label=chromosome)) + geom_text(size=3) + scale_color_discrete(guide=FALSE) + geom_smooth(method='lm',se=FALSE,color="black",linetype="dashed") + xlab("Chromosome size (Mbp)") + ylab("Peaks (millions)")  + scale_y_continuous(limits=c(0,1.5))

d = ggplot(chr_state_H3K27ac_total,aes(x=size/1000000,y=Peaks/1000000,color=chromosome,label=chromosome)) + geom_text(size=3) + scale_color_discrete(guide=FALSE) + geom_smooth(method='lm',se=FALSE,color="black",linetype="dashed") + xlab("Chromosome size (Mbp)") + ylab("Peaks (millions)") + scale_y_continuous(limits=c(0,1.5))

# Chromosome vs. states
e = ggplot(chromosome_potential,aes(x=State,y=Zscore,color=chromosome,label=chromosome)) + geom_point() + geom_text(angle=45,hjust=0,vjust=0,size=3) + scale_color_discrete(guide=FALSE) + scale_x_discrete(limits=rev(states),labels=all_state_labels) + scale_y_continuous(limits=c(-5,5)) + theme(axis.title.y = element_blank(),panel.grid.major.y=element_line(color="grey")) + ylab("Z-score, % of samples in state") + geom_hline(yintercept = 1,color="grey") + geom_hline(yintercept = 2,color="grey") + geom_hline(yintercept = -1,color="grey") + geom_hline(yintercept = -2,color="grey") + geom_hline(yintercept = -3,color="grey") + geom_hline(yintercept = 3,color="grey") + geom_hline(yintercept = -4,color="grey") + geom_hline(yintercept = 4,color="grey") + coord_flip() + geom_hline(yintercept = -5,color="grey") + geom_hline(yintercept = 5,color="grey")

# Feature vs. state
f = ggplot(na.omit(feature_state_mean[which(feature_state_mean$Coding == "All"),]),aes(x=State,y=Feature,fill=Enrichment*100)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(name="% increase",low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-100,4000)) + scale_y_discrete(limits=rev(levels(feature_state_mean$Feature)),labels=genic_labels) + scale_x_discrete(labels=all_state_labels) 

grid.arrange(a,b,c,d,e,f,layout_matrix=rbind(c(1,3),c(2,4),c(5,6)),heights=c(0.3,0.3,0.4))
```

### Supplementary Figure 15

```{r Figure S15, echo=FALSE}
#source("R_scripts/state_metric_correlation.R")

a = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background=element_rect(fill="lightgrey"),aspect.ratio = 1) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs"))

b = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background=element_rect(fill="lightgrey"),aspect.ratio = 1) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs")) + facet_wrap(~class_update)

# Metric vs. state
c = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update == "All" & correlate_TE_state$State %in% meth_states),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background=element_rect(fill="lightgrey"),aspect.ratio = 1) + scale_y_discrete(limits=rev(meth_states)) + scale_x_discrete(limits=measure_metrics[c(1,4:5,2:3)],labels=c("Length","CpGs","CpGs/bp","Mappability","Age"))

d = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update != "All" & correlate_TE_state$State %in% meth_states),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),panel.background=element_rect(fill="lightgrey"),aspect.ratio = 1) + scale_y_discrete(limits=rev(meth_states)) + scale_x_discrete(limits=measure_metrics[c(1,4:5,2:3)],labels=c("Length","CpGs","CpGs/bp","Mappability","Age")) + facet_wrap(~class_update) 

grid.arrange(a,b,c,d,nrow=2)
```

### TE feature stats

```{r analysis Fig S15, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 6","Figure S15")}
# States
apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y > 0),states],na.rm=TRUE)) - apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y == 0),states],na.rm=TRUE))

dcast(feature_state_mean[which(feature_state_mean$Coding == "All"),],Feature~State,value.var = "Enrichment")
feature_state_mean_class[which(feature_state_mean_class$Coding == "All"),]

# Proportion of TEs ever in state overlapping each feature
apply(rmsk_TE_measure[,states],2,function(x) apply(rmsk_TE_measure[,cohorts],2,function(y) dim(rmsk_TE_measure[which(x > 0 & y > 0),])[1]/dim(rmsk_TE_measure[which(x > 0),])[1]))
```

### TE feature vs. metric

```{r, eval=FALSE}
source("R_scripts/feature_overlap_metric.R") #Requires TE_correlation.R
feature_metric_mean[order(feature_metric_mean$Enrichment),]
feature_metric_mean_class[order(feature_metric_mean_class$Enrichment),]

apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y > 0),measure_metrics],na.rm=TRUE)) - apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y == 0),measure_metrics],na.rm=TRUE))
```

### Chromosome stats

```{r chromosome stats}
# Proportion of TEs relative to genome length
hg19_TEs[order(hg19_TEs$Prop_TEs/hg19_TEs$Prop),]

# Z-scores
chromosome_potential[order(chromosome_potential$Zscore),]

# Proportion of TEs in each state on each chromosome
apply(rmsk_TE_measure[,states],2,function(x) table(rmsk_TE_measure[which(x > 0),]$chromosome)/dim(rmsk_TE_measure[which(x > 0),])[1])[standard_chromosomes[1:24],]
```

### Potential without chrY test

```{r eval=FALSE}
# Potential without chrY
test = chromHMM_TE_state[which(chromHMM_TE_state$chromosome != "chrY"),]
test.a = sample_distribution(test,c(8:22),sample_counts["All","chromHMM"])
test.b = cumulative_distribution(test,c(8:22),sample_counts["All","chromHMM"])
test.b$State = factor(rep(chromHMM_states,each=sample_counts["All","chromHMM"]),levels=chromHMM_states)
test.c = potential_stats(test.a,15,sample_counts["All","chromHMM"])
test.c$State = factor(chromHMM_states,levels=chromHMM_states)
chromHMM_TE_state_dist_stats-test.c

# Average samples in state for chrY TEs
sort(colMeans(chromHMM_TE_state[which(chromHMM_TE_state$chromosome == "chrY"),8:22]))

# % of TEs overlapping DNase/H3K27ac on chrY, expressed RPKM > 1
ADD
```

## Mouse

```{r mouse scripts, cache=TRUE, cache.lazy=FALSE}
#source("R_scripts/proportion.R")
#source("R_scripts/subfamily_enrichment.R")
#source("R_scripts/WGBS_subfamily_enrichment_TE.R")
source("R_scripts/human_mouse_orthologs.R")
source("R_scripts/human_mouse_ortholog_WGBS.R") 
source("R_scripts/human_mouse_ortholog_chromHMM.R")
source("R_scripts/human_mouse_subfamily.R")

# TE ortholog analysis
rmsk_TE_class$Mouse_orthologs = table(unique(human_mouse_orthologs_mm10[,c(hg19_coordinates,"class_update")])$class_update)[as.vector(rmsk_TE_class$class_update)]
rmsk_TE_class = merge(rmsk_TE_class,ddply(rmsk_TE_subfamily[which(rmsk_TE_subfamily$subfamily %in% mm10_rmsk_TE$subfamily),],~class_update,summarize,
                                          Mouse_ortholog_subfamily = length(subfamily)),by="class_update",all.x=TRUE)
rmsk_TE_class[which(is.na(rmsk_TE_class$Mouse_ortholog_subfamily)),]$Mouse_ortholog_subfamily = 0

# Table of tissue-specific TEs
hg19_orthologs_specific_active = as.data.frame(t(sapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) tissue_matrix(x,hg19_mm10_chromHMM_active))))
hg19_orthologs_specific_active$Tissue = rownames(hg19_orthologs_specific_active)
hg19_orthologs_specific_active$Other = hg19_orthologs_specific_active$All - hg19_orthologs_specific_active$Specific - hg19_orthologs_specific_active$On - hg19_orthologs_specific_active$Off
hg19_orthologs_specific_active = melt(hg19_orthologs_specific_active[,c("Specific","On","Off","Other","Tissue")],id.vars="Tissue")
colnames(hg19_orthologs_specific_active)[2:3] = c("Subset","Count")
hg19_orthologs_specific_active$State = rep("Active",dim(hg19_orthologs_specific_active)[1])

hg19_orthologs_specific_promoter = as.data.frame(t(sapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) tissue_matrix(x,hg19_mm10_chromHMM_promoter))))
hg19_orthologs_specific_promoter$Tissue = rownames(hg19_orthologs_specific_promoter)
hg19_orthologs_specific_promoter$Other = hg19_orthologs_specific_promoter$All - hg19_orthologs_specific_promoter$Specific - hg19_orthologs_specific_promoter$On - hg19_orthologs_specific_promoter$Off
hg19_orthologs_specific_promoter = melt(hg19_orthologs_specific_promoter[,c("Specific","On","Off","Other","Tissue")],id.vars="Tissue")
colnames(hg19_orthologs_specific_promoter)[2:3] = c("Subset","Count")
hg19_orthologs_specific_promoter$State = rep("Promoter",dim(hg19_orthologs_specific_promoter)[1])

hg19_orthologs_specific = rbind(hg19_orthologs_specific_active,hg19_orthologs_specific_promoter)
hg19_orthologs_specific$Subset = factor(hg19_orthologs_specific$Subset,levels=c("Off","Specific","Other","On"))
```

### Figure 7

A) Axes are individual TE methylation level - all 13 samples pairs. Excludes TEs missing methylation in either sample of a pair.
B) Y-axis is % of human TEs in each state in that state in mouse; colors as before - all 13 sample pairs
C) Axes are subfamily % hypomethylated (of those with CpGs)

```{r Figure 7, echo=FALSE}
a = ggplot(human_mouse_orthologs_WGBS_paired,aes(x=Human_methylation,y=Mouse_methylation)) + stat_bin2d(bins=50) + scale_fill_gradient(low="white",high="darkblue",name="TEs") + xlab("Human") + ylab("Mouse") + coord_equal() + theme(legend.position = "bottom",legend.direction="horizontal")

state_freq = ddply(human_mouse_orthologs_WGBS_paired,.(Human_state_WGBS),summarise,Entries=length(Human_state_WGBS)/dim(human_mouse_orthologs_WGBS_paired)[1])

b = ggplot(human_mouse_orthologs_WGBS_paired,aes(x=Human_state_WGBS,fill=Mouse_state_WGBS)) + geom_bar(position="fill") + scale_fill_manual(values=meth_colors,guide=FALSE) + xlab("Human") + ylab("Mouse") + scale_x_discrete(labels=as.vector(apply(state_freq,1,function(x) paste(x[1],"\n(",round(as.numeric(x[2])*100,0),"%)",sep="")))) + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),axis.title.x=element_blank())

state_freq_chromHMM = ddply(human_mouse_orthologs_chromHMM_paired,.(Human_state_chromHMM),summarise,Entries=length(Human_state_chromHMM)/dim(human_mouse_orthologs_chromHMM_paired)[1])

c = ggplot(human_mouse_orthologs_chromHMM_paired,aes(x=Human_state_chromHMM,fill=Mouse_state_chromHMM)) + geom_bar(position="fill") + xlab("Human") + ylab("Mouse") + theme(axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),legend.direction="horizontal",legend.position="top",legend.title=element_blank(),axis.title.x = element_blank()) + scale_x_discrete(labels=as.vector(apply(state_freq_chromHMM,1,function(x) paste(x[1],"\n(",round(as.numeric(x[2])*100,1),"%)",sep="")))) + guides(fill = guide_legend(nrow=3)) + scale_fill_manual(values=mouse_chromHMM_colors)

d = ggplot(hg19_orthologs_specific,aes(x=factor(1),y=Count,fill=Subset)) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank(),legend.title = element_blank(),strip.text.x = element_text(size = 8)) + facet_grid(State~Tissue,labeller=labeller(Tissue = setNames(c("Brain","Instestine","Stomach","Heart","Lung"),unique(hg19_orthologs_specific$Tissue)))) + scale_fill_discrete(labels=c("<2 samples","Tissue-specific","2-8 samples","> 8 samples")) + guides(fill = guide_legend(nrow = 4))

e = ggplot(hg19_mm10_TE_WGBS_subfamily_hypo_paired,aes(x=Human_hypo,y=Mouse_hypo,color=Anatomy,shape=Age)) + geom_point() + scale_color_manual(values=anatomy_colors,guide=FALSE) + xlab("Human") + ylab("Mouse") + geom_abline(slope=1,linetype="dotted") + geom_smooth(aes(linetype=Age),method='lm',fullrange=TRUE,alpha=0.1) + xlim(0,0.4) + ylim(0,0.4) + guides(color=FALSE,shape=FALSE,linetype=FALSE) + coord_equal()

f = ggplot(hg19_mm10_chromHMM_subfamily[which(hg19_mm10_chromHMM_subfamily$Human_state_chromHMM == "1_TssA" & hg19_mm10_chromHMM_subfamily$Mouse_state_chromHMM == "TssA"),],aes(x=Percent.x,y=Percent.y,color=Anatomy,shape=Age)) + geom_point() + scale_color_manual(values=anatomy_colors,guide=FALSE) + xlab("Human") + ylab("Mouse") + geom_abline(slope=1,linetype="dotted") + geom_smooth(aes(linetype=Age),method='lm',fullrange=TRUE,alpha=0.1) + xlim(0,0.05) + ylim(0,0.05) + guides(color=FALSE,shape=FALSE,linetype=FALSE) + coord_equal()

g = ggplot(hg19_mm10_chromHMM_subfamily_active,aes(x=Percent.x,y=Percent.y,color=Anatomy,shape=Age)) + geom_point() + scale_color_manual(values=anatomy_colors,guide=FALSE) + xlab("Human") + ylab("Mouse") + geom_abline(slope=1,linetype="dotted") + geom_smooth(aes(linetype=Age),method='lm',fullrange=TRUE,alpha=0.1) + xlim(0,0.4) + ylim(0,0.4) + guides(color=FALSE,shape=FALSE,linetype=FALSE) + coord_equal()

legend = get_legend(ggplot(hg19_mm10_chromHMM_subfamily_active,aes(x=Percent.x,y=Percent.y,color=Anatomy,shape=Age)) + geom_point() + scale_color_manual(values=anatomy_colors) + geom_smooth(aes(linetype=Age),method='lm',fullrange=TRUE,alpha=0.1) + xlab("Human") + ylab("Mouse") + theme(legend.position = "bottom",legend.direction = "horizontal",legend.margin=margin(0,0,0,0),legend.key.size = unit(1,'mm')))

grid.arrange(a,b,c,d,e,f,g,legend,nrow=4,layout_matrix=rbind(c(1,3,3),c(2,4,4),c(5,6,7),c(8,8,8)),heights=c(0.35,0.3,0.3,0.05))
```

### Supplementary Figure 16

```{r Figure S16, echo=FALSE}
a = ggplot(melt(rmsk_TE_class[,c("class_update","Count","Mouse_orthologs")],id.var="class_update"),aes(x=variable,y=value,fill=class_update)) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(name="Class",values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank())
  
b = ggplot(melt(rmsk_TE_class[,c("class_update","Subfamilies","Mouse_ortholog_subfamily")],id.var="class_update"),aes(x=variable,y=value,fill=class_update)) + geom_bar(stat="identity",position="fill",width=0.8) + coord_polar(theta="y") + scale_fill_manual(name="Class",values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank())

c = ggplot(human_mouse_spearman,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + ylab("Spearman correlation") + theme(axis.title.x = element_blank()) + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse")) + scale_y_continuous(limits=c(0,1)) + scale_fill_discrete(guide=FALSE) + geom_segment(x=1,xend=1.25,y=0.8,yend=0.8,color="red") + annotate("text",x=1.125,y=0.81,label="*",color="red",size=5)

d = ggplot(human_mouse_spearman_subfamily,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + ylab("Spearman correlation") + theme(axis.title.x = element_blank()) + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse")) + scale_y_continuous(limits=c(0,1)) + scale_fill_discrete(guide=FALSE) + geom_segment(x=1,xend=1.25,y=0.87,yend=0.87,color="red") + annotate("text",x=1.125,y=0.88,label="*",color="red",size=5)

e = ggplot(human_mouse_spearman_subfamily_TssA,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + ylab("Spearman correlation") + theme(axis.title.x = element_blank()) + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse")) + scale_y_continuous(limits=c(0,1)) + scale_fill_discrete(guide=FALSE) + geom_segment(x=1.75,xend=1.99,y=0.85,yend=0.85,color="red") + geom_segment(x=2.01,xend=2.25,y=0.85,yend=0.85,color="red") + geom_segment(x=1.75,xend=2.25,y=0.9,yend=0.9,color="red") + annotate("text",x=1.875,y=0.86,label="*",color="red",size=5) + annotate("text",x=2,y=0.91,label="*",color="red",size=5) + annotate("text",x=2.125,y=0.86,label="*",color="red",size=5)

f = ggplot(human_mouse_spearman_subfamily_active,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + ylab("Spearman correlation") + theme(axis.title.x = element_blank()) + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse")) + scale_y_continuous(limits=c(0,1)) + scale_fill_discrete(guide=FALSE) + geom_segment(x=1,xend=1.25,y=0.9,yend=0.9,color="red") + annotate("text",x=1.125,y=0.91,label="*",color="red",size=5) + geom_segment(x=1.75,xend=1.99,y=0.85,yend=0.85,color="red") + geom_segment(x=2.01,xend=2.25,y=0.85,yend=0.85,color="red") + geom_segment(x=1.75,xend=2.25,y=0.9,yend=0.9,color="red") + annotate("text",x=1.875,y=0.86,label="*",color="red",size=5) + annotate("text",x=2,y=0.91,label="*",color="red",size=5) + annotate("text",x=2.125,y=0.86,label="*",color="red",size=5)

legend = get_legend(ggplot(human_mouse_spearman_subfamily_active,aes(x=Test,y=Correlation,fill=Tissue.Age)) + geom_boxplot() + theme(legend.direction = "horizontal") + scale_fill_discrete(name="Tissue/Age",labels=c("Same/Same","Same/Different","Different/Same","Different/Different")))

grid.arrange(a,b,c,d,e,f,legend,nrow=4,layout_matrix=rbind(c(1,2),c(3,4),c(5,6),c(7,7)),heights=c(0.3,0.3,0.3,0.1))
```

### Human-mouse ortholog analysis

```{r mouse ortholog analysis}
# Number of mouse TEs
MOUSE_TE = dim(mm10_rmsk_TE)[1]
MOUSE_TE

# Number of human TEs with orthologs
HG19_TE = dim(unique(human_mouse_orthologs_mm10[,hg19_coordinates]))[1]

# Number of mouse TEs with orthologs
MM10_TE = dim(unique(human_mouse_orthologs_mm10[,mm10_coordinates]))[1]

# Number of hg19 TEs that liftover
# (wc -l Mouse/liftover/rmsk_TE_hg19tomm10.bed)
1265775
1265775/NUM_TE

# Number of unique orthologous TEs in hg19->mm10, hg19, and mm10
dim(unique(human_mouse_orthologs_mm10[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")]))[1]
HG19_TE
HG19_TE/NUM_TE
MM10_TE

# Number, % of human TEs with 2+ human mm10 TEs
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_hg19))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_mm10))[2:11])/length(unique(human_mouse_orthologs_mm10$human_TE_mm10))

# Number, % of human TEs with 2+ mouse TEs
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$human_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$human_TE_mm10))[2:10])/length(unique(human_mouse_orthologs_mm10$human_TE_mm10))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$human_TE_hg19))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$human_TE_hg19))[2:10])/length(unique(human_mouse_orthologs_mm10$human_TE_hg19))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$mouse_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$mouse_TE_mm10))[2:15])/length(unique(human_mouse_orthologs_mm10$mouse_TE_mm10))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$mouse_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$mouse_TE_mm10))[2:25])/length(unique(human_mouse_orthologs_mm10$mouse_TE_mm10))

# Average length of those TEs
median(unique(human_mouse_orthologs_mm10[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")])$human_stop_mm10-
         unique(human_mouse_orthologs_mm10[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")])$human_start_mm10)
median(unique(human_mouse_orthologs_mm10[,hg19_coordinates])$human_stop_hg19-unique(human_mouse_orthologs_mm10[,hg19_coordinates])$human_start_hg19)
median(unique(human_mouse_orthologs_mm10[,mm10_coordinates])$mouse_stop_mm10-unique(human_mouse_orthologs_mm10[,mm10_coordinates])$mouse_start_mm10)

# Class distribution
test = rmsk_TE_class[,c("class_update","Count","Mouse_orthologs","Subfamilies","Mouse_ortholog_subfamily")]
test[,2:5] = apply(test[,2:5],2,function(x) x/sum(x))
test 

# Proportion of hg19 TEs and subfamilies with ortholog in mouse, by class
rmsk_TE_class$Mouse_orthologs/rmsk_TE_class$Count
rmsk_TE_class$Mouse_ortholog_subfamily/rmsk_TE_class$Subfamilies

# Average age of TEs with mouse orthologs
median(rmsk_TE$JC_distance)
median(merge(rmsk_TE,unique(human_mouse_orthologs_mm10[,hg19_coordinates]),by.x=TE_coordinates,by.y=hg19_coordinates[c(1:4,6,5,7)])$JC_distance)
```

### Human-mouse ortholog WGBS analysis

```{r mouse WGBS analysis}
# Number, % of hg19 orthologs with CpGs
x = merge(TE_meth_average[,c(TE_coordinates,"CpGs")],unique(human_mouse_orthologs_mm10[,hg19_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=hg19_coordinates)
dim(x)
dim(x)[1]/HG19_TE

# Average number of CpGs per hg19 ortholog with CpGs
median(x$CpGs)
rm(x)

# Number, % of mm10 orthologs with CpGs
y = merge(mm10_rmsk_TE_WGBS,unique(human_mouse_orthologs_mm10[,mm10_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=mm10_coordinates)
dim(y)
dim(y)[1]/MM10_TE

# Average number of CpGs per mm10 ortholog with CpGs
median(y$CpGs)
rm(y)

# Number of unique orthologous TEs with CpGs in mm10 and hg19 (both have CpGs)
dim(human_mouse_orthologs_WGBS)[1]
dim(unique(human_mouse_orthologs_WGBS[,mm10_coordinates]))[1]
dim(unique(human_mouse_orthologs_WGBS[,hg19_coordinates]))[1]

# TEs with meth data in at least one mouse/human sample
test = human_mouse_orthologs_WGBS[which(apply(human_mouse_orthologs_WGBS,1,function(x) sum(is.na(x[17:23])) < 7) & apply(human_mouse_orthologs_WGBS,1,function(x) sum(is.na(x[24:30])) < 7)),]

## Number of human (hg19) TEs
dim(unique(human_mouse_orthologs_WGBS[,hg19_coordinates]))[1]-dim(unique(test[,hg19_coordinates]))[1]

## Number of mouse TEs
dim(unique(human_mouse_orthologs_WGBS[,mm10_coordinates]))[1]-dim(unique(test[,mm10_coordinates]))[1]

# Number of TEs with methylation data per sample
ddply(human_mouse_orthologs_WGBS_paired,.(Human_sample,Mouse_sample_WGBS),summarise,Both=sum(!is.na(Human_methylation) & !is.na(Mouse_methylation)))

# Table of human/mouse states, overall
table(human_mouse_orthologs_WGBS_paired[,c("Mouse_state_WGBS","Human_state_WGBS")])

# Chi-sq test for human-mouse pairs, all methylation states, overall and by sample
ddply(human_mouse_orthologs_WGBS_paired,.(),summarise,Pvalue=unlist(chisq.test(x=Human_state_WGBS,y=Mouse_state_WGBS))["p.value"])
ddply(human_mouse_orthologs_WGBS_paired,.(Human_sample),summarise,Pvalue=unlist(chisq.test(x=Human_state_WGBS,y=Mouse_state_WGBS))["p.value"])

# Hypomethylated vs. all other states
test = human_mouse_orthologs_WGBS_paired[,c("Human_sample","Human_state_WGBS","Mouse_state_WGBS")]
test$Human_state_WGBS = mapvalues(test$Human_state_WGBS,meth_states[2:3],rep("Missing",2))
test$Mouse_state_WGBS = mapvalues(test$Mouse_state_WGBS,meth_states[2:3],rep("Missing",2))

ddply(test,.(),summarise,Pvalue=unlist(chisq.test(x=Human_state_WGBS,y=Mouse_state_WGBS))["p.value"])
ddply(test,.(Human_sample),summarise,Pvalue=unlist(chisq.test(x=Human_state_WGBS,y=Mouse_state_WGBS))["p.value"])

# TE orthologs hypomethylated in both human and mouse
human_mouse_orthologs_WGBS_hypo = human_mouse_orthologs_WGBS_paired[which(human_mouse_orthologs_WGBS_paired$Mouse_state_WGBS == "Hypomethylated" & human_mouse_orthologs_WGBS_paired$Human_state_WGBS == "Hypomethylated"),]

# Number of pairs where TE is hypomethylated in both
dim(human_mouse_orthologs_WGBS_hypo)
dim(unique(human_mouse_orthologs_WGBS_hypo[,hg19_coordinates]))
dim(unique(human_mouse_orthologs_WGBS_hypo[,mm10_coordinates]))
ddply(human_mouse_orthologs_WGBS_hypo,.(Human_sample),function(x) dim(x)[1])
```

### Age analysis (not shown)

```{r age analysis, eval=FALSE}
test = ddply(human_mouse_orthologs_WGBS_paired,.(human_chr_hg19,human_start_hg19,human_subfamily,human_class,mouse_chr_mm10,mouse_start_mm10,JC_distance),summarise,Concordant=sum(Human_state_WGBS == Mouse_state_WGBS),Con_hypo=sum(Human_state_WGBS == Mouse_state_WGBS & Human_state_WGBS == "Hypomethylated"))
cor.test(test$JC_distance,test$Con_hypo,method="spearman")
cor.test(test$JC_distance,test$Concordant,method="spearman")
by(test,test$human_class,function(x) cor.test(x$JC_distance,x$Con_hypo,method="spearman"))
by(test,test$human_class,function(x) cor.test(x$JC_distance,x$Concordant,method="spearman"))
```

### Human-mouse ortholog chromHMM analysis

```{r mouse chromHMM analysis}
# Confirming all orthologous TEs have chromHMM state
count_na(human_mouse_orthologs_chromHMM_paired)
dim(unique(hg19_orthologs_chromHMM[,hg19_coordinates]))[1]
dim(hg19_orthologs_chromHMM)
dim(unique(human_mouse_orthologs_chromHMM_paired[,c(hg19_coordinates,"Human_sample","Human_state_chromHMM")]))[1]
dim(unique(mm10_orthologs_chromHMM[,mm10_coordinates]))[1]
dim(mm10_orthologs_chromHMM)
dim(unique(human_mouse_orthologs_chromHMM_paired[,c(mm10_coordinates,"Mouse_sample_chromHMM","Mouse_state_chromHMM")]))[1]

# Number of orthologous TEs in each chromHMM state
human_mouse_chromHMM_table = table(human_mouse_orthologs_chromHMM_paired$Human_state_chromHMM,human_mouse_orthologs_chromHMM_paired$Mouse_state_chromHMM)
human_mouse_chromHMM_table

# Chi-squared, overall and by sample
ddply(human_mouse_orthologs_chromHMM_paired,.(),summarise,Pvalue=unlist(chisq.test(x=Human_state_chromHMM,y=Mouse_state_chromHMM))["p.value"])
ddply(human_mouse_orthologs_chromHMM_paired,.(Human_sample),summarise,Pvalue=unlist(chisq.test(x=Human_state_chromHMM,y=Mouse_state_chromHMM))["p.value"])

# Number of pairs and human/mouse TEs, both in promoter state
x = human_mouse_orthologs_chromHMM_paired[which(human_mouse_orthologs_chromHMM_paired$Human_state_chromHMM == "1_TssA" & human_mouse_orthologs_chromHMM_paired$Mouse_state_chromHMM == "TssA"),]
dim(x)[1]
dim(unique(x[,hg19_coordinates]))[1]
dim(unique(x[,mm10_coordinates]))[1]
ddply(x,.(Human_sample),function(x) dim(x)[1])

# hg19 TE class per state
table(unique(x[,hg19_coordinates])$human_class)

# Number of pairs and human/mouse TEs, both in active regulatory state
y = unique(human_mouse_orthologs_chromHMM_paired[which(human_mouse_orthologs_chromHMM_paired$Human_state_chromHMM %in% chromHMM_states[c(1:3,6:7)] & human_mouse_orthologs_chromHMM_paired$Mouse_state_chromHMM %in% mouse_chromHMM_states[c(1:3,6:8)]),c(hg19_coordinates,mm10_coordinates,"Human_sample","Mouse_sample_chromHMM")])
dim(y)[1]
dim(unique(y[,hg19_coordinates]))[1]
dim(unique(y[,mm10_coordinates]))[1]
ddply(y,.(Human_sample),function(z) dim(z)[1])

# hg19 TE class per state
table(unique(y[,hg19_coordinates])$human_class)

# Tissue-specific TEs
## Promoter state
as.data.frame(t(sapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) tissue_matrix(x,hg19_mm10_chromHMM_promoter))))

## Active regulatory state
as.data.frame(t(sapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) tissue_matrix(x,hg19_mm10_chromHMM_active))))

# TEs constitutively on/off
## Promoter state
dim(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples > 8 & hg19_mm10_chromHMM_promoter$Mouse_samples > 8),])
dim(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples > 8 & hg19_mm10_chromHMM_promoter$Mouse_samples < 2),])
dim(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples < 2 & hg19_mm10_chromHMM_promoter$Mouse_samples > 8),])

## Active regulatory states
dim(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples > 8 & hg19_mm10_chromHMM_active$Mouse_samples > 8),])
dim(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples > 8 & hg19_mm10_chromHMM_active$Mouse_samples < 2),])
dim(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples < 2 & hg19_mm10_chromHMM_active$Mouse_samples > 8),])
```

### Subfamily analysis

```{r ortholog subfamily analysis}
# Number of mouse subfamilies
length(unique(mm10_rmsk_TE$subfamily))

# Number of subfamilies shared between human and mouse
dim(hg19_mm10_subfamily_count)[1]

# Average age of subfamilies also in mouse 
median(rmsk_TE_subfamily$Age)
median(rmsk_TE_subfamily[which(rmsk_TE_subfamily$subfamily %in% mm10_rmsk_TE$subfamily),]$Age)

# Number of subfamilies with at least 30 members in both
dim(hg19_mm10_subfamily_count[which(hg19_mm10_subfamily_count$Count.x > THRESHOLD_IK_MEMBER & hg19_mm10_subfamily_count$Count.y > THRESHOLD_IK_MEMBER),])[1]
dim(hg19_mm10_subfamily_count[which(hg19_mm10_subfamily_count$Count_CpGs.x > THRESHOLD_IK_MEMBER & hg19_mm10_subfamily_count$Count_CpGs.y > THRESHOLD_IK_MEMBER),])[1]

# Average ratio of subfamily members for shared subfamilies
median(hg19_mm10_subfamily_count$Count.y/hg19_mm10_subfamily_count$Count.x)
median(hg19_mm10_subfamily_count$Count_CpGs.y/hg19_mm10_subfamily_count$Count_CpGs.x)
head(hg19_mm10_subfamily_count[order(hg19_mm10_subfamily_count$Count.y/hg19_mm10_subfamily_count$Count.x),])
tail(hg19_mm10_subfamily_count[order(hg19_mm10_subfamily_count$Count.y/hg19_mm10_subfamily_count$Count.x),])

# Linear correlations
by(hg19_mm10_TE_WGBS_subfamily_hypo_paired,hg19_mm10_TE_WGBS_subfamily_hypo_paired$Human_sample,function(x) lm(Mouse_hypo~Human_hypo,x))
by(hg19_mm10_chromHMM_subfamily[which(hg19_mm10_chromHMM_subfamily$Human_state_chromHMM == "1_TssA" & hg19_mm10_chromHMM_subfamily$Mouse_state_chromHMM == "TssA"),],droplevels(hg19_mm10_chromHMM_subfamily[which(hg19_mm10_chromHMM_subfamily$Human_state_chromHMM == "1_TssA" & hg19_mm10_chromHMM_subfamily$Mouse_state_chromHMM == "TssA"),]$Human_sample),function(x) lm(Percent.y~Percent.x,x))
by(hg19_mm10_chromHMM_subfamily_active,hg19_mm10_chromHMM_subfamily_active$Human_sample,function(x) lm(Percent.y~Percent.x,x))
```

### Correlation analysis

```{r correlation boxplots}
# Individual TE Spearman correlations
ddply(human_mouse_spearman,.(Test),summarise,Median=median(Correlation),IQR=IQR(Correlation))
pairwise.wilcox.test(human_mouse_spearman$Correlation,human_mouse_spearman$Test)
by(human_mouse_spearman,human_mouse_spearman$Test,function(x) pairwise.wilcox.test(x$Correlation,x$Tissue.Age))

# Subfamily Spearman correlations 
## Hypomethylated
ddply(human_mouse_spearman_subfamily,.(Test),summarise,Median=median(Correlation),IQR=IQR(Correlation))
pairwise.wilcox.test(human_mouse_spearman_subfamily$Correlation,human_mouse_spearman_subfamily$Test)
by(human_mouse_spearman_subfamily,human_mouse_spearman_subfamily$Test,function(x) pairwise.wilcox.test(x$Correlation,x$Tissue.Age))

## 1_TssA
ddply(human_mouse_spearman_subfamily_TssA,.(Test),summarise,Median=median(Correlation),IQR=IQR(Correlation))
pairwise.wilcox.test(human_mouse_spearman_subfamily_TssA$Correlation,human_mouse_spearman_subfamily_TssA$Test)
by(human_mouse_spearman_subfamily_TssA,human_mouse_spearman_subfamily_TssA$Test,function(x) pairwise.wilcox.test(x$Correlation,x$Tissue.Age))

## Active regulatory
ddply(human_mouse_spearman_subfamily_active,.(Test),summarise,Median=median(Correlation),IQR=IQR(Correlation))
pairwise.wilcox.test(human_mouse_spearman_subfamily_active$Correlation,human_mouse_spearman_subfamily_active$Test)
by(human_mouse_spearman_subfamily_active,human_mouse_spearman_subfamily_active$Test,function(x) pairwise.wilcox.test(x$Correlation,x$Tissue.Age))
```

### Human-mouse tissue-specific orthologs in same chromHMM state

```{r print candidates}
write.table(ldply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples < 5 & hg19_mm10_chromHMM_promoter[[paste(x,".x",sep="")]] == 2 & hg19_mm10_chromHMM_promoter$Mouse_samples < 5 & hg19_mm10_chromHMM_promoter[[paste(x,".y",sep="")]] == 2),]),file="Mouse/tissue_specific_promoter.txt",sep='\t',quote=FALSE,row.names=FALSE)

lapply(c("BRAIN","GI_INTESTINE","GI_STOMACH","HEART","LUNG"),function(x) write.table(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples < 5 & hg19_mm10_chromHMM_active[[paste(x,".x",sep="")]] == 2 & hg19_mm10_chromHMM_active$Mouse_samples < 5 & hg19_mm10_chromHMM_active[[paste(x,".y",sep="")]] == 2),],file=paste("Mouse/tissue_specific_active_",x,".txt",sep=""),sep='\t',quote=FALSE,row.names=FALSE))
```

### Human-mouse constitutive orthologs

```{r print constitutive}
## Promoter state
write.table(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples > 8 & hg19_mm10_chromHMM_promoter$Mouse_samples > 8),],file="Mouse/constitutive_both_promoter.txt",sep='\t',quote=FALSE,row.names=FALSE)
write.table(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples > 8 & hg19_mm10_chromHMM_promoter$Mouse_samples < 2),],file="Mouse/constitutive_human_promoter.txt",sep='\t',quote=FALSE,row.names=FALSE)
write.table(hg19_mm10_chromHMM_promoter[which(hg19_mm10_chromHMM_promoter$Human_samples < 2 & hg19_mm10_chromHMM_promoter$Mouse_samples > 8),],file="Mouse/constitutive_mouse_promoter.txt",sep='\t',quote=FALSE,row.names=FALSE)

## Active regulatory states
write.table(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples > 8 & hg19_mm10_chromHMM_active$Mouse_samples > 8),],file="Mouse/constitutive_both_active.txt",sep='\t',quote=FALSE,row.names=FALSE)
write.table(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples > 8 & hg19_mm10_chromHMM_active$Mouse_samples < 2),],file="Mouse/constitutive_human_active.txt",sep='\t',quote=FALSE,row.names=FALSE)
write.table(hg19_mm10_chromHMM_active[which(hg19_mm10_chromHMM_active$Human_samples < 2 & hg19_mm10_chromHMM_active$Mouse_samples > 8),],file="Mouse/constitutive_mouse_active.txt",sep='\t',quote=FALSE,row.names=FALSE)
```

http://epigenomegateway.wustl.edu/browser/?genome=hg19&session=CrUnGegbbU&statusId=1624390483 (Remy the chef)

# Methods 

## Basic statistics

```{r QC statistics}
# Proportion of TEs on chrY
NUM_TE-NUM_TE_noY
(NUM_TE-NUM_TE_noY)/NUM_TE

# Total bp in TEs
MERGED_TE_WIDTH

# Total TE bases (redundant)
sum(rmsk_TE$Length)-MERGED_TE_WIDTH
sum(rmsk_TE$Length)/MERGED_TE_WIDTH

# CpGs in TEs, genome
TE_CPGS
ALL_CPGS
```

## Blacklist TEs

```{r blacklist}
# Number of TEs overlapping blacklist
dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1]/NUM_TE

# Number of subfamilies overlapping blacklist
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$subfamily)))

# Which chromosomes?
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$chromosome)))

dim(rmsk_TE_subfamily_measure[which(rmsk_TE_subfamily_measure$blacklist > 0.01),c("subfamily","family","class_update","Count","blacklist")])
tail(rmsk_TE_subfamily_measure[order(rmsk_TE_subfamily_measure$blacklist),c("subfamily","family","class_update","Count","blacklist")])
```

# Supplement

### Supplementary Figure 18

```{r Fig S18, echo=FALSE, cache=TRUE, cache.lazy=FALSE}
source("R_scripts/state_lengths.R")

# Number vs. length of blocks/peaks by sample
all_lengths_blocks = ddply(all_lengths,.(State,Sample),summarise,Mean=mean(Length),SD=sd(Length),Median=median(Length),Blocks=length(Length))
all_lengths_blocks = merge(all_lengths_blocks,metadata[,c("Sample",sample_categories)],by="Sample",all.x=TRUE)

all_lengths_blocks_corr = ddply(all_lengths_blocks,.(State),summarise,Cor=as.numeric(unlist(cor.test(Blocks,Median,method="spearman"))["estimate.rho"]),Pvalue=as.numeric(unlist(cor.test(Blocks,Median,method="spearman"))["p.value"]))

ggplot(all_lengths_blocks,aes(x=log10(Blocks),y=log10(Median))) + geom_point(aes(color=Group)) + xlab("Number of blocks/peaks (log10)") + ylab("Median length of blocks/peaks (log10(bp))") + scale_color_manual(values=group_colors,guide=FALSE) + facet_wrap(~State,scales="free") + geom_smooth(method="lm",se=FALSE,color="black") + geom_text(data=all_lengths_blocks_corr,aes(x = -Inf, y = Inf,label=round(Cor,2)),hjust=-0.1,vjust=1.5,size=3) + scale_y_continuous(breaks=pretty_breaks(n=3)) + scale_x_continuous(breaks=pretty_breaks(n=3))
```

### State/peak lengths analysis

```{r analysis Fig S1, cache=TRUE}
# Length of states
ddply(all_lengths,.(State),summarise,Mean=mean(Length),SD=sd(Length),Median=median(Length))

# Differences in length of states by sample
ddply(all_lengths,.(State),summarise,KW_length=unlist(kruskal.test(Length~Sample))["p.value"])

# Differences in number of blocks by sample
test = ddply(all_lengths_blocks,.(State),summarise,Range_number=max(Blocks)-min(Blocks),CV_number=sd(Blocks)/mean(Blocks),Range_length=max(Median)-min(Median),CV_length=sd(Median)/mean(Median))
test[order(test$CV_number),]
test[order(test$CV_length),]

# Comparing number of blocks to length of states
all_lengths_blocks_corr[order(as.numeric(all_lengths_blocks_corr$Cor)),]

rm(all_lengths)
```

### Mappability of genome/TE bases

```{r mappability}
# Mappability whole genome, TEs (8/27/17)
# (awk '{sum+=($4*($3-$2));total+=$3-$2;}END{print sum/total}' wgEncodeCrgMapabilityAlign36mer.bedGraph)
GENOME_MAP=0.789192
GENOME_MAP

# (for file in x*; do bedtools intersect -wo -a rmsk_TEother_merge.txt -b $file >> rmsk_TEother_merge_map.txt; done)
# (awk '{sum+=($7*$8);total+=$8;}END{print sum/total}' rmsk_TEother_merge_map.txt)
TE_MAP=0.624207
TE_MAP

# Mappability vs. length by class
cor.test(rmsk_TE$Length,rmsk_TE$mappability,method="spearman")
by(rmsk_TE,rmsk_TE$class_update,function(x) cor.test(x$Length,x$mappability,method="spearman"))

# Correlations between mappability, CpGs, CpGs per length, number of samples Missing (WGBS)
correlate_TE_state[which(correlate_TE_state$State == "Missing" & correlate_TE_state$Metric %in% c("mappability","CpGs","CpGs_per_length")),]

# Correlation between mappability, number of samples Quiescent
correlate_TE_state[which(correlate_TE_state$Metric == "mappability" & correlate_TE_state$State == "15_Quies"),]
```