---
title: "TE landscape Draft"
author: "Erica Pehrsson"
date: "`r Sys.Date()`"
output: 
  word_document: 
    keep_md: yes
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(fig.width=7, fig.height=9, fig.path='TE_landscape_draft_figures/', cache.path="TE_landscape_draft_cache/", error=TRUE)
```

# Setup

## Load required libraries

```{r load libraries}
library(plyr)
packageVersion("plyr")
library(reshape2)
packageVersion("reshape2")
library(vcd)
packageVersion("vcd")
library(ggplot2)
packageVersion("ggplot2")
library(NMF)
packageVersion("NMF")
library(gridExtra)
packageVersion("gridExtra")
library(RColorBrewer)
packageVersion("RColorBrewer")
```

## Set universal plot parameters

```{r set plot parameters}
theme_set(theme_bw() %+replace%  theme(panel.background=element_rect(fill="white"),panel.border=element_rect(color="black",fill=NA),panel.grid=element_blank()))
```

## Load functions

```{r load functions}
source("R_scripts/functions.R")
```

## Create essential matrices

```{r create essential matrices, eval=FALSE}
source("R_scripts/metadata.R")
source("R_scripts/TE_matrix.R")
source("R_scripts/chromHMM_matrix.R")
source("R_scripts/WGBS_TE_avg_methylation.R")
source("R_scripts/DNase_peaks.R")
source("R_scripts/H3K27ac_peaks.R")
source("R_scripts/RNAseq.R")
source("R_scripts/promoter_matrices.R")
source("R_scripts/shuffled_matrix.R")
source("R_scripts/shuffled_matrix_CpG.R")
source("R_scripts/shuffled_matrix_mappability.R")
source("R_scripts/shuffled_matrix_features.R")
source("R_scripts/shuffled_matrix_chromHMM.R")
source("R_scripts/shuffled_matrix_WGBS.R")
source("R_scripts/shuffled_matrix_DNase.R")
source("R_scripts/shuffled_matrix_H3K27ac.R")
```

## Load metadata

```{r load metadata}
load("R_datasets/metadata.RData")
```

## Load essential matrices

```{r load essential matrices}
load("R_datasets/rmsk_TE.RData")
load("R_datasets/chromHMM_TE_state.RData")
load("R_datasets/TE_meth_average.RData")
load("R_datasets/TE_DNase_peaks.RData")
load("R_datasets/TE_H3K27ac_peaks.RData")
load("R_datasets/rna.RData")
```

## Define color and label vectors

```{r define vectors}
# Chromosomes
standard_chromosomes = c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY","chrM")

# chromHMM
chromHMM_states = c("1_TssA","2_TssAFlnk","3_TxFlnk","4_Tx","5_TxWk","6_EnhG","7_Enh","8_ZNF/Rpts","9_Het","10_TssBiv","11_BivFlnk","12_EnhBiv","13_ReprPC","14_ReprPCWk","15_Quies")
chromHMM_labels = setNames(c("Active TSS","Flanking active TSS","Transcription at gene 5'/3'","Strong transcription","Weak transcription","Genic enhancer","Enhancer","ZNF genes/repeats","Heterochromatin","Poised TSS","Flanking poised TSS/enhancer","Poised enhancer","Repressed Polycomb","Weak repressed Polycomb","Quiescent"),chromHMM_states)
chromHMM_colors = setNames(c("#66A61E","#1B9E77","#005A32","#E7298A","#F781BF","#41B6C4","#045A8D","#FFD92F","#99000D","#D95F02","#E6AB02","#A6761D","#6A3D9A","#CAB2D6","#666666"),chromHMM_states)
chromHMM_composite = setNames(c("Active regulatory","Active regulatory","Active regulatory","Transcribed","Transcribed","Active regulatory","Active regulatory","Zinc-finger/repeats","Repressed","Poised regulatory","Poised regulatory","Poised regulatory","Repressed","Repressed","Quiescent"),chromHMM_states)

# WGBS
meth_states = c("Hypomethylated","Intermediate","Hypermethylated","Missing")
meth_labels = setNames(c("Hypomethylated (<30%)","Intermediately methylated (30-70%)","Hypermethylated (>70%)","Missing CpG data"),meth_states)
meth_colors = setNames(c("#F8766D","#00BFC4","#7CAE00","#C77CFF"),meth_states)

# All states
all_state_colors = c(chromHMM_colors,meth_colors,"aquamarine","tan1","black")
names(all_state_colors)[20:22] = c("DNase","H3K27ac","Expression")
all_state_labels = setNames(c(chromHMM_states,meth_labels,"DNase","H3K27ac","RPKM > 1"),c(chromHMM_states,meth_states,"DNase","H3K27ac","Expression"))

# TE classification
class_colors = setNames(c("gold1","azure3","peru","lemonchiffon3","seagreen","lavenderblush4","brown1"),c("DNA","LINE","LTR","SINE","SVA","Other","RC"))

# Sample classification
group_colors = setNames(c("#924965","#4178AE","#E41A1C","#69608A","#B65C73","#FF9D0C","#678C69","#55A354","#E67326","#FFD924","#AF5B39","#D56F80","#999999","#C5912B","#C58DAA","#F182BC","#C2655D","#DAB92E","#000000"),c("ESC","ES-deriv","IMR90","iPSC","Mesench","Epithelial","HSC & B-cell","Blood & T-cell","Myosat","Neurosph","Adipose","Heart","Other","Brain","Digestive","Sm. Muscle","Muscle","Thymus","ENCODE2012"))
anatomy_colors = setNames(c("#B2DF8A","#33A02C","#FB9A99","#FFED6F","#80B1D3","#666666","#924965","#4178AE","#FDBF6F","#CAB2D6","#6A3D9A","#BEBADA","#FCCDE5","#BC80BD","#E7298A","#FF7F00","#69608A","#B3DE69","#CCEBC5","#000000","#FB8072","#FDB462","#8DD3C7","#D9D9D9","#E31A1C","#A6CEE3","#D95F02","#E6AB02","#66A61E","#B15928","#E31A1C","#8DD3C7","#E7298A","#A6761D","#666666"),c("ADRENAL","BLOOD","BONE","BRAIN","BREAST","CERVIX","ESC","ESC_DERIVED","FAT","GI_COLON","GI_DUODENUM","GI_ESOPHAGUS","GI_INTESTINE","GI_RECTUM","GI_STOMACH","HEART","IPSC","KIDNEY","LIVER","LUNG","MUSCLE","MUSCLE_LEG","OVARY","PANCREAS","PLACENTA","SKIN","SPLEEN","STROMAL_CONNECTIVE","THYMUS","VASCULAR",NA,NA,NA,NA,NA))

# Genic features
genic_labels = setNames(c("Promoter","5'UTR","CDS","3'UTR","Exon","Intron","Intergenic","CpG island"),c("promoters","5UTR","coding_exon","3UTR","exons","introns","intergenic","cpgIsland"))

# C-Gate subfamilies
cgate_subfams = c('AluJb','AluJo','AluS','AluSc','AluSg','AluSg1','AluSp','AluSq','AluSx','AluSz','AluY','FLAM_A','FRAM','HERV3-int','HERV9-int','HERVE-int','HERVH-int','L1HS','L1MB4','L1PA6','L2a','LTR10A','LTR16A','LTR22B','LTR31A','LTR5','LTR5_Hs','LTR9','MER11A','MER21A','MER34','MER39B','MER4B','MER54B','MER74C','Mir m','MIR3','MIRb','MLT1C','MLT2B2','MLT2B3','MSTB2','Plat_L3','RTVL-H','THE1A','THE1C','THE1D','AluJr','AluSq2','L1MB7','L1MC4','L1ME3A','L1PA11','L1PA4','LTR12','LTR12C','LTR12D','LTR13A','LTR2','LTR2B','MIR','MIRc','MLT1B')
```

## Define facet labels

```{r define facet labels}
measure_labels = c("% genome","% TEs (either strand)","% TEs (same strand)","% feature in TEs")
names(measure_labels) = c("Genome_percent_genome","TEs_percent_TE","TEs_strand_percent_TE","Genome_percent_TE")

mark_labels = c("chromHMM (n=127)","WGBS (n=37)","DNase (n=53)","H3K27ac (n=98)")
names(mark_labels) = c("chromHMM","WGBS","DNase","H3K27ac")

group_labels = c("chromHMM","WGBS","DNase","H3K27ac","RPKM > 1")
names(group_labels) = c("chromHMM","WGBS","DNase","H3K27ac","Expression")

coding_labels = c("All","Protein-coding","Non-coding")
names(coding_labels) = c("All","pc","nc")

metric_labels = c("Proportion of TEs in state in any sample","Average % of samples in state","Average % of samples in state if ever in state")
names(metric_labels) = c("Proportion_ever","Samples_avg_all","Samples_avg_ever")

variable_labels = c("Mappability","Jukes-Cantor evolutionary distance","Length (bp)","CpGs","CpGs/bp")
names(variable_labels) = c("mappability","JC_distance","Length","CpGs","CpGs_per_length")
```

## Load feature sizes

```{r load feature sizes}
# Genome
hg19_genome = read.table("raw_data/hg19.genome",sep='\t',header=TRUE)
GENOME_WIDTH = sum(as.numeric(hg19_genome[which(hg19_genome$chrom %in% standard_chromosomes),]$size))
GENOME_WIDTH_noY = sum(as.numeric(hg19_genome[which(hg19_genome$chrom %in% standard_chromosomes & hg19_genome$chrom != "chrY"),]$size))

# CpGs
ALL_CPGS = 56434896
TE_CPGS = 28373958

# TEs 
NUM_TE = dim(rmsk_TE)[1]
NUM_TE_noY = dim(rmsk_TE[which(rmsk_TE$chromosome != "chrY"),])[1]
NUM_TE_WGBS = dim(TE_meth_average)[1]
  
# awk '{sum+=$3-$2}END{print sum}' features/TEs/merge/rmsk_TEother_merge.txt
MERGED_TE_WIDTH = 1389947349
# awk '{if($1 != "chrY") sum+=$3-$2}END{print sum}' features/TEs/merge/rmsk_TEother_merge.txt
MERGED_TE_WIDTH_noY = 1375898142
```

## Calculate TE class and subfamily stats

```{r calculate TE stats}
source("R_scripts/TE_class_stats.R")
source("R_scripts/TE_subfamily_stats.R")
```

## Generate other matrices

```{r generate other matrices}
source("~/TE_landscape/R_scripts/WGBS_sample_CpG_state.R")
source("~/TE_landscape/R_scripts/DNase_overlap.R")
source("~/TE_landscape/R_scripts/H3K27ac_overlap.R")
```

# Introduction

## Basic sample statistics

```{r basic sample stats}
# Samples with each metric
apply(metadata[,c(1,10:13)],2,function(x) dim(metadata[which(!is.na(x)),])[1])

# Samples with each metric, not excluded 
apply(metadata[,c(1,10:13)],2,function(x) dim(metadata[which(!is.na(x) & metadata$Exclude == "Include"),])[1])
```

## Basic TE statistics

```{r basic TE stats}
# Number of TEs
NUM_TE

# Proportion of TEs on chrY
NUM_TE-NUM_TE_noY
(NUM_TE-NUM_TE_noY)/NUM_TE

# Total bp in TEs
MERGED_TE_WIDTH

# % of genome
MERGED_TE_WIDTH/GENOME_WIDTH

# Total TE bases (redundant)
sum(rmsk_TE$Length)-MERGED_TE_WIDTH
sum(rmsk_TE$Length)/MERGED_TE_WIDTH

# % CpGs overlapping TEs
TE_CPGS
ALL_CPGS
TE_CPGS/ALL_CPGS

# Number of families
length(unique(rmsk_TE$family))

# Number of subfamilies
length(unique(rmsk_TE$subfamily))
```

# Results

## Contribution

### Figure 1

**Figure 1. Contribution of TEs to active and regulatory epigenetic states.** (A) Total proportion of TEs, the entire genome, and RefSeq genic features annotated with each chromHMM state, overlapping DNase or H3K27ac peaks, or in each methylation state (WGBS, CpGs overlapping feature) across all samples annotated with that measure. Normalized by total bases annotated with chromHMM for that feature. Regions of overlap with TEs were excluded from all RefSeq features, and features were merged to remove redundant bases. Promoters represent 2000bp upstream / 500bp downstream of transcription start sites. (B) Total proportion of each state within TEs across all samples, ordered by proportion in TEs. Horizontal line represents the proportion of all genomic bases in TEs (“Total”). 

```{r Figure 1, fig.width=7, fig.height=5, echo=FALSE}
source("R_scripts/proportion.R")
source("R_scripts/contribution.R")

a = ggplot(combined_proportion[which(combined_proportion$Coding == "All"),],aes(x=Cohort,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion of feature in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x=element_blank()) + scale_x_discrete(labels=c("TE","Genome",genic_labels[1:7])) + facet_wrap(~Group,nrow=2,scales="free",labeller=labeller(Group = mark_labels))

b = ggplot(contribution_TE[which(contribution_TE$Cohort == "TE"),],aes(x=State,y=Proportion,fill=State)) + geom_bar(stat="identity") + coord_flip() + ylab("Proportion of state in TEs") + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.ticks.x=element_blank()) + geom_hline(yintercept = MERGED_TE_WIDTH/GENOME_WIDTH) + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=contribution_TE$State[1:21]) + geom_text(aes(label=round(100*Proportion,1)),position="stack",hjust=1)

legend = get_legend(ggplot(combined_proportion[which(combined_proportion$Coding == "All"),],aes(x=Cohort,y=Proportion,fill=State)) + geom_bar(stat="identity") + scale_fill_manual(values=all_state_colors) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x=element_blank(),legend.title = element_blank(),legend.position="bottom",legend.direction = "horizontal") + facet_wrap(~Group,nrow=4))

grid.arrange(a,b,legend, nrow = 2,layout_matrix = rbind(c(1,2),c(3,3)),heights=c(0.9,0.1))
```

### Proportion/contribution analysis

```{r analysis Fig 1}
# Proportion of each feature in each state, across all samples
combined_proportion[which(combined_proportion$Coding == "All"),]

# Specific states, TEs, promoters, and exons
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "TE" & combined_proportion$State %in% chromHMM_states[c(1:3,6:7)]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "TE" & combined_proportion$State %in% chromHMM_states[4:5]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "promoters" & combined_proportion$State %in% chromHMM_states[c(1:3,6:7)]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "exons" & combined_proportion$State %in% chromHMM_states[4:5]),]$Proportion)

# Does any feature have less % in state than TEs?
ddply(combined_proportion,.(State),function(x) x[which(x$Proportion < sum(x[which(x$Feature == "TE"),]$Proportion)),])

# Proportion of all genomic bases in TEs
(121*MERGED_TE_WIDTH+5*MERGED_TE_WIDTH_noY)/(121*GENOME_WIDTH+5*GENOME_WIDTH_noY)

# Contribution
contribution_TE[which(contribution_TE$Cohort == "TE"),]

# Contribution, composite states
contribution_composite = as.data.frame(cbind(rowSums(contribution[,2:9]),rowSums(contribution[,10:16]),rowSums(contribution[,c(2:4,7:8)]),rowSums(contribution[,5:6]),rowSums(contribution[,c(10,14:15)]),rowSums(contribution[,11:13])))
colnames(contribution_composite) = c("Active","Inactive","Active Regulatory","Transcribed","Repressed","Poised Regulatory")
contribution_composite = as.data.frame(t(contribution_composite[1:3,]))
contribution_composite = (contribution_composite/contribution_composite$Total)[,2:3]
contribution_composite
```

### Average proportion of TE CpGs hypomethylated per sample

```{r CpG proportion sample}
mean(apply(TE_CpG_meth,1,function(x) x[1]/sum(x)))
sd(apply(TE_CpG_meth,1,function(x) x[1]/sum(x)))
```

### VISTA enhancers

```{r vista enhancers}
# wc -l features/vista_enhancers/vista_enhancers_human.bed
# cat features/intersect_features/vista_enhancers_rmsk_*.txt | awk '{print $1, $2, $3}' - | sort | uniq | wc -l
355

355/934
```

### Supplementary Figure 1

**Supplementary Figure 1. Confirmation of chromHMM annotations.** (A) Distribution of hg19 TE lengths by class, plotted on log10 scale. (B) Distribution of lengths of chromHMM state blocks, DNase peaks, and H3K27ac peaks from all samples, plotted on log10 scale. (C) Distribution of methylation levels for CpGs overlapping TEs, by sample. Excludes CpGs with missing data. (D) Fold-enrichment ratios over input for ChIP-seq of seven histone modifications and DNase hypersensitivity as well as DNA methylation level over 10kb regions centered on TEs in each chromHMM state (≥1bp) in each sample, calculated in bins of 50bp. Values are averages of all instances of TEs in the state in each of the 127 samples. 

```{r Figure S1a, echo=FALSE, cache=TRUE, dependson=c("setup","load libraries","set plot parameters","define vectors")}
source("R_scripts/state_lengths.R")

a = ggplot(all_lengths,aes(x=State,y=log10(Length),fill=State)) + geom_boxplot() + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_fill_manual(values=all_state_colors[c(1:15,20:21)],guide=FALSE) + ylab("Length (log10(bp))") 
```

```{r cpg density plot, echo=FALSE, eval=FALSE}
source("R_scripts/cpg_density.R")
load("R_datasets/cpg_density.RData")

ggplot(cpg,aes(x=Methylation,fill=Sample)) + geom_density(alpha=0.5,adjust=10) + xlab("Methylation level per CpG") + ylab("Density")
```

```{r Figure S1b-d, echo=FALSE}
source("R_scripts/proportion_investigation.R")

# Bases/CpGs in state by sample
b = ggplot(all_state_proportion[which(all_state_proportion$Cohort == "Genome"),],aes(x=Sample,y=Bases,fill=State)) + geom_bar(stat="identity") + ylab("Bases/CpGs in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.ticks.x=element_blank(),axis.title.x=element_blank()) + facet_wrap(~Mark,nrow=2,scales="free",labeller=labeller(Mark = mark_labels))

# Average mappability by chromHMM states
test = lapply(metadata$Sample[c(1:12,32:40,119:127)],function(x) read.table(file=paste("mappability/by_state/mappability_",x,"_average.txt",sep=""),sep='\t'))
names(test) = metadata$Sample[c(1:12,32:40,119:127)]
test = ldply(test)

c = ggplot(test,aes(x=V1,y=V2,fill=V1)) + geom_boxplot() + scale_x_discrete(limits=chromHMM_states) + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + ylab("Average mappability") + scale_fill_manual(values=chromHMM_colors,guide=FALSE)

source("R_scripts/histone_modification.R")

d = ggplot(histones,aes(x=Bin,y=Level,color=Mark,group=Mark))+geom_point(size=0.5)+geom_line(size=1.5)+theme(axis.text.x = element_blank(),axis.ticks.x=element_blank(),panel.grid=element_blank(),legend.position = "bottom",legend.title = element_blank())+ylab("Fold enrichment over input") + scale_color_manual(labels=c(levels(histones$Mark)[1:8],"DNA methylation"),values=c(gg_color_hue(8),"black")) + xlab("Genomic position (+/-5kb around TE)")+ylim(0,12) + facet_wrap(~ State,nrow=3)
```

```{r Figure S1 plot, echo=FALSE}
grid.arrange(a,b,c,d,nrow=3,ncol=2,layout_matrix=rbind(c(1,2),c(3,2),c(4,4)),heights=c(0.2,0.2,0.6),widths=c(0.5,0.5))
```

### State/peak lengths analysis

```{r analysis Fig S1}
# Length of TEs
min(rmsk_TE$Length)
median(rmsk_TE$Length)
max(rmsk_TE$Length)

# Length of states
ddply(all_lengths,.(State),summarise,Mean=mean(Length),SD=sd(Length),Median=median(Length))
```

### CpG state analysis

```{r}
apply(all_CpG_meth,1,function(x) x/sum(x))
median(t(apply(all_CpG_meth,1,function(x) x/sum(x)))[,2])
```

### By-sample analysis

```{r by-sample bases}
ddply(all_state_proportion,.(State),summarise,Mean=mean(Bases),SD=sd(Bases),Min=min(Bases),Max=max(Bases),CV=sd(Bases)/mean(Bases))
```

### Mappability of genome/TE bases

```{r}
# Mappability whole genome, TEs (8/27/17)
# (awk '{sum+=($4*($3-$2));total+=$3-$2;}END{print sum/total}' wgEncodeCrgMapabilityAlign36mer.bedGraph)
GENOME_MAP=0.789192
GENOME_MAP

# (for file in x*; do bedtools intersect -wo -a rmsk_TEother_merge.txt -b $file >> rmsk_TEother_merge_map.txt; done)
# (awk '{sum+=($7*$8);total+=$8;}END{print sum/total}' rmsk_TEother_merge_map.txt)
TE_MAP=0.624207
TE_MAP
```

### Supplementary Figure 2

**Supplementary Figure 2. Overlap of TEs with RefSeq genic features.** (A) Length of overlap between TEs and RefSeq features as proportion of genome, feature, or TEs. (B) Average proportion of TEs, the entire genome, and RefSeq genic features in each state across all samples annotated with that measure, split by protein-coding and non-coding features ("NM"/"NR"). Regions of overlap with TEs were excluded from all RefSeq features, and features were merged to remove redundant bases.

```{r Figure S2, echo=FALSE}
#source("R_scripts/proportion.R")
source("R_scripts/feature_overlap.R")
source("R_scripts/proportion_class.R") #Requires proportion.R

a = ggplot(feature_overlap_long,aes(x=Feature,y=Percent,fill=Cohort)) + geom_bar(position="dodge",stat="identity") + theme(axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1),legend.title = element_blank())  + scale_x_discrete(labels=genic_labels[1:7]) + scale_fill_discrete(labels=c("All","Protein-Coding","Non-coding")) + facet_wrap(~Measure,nrow=1,labeller=labeller(Measure=measure_labels))

b = ggplot(combined_proportion[which(combined_proportion$Cohort != "genome_noTE"),],aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion of feature in state") + scale_fill_manual(values=all_state_colors) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x=element_blank(),legend.position = "bottom") + scale_x_discrete(labels=c("TE","Genome",genic_labels[1:7])) + facet_grid(Coding~Group,labeller=labeller(Coding=coding_labels))

# Proportion in state by class
combine_class_proportion = rbind(contribution_class[,c(1:2,4)],class_CpG_meth_total,TE_DNase_class_total,TE_H3K27ac_class_total)
combine_class_proportion$Cohort = factor(c(rep("chromHMM",90),rep("WGBS",24),rep("DNase",6),rep("H3K27ac",6)),levels=c("chromHMM","WGBS","DNase","H3K27ac"))

c = ggplot(combine_class_proportion,aes(x=class,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion of class in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x = element_blank()) + facet_wrap(~Cohort,scales="free",nrow=1)

grid.arrange(a,b,c,nrow=3,heights=c(0.2,0.5,0.3))
```

### RefSeq feature analysis

```{r analysis Fig S2}
# Feature %
feature_overlap_long

# Proportion of each feature in each state, across all samples
combined_proportion[which(combined_proportion$Cohort != "genome_noTE"),]
```

## Potential

### Figure 2

**Figure 2. Potential for individual TEs to be in each epigenetic state.** (A) Boxplots indicate the proportion of all TEs (n=`NUM_TE`) in the state in each sample. Red dots are the fraction of all TEs in the state in any sample. For WGBS states, TEs are restricted to those with any CpG (n=`NUM_TE_WGBS`). (B) Average proportion of samples a TE is in a state. Total is >100% for chromHMM due to TEs in multiple states in a single sample. (C) For TEs in a chromHMM state in at least one sample, cumulative proportion of individual TEs in that state in at least that fraction of samples. Horizontal lines indicate the mean fraction of samples a TE is in each state if it is in that state in at least one sample. 

```{r Figure 2, fig.width=7, fig.height=5, echo=FALSE}
source("R_scripts/potential.R")

combine_boxplot = rbind(state_sample_count[,c(1:2,4)],TE_meth_average_state_long,TE_DNase_peaks_sample[,2:4],TE_H3K27ac_peaks_sample[,2:4],RNA_RPKM_sample[,2:4])
combine_stats = rbind(chromHMM_TE_state_dist_stats,TE_meth_average_category_stats,TE_DNase_potential_stats,TE_H3K27ac_potential_stats,RNA_potential_stats)
combine_stats$Group = factor(c(rep("chromHMM",15),rep("WGBS",4),"DNase","H3K27ac","Expression"),levels=c("chromHMM","WGBS","DNase","H3K27ac","Expression"))

a = ggplot(combine_boxplot,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="% of TEs in state") + coord_flip() + scale_x_discrete(limits = rev(levels(combine_boxplot$State)),labels=rev(all_state_labels)) + geom_point(data=combine_stats,aes(x=State,y=Proportion_ever),color="red",size=3)

b = ggplot(combine_stats,aes(x=1,y=Samples_avg_all,fill=State)) + geom_bar(stat="identity",show.legend=FALSE) + scale_fill_manual(values=all_state_colors) + theme(axis.title.x=element_blank(),axis.ticks.x=element_blank(),axis.text.x=element_blank()) + ylab("Average % of samples in state") + facet_wrap(~Group,nrow=1,labeller=labeller(Group = group_labels))

c = ggplot(chromHMM_TE_state_cum,aes(x=Sample_proportion,y=Proportion,color=State,group=State)) + geom_point(size=0.1) + geom_line(size=1.5) + labs(x="Proportion of all samples (n=127)",y="Cumulative proportion of TEs in state") + scale_color_manual(values=chromHMM_colors,guide=FALSE) + geom_vline(xintercept = chromHMM_TE_state_dist_stats$Samples_avg_ever/100,color=chromHMM_colors)

legend = get_legend(ggplot(combine_boxplot,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors) + theme(legend.direction = "horizontal",legend.title = element_blank(),legend.position="bottom"))

grid.arrange(a,b,c,legend, nrow = 3, layout_matrix=rbind(c(1,2),c(1,3),c(4,4)),heights=c(0.4,0.4,0.2))
```

### Potential analysis

```{r analysis Fig 2a}
aggregate(data=combine_boxplot,Proportion~State,mean)
combine_stats
```

### Number of TEs ever in composite states

```{r analysis Fig 2b}
# Roadmap Active (states 1-8)
sum(apply(chromHMM_TE_state[8:15],1,function(x) sum(x > 0)) > 0)/NUM_TE
# Roadmap Inactive (states 9-15)
sum(apply(chromHMM_TE_state[16:22],1,function(x) sum(x > 0)) > 0)/NUM_TE
# Active regulatory states (states 1-3, 6-7)
sum(apply(chromHMM_TE_state[c(8:10,13:14)],1,function(x) sum(x > 0)) > 0)/NUM_TE
# Transcribed states (states 4-5)
sum(apply(chromHMM_TE_state[11:12],1,function(x) sum(x > 0)) > 0)/NUM_TE
# Poised regulatory states (states 10-12)
sum(apply(chromHMM_TE_state[17:19],1,function(x) sum(x > 0)) > 0)/NUM_TE
# Repressed states (states 9, 13-14)
sum(apply(chromHMM_TE_state[c(16,20,21)],1,function(x) sum(x > 0)) > 0)/NUM_TE
```

### Additional RNA analysis

```{r analysis Fig 2c}
# Average number of samples a TE is expressed RPKM > 0
mean(apply(RNA_TE_agnostic[,9:60],1,function(x) sum(x > 0)))

table(RNA_TE_agnostic$Expressed_samples)
table(RNA_TE_agnostic$Expressed_samples)/dim(RNA_TE_agnostic)[1]

# Maximum expression
#hist(RNA_TE_agnostic$Max_expression)
median(RNA_TE_agnostic$Max_expression)
median(RNA_TE_agnostic[which(RNA_TE_agnostic$Expressed_samples > 0),]$Max_expression)
max(RNA_TE_agnostic$Max_expression)
tail(RNA_TE_agnostic[order(RNA_TE_agnostic$Max_expression),],n=20)

cor.test(RNA_TE_agnostic$Max_expression,RNA_TE_agnostic$Expressed_samples,method="spearman")
```

### Supplementary Figure 3

**Supplmentary Figure 3. Comparison of chromHMM, WGBS, DNase, H3K27ac states and RNA expression.** (A-T) Proportion of all TEs x sample in each metric 1 state annotated with each metric 2 state in samples with data for both metrics. ChromHMM comparisons are normalized for TEs annotated with multiple states. (U) Comparison of all states. 

```{r Figure S3, echo=FALSE}
source("R_scripts/compare_marks.R")

a = xtabs(TE_sample_norm ~ chromHMM+WGBS, data=compare_marks_all)
names(dimnames(test)) = c("chromHMM","WGBS")

b = xtabs(TE_sample_norm ~ WGBS+chromHMM, data=compare_marks_all)
names(dimnames(test)) = c("WGBS","chromHMM")

c = xtabs(TE_sample_norm ~ DNase+chromHMM, data=compare_marks_all)
names(dimnames(test)) = c("DNase","chromHMM")

d = xtabs(TE_sample_norm ~ H3K27ac+chromHMM, data=compare_marks_all)
names(dimnames(test)) = c("H3K27ac","chromHMM")

e = xtabs(TE_sample_norm ~ RNA+chromHMM, data=compare_marks_all)
names(dimnames(test)) = c("RNA","chromHMM")

f = xtabs(TE_sample_norm ~ chromHMM+DNase, data=compare_marks_all)
names(dimnames(test)) = c("chromHMM","DNase")

g = xtabs(TE_sample ~ WGBS+DNase, data=compare_marks_unique)
names(dimnames(test)) = c("WGBS","DNase")

h = xtabs(TE_sample ~ DNase+WGBS, data=compare_marks_unique)
names(dimnames(test)) = c("DNase","WGBS")

i = xtabs(TE_sample ~ H3K27ac+WGBS, data=compare_marks_unique)
names(dimnames(test)) = c("H3K27ac","WGBS")

j = xtabs(TE_sample ~ RNA+WGBS, data=compare_marks_unique)
names(dimnames(test)) = c("RNA","WGBS")

k = xtabs(TE_sample_norm ~ chromHMM+H3K27ac, data=compare_marks_all)
names(dimnames(test)) = c("chromHMM","H3K27ac")

l = xtabs(TE_sample ~ WGBS+H3K27ac, data=compare_marks_unique)
names(dimnames(test)) = c("WGBS","H3K27ac")

m = xtabs(TE_sample ~ DNase+H3K27ac, data=compare_marks_unique)
names(dimnames(test)) = c("DNase","H3K27ac")

n = xtabs(TE_sample ~ H3K27ac+DNase, data=compare_marks_unique)
names(dimnames(test)) = c("H3K27ac","DNase")

o = xtabs(TE_sample ~ RNA+DNase, data=compare_marks_unique)
names(dimnames(test)) = c("RNA","DNase")

p = xtabs(TE_sample_norm ~ chromHMM+RNA, data=compare_marks_all)
names(dimnames(test)) = c("chromHMM","RNA")

q = xtabs(TE_sample ~ WGBS+RNA, data=compare_marks_unique)
names(dimnames(test)) = c("WGBS","RNA")

r = xtabs(TE_sample ~ DNase+RNA, data=compare_marks_unique)
names(dimnames(test)) = c("DNase","RNA")

s = xtabs(TE_sample ~ H3K27ac+RNA, data=compare_marks_unique)
names(dimnames(test)) = c("H3K27ac","RNA")

t = xtabs(TE_sample ~ RNA+H3K27ac, data=compare_marks_unique)
names(dimnames(test)) = c("RNA","H3K27ac")

mplot(mosaic(a,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(b,shade=TRUE,return_grob=TRUE,legend=FALSE),
      mosaic(c,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(d,shade=TRUE,return_grob=TRUE,legend=FALSE),
      mosaic(e,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(f,shade=TRUE,return_grob=TRUE,legend=FALSE),
      mosaic(g,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(h,shade=TRUE,return_grob=TRUE,legend=FALSE),
      mosaic(i,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(j,shade=TRUE,return_grob=TRUE,legend=FALSE),
      mosaic(k,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(l,shade=TRUE,return_grob=TRUE,legend=FALSE),
      mosaic(m,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(n,shade=TRUE,return_grob=TRUE,legend=FALSE),
      mosaic(o,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(p,shade=TRUE,return_grob=TRUE,legend=FALSE),
      mosaic(q,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(r,shade=TRUE,return_grob=TRUE,legend=FALSE),
      mosaic(s,shade=TRUE,return_grob=TRUE,legend=FALSE),mosaic(t,shade=TRUE,return_grob=TRUE,legend=FALSE),pin=c(7,9))

u = xtabs(TE_sample_norm ~ chromHMM+WGBS+DNase+H3K27ac+RNA, data=compare_marks_all)

mplot(mosaic(u,shade=TRUE,return_grob=TRUE,legend=FALSE),pin=c(7,9))
```

### Compare mark analysis (get p-values and residuals)

```{r compare mark analysis}
# Number of samples with each combination of data
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$DNase)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$H3K27ac)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$DNase) & !is.na(metadata$H3K27ac)),])[1]
dim(metadata[which(!is.na(metadata$DNase) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$H3K27ac) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$DNase) & !is.na(metadata$H3K27ac) & !is.na(metadata$RNA)),])[1]

# Tables for each mark
aggregate(data=compare_marks_unique,TE_sample~DNase,function(x) sum(x)/sum(compare_marks_unique[which(!is.na(compare_marks_unique$DNase)),]$TE_sample))
aggregate(data=compare_marks_unique,TE_sample~H3K27ac,function(x) sum(x)/sum(compare_marks_unique[which(!is.na(compare_marks_unique$H3K27ac)),]$TE_sample))
aggregate(data=compare_marks_unique,TE_sample~RNA,function(x) sum(x)/sum(compare_marks_unique[which(!is.na(compare_marks_unique$RNA)),]$TE_sample))
aggregate(data=compare_marks_unique,TE_sample~WGBS,function(x) sum(x)/sum(compare_marks_unique[which(!is.na(compare_marks_unique$WGBS)),]$TE_sample))
aggregate(data=compare_marks_all,TE_sample_norm~chromHMM,function(x) sum(x)/sum(compare_marks_all$TE_sample_norm))

# Chi-squared for each mark combination
table_list = list(xtabs(TE_sample~H3K27ac+DNase,data=aggregate(data=compare_marks_unique,TE_sample~H3K27ac+DNase,sum)),  
                  xtabs(TE_sample~H3K27ac+WGBS,data=aggregate(data=compare_marks_unique,TE_sample~H3K27ac+WGBS,sum)), 
                  xtabs(TE_sample~DNase+WGBS,data=aggregate(data=compare_marks_unique,TE_sample~DNase+WGBS,sum)), 
                  xtabs(TE_sample~DNase+RNA,data=aggregate(data=compare_marks_unique,TE_sample~DNase+RNA,sum)), 
                  xtabs(TE_sample~RNA+WGBS,data=aggregate(data=compare_marks_unique,TE_sample~RNA+WGBS,sum)), 
                  xtabs(TE_sample~RNA+H3K27ac,data=aggregate(data=compare_marks_unique,TE_sample~RNA+H3K27ac,sum)), 
                  xtabs(TE_sample_norm~chromHMM+DNase,data=aggregate(data=compare_marks_all,TE_sample_norm~chromHMM+DNase,sum)), 
                  xtabs(TE_sample_norm~chromHMM+H3K27ac,data=aggregate(data=compare_marks_all,TE_sample_norm~chromHMM+H3K27ac,sum)),
                  xtabs(TE_sample_norm~chromHMM+RNA,data=aggregate(data=compare_marks_all,TE_sample_norm~chromHMM+RNA,sum)),
                  xtabs(TE_sample_norm~chromHMM+WGBS,data=aggregate(data=compare_marks_all,TE_sample_norm~chromHMM+WGBS,sum)))
chisq_list = lapply(table_list,function(x) chisq.test(x))
lapply(chisq_list,function(x) x$p.value)
lapply(chisq_list,function(x) x$residuals)

# Proportion of each state in other states
lapply(table_list,function(x) x/rowSums(x))
lapply(table_list,function(x) t(x)/rowSums(t(x)))

# Proportion of TEs x sample in each chromHMM state also overlapping DNase peak and H3K27ac peak
test1 = ddply(compare_marks_all[which(!is.na(compare_marks_all$DNase) & !is.na(compare_marks_all$H3K27ac)),],.(chromHMM,DNase,H3K27ac),summarise,TE_sample=sum(TE_sample),TE_sample_norm=sum(TE_sample_norm))
test2 = merge(ddply(test1,.(chromHMM),summarise,TE_sample=sum(TE_sample),TE_sample_norm=sum(TE_sample_norm)),ddply(test1[which(test1$DNase == "yes" & test1$H3K27ac == "yes"),],.(chromHMM),summarise,TE_sample=sum(TE_sample),TE_sample_norm=sum(TE_sample_norm)),by="chromHMM")
test2$TE_sample_proportion = test2$TE_sample.y/test2$TE_sample.x
test2$TE_sample_norm_proportion = test2$TE_sample_norm.y/test2$TE_sample_norm.x
test2

# Proportion of TEs x sample in each WGBS state also overlapping DNase peak and H3K27ac peak
test1 = ddply(compare_marks_all[which(!is.na(compare_marks_all$DNase) & !is.na(compare_marks_all$H3K27ac) & !is.na(compare_marks_all$WGBS)),],.(WGBS,DNase,H3K27ac),summarise,TE_sample=sum(TE_sample),TE_sample_norm=sum(TE_sample_norm))
test2 = merge(ddply(test1,.(WGBS),summarise,TE_sample=sum(TE_sample),TE_sample_norm=sum(TE_sample_norm)),ddply(test1[which(test1$DNase == "yes" & test1$H3K27ac == "yes"),],.(WGBS),summarise,TE_sample=sum(TE_sample),TE_sample_norm=sum(TE_sample_norm)),by="WGBS")
test2$TE_sample_proportion = test2$TE_sample.y/test2$TE_sample.x
test2$TE_sample_norm_proportion = test2$TE_sample_norm.y/test2$TE_sample_norm.x
```

### Supplementary Figure 4

**Supplementary Figure 4. Potential controls.** (A-B) Potential metrics with shuffled TEs (n=`NUM_TE`), 10 iterations. Actual TE values are indicated with red dot. (C-D) Potential metrics with unique Refseq promoters. (E-F) Expression potential metrics with unique Refseq exons. 

```{r Figure S4, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 2","setup","load libraries","set plot parameters","load functions","create essential matrices","load metadata","load essential matrices","define vectors","define facet labels","load feature sizes","calculate TE stats","generate other matrices")}
source("R_scripts/potential_shuffled.R")
source("R_scripts/potential_promoter.R")
source("R_scripts/potential_exon.R")

# Shuffled TEs
combine_boxplot$Iteration = rep("True",dim(combine_boxplot)[1])
combine_shuffled_boxplot = rbind(state_sample_count_shuffled[,c(1:2,4:5)],shuffled_WGBS_average_count,shuffled_DNase_potential_count[,2:5],shuffled_H3K27ac_potential_count[,2:5],combine_boxplot[which(combine_boxplot$State != "Expression"),])
combine_shuffled_boxplot$Iteration = factor(combine_shuffled_boxplot$Iteration,levels=c(seq(1,10,1),"True"))

combine_shuffled_stats = rbind(shuffled_chromHMM_potential_stats,shuffled_WGBS_average_stats,shuffled_DNase_potential_stats,shuffled_H3K27ac_potential_stats)
combine_shuffled_stats$Group = factor(c(rep("chromHMM",150),rep("WGBS",40),rep("DNase",10),rep("H3K27ac",10)),levels=c("chromHMM","WGBS","DNase","H3K27ac"))
combine_shuffled_stats = melt(combine_shuffled_stats,id.vars=c("State","Group","Iteration"))
colnames(combine_shuffled_stats)[4:5] = c("Metric","Value")

combine_stats_long = melt(combine_stats,id.vars=c("State","Group"))
colnames(combine_stats_long)[3:4] = c("Metric","Value")

a = ggplot(combine_shuffled_boxplot,aes(x=State,y=Proportion*100,fill=Iteration,color=Iteration)) + geom_boxplot(position="dodge",outlier.size=0.1) + scale_x_discrete(limits=all_state_labels[1:21]) + labs(y="% of TEs in state") + theme(axis.title.x=element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_fill_manual(values=c(rep("white",10),"white"),guide=FALSE) + scale_color_manual(values=c(rep("grey",10),"red"),guide=FALSE)

b = ggplot(combine_shuffled_stats,aes(x=State,y=Value)) + geom_boxplot() + theme(axis.title.y = element_blank(),axis.title.x = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_x_discrete(limits=c(chromHMM_states,meth_states,"DNase","H3K27ac")) + geom_point(data=combine_stats_long[which(combine_stats_long$State != "Expression"),],aes(x=State,y=Value),color="red",size=1) + facet_wrap(~Metric,nrow=3,labeller=labeller(Metric=metric_labels))

# Promoters
combine_cum_prom = rbind(chromHMM_promoter_state_cum,promoter_meth_average_category_cum,promoter_DNase_potential_cum,promoter_H3K27ac_potential_cum)
combine_stats_prom = rbind(chromHMM_promoter_state_dist_stats,promoter_meth_average_category_stats,promoter_DNase_potential_stats,promoter_H3K27ac_potential_stats)
combine_stats_prom$Group = factor(c(rep("chromHMM",15),rep("WGBS",4),"DNase","H3K27ac"),levels=c("chromHMM","WGBS","DNase","H3K27ac"))
combine_boxplot_prom = rbind(promoter_state_sample_count[,c(1:2,4)],promoter_meth_average_state_long,promoter_DNase_peaks_sample[,2:4],promoter_H3K27ac_peaks_sample[,2:4])

c = ggplot(combine_boxplot_prom,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors[1:21],guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="% of promoters in state") + coord_flip() + scale_x_discrete(limits = rev(levels(combine_boxplot_prom$State))) + geom_point(data=combine_stats_prom,aes(x=State,y=Proportion_ever),color="red",size=2)

d = ggplot(combine_cum_prom,aes(x=Sample_proportion,y=Proportion,color=State,group=State)) + geom_point(size=0.1) + geom_line(size=1.5) + labs(x="Proportion of all samples",y="Cumulative proportion of promoters in state") + scale_color_manual(values=all_state_colors[1:21],guide=FALSE) + geom_vline(xintercept = combine_stats_prom$Samples_avg_ever/100,color=all_state_colors[as.vector(combine_stats_prom$State)])

# Exons
e = ggplot(RNA_exon_sample,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values="black",guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="% of exons in state") + coord_flip() + scale_x_discrete(labels="RPKM > 1") + geom_point(data=RNA_potential_exon_stats,aes(x=State,y=Proportion_ever),color="red",size=3) + scale_y_continuous(limits=c(0,100))

f = ggplot(RNA_potential_exon_cum,aes(x=Sample_proportion,y=Proportion,color=State,group=State)) + geom_point(size=0.1) + geom_line(size=1.5) + labs(x="Proportion of all samples (n=52)",y="Cumulative proportion of exons RPKM > 1") + scale_color_manual(values="black",guide=FALSE) + geom_vline(xintercept = RNA_potential_exon_stats$Samples_avg_ever/100,color="black")

legend = get_legend(ggplot(combine_boxplot,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors) + theme(legend.direction = "horizontal",legend.title = element_blank(),legend.position="bottom"))

grid.arrange(a,b,c,d,e,f,legend, nrow = 5, ncol=2, layout_matrix=cbind(c(1,3,3,4,4),c(2,5,6,6,7)),heights=c(0.5,0.1,0.15,0.1,0.15))
```

### Remove shuffled matrices

```{r}
rm(list=c("shuffled_WGBS_average","shuffled_DNase_potential","shuffled_H3K27ac_potential"))
```

### Potential control statistics

```{r potential control analysis}
ddply(combine_shuffled_boxplot,.(State,Iteration),summarise,Mean=mean(Proportion),SD=sd(Proportion))
ddply(combine_shuffled_stats,.(State,Metric),summarise,Mean=mean(Value),SD=sd(Value))
combine_stats_prom
RNA_potential_exon_stats
mean(RNA_exon_sample$Proportion)

# Number of unique promoters and exons
dim(chromHMM_promoter_state)[1]
dim(RNA_exon_unique)[1]

# Promoter CpG stats
promoter_CpG_count = read.table("WGBS/Refseq_promoters/refseq_promoter_unique_CpG_count.txt",sep='\t')
dim(promoter_CpG_count)[1] # Number of promoters with CpGs
dim(promoter_CpG_count)[1]/dim(chromHMM_promoter_state)[1]
min(promoter_CpG_count$V5)
median(promoter_CpG_count$V5)
mean(promoter_CpG_count$V5)
max(promoter_CpG_count$V5)
mean(c(promoter_CpG_count$V5,rep(0,(dim(chromHMM_promoter_state)[1]-dim(promoter_CpG_count)[1]))))
median(c(promoter_CpG_count$V5,rep(0,(dim(chromHMM_promoter_state)[1]-dim(promoter_CpG_count)[1]))))

# Exon CpG stats
exon_CpG_count = read.table("WGBS/refseq_exons_unique_CpG_count.txt",sep='\t')
dim(exon_CpG_count)[1] # Number of exons with CpGs
dim(exon_CpG_count)[1]/dim(RNA_exon_unique)[1]
min(exon_CpG_count$V5)
median(exon_CpG_count$V5)
mean(exon_CpG_count$V5)
max(exon_CpG_count$V5)
mean(c(exon_CpG_count$V5,rep(0,(dim(RNA_exon_unique)[1]-dim(exon_CpG_count)[1]))))
median(c(exon_CpG_count$V5,rep(0,(dim(RNA_exon_unique)[1]-dim(exon_CpG_count)[1]))))

# % TEs overlapping CpGs
NUM_TE_WGBS
NUM_TE_WGBS/NUM_TE
NUM_TE-NUM_TE_WGBS
min(TE_meth_average$CpGs)
median(TE_meth_average$CpGs)
max(TE_meth_average$CpGs)

# Shuffled TE CpG stats
load("R_datasets/shuffled_TEs_CpG.RData")

ldply(rmsk_TE_shuffled,summarise,TE_wCpG=length(CpGs[which(CpGs > 0)]),TE_wCpG_per=length(CpGs[which(CpGs > 0)])/length(CpGs),Median_all=median(CpGs),Mean_all=mean(CpGs),Median_wCpG=median(CpGs[which(CpGs > 0)]),Mean_wCpG=mean(CpGs[which(CpGs > 0)]))

rm(rmsk_TE_shuffled)
```

### Exon expression control analysis

```{r exon control analysis}
# Average number of samples an exon is expressed RPKM > 0
mean(apply(RNA_exon_unique[,5:56],1,function(x) sum(x > 0)))

table(RNA_exon_unique$Expressed_samples)
table(RNA_exon_unique$Expressed_samples)/dim(RNA_exon_unique)[1]

# Maximum expression
#hist(RNA_exon_unique$Max_expression)
median(RNA_exon_unique$Max_expression)
median(RNA_exon_unique[which(RNA_exon_unique$Expressed_samples > 0),]$Max_expression)
max(RNA_exon_unique$Max_expression)
tail(RNA_exon_unique[order(RNA_exon_unique$Max_expression),],n=20)

# Average TE vs. exon expression
TE_express = unlist(RNA_TE_agnostic[,9:60])
median(TE_express)
median(TE_express[which(TE_express > 1)])
rm(TE_express)
exon_express = unlist(RNA_exon_unique[,5:56])
median(exon_express)
median(exon_express[which(exon_express > 1)])
rm(exon_express)
```

### Supplementary Figure 5

**Supplementary Figure 5. Additional potential figures.** (A) Proportion of TEs with CpGs (n=`NUM_TE_WGBS`) in each methylation state by sample. Ordered by decreasing proportion of TEs intermediately methylated or hypomethylated. (B) For TEs in a methylation state, overlapping DNase or H3K27ac peaks, or expressed RPKM >1 in at least one sample, cumulative proportion of individual TEs in that state in at least that proportion of samples. Horizontal lines indicate the mean proportion of samples a TE is in each state if it is in that state in at least one sample. (C) Boxplots indicate the proportion of all TEs in the state per sample, excluding five cancer cell lines and IMR90 (chromHMM n=121, WGBS n=36, DNase n=48, H3K27ac n=92, RNA n=48).  Red dots are the fraction of all TEs in that state in any sample, excluding those six samples. (D) Average proportion of samples a TE is in a state, excluding cancer cell lines and IMR90. Total is >100% for chromHMM due to TEs in multiple states in a single sample.

```{r Figure S5, echo=FALSE}
# source("R_scripts/potential.R")

combine_cum = rbind(TE_meth_average_category_cum,TE_DNase_potential_cum,TE_H3K27ac_potential_cum,RNA_potential_cum)
combine_boxplot_noCancer_IMR90 = droplevels(combine_boxplot[which(metadata[match(combine_boxplot$Sample,metadata$Sample),]$Exclude == "Include"),])
combine_stats_noCancer_IMR90 = rbind(chromHMM_TE_state_dist_noCancer_stats,TE_meth_average_noIMR90_category_stats,TE_DNase_potential_noCancer_stats,TE_H3K27ac_potential_noCancer_stats,RNA_potential_noCancer_stats)
combine_stats_noCancer_IMR90$Group = factor(c(rep("chromHMM",15),rep("WGBS",4),"DNase","H3K27ac","Expression"),levels=c("chromHMM","WGBS","DNase","H3K27ac","Expression"))

a = ggplot(TE_meth_average_state_long,aes(x=Sample,y=Proportion,fill=State)) + geom_bar(stat="identity") + theme(axis.text.x=element_text(angle=90),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.title.x=element_blank()) + scale_fill_manual(values=meth_colors,guide=FALSE) + ylab("Proportion of TEs in methylation state")

b = ggplot(combine_cum,aes(x=Sample_proportion,y=Proportion,color=State,group=State)) + geom_point(size=0.1) + geom_line(size=1.5) + labs(x="Proportion of samples",y="Cumulative proportion of TEs in state") + scale_color_manual(values=all_state_colors[16:22],guide=FALSE) + geom_vline(xintercept = combine_stats$Samples_avg_ever[16:22]/100,color=all_state_colors[as.vector(combine_stats$State)[16:22]]) 

c = ggplot(combine_boxplot_noCancer_IMR90,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="% of TEs in state") + coord_flip() + scale_x_discrete(limits = rev(levels(combine_boxplot_noCancer_IMR90$State)),labels=rev(all_state_labels)) + geom_point(data=combine_stats_noCancer_IMR90,aes(x=State,y=Proportion_ever),color="red",size=3)

d = ggplot(combine_stats_noCancer_IMR90,aes(x=1,y=Samples_avg_all,fill=State)) + geom_bar(stat="identity",show.legend=FALSE) + scale_fill_manual(values=all_state_colors) + theme(axis.title.x=element_blank(),axis.ticks.x=element_blank(),axis.text.x=element_blank()) + ylab("Average % of samples in state") + facet_wrap(~Group,nrow=1,labeller=labeller(Group = group_labels))

legend = get_legend(ggplot(combine_boxplot,aes(x=State,y=Proportion*100,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors) + theme(legend.direction = "horizontal",legend.title = element_blank(),legend.position="bottom"))

grid.arrange(a,b,c,d,legend, nrow = 3,layout_matrix=rbind(c(1,2),c(3,4),c(5,5)),heights=c(0.4,0.4,0.2))
```

### Potential no cancer analysis

```{r potential no cancer}
combine_stats_noCancer_IMR90
```

### Average TE methylation by sample

```{r analysis Fig S5a}
# Average TE methylation level by sample
TE_meth_average_mean = apply(TE_meth_average[,c(8:44)],2,function(x) mean(na.omit(x)))
TE_meth_average_mean[which(TE_meth_average_mean < (mean(TE_meth_average_mean)-2*sd(TE_meth_average_mean)))]

# Summaries with/without IMR90
mean(TE_meth_average_mean)
sd(TE_meth_average_mean)
mean(TE_meth_average_mean[c(1:10,12:37)])
sd(TE_meth_average_mean[c(1:10,12:37)])
```

### Average proportion of TEs in state by sample

```{r analysis Fig S5b}
# Average proportion of TEs in each state per sample, no IMR90
ddply(TE_meth_average_state_long,.(State),summarise,Mean=mean(Proportion),SD=sd(Proportion),Outlier=mean(Proportion)+2*sd(Proportion))
TE_meth_average_state_long
ddply(TE_meth_average_state_long[which(TE_meth_average_state_long$Sample != "E017"),],.(State),summarise,Mean=mean(Proportion),SD=sd(Proportion))

# Proportion of TEs in each state, IMR90
TE_meth_average_state[1,]
```

### Supplementary Figure 6

**Supplementary Figure 6. State switching for chromHMM states.** (A) Total number of chromHMM states per TE across all samples (blue) and maximum number of chromHMM states per TE in one sample (red). (B) By class. (C) Total number of WGBS states per TE across all samples. (D) By class. (E) Proportion of instances where two chromHMM states are co-annotated on the same TE in the same sample (n=4,430,788 x 127). Rows normalized by total number of instances of TEs in that state in one sample (diagonal). Color is the proportion of total instances in the first state (rows) co-annotated with the second state (columns). (F) Proportion of TEs (n=4,430,788) in two chromHMM states in different samples. Rows normalized by number of TEs in that state in at least one sample. Color is the proportion of TEs in one state (rows) that are in the second state (columns) in a different sample. (G) By class. (H) (F) for WGBS. (I) (H) by class.

```{r Figure S6, echo=FALSE}
#source("R_scripts/potential.R")
source("R_scripts/state_switching.R")

# chromHMM states
a = ggplot(chromHMM_TE_state,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + geom_histogram(data = chromHMM_TE_state,aes(x=Max_states_intra),binwidth=1,fill="darkred",color="grey",alpha=0.5) + ylab("Number of TEs") + xlab("Number of chromHMM states")

# chromHMM states by class 
b = ggplot(chromHMM_TE_state,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + scale_x_continuous(limits=c(0,15)) + geom_histogram(data = chromHMM_TE_state,aes(x=Max_states_intra),binwidth=1,fill="darkred",color="grey",alpha=0.5) + ylab("Number of TEs") + xlab("Number of chromHMM states") + facet_wrap(~class_update,scales="free")

# WGBS states
TE_meth_average$States = apply(TE_meth_average[,46:49],1,function(x) sum(x > 0))
TE_meth_average$States_noIMR90 = apply(TE_meth_average[,50:53],1,function(x) sum(x > 0))

c = ggplot(TE_meth_average,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + ylab("Number of TEs") + xlab("Number of methylation states")

# WGBS states by class
d = ggplot(TE_meth_average,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + scale_x_continuous(limits=c(0,4)) + ylab("Number of TEs") + xlab("Number of methylation states") + facet_wrap(~class_update,scales="free")

# State switching, chromHMM
e = ggplot(melt(as.matrix(state_switching_intra)),aes(x=Var2,y=Var1,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(labels=rev(chromHMM_states),limits=rev(levels(melt(as.matrix(state_switching_intra))$Var1))) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1),guide=FALSE)

f = ggplot(melt(as.matrix(state_switching_inter)),aes(x=Var2,y=Var1,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(labels=rev(chromHMM_states),limits=rev(levels(melt(as.matrix(state_switching_inter))$Var1))) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1),guide=FALSE)

# State switching, chromHMM by class 
g = ggplot(melt(ss_inter_class,id.vars=c("Class","State")),aes(x=variable,y=State,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(limits=rev(chromHMM_states)) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1),guide=FALSE) + facet_wrap(~Class)

# State switching, WGBS
h = ggplot(melt(as.matrix(ss_inter_meth)),aes(x=Var2,y=Var1,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_y_discrete(limits=rev(levels(melt(as.matrix(ss_inter_meth))$Var1))) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1))

i = ggplot(melt(ss_inter_meth_class,id.vars=c("Class","State")),aes(x=variable,y=State,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_y_discrete(limits=rev(meth_states)) + scale_fill_gradient2(low="darkblue",mid="white",high="darkred",midpoint=0.5,limits=c(0,1),guide=FALSE) + facet_wrap(~Class)

grid.arrange(a,b,c,d,e,f,g,h,i, nrow=4, layout_matrix=rbind(c(1,2,2),c(3,4,4),c(5,6,8),c(7,7,9)),heights=c(0.2,0.2,0.3,0.3))
```

### State switching analysis

```{r state switching analysis}
state_switching_intra
state_switching_inter

# Number of TEs ever in each state
apply(chromHMM_TE_state_dist[,2:16],2,function(x) sum(x[2:128]))
```

### Shuffled number of states (not shown)

```{r shuffled number of states, eval=FALSE}
source("R_scripts/potential_shuffled.R")

# Shuffled chromHMM
shuffled_chromHMM_states = ldply(shuffled_chromHMM_potential,function(x) x[,23:24])
shuffled_chromHMM_states$Iteration = factor(rep(seq(1,10,1),each=4430788))

ggplot(shuffled_chromHMM_states,aes(x=States,fill=Iteration)) + geom_histogram(binwidth=1,alpha=0.5,position="dodge") + ylab("Number of TEs") + xlab("Number of chromHMM states")
ggplot(shuffled_chromHMM_states,aes(x=Max_states_intra,fill=Iteration)) + geom_histogram(binwidth=1,alpha=0.5,position="dodge") + ylab("Number of TEs") + xlab("Number of chromHMM states")

# Shuffled WGBS
shuffled_WGBS_average = lapply(shuffled_WGBS_average,function(x) transform(x,States = apply(x[,46:49],1,function(y) sum(y > 0))))
shuffled_WGBS_average = lapply(shuffled_WGBS_average,function(x) transform(x,States_noIMR90 = apply(x[,50:53],1,function(y) sum(y > 0))))
shuffled_WGBS_states = ldply(shuffled_WGBS_average,function(x) x[,54:55])
shuffled_WGBS_states$Iteration = factor(unlist(lapply(seq(1,10,1),function(y) rep(y,unlist(lapply(shuffled_WGBS_average,function(x) dim(x)[1]))[y]))))

ggplot(shuffled_WGBS_states,aes(x=States,fill=Iteration)) + geom_histogram(binwidth=1,alpha=0.5,position="dodge") + ylab("Number of shuffled TEs") + xlab("Number of methylation states")
```

### Number of states per TE x sample

```{r states per TE-sample}
# chromHMM
test = merge(read.table("chromHMM/all_chromHMM_other_state_counts.txt",sep='\t'),read.table("chromHMM/all_chromHMM_TE_state_counts.txt",sep='\t'),by="V1")
test$Total = test$V2.x+test$V2.y
test$Proportion = test$Total/sum(test$Total)
test
sum(test$Proportion[2:15])
# Proportion of those with >1 state in only 2 states
test$Proportion[2]/sum(test$Proportion[2:15])

# WGBS
test = read.table("WGBS/TE_CpG_state_counts.txt",sep='\t')
test$Proportion = test$V2/sum(test$V2)
test
sum(test$Proportion[2:4])
# Proportion of those with >1 state in only 2 states
test$Proportion[2]/sum(test$Proportion[2:4])
```

### Stats for number of chromHMM and WGBS states per TE, overall and by class

```{r analysis Fig S6a}
mean(chromHMM_TE_state$Max_states_intra)
sd(chromHMM_TE_state$Max_states_intra)
ddply(chromHMM_TE_state,.(class_update),summarise,Mean=mean(Max_states_intra),SD=sd(Max_states_intra))

mean(chromHMM_TE_state$States)
sd(chromHMM_TE_state$States)
ddply(chromHMM_TE_state,.(class_update),summarise,Mean=mean(States),SD=sd(States))

mean(TE_meth_average$States)
sd(TE_meth_average$States)
ddply(TE_meth_average,.(class_update),summarise,Mean=mean(States),SD=sd(States))

# TEs with >1 max states in a sample
dim(chromHMM_TE_state[which(chromHMM_TE_state$Max_states_intra > 1),])[1]
dim(chromHMM_TE_state[which(chromHMM_TE_state$Max_states_intra > 1),])[1]/NUM_TE
```

### Correlation of TE length with number of states ever, max in one sample

```{r analysis Fig S6b}
cor.test((chromHMM_TE_state$stop-chromHMM_TE_state$start),chromHMM_TE_state$States,method="spearman")
cor.test(chromHMM_TE_state$stop-chromHMM_TE_state$start,chromHMM_TE_state$Max_states_intra,method="spearman")
```

### TEs always in single state (update for chrY!)

```{r analysis Fig S6c}
# Methylation states
apply(TE_meth_average[,46:49],2,function(x) dim(TE_meth_average[which(x == 37),])[1])
apply(TE_meth_average[,46:49],2,function(x) dim(TE_meth_average[which(x == 37),])[1])/NUM_TE_WGBS
apply(TE_meth_average[,46:49],2,function(x) table(TE_meth_average[which(x == 37),]$class_update))
apply(TE_meth_average[,50:53],2,function(x) dim(TE_meth_average[which(x == 36),])[1])
apply(TE_meth_average[,50:53],2,function(x) dim(TE_meth_average[which(x == 36),])[1])/NUM_TE_WGBS

# DNase/H3K27ac
dim(TE_DNase_peaks[which(TE_DNase_peaks$Samples == 53),])[1]
dim(TE_DNase_peaks[which(TE_DNase_peaks$Samples == 53),])[1]/NUM_TE
dim(TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$Samples == 98),])[1]
dim(TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$Samples == 98),])[1]/NUM_TE

# chromHMM states
apply(chromHMM_TE_state[,8:22],2,function(x) dim(chromHMM_TE_state[which(x == chromHMM_TE_state$Tissues),])[1])
sum(apply(chromHMM_TE_state[,8:22],2,function(x) dim(chromHMM_TE_state[which(x == chromHMM_TE_state$Tissues),])[1])/NUM_TE)

# 5_TxWk
chromHMM_TE_state[which(chromHMM_TE_state$X5_TxWk == chromHMM_TE_state$Tissues),c(1:7,26)]

# Shuffled TE control
lapply(shuffled_chromHMM_potential,function(x) dim(x[which(x$X5_TxWk >= 121 & x$States == 1),]))
lapply(shuffled_chromHMM_potential,function(x) x[which(x$X5_TxWk >= 121 & x$States == 1),])
rm(shuffled_chromHMM_potential)

# 15_Quies (particular subfamilies or families)
test = merge(rmsk_TE_subfamily[,c(1:4,11)],aggregate(data=chromHMM_TE_state[which(chromHMM_TE_state$X15_Quies == chromHMM_TE_state$Tissues),c(1:7,26)],start~subfamily,length),by="subfamily",all.x=TRUE)
test[is.na(test)] = 0
test$quiescent = log2((test$start/sum(test$start))/(test$Count/NUM_TE))
dim(test[which(test$start > 0),][1]) # Number of subfamilies with a member always quiescent
tail(test[order(test$quiescent),],n=40)
rm(test)

test = merge(aggregate(data=rmsk_TE_subfamily[,c(2:3,4)],Count~family+class_update,sum),aggregate(data=chromHMM_TE_state[which(chromHMM_TE_state$X15_Quies == chromHMM_TE_state$Tissues),c(1:7,26)],start~family,length),by="family",all.x=TRUE)
test[is.na(test)] = 0
test$quiescent = log2((test$start/sum(test$start))/(test$Count/NUM_TE))
dim(test[which(test$start > 0),][1]) # Number of families with a member always quiescent
tail(test[order(test$quiescent),],n=44)
rm(test)
```

### Supplementary Figure XX 

Correlation of number of samples in state with other states

```{r Figure SXX, eval=FALSE}
# Requires TE_correlation.R
library(Hmisc)

TE_state_correlation = rcorr(as.matrix(rmsk_TE_measure[,35:63]),type="spearman")
TE_state_correlation = merge(merge(melt(TE_state_correlation$r),melt(TE_state_correlation$P),by=c("Var1","Var2")),melt(TE_state_correlation$n),by=c("Var1","Var2"))
colnames(TE_state_correlation) = c("State1","State2","Correlation","Pvalue","n")
TE_state_correlation$Padj = p.adjust(TE_state_correlation$Pvalue,method="bonf")

TE_state_correlation_filter = TE_state_correlation[which(TE_state_correlation$Padj < 0.05 & TE_state_correlation$State1 %in% levels(TE_state_correlation$State1)[c(1:15,18:21,26:28)] & TE_state_correlation$State2 %in% levels(TE_state_correlation$State1)[c(1:15,18:21,26:28)]),]

detach(package:Hmisc,unload=TRUE)

ggplot(TE_state_correlation_filter,aes(x=State1,y=State2,fill=Correlation)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_x_discrete(limits=levels(TE_state_correlation_filter$State1)[c(1:15,18,20,19,21,26:28)],labels=as.vector(all_state_labels)) + scale_y_discrete(limits=levels(TE_state_correlation_filter$State2)[c(1:15,18,20,19,21,26:28)],labels=as.vector(all_state_labels))
```

## Factors influencing potential

### Figure 3

**Figure 3. Correlation of TE activity with phylogeny and TE features.** (A) Total proportion of each TE class in each state across all samples. (B) Average proportion of samples a TE is in a state, by class. Ordered by decreasing range. Coefficient of variation in parentheses. (C) Difference in proportion of samples a TE is in a state by overlap with RefSeq genic features, normalized by overall average. (D) TE subfamily age (Jukes-Cantor evolutionary distance) vs. average number of CpGs per kbp. (E) TE subfamily age (Jukes-Cantor evolutionary distance) vs. range of proportion of TEs hypomethylated per sample (n=37). 

```{r Figure 3, fig.width=7, fig.height=7, echo=FALSE, cache=TRUE, dependson=c("Figure 1","Figure 2","Figure S6","setup","load libraries","set plot parameters","load functions","create essential matrices","load metadata","load essential matrices","define vectors","define facet labels","load feature sizes","calculate TE stats","generate other matrices")}
#source("R_scripts/potential_class.R") #Run in state_switching.R
#source("R_scripts/potential.R")
source("R_scripts/TE_correlation.R")
source("R_scripts/feature_overlap_state.R") #Requires potential.R
source("R_scripts/state_metric_correlation.R")

# Class potential
combine_stats_class = rbind(chromHMM_TE_state_class_stats,TE_meth_average_class_stats,potential_TE_DNase_class_stats,potential_TE_H3K27ac_class_stats,RNA_potential_class_stats)
combine_stats_class$Group = factor(c(rep("chromHMM",90),rep("WGBS",24),rep("DNase",6),rep("H3K27ac",6),rep("Expression",6)),levels=c("chromHMM","WGBS","DNase","H3K27ac","Expression"))
state_order = ddply(combine_stats_class,"State",summarise,CV=sd(Samples_avg_all)/mean(Samples_avg_all), Range=max(Samples_avg_all)-min(Samples_avg_all))

a = ggplot(combine_stats_class,aes(x=State,y=Samples_avg_all)) + geom_point(aes(color=class_update),size=4) + scale_color_manual(values=class_colors,name="Class") + scale_x_discrete(limits=state_order[order(state_order$Range),]$State,labels=all_state_labels) + scale_y_continuous(limits=c(0,100)) + theme(axis.title.y = element_blank()) + ylab("Average % samples in state") + geom_text(data=state_order,aes(x=State,y=100,label=paste("(",round(CV,2),")",sep=""))) + coord_flip() 

# Feature table
b = ggplot(feature_state_mean[which(feature_state_mean$Coding == "All"),],aes(x=State,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1)) + scale_fill_gradient2(name="Average % increase",low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-1,50)) + scale_y_discrete(limits=rev(levels(feature_state_mean$Feature)),labels=genic_labels) + scale_x_discrete(labels=all_state_labels)

# By-chromosome
chromosome_potential = aggregate(data=rmsk_TE_measure[,c(1,35:49,52:55,60:62)],.~chromosome,mean)
colnames(chromosome_potential) = c("chromosome",names(all_state_labels))
chromosome_potential_Zscore = as.data.frame(apply(chromosome_potential[,2:23],2,function(x) (x-mean(x))/sd(x)))
chromosome_potential_Zscore$chromosome = factor(chromosome_potential$chromosome,levels=standard_chromosomes)
chromosome_potential_long = melt(chromosome_potential_Zscore,id.var="chromosome")
colnames(chromosome_potential_long)[2:3] = c("State","Mean_samples")

c = ggplot(chromosome_potential_long,aes(x=State,y=Mean_samples,color=chromosome,label=chromosome)) + geom_point() + geom_text(angle=45,hjust=0,vjust=0,size=3) + scale_color_discrete(guide=FALSE) + scale_x_discrete(limits=rev(names(all_state_labels)),labels=all_state_labels) + theme(axis.title.y = element_blank(),panel.grid.major.y=element_line(color="grey")) + ylab("Z-score, mean samples in state") + coord_flip() + geom_hline(yintercept=2) + geom_hline(yintercept=-2)

# SINE/Alu age vs. TE metrics and number of samples in methylation states
rmsk_TE_measure$cpgIsland_binary = as.vector(rmsk_TE_measure$cpgIsland)
rmsk_TE_measure$cpgIsland_binary = gsub("no",0,rmsk_TE_measure$cpgIsland_binary)
rmsk_TE_measure$cpgIsland_binary = gsub("yes",1,rmsk_TE_measure$cpgIsland_binary)
rmsk_TE_measure$cpgIsland_binary = as.numeric(rmsk_TE_measure$cpgIsland_binary)

rmsk_TE_measure_SINE = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$class_update == "SINE"),c(1:10,12:14,52:55,65)])
rmsk_TE_measure_SINE$Length = rmsk_TE_measure_SINE$Length/1000

d = ggplot(rmsk_TE_measure_SINE,aes(x=JC_distance)) + geom_density(fill=class_colors["SINE"]) + theme(axis.title.x = element_blank(),axis.title.y = element_blank())

e = ggplot(melt(rmsk_TE_measure_SINE,id.vars=colnames(rmsk_TE_measure_SINE)[c(1:11,13,18)]),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=setNames(c("#000000",meth_colors),c("CpGs",names(meth_colors))),guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank()) 

f = ggplot(melt(rmsk_TE_measure_SINE,id.vars=colnames(rmsk_TE_measure_SINE)[c(1:7,9,11:12,14:18)])) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + geom_smooth(data=rmsk_TE_measure_SINE,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + theme(axis.title.y = element_blank())

rmsk_TE_measure_Alu = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$family == "Alu"),c(1:10,12:14,52:55,65)])
rmsk_TE_measure_Alu$Length = rmsk_TE_measure_Alu$Length/1000

g = ggplot(rmsk_TE_measure_Alu,aes(x=JC_distance)) + geom_density(fill="lightblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank())

h = ggplot(melt(rmsk_TE_measure_Alu,id.vars=colnames(rmsk_TE_measure_Alu)[c(1:11,13,18)]),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=setNames(c("#000000",meth_colors),c("CpGs",names(meth_colors))),guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank()) 

i = ggplot(melt(rmsk_TE_measure_Alu,id.vars=colnames(rmsk_TE_measure_Alu)[c(1:7,9,11:12,14:18)])) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + theme(axis.title.y = element_blank())

legend2 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=colnames(rmsk_TE_measure_Alu)[c(1:11,13,18)]),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=setNames(c("#000000",meth_colors),c("CpGs",names(meth_colors))),labels=c("CpGs/TE",meth_states[c(1,3,2,4)])) + theme(legend.title = element_blank(),legend.direction = "horizontal"))
  
legend3 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=colnames(rmsk_TE_measure_Alu)[c(1:7,9,11:12,14:18)])) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),labels=c("Length (kbp)","Mappability","CpGs/bp")) + ylab("Property") + theme(legend.title = element_blank(),legend.direction = "horizontal") + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")))

grid.arrange(a,b,c,d,e,f,g,h,i,legend2,legend3,nrow=6,layout_matrix=rbind(c(1,1,2,2),c(3,3,4,7),c(3,3,5,8),c(3,3,6,9),c(3,3,10,10),c(3,3,11,11)),heights=c(0.5,0.1,0.2,0.2,0.05,0.05),widths=c(0.2,0.2,0.3,0.3))
```

### Figure 3 analysis

```{r analysis Fig 3}
# By-class proportion
dcast(combine_class_proportion,State~class,value.var="Proportion")

# Fig 3A
combine_stats_class
state_order

# Number of samples in state by class, p-values
apply(rmsk_TE_measure[,35:63],2,function(x) unlist(kruskal.test(x~rmsk_TE_measure$class_update))["p.value"])

# Fig 3B
dcast(feature_state_mean[which(feature_state_mean$Coding == "All"),],Feature~State,value.var = "Enrichment")

# Fig 3C
lapply(colnames(chromosome_potential[,2:22]),function(x) {test = chromosome_potential[,c("chromosome",x)]; test$Zscore = (test[[x]]-mean(test[[x]]))/sd(test[[x]]); test = test[which(abs(test$Zscore) > 2),]; test[order(test$Zscore),]})

# Number of TEs on each chromosome
table(rmsk_TE$chromosome)
sort(table(rmsk_TE$chromosome)/NUM_TE)

# Proportion of TEs in each state on chr19
apply(rmsk_TE_measure[,c(35:49,52:55,60:62)],2,function(x) table(rmsk_TE_measure[which(x > 0),]$chromosome)/dim(rmsk_TE_measure[which(x > 0),])[1])

# Fig 3D

# Significant correlations between number of samples in state and metric
correlate_TE_state[which(correlate_TE_state$p.value < 0.001 & abs(correlate_TE_state$estimate.rho) > 0.3 & !(correlate_TE_state$State %in% c("Max_states_intra","Hypomethylated_noIMR90","Intermediate_noIMR90","Hypermethylated_noIMR90","Missing_noIMR90"))),]

# Overlap with CpG islands by age (logistic regression) (all, SINE, Alu)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure,family="binomial")
summary(test)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure_SINE,family="binomial")
summary(test)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure_Alu,family="binomial")
summary(test)
```

### Number of samples in state by genic feature, p-values (Too long!)

```{r anaylsis Fig3b, eval=FALSE}
TE_pvalue_all_state = apply(rmsk_TE_measure[,15:34],2,function(x) apply(rmsk_TE_measure[,35:63],2,function(y) unlist(wilcox.test(y~x))["p.value"]))
TE_pvalue_class_state = ddply(rmsk_TE_measure,~class_update,function(z) apply(z[,15:34],2,function(x) apply(z[,35:63],2,function(y) unlist(wilcox.test(y~x))["p.value"])))
```

### Supplementary Figure 7

**Supplementary Figure 7. Potential for an individual TE to be in an epigenetic state, by class.** (A) Proportion of TEs in each state in any sample, by class (DNA n = 456,948, LINE n = 1,480,369, LTR n = 708,210, SINE n = 1,769,839, SVA n = 3,608, Other n = 11,814; TEs with CpGs only; DNA n = 273,870, LINE n = 948,864, LTR n = 531,389, SINE n = 1,427,517, SVA n = 3,519, Other n = 6,525). (B) Average proportion of samples in which a TE is in each state, by class. (C) For TEs in a state in at least one sample, average proportion of samples the TE is in that state, by class.

```{r Figure S7, echo=FALSE}
#source("R_scripts/potential_class.R")

combine_boxplot_class = rbind(class_state_sample[,c(1:3,5)],TE_meth_average_state_class,TE_DNase_peaks_class,TE_H3K27ac_peaks_class,RNA_RPKM_class)

a = ggplot(combine_boxplot_class,aes(x=State,y=Proportion*100,fill=Class)) + geom_boxplot() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y = element_blank()) + coord_flip() + labs(y="% of TEs in state by sample") + scale_x_discrete(labels=rev(all_state_labels),limits=rev(levels(combine_boxplot_class$State)))

b = ggplot(melt(combine_stats_class),aes(x=State,y=value,fill=class_update)) + geom_bar(stat="identity",position="dodge") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1),axis.title.y=element_blank()) + scale_x_discrete(labels=all_state_labels) + scale_y_continuous(limits=c(0,100)) + facet_wrap(~variable,nrow=3,labeller=labeller(variable=metric_labels))

#c = ggplot(combine_stats_class,aes(x=class_update,y=Samples_avg_all,fill=State)) + geom_bar(stat="identity",show.legend=FALSE) + scale_fill_manual(values=all_state_colors) + theme(axis.title.x=element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + ylab("Average % of samples in state") + facet_wrap(~Group,nrow=1,scales="free",labeller=labeller(Group = group_labels))

legend1 = get_legend(ggplot(combine_stats_class,aes(x=State,y=Samples_avg_ever,fill=class_update)) + geom_bar(stat="identity",position="dodge") + scale_fill_manual(values=class_colors,name="Class") + theme(legend.position="bottom",legend.direction="horizontal"))

#legend2 = get_legend(ggplot(combine_stats_class,aes(x=class_update,y=Samples_avg_all,fill=State)) + geom_bar(stat="identity") + scale_fill_manual(values=all_state_colors) + theme(legend.title=element_blank(),legend.position="bottom",legend.direction = "horizontal"))

grid.arrange(a,b,legend1,nrow=2,layout_matrix=rbind(c(1,2),c(3,3)),heights=c(0.9,0.1))
```

### Max expression by class

```{r analysis Fig S7}
merge(ddply(RNA_TE_agnostic,.(class_update),summarise,Proportion_expressed=sum(Expressed_samples > 0)/length(Expressed_samples),Max_median=median(Max_expression)),ddply(RNA_TE_agnostic[which(RNA_TE_agnostic$Expressed_samples > 0),],.(class_update),summarise,Max_median_expressed=median(Max_expression)),by="class_update")
```

### Supplementary Figure 8

**Supplementary Figure 8. Epigenetic state by feature overlap.** (A) Proportion of TEs in each class overlapping RefSeq features. TEs may overlap multiple features. (B) Difference in proportion of samples a TE is in a state by overlap with RefSeq genic features, normalized by overall average, including features split by protein-coding/non-coding. (C) By class. 

```{r Figure S8, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 3","setup","load libraries","set plot parameters","load functions","create essential matrices","load metadata","load essential matrices","define vectors","define facet labels","load feature sizes","calculate TE stats","generate other matrices")}
#source("R_scripts/feature_overlap_state.R")
source("R_scripts/shuffled_feature_overlap.R")

rmsk_TE_class_feature = melt(rmsk_TE_class[,c(1,12:30)],id.vars="class_update")
colnames(rmsk_TE_class_feature) = c("Class","Cohort","Proportion")
rmsk_TE_class_feature = split_coding(rmsk_TE_class_feature,2)

a = ggplot(rmsk_TE_class_feature,aes(x=Class,y=Proportion,fill=Feature)) + geom_bar(stat="identity") + theme(axis.title.x = element_blank(),legend.title=element_blank()) +  ylab("Proportion of TEs overlapping feature") + scale_fill_manual(values=c(brewer.pal(8,"Spectral")),labels=genic_labels[c(8,1:7)]) + facet_wrap(~Coding,nrow=1,labeller=labeller(Coding=coding_labels))

feature_proportion = apply(rmsk_TE[,13:31],2,function(x) length(na.omit(x)))
feature_proportion_long = melt(as.matrix(feature_proportion/NUM_TE))
colnames(feature_proportion_long) = c("Feature","X","Proportion")

#b = ggplot(feature_proportion_shuffled_long,aes(x=Feature,y=Proportion)) + geom_boxplot() + geom_point(data=feature_proportion_long,aes(x=Feature,y=Proportion),size=3,color="red") + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_x_discrete(limits=levels(feature_proportion_shuffled_long$Feature)[c(17,19,18,5,7,6,8,2,4,3,10,12,11,14,16,15,13,1)])

rmsk_TE_class_feature_long = melt(rmsk_TE_class[,c(1,12:30)],id.var="class_update")
colnames(rmsk_TE_class_feature_long)[2:3] = c("Feature","Proportion")

c = ggplot(feature_proportion_shuffled_class_long,aes(x=Feature,y=Proportion)) + geom_boxplot() + geom_point(data=rmsk_TE_class_feature_long,aes(x=Feature,y=Proportion),size=2,color="red") + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_x_discrete(limits=levels(feature_proportion_shuffled_long$Feature)[c(17,19,18,5,7,6,8,2,4,3,10,12,11,14,16,15,13,1)]) + facet_wrap(~class_update)

d = ggplot(feature_state_mean,aes(x=State,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(name="Total % increase",low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-1,50)) + scale_y_discrete(limits=rev(levels(feature_state_mean$Feature)),labels=genic_labels) + scale_x_discrete(labels=all_state_labels) + facet_wrap(~Coding,labeller=labeller(Coding=coding_labels))

e = ggplot(feature_state_mean_class[which(feature_state_mean_class$Coding == "All"),],aes(x=State,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(name="Total % increase",low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-1,175)) + scale_y_discrete(limits=rev(levels(feature_state_mean_class$Feature)),labels=genic_labels) + scale_x_discrete(labels=all_state_labels) + facet_wrap(~Class)
feature_state_mean_class[which(feature_state_mean_class$Coding == "All"),]

grid.arrange(a,c,d,e,nrow=4,heights=c(0.15,0.30,0.25,0.30))
```

### TE feature stats

```{r analysis Fig S8}
# Any TEs not in a feature? Number of features each TE overlaps
table(apply(rmsk_TE[,14:31],1,function(x) sum(!is.na(x))))
table(apply(rmsk_TE[,names(genic_labels)[1:7]],1,function(x) sum(!is.na(x))))

# Number/proportion of TEs in each feature, overall and by class
feature_proportion/NUM_TE
rmsk_TE_class[,c(1,12:31)]

# Proportion of TEs ever in state overlapping each feature
apply(rmsk_TE_measure[,c(35:49,52:55,60:61)],2,function(x) apply(rmsk_TE_measure[,15:33],2,function(y) dim(rmsk_TE_measure[which(x > 0 & y == "yes"),])[1]/dim(rmsk_TE_measure[which(x > 0),])[1]))
```

### TE feature overlap control

```{r shuffled feature distribution}
# Number/proportion of shuffled TEs in each feature
feature_proportion_shuffled/NUM_TE

# Class
test = merge(ddply(feature_proportion_shuffled_class_long,.(Feature,class_update),summarise,Mean=mean(Proportion),SD=sd(Proportion)),rmsk_TE_class_feature_long,by=c("class_update","Feature"))
test$Zscore = (test$Proportion-test$Mean)/test$SD
test[order(test$Zscore),]

# Subfamily
rmsk_TE_subfamily_feature_long = melt(rmsk_TE_subfamily[,c(1,12:30)],id.var="subfamily")
colnames(rmsk_TE_subfamily_feature_long)[2:3] = c("Feature","Proportion")

test = merge(ddply(feature_proportion_shuffled_subfamily_long,.(Feature,subfamily),summarise,Mean=mean(Proportion),SD=sd(Proportion)),rmsk_TE_subfamily_feature_long,by=c("subfamily","Feature"))
test$Zscore = (test$Proportion-test$Mean)/test$SD
sort(table(test[which(test$Zscore > 2),]$Feature))
test[which(test$Zscore > 2 & test$Feature %in% c("exons_pc","5UTR_pc","3UTR_pc")),]
test[which(test$Zscore > 2 & test$Feature %in% c("exons_nc","5UTR_nc","3UTR_nc")),]
test[which(test$Zscore > 2 & test$Feature =="cpgIsland"),]
test[which(test$Zscore > 2 & test$Feature %in% c("promoter","promoter_pc","promoter_nc")),]
```

### Max expression by feature

```{r max expression by feature}
# Median max expression by feature
sort(apply(rmsk_TE_measure[,15:34],2,function(y) median(rmsk_TE_measure[which(y == "yes"),]$Max_expression,na.rm=TRUE)))

# Max expression by feature
sort(apply(rmsk_TE_measure[,15:34],2,function(y) max(rmsk_TE_measure[which(y == "yes"),]$Max_expression,na.rm=TRUE)))
```

### Supplementary Figure 9

**Supplementary Figure 9. TE metrics by class.** (A) Proportion of all TEs (left), TE subfamilies (middle), and length (bp; right) in each TE class. Blue slice represents the length of the non-TE genome. (B) Density plot of mappability per TE, by class. (D) Density plot of TE age (Jukes-Cantor evolutionary distance from consensus) per TE, by class. (G) CpGs per TE and proportion of TEs with CpGs by class. 

```{r Figure S9, echo=FALSE, cache=TRUE, dependson=c("Figure 3","Figure S8","setup","load libraries","set plot parameters","load functions","create essential matrices","load metadata","load essential matrices","define vectors","define facet labels","load feature sizes","calculate TE stats","generate other matrices")}
#source("R_scripts/TE_class_stats.R")
#source("R_scripts/TE_correlation.R")

par(mfrow=c(1,3)) 

pie(rmsk_TE_class$Count,labels=paste(rmsk_TE_class$Count,"(",round(rmsk_TE_class$Count/sum(rmsk_TE_class$Count)*100,0),"%) ",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="TEs",cex=1,cex.main=2,line=-20,radius=0.5)

pie(rmsk_TE_class$Subfamilies,labels=paste(rmsk_TE_class$Subfamilies,"(",round(rmsk_TE_class$Subfamilies/sum(rmsk_TE_class$Subfamilies)*100,0),"%) ",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="Subfamilies",cex=1,cex.main=2,line=-20,radius=0.5)
 
pie(c(3095693983-1389947349,rmsk_TE_class$Total_length),labels=paste(round(c(3095693983-1389947349,rmsk_TE_class$Total_length)/3095693983*100,0),"%",sep=""),col=c("lightblue",class_colors[as.vector(rmsk_TE_class$class_update)]),clockwise=TRUE,cex=1,cex.main=2,main="Length (bp)",line=-20,radius=0.5)

rmsk_TE_measure_metric = melt(rmsk_TE_measure[,c(8:10,12:14)],id.vars="class_update")
rmsk_TE_measure_metric$variable = factor(rmsk_TE_measure_metric$variable,levels(rmsk_TE_measure_metric$variable)[c(3,2,1,4,5)])

b = ggplot(rmsk_TE_measure,aes(x=JC_distance,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("Jukes-Cantor evolutionary distance") + ylab("Density") + scale_x_continuous(limits=c(0,1)) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

c = ggplot(rmsk_TE_measure,aes(x=mappability,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("Mappability (36bp)") + ylab("Density") + scale_x_continuous(limits=c(0,1)) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

d = ggplot(rmsk_TE_measure,aes(x=Length,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("Length (bp)") + ylab("Density") + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

e = ggplot(rmsk_TE_measure,aes(x=CpGs,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("CpGs") + ylab("Density") + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

f = ggplot(rmsk_TE_measure,aes(x=CpGs_per_length,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + xlab("CpGs/bp") + ylab("Density") + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + facet_wrap(~class_update,scales="free_y",nrow=1)

# Shuffled mappability
rmsk_TE_shuffled_map = melt(t(ldply(rmsk_TE_shuffled,function(x) x$mappability)))[,2:3]
test = as.data.frame(cbind(rep("Real",NUM_TE),rmsk_TE$mappability))
colnames(test) = c("Var2","value")
rmsk_TE_shuffled_map = rbind(rmsk_TE_shuffled_map,test)
rmsk_TE_shuffled_map$Iteration = factor(rmsk_TE_shuffled_map$Var2,levels=c(seq(1,10,1),"Real"))
rmsk_TE_shuffled_map$mappability = as.numeric(rmsk_TE_shuffled_map$value)

g = ggplot(rmsk_TE_shuffled_map,aes(x=mappability,fill=Iteration)) + geom_histogram(binwidth=0.05,position="dodge") + xlab("Mappability") + ylab("TEs") + scale_fill_manual(values=c(rep("grey",10),"red"))

grid.arrange(b,c,d,e,f,g,nrow=6,heights=c(rep(0.14,5),0.3))
```

### Mappability for shuffled TEs

```{r}
aggregate(data=rmsk_TE_shuffled_map,mappability~Iteration,mean)
```

### Remove shuffled matrices

```{r}
rm(rmsk_TE_shuffled_map)
rm(rmsk_TE_shuffled)
```

### CpGs per class

```{r analysis Fig S9a}
rmsk_TE_class[,c(1:2,12,33:39)]
```

### TE stats

```{r analysis Fig S9b}
# TE metrics by overall and by class, median and mean, with p-values
apply(rmsk_TE_measure[,c(8,10,12:14)],2,median)
aggregate(data=rmsk_TE_measure[,c(8:10,12:14)],.~class_update,median)

colMeans(rmsk_TE_measure[,c(8,10,12:14)])
aggregate(data=rmsk_TE_measure[,c(8:10,12:14)],.~class_update,mean)
sort(apply(aggregate(data=rmsk_TE_measure[,c(8:10,12:14)],.~class_update,mean)[,2:6],2,function(x) sd(x)/mean(x)))

apply(rmsk_TE_measure[,c(8,10,12:14)],2,function(x) unlist(kruskal.test(x~rmsk_TE_measure$class_update))["p.value"])

# Max TE length by class
aggregate(data=rmsk_TE_measure,Length~class_update,max)

# % of SINE that is Alu
dim(rmsk_TE[which(rmsk_TE$family == "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SINE"),])[1]

# Average age for older SINE, Alu, SVA
mean(rmsk_TE[which(rmsk_TE$class == "SINE" & rmsk_TE$family != "Alu"),]$JC_distance)
sd(rmsk_TE[which(rmsk_TE$class == "SINE" & rmsk_TE$family != "Alu"),]$JC_distance)
mean(rmsk_TE[which(rmsk_TE$family == "Alu"),]$JC_distance)
sd(rmsk_TE[which(rmsk_TE$family == "Alu"),]$JC_distance)
mean(rmsk_TE[which(rmsk_TE$class_update == "SVA"),]$JC_distance)
sd(rmsk_TE[which(rmsk_TE$class_update == "SVA"),]$JC_distance)

# Mappability vs. length
correlate_TE_feature[which(correlate_TE_feature$Metric == "Length" & correlate_TE_feature$State == "mappability"),]

# Age vs. overlap with CpG islands
wilcox.test(rmsk_TE_measure[which(rmsk_TE_measure$cpgIsland == "yes"),]$JC_distance,rmsk_TE_measure[which(rmsk_TE_measure$cpgIsland == "no"),]$JC_distance)
aggregate(data=rmsk_TE_measure,JC_distance~cpgIsland,mean)

# SVA/Alu overlapping CpG islands
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$family == "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$family == "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$family == "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SVA"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SVA"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SVA"),])[1]

# Proportion of all TEs overlapping CpG islands by family/subfamily
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),])[1]
sort(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),]$family)))/dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),])[1]
sort(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),]$subfamily)))/dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),])[1]
```

### Supplementary Figure 10

**Supplementary Figure 10. Correlations between TE metrics and number of samples in state, by class.** 

```{r Figure S10, echo=FALSE}
#source("R_scripts/state_metric_correlation.R")
source("R_scripts/feature_overlap_metric.R") #Requires TE_correlation.R

a = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update == "All" & correlate_TE_state$State %in% c("Hypomethylated","Intermediate","Hypermethylated","Missing")),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(limits=c("Missing","Hypermethylated","Intermediate","Hypomethylated")) + scale_x_discrete(limits=c("Length","CpGs","CpGs_per_length","JC_distance","mappability"),labels=c("Length","CpGs","CpGs/bp","Age","Mappability"))

b = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update != "All" & correlate_TE_state$State %in% c("Hypomethylated","Intermediate","Hypermethylated","Missing")),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(limits=c("Missing","Hypermethylated","Intermediate","Hypomethylated")) + scale_x_discrete(limits=c("Length","CpGs","CpGs_per_length","JC_distance","mappability"),labels=c("Length","CpGs","CpGs/bp","Age","Mappability")) + facet_wrap(~class_update) 

c = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs"))

d = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs")) + facet_wrap(~class_update)

e = ggplot(feature_metric_mean[which(feature_metric_mean$Coding == "All"),],aes(x=Metric,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(name="Total % increase",low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-1,15)) + scale_y_discrete(limits=rev(levels(feature_metric_mean$Feature)),labels=genic_labels) + scale_x_discrete(labels=c("Length (bp)","Mappability","Age","CpGs","CpGs/bp"))

f = ggplot(feature_metric_mean_class[which(feature_metric_mean_class$Coding == "All"),],aes(x=Metric,y=Feature,fill=Enrichment)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(name="Total % increase",low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-1,15),scale=FALSE) + scale_y_discrete(limits=rev(levels(feature_metric_mean_class$Feature)),labels=genic_labels) + scale_x_discrete(labels=c("Length (bp)","Mappability","Age","CpGs","CpGs/bp")) + facet_wrap(~Class)

grid.arrange(a,b,c,d,e,f,nrow=3)
```

### TE metrics by genic features, p-values (Too long!)

```{r analysis Fig S10a, eval=FALSE}
TE_pvalue_all_feature = apply(rmsk_TE_measure[,15:34],2,function(x) apply(rmsk_TE_measure[,c(8,10,12:14)],2,function(y) unlist(wilcox.test(y~x))["p.value"]))
TE_pvalue_class_feature = ddply(rmsk_TE_measure,~class_update,function(z) apply(z[,15:34],2,function(x) apply(z[,c(8,10,12:14)],2,function(y) unlist(wilcox.test(y~x))["p.value"])))
```

### TE stats

```{r analysis Fig S10b}
# From Figure S10
feature_metric_mean[which(feature_metric_mean$Coding == "All"),]
feature_metric_mean_class[which(feature_metric_mean_class$Coding == "All"),]

# Correlations between mappability, CpGs, CpGs per length, number of samples Missing (WGBS)
# Missing: All TEs, all classes except SVA; SVA - Length, mappability, CpGs_per_length
correlate_TE_state[which(correlate_TE_state$State == "Missing" & correlate_TE_state$Metric %in% c("mappability","CpGs","CpGs_per_length")),]

# Correlation between mappability, number of samples Quiescent
correlate_TE_state[which(correlate_TE_state$Metric == "mappability" & correlate_TE_state$State == "X15_Quies"),]

# Mappability of TEs always Quiescent vs not
wilcox.test(rmsk_TE_measure[which(apply(rmsk_TE_measure,1,function(x) sum(as.numeric(x[35:48]))) > 0),]$mappability,rmsk_TE_measure[which(rmsk_TE_measure$X15_Quies > 0 & rmsk_TE_measure$States == 1),]$mappability)
mean(rmsk_TE_measure[which(rmsk_TE_measure$X15_Quies > 0 & rmsk_TE_measure$States == 1),]$mappability)
mean(rmsk_TE_measure[which(apply(rmsk_TE_measure,1,function(x) sum(as.numeric(x[35:48]))) > 0),]$mappability)
```

## By-sample

### Figure 4

**Figure 4. Variation in TE activity by sample classification.** (A) Proportion of each state in TEs by sample, colored by Roadmap groups. Black line represents median across all samples for that state. For WGBS, proportion is CpGs instead of bases. (B) Average CpG methylation level by sample for all CpGs, CpGs in TEs, and CpGs not in TEs. (C) Principal coordinate analysis on TEs in the 6_EnhG or 7_Enh enhancer states in each sample. 

```{r Figure 4, fig.width=7, fig.height=5, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 1","setup","load libraries","set plot parameters","load functions","create essential matrices","load metadata","load essential matrices","define vectors","define facet labels","load feature sizes","calculate TE stats","generate other matrices")}
#source("R_scripts/proportion.R")
source("R_scripts/proportion_by_sample.R") #Requires proportion.R

by_sample_all = rbind(mnemonics_states_TE_proportion,WGBS_proportion,DNase_proportion,H3K27ac_proportion)
by_sample_all = merge(by_sample_all,metadata[,c(1,4:9)],by="Sample")
by_sample_all$Proportion = as.numeric(by_sample_all$Proportion)

a = ggplot(by_sample_all,aes(x=State,y=Proportion,color=Group)) + geom_point(position = position_jitter(width = 0.3)) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylim(0,1) + ylab("Proportion of state in TEs") + scale_color_manual(values=group_colors,guide=FALSE) + scale_x_discrete(limits=levels(by_sample_all$State)) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black",width = 0.8)
 
source("~/TE_landscape/R_scripts/WGBS_sample_CpG_average.R")

b = ggplot(CpG_meth_average,aes(x=Sample,y=Methylation,group=Cohort,color=Cohort)) + geom_point(size=3) + geom_line() + theme(axis.text.x=element_text(angle=90),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.title.x=element_blank(),legend.position = c(0.8, 0.2),legend.title=element_blank(),panel.grid.major.x = element_line(color="black"),legend.key = element_rect(fill = "transparent", colour = "transparent")) + ylab("Average methylation") + scale_color_manual(values=c("black","blue","red"),labels=c("All CpGs","Not in TEs","In TEs")) 

source("R_scripts/enhancer_PCA.R")

c = plot_pca(enhancer_matrix_pca,c(1,2),metadata,"Group","Roadmap Group",group_colors,FALSE)

legend = get_legend(ggplot(by_sample_all,aes(x=State,y=Proportion,color=Group)) + geom_point(position = position_jitter(width = 0.3)) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylim(0,1) + ylab("Proportion of state in TEs") + scale_color_manual(values=group_colors) + scale_x_discrete(limits=levels(by_sample_all$State)) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black",width = 0.4) + guides(color = guide_legend(override.aes = list(size=6.5))))

grid.arrange(a,b,c,legend,nrow=2,layout_matrix=rbind(c(1,1,4),c(2,3,4)),widths=c(0.4,0.4,0.2))
```

### By-sample analysis

```{r analysis Fig 4a}
ddply(by_sample_all,~State,function(x) x[which.min(x$Proportion),])
ddply(by_sample_all,~State,function(x) x[which.max(x$Proportion),])

# Stats
by_sample_stats = ddply(by_sample_all,~State,summarise, Range = max(Proportion)-min(Proportion), Median = median(Proportion), Mean = mean(Proportion), CV = sd(Proportion)/mean(Proportion))
by_sample_stats[order(by_sample_stats$Range),]

# Kruskal-Wallis test by sample classification
by_sample_KW = ddply(by_sample_all,~State,summarise,Group = p.adjust(unlist(kruskal.test(Proportion,Group))["p.value"],method="bonf"),
      Anatomy = p.adjust(unlist(kruskal.test(Proportion,Anatomy))["p.value"],method="bonf"),
      Type = p.adjust(unlist(kruskal.test(Proportion,Type))["p.value"],method="bonf"),
      Age = p.adjust(unlist(kruskal.test(Proportion,Age))["p.value"],method="bonf"),
      Germline = p.adjust(unlist(kruskal.test(Proportion,Germline))["p.value"],method="bonf"))
by_sample_KW = merge(by_sample_KW,ddply(droplevels(by_sample_all[which(!(by_sample_all$State %in% meth_states)),]),~State,summarise,Cancer = p.adjust(unlist(kruskal.test(Proportion,Cancer))["p.value"],method="bonf")),by="State",all.x=TRUE)

# Wilcox tests by sample classification
by_sample_Wilcox = list(
  merge(ddply(by_sample_all,~State,summarise,Group_pvalue=wilcox_to_all(Proportion,droplevels(Group)),Group=levels(droplevels(Group))),
  aggregate(data=by_sample_all,Proportion~State+Group,mean),by=c("State","Group"),all.x=TRUE),
  merge(ddply(by_sample_all,~State,summarise,Anatomy_pvalue=wilcox_to_all(Proportion,droplevels(Anatomy)),Anatomy=levels(droplevels(Anatomy))),
  aggregate(data=by_sample_all,Proportion~State+Anatomy,mean),by=c("State","Anatomy"),all.x=TRUE),
  merge(ddply(by_sample_all,~State,summarise,Type_pvalue=wilcox_to_all(Proportion,droplevels(Type)),Type=levels(droplevels(Type))),
  aggregate(data=by_sample_all,Proportion~State+Type,mean),by=c("State","Type"),all.x=TRUE),
  merge(ddply(by_sample_all,~State,summarise,Age_pvalue=wilcox_to_all(Proportion,droplevels(Age)),Age=levels(droplevels(Age))),
  aggregate(data=by_sample_all,Proportion~State+Age,mean),by=c("State","Age"),all.x=TRUE),
  merge(ddply(by_sample_all,~State,summarise,Germline_pvalue=wilcox_to_all(Proportion,droplevels(Germline)),Germline=levels(droplevels(Germline))),
  aggregate(data=by_sample_all,Proportion~State+Germline,mean),by=c("State","Germline"),all.x=TRUE))
names(by_sample_Wilcox) = c("Group","Anatomy","Type","Age","Germline")

by_sample_Wilcox$Cancer = 
  merge(ddply(by_sample_all[which(!(by_sample_all$State %in% meth_states)),],~State,summarise,Cancer_pvalue=p.adjust(unlist(wilcox.test(Proportion~Cancer))["p.value"],method="bonf")),
        aggregate(data=by_sample_all,Proportion~State+Cancer,mean),by="State",all.y=TRUE)

# For table by Group
merge(by_sample_stats,merge(by_sample_KW[,1:2],by_sample_Wilcox$Group[which(by_sample_Wilcox$Group$Group_pvalue < 0.05),],by="State",all=TRUE),by="State",all=TRUE)

# Higher % CpGs in each methylation state in TEs in ESC/iPSC? 
test = aggregate(data=by_sample_all,Proportion~State+Group,mean)
test[order(test$State,test$Proportion),]
```

### By-sample CpG methylation

```{r}
# Difference in TE/non-TE methylation across samples
test = dcast(CpG_meth_average,Sample~Cohort,value.var="Methylation")
test$Difference = test$noTE-test$TE
test[order(test$Difference),]

metadata[which(metadata$Sample %in% as.vector(tail(test[order(test$Difference),],n=10)$Sample)),]

cor.test(test$All,test$Difference)
```

# Compare TE proportion of state to genome, non-TE, and TE bases

```{r proportion investigation}
all_state_proportion_matrix = dcast(all_state_proportion,State+Sample+Group~Cohort,value.var="Bases")
ddply(all_state_proportion_matrix,.(State),summarise,Genome_corr = unlist(cor.test((TE/Genome),Genome))["estimate.cor"],Genome_pvalue = unlist(cor.test((TE/Genome),Genome))["p.value"],
      TE_corr = unlist(cor.test((TE/Genome),TE))["estimate.cor"],TE_pvalue = unlist(cor.test((TE/Genome),TE))["p.value"],
      NonTE_corr = unlist(cor.test((TE/Genome),(Genome-TE)))["estimate.cor"],NonTE_pvalue = unlist(cor.test((TE/Genome),(Genome-TE)))["p.value"])

# Plot bases in genome/TEs in each state by sample, increasing order
lapply(c(chromHMM_states,meth_states,"DNase","H3K27ac"),function(x) ggplot(all_state_proportion[which(all_state_proportion$State == x),],aes(x=reorder(Sample,Bases,max),y=log10(Bases),color=Group,group=Cohort,shape=Cohort)) + geom_point(size=2) + geom_line() + theme(axis.text.x = element_text(angle=90),axis.title.x = element_blank()) + scale_color_manual(values=group_colors))

#lapply(c(chromHMM_states,meth_states,"DNase","H3K27ac"),function(x) ggplot(all_state_proportion_matrix[which(all_state_proportion_matrix$State == x),],aes(x=reorder(Sample,TE/Genome,max),y=(TE/(Genome-TE)),color=Group)) + geom_point(shape=16) + scale_color_manual(values=group_colors) + theme(axis.text.x = element_text(angle=90),axis.title.x = element_blank()))
```

### Enhancer PCA analysis

```{r analysis Fig 4b}
# Proportion and cumulative proportion of variance explained by each PC
dim(enhancer_matrix)[1]
summary(enhancer_matrix_pca) 

# Contribution of TEs to PCs
head(enhancer_matrix_pca$rotation[order(enhancer_matrix_pca$rotation[,1]),])
tail(enhancer_matrix_pca$rotation[order(enhancer_matrix_pca$rotation[,1]),])
```

### Supplementary Figure 11

**Supplementary Figure 11. Methylation level of RefSeq features across samples.** (B) Average methylation level of CpGs overlapping RefSeq features by sample. 

### Sample groups in each category that are significantly different than all other samples in the proportion of that state in TEs (p<0.05, Wilcox tests with Bonferroni correction).

```{r Figure S11, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 3","Figure 4","setup","load libraries","set plot parameters","load functions","create essential matrices","load metadata","load essential matrices","define vectors","define facet labels","load feature sizes","calculate TE stats","generate other matrices")}
#source("R_scripts/proportion_class.R")
#source("R_scripts/WGBS_sample_CpG_average.R")
source("R_scripts/PCA.R")

# By-sample proportion by TE class
by_sample_class = rbind(class_chromHMM[,c(1:3,8)],class_CpG_meth[,c(1:2,4,8)],TE_DNase_class[,c(1:2,4,8)],TE_H3K27ac_class[,c(1:2,4,8)])
by_sample_class = merge(by_sample_class,metadata[,c(1,4:9)],by="Sample")
by_sample_class$Proportion = as.numeric(by_sample_class$Proportion_state)

a = ggplot(by_sample_class,aes(x=State,y=Proportion,color=Group)) + geom_point(position = position_jitter(width = 0.3)) + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=90,hjust=1)) + ylim(0,1) + ylab("Proportion of state in TEs") + scale_color_manual(values=group_colors,guide=FALSE) + scale_x_discrete(limits=levels(by_sample_class$State)) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black") + facet_wrap(~class)

# Feature methylation by sample
b =  ggplot(CpG_feature_meth_average,aes(x=Sample,y=Methylation,group=Cohort,color=Feature,shape=Coding)) + geom_point(size=3) + geom_line() + theme(axis.text.x=element_text(angle=90),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.title.x=element_blank(),legend.title=element_blank(),panel.grid.major.x = element_line(color="black")) + ylab("Average methylation") + scale_color_manual(values=c("black",brewer.pal(8,"Set1"))) + scale_shape_discrete(labels=c("All","Protein-coding","Non-coding")) 

# PCA on TEs in DNase, H3K27ac
c = plot_pca(DNase_pca,c(1,2),metadata[which(!is.na(metadata$DNase)),],"Group","Roadmap Group",group_colors,guide=F)
d = plot_pca(DNase_pca,c(3,2),metadata[which(!is.na(metadata$DNase)),],"Group","Roadmap Group",group_colors,guide=F)
e = plot_pca(DNase_pca,c(3,4),metadata[which(!is.na(metadata$DNase)),],"Group","Roadmap Group",group_colors,guide=F)
f = plot_pca(H3K27ac_pca,c(1,2),metadata[which(!is.na(metadata$H3K27ac)),],"Group","Roadmap Group",group_colors,guide=F)
g = plot_pca(H3K27ac_pca,c(3,2),metadata[which(!is.na(metadata$H3K27ac)),],"Group","Roadmap Group",group_colors,guide=F)
h = plot_pca(H3K27ac_pca,c(3,4),metadata[which(!is.na(metadata$H3K27ac)),],"Group","Roadmap Group",group_colors,guide=F)

grid.arrange(a,b,c,d,e,f,g,h,nrow=4,layout_matrix=rbind(c(1,1,1),c(2,2,2),c(3,4,5),c(6,7,8)),heights=c(0.3,0.3,0.2,0.2))
```

### Tests by sample classification and class

```{r analysis Fig S11a}
by_sample_stats_class = ddply(by_sample_class,.(State,class),summarise, Range = max(Proportion)-min(Proportion), Median = median(Proportion), Mean = mean(Proportion),CV = sd(Proportion)/mean(Proportion))
by_sample_stats_class[order(by_sample_stats_class$Range),]
by_sample_stats_class[order(by_sample_stats_class$CV),]

by_sample_KW_class = ddply(by_sample_class,.(State,class),summarise,Group = p.adjust(unlist(kruskal.test(Proportion,Group))["p.value"],method="bonf"))

by_sample_Wilcox_class = merge(ddply(by_sample_class,.(State,class),summarise,Group_pvalue=wilcox_to_all(Proportion,droplevels(Group)),Group=levels(droplevels(Group))),aggregate(data=by_sample_class,Proportion~State+Group+class,mean),by=c("State","Group","class"),all.x=TRUE)

# For table by Group
test = merge(by_sample_stats_class,merge(by_sample_KW_class,by_sample_Wilcox_class[which(by_sample_Wilcox_class$Group_pvalue < 0.05),],by=c("State","class"),all=TRUE),by=c("State","class"),all=TRUE)
test
```

### Enrichment of TE class x state x sample

```{r analysis Fig S11b}
# Kruskal-Wallis test of proportion in each state by class (obvious, classes are different sizes)
ddply(by_sample_class,.(State),summarise,Pvalue=unlist(kruskal.test(Proportion_state~class))["p.value"])

class_enrichment = rbind(class_chromHMM[,c(1:3,10)],class_CpG_meth[,c(1:2,4,10)],TE_DNase_class[,c(1:2,4,10)],TE_H3K27ac_class[,c(1:2,4,10)])
table(class_enrichment[which(class_enrichment$Enrichment > 1.5),2:3]) # Reduce threshold?
```

### CpG average methylation per sample

```{r analysis Fig S11c}
# Stats (TEs and all)
CpG_feature_meth_average
ddply(CpG_meth_average,~Cohort,summarise,Mean=mean(Methylation),SD=sd(Methylation),Outlier=mean(Methylation)-2*sd(Methylation))
ddply(CpG_meth_average[which(CpG_meth_average$Sample != "E017"),],~Cohort,summarise,Mean=mean(Methylation),SD=sd(Methylation)) #Without IMR90
ddply(CpG_feature_meth_average,"Cohort",summarise,Mean=mean(Methylation),SD=sd(Methylation),Outlier=mean(Methylation)-2*sd(Methylation))

# Identifying differences by sample Group
ddply(CpG_feature_meth_average,~Cohort+metadata[match(Sample,metadata$Sample),]$Group,summarise,Mean=mean(Methylation))
ddply(CpG_feature_meth_average,~Cohort,summarise,Group=p.adjust(unlist(kruskal.test(Methylation,metadata[match(Sample,metadata$Sample),]$Group))["p.value"]))
ddply(CpG_feature_meth_average,~Cohort,summarise,Pvalue=wilcox_to_all(Methylation,droplevels(metadata[match(Sample,metadata$Sample),]$Group)),Group=levels(droplevels(metadata[match(Sample,metadata$Sample),]$Group)))
```

### DNase/H3K27ac PCA analysis

```{r analysis Fig S11d}
# TEs ever in state
dim(TE_DNase_peaks)[1]
dim(TE_DNase_binary)[1]
dim(TE_H3K27ac_peaks)[1]
dim(TE_H3K27ac_binary)[1]

summary(DNase_pca)
summary(H3K27ac_pca) 

# Contribution of TEs to PCs
head(DNase_pca$rotation[order(DNase_pca$rotation[,1]),])
tail(DNase_pca$rotation[order(DNase_pca$rotation[,1]),])

# Contribution of TEs to PCs
head(H3K27ac_pca$rotation[order(H3K27ac_pca$rotation[,1]),])
tail(H3K27ac_pca$rotation[order(H3K27ac_pca$rotation[,1]),])

# DNase outliers
DNase_pca$x[order(DNase_pca$x[,1]),1:2]
sort(colSums(TE_DNase_peaks[,8:60]))
DNase_stats[order(DNase_stats$Peaks),]

rm(list=c("DNase_pca","H3K27ac_pca","enhancer_matrix_pca","enhancer_matrix"))
```

## Subfamily enrichment

### Set thresholds 

```{r set enrichment thresholds}
THRESHOLD_IJK_BASE = 0
THRESHOLD_IK_BASE = 0
THRESHOLD_IJK_CPG = 0
THRESHOLD_IK_CPG = 0
THRESHOLD_IJK_MEMBERS = 3
THRESHOLD_LOR = 0
THRESHOLD_PC = 0.01
```

### Figure 5

**Figure 5. Enrichment of TE subfamilies in epigenetic states in individual samples.** (A) Number of samples each subfamily is enriched in each state (log odds ratio > 1.5). The total number of subfamilies enriched LOR > 1.5 in at least one sample are listed under each state. (B) Subfamilies with at least two LOR scores >1.5 in the 7_Enh state (blue squares indicate enrichment in that sample). Binary clustering, complete linkage. Columns are colored by sample metadata, and rows are colored by subfamily class. 

```{r Figure 5, fig.width=7, fig.height=9, echo=FALSE}
source("R_scripts/subfamily_enrichment.R")

a = ggplot(subfamily_state_sample_counts,aes(x=State,y=V1,color=class_update)) + geom_point(position=position_jitterdodge(jitter.width=0.5)) + coord_flip() + theme(axis.title.y=element_blank()) + ylab("Number of samples enriched LOR > 1.5") + scale_color_manual(values=class_colors,name="Class") + scale_y_continuous(breaks=c(0,seq(10,120,10),127),expand = c(0.01,0.01)) + geom_rect(xmin=0.5,xmax=1.5,ymin=99,ymax=127,fill="grey",color=NA) + geom_rect(xmin=1.5,xmax=2.5,ymin=54,ymax=127,fill="grey",color=NA) + geom_rect(xmin=2.5,xmax=6.5,ymin=38,ymax=127,fill="grey",color=NA) + geom_vline(xintercept=seq(1.5,20.5,1), colour='grey') + scale_x_discrete(limits = rev(levels(subfamily_state_sample_counts$State)),labels=as.vector(rev(apply(subfamily_state_sample_counts_combine,1,function(x) paste(x[1],"\n(n=",x[2],")",sep="")))))

grid.arrange(a)

# Legends for metadata categories
legend_group = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Group)) + geom_tile() + scale_fill_manual(values=group_colors) + theme(legend.direction = "horizontal"))
legend_anatomy = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Anatomy)) + geom_tile() + scale_fill_manual(values=anatomy_colors) + theme(legend.direction = "horizontal"))
legend_type = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Type)) + geom_tile() + scale_fill_manual(values=brewer.pal(5,"Greens")) + theme(legend.direction = "horizontal"))
legend_age = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Age)) + geom_tile() + scale_fill_manual(values=brewer.pal(4,"YlOrRd"),labels=c("Cell line","Fetal","Non-fetal","Unknown")) + theme(legend.direction = "horizontal"))
legend_cancer = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Cancer)) + geom_tile() + scale_fill_manual(values=c("white","red")) + theme(legend.direction = "horizontal"))
legend_germline = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Germline)) + geom_tile() + scale_fill_manual(values=brewer.pal(6,"Dark2")) + theme(legend.direction = "horizontal"))

b = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","7_Enh",2,127,THRESHOLD_LOR)

# Arrange column legends
grid.arrange(legend_group,legend_anatomy,legend_age,legend_cancer,legend_germline,legend_type,ncol=2,widths=c(0.6,0.4),layout_matrix=cbind(c(1,2,6),c(3,4,5)))
```

### Number of subfamilies excluded by enrichment thresholds

```{r excluded subfamilies}
# chromHMM, all chromosomes
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Total_length <= THRESHOLD_IK_BASE),])[1]
# chromHMM, no Y
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Total_length > THRESHOLD_IK_BASE & rmsk_TE_subfamily$Total_length_noY <= THRESHOLD_IK_BASE),])[1]
# CpGs
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$CpGs <= THRESHOLD_IK_CPG),])[1]

rmsk_TE_subfamily[which(rmsk_TE_subfamily$Total_length_noY <= THRESHOLD_IK_BASE | rmsk_TE_subfamily$CpGs <= THRESHOLD_IK_CPG),c("subfamily","Count","Count_noY","Total_length","Total_length_noY","CpGs")]

# Subfamilies with no CpGs
rmsk_TE_subfamily[which(rmsk_TE_subfamily$CpGs == 0),c("subfamily","Count","Count_noY","Total_length","Total_length_noY","CpGs")]

# Subfamilies with fewer members than member threshold
rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count < THRESHOLD_IJK_MEMBERS | rmsk_TE_subfamily$Count_noY < THRESHOLD_IJK_MEMBERS),c("subfamily","Count","Count_noY","Total_length","Total_length_noY","CpGs")]
```

### Number of enrichments excluded by enrichment thresholds

```{r excluded enrichments}
test = table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR & subfamily_state_sample_combined$Length_ik > THRESHOLD_IK_BASE & subfamily_state_sample_combined$Length_ijk < THRESHOLD_IJK_BASE),]$State)
sum(test)
test

test = table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR & subfamily_state_sample_combined$Length_ik > THRESHOLD_IK_CPG & subfamily_state_sample_combined$Length_ijk < THRESHOLD_IJK_CPG & subfamily_state_sample_combined$State %in% meth_states),]$State)
sum(test)
test

test = table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR & subfamily_state_sample_combined$Members < THRESHOLD_IJK_MEMBERS),]$State)
sum(test)
test
```

### Number of enrichments by state and number of enriched subfamilies by state, overall and by class

```{r analysis Fig 5}
enrichment_table = merge(as.data.frame(table(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),]$State)),subfamily_state_sample_counts_combine,by.x="Var1",by.y="State")
names(enrichment_table)[1:3] = c("State","Enrichments","Subfamilies")

sum(enrichment_table$Enrichments)
enrichment_table$Enrichments_pc = enrichment_table$Enrichments/sum(enrichment_table$Enrichments)
enrichment_table

# Subfamilies with 2+ enrichments in state
aggregate(data=subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 1),],subfamily~State,length)

# Number of subfamilies enriched at least once, by composite states
length(unique(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),]$subfamily))
length(unique(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR & subfamily_state_sample_filter$State %in% c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")),]$subfamily))

# Enriched classes
test = dcast(aggregate(data=subfamily_state_sample_counts,V1~State+class_update,function(x) sum(x > 0)),class_update~State,value.var = "V1")
table(rmsk_TE_subfamily$class_update)/968
apply(test[,2:22],2,function(x) log2((x/sum(x))/(table(rmsk_TE_subfamily$class_update)/968)))
apply(enrichment_table[,4:9],1,function(x) x/sum(x))

# Enriched families
test = dcast(aggregate(data=subfamily_state_sample_counts,V1~State+family,function(x) sum(x > 0)),family~State,value.var = "V1")
table(rmsk_TE_subfamily$family)/968
test_log = as.data.frame(apply(test[,2:22],2,function(x) log2((x/sum(x))/(table(rmsk_TE_subfamily$family)/968))))
test_log$Subfamilies = as.numeric(table(rmsk_TE_subfamily$family)[rownames(test_log)])
lapply(colnames(test_log)[1:21],function(x) test_log[which(test_log[,x] > 0 & test_log$Subfamilies > 0),c(x,"Subfamilies")])
```

### Supplementary Figure 12

**Supplementary Figure 12. Subfamilies with at least two logs odds ratio scores >1.5** in the (A) 1_TssA, (B) 2_TssAFlnk, (C) 3_TxFlnk, (D) 4_Tx, (E) 5_TxWk, or (F) 6_EnhG chromHMM states, (G) hypomethylation, (H) DNase peaks, or (I) H3K27ac peaks. Blue squares indicate enrichment in that sample. Binary clustering, complete linkage. Columns are colored by sample metadata, and rows are colored by subfamily class. 

```{r Figure S12, echo=FALSE}
#source("R_scripts/subfamily_enrichment.R")

a = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","1_TssA",2,127,THRESHOLD_LOR)
b = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","2_TssAFlnk",2,127,THRESHOLD_LOR)
c = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","3_TxFlnk",2,127,THRESHOLD_LOR)
d = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","6_EnhG",2,127,THRESHOLD_LOR)
e = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","4_Tx",2,127,THRESHOLD_LOR)
f = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","5_TxWk",2,127,THRESHOLD_LOR)
g = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Hypomethylated",2,37,THRESHOLD_LOR)
h = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Intermediate",2,37,THRESHOLD_LOR)
i = plot_binary_heatmap(subfamily_state_sample_filter,"DNase","DNase",2,53,THRESHOLD_LOR)
j = plot_binary_heatmap(subfamily_state_sample_filter,"H3K27ac","H3K27ac",2,98,THRESHOLD_LOR)

legend_class = get_legend(ggplot(rmsk_TE_subfamily,aes(x=subfamily,y=class_update,fill=class_update)) + geom_tile() + scale_fill_manual(values=class_colors))

# Arrange column legends
grid.arrange(legend_group,legend_anatomy,legend_age,legend_cancer,legend_germline,legend_type,ncol=2,widths=c(0.6,0.4),layout_matrix=cbind(c(1,2,6),c(3,4,5)))
```

### Average bp of state and proportion of members in a state for subfamilies in a state

```{r bp members percent in state}
# All instances
ddply(subfamily_state_sample_combined,~State,summarize,Length_ijk_median=median(na.omit(Length_ijk)),Length_ijk_min=min(na.omit(Length_ijk)),Length_ijk_max=max(na.omit(Length_ijk)),Percent_mean=mean(na.omit(Percent)),Members_mean=mean(na.omit(Members)),Percent_median=median(na.omit(Percent)), Members_median=median(na.omit(Members)),Percent_min=min(na.omit(Percent)), Members_min=min(na.omit(Members)), Percent_max=max(na.omit(Percent)), Members_max=max(na.omit(Members)))   

# Enriched
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],~State,summarise,Length_ijk_median=median(Length_ijk),Length_ijk_min=min(Length_ijk),Length_ijk_max=max(Length_ijk), Percent_mean=mean(Percent),Members_mean=mean(Members), Percent_median=median(Percent), Members_median=median(Members),Percent_min=min(Percent), Members_min=min(Members), Percent_max=max(Percent), Members_max=max(Members))

# >1%
ddply(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],~State,summarise,Length_ijk_median=median(Length_ijk),Length_ijk_min=min(Length_ijk),Length_ijk_max=max(Length_ijk),Percent_mean=mean(Percent), Members_mean=mean(Members), Percent_median=median(Percent),Members_median=median(Members),Percent_min=min(Percent), Members_min=min(Members), Percent_max=max(Percent), Members_max=max(Members))

# Max by state
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],.(State),function(x) x[which.max(x$Percent),])
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],.(State),function(x) tail(x[order(x$Percent),c(1:11,16:17)],n=10))

test = ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),c(1:11,16:17)],.(State),function(x) x[which(x$Percent > mean(x$Percent)+2*sd(x$Percent)),])
ddply(test,.(State),function(x) length(unique(x$subfamily)))

# For each subfamily x state, members ever in state vs. mean members in state in a single sample
test = rmsk_TE_measure[,c(4,35:49,52:55,60:61)]
colnames(test)[2:16] = chromHMM_states

subfamily_state_members = merge(aggregate(data=melt(test,id.var="subfamily"),value~variable+subfamily,function(x) length(x[which(x > 0)])),ddply(subfamily_state_sample_combined,.(subfamily,State),summarise,Members_mean=mean(Members),Members_max=max(Members)),by.x=c("subfamily","variable"),by.y=c("subfamily","State"))

mean(subfamily_state_members[which(subfamily_state_members$Members_max > 0),]$value/subfamily_state_members[which(subfamily_state_members$Members_max > 0),]$Members_mean)
sd(subfamily_state_members[which(subfamily_state_members$Members_max > 0),]$value/subfamily_state_members[which(subfamily_state_members$Members_max > 0),]$Members_mean)

ddply(subfamily_state_members,.(variable),summarise,Mean=mean(na.omit(value/Members_mean)),SD=sd(na.omit(value/Members_mean)),Max=max(na.omit(value/Members_max)))
```

### Supplementary Figure 13

**Supplementary Figure 13. Involvement of subfamily members enriched in epigenetic states.** (A) Proportion and (B) number of subfamily members in a state for subfamilies enriched in each state. If a subfamily is enriched in a state in multiple samples, each instance is included. No subfamily is enriched in the 15_Quies chromHMM or hypermethylated CpG states. 

```{r Figure S13, echo=FALSE}

a = ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=log10(Length_ijk),fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Length of state in subfamily in sample, log10(bp/CpGs)") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=names(all_state_labels)[1:21]) + ylim(0,7)
  
b = ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=Percent,fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Proportion of subfamily members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=names(all_state_labels)[1:21])

c = ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=log10(Members),fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Log10 number of subfamily members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=names(all_state_labels)[1:21])

d = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],aes(x=State,y=log10(Length_ijk),fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Length of state in subfamily in sample, log10(bp/CpGs)") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=names(all_state_labels)[1:21]) + ylim(0,7)
  
e = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],aes(x=State,y=Percent,fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Proportion of subfamily members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=names(all_state_labels)[1:21])

f = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],aes(x=State,y=log10(Members),fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Log10 number of subfamily members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=names(all_state_labels)[1:21])

legend = get_legend(ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=Percent,fill=State)) + geom_boxplot() + scale_fill_manual(values=all_state_colors))

grid.arrange(a,b,c,d,e,f,legend,nrow=3,ncol=3,layout_matrix=cbind(c(1,2,3),c(4,5,6),c(7,7,7)),widths=c(0.4,0.4,0.2))
```

### WGBS subfamily state stats

```{r analysis Fig S13c, cache=TRUE, cache.lazy=FALSE, dependson=c("setup","load libraries","set plot parameters","load functions","create essential matrices","load metadata","load essential matrices","define vectors","define facet labels","load feature sizes","calculate TE stats","generate other matrices")}
source("R_scripts/WGBS_subfamily_enrichment_TE.R")

# Subfamilies with all members in a state in a sample
TE_meth_subfamily = lapply(TE_meth_subfamily,function(y) transform(y,All_members = apply(y[,4:40],1,function(x) sum(na.omit(x) == 1))))

lapply(TE_meth_subfamily,function(y) dim(y[which(y$All_members > 0),])[1])

lapply(TE_meth_subfamily,function(y) merge(y[which(y$All_members > 0),c("subfamily","family","class_update","All_members")],rmsk_TE_subfamily[,c("subfamily","Count","Count_CpGs","Total_length","Total_length_noY","CpGs")],by="subfamily"))

# Average proportion in state, with/without IMR90
ldply(TE_meth_subfamily,summarise,SD=sd(Mean),SD_noIMR90=sd(Mean_noIMR90),Mean=mean(Mean),Mean_noIMR90=mean(Mean_noIMR90),Range=mean(Range),Range_noIMR90=mean(Range_noIMR90),Max=mean(Max),Max_noIMR90=mean(Max_noIMR90))
```

### Supplementary Figure 14

**Supplementary Figure 14. Subfamilies representing >1% of a state.** (A) Subfamilies representing more than 1% of a chromHMM state, hypomethylated or intermediately methylated CpGs, or DNase/H3K27ac peak width in a sample. Active regulatory and transcribed chromHMM states are presented. Color represents proportion of the state in that sample in the subfamily. Subfamilies are ordered by decreasing total length.

```{r Figure S14, echo=FALSE}
ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_combined$State %in% c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")),],aes(x=Sample,y=reorder(subfamily,rmsk_TE_subfamily[match(subfamily,rmsk_TE_subfamily$subfamily),]$Total_length),fill=Length_percent_jk)) + geom_tile(linetype="blank") + theme(axis.ticks.x=element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank()) + scale_fill_gradient(name="% state",low="lightblue",high="darkblue",limits=c(0.009,0.12)) +  facet_wrap(~State)
```

### Subfamily stats, >1%

```{r analysis Fig S14a}
# Max % in subfamily by state
ddply(subfamily_state_sample_combined,~State,function(x) x[which.max(x$Length_percent_jk),])

# Max % of state in subfamily without enrichment
ddply(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment <= THRESHOLD_LOR),],~State,function(x) x[which.max(x$Length_percent_jk),])

# Number of instances by state and number of >1% subfamilies by state, overall and by class
pc_table = merge(as.data.frame(table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),]$State)),subfamily_state_sample_counts_pc_combine,by.x="Var1",by.y="State")
names(pc_table)[1:3] = c("State","Instances","Subfamilies")
sum(pc_table$Instances)
sum(pc_table[which(pc_table$State != c("8_ZNF/Rpts","9_Het","Missing")),]$Instances)

# Instances >1% bp of state, also enriched
pc_table = merge(pc_table,as.data.frame(table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR),]$State)),by.x="State",by.y="Var1",all=TRUE)
colnames(pc_table)[10] = "Enriched_Instances"
pc_table$Percent = pc_table$Enriched_Instances/pc_table$Instances
pc_table

# Number of subfamilies >1% across all states
length(unique(subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0),]$subfamily))
length(unique(subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0 & !(subfamily_state_sample_counts_pc$State %in% c("8_ZNF/Rpts","9_Het","Missing"))),]$subfamily))

# Subfamilies with >1% of bases in genome
GENOME_WIDTH*THRESHOLD_PC
rmsk_TE_subfamily[which(rmsk_TE_subfamily$Total_length > GENOME_WIDTH*THRESHOLD_PC),c("subfamily","Count","Total_length","Total_length_noY","CpGs")]

# Subfamilies with >1% of CpGs in genome
ALL_CPGS*THRESHOLD_PC
rmsk_TE_subfamily[which(rmsk_TE_subfamily$CpGs > ALL_CPGS*THRESHOLD_PC),c("subfamily","Count","Total_length","Total_length_noY","CpGs")]

# Largest subfamilies vs. >1% of state
test = merge(dcast(subfamily_state_sample_counts_pc,subfamily+family+class_update~State,value.var = "V1"),rmsk_TE_subfamily[,c("subfamily","Count","Total_length","CpGs")],by="subfamily",all=TRUE)

test = test[order(test$Total_length,decreasing=TRUE),]
rownames(test) <- NULL
test$Order = rownames(test)
lapply(c(chromHMM_states,"DNase","H3K27ac"),function(x) if(pc_table[which(pc_table$State == x),]$Enriched_Instances > 0) {join(test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)],aggregate(data=subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x & subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],Enrichment~subfamily,function(y) sum(y > THRESHOLD_LOR)),by="subfamily",type="left")} else{test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)]})

test = test[order(test$CpGs,decreasing=TRUE),]
rownames(test) <- NULL
test$Order = rownames(test)
lapply(meth_states,function(x) if(pc_table[which(pc_table$State == x),]$Enriched_Instances > 0) {join(test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)],aggregate(data=subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x & subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],Enrichment~subfamily,function(y) sum(y > THRESHOLD_LOR)),by="subfamily",type="left")} else{test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)]})
```

### Subfamily enrichment clustering and stats (for Excel)

```{r enrichment clusters, eval=FALSE}
#source("R_scripts/subfamily_enrichment.R")
source("R_scripts/TE_correlation_subfamily.R") #Requires subfamily enrichment

enrichment_lor = c(lapply(chromHMM_states,function(x) enrichment_proportion(subfamily_state_sample_filter[which(subfamily_state_sample_filter$State == x),],"Enrichment",THRESHOLD_LOR,"chromHMM")),
                      lapply(meth_states,function(x) enrichment_proportion(subfamily_state_sample_filter[which(subfamily_state_sample_filter$State == x),],"Enrichment",THRESHOLD_LOR,"WGBS")),
                      lapply(c("DNase","H3K27ac"),function(x) enrichment_proportion(subfamily_state_sample_filter[which(subfamily_state_sample_filter$State == x),],"Enrichment",THRESHOLD_LOR,x)))
names(enrichment_lor) = c(chromHMM_states,meth_states,"DNase","H3K27ac")
enrichment_lor_0.5 = ldply(enrichment_lor,function(x) x[which(x$Proportion > 0.5 & x$Enriched > 1),])
enrichment_lor_0.5 = dcast(aggregate(data = enrichment_lor_0.5[,c(1:3,5)], Category~.id+subfamily+Metadata, paste, collapse = ','),.id+subfamily~Metadata,value.var="Category")
colnames(enrichment_lor_0.5)[1] = "State"

enrichment_pc = c(lapply(chromHMM_states,function(x) enrichment_proportion(subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x),],"Length_percent_jk",THRESHOLD_PC,"chromHMM")),
                      lapply(meth_states,function(x) enrichment_proportion(subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x),],"Length_percent_jk",THRESHOLD_PC,"WGBS")),
                      lapply(c("DNase","H3K27ac"), function(x) enrichment_proportion(subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x),],"Length_percent_jk",THRESHOLD_PC,x)))
names(enrichment_pc) = c(chromHMM_states,meth_states,"DNase","H3K27ac")
enrichment_pc_0.5 = ldply(enrichment_pc,function(x) x[which(x$Proportion > 0.5 & x$Enriched > 1),])
enrichment_pc_0.5 = dcast(aggregate(data = enrichment_pc_0.5[,c(1:3,5)], Category~.id+subfamily+Metadata, paste, collapse = ','),.id+subfamily~Metadata,value.var="Category")
colnames(enrichment_pc_0.5) = c("State","subfamily","Age_PC","Anatomy_PC","Cancer_PC","Germline_PC","Group_PC","Type_PC")

test = melt(rmsk_TE_subfamily_measure[,c(1,60:80)],id.vars=c("subfamily"))
test$variable = gsub("_PC","",test$variable)
test = merge(melt(rmsk_TE_subfamily_measure[,c(1,39:59)],id.vars=c("subfamily")),test,by=c("subfamily","variable"))
colnames(test)[2:4] = c("State","Samples_enriched","Samples_1pc")

test = merge(test,enrichment_lor_0.5,by=c("State","subfamily"),all.x=TRUE)
test = merge(test,enrichment_pc_0.5,by=c("State","subfamily"),all.x=TRUE)
write.table(test,file="enrichment/state_subfamily_clusters.txt",sep='\t',row.names = FALSE,quote = FALSE)

# Subfamily features
write.table(rmsk_TE_subfamily_measure[,c(1:5,35,33:34,37:38,7,9,12:13,16,19,21,24:25,28,31:32,81)],file="enrichment/subfamily_features.txt",sep='\t',row.names=FALSE,quote=FALSE)
```

### Investigate individual subfamilies (for Excel)
### RERUN when updated

```{r analysis Fig S14d, eval=FALSE}
#source("R_scripts/TE_correlation.R")
rmsk_TE_subfamily_ever = aggregate(data=rmsk_TE_measure[,c(4,35:49,52:55,60:62)],.~subfamily,function(x) sum(x > 0,na.rm=TRUE),na.action=na.pass)
colnames(rmsk_TE_subfamily_ever)[9] = "X8_ZNF/Rpts"
colnames(rmsk_TE_measure)[42] = "X8_ZNF/Rpts"

lapply(names(all_state_labels)[1:21],function(x) write_subfamily_candidates(subfamily_state_sample_counts[which(subfamily_state_sample_counts$State == x & subfamily_state_sample_counts$V1 > 1),]$subfamily,x,FALSE))
```

### Core subfamily members
### Relies on intervening unix analysis

```{r core subfamily members}
# See 8/29/2016, 9/8/2016, 1/20/2017, 2/10/2017, 3/16/2017

# Subfamily members in the state when the subfamily is enriched
subfamily_state_individual_V1 = read.table(file="enrichment/state_enriched.bed",sep='\t')
colnames(subfamily_state_individual_V1) = c("State","chromosome","start","stop","subfamily","Samples","strand")
subfamily_state_individual_V1$State = gsub("8_ZNF.Rpts","8_ZNF/Rpts",subfamily_state_individual_V1$State)
subfamily_state_individual_V1 = merge(subfamily_state_individual_V1,subfamily_state_sample_counts,by=c("subfamily","State"),all.x=TRUE)
subfamily_state_individual_V1 = subfamily_state_individual_V1[,c(3:5,1,8:9,7,2,6,10)]
subfamily_state_individual_V1$Pc_of_enriched = subfamily_state_individual_V1$Samples/subfamily_state_individual_V1$V1
subfamily_state_individual_V1$State = factor(subfamily_state_individual_V1$State,levels=c(chromHMM_states,meth_states,"DNase","H3K27ac"))

# Average % of individual members in state in at least 25%, 50%, 75% enriched samples, by state
test = ddply(subfamily_state_individual_V1,.(subfamily,State),summarise,Quarter1=length(Pc_of_enriched[which(Pc_of_enriched > 0.25)])/length(Pc_of_enriched),Half=length(Pc_of_enriched[which(Pc_of_enriched > 0.5)])/length(Pc_of_enriched),Quarter2=length(Pc_of_enriched[which(Pc_of_enriched > 0.75)])/length(Pc_of_enriched))

aggregate(data=test[,2:5],.~State,mean)
```

### Supplementary Figure 15

```{r Figure S15, echo=FALSE}
ggplot(subfamily_state_individual_V1,aes(x=Pc_of_enriched)) + geom_histogram(bins=10) + facet_wrap(~State,scale="free_y") + ylab("TEs") + xlab("Percent of enriched samples TE is in state")
```

### Core subfamily members (for Excel)

```{r, eval=FALSE}
write.table(ddply(subfamily_state_individual_V1,.(subfamily,State),summarise,Quarter1=length(Pc_of_enriched[which(Pc_of_enriched > 0.25)]),Half=length(Pc_of_enriched[which(Pc_of_enriched > 0.5)]),Quarter2=length(Pc_of_enriched[which(Pc_of_enriched > 0.75)])),file="enrichment/subfamily_state_individual_V1_core.txt",sep='\t',quote=FALSE,row.names=FALSE)
```

## Enrichment examples

### Figure 6

```{r Figure 6, echo=FALSE}
tf = read.table("enrichment/TF_matrix.txt",sep='\t',header=TRUE)
tf = tf[,1:4]
tf$Target_PC = as.numeric(gsub("%","",tf$Target_PC))
tf$Background_PC = as.numeric(gsub("%","",tf$Background_PC))
tf = melt(tf,id.vars=c("subfamily","TF"))

ggplot(tf,aes(x=TF,y=value,fill=variable)) + geom_bar(stat="identity",position="dodge") + facet_grid(~subfamily,scales="free_x",space="free_x") + ylab("% TEs with motif") + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x = element_blank()) + scale_fill_discrete(guide=FALSE)
```

### Supplementary Figure 16

**Supplementary Figure 16. Subfamily metrics.** (C) Density plot of mean subfamily mappability, by class. (E) Density plot of mean subfamily age, by class.

```{r Figure S16, echo=FALSE}
#source("R_scripts/TE_subfamily_stats.R")

a = ggplot(rmsk_TE_subfamily,aes(x=log10(Count),fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + xlab("Number of TEs (log10)") + ylab("Density") + facet_wrap(~class_update,nrow=1,scales="free_y")

b = ggplot(rmsk_TE_subfamily,aes(x=log10(Total_length),fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + xlab("Total length (bp) (log10)") + ylab("Density") + facet_wrap(~class_update,nrow=1,scales="free_y")

c = ggplot(rmsk_TE_subfamily,aes(x=log10(CpGs),fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + xlab("Total CpGs (log10)") + ylab("Density") + facet_wrap(~class_update,nrow=1,scales="free_y")

d = ggplot(rmsk_TE_subfamily,aes(x=Mappability,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + xlab("Mean mappability") + ylab("Density") + xlim(0,1) + facet_wrap(~class_update,nrow=1,scales="free_y")

e = ggplot(rmsk_TE_subfamily,aes(x=Age,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) +  xlab("Mean Jukes-Cantor evolutionary distance") + ylab("Density") + facet_wrap(~class_update,nrow=1,scales="free_y")

grid.arrange(a,b,c,d,e,ncol=1)
```

### Summary of subfamily metrics

```{r subfamily metrics}
source("R_scripts/TE_correlation_subfamily.R")

# Subfamily metrics, overall and by class, median and mean, with p-values
apply(rmsk_TE_subfamily[,c(4:5,7,9,33:35,37:38)],2,median)
aggregate(data=rmsk_TE_subfamily[,c(3:5,7,9,33:35,37:38)],.~class_update,median)

colMeans(rmsk_TE_subfamily[,c(4:5,7,9,33:35,37:38)])
aggregate(data=rmsk_TE_subfamily[,c(3:5,7,9,33:35,37:38)],.~class_update,mean)
sort(apply(aggregate(data=rmsk_TE_subfamily[,c(3:5,7,9,33:35,37:38)],.~class_update,mean)[,2:10],2,function(x) sd(x)/mean(x)))

apply(rmsk_TE_subfamily[,c(4:5,7,9,33:35,37:38)],2,function(x) unlist(kruskal.test(x~rmsk_TE_subfamily$class_update))["p.value"])

# Subfamily metric outliers
lapply(c(4:5,7,9,33:35,37:38),function(x) rmsk_TE_subfamily[which(rmsk_TE_subfamily[,x] > mean(rmsk_TE_subfamily[,x])+2*sd(rmsk_TE_subfamily[,x])),c(1:4,x)])
lapply(c(4:5,7,9,33:35,37:38),function(x) rmsk_TE_subfamily[which(rmsk_TE_subfamily[,x] < mean(rmsk_TE_subfamily[,x])-2*sd(rmsk_TE_subfamily[,x])),c(1:4,x)])

# Subfamilies with very low mappability
rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),c("subfamily","family","class_update","Count","Mappability")]
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),])[1]
table(droplevels(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),]$family))
#ggplot(rmsk_TE_subfamily,aes(x=Age,y=Mappability)) + geom_point(size=0.5) + geom_text(aes(label=subfamily),size=3)

# Age of SINE subfamilies
test = rmsk_TE_subfamily[which(rmsk_TE_subfamily$class_update == "SINE"),]
test[order(test$Age),c("subfamily","family","class_update","Count","Age")]

# Subfamily feature overlap by class
apply(rmsk_TE_subfamily[,12:32],2,median)
aggregate(data=rmsk_TE_subfamily[,c(3,12:32)],.~class_update,median)

colMeans(rmsk_TE_subfamily[,12:32])
aggregate(data=rmsk_TE_subfamily_measure[,c(3,12:32)],.~class_update,mean)
sort(apply(aggregate(data=rmsk_TE_subfamily_measure[,c(3,12:32)],.~class_update,mean)[,2:22],2,function(x) sd(x)/mean(x)))

sort(apply(rmsk_TE_subfamily_measure[,12:32],2,function(x) as.numeric(unlist(kruskal.test(x~rmsk_TE_subfamily_measure$class_update))["p.value"])))

# Subfamily feature overlap outliers
#ggplot(melt(rmsk_TE_subfamily[,12:32]),aes(x=value)) + geom_histogram() + facet_wrap(~variable)
lapply(c(12:13,16,19,21,24:25,28,31:32),function(x) rmsk_TE_subfamily[which(rmsk_TE_subfamily[,x] > 2*sd(rmsk_TE_subfamily[,x])+mean(rmsk_TE_subfamily[,x])),c(1:4,x)])

# CGate subfamilies
length(cgate_subfams)
```

### Subfamily enrichment correlation

```{r analysis Fig S15a}
# Metrics - 4:5,7,9,33:35,37:38
# Features - 12:32 (81)
# Enrichment samples - 39:59
# >1% samples - 60:80

# Number of samples by class
sort(apply(rmsk_TE_subfamily_measure[,39:80],2,function(x) as.numeric(unlist(kruskal.test(x~rmsk_TE_subfamily_measure$class_update))["p.value"])))
aggregate(data=rmsk_TE_subfamily_measure[,c(3,39:80)],.~class_update,mean)
sort(apply(aggregate(data=rmsk_TE_subfamily_measure[,c(3,39:80)],.~class_update,mean)[,2:43],2,function(x) sd(x)/mean(x)))

# Correlate samples enriched/>1% in state with subfamily features
correlate_subfamily = ldply(colnames(rmsk_TE_subfamily_measure)[c(4:5,7,9,12:35,37:38)],function(x) correlate_spearman(rmsk_TE_subfamily_measure,x,39:80))
correlate_subfamily$Metric = rep(colnames(rmsk_TE_subfamily_measure)[c(4:5,7,9,12:35,37:38)],each=294)
correlate_subfamily$Metric = factor(correlate_subfamily$Metric,levels=unique(correlate_subfamily$Metric)[c(1:2,28,3:4,26:27,29:30,5:25)])
correlate_subfamily$p.value = as.numeric(as.character(correlate_subfamily$p.value))
correlate_subfamily$estimate.rho = as.numeric(as.character(correlate_subfamily$estimate.rho))
correlate_subfamily$class_update = factor(correlate_subfamily$class_update,levels=c("DNA","LINE","LTR","SINE","SVA","Other","All"))

correlate_subfamily[which(correlate_subfamily$p.value < 0.001 & abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfamily$Metric %in% unique(correlate_subfamily$Metric)[c(1:4,26:30)] & correlate_subfamily$class_update == "All"),]

correlate_subfamily[which(correlate_subfamily$p.value < 0.001 & abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfamily$Metric %in% unique(correlate_subfamily$Metric)[c(1:4,26:30)] & correlate_subfamily$class_update != "All"),]

correlate_subfamily[which(correlate_subfamily$p.value < 0.001 & abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfamily$Metric %in% names(genic_labels) & correlate_subfamily$class_update == "All"),]

correlate_subfamily[which(correlate_subfamily$p.value < 0.001 & abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfamily$Metric %in% names(genic_labels) & correlate_subfamily$class_update != "All"),]

# Correlation between input variables (including feature ratio)
correlate_subfam_feature = ldply(colnames(rmsk_TE_subfamily_measure)[c(4:5,7,9,12:35,37:38)],function(x) correlate_spearman(rmsk_TE_subfamily_measure,x,c(4:5,7,9,12:35,37:38)))
correlate_subfam_feature$Metric = rep(colnames(rmsk_TE_subfamily_measure)[c(4:5,7,9,12:35,37:38)],each=210)
correlate_subfam_feature$p.value = as.numeric(as.character(correlate_subfam_feature$p.value))
correlate_subfam_feature$estimate.rho = as.numeric(as.character(correlate_subfam_feature$estimate.rho))
correlate_subfam_feature$class_update = factor(correlate_subfam_feature$class_update,levels=c("DNA","LINE","LTR","SINE","SVA","Other","All"))

correlate_subfam_feature[which(correlate_subfam_feature$p.value < 0.001 & abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% unique(correlate_subfamily$Metric)[c(1:4,26:30)] & correlate_subfam_feature$Metric %in% unique(correlate_subfamily$Metric)[c(1:4,26:30)] & correlate_subfam_feature$class_update == "All"),]

# Compare states to each other
correlate_subfam_state = ldply(colnames(rmsk_TE_subfamily_measure)[39:80],function(x) correlate_spearman(rmsk_TE_subfamily_measure,x,39:80))
correlate_subfam_state$State2 = rep(colnames(rmsk_TE_subfamily_measure)[39:80],each=294)
correlate_subfam_state$p.value = as.numeric(as.character(correlate_subfam_state$p.value))
correlate_subfam_state$estimate.rho = as.numeric(as.character(correlate_subfam_state$estimate.rho))
correlate_subfam_state$class_update = factor(correlate_subfam_state$class_update,levels=c("DNA","LINE","LTR","SINE","SVA","Other","All"))
```

### Supplementary Figure 17

```{r Figure S17, echo=FALSE}
a = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfamily$Metric %in% unique(correlate_subfamily$Metric)[c(1:4,26:30)] & correlate_subfamily$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(c(chromHMM_states,meth_states,"DNase","H3K27ac"))) + scale_x_discrete(limits=levels(correlate_subfamily$Metric)[c(1,3,6,4:5,2,7:9)])

b = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfamily$Metric %in% unique(correlate_subfamily$Metric)[c(1:4,26:30)] & correlate_subfamily$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(c(chromHMM_states,meth_states,"DNase","H3K27ac"))) + scale_x_discrete(limits=levels(correlate_subfamily$Metric)[c(1,3,6,4:5,2,7:9)]) + facet_wrap(~class_update)

c = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfamily$Metric %in% c(names(genic_labels),"Vista_enhancers") & correlate_subfamily$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(c(chromHMM_states,meth_states,"DNase","H3K27ac"))) + scale_x_discrete(limits=c(names(genic_labels),"Vista_enhancers"),labels=c(genic_labels,"Vista enhancers"))

d = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfamily$Metric %in% c(names(genic_labels),"Vista_enhancers") & correlate_subfamily$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(c(chromHMM_states,meth_states,"DNase","H3K27ac"))) + scale_x_discrete(limits=c(names(genic_labels),"Vista_enhancers"),labels=c(genic_labels,"Vista enhancers")) + facet_wrap(~class_update)

# Metric vs. metric
e = ggplot(correlate_subfam_feature[which(correlate_subfam_feature$State %in% unique(correlate_subfamily$Metric)[c(1:4,26:30)] & correlate_subfam_feature$Metric %in% unique(correlate_subfamily$Metric)[c(1:4,26:30)] & correlate_subfam_feature$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_x_discrete(limits=levels(correlate_subfamily$Metric)[c(1,3,6,4:5,2,7:9)]) + scale_y_discrete(limits=levels(correlate_subfamily$Metric)[rev(c(1,3,6,4:5,2,7:9))])

# State vs. state, PC
f = ggplot(correlate_subfam_state[which(correlate_subfam_state$State %in% unique(correlate_subfamily$State)[22:42] & correlate_subfam_state$State2 %in% unique(correlate_subfamily$State)[22:42] & correlate_subfam_state$class_update == "All"),],aes(x=State,y=State2,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1))

# State vs. state
g = ggplot(correlate_subfam_state[which(correlate_subfam_state$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfam_state$State2 %in% unique(correlate_subfamily$State)[1:21] & correlate_subfam_state$class_update == "All"),],aes(x=State,y=State2,fill=estimate.rho)) + geom_tile() + facet_wrap(~class_update) + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1))

h = ggplot(correlate_subfam_state[which(correlate_subfam_state$State %in% unique(correlate_subfamily$State)[1:21] & correlate_subfam_state$State2 %in% unique(correlate_subfamily$State)[1:21] & correlate_subfam_state$class_update != "All"),],aes(x=State,y=State2,fill=estimate.rho)) + geom_tile() + facet_wrap(~class_update) + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1))

grid.arrange(a,b,c,d,e,f,g,h,nrow=4,layout_matrix=cbind(c(1,3,5,7),c(2,4,6,8)))
```

### Analysis

```{r analysis Fig S15b}
# Age vs. number of members
correlate_subfam_feature[which(correlate_subfam_feature$Metric == "Count" & correlate_subfam_feature$State == "Age"),]

# Size vs. # samples enriched
correlate_subfamily[which(correlate_subfamily$Metric == "Total_length" & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.01 & correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21]),]

# Overlap with VISTA enhancers
dim(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),c(1:7,33)])
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),c(1:7,33)]$subfamily)))
table(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),c(1:7,33)]$subfamily)))

correlate_subfamily[which(correlate_subfamily$Metric == "Vista_enhancers" & correlate_subfamily$State %in% unique(correlate_subfamily$State)[1:21] & abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$p.value < 0.01),]
```

### MER57E3

```{r mer57e3}
subfamily_state_sample_counts[which(subfamily_state_sample_counts$subfamily == "MER57E3"),]
MER57E3 = rmsk_TE[which(rmsk_TE$subfamily == "MER57E3"),]
dim(MER57E3)[1]
table(MER57E3$chromosome)
apply(MER57E3[,15:33],2,function(x) dim(MER57E3[which(!is.na(x)),])[1])
```

### Young subfamilies

```{r young subfamilies}
young_subfamilies = as.vector(rmsk_TE_subfamily_measure[which(rmsk_TE_subfamily_measure$Age < mean(rmsk_TE_subfamily_measure$Age)-2*sd(rmsk_TE_subfamily_measure$Age)),]$subfamily)
test = rmsk_TE_subfamily_measure[which(rmsk_TE_subfamily_measure$subfamily %in% young_subfamilies),c(1:3,39:80)]
test[,which(apply(test,2,function(x) sum(x > 0) > 0))]
```

## Mouse

### Figure 7

**Figure 7. Epigenetic state of orthologous TEs in human and mouse.** (A) Density plot of the methylation level of TE orthologs in fetal forebrain (ENCFF592SYK) /ganglion eminence-derived primary cultured neurospheres (E054) in human and mouse. (B) Pearson correlation methylation level of orthologous TEs within species, across species in equivalent samples, or across species in randomly paired samples. (C) Pearson correlation for subfamilies present in both human and mouse, proportion of members hypomethylated. (D) Scatter plot of the proportion of hypomethylated members for subfamilies present in both human in mouse in equivalent samples. (E) An example TE hypomethylated in both human and mouse and in a promoter chromHMM state in both. 

```{r Figure 7, echo=FALSE}
#source("R_scripts/WGBS_subfamily_enrichment_TE.R")
source("R_scripts/human_mouse_ortholog_WGBS.R") #Requires WGBS_subfamily_enrichment_TE.R

a = ggplot(human_mouse_orthologs_WGBS,aes(x=E054,y=ENCFF592SYK/100)) + geom_point(alpha=0.1) + ylab("Mouse TE methylation level") + xlab("Human TE methylation level") + geom_density2d()

b = ggplot(hg19_mm10_TE_WGBS_subfamily_hypo_paired,aes(x=Human,y=Mouse,color=Pair)) + geom_point(alpha=0.5) + scale_color_discrete(guide=FALSE) + xlab("Human subfamily % hypomethylated") + ylab("Mouse subfamily % hypomethylated")

grid.arrange(a,b,nrow=1)

# Mosaic plot
human_mouse_orthologs_WGBS_category_pairs = lapply(seq(1,13,1),function(x) table(human_mouse_orthologs_WGBS_category[,c(as.vector(human_mouse_samples[x,1]),as.vector(human_mouse_samples[x,2]))]))
human_mouse_orthologs_WGBS_category_pairs_summed = Reduce('+',human_mouse_orthologs_WGBS_category_pairs)
names(attributes(human_mouse_orthologs_WGBS_category_pairs_summed)$dimnames) <- c("Mouse","Human")

c = mosaic(human_mouse_orthologs_WGBS_category_pairs_summed[c(2,3,1),c(2,3,1)],shade=TRUE,return_grob=TRUE)
```

### Supplementary Figure 18

**Supplementary Figure 18. Mouse TE statistics.** (A) Proportion of human TEs and TE subfamilies with orthologs in mouse, by class. (B) chromHMM states in the 7-state model, defining histone modifications, and the 15-state model state with a similar definition.

```{r Figure S18, echo=FALSE}
source("R_scripts/TE_ortholog_class_stats.R")
#source("R_scripts/human_mouse_ortholog_WGBS.R")

par(mfrow=c(1,2)) 

pie(rmsk_TE_class$Mouse_orthologs,labels=paste(rmsk_TE_class$Mouse_orthologs,"(",round(rmsk_TE_class$Mouse_orthologs/sum(rmsk_TE_class$Mouse_orthologs)*100,0),"%)",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="Orthologous TEs",cex=1,cex.main=2,line=-10,radius=0.5)

pie(rmsk_TE_class$Mouse_ortholog_subfamily,labels=paste(rmsk_TE_class$Mouse_ortholog_subfamily,"(",round(rmsk_TE_class$Mouse_ortholog_subfamily/sum(rmsk_TE_class$Mouse_ortholog_subfamily)*100,0),"%)",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="Orthologous subfamilies",cex=1,cex.main=2,line=-10,radius=0.5)

b = ggplot(human_mouse_samples,aes(x=Test,y=Spearman)) + geom_boxplot() + ylab("Spearman correlation across TEs") + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse, same tissue","Human-mouse, random tissue")) + scale_y_continuous(limits=c(0,1))

c = ggplot(human_mouse_samples,aes(x=Test,y=Subfamily_Spearman)) + geom_boxplot() + ylab("Spearman correlation across subfamilies") + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_x_discrete(labels=c("Human-human, random tissue","Mouse-mouse, random tissue","Human-mouse, same tissue","Human-mouse, random tissue")) + scale_y_continuous(limits=c(0,1))

hg19_mm10_TE_WGBS_subfamily_hypo_paired_age = merge(ddply(hg19_mm10_TE_WGBS_subfamily_hypo_paired,.(subfamily),summarise,unlist(cor.test(Mouse,Human,method="spearman"))["estimate.rho"]),rmsk_TE_subfamily[c("subfamily","Age")],by="subfamily")
colnames(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age)[2] = "Spearman_correlation"
hg19_mm10_TE_WGBS_subfamily_hypo_paired_age$Spearman_correlation = as.numeric(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age$Spearman_correlation)

d = ggplot(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age,aes(x=Age,y=Spearman_correlation)) + geom_point() + ylab("Spearman's rho")

grid.arrange(b,c,d,nrow=2,layout_matrix=rbind(c(1,2),c(3,3)))
```

### Mouse ortholog analysis

```{r analysis Fig S16a}
# Number of mouse TEs
MOUSE_TE = dim(mm10_rmsk_TE)[1]
MOUSE_TE

# Number of hg19 TEs that liftover
# (wc -l Mouse/liftover/rmsk_TE_hg19tomm10.bed)
1265775
1265775/NUM_TE

# Number of unique orthologous TEs in hg19->mm10, hg19, and mm10
dim(unique(human_mouse_orthologs_mm10[,1:4]))[1]
dim(unique(human_mouse_orthologs_mm10[,5:11]))[1]
dim(unique(human_mouse_orthologs_mm10[,5:11]))[1]/NUM_TE
dim(unique(human_mouse_orthologs_mm10[,12:18]))[1]

# Number, % of human TEs with 2+ human mm10 TEs
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_hg19))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_mm10))[2:11])/length(unique(human_mouse_orthologs_mm10$human_TE_mm10))

# Number, % of human TEs with 2+ mouse TEs
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$human_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$human_TE_mm10))[2:10])/length(unique(human_mouse_orthologs_mm10$human_TE_mm10))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$human_TE_hg19))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$human_TE_hg19))[2:10])/length(unique(human_mouse_orthologs_mm10$human_TE_hg19))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$mouse_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$mouse_TE_mm10))[2:15])/length(unique(human_mouse_orthologs_mm10$mouse_TE_mm10))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$mouse_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$mouse_TE_mm10))[2:25])/length(unique(human_mouse_orthologs_mm10$mouse_TE_mm10))

# Average length of those TEs
median(unique(human_mouse_orthologs_mm10[,1:4])$human_stop_mm10-unique(human_mouse_orthologs_mm10[,1:4])$human_start_mm10)
median(unique(human_mouse_orthologs_mm10[,5:11])$human_stop_hg19-unique(human_mouse_orthologs_mm10[,5:11])$human_start_hg19)
median(unique(human_mouse_orthologs_mm10[,12:18])$mouse_stop_mm10-unique(human_mouse_orthologs_mm10[,12:18])$mouse_start_mm10)

# Average age of TEs with mouse orthologs
median(merge(rmsk_TE_measure,unique(human_mouse_orthologs_mm10[,5:11]),by.x=c("chromosome","start","stop","subfamily","family","class","strand"),by.y=c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_family","human_class","human_strand_hg19"))$JC_distance)

# Number of subfamilies shared between human and mouse
length(intersect(levels(mm10_rmsk_TE$subfamily),levels(rmsk_TE$subfamily)))

# Average age of subfamilies also in mouse 
median(rmsk_TE_subfamily[which(rmsk_TE_subfamily$subfamily %in% mm10_rmsk_TE$subfamily),]$Age)

# Proportion of hg19 TEs and subfamilies with ortholog in mouse, by class
rmsk_TE_class$Mouse_orthologs/rmsk_TE_class$Count
rmsk_TE_class$Mouse_ortholog_subfamily/rmsk_TE_class$Subfamilies
```

### Mouse WGBS analysis

```{r analysis Fig S16b}
# Number, % of mm10 TEs with CpGs
dim(mm10_rmsk_TE_CpG_count)[1]
dim(mm10_rmsk_TE_CpG_count)[1]/MOUSE_TE

# Number, % of hg19 orthologs with CpGs
dim(merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,5:11]),by.x=colnames(TE_CpG_count)[1:7],by.y=colnames(human_mouse_orthologs_mm10)[5:11]))
dim(merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,5:11]),by.x=colnames(TE_CpG_count)[1:7],by.y=colnames(human_mouse_orthologs_mm10)[5:11]))[1]/dim(unique(human_mouse_orthologs_mm10[,5:11]))[1]

# Average number of CpGs per hg19 ortholog with CpGs
mean(merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,5:11]),by.x=colnames(TE_CpG_count)[1:7],by.y=colnames(human_mouse_orthologs_mm10)[5:11])$CpGs)
median(merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,5:11]),by.x=colnames(TE_CpG_count)[1:7],by.y=colnames(human_mouse_orthologs_mm10)[5:11])$CpGs)

# Number, % of mm10 orthologs with CpGs
dim(merge(mm10_rmsk_TE_CpG_count,unique(human_mouse_orthologs_mm10[,12:18]),by.x=colnames(mm10_rmsk_TE_CpG_count)[1:7],by.y=colnames(human_mouse_orthologs_mm10)[12:18]))
dim(merge(mm10_rmsk_TE_CpG_count,unique(human_mouse_orthologs_mm10[,12:18]),by.x=colnames(mm10_rmsk_TE_CpG_count)[1:7],by.y=colnames(human_mouse_orthologs_mm10)[12:18]))[1]/dim(unique(human_mouse_orthologs_mm10[,12:18]))[1]

# Number of unique orthologous TEs with CpGs in mm10, hg19, and hg19->mm10
dim(unique(human_mouse_orthologs_WGBS[,1:7]))[1]
dim(unique(human_mouse_orthologs_WGBS[,8:14]))[1]
dim(unique(human_mouse_orthologs_WGBS[,15:18]))[1]

# TEs with meth data in at least one mouse/human sample
test = human_mouse_orthologs_WGBS[which(apply(human_mouse_orthologs_WGBS,1,function(x) sum(is.na(x[20:32])) < 13) & apply(human_mouse_orthologs_WGBS,1,function(x) sum(is.na(x[33:41])) < 9)),]

# Number of human (hg19) TEs
dim(unique(human_mouse_orthologs_WGBS[,8:14]))[1]-dim(unique(test[,8:14]))[1]

# Number of mouse TEs
dim(unique(human_mouse_orthologs_WGBS[,1:7]))[1]-dim(unique(test[,1:7]))[1]

# Number of TEs with methylation data in E054 and ENCFF592SYK
dim(human_mouse_orthologs_WGBS[which(!is.na(human_mouse_orthologs_WGBS$E054) & !is.na(human_mouse_orthologs_WGBS$ENCFF592SYK)),])

# Individual TE Spearman correlations
ddply(human_mouse_samples,.(Test),summarise,Mean=mean(Spearman),SD=sd(Spearman))

# Number of pairs where TE is hypomethylated in both
dim(human_mouse_orthologs_mm10_hypo)
dim(unique(human_mouse_orthologs_mm10_hypo[,c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")]))
dim(unique(human_mouse_orthologs_mm10_hypo[,c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10")]))
dim(unique(human_mouse_orthologs_mm10_hypo[,c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19","Human_sample")]))
dim(unique(human_mouse_orthologs_mm10_hypo[,c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10","Mouse_sample_mm10")]))

# Number of subfamilies shared between human and mouse with CpGs in both 
dim(hg19_mm10_TE_WGBS_subfamily_hypo)[1]
setdiff(intersect(levels(mm10_rmsk_TE$subfamily),levels(rmsk_TE$subfamily)),hg19_mm10_TE_WGBS_subfamily_hypo$subfamily)

# Subfamily Spearman correlations 
ddply(human_mouse_samples,.(Test),summarise,Mean=mean(Subfamily_Spearman),SD=sd(Subfamily_Spearman))

# Subfamily age vs. degree of correlation with mouse
cor.test(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age$Age,as.numeric(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age$Spearman_correlation),method="spearman")
```

### Chi-sq test for human-mouse pairs (WGBS)

```{r analysis Fig S16c}
# All methylation states
lapply(human_mouse_orthologs_WGBS_category_pairs,function(x) unlist(chisq.test(x))["p.value"])

# Hypomethylated vs. all other states
apply(human_mouse_samples[1:13,],1,function(x) unlist(chisq.test(matrix(c(nrow(human_mouse_orthologs_WGBS_category[which(human_mouse_orthologs_WGBS_category[,x[1]] == "Hypomethylated" & human_mouse_orthologs_WGBS_category[,x[2]] == "Hypomethylated"),]),nrow(human_mouse_orthologs_WGBS_category[which(human_mouse_orthologs_WGBS_category[,x[1]] == "Hypomethylated" & human_mouse_orthologs_WGBS_category[,x[2]] %in% c("Hypermethylated","Intermediate")),]),nrow(human_mouse_orthologs_WGBS_category[which(human_mouse_orthologs_WGBS_category[,x[1]] %in% c("Hypermethylated","Intermediate") & human_mouse_orthologs_WGBS_category[,x[2]] == "Hypomethylated"),]),nrow(human_mouse_orthologs_WGBS_category[which(human_mouse_orthologs_WGBS_category[,x[1]] %in% c("Hypermethylated","Intermediate") & human_mouse_orthologs_WGBS_category[,x[2]] %in% c("Hypermethylated","Intermediate")),])),nrow=2)))["p.value"])
```

### Comparison of human-mouse TE ortholog methylation and chromHMM state
### RERUN when updated

```{r write chromHMM, eval=FALSE}
write.table(unique(human_mouse_orthologs_mm10_hypo[,c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19","Human_sample")]),file="Mouse/hg19_ortholog_hypo.txt",sep="\t",row.names=FALSE,col.names=FALSE,quote=FALSE)
write.table(unique(human_mouse_orthologs_mm10_hypo[,c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10")]),file="Mouse/mm10_ortholog_hypo.txt",sep="\t",col.names=FALSE,row.names=FALSE,quote=FALSE)
```

### Relies on intervening unix analysis

```{r}
# Human chromHMM state for hg19 TEs hypomethylated in mouse/human
hg19_orthologs_hypo_chromHMM = read.table("Mouse/hg19_ortholog_hypo_chromHMM.txt",sep='\t')[,1:9]
colnames(hg19_orthologs_hypo_chromHMM) = c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19","Human_sample","Human_state")

# mm9 chromHMM state for mm10 TEs hypomethylated in mouse/human
mm10_orthologs_hypo_chromHMM = read.table("Mouse/mm10_ortholog_hypo_chromHMM.txt",sep='\t')[,1:9]
colnames(mm10_orthologs_hypo_chromHMM) = c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10","Mouse_State","Mouse_sample_mm9")

# Combine chromHMM and WGBS
human_mouse_orthologs_mm10_hypo_chromHMM = merge(human_mouse_orthologs_mm10_hypo,hg19_orthologs_hypo_chromHMM,by=c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19","Human_sample"),all.x=TRUE)
human_mouse_orthologs_mm10_hypo_chromHMM = merge(human_mouse_orthologs_mm10_hypo_chromHMM,mm10_orthologs_hypo_chromHMM,by=c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10","Mouse_sample_mm9"),all.x=TRUE)
human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State = factor(human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State)
human_mouse_orthologs_mm10_hypo_chromHMM$Human_state = factor(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state,chromHMM_states)
```

### Supplementary Figure 19

**Supplementary Figure 19. ChromHMM states of human-mouse orthologs.**

```{r Figure S19, echo=FALSE}
# Mouse-human chromHMM states
human_mouse_chromHMM_states = matrix(c(chromHMM_states[c(1:2,7,6,4,15,13)],"1","2","3","4","5","6","7"),nrow=7)

mosaic(table(human_mouse_orthologs_mm10_hypo_chromHMM[,c("Human_state","Mouse_State")]),shade=TRUE,return_grob=TRUE)
```

### Human-mouse ortholog chromHMM analysis

```{r analysis Fig S17}
# Corresponding samples
chromHMM_mouse_samples = unique(human_mouse_orthologs_mm10_hypo_chromHMM[,c("Human_sample","Mouse_sample_mm9")])
table(human_mouse_orthologs_mm10_hypo_chromHMM$Human_sample,human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_sample_mm9)

# Number of pairs and unique TEs in each state
dim(human_mouse_orthologs_mm10_hypo_chromHMM)
dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[,c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")]))
dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[,c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19","Human_sample")]))
dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[,c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10")]))
dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[,c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10","Mouse_sample_mm9")]))

apply(human_mouse_chromHMM_states,1,function(x) { c(
  pairs = dim(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),])[1],
  human = dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")]))[1],
  mouse = dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10")]))[1]
)})

# Chi-squared for each sample pair (rows and columns with no data removed)
lapply(lapply(seq(1,11,1),function(x) table(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_sample == chromHMM_mouse_samples[x,1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_sample_mm9 == chromHMM_mouse_samples[x,2]),c("Human_state","Mouse_State")])),function(x) chisq.test(x[which(rowSums(x) > 0),which(colSums(x) > 0)]))

# Summarise hg19 TE class by state
apply(human_mouse_chromHMM_states,1,function(x) table(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")])$human_class))

# Summarise hg19 TE subfamily by state
apply(human_mouse_chromHMM_states,1,function(x) sort(table(droplevels(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")])$human_subfamily))))

test = apply(human_mouse_chromHMM_states,1,function(x) log2((table(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")])$human_subfamily)/sum(table(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")])$human_subfamily)))/(table(unique(human_mouse_orthologs_mm10[,c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")])$human_subfamily)/sum(table(unique(human_mouse_orthologs_mm10[,c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")])$human_subfamily)))))

test[order(test[,1]),]
test[order(test[,3]),]
test[order(test[,4]),]
```

```{r overlap with GENCODE,eval=FALSE}
write.table(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == "1_TssA" & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == "1"),c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")]),row.names=FALSE,col.names=FALSE,quote=FALSE,sep='\t',file="Mouse/hg19_orthologs_hypo_1TssA.bed")
```

# Methods 

## Blacklist TEs

```{r blacklist}
# Number of TEs overlapping blacklist
dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1]/NUM_TE

# Number of subfamilies overlapping blacklist
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$subfamily)))

# Which chromosomes?
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$chromosome)))

# Enrichment over blacklist
sort(log2((table(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$subfamily)/dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1])/(table(rmsk_TE$subfamily)/NUM_TE)))

# Blacklist overlap % outliers
rmsk_TE_subfamily[which(rmsk_TE_subfamily$blacklist > 2*sd(rmsk_TE_subfamily$blacklist)+mean(rmsk_TE_subfamily$blacklist)),c(1:4,31)]
```
