---
title: "TE landscape Draft"
author: "Erica Pehrsson"
date: "`r Sys.Date()`"
output: 
  word_document: 
    keep_md: yes
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(fig.width=7, fig.height=8, fig.path='TE_landscape_draft_figures/', cache.path="TE_landscape_draft_cache/")
                      #,error=TRUE)
```

# Setup

## Load required libraries

```{r load libraries}
library(plyr)
packageVersion("plyr")
library(reshape2)
packageVersion("reshape2")
library(vcd)
packageVersion("vcd")
library(ggplot2)
packageVersion("ggplot2")
library(NMF)
packageVersion("NMF")
library(gridExtra)
packageVersion("gridExtra")
library(RColorBrewer)
packageVersion("RColorBrewer")
library(Rtsne)
packageVersion("Rtsne")
```

## Set universal plot parameters

```{r set plot parameters}
theme_set(theme_bw() %+replace%  theme(panel.background=element_rect(fill="white"),panel.border=element_rect(color="black",fill=NA),panel.grid=element_blank()))
```

## Define color and label vectors

```{r define vectors}
# TE coordinates
TE_coordinates = c("chromosome","start","stop","subfamily","family","class","strand")
hg19_coordinates = c("human_chr_hg19","human_start_hg19","human_stop_hg19","human_subfamily","human_class","human_family","human_strand_hg19")
mm10_coordinates = c("mouse_chr_mm10","mouse_start_mm10","mouse_stop_mm10","mouse_subfamily","mouse_class","mouse_family","mouse_strand_mm10")

# Chromosomes
standard_chromosomes = c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY","chrM")

# chromHMM
chromHMM_states = c("1_TssA","2_TssAFlnk","3_TxFlnk","4_Tx","5_TxWk","6_EnhG","7_Enh","8_ZNF/Rpts","9_Het","10_TssBiv","11_BivFlnk","12_EnhBiv","13_ReprPC","14_ReprPCWk","15_Quies")
chromHMM_labels = setNames(c("Active TSS","Flanking active TSS","Transcription at gene 5'/3'","Strong transcription","Weak transcription","Genic enhancer","Enhancer","ZNF genes/repeats","Heterochromatin","Poised TSS","Flanking poised TSS/enhancer","Poised enhancer","Repressed Polycomb","Weak repressed Polycomb","Quiescent"),chromHMM_states)
chromHMM_colors = setNames(c(rgb(255,0,0,maxColorValue=255),rgb(255,69,0,maxColorValue=255),rgb(50,205,50,maxColorValue=255),rgb(0,128,0,maxColorValue=255),rgb(0,100,0,maxColorValue=255),rgb(194,225,5,maxColorValue=255),rgb(255,255,0,maxColorValue=255),rgb(102,205,170,maxColorValue=255),rgb(138,145,208,maxColorValue=255),rgb(205,92,92,maxColorValue=255),rgb(233,150,122,maxColorValue=255),rgb(189,183,107,maxColorValue=255),rgb(128,128,128,maxColorValue=255),rgb(192,192,192,maxColorValue=255),"grey90"),chromHMM_states)
chromHMM_composite = setNames(c("Active regulatory","Active regulatory","Active regulatory","Transcribed","Transcribed","Active regulatory","Active regulatory","Zinc-finger/repeats","Repressed","Poised regulatory","Poised regulatory","Poised regulatory","Repressed","Repressed","Quiescent"),chromHMM_states)

# WGBS
meth_states = c("Hypomethylated","Intermediate","Hypermethylated","Missing")
meth_labels = setNames(c("Hypomethylated","Intermediately methylated","Hypermethylated","Missing CpG data"),meth_states)
meth_colors = setNames(c("#F8766D","#00BFC4","#7CAE00","#C77CFF"),meth_states)

# All states
states = c(chromHMM_states,meth_states,"DNase","H3K27ac","Expressed_samples")
all_state_colors = c(chromHMM_colors,meth_colors,"aquamarine","tan1","black","gray28") 
names(all_state_colors)[20:23] = c("DNase","H3K27ac","Expressed_samples","Total")
all_state_labels = setNames(c(chromHMM_states,meth_labels,"DNase","H3K27ac","RPKM > 1"),states)
all_pc_states = c(states[1:21],unlist(lapply(states[1:21],function(x) paste(x,"PC",sep="_"))))

# TE classification
class_colors = setNames(c("#4A72E8","#FF6600","#006600","#cc0000","lightseagreen","#5C5C5C"),c("DNA","LINE","LTR","SINE","SVA","Other"))

# Sample classification
sample_categories = c("Group","Anatomy","Type","Age","Cancer","Germline")

group_colors = setNames(c("#924965","#4178AE","#E41A1C","#69608A","#B65C73","#FF9D0C","#678C69","#55A354","#E67326","#FFD924","#AF5B39","#D56F80","#999999","#C5912B","#C58DAA","#F182BC","#C2655D","#DAB92E","#000000"),c("ESC","ES-deriv","IMR90","iPSC","Mesench","Epithelial","HSC & B-cell","Blood & T-cell","Myosat","Neurosph","Adipose","Heart","Other","Brain","Digestive","Sm. Muscle","Muscle","Thymus","ENCODE2012"))
anatomy_colors = setNames(c("#B2DF8A","#33A02C","#FB9A99","#FFED6F","#80B1D3","#666666","#924965","#4178AE","#FDBF6F","#CAB2D6","#6A3D9A","#BEBADA","#FCCDE5","#BC80BD","#E7298A","#FF7F00","#69608A","#B3DE69","#CCEBC5","#000000","#FB8072","#FDB462","#8DD3C7","#D9D9D9","#E31A1C","#A6CEE3","#D95F02","#E6AB02","#66A61E","#B15928","#E31A1C","#8DD3C7","#E7298A","#A6761D","#666666"),c("ADRENAL","BLOOD","BONE","BRAIN","BREAST","CERVIX","ESC","ESC_DERIVED","FAT","GI_COLON","GI_DUODENUM","GI_ESOPHAGUS","GI_INTESTINE","GI_RECTUM","GI_STOMACH","HEART","IPSC","KIDNEY","LIVER","LUNG","MUSCLE","MUSCLE_LEG","OVARY","PANCREAS","PLACENTA","SKIN","SPLEEN","STROMAL_CONNECTIVE","THYMUS","VASCULAR",NA,NA,NA,NA,NA))
type_colors = setNames(brewer.pal(5,"Greens"),c("PrimaryCulture","PrimaryCell","PrimaryTissue","CellLine","ESCDerived"))
age_colors = setNames(brewer.pal(4,"YlOrRd"),c("Cell_line","Unknown","Non_fetal","Fetal"))
cancer_colors = setNames(c("lightgrey","red"),c("No","Yes"))
germline_colors = setNames(brewer.pal(6,"Dark2"),c("Pluripotent","Mesoderm","Ectoderm","Endoderm","Mixed","Other"))

# Genic features
features = c("cpgIsland","promoters","5UTR","CDS","3UTR","exons","introns","intergenic")
genic_labels = setNames(c("CpG island","Promoter","5'UTR","CDS","3'UTR","Exon","Intron","Intergenic"),features)

cohorts = c("cpgIsland","promoters","promoters_pc","promoters_nc","5UTR","5UTR_pc","5UTR_nc","coding_exon","coding_exon_pc","3UTR","3UTR_pc","3UTR_nc",
                     "exons","exons_pc","exons_nc","introns","introns_pc","introns_nc","intergenic","blacklist","Vista_enhancers")

# C-Gate subfamilies
cgate_subfams = c('AluJb','AluJo','AluS','AluSc','AluSg','AluSg1','AluSp','AluSq','AluSx','AluSz','AluY','FLAM_A','FRAM','HERV3-int','HERV9-int','HERVE-int','HERVH-int','L1HS','L1MB4','L1PA6','L2a','LTR10A','LTR16A','LTR22B','LTR31A','LTR5','LTR5_Hs','LTR9','MER11A','MER21A','MER34','MER39B','MER4B','MER54B','MER74C','Mir m','MIR3','MIRb','MLT1C','MLT2B2','MLT2B3','MSTB2','Plat_L3','RTVL-H','THE1A','THE1C','THE1D','AluJr','AluSq2','L1MB7','L1MC4','L1ME3A','L1PA11','L1PA4','LTR12','LTR12C','LTR12D','LTR13A','LTR2','LTR2B','MIR','MIRc','MLT1B')

# Colnames vectors
measure_states_extra = c("States.chromHMM","Max_states_intra","States.WGBS","Max_expression")
measure_metrics = c("Length","mappability","JC_distance","CpGs","CpGs_per_length")
measure_metrics_subfam = c("Count","Total_length","Length","Mappability","Age","CpGs","Count_CpGs","CpGs_per_TE","CpGs_per_kbp")
enrichment_names = c("Length_ijk","Length_ik","Length_jk","Length_k","Enrichment","Length_percent_jk","Members","Count","Percent")
```

## Load functions

```{r load functions}
source("R_scripts/functions.R")
```

## Create essential matrices

```{r create essential matrices, eval=FALSE}
source("R_scripts/metadata.R")
source("R_scripts/TE_matrix.R")
source("R_scripts/chromHMM_matrix.R")
source("R_scripts/WGBS_TE_avg_methylation.R")
source("R_scripts/DNase_peaks.R")
source("R_scripts/H3K27ac_peaks.R")
source("R_scripts/RNAseq.R")
source("R_scripts/promoter_matrices.R")
```

## Load metadata

```{r load metadata}
load("R_datasets/metadata.RData")
```

## Load sample statistics

```{r load sample stats}
# Samples with each metric: all, not excluded, no Y, both
sample_counts = rbind(apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x)),])[1]),
                      apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x) & metadata$Exclude == "Include"),])[1]),
                      apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x) & metadata$chrY == "Yes"),])[1]),
                      apply(metadata[,c("Sample","WGBS","DNase","H3K27ac","RNA")],2,function(x) dim(metadata[which(!is.na(x) & metadata$chrY == "Yes" & metadata$Exclude == "Include"),])[1]))
colnames(sample_counts)[1] = "chromHMM"
rownames(sample_counts) = c("All","Include","chrY","Include.chrY")
```

## Load essential matrices

```{r load essential matrices}
load("R_datasets/rmsk_TE.RData")
load("R_datasets/chromHMM_TE_state.RData")
load("R_datasets/TE_meth_average.RData")
load("R_datasets/TE_DNase_peaks.RData")
load("R_datasets/TE_H3K27ac_peaks.RData")
load("R_datasets/rna.RData")
```

## Define facet labels

```{r define facet labels}
measure_labels = c("% genome","% TEs (either strand)","% TEs (same strand)","% feature in TEs")
names(measure_labels) = c("Genome_percent_genome","TEs_percent_TE","TEs_strand_percent_TE","Genome_percent_TE")

mark_labels = c(paste("chromHMM (n=",sample_counts["All","chromHMM"],")",sep=""),
                paste("WGBS (n=",sample_counts["All","WGBS"],")",sep=""),
                paste("DNase (n=",sample_counts["All","DNase"],")",sep=""),
                paste("H3K27ac (n=",sample_counts["All","H3K27ac"],")",sep=""))
names(mark_labels) = c("chromHMM","WGBS","DNase","H3K27ac")

group_labels = c("chromHMM","WGBS","DNase","H3K27ac","RPKM > 1")
names(group_labels) = c("chromHMM","WGBS","DNase","H3K27ac","Expressed_samples")

coding_labels = c("All","Protein-coding","Non-coding")
names(coding_labels) = c("All","pc","nc")

metric_labels = c("Proportion of TEs in state in any sample","Average % of samples in state","Average % of samples in state if ever in state")
names(metric_labels) = c("Proportion_ever","Samples_avg_all","Samples_avg_ever")

variable_labels = c("Mappability","Jukes-Cantor evolutionary distance","Length (bp)","CpGs","CpGs/bp")
names(variable_labels) = c("mappability","JC_distance","Length","CpGs","CpGs_per_length")
```

## Load feature sizes

```{r load feature sizes}
# Genome
hg19_genome = read.table("raw_data/hg19.genome",sep='\t',header=TRUE)
GENOME_WIDTH = sum(as.numeric(hg19_genome[which(hg19_genome$chrom %in% standard_chromosomes),]$size))
GENOME_WIDTH_noY = sum(as.numeric(hg19_genome[which(hg19_genome$chrom %in% standard_chromosomes & hg19_genome$chrom != "chrY"),]$size))

# Total widths
CHROMHMM_TOTAL_WIDTH = sample_counts["chrY","chromHMM"]*GENOME_WIDTH + (sample_counts["All","chromHMM"]-sample_counts["chrY","chromHMM"])*GENOME_WIDTH_noY
DNASE_TOTAL_WIDTH = sample_counts["chrY","DNase"]*GENOME_WIDTH + (sample_counts["All","DNase"]-sample_counts["chrY","DNase"])*GENOME_WIDTH_noY
H3K27AC_TOTAL_WIDTH = sample_counts["chrY","H3K27ac"]*GENOME_WIDTH + (sample_counts["All","H3K27ac"]-sample_counts["chrY","H3K27ac"])*GENOME_WIDTH_noY

# TEs 
NUM_TE = dim(rmsk_TE)[1]
NUM_TE_noY = dim(rmsk_TE[which(rmsk_TE$chromosome != "chrY"),])[1]
NUM_TE_WGBS = dim(TE_meth_average)[1]
  
# awk '{sum+=$3-$2}END{print sum}' features/TEs/merge/rmsk_TEother_merge.txt
MERGED_TE_WIDTH = 1389947349
# awk '{if($1 != "chrY") sum+=$3-$2}END{print sum}' features/TEs/merge/rmsk_TEother_merge.txt
MERGED_TE_WIDTH_noY = 1375898142

# Total widths
CHROMHMM_TE_WIDTH = sample_counts["chrY","chromHMM"]*MERGED_TE_WIDTH + (sample_counts["All","chromHMM"]-sample_counts["chrY","chromHMM"])*MERGED_TE_WIDTH_noY
DNASE_TE_WIDTH = sample_counts["chrY","DNase"]*MERGED_TE_WIDTH + (sample_counts["All","DNase"]-sample_counts["chrY","DNase"])*MERGED_TE_WIDTH_noY
H3K27AC_TE_WIDTH = sample_counts["chrY","H3K27ac"]*MERGED_TE_WIDTH + (sample_counts["All","H3K27ac"]-sample_counts["chrY","H3K27ac"])*MERGED_TE_WIDTH_noY

# CpGs
ALL_CPGS = 56434896/2
TE_CPGS = 28373958/2

# Feature lengths
feature_overlap = as.data.frame(matrix(readLines("features/intersect_features/feature_overlap.txt"),ncol=5,byrow=TRUE))
colnames(feature_overlap) = c("Cohort","Genome","Genome_noY","TEs","TEs_stranded")
feature_overlap$Cohort = gsub("refseq_","",feature_overlap$Cohort)
feature_overlap = split_coding(feature_overlap)
feature_overlap[,2:5] = apply(feature_overlap[,2:5],2,function(x) as.numeric(x))
feature_overlap = feature_overlap[,c(1,6:7,2:5)]
```

## Calculate TE class and subfamily stats

```{r calculate TE stats}
source("R_scripts/TE_class_stats.R")
source("R_scripts/TE_subfamily_stats.R")
```

# Introduction

## Basic sample stats

```{r basic sample stats}
sample_counts
```

## Basic TE statistics

```{r basic TE stats}
# Number of TEs
NUM_TE

# % of genome
MERGED_TE_WIDTH/GENOME_WIDTH

# % CpGs overlapping TEs
TE_CPGS/ALL_CPGS

# Number of families
length(unique(rmsk_TE$family))

# Number of subfamilies
length(unique(rmsk_TE$subfamily))

# Length of TEs
ddply(rmsk_TE,.(),summarise,Min=min(Length),Median=median(Length),Max=max(Length))

# % TEs overlapping CpGs
NUM_TE_WGBS
NUM_TE_WGBS/NUM_TE
NUM_TE-NUM_TE_WGBS
ddply(TE_meth_average,.(),summarise,Min=min(CpGs),Median=median(CpGs),Max=max(CpGs))
```

# Results

## Contribution

### Figure 1

```{r Figure 1, echo=FALSE}
source("R_scripts/proportion.R")
source("R_scripts/proportion_class.R") 

a = ggplot(combined_proportion[which(combined_proportion$Coding == "All"),],aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion of feature in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x=element_blank()) + scale_x_discrete(labels=c("TE","Genome",genic_labels)) + facet_wrap(~Mark,nrow=1,labeller=labeller(Mark = mark_labels))

b = ggplot(combine_class_proportion,aes(x=class,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion of class in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x = element_blank()) + facet_wrap(~Cohort,nrow=1)

c = ggplot(contribution,aes(x=reorder(State,-Proportion,max),y=Proportion,fill=State)) + geom_bar(stat="identity") + coord_flip() + scale_y_reverse() + ylab("Proportion of state in TEs") + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.ticks.x=element_blank()) + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(labels=all_state_labels) + geom_text(aes(label=round(100*Proportion,1)),position=position_stack(reverse=TRUE,vjust=0.1))

d = ggplot(contribution_class,aes(x=State,y=Proportion,fill=class)) + geom_bar(stat="identity") + coord_flip() + ylab("Proportion in each TE class") + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.text.y=element_blank()) + scale_x_discrete(limits=as.vector(contribution[order(contribution$Proportion,decreasing = TRUE),]$State)) + scale_fill_manual(values=class_colors,guide=FALSE)

legend_class = get_legend(ggplot(contribution_class,aes(x=State,y=Proportion,fill=class)) + geom_bar(stat="identity") + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.ticks.x=element_blank(),axis.text.y=element_blank(),legend.position="bottom",legend.direction = "horizontal") + scale_fill_manual(values=class_colors,name="Class"))

grid.arrange(a,b,c,d,legend_class, nrow = 4,layout_matrix = rbind(c(1,1),c(2,2),c(3,4),c(6,6)),heights=c(0.3,0.3,0.35,0.05),widths=c(0.6,0.4))
```

### Proportion/contribution analysis

```{r analysis Fig 1}
# Proportion of each feature in each state, across all samples
combined_proportion[which(combined_proportion$Coding == "All"),]

# Specific states, TEs, promoters, and exons
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "TE" & combined_proportion$State %in% chromHMM_states[c(1:3,6:7)]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "TE" & combined_proportion$State %in% chromHMM_states[4:5]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "promoters" & combined_proportion$State %in% chromHMM_states[c(1:3,6:7)]),]$Proportion)
sum(combined_proportion[which(combined_proportion$Coding == "All" & combined_proportion$Feature == "exons" & combined_proportion$State %in% chromHMM_states[4:5]),]$Proportion)

# Does any feature have less % in state than TEs?
ddply(combined_proportion[which(combined_proportion$Coding == "All"),],.(State),function(x) x[which(x$Proportion > x[which(x$Feature == "TE"),]$Proportion),c("State","Cohort","Feature","Coding","Proportion")])

# By-class proportion
dcast(combine_class_proportion,class~State,value.var="Proportion")
test = ddply(combine_class_proportion,.(State),function(x) x[which.max(x$Proportion),])
test[order(test$class),]

# Proportion of all genomic bases in TEs
CHROMHMM_TE_WIDTH/CHROMHMM_TOTAL_WIDTH

# Contribution
contribution

# Contribution, composite states
contribution_composite
```

### VISTA enhancers

```{r vista enhancers}
# wc -l features/vista_enhancers/vista_enhancers_human.bed
# cat features/intersect_features/vista_enhancers_rmsk_*.txt | awk '{print $1, $2, $3}' - | sort | uniq | wc -l
355

355/934
```

### Supplementary Figure 1

```{r Figure S1, echo=FALSE}
source("R_scripts/histone_modification.R")

## Average proportion by sample
a = ggplot(ddply(all_state_proportion[which(all_state_proportion$Coding == "All"),],.(Mark,State,Feature),summarise,Proportion=mean(Proportion)),aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Average proportion of feature in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x=element_blank()) + scale_x_discrete(labels=c("TE","Genome",genic_labels)) + facet_wrap(~Mark,nrow=1,labeller=labeller(Mark = mark_labels))

# Average mappability by chromHMM states
test = lapply(metadata$Sample[c(1:12,32:40,119:127)],function(x) read.table(file=paste("mappability/by_state/mappability_",x,"_average.txt",sep=""),sep='\t'))
names(test) = metadata$Sample[c(1:12,32:40,119:127)]
test = ldply(test)

b = ggplot(test,aes(x=V1,y=V2,fill=V1)) + geom_boxplot() + scale_x_discrete(limits=chromHMM_states) + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + ylab("Average mappability") + scale_fill_manual(values=chromHMM_colors,guide=FALSE)

c = ggplot(histones,aes(x=Bin,y=Level,color=Mark,group=Mark))+geom_point(size=0.5)+geom_line(size=1.5)+theme(axis.text.x = element_blank(),axis.ticks.x=element_blank(),panel.grid=element_blank(),legend.position = "bottom",legend.title = element_blank())+ylab("Fold enrichment over input") + scale_color_manual(labels=c(levels(histones$Mark)[1:8],"DNA methylation"),values=c(gg_color_hue(8),"black")) + xlab("Genomic position (+/-5kb around TE)")+ylim(0,12) + facet_wrap(~ State,nrow=3)

grid.arrange(a,b,c,nrow=3,heights=c(0.3,0.2,0.5))
```

### Mappability of genome/TE bases

```{r mappability}
# Mappability whole genome, TEs (8/27/17)
# (awk '{sum+=($4*($3-$2));total+=$3-$2;}END{print sum/total}' wgEncodeCrgMapabilityAlign36mer.bedGraph)
GENOME_MAP=0.789192
GENOME_MAP

# (for file in x*; do bedtools intersect -wo -a rmsk_TEother_merge.txt -b $file >> rmsk_TEother_merge_map.txt; done)
# (awk '{sum+=($7*$8);total+=$8;}END{print sum/total}' rmsk_TEother_merge_map.txt)
TE_MAP=0.624207
TE_MAP

# Mappability vs. length by class
cor.test(rmsk_TE$Length,rmsk_TE$mappability)
by(rmsk_TE,rmsk_TE$class_update,function(x) cor.test(x$Length,x$mappability))
```

### Supplementary Figure 2

```{r Figure S2, echo=FALSE}
source("R_scripts/feature_overlap.R")

par(mfrow=c(1,3)) 

pie(rmsk_TE_class$Count,labels=paste(rmsk_TE_class$Count,"(",round(rmsk_TE_class$Count/sum(rmsk_TE_class$Count)*100,0),"%) ",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="TEs",cex=1,cex.main=2,line=-20,radius=0.5)

pie(rmsk_TE_class$Subfamilies,labels=paste(rmsk_TE_class$Subfamilies,"(",round(rmsk_TE_class$Subfamilies/sum(rmsk_TE_class$Subfamilies)*100,0),"%) ",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="Subfamilies",cex=1,cex.main=2,line=-20,radius=0.5)
 
pie(c(GENOME_WIDTH-MERGED_TE_WIDTH,rmsk_TE_class$Total_length),labels=paste(round(c(GENOME_WIDTH-MERGED_TE_WIDTH,rmsk_TE_class$Total_length)/GENOME_WIDTH*100,0),"%",sep=""),col=c("lightblue",class_colors[as.vector(rmsk_TE_class$class_update)]),clockwise=TRUE,cex=1,cex.main=2,main="Length (bp)",line=-20,radius=0.5)

a = ggplot(feature_overlap_long,aes(x=Feature,y=Percent,fill=Coding)) + geom_bar(position="dodge",stat="identity") + theme(axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1),legend.title = element_blank())  + scale_x_discrete(labels=genic_labels) + scale_fill_discrete(labels=coding_labels) + facet_wrap(~Measure,nrow=1,labeller=labeller(Measure=measure_labels))

b = ggplot(combined_proportion,aes(x=Feature,y=Proportion,fill=State)) + geom_bar(stat="identity") + ylab("Proportion of feature in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x=element_blank()) + scale_x_discrete(labels=c("TE","Genome",genic_labels[1:7])) + facet_grid(Coding~Mark,labeller=labeller(Coding=coding_labels))

grid.arrange(a,b,nrow=2,heights=c(0.3,0.7))
```

### RefSeq feature analysis

```{r analysis Fig S2}
# Feature %
feature_overlap_long

# Proportion of each feature in each state, across all samples
combined_proportion
```

## Potential

### Figure 2

```{r Figure 2, echo=FALSE, cache=TRUE}
source("R_scripts/potential.R")
source("R_scripts/potential_class.R")

a = ggplot(combine_boxplot,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="Proportion of TEs in state") + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + geom_point(data=combine_stats,aes(x=State,y=Proportion_ever),color="red",size=3)

b = ggplot(combine_class_ever,aes(x=State,y=Proportion,fill=Class)) + geom_bar(stat="identity") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + labs(y="Proportion of TEs ever in state by class") + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels))

c = ggplot(combine_potential[which(combine_potential$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + geom_boxplot(width=0.2,outlier.size = 0.1)

d = ggplot(combine_potential[which(combine_potential$Sample.Proportion == 1),],aes(x=State,y=log10(Count+1),fill=State)) + geom_bar(stat="identity") + geom_text(aes(label=Count),position=position_stack(reverse=TRUE,vjust=-0.1)) + scale_y_reverse() + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + ylab("TEs in state in 100% of samples (log10)") 

e = ggplot(ddply(combine_potential[which(combine_potential$Sample.Proportion >= 0.9),],.(State),summarise,Count=sum(Count)),aes(x=State,y=log10(Count+1),fill=State)) + geom_bar(stat="identity") + geom_text(aes(label=Count),position=position_stack(vjust=1.1)) + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("TEs in state in 90% of samples (log10)") 

f = ggplot(ddply(combine_potential_class[which(combine_potential_class$Sample.Proportion >= 0.9),],.(State,Class),summarise,TEs.Proportion=sum(Count)/sum(Total)),aes(x=State,y=TEs.Proportion,fill=Class)) + geom_bar(stat="identity") + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + labs(y="Proportion in each class (90%)") + coord_flip() + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels))

grid.arrange(a,b,c,d,e,f, nrow = 2, layout_matrix=rbind(c(1,2,3),c(4,5,6)),widths=c(0.45,0.275,0.275))
```

### Potential analysis

```{r analysis Fig 2a}
aggregate(data=combine_boxplot,Proportion~State,mean)

# Number of TEs ever in composite states
## Roadmap Active (states 1-8)
sum(apply(chromHMM_TE_state[8:15],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Roadmap Inactive (states 9-15)
sum(apply(chromHMM_TE_state[16:22],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Active regulatory states (states 1-3, 6-7)
sum(apply(chromHMM_TE_state[c(8:10,13:14)],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Transcribed states (states 4-5)
sum(apply(chromHMM_TE_state[11:12],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Poised regulatory states (states 10-12)
sum(apply(chromHMM_TE_state[17:19],1,function(x) sum(x > 0)) > 0)/NUM_TE
## Repressed states (states 9, 13-14)
sum(apply(chromHMM_TE_state[c(16,20,21)],1,function(x) sum(x > 0)) > 0)/NUM_TE

# Number of TEs ever in each state
ddply(combine_potential,.(State),summarise,Count=sum(Count[which(Samples != 0)]))
combine_stats
ddply(rmsk_TE_class,.(class_update,Count),summarise,TE_prop=Count/NUM_TE)
dcast(combine_class_ever,Class~State,value.var="Proportion")
combine_potential_class[which(combine_potential_class$Sample.Proportion == 1),]

# Median samples in state (ever)
ddply(combine_potential[which(combine_potential$Samples != 0),],.(State),summarise,first=quantile(rep(Sample.Proportion,times=Count),0.25),Median=median(rep(Sample.Proportion,times=Count)),third=quantile(rep(Sample.Proportion,times=Count),0.75))
dcast(ddply(combine_potential_class[which(combine_potential_class$Samples != 0),],.(State,Class),summarise,Median=median(rep(Sample.Proportion,times=Count))),Class~State,value.var="Median")
```

### TEs in each state in x% of samples

```{r potential percent}
ddply(combine_potential,.(State),summarise,In_100=sum(Count[which(Sample.Proportion == 1)]),
      In_90=sum(Count[which(Sample.Proportion >= 0.9)]),
      In_75=sum(Count[which(Sample.Proportion >= 0.75)]),
      In_50=sum(Count[which(Sample.Proportion >= 0.5)]),
      In_25=sum(Count[which(Sample.Proportion >= 0.25)]))
```

### TEs always in single state

```{r analysis Fig 2c}
always_TEs = c(setNames(lapply(chromHMM_states,function(x) chromHMM_TE_state[which(chromHMM_TE_state[[x]] == sample_counts["All","chromHMM"]),TE_coordinates]),chromHMM_states),
                  setNames(lapply(meth_states,function(x) TE_meth_average[which(TE_meth_average[[x]] == sample_counts["All","WGBS"]),TE_coordinates]),meth_states))
always_TEs$DNase = TE_DNase_peaks[which(TE_DNase_peaks$Samples == sample_counts["All","DNase"]),TE_coordinates]
always_TEs$H3K27ac = TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$Samples == sample_counts["All","H3K27ac"]),TE_coordinates]
always_TEs$RNA = RNA_TE[which(RNA_TE$Expressed_samples == sample_counts["All","RNA"]),TE_coordinates]
always_TEs = ldply(always_TEs)
colnames(always_TEs)[1] = "State"

always_TEs_noY = setNames(lapply(chromHMM_states,function(x) chromHMM_TE_state[which(chromHMM_TE_state$chromosome == "chrY" & chromHMM_TE_state[[x]] == sample_counts["chrY","chromHMM"]),TE_coordinates]),chromHMM_states)
always_TEs_noY$DNase = TE_DNase_peaks[which(TE_DNase_peaks$chromosome == "chrY" & TE_DNase_peaks$Samples == sample_counts["chrY","DNase"]),TE_coordinates]
always_TEs_noY$H3K27ac = TE_H3K27ac_peaks[which(TE_H3K27ac_peaks$chromosome == "chrY" & TE_H3K27ac_peaks$Samples == sample_counts["chrY","H3K27ac"]),TE_coordinates]
always_TEs_noY$RNA = RNA_TE[which(RNA_TE$chromosome == "chrY" & RNA_TE$Expressed_samples == sample_counts["chrY","RNA"]),TE_coordinates]
always_TEs_noY = ldply(always_TEs_noY)
colnames(always_TEs_noY)[1] = "State"

## Summarise
ddply(always_TEs,.(State),summarise,Count=length(chromosome),Proportion=length(chromosome)/NUM_TE,Proportion_CpG=length(chromosome)/NUM_TE_WGBS)
ddply(always_TEs_noY,.(State),summarise,Count=length(chromosome),Proportion=length(chromosome)/NUM_TE,Proportion_CpG=length(chromosome)/NUM_TE_WGBS)

## Genomic locations
always_TEs_location = merge(always_TEs,rmsk_TE[,c(TE_coordinates,"class_update","Length",cohorts)],by=TE_coordinates,all.x=TRUE)
ddply(always_TEs_location,.(State),function(x) apply(x[,cohorts],2,function(y) sum(!is.na(y))))

## Phylogeny
ddply(always_TEs_location,.(State,class_update),summarise,Count=length(chromosome))

always_TE_class = ddply(always_TEs_location[,c(TE_coordinates,"class_update","State")],.(State,class_update),summarise,Count=length(chromosome))
always_TE_class = merge(rmsk_TE_class[,c("class_update","Count")],always_TE_class,by="class_update",all.y=TRUE)
always_TE_class[is.na(always_TE_class)] = 0
colnames(always_TE_class)[c(2,4)] = c("Count.Class","Count")
always_TE_class = ddply(always_TE_class,.(State),transform,Count.State=sum(Count))
always_TE_class$Enrichment = log2((always_TE_class$Count/always_TE_class$Count.State)/(always_TE_class$Count.Class/NUM_TE))
always_TE_class = always_TE_class[order(always_TE_class$Enrichment),]
dim(always_TE_class[which(always_TE_class$Enrichment > 0),])
always_TE_class[which(always_TE_class$Enrichment > 5),]

always_TE_family = ddply(always_TEs_location[,c(TE_coordinates,"class_update","State")],.(State,class_update,family),summarise,Count=length(chromosome))
always_TE_family = merge(ddply(rmsk_TE_subfamily,.(family,class_update),summarise,Count=sum(Count)),always_TE_family,by=c("family","class_update"),all.y=TRUE)
always_TE_family[is.na(always_TE_family)] = 0
colnames(always_TE_family)[c(3,5)] = c("Count.Family","Count")
always_TE_family = ddply(always_TE_family,.(State),transform,Count.State=sum(Count))
always_TE_family$Enrichment = log2((always_TE_family$Count/always_TE_family$Count.State)/(always_TE_family$Count.Family/NUM_TE))
always_TE_family = always_TE_family[order(always_TE_family$Enrichment),]
dim(always_TE_family[which(always_TE_family$Enrichment > 0),])
always_TE_family[which(always_TE_family$Enrichment > 0.5 & always_TE_family$Count > 10),]

always_TE_subfamily = ddply(always_TEs_location[,c(TE_coordinates,"class_update","State")],.(State,class_update,family,subfamily),summarise,Count=length(chromosome))
always_TE_subfamily = merge(rmsk_TE_subfamily[,c("subfamily","family","class_update","Count")],always_TE_subfamily,by=c("subfamily","family","class_update"),all.y=TRUE)
always_TE_subfamily[is.na(always_TE_subfamily)] = 0
colnames(always_TE_subfamily)[c(4,6)] = c("Count.Subfamily","Count")
always_TE_subfamily = ddply(always_TE_subfamily,.(State),transform,Count.State=sum(Count))
always_TE_subfamily$Enrichment = log2((always_TE_subfamily$Count/always_TE_subfamily$Count.State)/(always_TE_subfamily$Count.Subfamily/NUM_TE))
dim(always_TE_subfamily[which(always_TE_subfamily$Enrichment > 0),])
always_TE_subfamily = always_TE_subfamily[order(always_TE_subfamily$State,always_TE_subfamily$Enrichment),]
always_TE_subfamily[which(always_TE_subfamily$Enrichment > 0.5 & always_TE_subfamily$Count > 10),]

## Print out TEs
write.table(always_TEs[,c(TE_coordinates,"State")],file="always_TEs.bed",sep='\t',row.names=FALSE,quote=FALSE,col.names = FALSE)
```

### Supplementary Figure 3

```{r Figure S3, echo=FALSE}
#source("R_scripts/potential_class.R")

a = ggplot(combine_boxplot_class,aes(x=State,y=Proportion,fill=Class)) + geom_boxplot() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.title.y = element_blank()) + coord_flip() + labs(y="Proportion of TEs in state by sample") + scale_x_discrete(labels=rev(all_state_labels),limits=rev(levels(combine_boxplot_class$State))) + geom_point(data=combine_stats_class,aes(x=State,y=Proportion_ever,color=Class),position=position_dodge(width=0.5)) + scale_color_manual(values=class_colors,guide=FALSE)

state_order = ddply(combine_stats_class,"State",summarise,CV=sd(Samples_avg_all)/mean(Samples_avg_all), Range=max(Samples_avg_all)-min(Samples_avg_all))

b = ggplot(combine_stats_class,aes(x=State,y=Samples_avg_all)) + geom_point(aes(color=Class),size=3) + scale_color_manual(values=class_colors,guide=FALSE) + scale_x_discrete(limits=state_order[order(state_order$Range),]$State,labels=all_state_labels) + scale_y_continuous(limits=c(0,1)) + theme(axis.title.y = element_blank()) + ylab("Average proportion of samples in state") + geom_text(data=state_order,aes(x=State,y=1,label=paste("(",round(CV,2),")",sep=""))) + coord_flip() 

c = ggplot(combine_potential_class[which(combine_potential_class$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + facet_wrap(~Class,nrow=1) + geom_boxplot(width=0.2,outlier.size = 0.1)

rmsk_TE_class_feature = melt(rmsk_TE_class[,c("class_update",cohorts[1:19])],id.vars="class_update")
colnames(rmsk_TE_class_feature) = c("Class","Cohort","Proportion")
rmsk_TE_class_feature = split_coding(rmsk_TE_class_feature)

d = ggplot(rmsk_TE_class_feature,aes(x=Class,y=Proportion,fill=Feature)) + geom_bar(stat="identity") + theme(axis.title.x = element_blank(),legend.title=element_blank()) +  ylab("Proportion of TEs overlapping feature") + scale_fill_manual(values=c(brewer.pal(8,"Spectral")),labels=genic_labels[c(8,1:7)]) + facet_wrap(~Coding,nrow=1,labeller=labeller(Coding=coding_labels))

grid.arrange(a,b,c,d,nrow=3,layout_matrix=rbind(c(1,2),c(3,3),c(4,4)),heights=c(0.35,0.4,0.25))
```

### Class potential analysis

```{r class potential analysis}
combine_stats_class
state_order
```

### Feature overlap and CpGs per class

```{r analysis Fig S9a,eval=FALSE}
# Any TEs not in a feature? Number of features each TE overlaps
table(apply(rmsk_TE[,cohorts[2:19]],1,function(x) sum(!is.na(x))))
table(apply(rmsk_TE[,c(features[2:3],"coding_exon",features[5:8])],1,function(x) sum(!is.na(x))))

# Number/proportion of TEs in each feature, overall and by class
feature_proportion = apply(rmsk_TE[,cohorts[1:19]],2,function(x) length(na.omit(x)))
feature_proportion/NUM_TE
rmsk_TE_class[,c("class_update",cohorts[1:20])]

rmsk_TE_class[,c("class_update","Count","cpgIsland","CpGs","Percent_TE_CpGs","Percent_all_CpGs","TEs_wCpG","TEs_wCpG_per","Mean_CpG","Mean_CpG_wCpG")]
```

### Supplementary Figure 4

```{r Figure S4, echo=FALSE, cache=TRUE}
source("R_scripts/potential_promoter.R")
source("R_scripts/potential_exon.R")

# No cancer/IMR90
a = ggplot(combine_boxplot_noCancer_IMR90,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="Proportion of TEs in state") + coord_flip() + scale_x_discrete(limits = rev(levels(combine_boxplot_noCancer_IMR90$State)),labels=rev(all_state_labels)) + geom_point(data=combine_stats_noCancer_IMR90,aes(x=State,y=Proportion_ever),color="red",size=2)

b = ggplot(combine_potential_noCancer[which(combine_potential_noCancer$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors,guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states),labels=rev(all_state_labels)) + geom_boxplot(width=0.2,outlier.size = 0.1)

# Promoters
c = ggplot(combine_boxplot_prom,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values=all_state_colors[1:21],guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="Proportion of promoters in state") + coord_flip() + scale_x_discrete(limits=rev(states[1:21]),labels=all_state_labels[1:21]) + geom_point(data=combine_stats_prom,aes(x=State,y=Proportion_ever),color="red",size=2)

d = ggplot(combine_potential_prom[which(combine_potential_prom$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values=all_state_colors[1:21],guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples in state") + scale_x_discrete(limits = rev(states[1:21]),labels=all_state_labels[1:21]) + geom_boxplot(width=0.2,outlier.size = 0.1)

# Exons
e = ggplot(RNA_exon_sample,aes(x=State,y=Proportion,fill=State)) + geom_boxplot() +scale_fill_manual(values="black",guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank()) + labs(y="Proportion of exons in state") + coord_flip() + scale_x_discrete(labels="RPKM > 1") + geom_point(data=RNA_potential_exon_stats,aes(x=State,y=Proportion_ever),color="red",size=3) + scale_y_continuous(limits=c(0,1))

f = ggplot(RNA_potential_exon_long[which(RNA_potential_exon_long$Samples != 0),],aes(x=State,y=Sample.Proportion,weight=Count,fill=State)) + geom_violin(scale="width") + coord_flip() + scale_fill_manual(values="black",guide=FALSE) + theme(axis.title.y=element_blank(),axis.ticks.y=element_blank(),axis.text.y = element_blank()) + ylab("Proportion of samples RPKM > 1") + scale_x_discrete(labels=all_state_labels[22]) + geom_boxplot(width=0.2,outlier.size = 0.1)

grid.arrange(a,b,c,d,e,f,nrow = 3,layout_matrix=rbind(c(1,2),c(3,4),c(5,6)),heights=c(0.45,0.45,0.1),widths=c(0.6,0.4))
```

### Promoter statistics

```{r promoter analysis}
# Number of unique promoters
NUM_PROMOTER

combine_stats_prom
ddply(combine_boxplot_prom,.(State),summarise,Proportion=mean(Proportion))

# Promoter CpG stats
promoter_CpG_count = read.table("WGBS/Refseq_promoters/refseq_promoter_unique_CpG_count.txt",sep='\t')
promoter_CpG_count$V5 = promoter_CpG_count$V5/2
dim(promoter_CpG_count)[1] # Number of promoters with CpGs
dim(promoter_CpG_count)[1]/NUM_PROMOTER # % of promoters with CpGs
ddply(promoter_CpG_count,.(),summarise,Min=min(V5),Median=median(V5),Mean=mean(V5),Max=max(V5))
mean(c(promoter_CpG_count$V5,rep(0,(NUM_PROMOTER-dim(promoter_CpG_count)[1]))))
median(c(promoter_CpG_count$V5,rep(0,(NUM_PROMOTER-dim(promoter_CpG_count)[1]))))
```

### Exon statistics (expression potential control)

```{r exon control analysis, cache=TRUE}
# Number of unique exons
NUM_EXON

RNA_potential_exon_stats
mean(RNA_exon_sample$Proportion)

# Exon CpG stats
exon_CpG_count = read.table("WGBS/refseq_exons_unique_CpG_count.txt",sep='\t')
exon_CpG_count$V5 = exon_CpG_count$V5/2
dim(exon_CpG_count)[1] # Number of exons with CpGs
dim(exon_CpG_count)[1]/NUM_EXON # % of exons with CpGs
ddply(exon_CpG_count,.(),summarise,Min=min(V5),Median=median(V5),Mean=mean(V5),Max=max(V5))
mean(c(exon_CpG_count$V5,rep(0,(NUM_EXON-dim(exon_CpG_count)[1]))))
median(c(exon_CpG_count$V5,rep(0,(NUM_EXON-dim(exon_CpG_count)[1]))))

# Number of samples expressed
table(RNA_TE$Expressed_samples)
table(RNA_TE$Expressed_samples)/dim(RNA_TE)[1]

table(RNA_refseq_exon$Expressed_samples)
table(RNA_refseq_exon$Expressed_samples)/NUM_EXON

# Maximum expression
median(RNA_TE$Max_expression)
median(RNA_TE[which(RNA_TE$Expressed_samples > 0),]$Max_expression)
max(RNA_TE$Max_expression)
tail(RNA_TE[order(RNA_TE$Max_expression),],n=20)

median(RNA_refseq_exon$Max_expression)
median(RNA_refseq_exon[which(RNA_refseq_exon$Expressed_samples > 0),]$Max_expression)
max(RNA_refseq_exon$Max_expression)
tail(RNA_refseq_exon[order(RNA_refseq_exon$Max_expression),],n=20)

cor.test(RNA_TE$Max_expression,RNA_TE$Expressed_samples,method="spearman")

# Max expression by class
merge(ddply(RNA_TE,.(class_update),summarise,Proportion_expressed=sum(Expressed_samples > 0)/length(Expressed_samples),Max_median=median(Max_expression)),
      ddply(RNA_TE[which(RNA_TE$Expressed_samples > 0),],.(class_update),summarise,Max_median_expressed=median(Max_expression)),by="class_update")

# Average TE vs. exon expression
TE_express = unlist(RNA_TE[,8:63])
median(TE_express)
median(TE_express[which(TE_express > 1)])
rm(TE_express)
exon_express = unlist(RNA_refseq_exon[,5:60])
median(exon_express)
median(exon_express[which(exon_express > 1)])
rm(exon_express)
```

### Cancer/IMR90 outlier analysis

```{r analysis Fig S5a}
combine_stats_noCancer_IMR90

# Average proportion of TEs in each state per sample, no IMR90
ddply(combine_boxplot,.(State),summarise,Mean=mean(Proportion),SD=sd(Proportion),Outlier=mean(Proportion)+2*sd(Proportion))
ddply(combine_boxplot[which(combine_boxplot$Sample != "E017"),],.(State),summarise,Mean=mean(Proportion),SD=sd(Proportion))

# Proportion of TEs in each state, IMR90
combine_boxplot[which(combine_boxplot$Sample == "E017"),]
```

### Figure 3

```{r Figure 3, echo=FALSE}
source("R_scripts/state_switching.R")

# chromHMM states
a = ggplot(chromHMM_TE_state,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + geom_histogram(data = chromHMM_TE_state,aes(x=Max_states_intra),binwidth=1,fill="darkred",color="grey",alpha=0.5) + ylab("Number of TEs") + xlab("Number of chromHMM states")

# chromHMM states by class 
b = ggplot(chromHMM_TE_state,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + scale_x_continuous(limits=c(0,15)) + geom_histogram(data = chromHMM_TE_state,aes(x=Max_states_intra),binwidth=1,fill="darkred",color="grey",alpha=0.5) + theme(axis.title.y = element_blank()) + xlab("Number of chromHMM states") + facet_wrap(~class_update,scales="free")

# State switching inter, chromHMM
c = ggplot(melt(as.matrix(state_switching_inter)),aes(x=Var2,y=Var1,fill=value/sample_counts["All","chromHMM"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(limits=rev(chromHMM_states)) + scale_fill_gradient(low="white",high="darkred",limits=c(0,1),guide=FALSE)

# State switching inter, WGBS
d = ggplot(melt(as.matrix(ss_inter_meth)),aes(x=Var2,y=Var1,fill=value/sample_counts["All","WGBS"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(meth_states)) + scale_fill_gradient2(low="white",high="darkred",limits=c(0,1),guide=FALSE)

legend = get_legend(ggplot(melt(as.matrix(ss_inter_meth)),aes(x=Var2,y=Var1,fill=value/sample_counts["All","WGBS"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank(),legend.position = "bottom",legend.direction = "horizontal") + scale_fill_gradient2(low="white",high="darkred",limits=c(0,1)))

grid.arrange(a,b,c,d,legend,nrow = 3, layout_matrix=rbind(c(1,2),c(3,4),c(5,5)),heights=c(0.3,0.6,0.1))
```

### Supplementary Figure 5

```{r Figure S5, echo=FALSE}
# WGBS states
a = ggplot(TE_meth_average,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + ylab("Number of TEs") + xlab("Number of methylation states")

# WGBS states by class
b = ggplot(TE_meth_average,aes(x=States)) + geom_histogram(binwidth=1,fill="blue",color="grey",alpha=0.5) + scale_x_continuous(limits=c(0,4)) + ylab("Number of TEs") + xlab("Number of methylation states") + facet_wrap(~class_update,scales="free_y")

# State switching intra, chromHMM
c = ggplot(melt(as.matrix(state_switching_intra)),aes(x=Var2,y=Var1,fill=value)) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(labels=rev(chromHMM_states),limits=rev(levels(melt(as.matrix(state_switching_intra))$Var1))) + scale_fill_gradient(low="white",high="darkred",limits=c(0,1))

# State switching inter, chromHMM, by class 
d = ggplot(melt(ss_inter_class,id.vars=c("Class","State")),aes(x=variable,y=State,fill=value/sample_counts["All","chromHMM"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_x_discrete(labels=chromHMM_states) + scale_y_discrete(limits=rev(chromHMM_states)) + scale_fill_gradient(low="white",high="darkred",guide=FALSE) + facet_wrap(~Class)

# State switching inter, WGBS, by class
e = ggplot(melt(ss_inter_meth_class,id.vars=c("Class","State")),aes(x=variable,y=State,fill=value/sample_counts["All","WGBS"])) + geom_tile() + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),legend.title = element_blank()) + scale_y_discrete(limits=rev(meth_states)) + scale_fill_gradient2(low="white",high="darkred",limits=c(0,1),guide=FALSE) + facet_wrap(~Class)

grid.arrange(a,b,c,d,e, nrow=3, layout_matrix=rbind(c(1,2,3),c(4,4,4),c(5,5,5)),heights=c(0.2,0.4,0.4),widths=c(0.3,0.35,0.35))
```

### State switching analysis

```{r state switching analysis}
state_switching_intra
state_switching_inter

dim(chromHMM_TE_state[which(chromHMM_TE_state$`10_TssBiv` > 0 | chromHMM_TE_state$`11_BivFlnk` > 0),])
```

### Number of states per TE, overall and by class, chromHMM and WGBS

```{r analysis Fig S6a}
table(chromHMM_TE_state$States)
ddply(chromHMM_TE_state,.(),summarise,Mean=mean(States),SD=sd(States))
ddply(chromHMM_TE_state,.(class_update),summarise,Mean=mean(States),SD=sd(States))

states_outlier = chromHMM_TE_state[which(chromHMM_TE_state$States == 15),]
ddply(states_outlier,.(),summarise,Length=median(stop-start),States_intra=mean(Tissues)/127,Max_states_intra=mean(Max_states_intra))
states_outlier = merge(states_outlier,rmsk_TE[,c(TE_coordinates,cohorts)],by=TE_coordinates,all.x=TRUE)
states_outlier
write.table(states_outlier[,c(TE_coordinates,"class_update",cohorts)],file="States_15_TEs.bed",sep='\t',row.names=FALSE,quote=FALSE)

table(TE_meth_average$States)
ddply(TE_meth_average,.(),summarise,Mean=mean(States),SD=sd(States))
ddply(TE_meth_average,.(class_update),summarise,Mean=mean(States),SD=sd(States))

# Correlation of TE length with number of states ever
cor.test((chromHMM_TE_state$stop-chromHMM_TE_state$start),chromHMM_TE_state$States,method="spearman")
cor.test((TE_meth_average$stop-TE_meth_average$start),TE_meth_average$States,method="spearman")

# Correlation of number of states ever, max in one sample
cor.test(chromHMM_TE_state$Max_states_intra,chromHMM_TE_state$States,method="spearman")

# By chromosome
ddply(chromHMM_TE_state,.(chromosome,Category),summarise,Max_states_intra=mean(Max_states_intra),States=mean(States))
```

### Number of states per TE x sample

```{r states per TE-sample}
# chromHMM
test = read.table("chromHMM/rmsk_TEother_chromHMM_summit_state_counts.txt",sep='\t')
colnames(test) = c("States","Total")
test$Proportion = test$Total/sum(test$Total)
test
## Proportion of those with >1 state in only 2 states
test$Total[2]/sum(test$Total[2:15])

table(chromHMM_TE_state$Max_states_intra)
max_intra_outlier = chromHMM_TE_state[which(chromHMM_TE_state$Max_states_intra > 7),]
ddply(max_intra_outlier,.(),summarise,Length=median(stop-start),Tissues=mean(Tissues)/127)
max_intra_outlier = merge(max_intra_outlier,rmsk_TE[,c(TE_coordinates,cohorts)],by=TE_coordinates,all.x=TRUE)
max_intra_outlier
write.table(max_intra_outlier[,c(TE_coordinates,"class_update",cohorts)],file="States_intra_8_TEs.bed",sep='\t',row.names=FALSE,quote=FALSE)

## Max states per TE
table(chromHMM_TE_state$Category)/NUM_TE

ddply(chromHMM_TE_state,.(Category),summarise,Mean=mean(Max_states_intra),SD=sd(Max_states_intra))
ddply(chromHMM_TE_state,.(class_update,Category),summarise,Count=length(Max_states_intra),Mean=mean(Max_states_intra),SD=sd(Max_states_intra))

## TEs with >1 max states in a sample
dim(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit" & chromHMM_TE_state$Max_states_intra > 1),])[1]
dim(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit" & chromHMM_TE_state$Max_states_intra > 1),])[1]/dim(chromHMM_TE_state[which(chromHMM_TE_state$Category == "summit"),])[1]

## Correlation of TE length with max in one sample
by(data=chromHMM_TE_state,chromHMM_TE_state$Category,function(x) cor.test(x$stop-x$start,x$Max_states_intra,method="spearman"))

# WGBS
test = read.table("WGBS/TE_CpG_state_counts.txt",sep='\t')
test$Proportion = test$V2/sum(test$V2)
test
sum(test$Proportion[2:4])

## Proportion of those with >2 states
sum(test$Proportion[3:4])/sum(test$Proportion)
```

### DNase/H3K27ac peak analysis

```{r peak analysis}
# Number of DNase peaks per TE
d = table(unlist(TE_DNase_peaks[,8:60]))
d

# % of TE x sample with 1 peak
d[2]/sum(d[2:length(d)])

# Number of H3K27ac peaks per TE
e = table(unlist(TE_H3K27ac_peaks[,8:105]))
e

# % of TE x sample with 1 peak
e[2]/sum(e[2:length(e)])
```

### Supplementary Figure 6

```{r Figure S6, echo=FALSE}
source("R_scripts/compare_marks.R")

a = ggplot(by_chromHMM,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=3) + facet_grid(Mark~chromHMM) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank()) + ylab("Proportion of TE x sample") + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE)

b = ggplot(by_WGBS,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=3) + facet_grid(Mark~WGBS) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank()) + ylab("Proportion of TE x sample") + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE)

c = ggplot(by_DNase,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=3) + facet_grid(Mark~DNase) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),axis.title.y = element_blank()) + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE)

d = ggplot(by_H3K27ac,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=3) + facet_grid(Mark~H3K27ac) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),axis.title.y = element_blank()) + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE)

e = ggplot(by_RNA,aes(x=1,y=Proportion,fill=State)) + geom_bar(stat="identity",position="stack") + 
  geom_text(aes(label = round(Proportion,2)), position = position_stack(vjust=0.5),size=3) + facet_grid(Mark~RNA) + 
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),axis.title.y = element_blank()) + 
  scale_fill_manual(values=c(chromHMM_colors,meth_colors,setNames(c("gold","darkgrey"),c("True","False"))),guide=FALSE)

grid.arrange(a,b,c,d,e, nrow=2, layout_matrix=rbind(c(1,1,1,1),c(2,3,4,5)),widths=c(0.4,0.2,0.2,0.2))
```

### Compare mark analysis (get p-values and residuals)

```{r compare mark analysis}
# Number of samples with each combination of data
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$DNase)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$H3K27ac)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$DNase) & !is.na(metadata$H3K27ac)),])[1]
dim(metadata[which(!is.na(metadata$DNase) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$H3K27ac) & !is.na(metadata$RNA)),])[1]
dim(metadata[which(!is.na(metadata$WGBS) & !is.na(metadata$DNase) & !is.na(metadata$H3K27ac) & !is.na(metadata$RNA)),])[1]

# Tables for each mark
ddply(WGBS_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(DNase_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(H3K27ac_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(RNA_norm,.(),transform,Proportion=TE_sample/sum(TE_sample))
ddply(compare_marks_all,.(chromHMM),summarise,TE_sample=sum(TE_sample),Proportion=sum(TE_sample)/sum(compare_marks_unique$TE_sample))

# Proportion of each state in other states
by_chromHMM
by_WGBS
by_DNase
by_H3K27ac
by_RNA

# Chi-squared for each mark combination (chromHMM not normalized)
table_list = list(xtabs(TE_sample~chromHMM+WGBS,compare_marks_all),
                  xtabs(TE_sample~chromHMM+DNase,compare_marks_all),
                  xtabs(TE_sample~chromHMM+H3K27ac,compare_marks_all),
                  xtabs(TE_sample~chromHMM+RNA,compare_marks_all),
                  xtabs(TE_sample~WGBS+DNase,compare_marks_unique),
                  xtabs(TE_sample~WGBS+H3K27ac,compare_marks_unique),
                  xtabs(TE_sample~WGBS+RNA,compare_marks_unique),
                  xtabs(TE_sample~DNase+H3K27ac,compare_marks_unique),
                  xtabs(TE_sample~DNase+RNA,compare_marks_unique),
                  xtabs(TE_sample~H3K27ac+RNA,compare_marks_unique))
chisq_list = lapply(table_list,function(x) chisq.test(x))
p.adjust(unlist(lapply(chisq_list,function(x) x$p.value)),method="bonf")
lapply(chisq_list,function(x) x$residuals)

# Proportion of TEs x sample in each chromHMM state also overlapping DNase peak and H3K27ac peak
test1 = ddply(compare_marks_all[which(!is.na(compare_marks_all$DNase) & !is.na(compare_marks_all$H3K27ac)),],.(chromHMM,DNase,H3K27ac),summarise,TE_sample=sum(TE_sample))
test2 = merge(ddply(test1,.(chromHMM),summarise,TE_sample=sum(TE_sample)),ddply(test1[which(test1$DNase == "True" & test1$H3K27ac == "True"),],.(chromHMM),summarise,TE_sample=sum(TE_sample)),by="chromHMM")
test2$TE_sample_proportion = test2$TE_sample.y/test2$TE_sample.x
test2[order(test2$TE_sample_proportion),]

# Proportion of TEs x sample in each WGBS state also overlapping DNase peak and H3K27ac peak
test1 = ddply(compare_marks_unique[which(!is.na(compare_marks_unique$DNase) & !is.na(compare_marks_unique$H3K27ac) & !is.na(compare_marks_unique$WGBS)),],.(WGBS,DNase,H3K27ac),summarise,TE_sample=sum(TE_sample))
test2 = merge(ddply(test1,.(WGBS),summarise,TE_sample=sum(TE_sample)),ddply(test1[which(test1$DNase == "True" & test1$H3K27ac == "True"),],.(WGBS),summarise,TE_sample=sum(TE_sample)),by="WGBS")
test2$TE_sample_proportion = test2$TE_sample.y/test2$TE_sample.x
test2[order(test2$TE_sample_proportion),]

# QC
## Total TE x sample x chromHMM states = 570405028 - matches total TEs in rmsk_TEother_chromHMM_summit_sorted.txt and sum(compare_marks_all$TE_sample)
## Total TE x sample = 562520596 - matches total unique TEs (121*NUM_TE + 6*NUM_TE_noY) and sum(compare_marks_unique$TE_sample)
## Total TE x sample correct for each metric
apply(sample_counts,2,function(x) (x[3]*NUM_TE) + ((x[1]-x[3])*NUM_TE_noY))
apply(compare_marks_unique,2,function(x) sum(compare_marks_unique[which(!is.na(x)),]$TE_sample))
```

## By sample

### Figure 4

```{r Figure 4, echo=FALSE}
#source("R_scripts/proportion.R")
#source("R_scripts/proportion_class.R")
source("R_scripts/RNA_coverage.R")

a = ggplot(by_sample_all,aes(x=reorder(State,-Contribution,max),y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank()) + ylab("Proportion of state in TEs") + scale_color_manual(values=group_colors,guide=FALSE) + geom_errorbar(aes(ymax=Contribution,ymin=Contribution),position=position_dodge(),color="black") + geom_hline(yintercept = MERGED_TE_WIDTH/GENOME_WIDTH,linetype="dashed") + geom_hline(yintercept = TE_CPGS/ALL_CPGS,linetype="dotdash") + coord_flip() + scale_y_reverse(limits=c(1,0))

b = ggplot(by_sample_class,aes(x=State,y=Proportion_state,color=Group)) + geom_point(position = position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank()) + ylab("Proportion of state in class") + scale_color_manual(values=group_colors,guide=FALSE) + scale_x_discrete(limits=rev(as.character(unique(by_sample_all[order(by_sample_all$Contribution),]$State)))) + geom_errorbar(aes(ymax=Contribution,ymin=Contribution),position=position_dodge(),color="black") + facet_wrap(~class,nrow=1) + coord_flip() + geom_hline(data=by_sample_class[which(by_sample_class$State == "1_TssA" & by_sample_class$Sample == "E001"),],aes(yintercept=Bases_class/GENOME_WIDTH),linetype="dashed") + geom_hline(data=by_sample_class[which(by_sample_class$State == "Hypomethylated" & by_sample_class$Sample == "E003"),],aes(yintercept=Bases_class/ALL_CPGS),linetype="dotdash")

# Average expression of TEs vs. genome
c = ggplot(RNA_proportion,aes(x=1,y=Ratio,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank()) + scale_color_manual(values=group_colors,guide=FALSE) + coord_flip() + ylab("TE:Genome average expression") + scale_y_continuous(limits=c(0,1)) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black")

d = ggplot(class_RNA,aes(x=1,y=Ratio,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),axis.text.y = element_blank(),axis.ticks.y = element_blank()) + scale_color_manual(values=group_colors,guide=FALSE) + coord_flip() + ylab("Class:Genome average expression") + ylim(0,1) + facet_wrap(~class_update,nrow=2) + stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom= "crossbar",color="black")

legend = get_legend(ggplot(by_sample_all,aes(x=State,y=Proportion,color=Group)) + geom_point(position = position_jitter(width = 0.3),size=3) + scale_color_manual(values=group_colors) + theme(legend.margin=margin(0,0,0,0))) 

grid.arrange(a,b,c,d,legend,nrow=3,ncol=3,layout_matrix=rbind(c(1,3,5),c(1,4,5),c(2,2,5)),heights=c(0.2,0.3,0.5),widths=c(0.45,0.45,0.1))
```

### By-sample analysis

```{r analysis Fig 4a, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 4")}
# Stats
by_sample_stats = ddply(by_sample_all,~State,summarise, Range = max(Proportion)-min(Proportion), Median = median(Proportion), Mean = mean(Proportion), CV = sd(Proportion)/mean(Proportion))
by_sample_stats[order(by_sample_stats$Range),]

ddply(by_sample_all,~State,function(x) x[which.min(x$Proportion),])
ddply(by_sample_all,~State,function(x) x[which.max(x$Proportion),])

# Kruskal-Wallis test by sample classification
# MHC adjusted for every state (not category)
by_sample_KW = ddply(by_sample_all,~State,summarise,Group = p.adjust(unlist(kruskal.test(Proportion,Group))["p.value"],method="bonf"),
      Anatomy = p.adjust(unlist(kruskal.test(Proportion,Anatomy))["p.value"],method="bonf"),
      Type = p.adjust(unlist(kruskal.test(Proportion,Type))["p.value"],method="bonf"),
      Age = p.adjust(unlist(kruskal.test(Proportion,Age))["p.value"],method="bonf"),
      Germline = p.adjust(unlist(kruskal.test(Proportion,Germline))["p.value"],method="bonf"))
by_sample_KW = merge(by_sample_KW,ddply(droplevels(by_sample_all[which(!(by_sample_all$State %in% meth_states)),]),~State,summarise,Cancer = p.adjust(unlist(kruskal.test(Proportion,Cancer))["p.value"],method="bonf")),by="State",all.x=TRUE)
by_sample_KW = melt(by_sample_KW,id.var="State")
colnames(by_sample_KW)[2:3] = c("Category","Pvalue")
by_sample_KW[which(by_sample_KW$Pvalue < 0.05),]

# Permutation tests by sample classification
# MHC adjusted for category and state, not direction
permute_up = ddply(by_sample_all[,c("State","Sample","Proportion","Contribution")],.(State),function(y) permute_by_sample(y,"Proportion","+",unique(y$Contribution),"Proportion",-1))
permute_up = permute_up[which(permute_up$Samples.x > 0),]
permute_up$Padjust = p.adjust(permute_up$Pvalue,method="fdr")
permute_up[which(permute_up$Padjust < 0.1 & permute_up$Category == "Group" & permute_up$State %in% states[c(1:7,16:17,20:21)]),]

permute_down = ddply(by_sample_all[,c("State","Sample","Proportion","Contribution")],.(State),function(y) permute_by_sample(y,"Proportion","-",unique(y$Contribution),"Proportion",-1))
permute_down = permute_down[which(permute_down$Samples.y - permute_down$Samples.x > 0),]
permute_down$Padjust = p.adjust(permute_down$Pvalue,method="fdr")
permute_down[which(permute_down$Padjust < 0.1 & permute_down$Category == "Group" & permute_down$State %in% states[c(1:7,16:17,20:21)]),]

# Mean Group proportion by state 
test = aggregate(data=by_sample_all,Proportion~State+Group,mean)
test[order(test$State,test$Proportion),]
```

### By sample and class analysis

```{r analysis Fig S11a, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 4")}
by_sample_stats_class = ddply(na.omit(by_sample_class),.(State,class),summarise, Range = max(Proportion_state)-min(Proportion_state), Median = median(Proportion_state), Mean = mean(Proportion_state),CV = sd(Proportion_state)/mean(Proportion_state))
by_sample_stats_class[order(by_sample_stats_class$Range),]
by_sample_stats_class[order(by_sample_stats_class$CV),]

# Kruskal-Wallis
# MHC adjusted for state and class
by_sample_KW_class = ddply(by_sample_class,.(State,class),summarise,Group = p.adjust(unlist(kruskal.test(Proportion_state,Group))["p.value"],method="bonf"))
by_sample_KW_class[which(by_sample_KW_class$Group < 0.05),]

# Number of enriched samples by Group
# MHC adjusted for state and class
permute_up = permute_up = ddply(by_sample_class[,c("class","State","Sample","Proportion_state","Contribution")],.(State,class),function(y) permute_by_sample(y,"Proportion_state","+",unique(y$Contribution),"Proportion_state",-1))
permute_up = permute_up[which(permute_up$Samples.x > 0),]
permute_up$Padjust = p.adjust(permute_up$Pvalue,method="fdr")
permute_up[which(permute_up$Padjust < 0.1 & permute_up$Category == "Group" & permute_up$State %in% states[c(1:7,16:17,20:21)]),]

permute_down =  permute_down = ddply(by_sample_class[,c("class","State","Sample","Proportion_state","Contribution")],.(State,class),function(y) permute_by_sample(y,"Proportion_state","-",unique(y$Contribution),"Proportion_state",-1))
permute_down = permute_down[which(permute_down$Samples.y - permute_down$Samples.x > 0),]
permute_down$Padjust = p.adjust(permute_down$Pvalue,method="fdr")
permute_down[which(permute_down$Padjust < 0.1 & permute_down$Category == "Group" & permute_down$State %in% states[c(1:7,16:17,20:21)]),]

test = ddply(by_sample_class,.(State,class,Group),summarise,Proportion_state=mean(Proportion_state),Contribution=unique(Contribution))
test[order(test$State,test$class,test$Proportion),]

# Enrichments over expectation
table(by_sample_class[which(by_sample_class$Enrichment > 0),c("State","class")])

by_sample_class[which(by_sample_class$class %in% c("DNA","LINE","SINE","LTR") & by_sample_class$State %in% states[c(1:3,6:7,16,20:21)] & by_sample_class$Enrichment > 0),c("Sample","class","State","Enrichment",sample_categories)]
```

### DNase/H3K27ac peak analysis (vs. bp)

```{r peak analysis pt2}
# % of DNase peaks overlapping TEs, overall
sum(DNase_stats$Summit_in_TE)/sum(DNase_stats$Peaks)

## By sample
ddply(DNase_stats,.(),summarise,Median=median(Summit_in_TE/Peaks),Mean=mean(Summit_in_TE/Peaks),SD=sd(Summit_in_TE/Peaks),Min=min(Summit_in_TE/Peaks),Max=max(Summit_in_TE/Peaks))
DNase_stats[which.min(DNase_stats$Summit_in_TE/DNase_stats$Peaks),]
DNase_stats[which.max(DNase_stats$Summit_in_TE/DNase_stats$Peaks),]

# % of H3K27ac peaks overlapping TEs, overall
sum(H3K27ac_stats$Summit_in_TE)/sum(H3K27ac_stats$Peaks)

## By sample
ddply(H3K27ac_stats,.(),summarise,Median=median(Summit_in_TE/Peaks),Mean=mean(Summit_in_TE/Peaks),SD=sd(Summit_in_TE/Peaks),Min=min(Summit_in_TE/Peaks),Max=max(Summit_in_TE/Peaks))
H3K27ac_stats[which.min(H3K27ac_stats$Summit_in_TE/H3K27ac_stats$Peaks),]
H3K27ac_stats[which.max(H3K27ac_stats$Summit_in_TE/H3K27ac_stats$Peaks),]
```

### Expression analysis

```{r ratio expression}
median(RNA_proportion$Ratio)
ddply(RNA_proportion,.(Group),summarise,Median=median(Ratio))

ddply(class_RNA,.(class_update),summarise,Median=median(Ratio))
dcast(ddply(class_RNA,.(Group,class_update),summarise,Median=median(Ratio)),class_update~Group,value.var="Median")
```

### Supplementary Figure 7

```{r Figure S7, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 4","setup","load libraries","set plot parameters","load functions","create essential matrices","load metadata","load essential matrices","define vectors","define facet labels","load feature sizes","calculate TE stats")}
source("R_scripts/state_lengths.R")

# Plot bases in genome/TEs in each state by sample, increasing order
all_state_proportion_matrix = dcast(all_state_proportion[which(all_state_proportion$Cohort %in% c("Genome","TE")),],State+Sample+Group~Cohort,value.var="Bases")

a = ggplot(all_state_proportion_matrix,aes(x=log10(Genome),y=log10(TE),color=Group)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + theme(axis.text.x = element_text(angle=90)) + facet_wrap(~State,scale="free",ncol=6) + xlab("Bases/CpGs in genome (log10)") + ylab("Bases/CpGs in TEs (log10)")

# Number vs. length of blocks/peaks by sample
all_lengths_blocks = ddply(all_lengths,.(State,Sample),summarise,Mean=mean(Length),SD=sd(Length),Median=median(Length),Blocks=length(Length))
all_lengths_blocks = merge(all_lengths_blocks,metadata[,c("Sample",sample_categories)],by="Sample",all.x=TRUE)

b = ggplot(all_lengths_blocks,aes(x=log10(Blocks),y=log10(Median),color=Group)) + geom_point() + xlab("Number of blocks/peaks (log10)") + ylab("Median length of blocks/peaks (log10(bp))") + scale_color_manual(values=group_colors,guide=FALSE) + facet_wrap(~State,scales="free",ncol=6) 

grid.arrange(a,b,heights=c(0.55,0.45))
```

### Compare TE bases to non-TE bases by sample

```{r proportion investigation}
ddply(all_state_proportion[which(all_state_proportion$Cohort %in% c("Genome","TE")),],.(State,Cohort),summarise,Median=median(Bases),Mean=mean(Bases),SD=sd(Bases),Min=min(Bases),Max=max(Bases),CV=sd(Bases)/mean(Bases))

dcast(all_state_proportion[which(all_state_proportion$Mark == "WGBS" & all_state_proportion$Cohort == "Genome"),c("Sample","State","Proportion")],Sample~State,value.var="Proportion")

test = ddply(all_state_proportion[which(all_state_proportion$Feature == "Genome"),],.(State),transform,Mean=mean(Proportion),SD=sd(Proportion))
tail(test[order(abs(test$Proportion-test$Mean)/test$SD),c("Sample","State","Proportion","Mean","SD")],n=15)

test = ddply(all_state_proportion_matrix,.(State,Group),summarise,Genome_median=median(Genome))
test[order(test$State,test$Genome_median),]

# Not MHC corrected (0.05/21 ~ 0.0024)
ddply(all_state_proportion_matrix,.(State),summarise,Genome_corr = unlist(cor.test(Genome-TE,TE))["estimate.cor"],Genome_pvalue = unlist(cor.test(Genome-TE,TE)["p.value"]))
ddply(all_state_proportion_matrix,.(State),summarise,Genome_corr = unlist(cor.test(Genome,TE/Genome))["estimate.cor"],Genome_pvalue = unlist(cor.test(Genome,TE/Genome)["p.value"]))
```

### State/peak lengths analysis

```{r analysis Fig S1}
# Length of states
ddply(all_lengths,.(State),summarise,Mean=mean(Length),SD=sd(Length),Median=median(Length))

# Differences in length of states by sample
ddply(all_lengths,.(State),summarise,KW_length=unlist(kruskal.test(Length~Sample))["p.value"])

# Differences in number of blocks by sample
test = ddply(all_lengths_blocks,.(State),summarise,Range_number=max(Blocks)-min(Blocks),CV_number=sd(Blocks)/mean(Blocks),Range_length=max(Median)-min(Median),CV_length=sd(Median)/mean(Median))
test[order(test$CV_number),]
test[order(test$CV_length),]

# Comparing number of blocks to length of states
test = ddply(all_lengths_blocks,.(State),summarise,Cor=unlist(cor.test(Blocks,Median,method="spearman"))["estimate.rho"],Pvalue=unlist(cor.test(Blocks,Median,method="spearman"))["p.value"])
test[order(as.numeric(test$Cor)),]

rm(all_lengths)
```

### Supplementary Figure 8

```{r Figure S8, echo=FALSE}
source("R_scripts/WGBS_sample_CpG_average.R")

# Other sample classifications
by_sample_all_long = melt(by_sample_all[,c("State","Sample","Proportion","Contribution",sample_categories[2:6])],id.vars=c("State","Sample","Proportion","Contribution"))
colnames(by_sample_all_long)[5:6] = c("Category","Group")

a = ggplot(by_sample_all_long,aes(x=reorder(State,-Contribution,max),y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank()) + ylab("Proportion of state in TEs") + scale_color_manual(values=c(anatomy_colors,type_colors,age_colors,cancer_colors,germline_colors),guide=FALSE) + geom_errorbar(aes(ymax=Contribution,ymin=Contribution),position=position_dodge(),color="black") + geom_hline(yintercept = MERGED_TE_WIDTH/GENOME_WIDTH,linetype="dashed") + geom_hline(yintercept = TE_CPGS/ALL_CPGS,linetype="dotdash") + coord_flip() + scale_y_reverse(limits=c(1,0)) + facet_wrap(~Category,nrow=1)

legend1 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Anatomy"),],aes(x=reorder(State,-Contribution,max),y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0)) + scale_color_manual(values=c(anatomy_colors),name="Anatomy")) 

legend2 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Type"),],aes(x=reorder(State,-Contribution,max),y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0)) + scale_color_manual(values=c(type_colors),name="Type"))

legend3 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Age"),],aes(x=reorder(State,-Contribution,max),y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0)) + scale_color_manual(values=c(age_colors),name="Age"))

legend4 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Cancer"),],aes(x=reorder(State,-Contribution,max),y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0)) + scale_color_manual(values=c(cancer_colors),name="Cancer"))

legend5 = get_legend(ggplot(by_sample_all_long[which(by_sample_all_long$Category == "Germline"),],aes(x=reorder(State,-Contribution,max),y=Proportion,color=Group)) + geom_point(position=position_jitterdodge(dodge.width = 0.6)) + theme(panel.grid=element_blank(),axis.title.y=element_blank(),legend.position = "bottom",legend.margin=margin(0,0,0,0)) + scale_color_manual(values=c(germline_colors),name="Germline"))

grid.arrange(a,legend1,legend2,legend3,legend4,legend5,nrow=4,layout_matrix=rbind(c(1,1),c(2,2),c(3,5),c(6,4)),heights=c(0.7,0.2,0.05,0.05),widths=c(0.6,0.4))
```

## Subfamily enrichment

### Set thresholds 

```{r set enrichment thresholds}
THRESHOLD_IJK_MEMBER = 10
THRESHOLD_IK_MEMBER = 30
THRESHOLD_LOR = 1.5
THRESHOLD_PC = 0.01

THRESHOLD_IJK_BASE = 600
THRESHOLD_IK_BASE = 5000
THRESHOLD_IJK_CPG = 6
THRESHOLD_IK_CPG = 50
```

### Figure 5

```{r Figure 5, echo=FALSE}
source("R_scripts/subfamily_enrichment.R") #Requires proportion.R
source("R_scripts/subfamily_potential.R")
source("R_scripts/subfamily_PCA.R")

# Potential
a = ggplot(subfam_states,aes(x=value,fill=variable)) + geom_histogram(position="dodge",binwidth=1) + ylab("Subfamilies") + scale_fill_discrete(guide=FALSE) + xlab("States")

# Clustering
family_shapes=setNames(c(17,rep(16,20),18,rep(16,23)),sort(as.vector(unique(rmsk_TE_subfamily$family))))

b = ggplot(sample_chromHMM_tsne,aes(x=V1,y=V2,color=Group,shape=Age)) + geom_point(size=3) + scale_color_manual(values=group_colors,guide=FALSE) + scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=c("Cell line","Fetal","Non-fetal","Unknown"),guide=FALSE) + xlab("Axis 1") + ylab("Axis 2")

c = ggplot(subfamily_chromHMM_tsne,aes(x=V1,y=V2,color=class_update,shape=family)) + geom_point(size=3) + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE) + xlab("Axis 1") + ylab("Axis 2")

# Enrichment
d = ggplot(subfamily_state_sample_counts,aes(x=State,y=V1,color=class_update)) + geom_point(position=position_jitterdodge(jitter.width=0.5)) + coord_flip() + theme(axis.title.y=element_blank()) + ylab("Number of samples enriched LOR > 1.5") + scale_color_manual(values=class_colors,guide=FALSE) + scale_y_continuous(breaks=c(0,seq(10,120,10),sample_counts["All","chromHMM"]),expand = c(0.01,0.01)) + geom_rect(xmin=0.5,xmax=1.5,ymin=sample_counts["All","H3K27ac"]+1,ymax=sample_counts["All","chromHMM"],fill="grey",color=NA) + geom_rect(xmin=1.5,xmax=2.5,ymin=sample_counts["All","DNase"]+1,ymax=sample_counts["All","chromHMM"],fill="grey",color=NA) + geom_rect(xmin=2.5,xmax=6.5,ymin=sample_counts["All","WGBS"]+1,ymax=sample_counts["All","chromHMM"],fill="grey",color=NA) + geom_vline(xintercept=seq(1.5,20.5,1), colour='grey') + scale_x_discrete(limits = rev(levels(subfamily_state_sample_counts$State)),labels=as.vector(rev(apply(subfamily_state_sample_counts_combine,1,function(x) paste(x[1],"\n(n=",x[2],")",sep="")))))

# HOMER
tf = read.table("enrichment/TF_matrix.txt",sep='\t',header=TRUE)
tf = tf[,1:4]
tf$Target_PC = as.numeric(gsub("%","",tf$Target_PC))
tf$Background_PC = as.numeric(gsub("%","",tf$Background_PC))
tf = melt(tf,id.vars=c("subfamily","TF"))

e = ggplot(tf,aes(x=TF,y=value,fill=variable)) + geom_bar(stat="identity",position="dodge") + facet_grid(~subfamily,scales="free_x",space="free_x") + ylab("% TEs with motif") + theme(axis.text.x = element_text(angle=90,hjust=1),axis.title.x = element_blank()) + scale_fill_discrete(guide=FALSE)

grid.arrange(a,b,c,d,e,nrow=2,ncol=3,layout_matrix=rbind(c(1,2,3),c(4,4,5)),heights=c(0.4,0.6),widths=c(0.3,0.35,0.35))
```

### Subfamily potential

```{r subfam potential}
# Number of subfamilies ever in each state and median samples in state
ddply(subfamily_state_potential,.(State),summarise,Subfamilies=sum(Samples > 0),Samples=median(Samples))

# Number of states per subfamily (across all samples, max in one sample)
by(subfam_states,subfam_states$variable,function(x) table(x$value))

ddply(subfam_states,.(variable),summarise,Median=median(value))

# Outliers
subfam_states[which(subfam_states$value < 15 & subfam_states$variable == "States"),]
subfam_states[which(subfam_states$value < 10 & subfam_states$variable == "Intra"),]

# Correlation with length
by(subfam_states,subfam_states$variable,function(x) cor.test(x$value,x$Total_length))
by(subfam_states,subfam_states$variable,function(x) cor.test(x$value,x$CpGs))

# Distribution of samples in state
ggplot(subfamily_state_potential,aes(x=Samples)) + geom_histogram() + facet_wrap(~State,scales="free_y") + ylab("Subfamilies")

ddply(subfamily_state_potential[which(subfamily_state_potential$State %in% chromHMM_states & subfamily_state_potential$Samples >= sample_counts["All","chromHMM"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
ddply(subfamily_state_potential[which(subfamily_state_potential$State %in% meth_states & subfamily_state_potential$Samples >= sample_counts["All","WGBS"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
ddply(subfamily_state_potential[which(subfamily_state_potential$State == "DNase" & subfamily_state_potential$Samples >= sample_counts["All","DNase"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
ddply(subfamily_state_potential[which(subfamily_state_potential$State == "H3K27ac" & subfamily_state_potential$Samples >= sample_counts["All","H3K27ac"]*0.75),],.(State),summarise,Subfamilies=length(Samples))
```

### tSNE

```{r subfam clustering}
# Matrix dimensions
dim(sample_chromHMM_3D)
dim(subfamily_chromHMM_3D)
dim(sample_WGBS_3D)
dim(subfamily_WGBS_3D)
dim(subfamily_DNase_enrichment)
dim(subfamily_H3K27ac_enrichment)
```

### Supplementary Figure 9

```{r Figure S9, echo=FALSE}
## Samples
a = ggplot(sample_WGBS_tsne,aes(x=V1,y=V2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + xlab("Axis 1") + ylab("Axis 2") + scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=c("Cell line","Fetal","Non-fetal","Unknown"),guide=FALSE)
b = ggplot(sample_DNase_tsne,aes(x=V1,y=V2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + xlab("Axis 1") + ylab("Axis 2") + scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=c("Cell line","Fetal","Non-fetal","Unknown"),guide=FALSE)
c = ggplot(sample_H3K27ac_tsne,aes(x=V1,y=V2,color=Group,shape=Age)) + geom_point() + scale_color_manual(values=group_colors,guide=FALSE) + xlab("Axis 1") + ylab("Axis 2") + scale_shape_manual(values=setNames(c(16,17,18,16),c("Non_fetal","Fetal","Cell_line","Unknown")),labels=c("Cell line","Fetal","Non-fetal","Unknown"),guide=FALSE)

## Subfamilies
d = ggplot(subfamily_WGBS_tsne,aes(x=V1,y=V2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE)+ xlab("Axis 1") + ylab("Axis 2")
e = ggplot(subfamily_DNase_tsne,aes(x=V1,y=V2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE)+ xlab("Axis 1") + ylab("Axis 2")
f = ggplot(subfamily_H3K27ac_tsne,aes(x=V1,y=V2,color=class_update,shape=family)) + geom_point() + scale_color_manual(values=class_colors,guide=FALSE) + scale_shape_manual(values=family_shapes,guide=FALSE)+ xlab("Axis 1") + ylab("Axis 2")

## chromHMM other groupings
g = ggplot(sample_chromHMM_tsne,aes(x=V1,y=V2,color=Anatomy)) + geom_point() + scale_color_manual(values=anatomy_colors,guide=FALSE) + xlab("Axis 1") + ylab("Axis 2")
h = ggplot(sample_chromHMM_tsne,aes(x=V1,y=V2,color=Type)) + geom_point() + scale_color_manual(values=type_colors,guide=FALSE) + xlab("Axis 1") + ylab("Axis 2")
i = ggplot(sample_chromHMM_tsne,aes(x=V1,y=V2,color=Germline)) + geom_point() + scale_color_manual(values=germline_colors,guide=FALSE) + xlab("Axis 1") + ylab("Axis 2")
j = ggplot(sample_chromHMM_tsne,aes(x=V1,y=V2,color=Cancer)) + geom_point() + scale_color_manual(values=cancer_colors,guide=FALSE) + xlab("Axis 1") + ylab("Axis 2")

grid.arrange(a,b,c,d,e,f,g,h,i,j,nrow=4,layout_matrix=rbind(c(1,2,3),c(4,5,6),c(7,7,8),c(9,9,10)))
```

### Enrichment thresholds

```{r excluded subfamilies}
# Number of subfamilies excluded by enrichment thresholds
# chromHMM, all chromosomes
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count <= THRESHOLD_IK_MEMBER),])[1]
# chromHMM, no Y
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count > THRESHOLD_IK_MEMBER & rmsk_TE_subfamily$Count_noY <= THRESHOLD_IK_MEMBER),])[1]

# WGBS, all chromosomes (excluded only from WGBS)
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count_CpGs <= THRESHOLD_IK_MEMBER & rmsk_TE_subfamily$Count > THRESHOLD_IK_MEMBER),])[1]

rmsk_TE_subfamily[which(rmsk_TE_subfamily$Count_noY <= THRESHOLD_IK_MEMBER | rmsk_TE_subfamily$Count_CpGs <= THRESHOLD_IK_MEMBER),c("subfamily","family","class_update","Count","Count_noY","Count_CpGs","Total_length","Total_length_noY","CpGs")]

# Subfamilies with no CpGs
rmsk_TE_subfamily[which(rmsk_TE_subfamily$CpGs == 0),c("subfamily","Count","Count_noY","Total_length","Total_length_noY","CpGs")]

# Number of enrichments excluded by enrichment thresholds
test = table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR & subfamily_state_sample_combined$Count > THRESHOLD_IK_MEMBER & subfamily_state_sample_combined$Members <= THRESHOLD_IJK_MEMBER),]$State)
test
sum(test)
```

### Number of enrichments by state and number of enriched subfamilies by state, overall and by class

```{r analysis Fig 5}
enrichment_table = merge(as.data.frame(table(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),]$State)),subfamily_state_sample_counts_combine,by.x="Var1",by.y="State")
colnames(enrichment_table)[1:3] = c("State","Enrichments","Subfamilies")
enrichment_table = enrichment_table[match(states[1:21],enrichment_table$State),]

sum(enrichment_table$Enrichments)
enrichment_table$Enrichments_pc = enrichment_table$Enrichments/sum(enrichment_table$Enrichments)
enrichment_table

# Number of subfamilies enriched at least once, overall and active states
length(unique(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),]$subfamily))
length(unique(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR & subfamily_state_sample_filter$State %in% c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")),]$subfamily))

# Enriched classes
test = dcast(aggregate(data=subfamily_state_sample_counts,V1~State+class_update,function(x) sum(x > 0)),class_update~State,value.var = "V1")
table(rmsk_TE_subfamily$class_update)/968
apply(enrichment_table[,4:9],1,function(x) x/sum(x))
apply(test[,2:22],2,function(x) log2((x/sum(x))/(table(rmsk_TE_subfamily$class_update)/968)))

# Enriched families
test = dcast(aggregate(data=subfamily_state_sample_counts,V1~State+family,function(x) sum(x > 0)),family~State,value.var = "V1")
table(rmsk_TE_subfamily$family)/968
test_log = as.data.frame(apply(test[,2:22],2,function(x) log2((x/sum(x))/(table(rmsk_TE_subfamily$family)/968))))
test_log$Subfamilies = as.numeric(table(rmsk_TE_subfamily$family)[rownames(test_log)])
lapply(colnames(test_log)[1:21],function(x) test_log[which(test_log[,x] > 0 & test_log$Subfamilies > 0),c(x,"Subfamilies")])
```

### Supplementary Figure 10

```{r Figure S10, echo=FALSE}
a = ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=Percent,fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Proportion of members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=states[1:21])

b = ggplot(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],aes(x=State,y=log10(Members),fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Log10 number of members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=states[1:21])

c = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],aes(x=State,y=Percent,fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Proportion of members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=states[1:21])

d = ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],aes(x=State,y=log10(Members),fill=State)) + geom_boxplot() + theme(panel.grid=element_blank(),axis.title.x=element_blank(),axis.text.x=element_text(angle=45,hjust=1)) + ylab("Log10 number of members in state") + scale_fill_manual(values=all_state_colors,guide=FALSE) + scale_x_discrete(limits=states[1:21])

grid.arrange(a,c,b,d,nrow=2,layout_matrix=cbind(c(1,2),c(3,4)))
```

### Average members in a state for subfamilies in a state

```{r bp members percent in state}
# All instances
ddply(subfamily_state_sample_combined,~State,summarize,Percent_median=median(na.omit(Percent)), Members_median=median(na.omit(Members)),Percent_min=min(na.omit(Percent)), Members_min=min(na.omit(Members)), Percent_max=max(na.omit(Percent)), Members_max=max(na.omit(Members)))   

# Enriched
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],~State,summarise,Percent_median=median(Percent), Members_median=median(Members),Percent_min=min(Percent), Members_min=min(Members), Percent_max=max(Percent), Members_max=max(Members))

# >1%
ddply(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],~State,summarise, Percent_median=median(Percent),Members_median=median(Members),Percent_min=min(Percent), Members_min=min(Members), Percent_max=max(Percent), Members_max=max(Members))

# Max by state (outliers)
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],.(State),function(x) x[which.max(x$Percent),])
ddply(subfamily_state_sample_filter[which(subfamily_state_sample_filter$Enrichment > THRESHOLD_LOR),],.(State),function(x) tail(x[order(x$Percent),c("subfamily","family","class_update","State","Sample",sample_categories,"Members","Percent")],n=10))
```

### WGBS subfamily state stats

```{r analysis Fig S13c}
TE_meth_subfamily_percent = subfamily_CpG_meth[,c("subfamily","family","class_update","State","Sample","Percent")]

# Average, max, and range of proportion of subfamily in methylation state plus all members in a state in a sample
TE_meth_subfamily_stats = ddply(TE_meth_subfamily_percent,.(subfamily,State),summarise,Median=median(Percent),Range=max(Percent)-min(Percent),Max=max(Percent),All_members=sum(Percent == 1))

table(TE_meth_subfamily_stats[which(TE_meth_subfamily_stats$All_members > 0),]$State)
merge(TE_meth_subfamily_stats[which(TE_meth_subfamily_stats$All_members > 0),],rmsk_TE_subfamily[,c("subfamily","Count","Count_CpGs","Total_length","Total_length_noY","CpGs")],by="subfamily")

# Average proportion in state, with/without IMR90
ddply(TE_meth_subfamily_stats,.(State),summarise,Median=mean(Median),Range=mean(Range),Max=mean(Max))
```

### Supplementary Figure 11

```{r Figure S11, echo=FALSE}
ggplot(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_combined$State %in% c(chromHMM_states[1:7],meth_states[1:2],"DNase","H3K27ac")),],aes(x=Sample,y=reorder(subfamily,rmsk_TE_subfamily[match(subfamily,rmsk_TE_subfamily$subfamily),]$Total_length),fill=Length_percent_jk)) + geom_tile(linetype="blank") + theme(axis.ticks.x=element_blank(),axis.title.y = element_blank(),axis.text.x = element_blank()) + scale_fill_gradient(name="% state",low="lightblue",high="darkblue",limits=c(0.009,0.12)) +  facet_wrap(~State)
```

### Subfamily stats, >1%

```{r analysis Fig S14a}
# Max % in subfamily by state
ddply(subfamily_state_sample_combined,~State,function(x) x[which.max(x$Length_percent_jk),])

# Max % of state in subfamily without enrichment
ddply(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Enrichment <= THRESHOLD_LOR),],~State,function(x) x[which.max(x$Length_percent_jk),])

# Number of instances by state and number of >1% subfamilies by state, overall and by class
pc_table = merge(as.data.frame(table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),]$State)),subfamily_state_sample_counts_pc_combine,by.x="Var1",by.y="State")
colnames(pc_table)[1:3] = c("State","Instances","Subfamilies")
sum(pc_table$Instances)

# Instances >1% bp of state, also enriched
pc_table = merge(pc_table,as.data.frame(table(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC & subfamily_state_sample_combined$Enrichment > THRESHOLD_LOR),]$State)),by.x="State",by.y="Var1",all=TRUE)
colnames(pc_table)[10] = "Enriched_Instances"
pc_table$Percent = pc_table$Enriched_Instances/pc_table$Instances
pc_table

# Number of subfamilies >1% across all states
length(unique(subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0),]$subfamily))
length(unique(subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0 & subfamily_state_sample_counts_pc$State %in% states[c(1:7,16:17,20:21)]),]$subfamily))

# Subfamilies with >1% of bases in genome
GENOME_WIDTH*THRESHOLD_PC
rmsk_TE_subfamily[which(rmsk_TE_subfamily$Total_length > GENOME_WIDTH*THRESHOLD_PC),c("subfamily","Count","Total_length","Total_length_noY","CpGs")]

# Subfamilies with >1% of CpGs in genome
ALL_CPGS*THRESHOLD_PC
rmsk_TE_subfamily[which(rmsk_TE_subfamily$CpGs > ALL_CPGS*THRESHOLD_PC),c("subfamily","Count","Total_length","Total_length_noY","CpGs")]

# Largest subfamilies vs. >1% of state
test = merge(dcast(subfamily_state_sample_counts_pc,subfamily+family+class_update~State,value.var = "V1"),rmsk_TE_subfamily[,c("subfamily","Count","Total_length","CpGs","Count_CpGs")],by="subfamily",all=TRUE)

test = test[order(test$Total_length,decreasing=TRUE),]
rownames(test) <- NULL
test$Order = rownames(test)
lapply(c(chromHMM_states,"DNase","H3K27ac"),function(x) if(pc_table[which(pc_table$State == x),]$Enriched_Instances > 0) {join(test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs","Count_CpGs",x)],aggregate(data=subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x & subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],Enrichment~subfamily,function(y) sum(y > THRESHOLD_LOR)),by="subfamily",type="left")} else{test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)]})

test = test[order(test$CpGs,decreasing=TRUE),]
rownames(test) <- NULL
test$Order = rownames(test)
lapply(meth_states,function(x) if(pc_table[which(pc_table$State == x),]$Enriched_Instances > 0) {join(test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs","Count_CpGs",x)],aggregate(data=subfamily_state_sample_combined[which(subfamily_state_sample_combined$State == x & subfamily_state_sample_combined$Length_percent_jk > THRESHOLD_PC),],Enrichment~subfamily,function(y) sum(y > THRESHOLD_LOR)),by="subfamily",type="left")} else{test[which(test[,x] > 0),c("Order","subfamily","family","class_update","Count","Total_length","CpGs",x)]})
```

### Subfamily permutation for candidates
### RERUN when updated

Only subfamilies ever enriched in state, active states only, categories with 1+ enrichments. MHC across all subfamilies, states, and categories/groupings with at least one enrichment. 

```{r subfam permutation, cache=TRUE, cache.lazy=FALSE, eval=FALSE}
# Subfamilies ever enriched in a state
subfamily_state_ever = subfamily_state_sample_counts[which(subfamily_state_sample_counts$V1 > 0 & subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)]),c("subfamily","State")]
permutation_input = merge(subfamily_state_sample_combined[which(subfamily_state_sample_combined$Count > THRESHOLD_IK_MEMBER),],subfamily_state_ever,by=c("subfamily","State"))

# Interesting candidates
permute_subfam = ddply(permutation_input[,c("subfamily","State","Sample","Enrichment","Members")],.(subfamily,State),function(x) permute_by_sample(matrix=x,metric="Enrichment",direction="+",threshold=THRESHOLD_LOR,filtering="Members",threshold2=THRESHOLD_IJK_MEMBER))
permute_subfam = permute_subfam[which(permute_subfam$Samples.x > 0),]
permute_subfam$Padjust = p.adjust(permute_subfam$Pvalue,method="fdr")
dim(permute_subfam[which(permute_subfam$Padjust < 0.05),])

permute_subfam_wide = dcast(aggregate(data = permute_subfam[which(permute_subfam$Padjust < 0.05),],Grouping~State+subfamily+Category, paste, collapse = ','),State+subfamily~Category,value.var="Grouping")
permute_subfam_wide = merge(permute_subfam_wide,subfamily_state_sample_counts,by=c("subfamily","State"))[,c(1,10,9,2:8,11)]
write.table(permute_subfam_wide,file="enrichment/state_subfamily_clusters_enriched.txt",sep='\t',row.names = FALSE,quote = FALSE)
```

```{r subfam permutation pc, cache=TRUE, cache.lazy=FALSE, eval=FALSE}
# Subfamilies ever enriched in a state
subfamily_state_ever_pc = subfamily_state_sample_counts_pc[which(subfamily_state_sample_counts_pc$V1 > 0 & subfamily_state_sample_counts_pc$State %in% states[c(1:7,16:17,20:21)]),c("subfamily","State")]
permutation_input_pc = merge(subfamily_state_sample_combined,subfamily_state_ever_pc,by=c("subfamily","State"))

# Interesting candidates
permute_subfam_pc = ddply(permutation_input_pc[,c("subfamily","State","Sample","Length_percent_jk","Members")],.(subfamily,State),function(x) permute_by_sample(x,"Length_percent_jk","+",THRESHOLD_PC,"Members",0))
permute_subfam_pc = permute_subfam_pc[which(permute_subfam_pc$Samples.x > 0),]
permute_subfam_pc$Padjust = p.adjust(permute_subfam_pc$Pvalue,method="fdr")
dim(permute_subfam_pc[which(permute_subfam_pc$Padjust < 0.05),])

permute_subfam_pc_wide = dcast(aggregate(data = permute_subfam_pc[which(permute_subfam_pc$Padjust < 0.05),],Grouping~State+subfamily+Category, paste, collapse = ','),State+subfamily~Category,value.var="Grouping")
permute_subfam_pc_wide = merge(permute_subfam_pc_wide,subfamily_state_sample_counts_pc,by=c("subfamily","State"))[,c(1,10,9,2:8,11)]
write.table(permute_subfam_pc_wide,file="enrichment/state_subfamily_clusters_pc.txt",sep='\t',row.names = FALSE,quote = FALSE)
```

```{r print bedfiles, eval=FALSE}
source('R_scripts/TE_correlation.R')

candidates = read.table(file="enrichment/enrichment_candidates.txt",sep='\t') #Results of candidate Excel analysis
colnames(candidates) = c("State","subfamily")

# Enriched bedfiles
candidates_individual = apply(candidates,1,function(x) get_subfamily_enriched(x[2],x[1]))
names(candidates_individual) = candidates$State
candidates_individual = ldply(candidates_individual)
colnames(candidates_individual)[1] = "State"

# Never in state bedfiles
apply(candidates,1,function(x) write.table(rmsk_TE_measure[which(rmsk_TE_measure$subfamily == x[2] & rmsk_TE_measure[[x[1]]] == 0),c(colnames(rmsk_TE_measure)[1:4],x[1],"strand")],sep='\t',row.names=FALSE,col.names=FALSE,quote=FALSE,file=paste("enrichment/bedfiles/",x[2],"_never_",x[1],".bed",sep="")))

# Subfamilies enriched in many samples
candidates_many = subfamily_state_sample_counts[which(subfamily_state_sample_counts$State %in% states[c(1:7,16:17,20:21)] & subfamily_state_sample_counts$V1 > 30),]
candidates_many[order(candidates_many$State,candidates_many$V1),]

# Enriched bedfiles, large
many_individual = apply(candidates_many,1,function(x) get_subfamily_enriched(x[1],x[4]))
names(many_individual) = candidates_many$State
many_individual = ldply(many_individual)
colnames(many_individual)[1] = "State"

# Never in state bedfiles, large
apply(candidates_many,1,function(x) write.table(rmsk_TE_measure[which(rmsk_TE_measure$subfamily == x[1] & rmsk_TE_measure[[x[4]]] == 0),c(colnames(rmsk_TE_measure)[1:4],x[4],"strand")],sep='\t',row.names=FALSE,col.names=FALSE,quote=FALSE,file=paste("enrichment/bedfiles/",x[1],"_never_",x[4],".bed",sep="")))

write.table(rbind(candidates_individual,many_individual),file="enrichment/enrichment_candidates_indv.txt",sep='\t',row.names = FALSE,quote = FALSE)
```

### Explore candidates

```{r explore candidates, eval=FALSE}
# Genic features
candidates_features = merge(candidates_individual,rmsk_TE[,c("chromosome","start","stop","subfamily","strand","family","class_update",features[c(1:3,5:8)],"coding_exon","Vista_enhancers")],all.x=TRUE,by=c("chromosome","start","stop","subfamily","strand"))
candidates_features_summary = ddply(candidates_features,.(subfamily,class_update,State),function(x) apply(x[,c(features[c(1:3,5:8)],"coding_exon","Vista_enhancers")],2,function(x) sum(!is.na(x))/length(x)))
```

### Subfamily heatmaps

```{r heatmaps, echo=FALSE, eval=FALSE}
# Legends for metadata categories
legend_group = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Group)) + geom_tile() + scale_fill_manual(values=group_colors) + theme(legend.direction = "horizontal"))
legend_anatomy = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Anatomy)) + geom_tile() + scale_fill_manual(values=anatomy_colors) + theme(legend.direction = "horizontal"))
legend_type = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Type)) + geom_tile() + scale_fill_manual(values=brewer.pal(5,"Greens")) + theme(legend.direction = "horizontal"))
legend_age = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Age)) + geom_tile() + scale_fill_manual(values=brewer.pal(4,"YlOrRd"),labels=c("Cell line","Fetal","Non-fetal","Unknown")) + theme(legend.direction = "horizontal"))
legend_cancer = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Cancer)) + geom_tile() + scale_fill_manual(values=c("white","red")) + theme(legend.direction = "horizontal"))
legend_germline = get_legend(ggplot(metadata,aes(x=Sample,y=Exclude,fill=Germline)) + geom_tile() + scale_fill_manual(values=brewer.pal(6,"Dark2")) + theme(legend.direction = "horizontal"))

a = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","1_TssA",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
b = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","2_TssAFlnk",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
c = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","3_TxFlnk",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
d = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","6_EnhG",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
e = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","7_Enh",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
f = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","4_Tx",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
g = plot_binary_heatmap(subfamily_state_sample_filter,"chromHMM","5_TxWk",1,sample_counts["All","chromHMM"],THRESHOLD_LOR)
h = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Hypomethylated",1,sample_counts["All","WGBS"],THRESHOLD_LOR)
i = plot_binary_heatmap(subfamily_state_sample_filter,"WGBS","Intermediate",1,sample_counts["All","WGBS"],THRESHOLD_LOR)
j = plot_binary_heatmap(subfamily_state_sample_filter,"DNase","DNase",1,sample_counts["All","DNase"],THRESHOLD_LOR)
k = plot_binary_heatmap(subfamily_state_sample_filter,"H3K27ac","H3K27ac",1,sample_counts["All","H3K27ac"],THRESHOLD_LOR)

legend_class = get_legend(ggplot(rmsk_TE_subfamily,aes(x=subfamily,y=class_update,fill=class_update)) + geom_tile() + scale_fill_manual(values=class_colors))

# Arrange column legends
grid.arrange(legend_group,legend_anatomy,legend_age,legend_cancer,legend_germline,legend_type,ncol=2,widths=c(0.6,0.4),layout_matrix=cbind(c(1,2,6),c(3,4,5)))
```

### Subfamily factors

```{r subfamily correlation}
source("R_scripts/TE_correlation_subfamily.R") # Requires subfamily_enrichment.R
source("R_scripts/subfamily_correlate.R") # Requires TE_correlation_subfamily.R

subfamily_measure_long = melt(rmsk_TE_subfamily_measure[,c("class_update","family","subfamily",cohorts,standard_chromosomes[1:24],measure_metrics_subfam,states[1:21])],id.vars=c("class_update","family","subfamily",cohorts,standard_chromosomes[1:24],measure_metrics_subfam))
subfamily_measure_long = melt(subfamily_measure_long,id.vars=c("class_update","family","subfamily","variable","value"))
colnames(subfamily_measure_long)[4:7] = c("State","Samples","Metric","Value")
```

### MER57E3

```{r mer57e3}
subfamily_state_sample_counts[which(subfamily_state_sample_counts$subfamily == "MER57E3"),]
MER57E3 = rmsk_TE[which(rmsk_TE$subfamily == "MER57E3"),]
dim(MER57E3)[1]
table(MER57E3$chromosome)
apply(MER57E3[,cohorts],2,function(x) dim(MER57E3[which(!is.na(x)),])[1])
```

### Young subfamilies

```{r young subfamilies}
young_subfamilies = as.vector(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Age < quantile(rmsk_TE_subfamily$Age,0.05)),]$subfamily)
rmsk_TE_subfamily[which(rmsk_TE_subfamily$subfamily %in% young_subfamilies),c("subfamily","family","class_update","Age")]

young_subfamilies_enrichment = dcast(subfamily_state_sample_counts[which(subfamily_state_sample_counts$subfamily %in% young_subfamilies & subfamily_state_sample_counts$Sample.Proportion > 0),],subfamily~State,value.var="Sample.Proportion")
rownames(young_subfamilies_enrichment) = young_subfamilies_enrichment$subfamily
young_subfamilies_enrichment = young_subfamilies_enrichment[,2:16]

sort(apply(young_subfamilies_enrichment,2,function(x) sum(!is.na(x))))
sort(apply(young_subfamilies_enrichment,1,function(x) sum(!is.na(x))))

young_subfamilies_enrichment
```

### VISTA enhancers

```{r vista enhancers}
dim(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),])
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),c(TE_coordinates,"Vista_enhancers")]$subfamily)))
sort(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$Vista_enhancers)),c(TE_coordinates,"Vista_enhancers")]$subfamily)))

correlate_subfamily[which(correlate_subfamily$Metric == "Vista_enhancers" & correlate_subfamily$State %in% states & abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$p.value < 0.05),] # Not MHC corrected
```

### Supplementary Figure 16

Metric v. metric - pretty similar across class, SINE stands out

Chromosome % vs. chromosome % - no obvious pattern overall

State v. state - by class - similar to overall

```{r Figure S16, echo=FALSE,eval=FALSE}
# State vs. metric
a = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% measure_metrics_subfam & correlate_subfamily$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(states[c(1:14,16:17,19:21)])) + scale_x_discrete(limits=measure_metrics_subfam)

b = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% measure_metrics_subfam & correlate_subfamily$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(states[c(1:14,16:17,19:21)])) + scale_x_discrete(limits=measure_metrics_subfam) + facet_wrap(~class_update)

# State vs. feature
c = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfamily$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(states[c(1:14,16:17,19:21)])) + scale_x_discrete(limits=c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers"),labels=c(genic_labels,"Vista enhancers"))

d = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfamily$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(states[c(1:14,16:17,19:21)])) + scale_x_discrete(limits=c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers"),labels=c(genic_labels,"Vista enhancers")) + facet_wrap(~class_update)

# State vs. chromosome
e = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% standard_chromosomes & correlate_subfamily$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(states[c(1:14,16:17,19:21)]))

f = ggplot(correlate_subfamily[which(correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% standard_chromosomes & correlate_subfamily$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_y_discrete(limits=rev(states[c(1:14,16:17,19:21)])) + facet_wrap(~class_update)

# Metric vs. metric
g = ggplot(correlate_subfam_feature[which(correlate_subfam_feature$State %in% measure_metrics_subfam & correlate_subfam_feature$Metric %in% measure_metrics_subfam & correlate_subfam_feature$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_x_discrete(limits=measure_metrics_subfam) + scale_y_discrete(limits=rev(measure_metrics_subfam))

# State vs. state
h = ggplot(correlate_subfam_state[which(correlate_subfam_state$State %in% states & correlate_subfam_state$State2 %in% states & correlate_subfam_state$class_update == "All"),],aes(x=State,y=State2,fill=estimate.rho)) + geom_tile() + facet_wrap(~class_update) + scale_fill_gradient2(name="Rho",limits=c(-1,1),high="darkred",low="darkblue",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_x_discrete(limits=states[1:21]) + scale_y_discrete(limits=states[1:21])

grid.arrange(a,b,c,d,e,f,g,h,nrow=4,layout_matrix=cbind(c(1,3,5,7),c(2,4,6,8)),widths=c(0.45,0.55))
```

### Subfamily enrichment correlation

```{r analysis Fig S15a,eval=FALSE}
## Not MHC corrected

# State vs. metric
correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% measure_metrics_subfam & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.05),]

correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% measure_metrics_subfam & correlate_subfamily$class_update != "All"  & correlate_subfamily$p.value < 0.05),]

# State vs. feature
correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.05),]

correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% c(features[1:3],"coding_exon",features[5:8],"Vista_enhancers") & correlate_subfamily$class_update != "All" & correlate_subfamily$p.value < 0.05),]

# State vs. chromosome
correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% standard_chromosomes & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.05),]

correlate_subfamily[which(abs(correlate_subfamily$estimate.rho) > 0.3 & correlate_subfamily$State %in% states & correlate_subfamily$Metric %in% standard_chromosomes & correlate_subfamily$class_update != "All" & correlate_subfamily$p.value < 0.05),]

# Metric vs. metric
correlate_subfam_feature[which(abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% measure_metrics_subfam & correlate_subfam_feature$Metric %in% measure_metrics_subfam & correlate_subfam_feature$class_update == "All" & correlate_subfam_feature$p.value < 0.05),]

correlate_subfam_feature[which(abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% measure_metrics_subfam & correlate_subfam_feature$Metric %in% measure_metrics_subfam & correlate_subfam_feature$class_update != "All" & correlate_subfam_feature$p.value < 0.05),]

# Chromosome vs. chromosome
correlate_subfam_feature[which(abs(correlate_subfam_feature$estimate.rho) > 0.3 & correlate_subfam_feature$State %in% standard_chromosomes & correlate_subfam_feature$Metric %in% standard_chromosomes & correlate_subfam_feature$class_update == "All" & correlate_subfam_feature$State != correlate_subfam_feature$Metric & correlate_subfam_feature$p.value < 0.05),]

# State v state
correlate_subfam_state[which(abs(correlate_subfam_state$estimate.rho) > 0.3 & correlate_subfam_state$State %in% states & correlate_subfam_state$State2 %in% states & correlate_subfam_state$class_update == "All" & correlate_subfam_state$State != correlate_subfam_state$State2 & correlate_subfam_state$p.value < 0.05),]
```

### Summary of subfamily metrics

```{r subfamily metrics,eval=FALSE}
# Number of samples by class
sort(p.adjust(apply(rmsk_TE_subfamily_measure[,all_pc_states],2,function(x) as.numeric(unlist(kruskal.test(x~rmsk_TE_subfamily_measure$class_update))["p.value"])),method="bonf"))
aggregate(data=rmsk_TE_subfamily_measure[,c("class_update",all_pc_states)],.~class_update,function(x) mean(na.omit(x)),na.action = na.pass)

sort(apply(aggregate(data=rmsk_TE_subfamily_measure[,c("class_update",all_pc_states)],.~class_update,function(x) mean(na.omit(x)),na.action = na.pass)[,2:43],2,function(x) sd(x)/mean(x)))

# Subfamily metrics, overall and by class, median and mean, with p-values
apply(rmsk_TE_subfamily[,measure_metrics_subfam],2,median)
aggregate(data=rmsk_TE_subfamily[,c("class_update",measure_metrics_subfam)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)

colMeans(rmsk_TE_subfamily[,measure_metrics_subfam])
aggregate(data=rmsk_TE_subfamily[,c("class_update",measure_metrics_subfam)],.~class_update,function(x) mean(na.omit(x)),na.action = na.pass)
sort(apply(aggregate(data=rmsk_TE_subfamily[,c("class_update",measure_metrics_subfam)],.~class_update,function(x) mean(na.omit(x)),na.action=na.pass)[,2:10],2,function(x) sd(x)/mean(x)))

p.adjust(apply(rmsk_TE_subfamily[,measure_metrics_subfam],2,function(x) unlist(kruskal.test(x~rmsk_TE_subfamily$class_update))["p.value"]),method="bonf")

# Subfamily metric outliers
lapply(measure_metrics_subfam,function(x) rmsk_TE_subfamily[which(rmsk_TE_subfamily[,x] > mean(rmsk_TE_subfamily[,x])+2*sd(rmsk_TE_subfamily[,x])),c(1:4,x)])
lapply(measure_metrics_subfam,function(x) rmsk_TE_subfamily[which(rmsk_TE_subfamily[,x] < mean(rmsk_TE_subfamily[,x])-2*sd(rmsk_TE_subfamily[,x])),c(1:4,x)])

# Subfamilies with very low mappability
rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),c("subfamily","family","class_update","Count","Mappability")]
dim(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),])[1]
table(droplevels(rmsk_TE_subfamily[which(rmsk_TE_subfamily$Mappability < 0.25),]$family))
#ggplot(rmsk_TE_subfamily,aes(x=Age,y=Mappability)) + geom_point(size=0.5) + geom_text(aes(label=subfamily),size=3)

# Age of SINE subfamilies
test = rmsk_TE_subfamily[which(rmsk_TE_subfamily$class_update == "SINE"),]
test[order(test$Age),c("subfamily","family","class_update","Count","Age")]

# Subfamily feature overlap by class
apply(rmsk_TE_subfamily[,cohorts],2,median)
aggregate(data=rmsk_TE_subfamily[,c("class_update",cohorts)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)

colMeans(rmsk_TE_subfamily[,cohorts])
aggregate(data=rmsk_TE_subfamily_measure[,c("class_update",cohorts)],.~class_update,function(x) mean(na.omit(x)),na.action = na.pass)
sort(apply(aggregate(data=rmsk_TE_subfamily_measure[,c("class_update",cohorts)],.~class_update,function(x) mean(na.omit(x)),na.action = na.pass)[,2:22],2,function(x) sd(x)/mean(x)))

sort(p.adjust(apply(rmsk_TE_subfamily_measure[,cohorts],2,function(x) as.numeric(unlist(kruskal.test(x~rmsk_TE_subfamily_measure$class_update))["p.value"])),method="bonf"))

# Subfamily feature overlap outliers
#ggplot(melt(rmsk_TE_subfamily[,cohorts]),aes(x=value)) + geom_histogram() + facet_wrap(~variable)
lapply(match(cohorts,colnames(rmsk_TE_subfamily)),function(x) rmsk_TE_subfamily[which(rmsk_TE_subfamily[[x]] > 2*sd(rmsk_TE_subfamily[[x]])+mean(rmsk_TE_subfamily[[x]])),c(1:4,x)])

# Subfamilies by chromosome...

# CGate subfamilies
length(cgate_subfams)
```

### Analysis

```{r analysis Fig S15b,eval=FALSE}
# Proportion of all TEs overlapping CpG islands by family/subfamily
sort(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),]$family)))/dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),])[1]
sort(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),]$subfamily)))/dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),])[1]

# Age vs. number of members
correlate_subfam_feature[which(correlate_subfam_feature$Metric == "Count" & correlate_subfam_feature$State == "Age"),]

# Size vs. # samples enriched (not MHC corrected)
correlate_subfamily[which(correlate_subfamily$Metric == "Total_length" & correlate_subfamily$class_update == "All" & correlate_subfamily$p.value < 0.01 & correlate_subfamily$State %in% states[1:21]),]
```

## Factors influencing potential

### Figure 6

```{r Figure 6, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 1","Figure 2","Figure 5")}
source("R_scripts/TE_correlation.R")
source("R_scripts/state_metric_correlation.R")

# SINE/Alu age vs. TE metrics and number of samples in methylation states
rmsk_TE_measure$cpgIsland_binary = as.numeric(ifelse(rmsk_TE_measure$cpgIsland == 0,0,1))

rmsk_TE_measure_SINE = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$class_update == "SINE"),c(TE_coordinates,"class_update",measure_metrics,meth_states,"cpgIsland_binary")])
rmsk_TE_measure_SINE$Length = rmsk_TE_measure_SINE$Length/1000

a = ggplot(rmsk_TE_measure_SINE,aes(x=JC_distance,y=..scaled..)) + geom_density(fill=class_colors["SINE"]) + theme(axis.title.x = element_blank(),axis.title.y = element_blank())

b = ggplot(melt(rmsk_TE_measure_SINE,id.vars=c(TE_coordinates,"class_update","Length","mappability","JC_distance","CpGs","CpGs_per_length","cpgIsland_binary")),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=meth_colors, guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank()) 

c = ggplot(melt(rmsk_TE_measure_SINE,id.vars=c(TE_coordinates,"class_update","JC_distance","CpGs",meth_states,"cpgIsland_binary"))) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + geom_smooth(data=rmsk_TE_measure_SINE,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + theme(axis.title.y = element_blank())

rmsk_TE_measure_Alu = na.omit(rmsk_TE_measure[which(rmsk_TE_measure$family == "Alu"),c(TE_coordinates,"class_update",measure_metrics,meth_states,"cpgIsland_binary")])
rmsk_TE_measure_Alu$Length = rmsk_TE_measure_Alu$Length/1000

d = ggplot(rmsk_TE_measure_Alu,aes(x=JC_distance,y=..scaled..)) + geom_density(fill="lightblue") + theme(axis.title.x = element_blank(),axis.title.y = element_blank())

e = ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","Length","mappability","JC_distance","CpGs","CpGs_per_length","cpgIsland_binary")),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=meth_colors, guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank()) 

f = ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","JC_distance","CpGs",meth_states,"cpgIsland_binary"))) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),guide=FALSE) + xlab("Jukes-Cantor Evolutionary Distance") + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")) + theme(axis.title.y = element_blank()) 

legend1 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","Length","mappability","JC_distance","CpGs","CpGs_per_length","cpgIsland_binary")),aes(x=JC_distance,y=value,color=variable)) + geom_smooth() + scale_color_manual(values=meth_colors) + theme(legend.title = element_blank(),legend.direction = "horizontal"))
  
legend2 = get_legend(ggplot(melt(rmsk_TE_measure_Alu,id.vars=c(TE_coordinates,"class_update","JC_distance","CpGs",meth_states,"cpgIsland_binary"))) + geom_smooth(aes(x=JC_distance,y=value,linetype=variable),color="black") + scale_linetype_manual(values=c(2,3,6),labels=c("Length (kbp)","Mappability","CpGs/bp")) + ylab("Property") + theme(legend.title = element_blank(),legend.direction = "horizontal") + geom_smooth(data=rmsk_TE_measure_Alu,aes(x=JC_distance,y=cpgIsland_binary),color="pink",method="glm",method.args = list(family = "binomial")))

grid.arrange(a,b,c,d,e,f,legend1,legend2,nrow=5,layout_matrix=rbind(c(1,4),c(2,5),c(3,6),c(7,7),c(8,8)),heights=c(0.2,0.3,0.4,0.05,0.05))
```

### Figure 6 analysis

```{r analysis Fig 6}
# Significant correlations between number of samples in state and metric (not MHC corrected)
correlate_TE_state[which(abs(correlate_TE_state$estimate.rho) > 0.3 & correlate_TE_state$State %in% states),]

# % of SINE that is Alu
dim(rmsk_TE[which(rmsk_TE$family == "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SINE"),])[1]

# Average age for older SINE, Alu, SVA
median(rmsk_TE[which(rmsk_TE$class == "SINE"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$class == "SINE"),]$JC_distance)
median(rmsk_TE[which(rmsk_TE$class == "SINE" & rmsk_TE$family != "Alu"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$class == "SINE" & rmsk_TE$family != "Alu"),]$JC_distance)
median(rmsk_TE[which(rmsk_TE$family == "Alu"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$family == "Alu"),]$JC_distance)
median(rmsk_TE[which(rmsk_TE$class_update == "SVA"),]$JC_distance)
IQR(rmsk_TE[which(rmsk_TE$class_update == "SVA"),]$JC_distance)

# Proportion of SINE, Alu, SVA overlapping CpG islands
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland)),])[1]/NUM_TE
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SINE"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SINE"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SINE"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SINE" & rmsk_TE$family != "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SINE" & rmsk_TE$family != "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SINE" & rmsk_TE$family != "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$family == "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$family == "Alu"),])[1]/dim(rmsk_TE[which(rmsk_TE$family == "Alu"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SVA"),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$cpgIsland) & rmsk_TE$class_update == "SVA"),])[1]/dim(rmsk_TE[which(rmsk_TE$class_update == "SVA"),])[1]

# Overlap with CpG islands by age (logistic regression) (all, SINE, Alu)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure,family="binomial")
summary(test)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure_SINE,family="binomial")
summary(test)
test = glm(as.factor(cpgIsland_binary)~JC_distance,data=rmsk_TE_measure_Alu,family="binomial")
summary(test)

# Age vs. overlap with CpG islands
ddply(rmsk_TE_measure,.(cpgIsland_binary),summarise,Age=median(JC_distance))
ddply(rmsk_TE_measure[which(rmsk_TE_measure$class_update == "SINE"),],.(cpgIsland_binary),summarise,Age=median(JC_distance))
ddply(rmsk_TE_measure[which(rmsk_TE_measure$family == "Alu"),],.(cpgIsland_binary),summarise,Age=median(JC_distance))
```

### Supplementary Figure 12

```{r Figure S12, echo=FALSE}
a = ggplot(rmsk_TE_measure,aes(x=Length/1000,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + xlab("Length (kbp)") + ylab("Density") + facet_wrap(~class_update,nrow=1,scales="free_y")

b = ggplot(rmsk_TE_measure,aes(x=CpGs,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + xlab("CpGs") + ylab("Density") + facet_wrap(~class_update,nrow=1,scales="free_y")

c = ggplot(rmsk_TE_measure,aes(x=CpGs_per_length,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + xlab("CpGs/bp") + ylab("Density") + facet_wrap(~class_update,nrow=1,scales="free_y")

d = ggplot(rmsk_TE_measure,aes(x=mappability,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) + xlab("Mappability") + ylab("Density") + xlim(0,1) + facet_wrap(~class_update,nrow=1,scales="free_y")

e = ggplot(rmsk_TE_measure,aes(x=JC_distance,fill=class_update)) + geom_density() + scale_fill_manual(values=class_colors,guide=FALSE) + theme(axis.text.y = element_blank(),axis.ticks.y = element_blank()) +  xlab("Jukes-Cantor evolutionary distance") + ylab("Density") + facet_wrap(~class_update,nrow=1,scales="free_y")

f = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update == "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs"))

g = ggplot(correlate_TE_feature[which(correlate_TE_feature$class_update != "All"),],aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(labels=c("Mappability","Age","CpGs","CpGs/bp")) + scale_x_discrete(labels=c("Length","Mappability","Age","CpGs")) + facet_wrap(~class_update)

grid.arrange(a,b,c,d,e,f,g,ncol=2,layout_matrix=rbind(c(1,1),c(2,2),c(3,3),c(4,4),c(5,5),c(6,7)),heights=c(0.15,0.15,0.15,0.15,0.15,0.25))
```

### TE metric stats
Missing: All TEs, all classes except SVA; SVA - Length, mappability, CpGs_per_length

```{r metrics analysis}
# Median TE metrics by overall and by class, with p-values
p.adjust(apply(rmsk_TE_measure[,measure_metrics],2,function(x) unlist(kruskal.test(x~rmsk_TE_measure$class_update))["p.value"]),method="bonf")

apply(rmsk_TE_measure[,measure_metrics],2,median)
aggregate(data=rmsk_TE_measure[,c("class_update",measure_metrics)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)
sort(apply(aggregate(data=rmsk_TE_measure[,c("class_update",measure_metrics)],.~class_update,function(x) median(na.omit(x)),na.action=na.pass)[,2:6],2,function(x) IQR(x)/median(x)))

# Max TE length by class
aggregate(data=rmsk_TE_measure,Length~class_update,max)

# Correlations between mappability, CpGs, CpGs per length, number of samples Missing (WGBS)
correlate_TE_state[which(correlate_TE_state$State == "Missing" & correlate_TE_state$Metric %in% c("mappability","CpGs","CpGs_per_length")),]

# Correlation between mappability, number of samples Quiescent
correlate_TE_state[which(correlate_TE_state$Metric == "mappability" & correlate_TE_state$State == "15_Quies"),]
```

### Supplementary Figure 13

Feature-state patterns roughly the same across class (not shown). Stronger for protein-coding than non-coding (not shown).

Feature-metric: Besides CpG islands, not much to write home about. Genic features have more CpGs and higher CpG density than intergenic regions, particularly promoters and 5'UTRs, and genic TEs are shorter than intergenic TEs. Stronger for protein-coding than non-coding genes. However, these changes are between -23% and 71%. Mappability and age don't really change.

Legends: a) Spearman's rho, c/d) Average % incrase

```{r Figure S13, echo=FALSE, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 6")}
#source("R_scripts/state_metric_correlation.R")
source("R_scripts/feature_overlap_state.R") #Requires potential.R
source("R_scripts/feature_overlap_metric.R") #Requires TE_correlation.R
source("R_scripts/chromosome_potential.R")

# Metric vs. state
a = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update == "All" & correlate_TE_state$State %in% meth_states),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white",guide=FALSE) + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(limits=rev(meth_states)) + scale_x_discrete(limits=measure_metrics[c(1,4:5,2:3)],labels=c("Length","CpGs","CpGs/bp","Mappability","Age"))

b = ggplot(droplevels(correlate_TE_state[which(correlate_TE_state$class_update != "All" & correlate_TE_state$State %in% meth_states),]),aes(x=Metric,y=State,fill=estimate.rho)) + geom_tile() + scale_fill_gradient2(name="Spearman's rho",limits=c(-1,1),high="darkred",low="darkblue",mid="white") + theme(axis.title.x = element_blank(),axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1),panel.background=element_rect(fill="lightgrey")) + scale_y_discrete(limits=rev(meth_states)) + scale_x_discrete(limits=measure_metrics[c(1,4:5,2:3)],labels=c("Length","CpGs","CpGs/bp","Mappability","Age")) + facet_wrap(~class_update) 

# Feature vs. state
c = ggplot(na.omit(feature_state_mean[which(feature_state_mean$Coding == "All"),]),aes(x=State,y=Feature,fill=Enrichment*100)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=45,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(name="% increase",low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-100,4000)) + scale_y_discrete(limits=rev(levels(feature_state_mean$Feature)),labels=genic_labels) + scale_x_discrete(labels=all_state_labels)

# Feature vs. metric
d = ggplot(feature_metric_mean_class[which(feature_metric_mean_class$Coding == "All"),],aes(x=Metric,y=Feature,fill=Enrichment*100)) + geom_tile() + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle=90,hjust=1),panel.background = element_rect(fill="lightgrey")) + scale_fill_gradient2(name="% increase",low="darkblue",high="darkred",mid="white",midpoint=0,limits=c(-100,1500)) + scale_y_discrete(limits=rev(levels(feature_metric_mean_class$Feature)),labels=genic_labels) + scale_x_discrete(limits=measure_metrics[c(1,4:5,2:3)],labels=c("Length","CpGs","CpGs/bp","Mappability","Age")) + facet_wrap(~Class)

# Epigenetic state by chromosome
e = ggplot(chrom_state_total,aes(x=chromosome,y=Bases,fill=State)) + geom_bar(stat="identity",position="fill") + scale_fill_manual(values = chromHMM_colors,guide=FALSE) + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + ylab("Proportion")

# Chromosome vs. states
f = ggplot(chromosome_potential,aes(x=State,y=Zscore,color=chromosome,label=chromosome)) + geom_point() + geom_text(angle=45,hjust=0,vjust=0,size=3) + scale_color_discrete(guide=FALSE) + scale_x_discrete(limits=rev(states),labels=all_state_labels) + theme(axis.title.y = element_blank(),panel.grid.major.y=element_line(color="grey")) + ylab("Z-score, % of samples in state") + geom_hline(yintercept = 1,color="grey") + geom_hline(yintercept = 2,color="grey") + geom_hline(yintercept = -1,color="grey") + geom_hline(yintercept = -2,color="grey") + geom_hline(yintercept = -3,color="grey") + geom_hline(yintercept = 3,color="grey") + geom_hline(yintercept = -4,color="grey") + geom_hline(yintercept = 4,color="grey") + coord_flip()

grid.arrange(a,b,c,d,e,f,layout_matrix=rbind(c(1,2),c(3,4),c(5,6)),widths=c(0.4,0.6))
```

### TE feature stats

```{r analysis Fig S13, cache=TRUE, cache.lazy=FALSE, dependson=c("Figure 6","Figure S13")}
# States
apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y > 0),states],na.rm=TRUE)) - apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y == 0),states],na.rm=TRUE))

dcast(feature_state_mean[which(feature_state_mean$Coding == "All"),],Feature~State,value.var = "Enrichment")
feature_state_mean_class[which(feature_state_mean_class$Coding == "All"),]

# Proportion of TEs ever in state overlapping each feature
apply(rmsk_TE_measure[,states],2,function(x) apply(rmsk_TE_measure[,cohorts],2,function(y) dim(rmsk_TE_measure[which(x > 0 & y > 0),])[1]/dim(rmsk_TE_measure[which(x > 0),])[1]))

# Metrics
apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y > 0),measure_metrics],na.rm=TRUE)) - apply(rmsk_TE_measure[,cohorts],2,function(y) colMeans(rmsk_TE_measure[which(y == 0),measure_metrics],na.rm=TRUE))

feature_metric_mean[order(feature_metric_mean$Enrichment),]
feature_metric_mean_class[order(feature_metric_mean_class$Enrichment),]
```

### Chromosome stats

```{r chromosome stats}
# Proportion of TEs relative to genome length
hg19_TEs[order(hg19_TEs$Prop_TEs/hg19_TEs$Prop),]

ggplot(hg19_TEs,aes(x=Prop,y=Prop_TEs,color=chrom,label=chrom)) + geom_point() + geom_text(nudge_y = 0.001) + geom_abline(slope=1,intercept=0) + xlim(0,0.1) + ylim(0,0.1) + scale_color_discrete(guide=FALSE) + ylab("Proportion of TEs") + xlab("Proportion of length")

ggplot(na.omit(hg19_class),aes(x=Prop,y=Prop_Class,color=chrom,label=chrom)) + geom_point() + geom_text(nudge_y = 0.001) + geom_abline(slope=1,intercept=0) + xlim(0,0.11) + ylim(0,0.11) + facet_wrap(~class_update) + ylab("Proportion of TEs") + xlab("Proportion of length") + scale_color_discrete(guide=FALSE)

# Z-scores
chromosome_potential[order(chromosome_potential$Zscore),]

# Proportion of TEs in each state on each chromosome
apply(rmsk_TE_measure[,states],2,function(x) table(rmsk_TE_measure[which(x > 0),]$chromosome)/dim(rmsk_TE_measure[which(x > 0),])[1])[standard_chromosomes[1:24],]

# Chromosome vs. feature
hg19_features[order(hg19_features$Zscore),]

ggplot(na.omit(hg19_features),aes(x=Feature,y=Prop_TEs_feature,color=chrom,label=chrom)) + geom_point() + geom_text() + theme(axis.title.y = element_blank(),axis.text.x = element_text(angle=90,hjust=1)) + scale_color_discrete(guide=FALSE) + ylab("Proportion of TEs in feature") + facet_wrap(~Coding,labeller=labeller(Coding=coding_labels)) + scale_x_discrete(limits=rev(names(genic_labels)),labels=genic_labels) + coord_flip()
```

### Potential without chrY test

```{r eval=FALSE}
# Potential without chrY
test = chromHMM_TE_state[which(chromHMM_TE_state$chromosome != "chrY"),]
test.a = sample_distribution(test,c(8:22),sample_counts["All","chromHMM"])
test.b = cumulative_distribution(test,c(8:22),sample_counts["All","chromHMM"])
test.b$State = factor(rep(chromHMM_states,each=sample_counts["All","chromHMM"]),levels=chromHMM_states)
test.c = potential_stats(test.a,15,sample_counts["All","chromHMM"])
test.c$State = factor(chromHMM_states,levels=chromHMM_states)
chromHMM_TE_state_dist_stats-test.c

# Average samples in state for chrY TEs
sort(colMeans(chromHMM_TE_state[which(chromHMM_TE_state$chromosome == "chrY"),8:22]))

# % of TEs overlapping DNase/H3K27ac on chrY, expressed RPKM > 1
ADD
```

## Mouse

### Figure 7

```{r Figure 7, echo=FALSE,eval=FALSE}
#source("R_scripts/WGBS_subfamily_enrichment_TE.R")
source("R_scripts/human_mouse_ortholog_WGBS.R") #Requires WGBS_subfamily_enrichment_TE.R

a = ggplot(human_mouse_orthologs_WGBS,aes(x=E054,y=ENCFF592SYK/100)) + geom_point(alpha=0.1) + ylab("Mouse TE methylation level") + xlab("Human TE methylation level") + geom_density2d()

b = ggplot(hg19_mm10_TE_WGBS_subfamily_hypo_paired,aes(x=Human,y=Mouse,color=Pair)) + geom_point(alpha=0.5) + scale_color_discrete(guide=FALSE) + xlab("Human subfamily % hypomethylated") + ylab("Mouse subfamily % hypomethylated")

grid.arrange(a,b,nrow=1)

# Mosaic plot
human_mouse_orthologs_WGBS_category_pairs = lapply(seq(1,13,1),function(x) table(human_mouse_orthologs_WGBS_category[,c(as.vector(human_mouse_samples[x,1]),as.vector(human_mouse_samples[x,2]))]))
human_mouse_orthologs_WGBS_category_pairs_summed = Reduce('+',human_mouse_orthologs_WGBS_category_pairs)
names(attributes(human_mouse_orthologs_WGBS_category_pairs_summed)$dimnames) <- c("Mouse","Human")

c = mosaic(human_mouse_orthologs_WGBS_category_pairs_summed[c(2,3,1),c(2,3,1)],shade=TRUE,return_grob=TRUE)
```

### Supplementary Figure 17

```{r Figure S17, echo=FALSE,eval=FALSE}
source("R_scripts/TE_ortholog_class_stats.R")
#source("R_scripts/human_mouse_ortholog_WGBS.R")

par(mfrow=c(1,2)) 

pie(rmsk_TE_class$Mouse_orthologs,labels=paste(rmsk_TE_class$Mouse_orthologs,"(",round(rmsk_TE_class$Mouse_orthologs/sum(rmsk_TE_class$Mouse_orthologs)*100,0),"%)",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="Orthologous TEs",cex=1,cex.main=2,line=-10,radius=0.5)

pie(rmsk_TE_class$Mouse_ortholog_subfamily,labels=paste(rmsk_TE_class$Mouse_ortholog_subfamily,"(",round(rmsk_TE_class$Mouse_ortholog_subfamily/sum(rmsk_TE_class$Mouse_ortholog_subfamily)*100,0),"%)",sep=""),col=class_colors[as.vector(rmsk_TE_class$class_update)],clockwise=TRUE,main="Orthologous subfamilies",cex=1,cex.main=2,line=-10,radius=0.5)

b = ggplot(human_mouse_samples,aes(x=Test,y=Spearman)) + geom_boxplot() + ylab("Spearman correlation across TEs") + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_x_discrete(labels=c("Human-human","Mouse-mouse","Human-mouse, same tissue","Human-mouse, random tissue")) + scale_y_continuous(limits=c(0,1))

c = ggplot(human_mouse_samples,aes(x=Test,y=Subfamily_Spearman)) + geom_boxplot() + ylab("Spearman correlation across subfamilies") + theme(axis.title.x = element_blank(),axis.text.x = element_text(angle=45,hjust=1)) + scale_x_discrete(labels=c("Human-human, random tissue","Mouse-mouse, random tissue","Human-mouse, same tissue","Human-mouse, random tissue")) + scale_y_continuous(limits=c(0,1))

hg19_mm10_TE_WGBS_subfamily_hypo_paired_age = merge(ddply(hg19_mm10_TE_WGBS_subfamily_hypo_paired,.(subfamily),summarise,unlist(cor.test(Mouse,Human,method="spearman"))["estimate.rho"]),rmsk_TE_subfamily[c("subfamily","Age")],by="subfamily")
colnames(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age)[2] = "Spearman_correlation"
hg19_mm10_TE_WGBS_subfamily_hypo_paired_age$Spearman_correlation = as.numeric(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age$Spearman_correlation)

d = ggplot(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age,aes(x=Age,y=Spearman_correlation)) + geom_point() + ylab("Spearman's rho")

grid.arrange(b,c,d,nrow=2,layout_matrix=rbind(c(1,2),c(3,3)))
```

### Mouse ortholog analysis

```{r analysis Fig S16a,eval=FALSE}
# Number of mouse TEs
MOUSE_TE = dim(mm10_rmsk_TE)[1]
MOUSE_TE

# Number of hg19 TEs that liftover
# (wc -l Mouse/liftover/rmsk_TE_hg19tomm10.bed)
1265775
1265775/NUM_TE

# Number of unique orthologous TEs in hg19->mm10, hg19, and mm10
dim(unique(human_mouse_orthologs_mm10[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")]))[1]
dim(unique(human_mouse_orthologs_mm10[,hg19_coordinates]))[1]
dim(unique(human_mouse_orthologs_mm10[,hg19_coordinates]))[1]/NUM_TE
dim(unique(human_mouse_orthologs_mm10[,mm10_coordinates]))[1]

# Number, % of human TEs with 2+ human mm10 TEs
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_hg19))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","human_TE_hg19")])$human_TE_mm10))[2:11])/length(unique(human_mouse_orthologs_mm10$human_TE_mm10))

# Number, % of human TEs with 2+ mouse TEs
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$human_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$human_TE_mm10))[2:10])/length(unique(human_mouse_orthologs_mm10$human_TE_mm10))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$human_TE_hg19))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$human_TE_hg19))[2:10])/length(unique(human_mouse_orthologs_mm10$human_TE_hg19))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$mouse_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_mm10","mouse_TE_mm10")])$mouse_TE_mm10))[2:15])/length(unique(human_mouse_orthologs_mm10$mouse_TE_mm10))
table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$mouse_TE_mm10))
sum(table(table(unique(human_mouse_orthologs_mm10[,c("human_TE_hg19","mouse_TE_mm10")])$mouse_TE_mm10))[2:25])/length(unique(human_mouse_orthologs_mm10$mouse_TE_mm10))

# Average length of those TEs
median(unique(human_mouse_orthologs_mm10[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")])$human_stop_mm10-
         unique(human_mouse_orthologs_mm10[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")])$human_start_mm10)
median(unique(human_mouse_orthologs_mm10[,hg19_coordinates])$human_stop_hg19-unique(human_mouse_orthologs_mm10[,hg19_coordinates])$human_start_hg19)
median(unique(human_mouse_orthologs_mm10[,mm10_coordinates])$mouse_stop_mm10-unique(human_mouse_orthologs_mm10[,mm10_coordinates])$mouse_start_mm10)

# Average age of TEs with mouse orthologs
median(merge(rmsk_TE_measure,unique(human_mouse_orthologs_mm10[,hg19_coordinates]),by.x=TE_coordinates,by.y=hg19_coordinates[c(1:4,6,5,7)])$JC_distance)

# Number of subfamilies shared between human and mouse
length(intersect(levels(mm10_rmsk_TE$subfamily),levels(rmsk_TE$subfamily)))

# Average age of subfamilies also in mouse 
median(rmsk_TE_subfamily[which(rmsk_TE_subfamily$subfamily %in% mm10_rmsk_TE$subfamily),]$Age)

# Proportion of hg19 TEs and subfamilies with ortholog in mouse, by class
rmsk_TE_class$Mouse_orthologs/rmsk_TE_class$Count
rmsk_TE_class$Mouse_ortholog_subfamily/rmsk_TE_class$Subfamilies
```

### Mouse WGBS analysis

```{r analysis Fig S16b,eval=FALSE}
# Number, % of mm10 TEs with CpGs
dim(mm10_rmsk_TE_CpG_count)[1]
dim(mm10_rmsk_TE_CpG_count)[1]/MOUSE_TE

# Number, % of hg19 orthologs with CpGs
dim(merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,hg19_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=hg19_coordinates))
dim(merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,hg19_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=hg19_coordinates))[1]/dim(unique(human_mouse_orthologs_mm10[,hg19_coordinates]))[1]

# Average number of CpGs per hg19 ortholog with CpGs
mean(merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,hg19_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=hg19_coordinates)$CpGs)
median(merge(TE_CpG_count,unique(human_mouse_orthologs_mm10[,hg19_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=hg19_coordinates)$CpGs)

# Number, % of mm10 orthologs with CpGs
dim(merge(mm10_rmsk_TE_CpG_count,unique(human_mouse_orthologs_mm10[,mm10_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=mm10_coordinates))
dim(merge(mm10_rmsk_TE_CpG_count,unique(human_mouse_orthologs_mm10[,mm10_coordinates]),by.x=TE_coordinates[c(1:4,6,5,7)],by.y=mm10_coordinates))[1]/dim(unique(human_mouse_orthologs_mm10[,mm10_coordinates]))[1]

# Number of unique orthologous TEs with CpGs in mm10, hg19, and hg19->mm10
dim(unique(human_mouse_orthologs_WGBS[,mm10_coordinates]))[1]
dim(unique(human_mouse_orthologs_WGBS[,hg19_coordinates]))[1]
dim(unique(human_mouse_orthologs_WGBS[,c("human_chr_mm10","human_start_mm10","human_stop_mm10","human_strand_mm10")]))[1]

# TEs with meth data in at least one mouse/human sample
test = human_mouse_orthologs_WGBS[which(apply(human_mouse_orthologs_WGBS,1,function(x) sum(is.na(x[20:32])) < 13) & apply(human_mouse_orthologs_WGBS,1,function(x) sum(is.na(x[33:41])) < 9)),]

# Number of human (hg19) TEs
dim(unique(human_mouse_orthologs_WGBS[,hg19_coordinates]))[1]-dim(unique(test[,hg19_coordinates]))[1]

# Number of mouse TEs
dim(unique(human_mouse_orthologs_WGBS[,mm10_coordinates]))[1]-dim(unique(test[,mm10_coordinates]))[1]

# Number of TEs with methylation data in E054 and ENCFF592SYK
dim(human_mouse_orthologs_WGBS[which(!is.na(human_mouse_orthologs_WGBS$E054) & !is.na(human_mouse_orthologs_WGBS$ENCFF592SYK)),])

# Individual TE Spearman correlations
ddply(human_mouse_samples,.(Test),summarise,Mean=mean(Spearman),SD=sd(Spearman))

# Number of pairs where TE is hypomethylated in both
dim(human_mouse_orthologs_mm10_hypo)
dim(unique(human_mouse_orthologs_mm10_hypo[,hg19_coordinates]))
dim(unique(human_mouse_orthologs_mm10_hypo[,mm10_coordinates]))
dim(unique(human_mouse_orthologs_mm10_hypo[,c(hg19_coordinates,"Human_sample")]))
dim(unique(human_mouse_orthologs_mm10_hypo[,c(mm10_coordinates,"Mouse_sample_mm10")]))

# Number of subfamilies shared between human and mouse with CpGs in both 
dim(hg19_mm10_TE_WGBS_subfamily_hypo)[1]
setdiff(intersect(levels(mm10_rmsk_TE$subfamily),levels(rmsk_TE$subfamily)),hg19_mm10_TE_WGBS_subfamily_hypo$subfamily)

# Subfamily Spearman correlations 
ddply(human_mouse_samples,.(Test),summarise,Mean=mean(Subfamily_Spearman),SD=sd(Subfamily_Spearman))

# Subfamily age vs. degree of correlation with mouse
cor.test(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age$Age,as.numeric(hg19_mm10_TE_WGBS_subfamily_hypo_paired_age$Spearman_correlation),method="spearman")
```

### Chi-sq test for human-mouse pairs (WGBS)

```{r analysis Fig S16c,eval=FALSE}
# All methylation states
p.adjust(unlist(lapply(human_mouse_orthologs_WGBS_category_pairs,function(x) unlist(chisq.test(x))["p.value"])),method="bonf")

# Hypomethylated vs. all other states
p.adjust(apply(human_mouse_samples,1,function(x) unlist(chisq.test(matrix(c(nrow(human_mouse_orthologs_WGBS_category[which(human_mouse_orthologs_WGBS_category[,x[1]] == "Hypomethylated" & human_mouse_orthologs_WGBS_category[,x[2]] == "Hypomethylated"),]),nrow(human_mouse_orthologs_WGBS_category[which(human_mouse_orthologs_WGBS_category[,x[1]] == "Hypomethylated" & human_mouse_orthologs_WGBS_category[,x[2]] %in% c("Hypermethylated","Intermediate")),]),nrow(human_mouse_orthologs_WGBS_category[which(human_mouse_orthologs_WGBS_category[,x[1]] %in% c("Hypermethylated","Intermediate") & human_mouse_orthologs_WGBS_category[,x[2]] == "Hypomethylated"),]),nrow(human_mouse_orthologs_WGBS_category[which(human_mouse_orthologs_WGBS_category[,x[1]] %in% c("Hypermethylated","Intermediate") & human_mouse_orthologs_WGBS_category[,x[2]] %in% c("Hypermethylated","Intermediate")),])),nrow=2)))["p.value"]),method="bonf")
```

### Comparison of human-mouse TE ortholog methylation and chromHMM state
### RERUN when updated

```{r write chromHMM, eval=FALSE}
write.table(unique(human_mouse_orthologs_mm10_hypo[,c(hg19_coordinates,"Human_sample")]),file="Mouse/hg19_ortholog_hypo.txt",sep="\t",row.names=FALSE,col.names=FALSE,quote=FALSE)
write.table(unique(human_mouse_orthologs_mm10_hypo[,mm10_coordinates]),file="Mouse/mm10_ortholog_hypo.txt",sep="\t",col.names=FALSE,row.names=FALSE,quote=FALSE)
```

### Relies on intervening unix analysis

```{r,eval=FALSE}
# Human chromHMM state for hg19 TEs hypomethylated in mouse/human
hg19_orthologs_hypo_chromHMM = read.table("Mouse/hg19_ortholog_hypo_chromHMM.txt",sep='\t')[,1:9]
colnames(hg19_orthologs_hypo_chromHMM) = c(hg19_coordinates,"Human_sample","Human_state")

# mm9 chromHMM state for mm10 TEs hypomethylated in mouse/human
mm10_orthologs_hypo_chromHMM = read.table("Mouse/mm10_ortholog_hypo_chromHMM.txt",sep='\t')[,1:9]
colnames(mm10_orthologs_hypo_chromHMM) = c(mm10_coordinates,"Mouse_State","Mouse_sample_mm9")

# Combine chromHMM and WGBS
human_mouse_orthologs_mm10_hypo_chromHMM = merge(human_mouse_orthologs_mm10_hypo,hg19_orthologs_hypo_chromHMM,by=c(hg19_coordinates,"Human_sample"),all.x=TRUE)
human_mouse_orthologs_mm10_hypo_chromHMM = merge(human_mouse_orthologs_mm10_hypo_chromHMM,mm10_orthologs_hypo_chromHMM,by=c(mm10_coordinates,"Mouse_sample_mm9"),all.x=TRUE)
human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State = factor(human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State)
human_mouse_orthologs_mm10_hypo_chromHMM$Human_state = factor(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state,chromHMM_states)
```

### Supplementary Figure 18

```{r Figure S18, echo=FALSE,eval=FALSE}
# Mouse-human chromHMM states
human_mouse_chromHMM_states = matrix(c(chromHMM_states[c(1:2,7,6,4,15,13)],"1","2","3","4","5","6","7"),nrow=7)

mosaic(table(human_mouse_orthologs_mm10_hypo_chromHMM[,c("Human_state","Mouse_State")]),shade=TRUE,return_grob=TRUE)
```

### Human-mouse ortholog chromHMM analysis

```{r analysis Fig S17,eval=FALSE}
# Corresponding samples
chromHMM_mouse_samples = unique(human_mouse_orthologs_mm10_hypo_chromHMM[,c("Human_sample","Mouse_sample_mm9")])
table(human_mouse_orthologs_mm10_hypo_chromHMM$Human_sample,human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_sample_mm9)

# Number of pairs and unique TEs in each state
dim(human_mouse_orthologs_mm10_hypo_chromHMM)
dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[,hg19_coordinates]))
dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[,c(hg19_coordinates,"Human_sample")]))
dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[,mm10_coordinates]))
dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[,c(mm10_coordinates,"Mouse_sample_mm9")]))

apply(human_mouse_chromHMM_states,1,function(x) { c(
  pairs = dim(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),])[1],
  human = dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),hg19_coordinates]))[1],
  mouse = dim(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),mm10_coordinates]))[1]
)})

# Chi-squared for each sample pair (rows and columns with no data removed)
p.adjust(unlist(lapply(lapply(seq(1,11,1),function(x) table(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_sample == chromHMM_mouse_samples[x,1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_sample_mm9 == chromHMM_mouse_samples[x,2]),c("Human_state","Mouse_State")])),function(x) unlist(chisq.test(x[which(rowSums(x) > 0),which(colSums(x) > 0)]))["p.value"])),method="bonf")

# Summarise hg19 TE class by state
apply(human_mouse_chromHMM_states,1,function(x) table(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),hg19_coordinates])$human_class))

# Summarise hg19 TE subfamily by state
apply(human_mouse_chromHMM_states,1,function(x) sort(table(droplevels(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),hg19_coordinates])$human_subfamily))))

test = apply(human_mouse_chromHMM_states,1,function(x) log2((table(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),hg19_coordinates])$human_subfamily)/sum(table(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == x[1] & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == x[2]),hg19_coordinates])$human_subfamily)))/(table(unique(human_mouse_orthologs_mm10[,hg19_coordinates])$human_subfamily)/sum(table(unique(human_mouse_orthologs_mm10[,hg19_coordinates])$human_subfamily)))))

test[order(test[,1]),]
test[order(test[,3]),]
test[order(test[,4]),]
```

```{r overlap with GENCODE,eval=FALSE}
write.table(unique(human_mouse_orthologs_mm10_hypo_chromHMM[which(human_mouse_orthologs_mm10_hypo_chromHMM$Human_state == "1_TssA" & human_mouse_orthologs_mm10_hypo_chromHMM$Mouse_State == "1"),hg19_coordinates]),row.names=FALSE,col.names=FALSE,quote=FALSE,sep='\t',file="Mouse/hg19_orthologs_hypo_1TssA.bed")
```

# Methods 

## Basic statistics

```{r QC statistics,eval=FALSE}
# Proportion of TEs on chrY
NUM_TE-NUM_TE_noY
(NUM_TE-NUM_TE_noY)/NUM_TE

# Total bp in TEs
MERGED_TE_WIDTH

# Total TE bases (redundant)
sum(rmsk_TE$Length)-MERGED_TE_WIDTH
sum(rmsk_TE$Length)/MERGED_TE_WIDTH

# CpGs in TEs, genome
TE_CPGS
ALL_CPGS
```

## Blacklist TEs

```{r blacklist,eval=FALSE}
# Number of TEs overlapping blacklist
dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1]
dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1]/NUM_TE

# Number of subfamilies overlapping blacklist
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$subfamily)))

# Which chromosomes?
length(table(droplevels(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$chromosome)))

# Enrichment over blacklist
sort(log2((table(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),]$subfamily)/dim(rmsk_TE[which(!is.na(rmsk_TE$blacklist)),])[1])/(table(rmsk_TE$subfamily)/NUM_TE)))

# Blacklist overlap % outliers
rmsk_TE_subfamily[which(rmsk_TE_subfamily$blacklist > 2*sd(rmsk_TE_subfamily$blacklist)+mean(rmsk_TE_subfamily$blacklist)),c("subfamily","family","class","blacklist")]
```
